---
title: "Barcode preprocessing - TF reporter library 90 TFs - KD screen HepG2"
author: 
  - name: "Max Trauernicht"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute - van Steensel lab"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
---

---

### Aim
35,500 reporters for 86 TFs were transfected into HepG2 cells that were previously treated with ~50 different siRNAs for TFs. In this script the barcode counts from these samples will be pre-processed.

---

## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}
```


### Libraries 

```{r setup, out.width= "80%", fig.align= "center",  warning= FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(data.table)
library(plyr)
library(stringr)
library(ggpubr)
library(GGally)
library(vwr)
library(dplyr)
library(tibble)
library(plotly)
library(ggbeeswarm)
library(haven)
library(readr)
library(parallel)
library(RColorBrewer)
library(gridExtra)
library(LncFinder)
library(tidyr)
library(grr)
library(viridis)
library(DESeq2)
library(PCAtools)
library(pheatmap)
```


### Functions

```{r out.width= "80%", fig.align= "center",  warning= FALSE, message=FALSE}
# Custom functions
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}

ReadFasta<-function(file) {
   # Read the file line by line
   fasta<-readLines(file)
   # Identify header lines
   ind<-grep(">", fasta)
   # Identify the sequence lines
   s<-data.frame(ind=ind, from=ind+1, to=c((ind-1)[-1], length(fasta)))
   # Process sequence lines
   seqs<-rep(NA, length(ind))
   for(i in 1:length(ind)) {
      seqs[i]<-paste(fasta[s$from[i]:s$to[i]], collapse="")
   }
   # Create a data frame 
   DF<-data.frame(name=gsub(">", "", fasta[ind]), sequence=seqs)
   # Return the data frame as a result object from the function
   return(DF)
}

# From Fede:
# ggpairs custom functions
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {

  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y, "pairwise.complete.obs")
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)

  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }

  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())

  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]

  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }

  p <- p +
    theme(panel.background = element_rect(fill = corCol))

  return(p)
}

# Custom ggplot2 themes
theme_classic_lines <- function() {
  theme_pubr(border = T, legend = "top") +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.1),
                  strip.background = element_rect(fill = "#ced4da"))
    
}

theme_classic_lines_45 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 45) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.1),
                  strip.background = element_rect(fill = "#ced4da"))
}

theme_classic_lines_90 <- function() {
  theme_pubr(border = T,legend = "top", x.text.angle = 90) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.1),
                  strip.background = element_rect(fill = "#ced4da"))
}

theme_set(theme_classic_lines())

colors_diverse <- c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51")

ggplot_custom <- function(...) ggplot2::ggplot(...) + 
  scale_color_manual(values = colors_diverse) + 
  scale_fill_manual(values = colors_diverse)


hline <- function(y = 0, color = "black") {
  list(
    type = "line",
    x0 = 0,
    x1 = 1,
    xref = "paper",
    y0 = y,
    y1 = y,
    line = list(color = color)
  )
}
```


### Loading data

```{r data import, out.width= "80%", fig.align= "center",  warning= FALSE, message=FALSE}
# Load metadata file that contains all required information about the sequenced samples
metadata_df <- read_csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6952_KD_HepG2/mt20220809_metadata_gcf6952_KD_e7.csv")

# Load in barcode counts
bc_files <- paste(metadata_df$path, metadata_df$file, sep = "")
bc_files <- lapply(bc_files, fread, header = FALSE)
names(bc_files) <- metadata_df$id


# Import barcode annotation of both libraries
bc_annotation <- read_csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6578_gen2_stimulation-2/mt20211021_bc_annotation_combined.csv") %>%
  dplyr::select(-library)
```


### Creating count data frames

```{r cluster_compare, out.width= "80%", fig.align= "center",  warning= FALSE, message=FALSE}
# Generate one long data frame from the list of data frames
bc_df <- bind_rows(bc_files, .id = "sample_id") %>%
  dplyr::select(sample_id, "barcode" = V1, "starcode_counts" = V2)

# Library 1 has barcodes of length 12 while library 2 has barcodes of length 13
# For sequencing samples that have library 1 and 2 mixed I extracted barcodes of length >=12
# So, I will now filter out all barcodes with length >13, because those are not relevant
bc_df$nchar <- nchar(bc_df$barcode)
bc_df <- bc_df %>%
  filter(nchar <= 13)

# Add barcode annotation to the data
bc_df <- merge(bc_df, bc_annotation, all = T)

# Add experiment annotation to the data
metadata_selected <- metadata_df %>%
  dplyr::select('sample_id' = id, gcf, replicate, condition, sample, library)
bc_df <- bc_df[!is.na(bc_df$sample_id),]
bc_df <- merge(bc_df, metadata_selected, all = T, by = "sample_id")

bc_df$starcode_counts[is.na(bc_df$starcode_counts)] <- 0

# First compute reads per million to estimate the relative counts in the respective sample
bc_df <- bc_df %>%
  mutate(rpm = ave(starcode_counts, sample_id, FUN = function(x) (x+1) / sum(x) *1e6 ))

bc_df_filt <- bc_df[!is.na(bc_df$tf),]
bc_df_unmatched <- bc_df[is.na(bc_df$tf),]
```


## Correlate and cluster individual samples

```{r out.width= "80%", fig.align= "center",  warning= FALSE, message=FALSE}
# # Make dataframe to compare raw counts between conditions
# bc_df_filt <- bc_df_filt[!is.na(bc_df_filt$sample),]
# bc_df_cor <- bc_df_filt %>%
#   dplyr::select(-starcode_counts)
# 
# bc_df_cor <- bc_df_cor %>%
#   dplyr::select(sample_id, barcode, library, rpm) %>%
#   unique() %>%
#   mutate(rpm = as.integer(rpm)) %>%
#   spread(sample_id, rpm) 
# 
# bc_df_cor[is.na(bc_df_cor)] <- 0
# 
# 
# ## Here I want to take a quick look how well individual samples correlate
# # Compare pDNA data
# plot(bc_df_cor$ATF2_r1_gcf6952, bc_df_cor$ATF4_r1_gcf6952)
# plot(bc_df_cor$`pMT02-09_v3_r1_gcf6952`, bc_df_cor$`pMT02-09_v2_r1_gcf6952`)
# 
# 
# # Convert into data.frame with row.names for deseq2
# bc_df_cor4 <- bc_df_cor %>%
#   filter(library == "1+2") %>%
#   dplyr::select(matches("barcode|gcf6952")) %>%
#   dplyr::select(-pMT02_3_r1_gcf6952, -pMT02_r1_gcf6952, -pMT02_v2_r1_gcf6952, -pMT02_v3_r1_gcf6952, -`pMT02-09_v2_r1_gcf6952`, -`pMT02-09_v3_r1_gcf6952`) %>%
#   column_to_rownames("barcode") 
# 
# 
# #######################################
# ## Initialize deseq2
# 
# # Prepare metadata ready for deseq
# metadata <- metadata_df %>%
#   filter(id %in% names(bc_df_cor4)) %>%
#   dplyr::select("replicate", "condition", "gcf", "id") %>%
#   unique()
# 
# # Initialize
# rnaseq_dds <- DESeqDataSetFromMatrix(countData = bc_df_cor4,
#                                      colData = metadata,
#                                      design= ~ condition)
# 
# # Execute deseq2
# rnaseq_dds <- DESeq(rnaseq_dds)
# 
# 
# #######################################
# ## PCA analysis - quality control
# 
# # Get the "normalized" values and create PCA plot
# rnaseq_dds_norm <- assay(vst(rnaseq_dds))
# metadata <- metadata %>% 
#            column_to_rownames("id")
# rownames(metadata) = colnames(bc_df_cor4)
# 
# p <- pca(bc_df_cor4, metadata, removeVar = 0.1)
# 
# biplot(p, 
#        colby = "condition", 
#        shape = "gcf", 
#        legendPosition = "right",
#        legendLabSize = 8,
#        legendTitleSize = 8,
#        legendIconSize = 3, 
#        lab = "")
# 
# 
# ## Correlation heatmap
# # bc_df_cor4$mean <- rowMeans(bc_df_cor4)
# # bc_df_cor4 <- bc_df_cor4 %>%
# #   filter(mean > 200) %>%
# #   dplyr::select(-mean)
# 
# cor_df <- cor(bc_df_cor4, method = "pearson")
# 
# pheatmap(cor_df, border_color = "black")
# 
# # Get cluster information
# clusters <- hclust(dist(cor_df))
# cutree(clusters, h = 7) 
# plot(clusters)
# abline(h = 3, col = "red")
# 
# # Apply cutree() to hclust.pokemon: cut.pokemon
# cut.pokemon<-cutree(clusters,k=3)
# ##############################################################
# # Initialize total within sum of squares error: wss
# wss <- 0
# # Look over 1 to 15 possible clusters
# for (i in 1:15) {
#   # Fit the model: km.pokemon
#   km.pokemon <- kmeans(cor_df, centers = i, nstart = 20, iter.max = 50)
#   # Save the within cluster sum of squares
#   wss[i] <- km.pokemon$tot.withinss
# }
# # Produce a scree plot
# plot(1:15, wss, type = "b", 
#      xlab = "Number of Clusters", 
#      ylab = "Within groups sum of squares")

```


---

## Data quality plots

```{r read count 2, out.width= "80%", fig.align= "center",  warning= FALSE, message= FALSE}
# I want to show the following:
## 1: Read distribution of matched barcodes vs. unmatched barcode

### a: total read counts per sample
for (i in unique(bc_df$library)) {
  bc_df_i <- bc_df[bc_df$library == i,]
  p <- plot_ly(bc_df_i %>%
         mutate(sum_counts = ave(starcode_counts, sample, FUN = function(x) sum(x))) %>%
         dplyr::select(sample, sum_counts) %>%
         unique(), 
        x = ~sum_counts, y = ~sample, type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = paste("Number of reads per barcode per sample", i),
         xaxis = list(title = "Expected number of reads per barcode per sample"),
         yaxis = list(title = "Sample"))
  print(p)
}


### b: read counts per TF
bc_df_filt <- bc_df[!is.na(bc_df$tf),]

ggplot_custom(bc_df_filt %>%
         filter(neg_ctrls == "No", native_enhancer == "No", commercial_reporter == "No", hPGK == "No", rand_promoter == "No"), aes(x = rpm, y = tf)) +
  geom_violin() +
  theme(axis.text.y = element_text(size = 6)) +
  xlim(0,5000) +
  facet_wrap(~condition)

bc_df_2 <- bc_df[bc_df$rpm <= 250,]
bc_df_2 <- bc_df_2[bc_df_2$rpm >= 0.5,]
bc_df_2 <- bc_df_2[!is.na(bc_df_2$tf),]

ggplot_custom(bc_df_2, aes(x = rpm)) +
  geom_histogram(binwidth = 20) +
  xlim(0,250)+
  ylim(0,6000)+
  facet_wrap(~sample_id)

ggplot_custom(bc_df[bc_df$rpm >= 500 & !is.na(bc_df$tf),], aes(x = rpm)) +
  geom_histogram(binwidth = 20) +
  xlim(250,1000)+
  ylim(0,50)+
  facet_wrap(~sample_id)

n_highly_expressed <- data.frame("sample_id" = unique(bc_df$sample_id),
                                 "n_bc" = "", stringsAsFactors = F)
for (i in unique(bc_df$sample_id)) {
  n_highly_expressed$n_bc[n_highly_expressed$sample_id == i] <- length(bc_df$barcode[bc_df$rpm > 500 & bc_df$sample_id == i])
}

plot_ly(n_highly_expressed, x = ~sample_id, y = ~as.numeric(n_bc), type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Highly expressed barcodes",
         xaxis = list(title = "Number of barcodes with > 500 rpm"),
         yaxis = list(title = "Condition")) %>%
  layout(shapes = list(hline(100)))


## 2: How many barcodes can I find back at which cutoff? + What is the percentage of barcode reads that match the design at which cutoff?
## Identify the unmapped fraction
bc_fraction <- bc_df %>%
  dplyr::select(sample_id, library) %>%
  unique() %>%
  na.omit() 
rpm_cutoff <- data.frame("cutoff" = seq(0.0001,15,3))
bc_fraction <- expand_grid(bc_fraction, rpm_cutoff)

bc_annotation2 <- bc_annotation %>%
  mutate(nchar = nchar(bc_annotation$barcode))

bc_counts_design <- data.frame("library" = c("1", "2", "1+2"),
                               "bc_counts" = c(nrow(bc_annotation2[bc_annotation2$nchar == 12,]),
                                               nrow(bc_annotation2[bc_annotation2$nchar == 13,]),
                                               nrow(bc_annotation2)))

bc_fraction_list <- list()
for (i in unique(bc_fraction$sample_id)) {
  print(i)
  test_x <- bc_df_filt %>%
    filter(sample_id == i)
  library <- unique(test_x$library)
  
  bc_fraction_list[[i]] <- lapply(unique(bc_fraction$cutoff), function(x) nrow(test_x[test_x$rpm >= x]) / bc_counts_design$bc_counts[bc_counts_design$library == library] *100)
  names(bc_fraction_list[[i]]) <- unique(bc_fraction$cutoff)
}

bc_fraction_2 <- bind_rows(bc_fraction_list)
rownames(bc_fraction_2) <- names(bc_fraction_list)
bc_fraction_2 <- bc_fraction_2 %>%
  rownames_to_column("sample_id") %>%
  pivot_longer(-sample_id, names_to = "cutoff", values_to = "bcs_found") %>%
  mutate(cutoff = as.numeric(cutoff))

bc_fraction <- merge(bc_fraction, bc_fraction_2, by = c("sample_id", "cutoff"))


## How many reads match to designed barcodes?
bc_reads <- bc_df %>%
  dplyr::select(sample_id, library) %>%
  unique() %>%
  na.omit()
rpm_cutoff <- data.frame("cutoff" = seq(0.0001,15,3))
bc_reads <- expand_grid(bc_reads, rpm_cutoff)

bc_df$match <- "Yes"
bc_df$match[is.na(bc_df$library)] <- "No"

bc_reads_list <- list()
for (i in unique(bc_reads$sample_id)) {
  print(i)
  test_x <- bc_df_filt %>%
    filter(sample_id == i)
  test_y <- bc_df %>%
    filter(sample_id == i)
  bc_reads_list[[i]] <- lapply(unique(bc_reads$cutoff), function(x) sum(na.omit(test_x$rpm[test_x$rpm >= x])) / sum(test_y$rpm[test_y$rpm >= x])  *100)
  names(bc_reads_list[[i]]) <- unique(bc_reads$cutoff)
}

bc_reads_2 <- bind_rows(bc_reads_list)
rownames(bc_reads_2) <- names(bc_reads_list)
bc_reads_2 <- bc_reads_2 %>%
  rownames_to_column("sample_id") %>%
  pivot_longer(-sample_id, names_to = "cutoff", values_to = "bc_reads") %>%
  mutate(cutoff = as.numeric(cutoff))

bc_reads <- merge(bc_reads, bc_reads_2, by = c("sample_id", "cutoff"))


bc_fraction <- merge(bc_fraction, bc_reads, by = c("sample_id", "cutoff"))
bc_fraction$bcs_found <- as.numeric(bc_fraction$bcs_found)
bc_fraction$bc_reads <- as.numeric(bc_fraction$bc_reads)

#c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B")
# Plot to evaluate data quality per cutoff
ggplot_custom(bc_fraction) +
  geom_point(aes(x = cutoff, y = bcs_found), color = '#1B998B', size = 1) +
  geom_line(aes(x = cutoff, y = bcs_found), color = '#1B998B', size = 1) +
  geom_point(aes(x = cutoff, y = bc_reads), color = 'black', size = 1) +
  geom_line(aes(x = cutoff, y = bc_reads), color = 'black', size = 1) +
  xlab("rpm cutoff")+
  ylab("total barcodes (green) and matched barcode reads (black) detected (%)")+
  facet_wrap(~sample_id)

## 3: What is the correlation of the cDNA bc counts with the pDNA bc counts? 
pMT09_02 <- bc_df_filt %>%
  filter(str_detect(sample_id, "pMT02-09_v2_r1_gcf6952"), library == "1+2") %>%
  dplyr::select(barcode, sample_id, rpm) %>%
  spread(sample_id, rpm) %>%
  #mutate(pMT09_02 = (`pMT02-09_v2_r1_gcf6952` + `pMT02-09_v3_r1_gcf6952`)/2) %>%
  dplyr::select(barcode, "pMT09_02" = `pMT02-09_v2_r1_gcf6952`)


bc_df_2 <- merge(pMT09_02, bc_df_filt, all = T, by = "barcode")
bc_df_2 <- bc_df_2[!is.na(bc_df_2$sample_id),]

for (i in unique(bc_df_2$library)) {
  bc_df_i <- bc_df_2[bc_df_2$library == i,] 
  if (i == "1") {
  p <- ggplot(bc_df_i, aes(x = pMT02, y = rpm)) +
  geom_bin2d(bins = 50)+
  xlim(0,500) +
  ylim(0,2000)+
  scale_color_viridis() +
  facet_wrap(~sample_id)
  
  print(p)
  }
    if (i == "2") {
  p <- ggplot(bc_df_i, aes(x = pMT09, y = rpm)) +
  geom_bin2d(bins = 50)+
  xlim(0,500) +
  ylim(0,2000)+
  scale_color_viridis() +
  facet_wrap(~sample_id)
  
  print(p)
    }
   if (i == "1+2"){
  p <- ggplot(bc_df_i, aes(x = pMT09_02, y = rpm)) +
  geom_bin2d(bins = 50)+
  xlim(0,500) +
  ylim(0,2000)+
  scale_color_viridis() +
  facet_wrap(~sample_id)
  
  print(p)
  }
  
}

# Generate correlation heatmaps 
for (i in unique(bc_df_2$library)) {
  bc_df_i <- bc_df_2[bc_df_2$library == i,] %>%
    dplyr::select(sample_id, rpm, barcode)  %>% 
    unique() %>%
    spread(sample_id, rpm) %>% 
    filter_all(any_vars(!is.na(.))) %>%
    column_to_rownames("barcode")
  
  x <- cor(bc_df_i, method = "pearson", use = "pairwise.complete.obs")
  
  p<- pheatmap(x, border = "black", main = i)
  print(p)
}
    


cor <- data.frame("sample_id" = unique(bc_df_2$sample_id), "cor" = "", stringsAsFactors = F)

for (i in unique(bc_df_2$sample_id)) {
      cor$cor[cor$sample_id == i] <- stats::cor(bc_df_2$rpm[bc_df_2$sample_id == i], bc_df_2$pMT09_02[bc_df_2$sample_id == i], use = "pairwise.complete.obs")
}

cor_2 <- cor %>%
  mutate(sample_id = gsub("(.*)_.*$", "\\1", sample_id))

bc_df <- merge(bc_df, cor, by = c("sample_id"), all = T)
```

---

### Sample filtering

```{r out.width= "80%", fig.align= "center",  warning= FALSE, message=FALSE}
# Remove samples with low dynamic range (this most likely is due to unhappy cells, pDNA contamination or some other mistake in the sample prep)
for (i in unique(bc_df$sample_id)) {
  bc_df$n_bc[bc_df$sample_id == i] <- length(bc_df$barcode[bc_df$rpm > 500 & bc_df$sample_id == i])
}
bc_df <- bc_df %>%
  filter(n_bc > 75 | str_detect(sample_id, "pMT")) %>% 
  dplyr::select(-n_bc)

# Remove samples with high pDNA contamination (most likely those samples were already removed in the previous step as this is a bit redundant)
bc_df <- bc_df %>%
  filter(cor <= 0.5 | str_detect(sample_id, "pMT")) %>% 
           dplyr::select(-cor)

# Manually remove CREB1 (too little reads)
bc_df <- bc_df %>%
  filter(sample_id != "CREB1_r1_gcf6952") %>%
  filter(!sample_id %in% c("pMT02_3_r1_gcf6952", "pMT02_r1_gcf6952", "pMT02_v2_r1_gcf6952", "pMT02_v3_r1_gcf6952"))

# Remove all non-matching reads
bc_df <- bc_df[!is.na(bc_df$tf),]
```

---


### Normalization of barcode counts:  
Divide cDNA barcode counts through pDNA barcode counts to get activity

```{r normalization, out.width= "80%", fig.align= "center",  warning= FALSE, message=FALSE}
# Prepare pDNA data and remove all barcodes with less than 20 counts
pMT09_02_1 <- bc_df %>%
  filter(sample_id == "pMT02-09_v2_r1_gcf6952") %>%
   filter(starcode_counts > 10) %>%
   dplyr::select(barcode, "pMT09_02_v2" = rpm)

# Add pDNA data as separate column - for both library 1 and library 2
bc_df <- merge(bc_df, pMT09_02, by = "barcode")
bc_df <- merge(bc_df, pMT09_02_1, by = "barcode")

# Compute activity by dividing cDNA bc counts through pDNA bc counts
bc_df$activity <- bc_df$rpm / bc_df$pMT09_02_v2

# Remove rows that could not be computed due to too little pDNA counts
bc_df <- bc_df %>%
  drop_na(activity)
```


---

## Calculate mean activity - filter out outlier barcodes 

```{r out.width= "80%", fig.align= "center",  warning= FALSE, message=FALSE}
# First identify and remove outlier barcodes - this removes the noise created by faulty barcode clustering etc. 
## Calculate mean and SD for each reporter
bc_df_cDNA <- bc_df %>%
  filter(str_detect(sample_id, "pMT", negate = T), str_detect(sample_id, "pDNA", negate = T))

bc_df_cDNA$reporter_id <- paste(bc_df_cDNA$tf, bc_df_cDNA$spacing, 
                                bc_df_cDNA$distance, bc_df_cDNA$promoter,
                                bc_df_cDNA$background, sep = "_")

bc_df_cDNA <- bc_df_cDNA %>%
  mutate(reporter_id = gsub("bc-[0-9]{1,2}_", "", reporter_id))

bc_df_cDNA$mean_activity <- ave(bc_df_cDNA$activity, bc_df_cDNA$reporter_id, 
                                bc_df_cDNA$sample_id, FUN =
                                  function(x) mean(x))
bc_df_cDNA$sd_activity <- ave(bc_df_cDNA$activity, bc_df_cDNA$reporter_id, 
                              bc_df_cDNA$sample_id,  FUN =
                                  function(x) sd(x))



## Remove data points that are 2xSD away from the mean
bc_df_cDNA$upper_activity <- bc_df_cDNA$mean_activity + (2 * bc_df_cDNA$sd_activity)
bc_df_cDNA$lower_activity <- bc_df_cDNA$mean_activity - (2 * bc_df_cDNA$sd_activity)

bc_df_cDNA$low_outlier <- bc_df_cDNA$activity - bc_df_cDNA$lower_activity
bc_df_cDNA$high_outlier <- bc_df_cDNA$upper_activity - bc_df_cDNA$activity

## Plot effect of highest outlier bc -> SNP in minP promoter
outlier <- bc_df_cDNA[bc_df_cDNA$reporter_id == "ZFP42_5bp_21bp_minP_3",] %>% 
  dplyr::select(barcode_number, activity, sample) %>%
  unique()

ggplot_custom(outlier, aes(x = barcode_number, y = activity)) +
  geom_quasirandom() +
  xlab("barcode #") + 
  labs(title = "Zfp42 reporter - TATA-box minP mutated", 
       subtitle = "bc8 AGAGGGTATATAAT -> AGAGGGGATATAAT")


## Plot effect of highest outlier bc -> Elk1 outlier
outlier <- bc_df_cDNA[bc_df_cDNA$reporter_id == "ELK1_5bp_21bp_mCMV_2",] %>% 
  dplyr::select(barcode_number, activity) %>%
  unique()

ggplot_custom(outlier, aes(x = barcode_number, y = activity)) +
  geom_quasirandom()+
  xlab("barcode #") + 
  labs(title = "Elk1 reporter - bc1 attached to Pax6 reporter")



## Plot effect of highest outlier bc -> Elk1 outlier
outlier <- bc_df_cDNA[bc_df_cDNA$reporter_id == "TCF3_5bp_10bp_hBGm_3",] %>% 
  dplyr::select(barcode_number, activity) %>%
  unique()

ggplot_custom(outlier, aes(x = barcode_number, y = activity)) +
  geom_quasirandom() +
  xlab("barcode #") + 
  labs(title = "Tcf3 reporter - bc1 attached to multiple (less active) reporters")



## Choose arbitrary cutoff to get rid of most extreme outliers
bc_df_cDNA_filt <- bc_df_cDNA[bc_df_cDNA$low_outlier > -0.3 & bc_df_cDNA$high_outlier > -2 & bc_df_cDNA$sd_activity < 100,]

## Recalculate mean and sd
bc_df_cDNA_filt$mean_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, 
                                bc_df_cDNA_filt$sample_id, FUN =
                                  function(x) mean(x))
bc_df_cDNA_filt$sd_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, 
                              bc_df_cDNA_filt$sample_id,  FUN =
                                  function(x) sd(x))
```


### Scaling data 

```{r out.width= "80%", fig.align= "center",  warning= FALSE}
## Correct native enhancer promoter names
bc_df_cDNA_filt2 <- bc_df_cDNA_filt %>%
  filter(native_enhancer == "Yes") %>%
  mutate(promoter = gsub(".*(minP|mCMV|hBGm|Random).*", "\\1", promoter))

bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  filter(native_enhancer == "No")

bc_df_cDNA_filt <- rbind(bc_df_cDNA_filt, bc_df_cDNA_filt2)

## Compute activity relative to minimal promoter background
bc_df_cDNA_filt_neg <- bc_df_cDNA_filt %>%
  dplyr::select(sample_id, activity, reporter_id, tf, promoter) %>%
  filter(str_detect(tf, "RANDOM")) %>%
  unique() %>%
  mutate(activity = ave(activity, promoter, sample_id, FUN = function(x) mean(x))) %>%
  dplyr::select(-reporter_id) %>%
  unique() %>%
  dplyr::select("tf_activity2" = activity, promoter, sample_id) %>%
  unique()


### For library 2 I don't have random motifs, so I will take the mean of the mutated motifs instead (get rid of the 2% highest active mutated controls)
bc_df_cDNA_filt_neg2 <- bc_df_cDNA_filt %>%
  dplyr::select(sample_id, reporter_id, tf, promoter, neg_ctrls, mean_activity) %>%
  filter(neg_ctrls == "Yes") %>%
  unique() %>%
  group_by(sample_id, promoter) %>%
  dplyr::top_frac(-0.75, mean_activity) %>%
  mutate(activity = ave(mean_activity, promoter, sample_id, FUN = function(x) mean(x))) %>%
  dplyr::select(-reporter_id) %>%
  unique() %>%
  dplyr::select("tf_activity" = activity, promoter, sample_id) %>%
  unique()

bc_df_cDNA_filt$promoter[bc_df_cDNA_filt$hPGK == "Yes"] <- "minP"


x <- merge(bc_df_cDNA_filt_neg, bc_df_cDNA_filt_neg2)
plot(x$tf_activity, x$tf_activity2)
### Normalization by random has some outliers (gcf6502) -> I can better for now normalize by mutated motifs since I also have data for all samples from that


bc_df_cDNA_filt <- merge(bc_df_cDNA_filt, bc_df_cDNA_filt_neg2, by = c("sample_id", "promoter"))
bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  mutate(minP_activity = activity / tf_activity)


### Compute a p-value for each reporter with respect to the background minimal promoter activity
bc_df_cDNA_filt$library[is.na(bc_df_cDNA_filt$library)] <- "1+2"
bc_df_cDNA_filt_neg2 <- bc_df_cDNA_filt %>%
  dplyr::select(sample_id, reporter_id, tf, promoter, neg_ctrls, mean_activity, condition, library) %>%
  filter(neg_ctrls == "Yes") %>%
  unique() %>%
  group_by(sample_id, promoter) %>%
  dplyr::top_frac(-0.75, mean_activity) %>%
  mutate(id = paste(condition, promoter, library, sep = "_"))

bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  mutate(reporter_id2 = gsub("(.*)_.*(_[1-3]{1}$)", "\\1\\2", reporter_id),
         id = paste(condition, promoter, library, sep = "_"))


pval_list <- list()
for (i in unique(bc_df_cDNA_filt$reporter_id2)) {
  print(i)
  test_x <- bc_df_cDNA_filt[bc_df_cDNA_filt$reporter_id2 == i,]
  ids <- unique(bc_df_cDNA_filt$id[bc_df_cDNA_filt$reporter_id2 == i])
  pval_list[[i]] <- lapply(ids, function(x) t.test(test_x$activity[test_x$id == x], 
                                                   bc_df_cDNA_filt_neg2$mean_activity[bc_df_cDNA_filt_neg2$id == x])$p.value)
  names(pval_list[[i]]) <- ids
}

pval_df <- data.frame(bind_rows(pval_list))
rownames(pval_df) <- names(pval_list)
pval_df <- pval_df %>%
  rownames_to_column("reporter_id2") %>%
  pivot_longer(-reporter_id2, names_to = "id", values_to = "pval") %>%
  mutate(id = gsub("1.2", "1+2", id)) %>%
  na.omit()

bc_df_cDNA_filt <- merge(bc_df_cDNA_filt, pval_df, by = c("reporter_id2", "id"), all = T)
bc_df_cDNA_filt <- bc_df_cDNA_filt[!is.na(bc_df_cDNA_filt$sample_id),]

bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  mutate(pval_adj = p.adjust(p = pval))

## Compute activity relative to NFYA reporter
bc_df_cDNA_filt_nfya <- bc_df_cDNA_filt %>%
  dplyr::select(sample_id, activity, reporter_id, tf, promoter) %>%
  filter(str_detect(tf, "NFYA"), promoter == "minP") %>%
  unique() %>%
  mutate(activity = ave(activity, sample_id, FUN = function(x) mean(x))) %>%
  dplyr::select(-reporter_id) %>%
  unique() %>%
  dplyr::select("nfya_activity" = activity, sample_id) %>%
  unique()

bc_df_cDNA_filt <- merge(bc_df_cDNA_filt, bc_df_cDNA_filt_nfya, by = "sample_id")
bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  mutate(nfya_activity = activity / nfya_activity)

# Compute mean of technical replicates
bc_df_cDNA_filt$mean_activity_sample_minP <- ave(bc_df_cDNA_filt$minP_activity, bc_df_cDNA_filt$reporter_id,
                                bc_df_cDNA_filt$sample_id, FUN =
                                  function(x) mean(x, na.rm = T))

bc_df_cDNA_filt$mean_activity_sample_nfya <- ave(bc_df_cDNA_filt$nfya_activity, bc_df_cDNA_filt$reporter_id,
                                bc_df_cDNA_filt$sample_id, FUN =
                                  function(x) mean(x, na.rm = T))
bc_df_cDNA_filt$mean_activity_sample <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id,
                                bc_df_cDNA_filt$sample_id, FUN =
                                  function(x) mean(x, na.rm = T))


# Compute activity relative to mutated motif
bc_df_cDNA_filt_mutated <- bc_df_cDNA_filt %>%
  filter(neg_ctrls == "Yes") %>%
  dplyr::select(reporter_id, 'mutated_activity' = mean_activity_sample, sample_id) %>%
  unique() %>%
  mutate(reporter_id = gsub("_neg", "", reporter_id)) %>%
  filter(str_detect(reporter_id, "MAFA_", negate = T), str_detect(reporter_id, "NR2E3_", negate = T), str_detect(reporter_id, "^T_", negate = T))

bc_df_cDNA_filt <- merge(bc_df_cDNA_filt, bc_df_cDNA_filt_mutated, by = c("reporter_id", "sample_id"), all = T)
bc_df_cDNA_filt <- bc_df_cDNA_filt[!is.na(bc_df_cDNA_filt$barcode),]
bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  mutate(mean_activity_sample_neg = mean_activity / mutated_activity)

## Statistical significance of reporters vs negative control
bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  mutate(reporter_id2 = gsub("_neg", "", reporter_id))

pval_list <- list()
for (i in unique(bc_df_cDNA_filt$reporter_id2)) {
  print(i)
  test_x <- bc_df_cDNA_filt[bc_df_cDNA_filt$reporter_id2 == i,]
  ids <- intersect(unique(bc_df_cDNA_filt$id[bc_df_cDNA_filt$reporter_id2 == i & bc_df_cDNA_filt$neg_ctrls == "Yes"]),
                     unique(bc_df_cDNA_filt$id[bc_df_cDNA_filt$reporter_id2 == i & bc_df_cDNA_filt$neg_ctrls == "No"]))
    pval_list[[i]] <- lapply(ids, function(x) t.test(test_x$activity[test_x$id == x & test_x$neg_ctrls == "No"], 
                                                   test_x$activity[test_x$id == x & test_x$neg_ctrls == "Yes"])$p.value)
    names(pval_list[[i]]) <- ids
}

pval_list <- pval_list[lengths(pval_list) > 0]
pval_df <- data.frame(bind_rows(pval_list))
rownames(pval_df) <- names(pval_list)
pval_df <- as.data.frame(pval_df)
pval_df <- pval_df %>%
  rownames_to_column("reporter_id2") %>%
  pivot_longer(-reporter_id2, names_to = "id", values_to = "pval_neg") %>%
  mutate(id = gsub("1.2", "1+2", id)) %>%
  na.omit()

bc_df_cDNA_filt <- merge(bc_df_cDNA_filt, pval_df, by = c("reporter_id2", "id"), all = T)
```


---


## Calculate correlations between technical replicates

```{r correlations_2, out.width= "80%", fig.align= "center",  warning= FALSE}
## Identify and relabel double reporters
bc_df_12 <- bc_df_cDNA_filt %>%
  filter(nchar == 12, library == "1+2")
bc_df_13 <- bc_df_cDNA_filt %>%
  filter(nchar == 13, library == "1+2")
double_reporters <- unique(bc_df_13$reporter_id[bc_df_13$reporter_id %in% bc_df_12$reporter_id])

double_reporters_df <- bc_df_cDNA_filt %>%
  filter(reporter_id %in% double_reporters, library == "1+2", nchar == 13)
double_reporters_df$reporter_id_2 <- paste(double_reporters_df$reporter_id, double_reporters_df$nchar)
bc_df_cDNA_filt$reporter_id_2 <- paste(bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$nchar)
bc_df_cDNA_filt <- bc_df_cDNA_filt[!bc_df_cDNA_filt$reporter_id_2 %in% double_reporters_df$reporter_id_2,] %>%
  dplyr::select(-reporter_id_2)
double_reporters_df <- double_reporters_df %>%
  dplyr::select(-reporter_id_2)

double_reporters_df$barcode_number[double_reporters_df$barcode_number == 1] <- 9
double_reporters_df$barcode_number[double_reporters_df$barcode_number == 2] <- 10
double_reporters_df$barcode_number[double_reporters_df$barcode_number == 3] <- 11
double_reporters_df$barcode_number[double_reporters_df$barcode_number == 4] <- 12
double_reporters_df$barcode_number[double_reporters_df$barcode_number == 5] <- 13

bc_df_cDNA_filt <- rbind(bc_df_cDNA_filt, double_reporters_df)

## Combine replicates in 8 different columns

bc_df_rep <- bc_df_cDNA_filt %>% 
  filter(commercial_reporter == "No", rand_promoter == "No", native_enhancer == "No", hPGK == "No") %>%
  dplyr::select(barcode_number, activity, sample_id, reporter_id, neg_ctrls, library) %>%
  spread(barcode_number, activity)
names(bc_df_rep) <- gsub("([1-9]{1})", "TR\\1", names(bc_df_rep))


for (i in unique(bc_df_rep$library)) {
  p <- ggscatter(bc_df_rep[bc_df_rep$library == i,], x = "TR1", y = "TR2",
                 add = "reg.line",
                 color = "neg_ctrls",
                 size = 0.5,
                 add.params = list(color = "blue", fill = "lightgray"), 
                 title = paste("rep1 vs rep2 - library:", i),
                 conf.int = TRUE, ylab = "rep2", xlab = "rep1") + 
    stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
    geom_abline(linetype = "dashed") +
    xlim(0,50) + ylim(0,50) +
    #scale_color_manual(values = colors) +
    facet_wrap(~sample_id)
  print(p)
}


# Correlation matrix plot
n <- sample(1:nrow(bc_df_rep), 5000)
boundaries <- seq(from = 0.8, by = 0.05, length.out = 4)
plt <- ggpairs(bc_df_rep %>% filter(library == "1+2") %>% dplyr::select('TR1', 'TR2', 'TR3', 'TR4', 'TR5', 'TR6', 'TR7', 'TR8'),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.1, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red") +
                   theme_bw()}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red") +
                   theme_bw()})) +
  ggtitle("Correlation Between Technial Replicates") +
  theme(text = element_text(size = 20)) +
  xlab("Reporter activity") +
  ylab("Reporter activity")

print(plt)
```


---

## Correlation between replicates

```{r correlations_3, out.width= "80%", fig.align= "center",  warning= FALSE}
# Correlation plots of the replicates
## Combine replicates of normalized data in 3 different columns
bc_df_rep <- bc_df_cDNA_filt %>% 
  dplyr::select(replicate, mean_activity_sample_minP, reporter_id, condition, neg_ctrls, library, gcf) %>% 
  unique() %>%
  spread(replicate, mean_activity_sample_minP)

colors <- c("#2D3047", "#1B998B")

for (i in unique(bc_df_rep$library)) {
  p <- ggscatter(bc_df_rep[bc_df_rep$library == i,], x = "r1", y = "r2",
                 add = "reg.line",
                 color = "neg_ctrls",
                 size = 0.5,
                 add.params = list(color = "blue", fill = "lightgray"), 
                 title = paste("rep1 vs rep2 - library:", i),
                 conf.int = TRUE, ylab = "rep2", xlab = "rep1") + 
    stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
    geom_abline(linetype = "dashed") +
    xlim(0,50) + ylim(0,50) +
    #scale_color_manual(values = colors) +
    facet_wrap(~condition)
  print(p)
}
```


```{r data export, out.width= "80%", fig.align= "center",  warning= FALSE}
# Mean of the three replicates
bc_df_cDNA_filt$reporter_activity <- ave(bc_df_cDNA_filt$mean_activity_sample, bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$condition, FUN = function(x) mean(x))
bc_df_cDNA_filt$reporter_activity_minP <- ave(bc_df_cDNA_filt$mean_activity_sample_minP, bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$condition, FUN = function(x) mean(x))
bc_df_cDNA_filt$reporter_activity_nfya <- ave(bc_df_cDNA_filt$mean_activity_sample_nfya, bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$condition, FUN = function(x) mean(x))
bc_df_cDNA_filt$reporter_activity_neg <- ave(bc_df_cDNA_filt$mean_activity_sample_neg, bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$condition, FUN = function(x) mean(x))

# Polish export dataframe
bc_df_cDNA_filt <- bc_df_cDNA_filt %>% 
  dplyr::select(-upper_activity, -lower_activity, 
         -high_outlier, -low_outlier, -match, -pMT09_02_v2, -pMT09_02, -mean_activity, -sd_activity)


# Export bc_df for cDNA analysis
filename <- SetFileName("_reporter_activity_filt_combined", "mt")
setwd("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6952_KD_HepG2/results/")
write.csv(bc_df_cDNA_filt, file = paste(filename,".csv", sep = ""), row.names = F)
```

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

