---
title: "RNA-seq analysis stimulated HepG2 cells"
author: 
  - name: "Max Trauernicht"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute - van Steensel lab"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
---

---

### Aim
I want to check if stimulation with 4 different nuclear receptor agonists induces changes in expression of NR target genes. This script contains basic DESeq2 analysis to do that. 

---


## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}
```

### Libraries

```{r set-up, message = F, warning = F}

# Load dependencies
library(tidyverse)
library(GenomicRanges)
library(rtracklayer)
library(ggplot2)
library(ggbeeswarm)
library(DESeq2)
library(RColorBrewer)
library(GGally)
library(ggpubr)
library(biomaRt)
library(clusterProfiler)
library(pathview)
library(enrichplot)
library(DOSE)
library(ggnewscale)
library(ggridges)
library(europepmc)
library(GENOVA)
library(ggh4x)
library(SummarizedExperiment)
library(seriation)

# # Prepare output 
output_dir <- "mt20210525_rnaseq_processing"
dir.create(output_dir, showWarnings = FALSE)

```

### Custom functions

```{r functions}

ExonLength <- function(genes) {
  # Get the combined exon length for the genes
  tib <- as_tibble(genes) %>%
    filter(type == "exon") %>%
    mutate(gene_id = factor(gene_id, levels = unique(gene_id))) %>%
    group_by(gene_id) %>%
    dplyr::summarise(exon_number = n(),
              exon_length = sum(width))
  
  tib
}

# From Fede:
# ggpairs custom functions
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {

  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y)
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)

  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }

  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())

  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]

  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[3]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }

  p <- p +
    theme(panel.background = element_rect(fill = corCol))

  return(p)
}

plotMAWithLabels <- function(results, main, alpha = 0.05) {
  
  # Plot MA plot with lamin genes highlighted 
  plotMA(results, ylim = c(-4, 4), main = main,
         ylab = "log2 fold change", alpha = alpha)
  points(results[c("ENSG00000165029", "ENSG00000160179"), "baseMean"],
         results[c("ENSG00000165029", "ENSG00000160179"), "log2FoldChange"],
         col = c("red"), cex = 1.5, pch = 19)
  legend("topright", 
         legend = c(paste0("non-sign (n=", sum(results$padj >= alpha, na.rm = T), ")"), 
                    paste0("sign (n=", sum(results$padj < alpha, na.rm = T), ")"), 
                    "ABCA1"), 
         col = c("black", "blue", "red"), pch = 19)
  
}

```


### Prepare data and DESeq2

Read data files and initialize the DESeq2 object.


```{r load rna-seq}

#######################################
## Prepare metadata

# Prepare metadata sheet
metadata <- read_tsv("ts210514_metadata_PE.tsv") %>%
  filter(grepl("HepG2", sample_name)) %>%
  dplyr::select(-fastq) %>%
  #mutate(sample_id_short = str_remove(sample_id, "_R1_001")) %>%
  separate(sample_name, c("cell", "condition", "replicate"), sep = "_", remove = F) %>%
  mutate(condition = factor(condition, levels = unique(condition)),
         replicate = factor(replicate, levels = unique(replicate)))


#######################################
## Prepare gene annotation

# Load genes
genes <- import("/home/t.v.schaik/mydata/data/gene_builds/GRCh38/gencode.v24.primary_assembly.annotation.gtf")

# Get gene lengths (total exon length)
gene_length <- ExonLength(genes)

# dplyr::select genes only
genes <- genes[genes$type == "gene"]

# Add seqinfo
# chrom_sizes <- read.table("/DATA/usr/t.v.schaik/data/genomes/GRCh38/hg38.chrom.sizes", sep = "\t")
# row.names(chrom_sizes) <- chrom_sizes[, 1]
# seqlengths(genes) <- chrom_sizes[seqlevels(genes), 2]


#######################################
## Prepare count matrix

# Read count matrix corresponding to the metadata
rnaseq_matrix <- read_tsv("/DATA/scratch/usr/t.v.schaik/proj/3D_nucleus/results/ts210514_RNAseq_Ki67_depletion/results_pe/Star/count_table.tsv") %>%
  dplyr::select(1, matches("HepG2")) %>%
  rename_at(vars(names(.)), ~ c("ensembl_id", metadata$sample_name))

# Filter for strange chromosomes
idx_chromosome <- which(seqnames(genes) %in% c(paste0("chr", 1:22),
                                               "chrX"))

genes <- genes[idx_chromosome]
rnaseq_matrix <- rnaseq_matrix[idx_chromosome, ]

# Filter for protein coding genes (for now)
filter <- c("protein_coding")
idx_filter <- which(genes$gene_type %in% filter)

genes <- genes[idx_filter]
rnaseq_matrix <- rnaseq_matrix[idx_filter, ]

# Filter for genes with no counts
tib_nocounts <- rnaseq_matrix %>%
  mutate(ensembl_id = factor(ensembl_id, levels = ensembl_id)) %>%
  gather(key, value, -ensembl_id) %>%
  group_by(ensembl_id) %>%
  dplyr::summarise(count = sum(value > 2) > 2)
idx_nocounts <- which(tib_nocounts$count)

genes <- genes[idx_nocounts]
rnaseq_matrix <- rnaseq_matrix[idx_nocounts, ]

# Add gene names
ensembl <- useMart("ensembl")
mart <- useDataset("hsapiens_gene_ensembl", mart = ensembl)
rnaseq_matrix <- rnaseq_matrix %>%
  mutate(ensembl_id = gsub(".[0-9]+$", "", ensembl_id))
gene <- rnaseq_matrix$ensembl_id
G_list <- getBM(filters= "ensembl_gene_id", attributes= c("ensembl_gene_id","hgnc_symbol"),values=gene,mart= mart)
rnaseq_matrix <- merge(rnaseq_matrix,G_list,by.x="ensembl_id",by.y="ensembl_gene_id")
```



## 1. PCA plot

```{r initialize deseq2, fig.width = 6, fig.height = 3.5}

#######################################
## Prepare count matrix

# Convert into data.frame with row.names for deseq2
rnaseq_counts_2 <- rnaseq_matrix %>%
  filter(hgnc_symbol != "")
rnaseq_counts <- rnaseq_counts_2 %>%
  dplyr::select(-ensembl_id, -hgnc_symbol)
rnaseq_counts <- data.frame(rnaseq_counts, row.names = rnaseq_counts_2$hgnc_symbol)


#######################################
## Initialize deseq2

# Prepare metadata ready for deseq
metadata_df <- data.frame(metadata)
metadata_df[] <- lapply(metadata_df, function(x) gsub("-", "_", x))

# Initialize
rnaseq_dds <- DESeqDataSetFromMatrix(countData = rnaseq_counts,
                                     colData = metadata_df,
                                     design= ~ condition)

# Execute deseq2
rnaseq_dds <- DESeq(rnaseq_dds)


#######################################
## PCA analysis - quality control

# Get the "normalized" values and create PCA plot
# Note that "normalized" is simply log2 + 0.01 transformed normalized counts!
rnaseq_dds_norm <- normTransform(rnaseq_dds, pc = 0.01)
plt <- plotPCA(rnaseq_dds_norm, intgroup = c("condition"), ntop = 100)

plt + 
  theme_bw() +
  theme(aspect.ratio = 1) +
  scale_shape_manual()

```

The PCA plot does not look perfect. But maybe PC2 is representing biological meaning.

---

## 2. Correlation plot replicates

A quick initial test: how well do the replicates correlate with each other and
between the different experiments?

```{r correlation plots, fig.width = 6, fig.height = 6, message = F, warning = F}

#######################################
## Make a correlation plot of all samples - quality control

# Get the "normalized" counts for each experiment - this is a different
# normalized from above!
rnaseq_norm <- as_tibble(assay(rnaseq_dds_norm)) %>%
  rename_at(vars(names(.)), ~ metadata$sample_name) %>%
  add_column(hgnc_symbol = rnaseq_counts_2$hgnc_symbol)

# rnaseq_norm <- rnaseq_norm[, order(samples.df$clone)]


# Also, combine the normalized counts
rnaseq_norm_combined <- do.call(cbind,
                                tapply(as.character(metadata$sample_name),
                                       metadata$condition,
                                       function(i) rowMeans(rnaseq_norm[, i])))

tib_norm_combined <- as_tibble(rnaseq_norm_combined, 
                               .name_repair = ~c("DMSO", "FXR", "LXR", "PPARa", "SXR")) %>%
  add_column(gene_id = rnaseq_counts_2$hgnc_symbol)

tib_norm_combined <- as_tibble(genes) %>%
  dplyr::select(seqnames, start, end, strand, gene_id, gene_name) %>%
  inner_join(tib_norm_combined, by = "gene_id")


# Get a sample for plotting dplyr::select points
n <- sample(1:nrow(rnaseq_norm), 1000)

# Use GGally to make correlation plots
boundaries <- seq(from = 0.9, by = 0.01, length.out = 4)
plt <- ggpairs(rnaseq_norm %>% dplyr::select(-hgnc_symbol),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.1, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red") +
                   theme_bw()}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red") +
                   theme_bw()})) +
  ggtitle("Correlation gene expression") +
  xlab("Gene expression") +
  ylab("Gene expression")
  # theme_bw()

print(plt)

```

The main point of the figure: there is no single replicate misbehaving.

---


## 3. Clustering of correlation matrix

Are the replicates clustering together?
```{r correlation matrix, fig.width = 5.5, fig.height = 5}
norm <- assay(vst(rnaseq_dds, nsub = 10))
dist <- as.dist(1 - cor(norm, method = "spearman"))

clust <- hclust(dist, method = "ward.D2")
clust <- reorder(clust, dist = dist, method = "OLO")

df <- reshape2::melt(cor(norm, method = 'spearman'))

ggplot(df, aes(Var1, Var2, fill = value)) +
  geom_raster() +
  guides(y.sec = "axis", x.sec = guide_axis(position = "bottom")) +
  GENOVA:::scale_fill_GENOVA_div(name = expression("Spearman's "*rho)) +
  scale_x_dendrogram(hclust = clust, name = NULL,
                     guide = guide_dendro(label = FALSE, position = "top")) +
  scale_y_dendrogram(hclust = clust, name = NULL,
                     guide = guide_dendro(label = FALSE)) +
  theme(axis.text.x.bottom = element_text(angle = 90, hjust = 1, vjust = 0.5))
```
It looks like replicate 3 clusters away from replicate 1 & 2. The stimulations do not really cluster together. If anything, LXR, SXR and PPARa agonists have similar effects and cluster away from FXR and DMSO.

---

## 4. Differential analysis

Identifying indivudal genes that change expression upon stimulation. 
  
```{r differential analysis, fig.width = 5.5, fig.height = 5}

#######################################
## Differential analysis - test for lfc > 0.5

# Differential analysis
diff_results <- results(rnaseq_dds,
                       contrast = c("condition", "LXR", "DMSO"))



# MA plots with significant hits
ggmaplot(diff_results, main = expression("DMSO" %->% "LXR"),
   fdr = 0.05, fc = 1.5, size = 1,
   palette = c("#B31B21", "#1465AC", "darkgray"),
   genenames = as.vector(diff_results$name),
   legend = "top", top = 100,
   font.label = c("bold", 11),
   font.legend = "bold",
   font.main = "bold",
   ggtheme = ggpubr::theme_pubr())


# Extract hits from all stimulations
df_changed_LXR <- data.frame("gene" = rownames(diff_results),
                         "basal_exp" = diff_results$baseMean,
                         "lfc" = diff_results$log2FoldChange, 
                         "p_value" = diff_results$pvalue,
                         "NR" = "LXR")


diff_results_PPARa <- results(rnaseq_dds,
                       contrast = c("condition", "PPARa", "DMSO"))

ggmaplot(diff_results_PPARa, main = expression("DMSO" %->% "PPARa"),
   fdr = 0.05, fc = 1.5, size = 1,
   palette = c("#B31B21", "#1465AC", "darkgray"),
   genenames = as.vector(diff_results$name),
   legend = "top", top = 100,
   font.label = c("bold", 11),
   font.legend = "bold",
   font.main = "bold",
   ggtheme = ggplot2::theme_bw())

df_changed_PPARa <- data.frame("gene" = rownames(diff_results_PPARa),
                         "basal_exp" = diff_results_PPARa$baseMean,
                         "lfc" = diff_results_PPARa$log2FoldChange, 
                         "p_value" = diff_results_PPARa$pvalue,
                         "NR" = "PPARa")

diff_results_SXR <- results(rnaseq_dds,
                       contrast = c("condition", "SXR", "DMSO"))

ggmaplot(diff_results_SXR, main = expression("DMSO" %->% "SXR"),
   fdr = 0.05, fc = 1.5, size = 1,
   palette = c("#B31B21", "#1465AC", "darkgray"),
   genenames = as.vector(diff_results$name),
   legend = "top", top = 100,
   font.label = c("bold", 11),
   font.legend = "bold",
   font.main = "bold",
   ggtheme = ggplot2::theme_bw())

df_changed_SXR <- data.frame("gene" = rownames(diff_results_SXR),
                         "basal_exp" = diff_results_SXR$baseMean,
                         "lfc" = diff_results_SXR$log2FoldChange, 
                         "p_value" = diff_results_SXR$pvalue,
                         "NR" = "SXR")

diff_results_FXR <- results(rnaseq_dds,
                       contrast = c("condition", "FXR", "DMSO"))

ggmaplot(diff_results_FXR, main = expression("DMSO" %->% "FXR"),
   fdr = 0.05, fc = 1.5, size = 1,
   palette = c("#B31B21", "#1465AC", "darkgray"),
   genenames = as.vector(diff_results$name),
   legend = "top", top = 100,
   font.label = c("bold", 11),
   font.legend = "bold",
   font.main = "bold",
   ggtheme = ggplot2::theme_bw())

df_changed_FXR <- data.frame("gene" = rownames(diff_results_FXR),
                         "basal_exp" = diff_results_FXR$baseMean,
                         "lfc" = diff_results_FXR$log2FoldChange, 
                         "p_value" = diff_results_FXR$pvalue,
                         "NR" = "FXR")

df_changed <- rbind(df_changed_LXR, df_changed_PPARa, df_changed_SXR, df_changed_FXR)


# df_changed <- df_changed %>%
#   filter(p_value <= 0.1)
```
Despite the poor global trends, there seem to be a fraction of genes significantly upregulated (even though FC are not very high). At a first glance, these genes seem to be relevant.

---

## 5. Gene ontology of changed set of genes

Can the upregulated genes be assigned to specific groups of genes? This might tell us if we should focus on a specific set of genes.

```{r gene ontology, fig.width = 5.5, fig.height = 5}
# SET THE DESIRED ORGANISM HERE
organism = "org.Hs.eg.db"
library(organism, character.only = TRUE)


# Gene set enrichment analysis for each stimulation
for (i in unique(df_changed$NR)) {
  if (i != "FXR") {
  print(i)
  
  # we want the log2 fold change 
  gene_list <- df_changed$lfc[df_changed$NR == i]
  
  # name the vector
  names(gene_list) <- df_changed$gene[df_changed$NR == i]
  gene_list <- na.omit(gene_list)
  
  # sort the list in decreasing order (required for clusterProfiler)
  gene_list = sort(gene_list, decreasing = TRUE)
  
  # Compute the gene set enrichment
  gse <- gseGO(geneList = gene_list, 
             ont = "MF",
             keyType = "SYMBOL", 
             OrgDb = org.Hs.eg.db)
  
  # Plot the importance of the gene sets
  p<-enrichplot::dotplot(gse, showCategory=10, split=".sign", orderBy = "x") + facet_grid(.~.sign) + ggtitle(paste(i))
  print(p)
  
  # Plot the interaction of genes within sets
  p<-cnetplot(gse, categorySize="pvalue", foldChange=gene_list, showCategory = 3) + ggtitle(paste(i))
  print(p)
  
  # Plot the enrichment distribution
  p<-ridgeplot(gse) + labs(x = "enrichment distribution") + ggtitle(paste(i))
  print(p)
  
  # Distribution of genes in most important gene set
  p<-gseaplot(gse, by = "all", title = paste(gse$Description[1], i), geneSetID = 1)
  print(p)
  
  # How relevant are the identified gene sets?
  terms <- gse$Description[1:3]
  p<-pmcplot(terms, 2010:2020, proportion=FALSE) + theme_bw() + ggtitle(paste(i))
  print(p)
  }
}
```





---




## 6. Overlap with nuclear receptor motif

Finally, let's ask the important question: do the upregulated genes have relevant NR motifs in their promoters/enhancers?

```{r motifs}



```








---






Plot the enrichment of differential genes within LADs.

```{r plot enrichment differential genes in LADs, fig.width = 5, fig.height = 3.5}

# #######################################
# ## Create summarizing tibble - only using active genes
# 
# tib <- as_tibble(diff_results) %>%
#   dplyr::select(baseMean, log2FoldChange, padj) %>%
#   add_column(ensembl_id = genes$gene_id,
#              overlaps_LAD = genes$overlaps_LAD,
#              overlaps_NAD = genes$overlaps_NAD) %>%
#   mutate(padj = replace_na(padj, 1))
# 
# tib_gather <- tib %>%
#   dplyr::select(-starts_with("log")) %>%
#   gather(key, value, contains("overlaps")) %>%
#   mutate(sign = padj < 0.05)
# 
# # Group by LAD overlap or significance
# tib_diff <- tib_gather %>%
#   group_by(key, value) %>%
#   dplyr::summarise(fraction = mean(sign),
#             total = sum(sign),
#             background = length(sign))
# 
# #######################################
# ## Plot results
# 
# tib_diff %>%
#   ggplot(aes(x = key, y = fraction, fill = value)) +
#     geom_bar(stat = "identity", position = "dodge", col = "black") +
#     xlab("Knockout cells") +
#     ylab("Fraction differentially expressed") +
#     scale_fill_manual(values = c("grey", "black"), name = "Overlaps LAD") +
#     theme_bw() +
#     theme(aspect.ratio = 1)
# 
# tib_diff %>%
#   ggplot(aes(x = key, y = total, fill = value)) +
#     geom_bar(stat = "identity", position = "dodge", col = "black") +
#     xlab("Knockout cells") +
#     ylab("Total differentially expressed") +
#     scale_fill_manual(values = c("grey", "black"), name = "Overlaps LAD") +
#     theme_bw() +
#     theme(aspect.ratio = 1)
# 
# 
# ```
# 
# Not significant.
# 
# The analysis above is based on differential gene expression and simply asking 
# whether LAD genes are over-represented. However, there is a bias in gene 
# expression levels in LAD genes compared to iLAD genes: they often are more lowly
# expressed. It could that this bias also affects the result. To rule this out, 
# let's create a matched set (Christ's script) and use that.
# 
# ```{r repeat with matched set, fig.width = 5, fig.height = 3.5}
# 
# ## Function by Christ Leemans
# ## Create a matched set
# ##
# ## get a table with matching sets
# ## table = complete table to take matching sets from
# ## class_col = column name of class of interest
# ## class = name of class to match the set on
# ## order_on = column name to order on
# ## bs = bin size to sample equal number of items from
# matchSet <- function(table, class_col, class, order_on, bs=10){
# 	# order by value of interest
# 	o_vec = order(table[,order_on])
# 	o_table = table[o_vec, ]
# 	set_A = which(o_table[,class_col]==class)
# 
# 	# define bins that cover the range of set A
# 	n = length(o_vec)
# 	bin_n = floor((n - set_A[1] - 1) / bs)
# 	seq_vec = seq(n-bin_n*bs, n, bs)
# 
# 	# take a matching set B
# 	set_B = c()
# 	for(i in 1:(length(seq_vec)-1)){
# 	  sub_table = o_table[(seq_vec[i] + 1):seq_vec[i + 1], ]
# 	  sub_A = which(sub_table[,class_col]==class)
# 	  if (length(sub_A) < bs/2){
# 	      sub_B = sample(which(sub_table[,class_col]!=class), length(sub_A))
# 	  } else {
# 	      sub_B = which(sub_table[,class_col]!=class)
# 	  }
# 	  set_B = c(set_B, sub_B + seq_vec[i])
# 	}
# 	## can also return o_table[c(setA, setB), ]
# 	## but this way order is perserved.
# 	i_vec = o_vec[c(set_A, set_B)]
# 	return(table[i_vec[order(i_vec)], ])
# }
# 
# tib_matched <- tibble()
# set.seed(123)
# for (i in 1:50) {
#   tmp <- matchSet(table = as.data.frame(tib), 
#                   class_col = "overlaps_LAD", 
#                   class = T, 
#                   order_on = "baseMean")
#   tmp <- as_tibble(tmp) %>%
#     mutate(match = i)
#   tib_matched <- bind_rows(tib_matched, tmp)
# }
# 
# # tib_matched <- matchSet(table = as.data.frame(tib), 
# #                         class_col = "overlaps_LAD", 
# #                         class = T, 
# #                         order_on = "baseMean")
# # tib_matched <- as_tibble(tib_matched)
# 
# # Check whether this worked
# tib %>% 
#   ggplot(aes(x = overlaps_LAD, y = log2(baseMean+1))) +
#   geom_violin(fill = "grey") +
#   theme_bw() +
#   theme(aspect.ratio = 2)
# 
# tib_matched %>% 
#   ggplot(aes(x = overlaps_LAD, y = log2(baseMean+1))) +
#   geom_violin(fill = "grey") +
#   theme_bw() +
#   theme(aspect.ratio = 2)
# 
# 
# 
# # Repeat enrichment plot
# tib_gather <- tib_matched %>%
#   dplyr::select(-starts_with("log")) %>%
#   gather(key, value, contains("overlaps")) %>%
#   mutate(sign = padj < 0.05)
# 
# # Group by LAD overlap or significance
# tib_diff <- tib_gather %>%
#   group_by(key, value) %>%
#   dplyr::summarise(fraction = mean(sign),
#             total = sum(sign),
#             background = length(sign))
# 
# #######################################
# ## Plot results
# 
# tib_diff %>%
#   ggplot(aes(x = key, y = fraction, fill = value)) +
#     geom_bar(stat = "identity", position = "dodge", col = "black") +
#     xlab("Knockout cells") +
#     ylab("Fraction differentially expressed") +
#     scale_fill_manual(values = c("grey", "black"), name = "Overlaps LAD") +
#     theme_bw() +
#     theme(aspect.ratio = 1)
# 
# tib_diff %>%
#   ggplot(aes(x = key, y = total, fill = value)) +
#     geom_bar(stat = "identity", position = "dodge", col = "black") +
#     xlab("Knockout cells") +
#     ylab("Total differentially expressed") +
#     scale_fill_manual(values = c("grey", "black"), name = "Overlaps LAD") +
#     theme_bw() +
#     theme(aspect.ratio = 1)

```

---

### Conclusion

Effects are mild, but make sense. 

---

### SessionInfo

```{r sessioninfo}

sessionInfo()

```
