---
title: "mt20230626_tf_abundances"
output: html_document
date: "2023-06-26"
---

Here I will load and pre-process all public RNA-seq and proteomics data for all TFs that I probed. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Load libraries
```{r}
library(edgeR)
library (EDASeq)
library(rtracklayer)
library(dplyr)
library(utils)
library(stringr)
library(tidyr)
library(readr)
library(ggplot2)
library(tidyverse)

ExonLength <- function(genes) {
  # Get the combined exon length for the genes
  tib <- as_tibble(genes) %>%
    filter(type == "exon") %>%
    mutate(gene_id = factor(gene_id, levels = unique(gene_id))) %>%
    group_by(gene_id) %>%
    dplyr::summarise(exon_number = n(),
              exon_length = sum(width))
  
  tib
}
```


## Process data

```{r cars}
# Load tf reporter data
# Import processed bc counts from the preprocessing step
cDNA_df <- read.csv("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf7124_stimulations/results/mt20231220_reporter_activity_filt_combined.csv", header = T)

cDNA_df$stimulation[is.na(cDNA_df$stimulation)] <- "no"

# Load in RNA-seq data
## Set the names of the TFs for which I want to retrieve the data
tfs <- cDNA_df %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  filter(neg_ctrls == "No", hPGK == "No", str_detect(tf, "RANDOM", negate = T)) %>%
  distinct(tf)

tfs <- tfs %>%
  mutate(tf2 = gsub(".*(::.*)", "\\1", tf)) %>%
  mutate(tf2 = gsub("::*", "", tf2)) %>%
  mutate(tf = gsub("::.*", "", tf)) %>%
  pivot_longer(cols = everything(.), names_to = "id", values_to = "tf") %>%
  distinct(tf) %>%
  mutate(tf = gsub(".[49|50]{2}", "", tf))

## Load mES data from Miguel
load('/DATA/usr/m.martinez.ara/GitLab/gurten/gurten/data/fc181121_epsure_tadec_rnaseq_expr_joshi.RData')

mES_rna <- tib_fpkm_joshi %>%
  mutate(twoi = (twoi_1 + twoi_2) / 2) %>%
  dplyr::select(symbol, twoi) %>%
  mutate(twoi = (twoi / sum(twoi))*10^6) %>% ## Transform from FPKM to TPM
  group_by(symbol) %>%
  arrange(desc(twoi)) %>%
  top_n(1) %>%
  ungroup()

mES_rna$symbol[mES_rna$symbol == "TRP53"] <- "TP53"

mES_rna <- mES_rna %>%
  dplyr::select("tf" = symbol, "TPM" = twoi) %>%
  #filter(mES_rna$tf %in% tfs$tf) %>%
  mutate(cell = "mES")

## Load data from protein atlas (TPM)
proteinatlas_rna <- read_tsv("/DATA/usr/m.trauernicht/data/RNA_seq/HPA_rna_celline_20230711.tsv") %>%
  dplyr::select('tf' = `Gene name`, TPM, "cell" = `Cell line`) %>%
  filter(cell %in% c("A-549", "U2OS", "Hep-G2", "HEK293", "K-562", "MCF-7", "HCT 116")) %>%
  mutate(cell = gsub(" ", "", cell)) %>%
  mutate(cell = gsub("-", "", cell)) %>%
  #filter(tf %in% tfs$tf)
  mutate(tf_cell = paste(tf, cell)) %>%
  group_by(tf_cell) %>%
  slice_max(n = 1, order_by = TPM) %>%
  ungroup() %>%
  dplyr::select(-tf_cell)

## diffTF-compatible output
library("org.Hs.eg.db")
proteinatlas_rna_export <- proteinatlas_rna %>%
  mutate(ENSEMBL = mapIds(org.Hs.eg.db, keys = proteinatlas_rna$tf, keytype = "SYMBOL", column="ENSEMBL")) %>%
  filter(!tf %in% c("BORCS8-MEF2B", "GAGE12J")) %>%
  dplyr::select(-tf) %>%
  na.omit() %>%
  distinct() %>%
  filter(ENSEMBL != "ENSG00000211689") %>%
  spread(key = "cell", value = "TPM")
  

## Load NPC RNA-seq data and calculate TPM
npc_rna <- read_delim("/shared/gcf/m.trauernicht/7086/genecounts.txt") %>%
  filter(gene_biotype == "protein_coding")

### Get gene lengths of all mm10 genes
genes_mouse <- import("/DATA/usr/m.trauernicht/data/genomes/mm10/gencode.vM10.annotation.gtf")
gene_length_mouse <- ExonLength(genes_mouse)
gene_length_mouse <- gene_length_mouse %>%
  mutate(gene_id = gsub("\\.[0-9]{1,2}$", "", gene_id)) %>%
  dplyr::select("ensembl_gene_id" = gene_id, exon_length)

npc_rna <- npc_rna %>%
  mutate(gene_length = end_position - start_position) %>%
  dplyr::select("tf" = external_gene_id, "NPC_r1" = `1_NPC_r1`, "NPC_r2" = `2_NPC_r2`, ensembl_gene_id, gene_length) %>%
  left_join(gene_length_mouse) %>%
  #left_join(transcript_lengths) %>%
  mutate(NPC_counts = (NPC_r1 + NPC_r2) / 2) %>%
  mutate(NPC_rpk = NPC_counts / (exon_length / 1000)) %>%
  mutate(pm = sum(NPC_rpk, na.rm = T) / 1e6) %>%
  mutate(TPM = NPC_rpk / pm) %>%
  mutate(cell = "NPC") %>%
  mutate(tf = ifelse(tf == "TRP53", "TP53", tf)) %>%
  distinct(tf, TPM, cell) %>%
  mutate(tf = toupper(tf))  %>%
  group_by(tf) %>%
  slice_max(n = 1, order_by = TPM) %>%
  ungroup()

## Load generated mESC RNA-seq data and calculate TPM
mES_rna_2 <- read_delim("/shared/gcf/m.trauernicht/7813/genecounts.txt") %>%
  filter(gene_biotype == "protein_coding")

mES_rna_2 <- mES_rna_2 %>%
  mutate(gene_length = end_position - start_position) %>%
  dplyr::select("tf" = external_gene_id, "mES_r3" = `3_mES_2i_LIF_r3`, "mES_r2" = `2_mES_2i_LIF_r2`,
                "mES_r1" = `1_mES_2i_LIF_r1`,
                ensembl_gene_id, gene_length) %>%
  left_join(gene_length_mouse) %>%
  #left_join(transcript_lengths) %>%
  mutate(mES_counts = (mES_r1 + mES_r2 + mES_r3) / 3) %>%
  mutate(mES_rpk = mES_counts / (exon_length / 1000)) %>%
  mutate(pm = sum(mES_rpk, na.rm = T) / 1e6) %>%
  mutate(TPM = mES_rpk / pm) %>%
  mutate(cell = "mESC_gen") %>%
  distinct(tf, TPM, cell) %>%
  mutate(tf = toupper(tf))  %>%
  mutate(tf = ifelse(tf == "TRP53", "TP53", tf)) %>%
  group_by(tf) %>%
  slice_max(n = 1, order_by = TPM) %>%
  ungroup()

## Load generated mESC RNA-seq data and calculate TPM
hepg2_gen <- read_delim("/shared/gcf/m.trauernicht/7535/genecounts.txt") %>%
  filter(gene_biotype == "protein_coding", chromosome_name != "MT")

### Get gene lengths of all mm10 genes
genes <- import("/DATA/usr/m.trauernicht/data/genomes/hg38/gencode.v38.annotation.gtf")
gene_length <- ExonLength(genes)
gene_length <- gene_length %>%
  mutate(gene_id = gsub("\\.[0-9]{1,2}$", "", gene_id)) %>%
  dplyr::select("ensembl_gene_id" = gene_id, exon_length)

hepg2_gen <- hepg2_gen %>%
  mutate(gene_length = end_position - start_position) %>%
  dplyr::select("tf" = external_gene_id, "HEPG2_r1" = `9_NT1_r1`, "HEPG2_r2" = `10_NT2_r1`,
                "HEPG2_r3" = `11_NT1_r2`, 
                ensembl_gene_id, gene_length) %>%
  left_join(gene_length) %>%
  #left_join(transcript_lengths) %>%
  mutate(HEPG2_counts = (HEPG2_r1 + HEPG2_r2 + HEPG2_r3) / 3) %>%
  mutate(HEPG2_rpk = HEPG2_counts / (exon_length / 1000)) %>%
  mutate(pm = sum(HEPG2_rpk, na.rm = T) / 1e6) %>%
  mutate(TPM = HEPG2_rpk / pm) %>%
  mutate(cell = "HEPG2_gen") %>%
  distinct(tf, TPM, cell) %>%
  mutate(tf = toupper(tf))  %>%
  group_by(tf) %>%
  slice_max(n = 1, order_by = TPM) %>%
  ungroup()


## Combine data from all cell types and change names back to original TF names
tf_rna <- rbind(mES_rna, mES_rna_2, proteinatlas_rna, npc_rna, hepg2_gen)
tf_rna$tf[tf_rna$tf == "AHR"] <- "AHR::ARNT"
tf_rna$tf[tf_rna$tf == "ARNT"] <- "AHR::ARNT"
tf_rna$tf[tf_rna$tf == "FOS"] <- "FOS::JUN"
tf_rna$tf[tf_rna$tf == "JUN"] <- "FOS::JUN"
tf_rna$tf[tf_rna$tf == "NFE2"] <- "MAF::NFE2"
tf_rna$tf[tf_rna$tf == "MAF"] <- "MAF::NFE2"
tf_rna$tf[tf_rna$tf == "NR1H4"] <- "NR1H4::RXRA"
tf_rna$tf[tf_rna$tf == "NR4A2"] <- "NR4A2::RXRA"
tf_rna$tf[tf_rna$tf == "PPARA"] <- "PPARA::RXRA"
tf_rna$tf[tf_rna$tf == "PPARG"] <- "PPARG::RXRA"
tf_rna$tf[tf_rna$tf == "SMAD2"] <- "SMAD2::SMAD3::SMAD4"
tf_rna$tf[tf_rna$tf == "SMAD3"] <- "SMAD2::SMAD3::SMAD4"
tf_rna$tf[tf_rna$tf == "STAT2"] <- "STAT1::STAT2"
tf_rna$tf[tf_rna$tf == "VDR"] <- "VDR::RXRA"

tf_rna2 <- tf_rna %>% filter(tf == "POU5F1")
tf_rna2$tf[tf_rna2$tf == "POU5F1"] <- "POU5F1::SOX2"
tf_rna <- rbind(tf_rna, tf_rna2)

tf_rna2 <- tf_rna %>% filter(tf == "SOX2")
tf_rna2$tf[tf_rna2$tf == "SOX2"] <- "POU5F1::SOX2"
tf_rna <- rbind(tf_rna, tf_rna2)

tf_rna2 <- tf_rna %>% filter(tf == "RARA")
tf_rna2$tf[tf_rna2$tf == "RARA"] <- "RARA:RXRA"
tf_rna <- rbind(tf_rna, tf_rna2)

tf_rna2 <- tf_rna %>% filter(tf == "SMAD4")
tf_rna2$tf[tf_rna2$tf == "SMAD4"] <- "SMAD2::SMAD3::SMAD4"
tf_rna <- rbind(tf_rna, tf_rna2)

tf_rna2 <- tf_rna %>% filter(tf == "STAT1")
tf_rna2$tf[tf_rna2$tf == "STAT1"] <- "STAT1::STAT2"
tf_rna <- rbind(tf_rna, tf_rna2)


# All unique cells and TFs
all_cells <- unique(tf_rna$cell)
all_tfs <- unique(tf_rna$tf)

# Create a complete data frame with all combinations of cell and tf
all_combinations <- expand.grid(cell = all_cells, tf = all_tfs)

# Join with the original data, filling missing values with 0
tf_rna_complete <- all_combinations %>%
  left_join(tf_rna, by = c("cell", "tf")) %>%
  replace_na(list(TPM = 0))

## For some heterodimeric TFs I set the TPM to the lowest of the two
tf_rna <- tf_rna_complete %>%
  ##mutate NA to 0
  mutate(TPM = ifelse(is.na(TPM), 0, TPM)) %>%
  mutate(TPM = ave(TPM, cell, tf, FUN = function(x) min (x, na.rm = T))) %>% 
  distinct()

tf_rna$TPM[is.na(tf_rna$TPM)] <- 0

## TPM normalization
tf_rna2 <- tf_rna %>%
  mutate(median_tpm = ave(TPM, cell, FUN = function(x) median(x, na.rm = T))) %>%
  mutate(max_median_tpm = max(median_tpm, na.rm = T)) %>%
  mutate(tpm_norm_factor = median_tpm/max_median_tpm) %>%
  mutate(TPM_norm = TPM/tpm_norm_factor)

# ggplot(tf_rna2,
#        aes(x = cell, y = log10(TPM))) +
#   geom_quasirandom()
  
tf_rna <- tf_rna2 %>%
  distinct(tf, cell, TPM, TPM_norm)

## TMM on TPM values
tf_rna_mat <- tf_rna %>%
  mutate(sum_cell = ave(TPM, cell, FUN = sum)) %>%
  mutate(mil_fac = 1e6/sum_cell) %>%
  mutate(pTPM = TPM * mil_fac) %>%
  distinct(tf, cell, pTPM) %>%
  spread(key = "cell", value = "pTPM") %>%
  column_to_rownames("tf") %>%
  replace(is.na(.), 0)

normfact <- data.frame(calcNormFactors(tf_rna_mat)) %>% 
  rownames_to_column("cell") %>% 
  dplyr::select(cell, "norm_factor" = calcNormFactors.tf_rna_mat.)

tf_rna <- tf_rna %>%
  left_join(normfact) %>%
  mutate(nTPM = TPM / norm_factor) %>%
  distinct(tf, cell, TPM, TPM_norm, nTPM)
```

## Correlate cells
```{r}
## mES cells
mES_rna_cor <- tf_rna %>%
  filter(cell %in% c("mES", "mESC_gen")) %>%
  distinct(tf, cell, nTPM) %>%
  pivot_wider(names_from = cell, values_from = nTPM)

mES_rna_cor2 <- mES_rna_cor %>% mutate(dif = mESC_gen / mES)
mES_rna_cor3 <- mES_rna_cor2 %>% mutate(tf_sel = ifelse(tf %in% cDNA_df$tf, "Yes", "No"))

ggplot() +
  geom_point(data = mES_rna_cor, aes(x = log10(mES), y = log10(mESC_gen)), 
             alpha = .1) +
  geom_point(data = mES_rna_cor %>%
         filter(tf %in% cDNA_df$tf), aes(x = log10(mES), y = log10(mESC_gen)),
         color = "red") +
  geom_abline(intercept = 0, slope = 1) +
  geom_smooth(method = "lm") +
  theme_pubr()

# Compute Pearson correlation
cor(mES_rna_cor$mES[mES_rna_cor$tf %in% cDNA_df$tf], mES_rna_cor$mESC_gen[mES_rna_cor$tf %in% cDNA_df$tf], use = "pairwise.complete.obs", method = "pearson")

## HEPG2 cells
hepg2_rna_cor <- tf_rna %>%
  filter(cell %in% c("HepG2", "HEPG2_gen")) %>%
  distinct(tf, cell, nTPM) %>%
  pivot_wider(names_from = cell, values_from = nTPM)

hepg2_rna_cor2 <- hepg2_rna_cor %>% mutate(dif = HEPG2_gen / HepG2)
hepg2_rna_cor3 <- hepg2_rna_cor2 %>% mutate(tf_sel = ifelse(tf %in% cDNA_df$tf, "Yes", "No"))

ggplot() +
  geom_point(data = hepg2_rna_cor, aes(x = log10(HepG2), y = log10(HEPG2_gen)), 
             alpha = .1) +
  geom_point(data = hepg2_rna_cor %>%
         filter(tf %in% cDNA_df$tf), aes(x = log10(HepG2), y = log10(HEPG2_gen)),
         color = "red") +
  geom_abline(intercept = 0, slope = 1) +
  geom_smooth(method = "lm") +
  theme_pubr()

# Compute Pearson correlation
cor(hepg2_rna_cor$HepG2[hepg2_rna_cor$tf %in% cDNA_df$tf], hepg2_rna_cor$HEPG2_gen[hepg2_rna_cor$tf %in% cDNA_df$tf], use = "pairwise.complete.obs")
```

## Export data
```{r}
tf_rna_export <- tf_rna %>%
  #filter(!cell %in% c("mES", "HepG2")) %>%
  #mutate(cell = ifelse(cell == "mESC_gen", "mES", cell)) %>%
  #mutate(cell = ifelse(cell == "HEPG2_gen", "HepG2", cell)) %>%
  distinct(tf, cell, nTPM)
  

write_tsv(tf_rna_export, "/DATA/usr/m.trauernicht/data/RNA_seq/rna_tpm_all_tfs.tsv")
```


## Process protein abundance data
```{r}
## Import proteomics data from depmap
proteomics_df <- read_csv("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/Proteomics_subsetted_NAsdropped.csv") %>%
  dplyr::select(-depmap_id, -lineage_1, -lineage_2, -lineage_3, -lineage_4, -lineage_5, -lineage_6) %>%
  pivot_longer(cols = -cell_line_display_name, names_to = "gene", values_to = "protein_abundance")

proteomics_df_long <- proteomics_df %>%
  mutate(gene = gsub(" (.*)", "", gene)) %>%
  filter(gene %in% tfs$tf) %>%
  mutate(min_abundance = min(protein_abundance, na.rm = T))

proteomics_df_long$cell_line_display_name[proteomics_df_long$cell_line_display_name == "HEPG2"] <- "HepG2"

proteomics_df_long$protein_abundance[is.na(proteomics_df_long$protein_abundance)] <- unique(proteomics_df_long$min_abundance)
  
proteomics_df_long <- proteomics_df_long %>%
  mutate(protein_abundance = ave(protein_abundance, gene, cell_line_display_name, FUN = function(x) mean(x, na.rm = T))) %>%
  dplyr::select(protein_abundance, "tf" = gene, "cell" = cell_line_display_name) %>%
  distinct() 

proteomics_df_long$tf[proteomics_df_long$tf == "AHR"] <- "AHR::ARNT"
proteomics_df_long$tf[proteomics_df_long$tf == "FOS"] <- "FOS::JUN"
proteomics_df_long$tf[proteomics_df_long$tf == "NFE2"] <- "MAF::NFE2"
proteomics_df_long$tf[proteomics_df_long$tf == "NR1H4"] <- "NR1H4::RXRA"
proteomics_df_long$tf[proteomics_df_long$tf == "NR4A2"] <- "NR4A2::RXRA"
proteomics_df_long$tf[proteomics_df_long$tf == "PPARA"] <- "PPARA::RXRA"
proteomics_df_long$tf[proteomics_df_long$tf == "PPARG"] <- "PPARG::RXRA"
proteomics_df_long$tf[proteomics_df_long$tf == "SMAD2"] <- "SMAD2::SMAD3::SMAD4"
proteomics_df_long$tf[proteomics_df_long$tf == "STAT1"] <- "STAT1::STAT2"
proteomics_df_long$tf[proteomics_df_long$tf == "VDR"] <- "VDR::RXRA"

write_tsv(proteomics_df_long, "/DATA/usr/m.trauernicht/data/RNA_seq/protein_abundance_all_tfs.tsv")
```

