---
title: "Barcode processing - TF reporter library - generation 2"
author: 
  - name: "Tea Filipovska & Max Trauernicht"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute - van Steensel lab"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
---

---

### Aim
18,000 TF reporters were transfected into HepG2 and K562 cells (in total 6 different conditions), sequencing data yielded barcode counts of these experiments. These counts will be processed in this script. 

---

## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}

```


### Libraries 

```{r setup, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(data.table)
library(plyr)
library(stringr)
library(ggpubr)
library(GGally)
library(vwr)
library(dplyr)
library(tibble)
library(plotly)
library(ggbeeswarm)
library(haven)
library(readr)
library(parallel)
library(RColorBrewer)
library(gridExtra)
library(LncFinder)
library(tidyr)
```


### Functions

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Custom functions
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}

ReadFasta<-function(file) {
   # Read the file line by line
   fasta<-readLines(file)
   # Identify header lines
   ind<-grep(">", fasta)
   # Identify the sequence lines
   s<-data.frame(ind=ind, from=ind+1, to=c((ind-1)[-1], length(fasta)))
   # Process sequence lines
   seqs<-rep(NA, length(ind))
   for(i in 1:length(ind)) {
      seqs[i]<-paste(fasta[s$from[i]:s$to[i]], collapse="")
   }
   # Create a data frame 
   DF<-data.frame(name=gsub(">", "", fasta[ind]), sequence=seqs)
   # Return the data frame as a result object from the function
   return(DF)
}

# From Fede:
# ggpairs custom functions
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {

  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y, "pairwise.complete.obs")
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)

  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }

  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())

  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]

  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }

  p <- p +
    theme(panel.background = element_rect(fill = corCol))

  return(p)
}
```


### Loading data

```{r data import, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Import barcode counts per experiment
## Do this for the 13 and 12 bp barcodes separately
bc_files_13 = list.files('/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/gcf6502/results_b12_d1/',
                       full.names=T, patter='*_barcode_counts.tsv')
bc_files_13 <- bc_files_13[-grep("pMT06", bc_files_13)]
bc_list_lib_13 <- lapply(bc_files_13, fread, header = FALSE)
names(bc_list_lib_13)<- gsub('.*//6502_[0-9]{2}_(.*?)_[CGAT]{6}-[CGAT]{8}.*_barcode_counts.tsv', 
                                    '\\1', 
                                    bc_files_13)

# this one is the average of 2 sequenced runs of the pDNA for this library
#pDNA_lib29 <- read_csv('/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6210_stimulation-3/results/mt20210312_reporter_activity_filt.csv')



bc_files = list.files('/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/gcf6301/results_pMT02_d1',
                       full.names=T, pattern='*_barcode_counts.tsv')
bc_list_2 <- lapply(bc_files, fread, header = FALSE)
names(bc_list_2)<- gsub('.*//6301_[0-9]{1,2}_(.*?)_[CGAT]{6}.*_barcode_counts.tsv', 
                                    '\\1', 
                                    bc_files)
bc_list_lib_13[21] <- bc_list_2[11]


pDNA_lib29 <- read_csv("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/pDNA_pMT02.csv")

# Import barcode annotation
bc_annotation_lib_29 <- read.csv("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/output/mt20191218_oligo_pool.csv", header = T) %>%
  mutate(barcode = gsub(".*([A-Z]{12})CATCGTCGCATCCAAGAG$", "\\1", seq.text),
         tf = gsub("(^.*)_.*_s-.*", "\\1", seq.name),
         spacing = gsub("^.*_s-(.*bp)_d.*", "\\1", seq.name),
         distance = gsub("^.*_d-(.*bp).*", "\\1", seq.name),
         barcode_number = gsub("^.*_bc-(.*)", "\\1", seq.name),
         promoter = gsub("^.*_(.*)_s-.*", "\\1", seq.name),
         background = gsub("^.*_bg-(.*)_.*", "\\1", seq.name),
         library = "1") %>%
  dplyr::select(-seq.name, -seq.text)

bc_annotation_lib_59 <- read.csv("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/gen-2/mt20210326_oligo_pool_gen2.csv", header = T) %>%
  mutate(barcode = gsub(".*([A-Z]{13})CATCGTCGCATCCAAGAG$", "\\1", seq.text),
         tf = gsub("(^.*)_s-.*", "\\1", seq.name),
         spacing = gsub("^.*_s-(.*bp)_d.*", "\\1", seq.name),
         distance = gsub("^.*_d-(.*bp).*", "\\1", seq.name),
         barcode_number = gsub("^.*_bc-(.*)", "\\1", seq.name),
         promoter = gsub("^.*_bg-[1-3]{1}_(.*)_bc.*", "\\1", seq.name),
         background = gsub("^.*_bg-(.*)_.*_.*", "\\1", seq.name),
         library = "2") %>%
  dplyr::select(-seq.name, -seq.text)

# Annotate positive controls
bc_annotation_lib_29$commercial_reporter <- "No"
bc_annotation_lib_59$commercial_reporter <- "No"
bc_annotation_lib_59$commercial_reporter[grep("TF-seq|romanov|promega", bc_annotation_lib_59$promoter)] <- "Yes"
bc_annotation_lib_59$tf[bc_annotation_lib_59$commercial_reporter == "Yes"] <- gsub("(.*)_TF-seq|romanov|promega", "\\1", bc_annotation_lib_59$tf[bc_annotation_lib_59$commercial_reporter == "Yes"])                                                                  
bc_annotation_lib_59$tf[bc_annotation_lib_59$commercial_reporter == "Yes"] <- gsub("_d-.*", "",bc_annotation_lib_59$tf[bc_annotation_lib_59$commercial_reporter == "Yes"])
                                                                                
bc_annotation_lib_59$spacing[bc_annotation_lib_59$commercial_reporter == "Yes"] <- ""
bc_annotation_lib_59$background[bc_annotation_lib_59$commercial_reporter == "Yes"] <- ""
bc_annotation_lib_59$promoter[bc_annotation_lib_59$commercial_reporter == "Yes"] <- gsub(".*bp_(.*)_bc.*", "\\1", bc_annotation_lib_59$promoter[bc_annotation_lib_59$commercial_reporter == "Yes"])

bc_annotation_lib_59$tf[grep("_", bc_annotation_lib_59$tf)]<- bc_annotation_lib_59$tf


bc_annotation_all <- rbind(bc_annotation_lib_29, bc_annotation_lib_59)
```


### Creating count data frames

```{r cluster_compare, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Generate one long data frame from the list of data frames and combine with barcode information
for (i in 1:length(bc_list_lib_13)) {
  if (i == 1) {
  bc_df <- data.frame(bc_list_lib_13[i])
  bc_df[3] <- names(bc_list_lib_13[i])
  names(bc_df) <- c("barcode", "starcode_counts", "sample")
  }
  else {
  bc_df_i <- data.frame(bc_list_lib_13[i])
  bc_df_i[3] <- names(bc_list_lib_13[i])
  names(bc_df_i) <- c("barcode", "starcode_counts", "sample")
  bc_df <- rbind(bc_df, bc_df_i)
  }
}
bc_df$sample[bc_df$sample == ""] <- "pMT02_pDNA"

bc_df <- merge(bc_df, bc_annotation_all, all = T)

bc_df$starcode_counts[is.na(bc_df$starcode_counts)] <- 0

bc_df_filt <- bc_df[!is.na(bc_df$tf),]
bc_df_unmatched <- bc_df[is.na(bc_df$tf),]

```

### Compare differently clustered pDNA data

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Make dataframe with columns for the together clustered pDNA data and individually clustered pDNA data
bc_df_pDNA <- bc_df_filt[grep("pDNA|pMT", bc_df_filt$sample),]
bc_df_pDNA <- bc_df_pDNA %>%
  spread(sample, starcode_counts)
bc_df_pDNA$R1_pDNA_TF_v2_08_04[is.na(bc_df_pDNA$R1_pDNA_TF_v2_08_04)] <- 0
bc_df_pDNA$R2_pDNA_TF_v2_08_04[is.na(bc_df_pDNA$R2_pDNA_TF_v2_08_04)] <- 0
bc_df_pDNA <- bc_df_pDNA %>%
  mutate(pMT09 = R1_pDNA_TF_v2_08_04 + R2_pDNA_TF_v2_08_04)

bc_df_pDNA$pMT02_pDNA[is.na(bc_df_pDNA$pMT02_pDNA)] <- 0
bc_df_pDNA <- bc_df_pDNA %>%
  mutate(pMT09_pMT02_combined = pMT09 + pMT02_pDNA)

# Compare library 1
#plot(bc_df_pDNA$pMT09_pMT02_combined[bc_df_pDNA$library == "1"], bc_df_pDNA$pMT02_pDNA[bc_df_pDNA$library == "1"])

# Compare library 2
#plot(bc_df_pDNA$pMT09_pMT02_combined[bc_df_pDNA$library == "2"], bc_df_pDNA$pMT09[bc_df_pDNA$library == "2"])
```


### Annotate data frame

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
## Annotate controls
# Annotate the mutated motif of each TF
bc_df$neg_ctrls <- "No"
bc_df$neg_ctrls[grep("neg", bc_df$tf)] <- "Yes"

# Annotate hPGK postive control
bc_df$hPGK <- "No"
bc_df$hPGK[grep("hPGK", bc_df$tf)] <- "Yes"

# Annotate enhancer controls
bc_df$native_enhancer <- "No"
bc_df$native_enhancer[grep("klf2", bc_df$tf)] <- "Yes"

# Annotate random promoter control
bc_df$rand_promoter <- "No"
bc_df$rand_promoter[grep("Random", bc_df$promoter)] <- "Yes"

# Add replicate and condition columns
bc_df$replicate <- gsub("(^R[1-3]).*", "\\1", bc_df$sample)
bc_df$condition <- gsub("^R[1-3]_(.*)", "\\1", bc_df$sample)

bc_df <- bc_df[!is.na(bc_df$sample),]

##remove all unmatched barcodes

bc_df_matched <- bc_df[!is.na(bc_df$tf),]
```


## Data quality plots

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message= FALSE}
##First compute reads per million to estimate the relative counts in their respective sample
bc_df <- bc_df %>%
  mutate(rpm = ave(starcode_counts, sample, FUN = function(x) (x+1) / sum(x) *1e6 ))


### a: total read counts per sample
plot_ly(bc_df[!is.na(bc_df$condition),] %>%
         mutate(sum_counts = ave(starcode_counts, sample, FUN = function(x) sum(x) / 36000)) %>%
         dplyr::select(sample, sum_counts) %>%
         unique(), 
        x = ~sum_counts, y = ~sample, type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Number of reads per barcode per sample",
         xaxis = list(title = "Expected number of reads per barcode per sample"),
         yaxis = list(title = "Sample"))


# I want to show the following:
## 1: Read distribution of matched barcodes vs. unmatched barcode
ggplot(bc_df[!is.na(bc_df$sample),] %>%
         filter(library =="1", neg_ctrls == "No", native_enhancer == "No", commercial_reporter == "No", hPGK == "No", rand_promoter == "No"), aes(x = tf, y = rpm)) +
  geom_violin() +
  theme_pubr(border = T) +
  ylim(0,2500) +
  theme(axis.text.y = element_text(size = 5), axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1)) +
  facet_wrap(~condition) +
  coord_flip() +
  ggtitle("rpm per TF per condition library 1")

ggplot(bc_df[!is.na(bc_df$sample),] %>%
         filter(library =="2", neg_ctrls == "No", native_enhancer == "No", commercial_reporter == "No", hPGK == "No", rand_promoter == "No"), aes(x = tf, y = rpm)) +
  geom_violin() +
  theme_pubr(border = T) +
  ylim(0,2500) +
  theme(axis.text.y = element_text(size = 5), axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1)) +
  facet_wrap(~condition) +
  coord_flip() +
  ggtitle("rpm per TF per condition library 2")


ggplot(bc_df, aes(x = rpm)) +
  geom_histogram(binwidth = 20) +
  theme_bw() +
  xlim(0,250)+
  ylim(0,6000)+
  facet_wrap(~sample)+
  theme(strip.background =element_rect(fill="#D6D5C9"))+
  ggtitle("<250 rpm ")

ggplot(bc_df[bc_df$rpm >= 500 & !is.na(bc_df$tf),], aes(x = rpm)) +
  geom_histogram(binwidth = 20) +
  theme_bw() +
  xlim(250,1000)+
  ylim(0,50)+
  facet_wrap(~sample)+
  theme(strip.background =element_rect(fill="#D6D5C9")) + 
  ggtitle(">500 rpm")

n_highly_expressed <- data.frame("sample" = unique(bc_df$sample),
                                 "n_bc" = "", stringsAsFactors = F)
for (i in unique(bc_df$sample)) {
  n_highly_expressed$n_bc[n_highly_expressed$sample == i] <- length(bc_df$barcode[bc_df$rpm > 500 & bc_df$sample == i])
}

plot_ly(n_highly_expressed, x = ~sample, y = ~as.numeric(n_bc), type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Highly expressed barcodes",
         xaxis = list(title = "Sample"),
         yaxis = list(title = "Number of barcodes with > 500 rpm"))
```


```{r read count 2, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message= FALSE}
## 2: How many barcodes can I find back at which cutoff? + What is the percentage of barcode reads that match the design at which cutoff?
## Identify the unmapped fraction

bc_fraction <- expand.grid(unique(bc_df$sample[bc_df$sample != "pMT09_pMT02_combined"]), unique(bc_df$library))
bc_fraction <- na.omit(bc_fraction)
names(bc_fraction) <- c("sample", "lib")
bc_fraction$bcs_found <- ""
rpm_cutoff <- data.frame("cutoff" = seq(0.0001,15,0.5))
bc_fraction <- merge(bc_fraction, rpm_cutoff)

for (i in unique(bc_fraction$cutoff)) {
  for (j in unique(bc_df$sample)) {
    for (k in unique(bc_fraction$lib)) {
    bc_fraction$bcs_found[bc_fraction$cutoff == i & bc_fraction$sample == j & bc_fraction$lib == k] <- 
      nrow(bc_df_filt[bc_df_filt$rpm >= i & bc_df_filt$sample == j & bc_df_filt$library == k,])/
      nrow(bc_annotation_all[bc_annotation_all$library == k,]) *100
    }
  }
}


## How many reads match to designed barcodes?
bc_reads <- expand.grid(unique(bc_df$sample[bc_df$sample != "pMT09_pMT02_combined"]), unique(bc_df$library))
bc_reads <- na.omit(bc_reads)
names(bc_reads) <- c("sample", "lib")
bc_reads$bc_reads <- ""
bc_reads <- merge(bc_reads, rpm_cutoff)

bc_df$match <- "Yes"
bc_df$match[is.na(bc_df$library)] <- "No"

for (i in unique(bc_reads$cutoff)) {
  for (j in unique(bc_df$sample)) {
    for (k in unique(bc_reads$lib)) {
      bc_reads$bc_reads[bc_reads$cutoff == i & bc_reads$sample == j & bc_reads$lib == k] <- 
        sum(na.omit(bc_df_filt$rpm[bc_df_filt$rpm >= i & bc_df_filt$sample == j & bc_df_filt$library == k]))/
      sum(bc_df$rpm[bc_df$rpm >= i & bc_df$sample == j & (bc_df$library == k | bc_df$match == "No")]) *100
    }
  }
}



bc_fraction <- merge(bc_fraction, bc_reads)
bc_fraction$bcs_found <- as.numeric(bc_fraction$bcs_found)
bc_fraction$bc_reads <- as.numeric(bc_fraction$bc_reads)

#c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B")
# Plot to evaluate data quality per cutoff
for (i in unique(bc_fraction$lib)) {
p <- ggplot(bc_fraction[bc_fraction$lib == i,]) +
  geom_point(aes(x = cutoff, y = bcs_found), color = '#1B998B', size = 1) +
  geom_line(aes(x = cutoff, y = bcs_found), color = '#1B998B', size = 1) +
  geom_point(aes(x = cutoff, y = bc_reads), color = 'black', size = 1) +
  geom_line(aes(x = cutoff, y = bc_reads), color = 'black', size = 1) +
  theme_bw()+
  xlab("rpm cutoff")+
  ylab("total barcodes (green) and matched barcode reads (black) detected (%)")+
  facet_wrap(~sample)+
  theme(strip.background =element_rect(fill="#D6D5C9"))+
  ggtitle(paste("library", i, sep = ": "))
print(p)
}


## 3: What is the correlation of the cDNA bc counts with the pDNA bc counts? 

pDNA_1 <- data.frame("pDNA" = bc_df$rpm[bc_df$condition == "pMT02_pDNA"],
                   "barcode"= bc_df$barcode[bc_df$condition == "pMT02_pDNA"], stringsAsFactors = F)
bc_df_1 <- merge(pDNA_1, bc_df, all = T)
bc_df_1 <- bc_df_1[!is.na(bc_df_1$sample),]
bc_df_1 <- na.omit(bc_df_1)

ggplot(bc_df_1, aes(x = pDNA, y = rpm)) +
  geom_bin2d(bins = 100)+
  xlim(0,400) +
  ylim(0,400)+
  theme_bw()+
  facet_wrap(~sample)+
  ggtitle("Correlation of cDNA barcode counts with pDNA barcode counts lib 1")

pDNA_2 <- data.frame("pDNA" = bc_df$rpm[bc_df$condition == "pDNA_TF_v2_08_04"],
                   "barcode"= bc_df$barcode[bc_df$condition == "pDNA_TF_v2_08_04"], stringsAsFactors = F)
bc_df_2 <- merge(pDNA_2, bc_df, all = T)
bc_df_2 <- bc_df_2[!is.na(bc_df_2$sample),]
bc_df_2 <- na.omit(bc_df_2)


ggplot(bc_df_2, aes(x = pDNA, y = rpm)) +
  geom_bin2d(bins = 100)+
  xlim(0,400) +
  ylim(0,400)+
  theme_bw()+
  facet_wrap(~sample)+
  ggtitle("Correlation of cDNA barcode counts with pDNA barcode counts lib 2")


cor <- data.frame("sample" = unique(bc_df_2$sample), "cor" = "", stringsAsFactors = F)

for (i in unique(bc_df_2$sample)) {
  x <- bc_df_2[bc_df_2$sample == i,] %>% dplyr::select(pDNA, rpm, barcode) %>% unique()
  cor$cor[cor$sample == i] <- cor(x$rpm, x$pDNA, "pairwise.complete.obs")
}


bc_df <- merge(bc_df, cor, by = "sample", all = T)

## 4: Correlation plots of the replicates
## Combine replicates of normalized data in 3 different columns

bc_df_rep <- bc_df[!is.na(bc_df$tf),] %>% 
  dplyr::select(replicate, rpm, barcode, condition, library, neg_ctrls) %>% 
  unique() %>%
  spread(replicate, rpm)

colors <- c("#2D3047", "#1B998B")

for (i in unique(bc_df_rep$library)) {
p <- ggscatter(bc_df_rep[bc_df_rep$library == i,], x = "R1", y = "R2",
   add = "reg.line",
   color = "neg_ctrls",
   size = 1,
   add.params = list(color = "blue", fill = "lightgray"), title = paste("rep1 vs rep2 - library: ", i, sep = ""),
   conf.int = TRUE, ylab = "rep2", xlab = "rep1") + 
  stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
  geom_abline(linetype = "dashed") +
  xlim(0,4000) + ylim(0,4000) +
  scale_color_manual(values = colors)+
  facet_wrap(~condition)
print(p)
}

for (i in unique(bc_df_rep$library)) {
p <- ggscatter(bc_df_rep[bc_df_rep$library == i,], x = "R1", y = "R3",
   add = "reg.line",
   color = "neg_ctrls",
   size = 1,
   add.params = list(color = "blue", fill = "lightgray"), title = paste("rep1 vs rep3 - library: ", i, sep = ""),
   conf.int = TRUE, ylab = "rep3", xlab = "rep1") + 
  stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
  geom_abline(linetype = "dashed") +
  xlim(0,4000) + ylim(0,4000) +
  scale_color_manual(values = colors)+
  facet_wrap(~condition)
print(p)
}

for (i in unique(bc_df_rep$library)) {
p <- ggscatter(bc_df_rep[bc_df_rep$library == i,], x = "R2", y = "R3",
   add = "reg.line",
   color = "neg_ctrls",
   size = 1,
   add.params = list(color = "blue", fill = "lightgray"), title = paste("rep2 vs rep3 - library: ", i, sep = ""),
   conf.int = TRUE, ylab = "rep3", xlab = "rep2") + 
  stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
  geom_abline(linetype = "dashed") +
  xlim(0,4000) + ylim(0,4000) +
  scale_color_manual(values = colors)+
  facet_wrap(~condition)
print(p)
}



# Plot pDNA distribution vs. cutoff

pDNA_fraction <- bc_df[grep("pDNA", bc_df$condition),]
pDNA_fraction <- pDNA_fraction[!is.na(pDNA_fraction$tf),]
pDNA_fraction <- pDNA_fraction %>% dplyr::select(barcode, replicate, rpm)
pDNA_fraction <- dcast(pDNA_fraction, barcode ~ replicate)
pDNA_fraction <- pDNA_fraction %>% 
  mutate(pDNA = (R1 + R2) / 2) %>%
  dplyr::select(-R1, -R2)
pDNA_fr <- data.frame("cutoff" = seq(0,100, 1),
                      "bcs_missing" = "", stringsAsFactors=FALSE)
for (i in pDNA_fr$cutoff) {
  pDNA_fr$bcs_missing[pDNA_fr$cutoff == i] <- length(pDNA_fraction$barcode) -
    length(pDNA_fraction$barcode[pDNA_fraction$pDNA >= i])
}

ggplot(pDNA_fr) +
  geom_point(aes(x = cutoff, y = as.numeric(bcs_missing)), color = '#1B998B', size = 1) +
  geom_line(aes(x = cutoff, y = as.numeric(bcs_missing)), color = '#1B998B', size = 1) +
  theme_bw()+
  xlab("rpm cutoff")+
  ylab("barcodes excluded from analysis (18,000 in total)") +
  geom_vline(xintercept = 10, linetype = "dashed", color = "black") +
  ggtitle("library 2")
```

---

### Sample filtering

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Remove samples with low dynamic range
for (i in unique(bc_df$sample)) {
  bc_df$n_bc[bc_df$sample == i] <- length(bc_df$barcode[bc_df$rpm > 500 & bc_df$sample == i])
}
bc_df <- bc_df[bc_df$n_bc >= 150 | bc_df$condition == "pDNA_TF_v2_08_04" | bc_df$condition == "pMT02_pDNA",] %>% dplyr::select(-n_bc)

# Remove samples with high pDNA contamination
bc_df <- bc_df[bc_df$cor <= 0.85 | bc_df$condition == "pDNA_TF_v2_08_04" | bc_df$condition == "pMT02_pDNA",] %>% dplyr::select(-cor)

# Remove all non-matching reads
bc_df <- bc_df[!is.na(bc_df$tf),]


```

---


### Normalization of barcode counts:  
Divide cDNA barcode counts through pDNA barcode counts to get activity

```{r normalization, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Add pDNA data as separate column
pDNA_lib59 <- bc_df[grep("pDNA_TF", bc_df$condition),] %>% 
  filter(library == "2") %>%
  dplyr::select(barcode, rpm, sample, condition) %>% unique() %>%
  spread(sample, rpm) %>% 
  mutate(pDNA_counts_rpm = (R1_pDNA_TF_v2_08_04 + R2_pDNA_TF_v2_08_04) / 2) %>%
  dplyr::select(barcode, pDNA_counts_rpm)

##merge pDNA data from both libraires 

pDNA_lib29 <- bc_df[bc_df$sample == "pMT02_pDNA",] %>%
  filter(library == "1") %>% 
  dplyr::select(barcode, 'pDNA_counts_rpm' = rpm) %>%
  unique()

pDNA <- rbind(pDNA_lib59, pDNA_lib29)

bc_df <- merge(pDNA, bc_df)

# Compute activity by dividing cDNA bc counts through pDNA bc counts
bc_df$activity <- bc_df$rpm / bc_df$pDNA_counts_rpm


# Remove all activities that were computed with low pDNA bc counts, this keeps the noise level in the data low
#bc_df$activity[bc_df$pDNA_counts_rpm <= 1] <- NA
bc_df$noisy <- "not noisy"
bc_df$noisy[bc_df$pDNA_counts_rpm <= 10] <- "potentially noisy"
```


---

## Calculate mean activity - filter out outlier barcodes 

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# First identify and remove outlier barcodes - this removes the noise created by faulty barcode clustering etc. 
## Calculate mean and SD for each reporter


bc_df_cDNA <- bc_df[-grep("pDNA", bc_df$sample),]
bc_df_cDNA$reporter_id <- paste(bc_df_cDNA$tf, bc_df_cDNA$spacing, 
                                bc_df_cDNA$distance, bc_df_cDNA$promoter,
                                bc_df_cDNA$background, sep = "_")

##calculate mean and sd
bc_df_cDNA$mean_activity <- ave(bc_df_cDNA$activity, bc_df_cDNA$reporter_id, 
                                bc_df_cDNA$sample, FUN =
                                  function(x) mean(x))
bc_df_cDNA$sd_activity <- ave(bc_df_cDNA$activity, bc_df_cDNA$reporter_id, 
                              bc_df_cDNA$sample,  FUN =
                                  function(x) sd(x))


## Remove data points that are 2xSD away from the mean
bc_df_cDNA$upper_activity <- bc_df_cDNA$mean_activity + (2 * bc_df_cDNA$sd_activity)
bc_df_cDNA$lower_activity <- bc_df_cDNA$mean_activity - (2 * bc_df_cDNA$sd_activity)

bc_df_cDNA$low_outlier <- bc_df_cDNA$activity - bc_df_cDNA$lower_activity
bc_df_cDNA$high_outlier <- bc_df_cDNA$upper_activity - bc_df_cDNA$activity



bc_df_cDNA_notnoisy <- bc_df_cDNA[bc_df_cDNA$noisy == "not noisy",]

## Plot effect of highest outlier bc 
outlier <- bc_df_cDNA_notnoisy[bc_df_cDNA_notnoisy$reporter_id == "Trp53_10bp_10bp_hBGm_1",] %>% 
  dplyr::select(barcode_number, activity) %>%
  unique()

ggplot(outlier, aes(x = barcode_number, y = activity)) +
  geom_quasirandom() + theme_bw() +
  scale_x_discrete(breaks = seq(1, 8, by = 1)) +
  xlab("barcode #") + 
  labs(title = "highest outlier after filtering: Trp53_10bp_10bp_hBGm_1")


## Plot effect of lowest outlier bc 
outlier2 <- bc_df_cDNA_notnoisy[bc_df_cDNA_notnoisy$reporter_id == "Klf4_10bp_10bp_minP_3",] %>% 
  dplyr::select(barcode_number, activity) %>%
  unique()

ggplot(outlier2, aes(x = barcode_number, y = activity)) +
  geom_quasirandom() + theme_bw() +
  scale_x_discrete(breaks = seq(1, 8, by = 1)) +
  xlab("barcode #") + 
  labs(title = "lowest outlier after filtering: Klf4_10bp_10bp_minP_3")




##To do: check in the pDNA data if the barcode is mismatched to another promoter or if something else is strange


## Choose arbitrary cutoff to get rid of most extreme outliers and keep all positive controls
bc_df_cDNA_filt <- bc_df_cDNA_notnoisy[bc_df_cDNA_notnoisy$low_outlier > -0.3 & bc_df_cDNA_notnoisy$high_outlier > -2,]


## Recalculate mean and sd
bc_df_cDNA_filt$mean_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, 
                                bc_df_cDNA_filt$condition, FUN =
                                  function(x) mean(x))
bc_df_cDNA_filt$sd_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, 
                              bc_df_cDNA_filt$condition,  FUN =
                                  function(x) sd(x))

bc_df_cDNA_filt <- bc_df_cDNA_filt[!is.na(bc_df_cDNA_filt$barcode),]

### re-plot effect of highest outlier bc 
outlier3 <- bc_df_cDNA_filt[bc_df_cDNA_notnoisy$reporter_id == "Trp53_10bp_10bp_hBGm_1",] %>% 
  dplyr::select(barcode_number, activity) %>%
  unique()

ggplot(outlier3, aes(x = barcode_number, y = activity)) +
  geom_quasirandom() + theme_bw() +
  scale_x_discrete(breaks = seq(1, 8, by = 1)) +
  xlab("barcode #") + 
  ylim(0,20) +
  labs(title = "highest outlier after filtering and outlier removal: Trp53_10bp_10bp_hBGm_1")


     
```


### Scaling data 

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE}
# Scale data to 1 for negative controls

bc_df_cDNA_filt_neg <- bc_df_cDNA_filt %>%
  dplyr::select(sample, activity, reporter_id, neg_ctrls, rand_promoter) %>%
  filter(neg_ctrls == "Yes", rand_promoter == "Yes") %>%
  unique() %>% 
  group_by(sample) %>%
  mutate(activity = mean(activity)) %>%
  ungroup() %>%
  dplyr::select(-reporter_id) %>%
  unique() %>%
  dplyr::select("background_activity" = activity, sample)

bc_df_cDNA_filt <- merge(bc_df_cDNA_filt, bc_df_cDNA_filt_neg)
bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  mutate(activity = activity / background_activity)

bc_df_cDNA_filt$mean_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, 
                                bc_df_cDNA_filt$sample, FUN =
                                  function(x) mean(x))
bc_df_cDNA_filt <- bc_df_cDNA_filt[!is.na(bc_df_cDNA_filt$sample),]

```

```{r}
## Re-label R3 K562

bc_df_cDNA_filt$sample[bc_df_cDNA_filt$sample == "R3_K562_DMSO"] <- "K562_DMSO_stressed"
bc_df_cDNA_filt$condition[bc_df_cDNA_filt$sample =="K562_DMSO_stressed"] <- "K562_stressed"   

```

---

## Calculate correlations between technical replicates for only library59

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE}
## Combine replicates in 8 different columns

bc_df_rep_59 <- bc_df_cDNA_filt[bc_df_cDNA_filt$native_enhancer == "No" &
                          bc_df_cDNA_filt$rand_promoter == "No",] %>% 
  dplyr::select(barcode_number, activity, sample, reporter_id, library)
bc_df_rep_59 <- bc_df_rep_59[!is.na(bc_df_rep_59$sample),]
bc_df_rep_59 <- bc_df_rep_59[bc_df_rep_59$library == "2",]


rep1 <- bc_df_rep_59[bc_df_rep_59$barcode_number == 1,] %>% dplyr::select(-barcode_number)
rep2 <- bc_df_rep_59[bc_df_rep_59$barcode_number == 2,] %>% dplyr::select(-barcode_number)
rep3 <- bc_df_rep_59[bc_df_rep_59$barcode_number == 3,] %>% dplyr::select(-barcode_number)
rep4 <- bc_df_rep_59[bc_df_rep_59$barcode_number == 4,] %>% dplyr::select(-barcode_number)
rep5 <- bc_df_rep_59[bc_df_rep_59$barcode_number == 5,] %>% dplyr::select(-barcode_number)


setnames(rep1, "activity", "bc1")
rep1 <- rep1[!is.na(rep1$sample),]
setnames(rep2, "activity", "bc2")
rep2 <- rep2[!is.na(rep2$sample),]
setnames(rep3, "activity", "bc3")
rep3 <- rep3[!is.na(rep3$sample),]
setnames(rep4, "activity", "bc4")
rep4 <- rep4[!is.na(rep4$sample),]
setnames(rep5, "activity", "bc5")
rep5 <- rep5[!is.na(rep5$sample),]



bc_df_rep_59 <- merge(rep1, rep2, all = T)
bc_df_rep_59 <- bc_df_rep_59[rowSums(is.na(bc_df_rep_59)) != ncol(bc_df_rep_59), ]
bc_df_rep_59 <- merge(bc_df_rep_59, rep3, all = T)
bc_df_rep_59 <- merge(bc_df_rep_59, rep4, all = T)
bc_df_rep_59 <- bc_df_rep_59[rowSums(is.na(bc_df_rep_59)) != ncol(bc_df_rep_59), ]
bc_df_rep_59 <- merge(bc_df_rep_59, rep5, all = T)




# Correlation matrix plot
n <- sample(1:nrow(bc_df_rep_59), 5000)
boundaries <- seq(from = 0.8, by = 0.05, length.out = 4)
plt <- ggpairs(bc_df_rep_59 %>% dplyr::select(bc1, bc2, bc3, bc4, bc5),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.1, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red") +
                   theme_bw()}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red") +
                   theme_bw()})) +
  ggtitle("Correlation Between Technial Replicates Library 2") +
  theme(text = element_text(size = 20)) +
  xlab("Reporter activity") +
  ylab("Reporter activity") + 
  theme_light()

print(plt)
```

## Calculate correlations between technical replicates for only library29
```{r}
## Combine replicates in 8 different columns

bc_df_rep_29 <- bc_df_cDNA_filt[bc_df_cDNA_filt$native_enhancer == "No" &
                          bc_df_cDNA_filt$rand_promoter == "No",] %>% 
  dplyr::select(barcode_number, activity, sample, reporter_id, library)
bc_df_rep_29 <- bc_df_rep_29[!is.na(bc_df_rep_29$sample),]
bc_df_rep_29 <- bc_df_rep_29[bc_df_rep_29$library == "1",]


rep1 <- bc_df_rep_29[bc_df_rep_29$barcode_number == 1,] %>% dplyr::select(-barcode_number)
rep2 <- bc_df_rep_29[bc_df_rep_29$barcode_number == 2,] %>% dplyr::select(-barcode_number)
rep3 <- bc_df_rep_29[bc_df_rep_29$barcode_number == 3,] %>% dplyr::select(-barcode_number)
rep4 <- bc_df_rep_29[bc_df_rep_29$barcode_number == 4,] %>% dplyr::select(-barcode_number)
rep5 <- bc_df_rep_29[bc_df_rep_29$barcode_number == 5,] %>% dplyr::select(-barcode_number)
rep6 <- bc_df_rep_29[bc_df_rep_29$barcode_number == 6,] %>% dplyr::select(-barcode_number)
rep7 <- bc_df_rep_29[bc_df_rep_29$barcode_number == 7,] %>% dplyr::select(-barcode_number)
rep8 <- bc_df_rep_29[bc_df_rep_29$barcode_number == 8,] %>% dplyr::select(-barcode_number)

setnames(rep1, "activity", "bc1")
rep1 <- rep1[!is.na(rep1$sample),]
setnames(rep2, "activity", "bc2")
rep2 <- rep2[!is.na(rep2$sample),]
setnames(rep3, "activity", "bc3")
rep3 <- rep3[!is.na(rep3$sample),]
setnames(rep4, "activity", "bc4")
rep4 <- rep4[!is.na(rep4$sample),]
setnames(rep5, "activity", "bc5")
rep5 <- rep5[!is.na(rep5$sample),]
setnames(rep6, "activity", "bc6")
rep6 <- rep6[!is.na(rep6$sample),]
setnames(rep7, "activity", "bc7")
rep7 <- rep7[!is.na(rep7$sample),]
setnames(rep8, "activity", "bc8")
rep8 <- rep8[!is.na(rep8$sample),]


bc_df_rep_29 <- merge(rep1, rep2, all = T)
bc_df_rep_29 <- bc_df_rep_29[rowSums(is.na(bc_df_rep_29)) != ncol(bc_df_rep_29), ]
bc_df_rep_29 <- merge(bc_df_rep_29, rep3, all = T)
bc_df_rep_29 <- merge(bc_df_rep_29, rep4, all = T)
bc_df_rep_29 <- bc_df_rep_29[rowSums(is.na(bc_df_rep_29)) != ncol(bc_df_rep_29), ]
bc_df_rep_29 <- merge(bc_df_rep_29, rep5, all = T)
bc_df_rep_29 <- merge(bc_df_rep_29, rep6, all = T)
bc_df_rep_29 <- bc_df_rep_29[rowSums(is.na(bc_df_rep_29)) != ncol(bc_df_rep_29), ]
bc_df_rep_29 <- merge(bc_df_rep_29, rep7, all = T)
bc_df_rep_29 <- merge(bc_df_rep_29, rep8, all = T)
bc_df_rep_29 <- bc_df_rep_29[rowSums(is.na(bc_df_rep_29)) != ncol(bc_df_rep_29), ]


# Correlation matrix plot
n <- sample(1:nrow(bc_df_rep_29), 5000)
boundaries <- seq(from = 0.8, by = 0.05, length.out = 4)
plt <- ggpairs(bc_df_rep_29 %>% dplyr::select(bc1, bc2, bc3, bc4, bc5, bc6, bc7, bc8),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.1, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red") +
                   theme_bw()}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red") +
                   theme_bw()})) +
  ggtitle("Correlation Between Technial Replicates Library 1") +
  theme(text = element_text(size = 20)) +
  xlab("Reporter activity") +
  ylab("Reporter activity") + 
  theme_light()

print(plt)
```


## Correlation between biological replicates for each condition

```{r correlations_4, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE}
# Correlation plots of the replicates of library 2
## Combine replicates of normalized data in 3 different columns

bc_df_rep <- bc_df_cDNA_filt %>% 
  dplyr::select(replicate, mean_activity, reporter_id, condition, neg_ctrls, library) %>% 
  unique() %>%
  spread(replicate, mean_activity)

colors <- c("#2D3047", "#1B998B")

for (i in unique(bc_df_rep$library)) {
  p <- ggscatter(bc_df_rep[bc_df_rep$library == i,], x = "R1", y = "R2",
                 add = "reg.line",
                 color = "neg_ctrls",
                 size = 0.5,
                 add.params = list(color = "blue", fill = "lightgray"), 
                 title = paste("rep1 vs rep2 - library:", i),
                 conf.int = TRUE, ylab = "rep2", xlab = "rep1") + 
    stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
    geom_abline(linetype = "dashed") +
    xlim(0,50) + ylim(0,50) +
    scale_color_manual(values = colors) +
    facet_wrap(~condition)
  print(p)
}

for (i in unique(bc_df_rep$library)) {
  p <- ggscatter(bc_df_rep[bc_df_rep$library == i,], x = "R1", y = "R3",
                 add = "reg.line",
                 color = "neg_ctrls",
                 size = 0.5,
                 add.params = list(color = "blue", fill = "lightgray"), 
                 title = paste("rep1 vs rep3 - library:", i),
                 conf.int = TRUE, ylab = "rep3", xlab = "rep1") + 
    stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
    geom_abline(linetype = "dashed") +
    xlim(0,50) + ylim(0,50) +
    scale_color_manual(values = colors) +
    facet_wrap(~condition)
  print(p)
}

for (i in unique(bc_df_rep$library)) {
  p <- ggscatter(bc_df_rep[bc_df_rep$library == i,], x = "R2", y = "R3",
                 add = "reg.line",
                 color = "neg_ctrls",
                 size = 0.5,
                 add.params = list(color = "blue", fill = "lightgray"), 
                 title = paste("rep2 vs rep3 - library:", i),
                 conf.int = TRUE, ylab = "rep3", xlab = "rep2") + 
    stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
    geom_abline(linetype = "dashed") +
    xlim(0,50) + ylim(0,50) +
    scale_color_manual(values = colors) +
    facet_wrap(~condition)
  print(p)
}


```


```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE}
# Mean of the three replicates
bc_df_cDNA_filt$reporter_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$condition, FUN = function(x) mean(x))
bc_df_cDNA_filt$reporter_activity_sd <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$condition, FUN = function(x) sd(x))

```


```{r data export, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE}
# Polish export dataframe
bc_df_cDNA_filt <- bc_df_cDNA_filt %>% 
  dplyr::select(-upper_activity, -lower_activity, 
         -high_outlier, -low_outlier,) %>% 
  setnames(old = c("mean_activity", "sd_activity"), 
           new = c("replicate_activity", "replicate_activity_sd")) %>% 
  mutate(log_activity = log2(activity),
         log_reporter_activity = log2(reporter_activity))
bc_df_cDNA_filt$condition <- gsub("^X", "", bc_df_cDNA_filt$condition)


# Export bc_df for cDNA analysis
filename <- SetFileName("_reporter_activity_filt", "tf")
setwd("/DATA/usr/t.filipovska/projects/TFreporter_assay_gen2/analysis")
write.csv(bc_df_cDNA_filt, file = paste(filename,".csv", sep = ""), row.names = F)
```

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

