---
title: "genomic P53 response element scoring"
author: 
  - name: "Max Trauernicht"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute - van Steensel lab"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
---

---

### Aim
I want to score the affinity genomic P53 response elements using the same algorithm that was used to create P53 reporters. I then want to conclude if I can say anything about P53 gene regulation with the P53 reporters that I generated or not.

---

## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}
```

### Libraries

```{r setup, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(GenomicRanges)
library(rtracklayer)
library(BSgenome.Hsapiens.UCSC.hg38)
library(GenomicFeatures)
library(BSgenome)
library(spgs)
library(dplyr)
library(ggplot2)
library(reshape)
library(ggbeeswarm)
library(biomaRt)
library(readr)
library(stringr)
library(maditr)
library(phylotools)
library(plyr)
library(tidyr)
library(ggpubr)
library(SummarizedExperiment)
```


### Custom functions

```{r, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}

ReadFasta<-function(file) {
   # Read the file line by line
   fasta<-readLines(file)
   # Identify header lines
   ind<-grep(">", fasta)
   # Identify the sequence lines
   s<-data.frame(ind=ind, from=ind+1, to=c((ind-1)[-1], length(fasta)))
   # Process sequence lines
   seqs<-rep(NA, length(ind))
   for(i in 1:length(ind)) {
      seqs[i]<-paste(fasta[s$from[i]:s$to[i]], collapse="")
   }
   # Create a data frame 
   DF<-data.frame(name=gsub(">", "", fasta[ind]), sequence=seqs)
   # Return the data frame as a result object from the function
   return(DF)
}


# Function to load PWM matrix
get_pwm_feature_matrix <- function(motif_meta_fn, fimo_fn, db = 2) {

  # validate args
  valid_dbs <- 1:2
  if(!db %in% valid_dbs)
    stop('Invalid db (database version). Please use db=1 (maintained for backward compatibility only) or db=2')

  # db=1 is maintained for backward compatibility only
  if(db == 1) {

    # read in motif metadata
    motif_meta    <- read.csv(motif_meta_fn)

    # check whether motif metadata contain essential annotations
    if(!all(c('PWM.ID', 'Cognate.TF') %in% colnames(motif_meta))) {
      message('The motif metadata file does not contain the essential columns PWM.ID and Cognate.TF')
    }

    motif_minimal <- motif_meta[, c('PWM.ID', 'Cognate.TF')]

    # load fimo output --> extract motif id, sequence id and p-value
    df <- read.table(fimo_fn)
    df <- df[, c(1, 2, 7)]

    colnames(df) <- c('PWM.ID', 'seqid', 'pval')

    # add TF id
    df <- merge(df, motif_minimal, by = 'PWM.ID')

    # group motif hits by sequence id
    l <- split(df, df[['seqid']])

    # multiple PWM and multiple hits possible. Reduce hits to one per TF, keeping best p-val only
    l <- lapply(l, function(x) {
      x_by_tf <- split(x, x[['Cognate.TF']], drop = TRUE)
      x_by_tf <- lapply(x_by_tf, function(y) y[which.min(y$pval), ])
      do.call('rbind', x_by_tf)
    })

    # initialize feature matrix
    n_tf          <- motif_minimal[['Cognate.TF']] %>%
      unique %>%
      length
    n_seq         <- length(l)
    pwm           <- matrix(1, nrow = n_seq, ncol = n_tf)
    colnames(pwm) <- (motif_minimal[['Cognate.TF']] %>% unique)

    # replace :: from names of composite motifs
    colnames(pwm) <- str_replace_all(colnames(pwm), '::', '_')

    # fill in feature matrix
    for(i in 1 : n_seq) {
      pwm[i, l[[i]][['Cognate.TF']]] <- l[[i]]$pval
    }

    # -log10 transform
    pwm           <- -1 * log10(pwm)

    # coerce to tib and return
    tib_fimo <- as_data_frame(pwm) %>%
      mutate(id = names(l))
      dplyr::select(id, everything())

  }

  # db = 2 (default)
  else {

    # load metadata
    tib_meta    <- read_csv(motif_meta_fn) %>%
      # extract tf symbol from motif id (Cognate_TF unsafe, it can be empty) and replace :: occurrences
      mutate(tf_symbol = str_remove(ID, '_[0-9]*'),
             tf_symbol = str_replace(tf_symbol, '::', '_')) %>%
      dplyr::select(motif_id = `PWM ID`, tf_symbol)

    # load fimo results
    tib_fimo <- read_tsv(fimo_fn) %>%
      # extract motif id, sequence id and p-value
      dplyr::select(motif_id, sequence_name, pval = `p-value`)

    # add tf symbol to fimo results
    tib_fimo <- tib_fimo %>%
      left_join(tib_meta, by = 'motif_id') %>%
      # remove hits with missing motif id (composite pwms)
      filter(!is.na(tf_symbol))

    # select best hit for each motif and sequence
    tib_fimo <- tib_fimo %>%
      group_by(sequence_name, tf_symbol) %>%
      dplyr::slice(which.min(pval)) %>%
      ungroup()

    # spread into feature matrix
    tib_fimo <- tib_fimo %>%
      mutate(pval = -1 * log10(pval)) %>%
      dplyr::select(-motif_id) %>%
      spread(key = tf_symbol, value = pval, fill = 0, drop = TRUE) %>%
      # perform cosmetics on the id
      mutate(id = sequence_name) %>%
      dplyr::select(-c(sequence_name)) %>%
      dplyr::select(id, everything())

  }

  return(tib_fimo)

}


theme_classic_lines <- function() {
  theme_pubr(border = T, legend = "top") +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.1),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_classic_lines_45 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 45) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.1),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_classic_lines_90 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 90) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.1),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_set(theme_classic_lines())

colors_diverse <- c("#264653", "#9AC1AE", "#5D987B", "#f2cc8f", "#e76f51")

ggplot_custom <- function(...) ggplot2::ggplot(...) + 
  scale_color_manual(values = colors_diverse) + 
  scale_fill_manual(values = colors_diverse)
```


### Run FIMO script
```{bash run fimo, eval = FALSE}
# fimo /DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/motifs/selected_motifs.meme  /DATA/usr/t.filipovska/projects/ATAC-seq_TF-footprinting/analysis/peak_sequences.fa
```


---


## Load FIMO results

```{r}
tib_fimo <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/fimo_out/fimo.tsv") %>%
  # extract motif id, sequence id and p-value
  dplyr::select(motif_id, sequence_name, pval = `p-value`, 'motif_start' = start, 'motif_stop' = stop) 
  # # select a p-value cutoff?
  # filter(pval < 1e-06)
  

# select best hit for each motif and sequence
tib_fimo <- tib_fimo %>%
      group_by(sequence_name, motif_id) %>%
      dplyr::slice(which.min(pval)) %>%
      ungroup()

# spread into feature matrix
tib_fimo <- tib_fimo %>%
      mutate(pval = -1 * log10(pval)) %>%
      spread(key = motif_id, value = pval, fill = 0, drop = TRUE) %>%
      # perform cosmetics on the id
      mutate(id = sequence_name) %>%
      dplyr::select(-c(sequence_name)) %>%
      dplyr::select(id, everything())

# convert to binary
#tib_fimo <- tib_fimo %>% mutate_if(is.numeric, function(x) ifelse((x==0 | is.na(x)),0,1)) 
```

---

## Load reads in peaks data and combine with FIMO scoring

```{r}
# Load data
exp <- readRDS("/DATA/usr/t.filipovska/projects/ATAC-seq_TF-footprinting/analysis/rds/experiment_6420_20210610.rds")

# Do the filtering as before
thres_count <- 10
thres_num <- 2

seqinfo <- seqinfo(SummarizedExperiment::rowRanges(exp))
seqinfo <- keepStandardChromosomes(seqinfo, "Homo_sapiens")
seqnames <- seqnames(seqinfo)
seqnames <- seqnames[seqnames != "chrM"]

row_keep <- rowSums(SummarizedExperiment::assay(exp)) > thres_count
row_keep <- row_keep & seqnames(rowRanges(exp)) %in% seqnames
row_keep <- row_keep & rowMeans(assay(exp)) < 1000



col_keep <- exp$replicate!="R3"
col_keep <- col_keep & exp$basename != "11_HepG2_Rifampicin_R2_AAGAGGCA-ATCTACAC_S3"

#col_keep <- exp$basename 
# col_keep <- col_keep & exp$tag != "NANOG"

exp <- exp[row_keep,col_keep]


exp_counts <- data.frame(assay(exp))
exp_counts <- cbind(exp_counts, data.frame(exp@rowRanges[,1])) %>%
  dplyr::select(-seqnames, -width, -start, -end, -strand) %>%
  dplyr::select('id' = name, everything())

# Calculate mean per condition
names(exp_counts) <- c("id", gsub("_R[1-3]", "", names(exp_counts[,-1])))
exp_counts_long <- melt(exp_counts, id.vars = "id", value.name = "reads_in_peaks", variable.name = "condition")
exp_counts_long$condition <- as.character(exp_counts_long$condition)
exp_counts_long <- exp_counts_long %>% 
  group_by(id, condition) %>% summarise_all(funs(mean))

for (i in unique(exp_counts_long$condition)) {
  exp_counts_long$rpm[exp_counts_long$condition == i] <- (exp_counts_long$reads_in_peaks[exp_counts_long$condition == i] + 1) / # Adds a pseudocount of 1
    sum(exp_counts_long$reads_in_peaks[exp_counts_long$condition == i]) *1e6
}

# Combine with fimo data
exp_counts_m <- merge(exp_counts_long, tib_fimo, all = T)
exp_counts_m[is.na(exp_counts_m)] <- 0

# Plot accessibility in peaks with and without LXR (NR1H2) motifs
ggplot(exp_counts_m %>%
         filter(WT1 > 4),
       aes(x = condition, y = rpm)) +
  geom_quasirandom() +
  coord_flip()
```

---

## Make bed files for tornado plots

```{r}
# Make .bed file from all motifs
tib_fimo <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/fimo_out/fimo.tsv") %>%
  # extract motif id, sequence id and p-value
  dplyr::select(motif_id, sequence_name, pval = `p-value`, 'motif_start' = start, 'motif_stop' = stop, matched_sequence) 
  # # select a p-value cutoff?
  # filter(pval < 1e-06)
  
# only select exact k-mer matches, so I only take direct comparisons with my reporters
kmer_motifs <- read_csv("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/gen-2/tf_motifs_all.csv") %>%
  filter(str_detect(TF, "neg", negate = T))
kmer_motifs2 <- read_csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/parameter_files/mt20191119_TFs_list_updated.csv") %>%
  dplyr::select(TF, 'motif' = Motif)
kmer_motifs <- rbind(kmer_motifs, kmer_motifs2)

tib_fimo_match <- tib_fimo[tib_fimo$matched_sequence %in% kmer_motifs$motif,]


# select best hit for each motif and sequence
tib_fimo <- tib_fimo %>%
      group_by(sequence_name, motif_id) %>%
      dplyr::slice(which.min(pval)) %>%
      ungroup()

# set stringent p-value cutoff

## Either top 1000 hits per motif
tib_fimo <- tib_fimo %>%
  group_by(motif_id) %>%
  dplyr::top_n(-1000, pval)

# ## Or based on p-value
# tib_fimo <- tib_fimo %>%
#   filter(pval < 1e-6)


peaks <- data.frame(exp@rowRanges[,1]) %>%
  dplyr::select('sequence_name' = name, everything())

tib_fimo_2 <- merge(peaks, tib_fimo, by = "sequence_name")

tib_fimo_2 <- tib_fimo_2 %>%
  mutate(motif_start2 = start + motif_start - 100,
         motif_stop2 = start + motif_stop + 100, 
         width = motif_stop2 - motif_start2)

tib_fimo_2 <- tib_fimo_2 %>%
  dplyr::select(motif_id, 'chr' = seqnames, 'start' = motif_start2, 'stop' = motif_stop2, 'name' = sequence_name, width, strand)

tib_fimo_2 <- tib_fimo_2 %>%
  mutate(motif_id = toupper(motif_id)) %>%
  unique()

# for (i in unique(tib_fimo_2$motif_id)) {
#   write.table(tib_fimo_2[tib_fimo_2$motif_id == i,] %>% dplyr::select(-motif_id),
#               file = paste("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/motifs/exact_reporter_match_only/", i, ".bed", sep = ""),
#               row.names = F,
#               col.names = F,
#               quote = F,
#               sep = '\t')
# }

```

---

## Extract only peaks in genes that are close to upregulated genes
```{r}
colors_diverse <- c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51")
# Load in upregulated genes
fxr_genes_up <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/HepG2_RNAseq/DE_genes_FXR_CDCA.csv", header = T)
lxr_genes_up <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/HepG2_RNAseq/DE_genes_LXR_T0901317.csv", header = T)
ppar_genes_up <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/HepG2_RNAseq/DE_genes_PPARa_Wy-14643.csv", header = T)
sxr_genes_up <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/HepG2_RNAseq/DE_genes_SXR_Rifampicin.csv", header = T)

# Import all TSSs 
tss_annotated <- read.delim2("/DATA/usr/m.trauernicht/projects/SuRE_deep_scan_trp53_gr/data/p53_affinity_scoring/gencode.v27-fantom-selection.txt", header = F, col.names = c("chr", "TSS_position", "transcript_id", "strand", "gene_id", "expr_avg", "n_tissues")) %>%
  mutate(gene_id = gsub(".[0-9]+$", "", gene_id))

# Keep only TSSs with highest expression
tss_annotated$expr_avg <- as.numeric(tss_annotated$expr_avg)
tss_annotated <- tss_annotated %>%
  arrange(gene_id, -expr_avg) %>%
  filter(duplicated(gene_id) == FALSE)

# Add 100 kb up and down of TSS of changed genes
tss_annotated <- tss_annotated %>%
  mutate(start = TSS_position - 100000,
         end = TSS_position + 100000)
tss_annotated$start[tss_annotated$start< 0] <- 0


ensembl <- useMart("ensembl")
mart <- useDataset("hsapiens_gene_ensembl", mart = ensembl)
genes <- tss_annotated$gene_id
G_list <- getBM(filters= "ensembl_gene_id", attributes= c("ensembl_gene_id","hgnc_symbol"),values=genes,mart= mart)
tss_annotated <- merge(tss_annotated,G_list,by.x="gene_id",by.y="ensembl_gene_id")
tss_annotated_fxr <- tss_annotated[tss_annotated$hgnc_symbol %in% fxr_genes_up$TFs,]
tss_annotated_lxr <- tss_annotated[tss_annotated$hgnc_symbol %in% lxr_genes_up$TFs,]
tss_annotated_ppar <- tss_annotated[tss_annotated$hgnc_symbol %in% ppar_genes_up$TFs,]
tss_annotated_sxr <- tss_annotated[tss_annotated$hgnc_symbol %in% sxr_genes_up$TFs,]

# Identify overlaps between regions in tss positions and peaks with motifs
## FXR
fxr_peaks <- as.data.frame(read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq/motifs/NR1H4::RXRA.bed",header = FALSE, sep="\t",stringsAsFactors=FALSE, quote=""))

gr_fxr_tss = with(tss_annotated_fxr, GRanges(chr, IRanges(start=start, end=end)))

gr_fxr_peaks = with(fxr_peaks, GRanges(V1, IRanges(start=V2, end=V3)))

hits_fxr <- data.frame(mergeByOverlaps(gr_fxr_tss, gr_fxr_peaks)) %>%
  dplyr::select("seqnames" = gr_fxr_peaks.seqnames, "start" = gr_fxr_peaks.start, "end" = gr_fxr_peaks.end) %>%
  mutate(tf = "NR1H4::RXRA")

## LXR
lxr_peaks <- as.data.frame(read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq/motifs/NR1H2.bed",header = FALSE, sep="\t",stringsAsFactors=FALSE, quote=""))

gr_lxr_tss = with(tss_annotated_lxr, GRanges(chr, IRanges(start=start, end=end)))

gr_lxr_peaks = with(lxr_peaks, GRanges(V1, IRanges(start=V2, end=V3)))

hits_lxr <- data.frame(mergeByOverlaps(gr_lxr_tss, gr_lxr_peaks)) %>%
  dplyr::select("seqnames" = gr_lxr_peaks.seqnames, "start" = gr_lxr_peaks.start, "end" = gr_lxr_peaks.end) %>%
  mutate(tf = "NR1H2")

## PPAR
ppar_peaks <- as.data.frame(read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq/motifs/PPARA::RXRA.bed",header = FALSE, sep="\t",stringsAsFactors=FALSE, quote=""))

gr_ppar_tss = with(tss_annotated_ppar, GRanges(chr, IRanges(start=start, end=end)))

gr_ppar_peaks = with(ppar_peaks, GRanges(V1, IRanges(start=V2, end=V3)))

hits_ppar <- data.frame(mergeByOverlaps(gr_ppar_tss, gr_ppar_peaks)) %>%
  dplyr::select("seqnames" = gr_ppar_peaks.seqnames, "start" = gr_ppar_peaks.start, "end" = gr_ppar_peaks.end) %>%
  mutate(tf = "PPARA::RXRA")

## SXR
sxr_peaks <- as.data.frame(read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq/motifs/NR1I2.bed",header = FALSE, sep="\t",stringsAsFactors=FALSE, quote=""))

gr_sxr_tss = with(tss_annotated_sxr, GRanges(chr, IRanges(start=start, end=end)))

gr_sxr_peaks = with(sxr_peaks, GRanges(V1, IRanges(start=V2, end=V3)))

hits_sxr <- data.frame(mergeByOverlaps(gr_sxr_tss, gr_sxr_peaks)) %>%
  dplyr::select("seqnames" = gr_sxr_peaks.seqnames, "start" = gr_sxr_peaks.start, "end" = gr_sxr_peaks.end) %>%
  mutate(tf = "NR1I2")

## Combine all and export
hits_all <- rbind(hits_fxr, hits_lxr, hits_ppar, hits_sxr)

for (i in unique(hits_all$tf)) {
  write.table(hits_all[hits_all$tf == i,] %>% dplyr::select(-tf),
              file = paste("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq/motifs/rna_seq_hits/", i, ".bed", sep = ""),
              row.names = F,
              col.names = F,
              quote = F,
              sep = '\t')
}
```

---

## AUC computation of accessibility around motifs in peaks

```{r}
hepg2_gata1 <- read.delim2("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq/motifs/matrix_out/GATA1.bed", sep = "\t", header = T) %>%
  dplyr::select(matches("HepG2"))
names(hepg2_gata1) <- 1:ncol(hepg2_gata1)
hepg2_gata1 <- hepg2_gata1 %>%
  melt(variable.name = "bin", value.name = "accessbility") %>%
  mutate(accessbility = ave(accessbility, bin, FUN = function(x) mean(x))) %>%
  unique()

hepg2_gata1$bin <- as.numeric(hepg2_gata1$bin)

auc_hepg2 <- sum(hepg2_gata1$accessbility[hepg2_gata1$bin > 50 & hepg2_gata1$bin < 350])


k562_gata1 <- read.delim("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq/motifs/matrix_out/GATA1_2.bed", sep = "\t", header = T) %>%
  dplyr::select(matches("K562"))
names(k562_gata1) <- 1:ncol(k562_gata1)
k562_gata1 <- k562_gata1 %>%
  melt(variable.name = "bin", value.name = "accessbility") %>%
  mutate(accessbility = ave(accessbility, bin, FUN = function(x) mean(x))) %>%
  unique()
k562_gata1$bin <- as.numeric(k562_gata1$bin)
auc_k562 <- sum(k562_gata1$accessbility[k562_gata1$bin > 50 & k562_gata1$bin < 350])

auc_k562/auc_hepg2
```



## DiffTF based TF activity computation
```{r}
## DiffTF was run, final output is imported:
difftf_activity <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/diffTF/output_rna_top1000_inverse/FINAL_OUTPUT/extension100/HepG2vsK562.top1000.summary.tsv") %>%
  dplyr::select(TF, 'diffTF_activity' = weighted_meanDifference, 'classification' = classification_q0.001_final)
difftf_activity$classification[difftf_activity$TF == "POU5F1::SOX2"] <- difftf_activity$classification[difftf_activity$TF == "POU5F1"]
difftf_activity$classification[difftf_activity$TF == "RAR:RXR"] <- difftf_activity$classification[difftf_activity$TF == "RARA"]

difftf_activity2 <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/diffTF/output_rna_allTFBS/FINAL_OUTPUT/extension100/HepG2vsK562.top1000.summary.tsv") %>%
  dplyr::select(TF, 'diffTF_activity2' = weighted_meanDifference, 'classification2' = classification_q0.001_final)

difftf_activity3 <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/diffTF/output_rna_exact_match_only/FINAL_OUTPUT/extension100/HepG2vsK562.top1000.summary.tsv") %>%
  dplyr::select(TF, 'diffTF_activity2' = weighted_meanDifference)

# Check if there is a difference between using only direct k-mers as input or using (more degenerate) motif matches
diff <- merge(difftf_activity, difftf_activity3, by = "TF")
ggplot(diff, aes(x = diffTF_activity, y = diffTF_activity2)) +
  geom_point() +
  xlab("DiffTF activity, input = regions similar to motif") +
  ylab("DiffTF activity, input = same sequence as in reporter") +
  geom_abline(linetype = "dashed")



# Import mean reporter activities per TF
reporter_activity_all <- read_csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/diffTF/tf_reporter_activities_ernst.csv")
reporter_activity_all_mean <- reporter_activity_all %>%
  pivot_wider(names_from = "condition", values_from = c("ernst_activity", "reporter_activity_minP")) %>%
  mutate(dif_ernst = log2(ernst_activity_K562 / ernst_activity_HepG2),
         dif_reporter = log2(reporter_activity_minP_K562 / reporter_activity_minP_HepG2)) %>%
  dplyr::select(dif_ernst, dif_reporter, 'TF' = tf) %>%
  unique()
reporter_activity_all_mean$TF <- gsub("::.*", "_", reporter_activity_all_mean$TF)
reporter_activity_all_mean$TF[reporter_activity_all_mean$TF == "POU5F1_"] <- "POU5F1::SOX2"
reporter_activity_all_mean$TF[reporter_activity_all_mean$TF == "RARA:RXRA"] <- "RAR:RXR"
reporter_activity_all_mean$TF[reporter_activity_all_mean$TF == "ETS1"] <- "ETS2"
reporter_activity_all_mean$TF[reporter_activity_all_mean$TF == "NFATc1"] <- "NFATC1"
reporter_activity_all_mean$TF[reporter_activity_all_mean$TF == "TCF3"] <- "TCF7L2"
reporter_activity_all_mean$TF <- gsub("_", "", reporter_activity_all_mean$TF)


activity_comp <- merge(difftf_activity, reporter_activity_all_mean, by = "TF", all = T) %>%
  na.omit()


# Add RNA-seq data from HepG2 and K562
rnaseq_matrix <- read_tsv("/DATA/scratch/usr/t.v.schaik/proj/3D_nucleus/results/ts210514_RNAseq_Ki67_depletion/results_pe/Star/count_table.tsv") %>%
  dplyr::select('ENSEMBL' = 1, 'HepG2_r1' = 2, 'HepG2_r2' = 8, 'K562_r1' = 7, 'K562_r2' = 13) %>%
  mutate(ENSEMBL = gsub(".[0-9]+$", "", ENSEMBL))

# Add gene names
ensembl <- useMart("ensembl")
mart <- useDataset("hsapiens_gene_ensembl", mart = ensembl)
gene <- rnaseq_matrix$ENSEMBL
G_list <- getBM(filters= "ensembl_gene_id", attributes= c("ensembl_gene_id","hgnc_symbol"),values=gene,mart= mart)
rnaseq_matrix <- merge(rnaseq_matrix,G_list,by.x="ENSEMBL",by.y="ensembl_gene_id")

rnaseq_matrix <- rnaseq_matrix[rnaseq_matrix$hgnc_symbol != "",]

# Filter for genes with no counts
tib_nocounts <- rnaseq_matrix %>%
  dplyr::select(-hgnc_symbol) %>%
  gather(key, value, -ENSEMBL) %>%
  group_by(ENSEMBL) %>%
  dplyr::summarise(count = sum(value > 2) > 1)
idx_nocounts <- which(tib_nocounts$count)

genes <- gene[idx_nocounts]
rnaseq_matrix <- rnaseq_matrix[idx_nocounts, ]

# index <- read.csv("/DATA/usr/m.trauernicht/software/diffTF/src/TF_Gene_TranslationTables/HOCOMOCO_v11/translationTable_hg38.csv", sep = " ")
# rnaseq_matrix <- rnaseq_matrix[rnaseq_matrix$ENSEMBL %in% index$ENSEMBL,]
# rnaseq_matrix <- rnaseq_matrix %>%
#   dplyr::select(-hgnc_symbol)
# write_tsv(rnaseq_matrix, "/DATA/usr/m.trauernicht/data/RNA_seq/hepG2_k562_counts.tsv")

activity_comp <- activity_comp %>%
  mutate(diffTF_activity_norm = (diffTF_activity-mean(diffTF_activity))/sd(diffTF_activity),
         reporter_activity_norm = (reporter_activity-mean(reporter_activity))/sd(reporter_activity))

activity_comp2 <- merge(activity_comp, rnaseq_matrix, by.x = 'TF', by.y = 'hgnc_symbol')
  
# Import class of TF
tf_class <- read_csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/garcia_tf-properties.csv") %>%
  dplyr::select(TF, chromatin_regulation_mode)

activity_comp <- merge(activity_comp, tf_class, all = T)
activity_comp$chromatin_regulation_mode[activity_comp$TF == "POU5F1::SOX2"] <- activity_comp$chromatin_regulation_mode[activity_comp$TF == "POU5F1"]
activity_comp$chromatin_regulation_mode[activity_comp$TF == "RAR:RXR"] <- activity_comp$chromatin_regulation_mode[activity_comp$TF == "RARA"]
activity_comp <- activity_comp %>% na.omit()


ggplot(activity_comp, aes(x = reporter_activity_norm, y = diffTF_activity_norm)) +
  geom_point(aes( color = chromatin_regulation_mode)) +
  geom_abline(linetype = "dashed") +
  geom_smooth(method = "lm", color = "grey", fill = "lightgrey") +
  xlab("Differential Reporter Activity (K562/HepG2)") +
  ylab("Differential Accessibility - diffTF (K562/HepG2)") +
  scale_color_manual(values = c("grey",colors_diverse[c(2,5,4)])) +
  facet_wrap(~classification)

for (i in unique(activity_comp$classification)) {
  print(paste("cor =", cor.test(activity_comp$reporter_activity_norm[activity_comp$classification == i], activity_comp$diffTF_activity_norm[activity_comp$classification == i])$estimate,
              "| p-value =", cor.test(activity_comp$reporter_activity_norm[activity_comp$classification == i], activity_comp$diffTF_activity_norm[activity_comp$classification == i])$p.value,
              "| classification =", i))
}


# Which of the TFs are correlating and which not?
activity_comp3 <- activity_comp2 %>%
  mutate(reporter_activity_norm = abs(reporter_activity_norm),
         diffTF_activity_norm = abs(diffTF_activity_norm),
         dif = reporter_activity_norm - diffTF_activity_norm,
         dif_rna = K562_r1 - HepG2_r1) %>%
  dplyr::select(dif, classification, TF, dif_rna) %>%
  unique()

ggplot(activity_comp3 %>% filter(classification != "not-expressed"), 
       aes(x = classification, y = dif)) +
  geom_quasirandom()

activity_comp3 <- activity_comp2 %>%
  mutate(dif_rna = K562_r1 - HepG2_r1) %>%
  dplyr::select(reporter_activity_norm, diffTF_activity_norm, classification, TF, dif_rna) %>%
  unique()

ggplot(activity_comp3 %>% filter(classification != "not-expressed"), 
       aes(x = reporter_activity_norm, y = dif_rna)) +
  geom_point()+
  geom_abline(linetype = "dashed") +
  geom_smooth(method = "lm", color = "grey", fill = "lightgrey") 

ggplot(activity_comp3 %>% filter(classification != "not-expressed"), 
       aes(x = diffTF_activity_norm, y = dif_rna)) +
  geom_point()  +
  geom_abline(linetype = "dashed") +
  geom_smooth(method = "lm", color = "grey", fill = "lightgrey") 



# The same for DMSO vs CDCA
difftf_activity_cdca <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/diffTF/output_fxr/FINAL_OUTPUT/extension100/DMSOvsCDCA.top1000.summary.tsv") %>%
  dplyr::select(TF, 'diffTF_activity' = weighted_meanDifference)

reporter_activity_dif <- reporter_activity %>%
  mutate(reporter_activity = log2(HepG2_CDCA / HepG2_DMSO)) %>%
  dplyr::select(TF, reporter_activity)

activity_comp_cdca <- merge(difftf_activity_cdca, reporter_activity_dif)


ggplot(activity_comp_cdca, aes(x = diffTF_activity, y = reporter_activity)) +
  geom_point(color = ifelse(activity_comp_cdca$reporter_activity > 1, "red", "black")) +
  geom_abline() +
  geom_smooth(method = "lm") +
  xlab("Differential Accessibility - diffTF (CDCA/DMSO)") +
  ylab("Differential Reporter Activity (CDCA/DMSO)") 

```

---

## ChromVAR differential TF activity computation
```{r}
library(chromVAR)
library(motifmatchr)
library(Matrix)
library(SummarizedExperiment)
library(BiocParallel)
set.seed(1252)

```







### Session Info

```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

