---
title: "Systematic comparison of TF reporter activity with ATAC-seq-inferred TF activity"
author: 
  - name: "Max Trauernicht"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute - van Steensel lab"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
---

---

### Aim
I performed reporter assays in HepG2 and K562 cells to directly measure the activity of 86 TFs. In parallel, I computed TF activities indirectly from accessibility using ATAC-seq. I now want to systematically compare these two measures in this script. 
The first part of this script describes how the candidate regions for the accessibility analysis were selected. 

---

## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}
```

### Libraries

```{r setup, out.width= "80%", fig.align= "center", warning= FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(chromVAR)
library(motifmatchr)
library(Matrix)
library(SummarizedExperiment)
library(BiocParallel)
library(BSgenome.Hsapiens.UCSC.hg38)
library(universalmotif)
library(ggseqlogo)
library(motifStack)
library(PWMEnrich)
set.seed(2023)
```


### Custom functions

```{r, out.width= "80%", fig.align= "center"}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}

ReadFasta<-function(file) {
   # Read the file line by line
   fasta<-readLines(file)
   # Identify header lines
   ind<-grep(">", fasta)
   # Identify the sequence lines
   s<-data.frame(ind=ind, from=ind+1, to=c((ind-1)[-1], length(fasta)))
   # Process sequence lines
   seqs<-rep(NA, length(ind))
   for(i in 1:length(ind)) {
      seqs[i]<-paste(fasta[s$from[i]:s$to[i]], collapse="")
   }
   # Create a data frame 
   DF<-data.frame(name=gsub(">", "", fasta[ind]), sequence=seqs)
   # Return the data frame as a result object from the function
   return(DF)
}


# Function to load PWM matrix
get_pwm_feature_matrix <- function(motif_meta_fn, fimo_fn, db = 2) {

  # validate args
  valid_dbs <- 1:2
  if(!db %in% valid_dbs)
    stop('Invalid db (database version). Please use db=1 (maintained for backward compatibility only) or db=2')

  # db=1 is maintained for backward compatibility only
  if(db == 1) {

    # read in motif metadata
    motif_meta    <- read.csv(motif_meta_fn)

    # check whether motif metadata contain essential annotations
    if(!all(c('PWM.ID', 'Cognate.TF') %in% colnames(motif_meta))) {
      message('The motif metadata file does not contain the essential columns PWM.ID and Cognate.TF')
    }

    motif_minimal <- motif_meta[, c('PWM.ID', 'Cognate.TF')]

    # load fimo output --> extract motif id, sequence id and p-value
    df <- read.table(fimo_fn)
    df <- df[, c(1, 2, 7)]

    colnames(df) <- c('PWM.ID', 'seqid', 'pval')

    # add TF id
    df <- merge(df, motif_minimal, by = 'PWM.ID')

    # group motif hits by sequence id
    l <- split(df, df[['seqid']])

    # multiple PWM and multiple hits possible. Reduce hits to one per TF, keeping best p-val only
    l <- lapply(l, function(x) {
      x_by_tf <- split(x, x[['Cognate.TF']], drop = TRUE)
      x_by_tf <- lapply(x_by_tf, function(y) y[which.min(y$pval), ])
      do.call('rbind', x_by_tf)
    })

    # initialize feature matrix
    n_tf          <- motif_minimal[['Cognate.TF']] %>%
      unique %>%
      length
    n_seq         <- length(l)
    pwm           <- matrix(1, nrow = n_seq, ncol = n_tf)
    colnames(pwm) <- (motif_minimal[['Cognate.TF']] %>% unique)

    # replace :: from names of composite motifs
    colnames(pwm) <- str_replace_all(colnames(pwm), '::', '_')

    # fill in feature matrix
    for(i in 1 : n_seq) {
      pwm[i, l[[i]][['Cognate.TF']]] <- l[[i]]$pval
    }

    # -log10 transform
    pwm           <- -1 * log10(pwm)

    # coerce to tib and return
    tib_fimo <- as_data_frame(pwm) %>%
      mutate(id = names(l))
      dplyr::select(id, everything())

  }

  # db = 2 (default)
  else {

    # load metadata
    tib_meta    <- read_csv(motif_meta_fn) %>%
      # extract tf symbol from motif id (Cognate_TF unsafe, it can be empty) and replace :: occurrences
      mutate(tf_symbol = str_remove(ID, '_[0-9]*'),
             tf_symbol = str_replace(tf_symbol, '::', '_')) %>%
      dplyr::select(motif_id = `PWM ID`, tf_symbol)

    # load fimo results
    tib_fimo <- read_tsv(fimo_fn) %>%
      # extract motif id, sequence id and p-value
      dplyr::select(motif_id, sequence_name, pval = `p-value`)

    # add tf symbol to fimo results
    tib_fimo <- tib_fimo %>%
      left_join(tib_meta, by = 'motif_id') %>%
      # remove hits with missing motif id (composite pwms)
      filter(!is.na(tf_symbol))

    # select best hit for each motif and sequence
    tib_fimo <- tib_fimo %>%
      group_by(sequence_name, tf_symbol) %>%
      dplyr::slice(which.min(pval)) %>%
      ungroup()

    # spread into feature matrix
    tib_fimo <- tib_fimo %>%
      mutate(pval = -1 * log10(pval)) %>%
      dplyr::select(-motif_id) %>%
      spread(key = tf_symbol, value = pval, fill = 0, drop = TRUE) %>%
      # perform cosmetics on the id
      mutate(id = sequence_name) %>%
      dplyr::select(-c(sequence_name)) %>%
      dplyr::select(id, everything())

  }

  return(tib_fimo)

}


theme_classic_lines <- function() {
  theme_pubr(border = T, legend = "top") +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.1),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_classic_lines_45 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 45) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.1),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_classic_lines_90 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 90) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.1),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_set(theme_classic_lines())

colors_diverse <- c("#264653", "#9AC1AE", "#5D987B", "#f2cc8f", "#e76f51")

ggplot_custom <- function(...) ggplot2::ggplot(...) + 
  scale_color_manual(values = colors_diverse) + 
  scale_fill_manual(values = colors_diverse)

register(MulticoreParam(8, progressbar = TRUE))
```


### Load peaks and counts
```{r peak, eval = T}
# Load and resize peaks using a ChromVAR function
peak_set <- readNarrowpeaks("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/bed_peaks/6420_peaks.narrowPeak")
peak_set <- peak_set[-119463]
peak_set <- peak_set[-119518]
peak_set <- peak_set[-119943]
peak_set <- peak_set[-119976]
peak_set <- peak_set[-120001]
peak_set <- peak_set[-120047]
peak_set <- peak_set[-120049]
peak_set <- peak_set[-120049]
peak_set <- peak_set[-120069]
peak_set <- peak_set[-120126]


# Load bam (count) files
## Generated K562 & HepG2 files
files_df <- readRDS("/DATA/usr/t.filipovska/projects/ATAC-seq_TF-footprinting/analysis/rds/bamfile_atacseq_metadata2.rds")
bam_files <- files_df$bam_file[files_df$treatment == "DMSO"]

## BAM files from public resources
bam_files_2 <- data.frame(path = c("/DATA/usr/m.trauernicht/data/ATAC/ENCFF530GWW.bam", "/DATA/usr/m.trauernicht/data/ATAC/ENCFF967ILH.bam"),
                          sample = c("HCT116_r1", "HCT116_r2"),
                          cell = c("HCT116", "HCT116"))

## Combine files
bam_files <- c(bam_files, bam_files_2$path)
cell_types <- c(files_df$celltype, bam_files_2$cell)

## Generate ChromVAR counts per peak file
fragment_counts <- getCounts(bam_files, 
                             peak_set, 
                             paired =  TRUE, 
                             #by_rg = TRUE, 
                             format = "bam", 
                             colData = DataFrame(celltype = files_df$celltype))

# idx_chromosome <- c(paste("chr", 1:22), "chrX")
# fragment_counts_filtered <- fragment_counts[fragment_counts@rowRanges@seqnames %in% idx_chromosome]

## Normalize counts
genome <- FaFile("/DATA/usr/m.trauernicht/data/genomes/hg38/hg38.fa")
fragment_counts_gc <- addGCBias(fragment_counts, 
                            genome = genome)

counts_filtered <- filterSamples(fragment_counts_gc, min_depth = 1500, 
                                 min_in_peaks = 0.15, shiny = FALSE)
```

---

## Load motifs and characterize + overlay with peaks

```{r motifs}
# Load motifs
motifs <- read_meme("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/motifs/selected_motifs.meme")
motifs <- motifs[-77] ## Remove NRF1 (double)

## Plot motif hierarchical tree
motifs_pfm <- convert_motifs(motifs = motifs, class = "motifStack-pfm")

hc <- clusterMotifs(motifs_pfm)
phylog <- ade4::hclust2phylog(hc)
plotMotifStackWithRadialPhylog(pfms = motifs_pfm, phylog = phylog)
plotMotifLogoStackWithTree(pfms = motifs_pfm, hc = hc)


## Compute motif similarity matrix
motifs_pwm <- convert_motifs(motifs = motifs, class = "PWMEnrich-PWM")

motif_similarity <- data.frame(merge(1:87, 1:87)) %>%
  dplyr::rename(motif_1 = x, motif_2 = y) %>%
  mutate(similarity = "", motif_1_name = "", motif_2_name = "")
  

for (i in 1:87) {
  for (j in 1:87) {
    print(i)
    x <- motifs_pwm[[i]]$pfm
    motif_1 <- motifs_pwm[[i]]$name
    y <- motifs_pwm[[j]]$pfm
    motif_2 <- motifs_pwm[[j]]$name
    storage.mode(x) <- "integer"
    storage.mode(y) <- "integer"
  
    motif_similarity$similarity[motif_similarity$motif_1 == i & motif_similarity$motif_2 == j] <- 
      motifSimilarity(m1 = x, m2 = y)
    
    motif_similarity$motif_1_name[motif_similarity$motif_1 == i & motif_similarity$motif_2 == j] <- motif_1
    
    motif_similarity$motif_2_name[motif_similarity$motif_1 == i & motif_similarity$motif_2 == j] <- motif_2
  
  }
}

motif_similarity_matrix <- motif_similarity %>%
  mutate(similarity = as.numeric(similarity)) %>%
  distinct(similarity, motif_1_name, motif_2_name) %>%
  spread(key = "motif_2_name", value = "similarity") %>%
  column_to_rownames("motif_1_name") 

## Plot motif similarity matrix
pheatmap(motif_similarity_matrix,
        border_color = NA,
        width = 10,
        height = 10)

## Calculate motifs per peak
pfm <- list()
motif_match <- data.frame(id = 1:87, tf = "")
for (i in 1:87) {
  print(i)
  print(motifs_pwm[[i]]$name)
  motif_match$tf[motif_match$id == i] <- motifs_pwm[[i]]$name
  pfm[[i]] <- PFMatrix(name=motifs_pwm[[i]]$name, 
                  bg=c(A=0.25, C=0.25, G=0.25, T=0.25),
                  profileMatrix=motifs_pwm[[i]]$pfm)
}

motifs_matrix <- NULL

for (i in seq(1,87,2)) {
  print(paste("i", i))
  print(paste("length", length(motifs_matrix)))
      if (i == 1) {
        
          motifs_matrix <- PFMatrixList(pfm[[1]], pfm[[2]])
      }
          if (i == 87) {
            motifs_matrix2 <- PFMatrixList(pfm[[87]])

            motifs_matrix <- c(motifs_matrix, motifs_matrix2)

          }
          if (i > 2 & i != 87) {
            
           motifs_matrix2 <- PFMatrixList(pfm[[length(motifs_matrix) +1 ]], pfm[[length(motifs_matrix) + 2]]) 
           
           motifs_matrix <- c(motifs_matrix, motifs_matrix2)
           
      }
}



motif_ix <- matchMotifs(motifs_matrix, counts_filtered, 
                        genome = genome)
```

---


## Compute motif z-scores
```{r}
## Compute motif deviations
dev <- computeDeviations(object = counts_filtered, annotations = motif_ix)

variability <- computeVariability(dev)

plotVariability(variability, use_plotly = T) 

motif_z_scores <- deviationScores(dev) %>%
  as.data.frame() %>%
  rownames_to_column("id") %>%
  mutate(id = as.numeric(id)) %>%
  left_join(motif_match) %>%
  pivot_longer(cols = c(-id, -tf), names_to = "sample", values_to = "motif_z-score")
```




### Session Info

```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

