---
title: "Barcode preprocessing - TF reporter library 90 TFs - KD screen HepG2"
author: 
  - name: "Max Trauernicht"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute - van Steensel lab"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
---

---

### Aim
35,500 reporters for 86 TFs were transfected into HepG2 cells that were previously treated with ~50 different siRNAs for TFs. In this script the barcode counts from these samples will be pre-processed.

---

## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}
```


### Libraries 

```{r setup, out.width= "80%", fig.align= "center",  warning= FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(data.table)
library(plyr)
library(stringr)
library(ggpubr)
library(GGally)
library(vwr)
library(dplyr)
library(tibble)
library(plotly)
library(ggbeeswarm)
library(haven)
library(readr)
library(parallel)
library(RColorBrewer)
library(gridExtra)
library(LncFinder)
library(tidyr)
library(grr)
library(viridis)
library(DESeq2)
library(PCAtools)
library(pheatmap)
library(IHW)
```


### Functions

```{r out.width= "80%", fig.align= "center",  warning= FALSE, message=FALSE}
# Custom functions
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}

ReadFasta<-function(file) {
   # Read the file line by line
   fasta<-readLines(file)
   # Identify header lines
   ind<-grep(">", fasta)
   # Identify the sequence lines
   s<-data.frame(ind=ind, from=ind+1, to=c((ind-1)[-1], length(fasta)))
   # Process sequence lines
   seqs<-rep(NA, length(ind))
   for(i in 1:length(ind)) {
      seqs[i]<-paste(fasta[s$from[i]:s$to[i]], collapse="")
   }
   # Create a data frame 
   DF<-data.frame(name=gsub(">", "", fasta[ind]), sequence=seqs)
   # Return the data frame as a result object from the function
   return(DF)
}

# From Fede:
# ggpairs custom functions
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {

  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y, "pairwise.complete.obs")
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)

  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }

  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())

  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]

  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }

  p <- p +
    theme(panel.background = element_rect(fill = corCol))

  return(p)
}

# Custom ggplot2 themes
theme_classic_lines <- function() {
  theme_pubr(border = T, legend = "top") +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.1),
                  strip.background = element_rect(fill = "#ced4da"))
    
}

theme_classic_lines_45 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 45) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.1),
                  strip.background = element_rect(fill = "#ced4da"))
}

theme_classic_lines_90 <- function() {
  theme_pubr(border = T,legend = "top", x.text.angle = 90) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.1),
                  strip.background = element_rect(fill = "#ced4da"))
}

theme_set(theme_classic_lines())

colors_diverse <- c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51")

ggplot_custom <- function(...) ggplot2::ggplot(...) + 
  scale_color_manual(values = colors_diverse) + 
  scale_fill_manual(values = colors_diverse)


hline <- function(y = 0, color = "black") {
  list(
    type = "line",
    x0 = 0,
    x1 = 1,
    xref = "paper",
    y0 = y,
    y1 = y,
    line = list(color = color)
  )
}
```


### Loading data

```{r data import, out.width= "80%", fig.align= "center",  warning= FALSE, message=FALSE}
# Load metadata file that contains all required information about the sequenced samples
metadata_df <- read_csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/mt20221011_metadata_all.csv")

# Load in barcode counts
bc_files <- paste(metadata_df$path, metadata_df$file, sep = "")
bc_files <- lapply(bc_files, fread, header = FALSE)
names(bc_files) <- metadata_df$sample_id

# Import barcode annotation of both libraries
bc_annotation <- read_csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6578_gen2_stimulation-2/mt20211021_bc_annotation_combined.csv") %>%
  dplyr::select(-library)
```


### Creating count data frames

```{r cluster_compare, out.width= "80%", fig.align= "center",  warning= FALSE, message=FALSE}
# Generate one long data frame from the list of data frames
bc_df <- bind_rows(bc_files, .id = "sample_id") %>%
  dplyr::select(sample_id, "barcode" = V1, "starcode_counts" = V2)

# Add barcode annotation to the data (also include barcodes that have 0 counts)
bc_annotation <- merge(bc_annotation, unique(bc_df$sample_id), all = T) %>%
  setnames("y", "sample_id")

bc_df <- merge(bc_df, bc_annotation, all = T, by = c("barcode", "sample_id"))

# Library 1 has barcodes of length 12 while library 2 has barcodes of length 13
# For sequencing samples that have library 1 and 2 mixed I extracted barcodes of length >=12
# So, I will now filter out all barcodes with length >13, because those are not relevant
bc_df$nchar <- nchar(bc_df$barcode)
bc_df <- bc_df %>%
  filter(nchar <= 13)

# Add experiment annotation to the data
bc_df <- bc_df[!is.na(bc_df$sample_id),]
bc_df <- merge(bc_df, metadata_df, all = T, by = "sample_id")

bc_df$starcode_counts[is.na(bc_df$starcode_counts)] <- 0
bc_df_remove <- bc_df %>%
  filter(nchar == 13 & bc_df$library == "1")
bc_df <- anti_join(bc_df, bc_df_remove)
bc_df_remove <- bc_df %>%
  filter(nchar == 12 & bc_df$library == "2")
bc_df <- anti_join(bc_df, bc_df_remove)

# Compute reads per million to estimate the relative counts in the respective sample
bc_df <- bc_df %>%
  mutate(rpm = ave(starcode_counts, sample_id, FUN = function(x) (x+1) / sum(x) *1e6 ))

bc_df_filt <- bc_df[!is.na(bc_df$tf),]

# Manually remove previously identified low quality samples
remove_samples <- c("HepG2_CREB1_r1_gcf6952", "K562_r3_gcf6502", "mES_TCF7L1_r2_gcf7027")
bc_df_filt <- bc_df_filt[!bc_df_filt$sample_id %in% remove_samples,]

# Filter TP53 from library 1+2 K562 data (we previously determined that we should filter these data)
bc_df_filt <- bc_df_filt %>%
  mutate(reporter_id = paste(tf, spacing, distance, promoter, background, sep = "_")) %>%
  mutate(reporter_id = gsub("bc-[0-9]{1,2}_", "", reporter_id))

bc_df_filt2 <- bc_df_filt %>%
  dplyr::filter(library == "1+2" & condition == "K562" & tf == "TP53")
bc_df_filt <- anti_join(bc_df_filt, bc_df_filt2, by = c("reporter_id", "sample_id"))
```


## Correlate pDNA data

```{r out.width= "80%", fig.align= "center",  warning= FALSE, message=FALSE}
# Make dataframe to compare raw counts between conditions
bc_df_filt <- bc_df_filt[!is.na(bc_df_filt$sample_id),]
bc_df_cor <- bc_df_filt %>%
  dplyr::select(-starcode_counts)

bc_df_cor <- bc_df_cor %>%
  filter(sample_id %in% c("pMT02_3_r1_gcf6952", "pMT02_r1_gcf6952", "pMT02_v2_r1_gcf6952", "pMT02_v3_r1_gcf6952", "pMT02-09_v2_r1_gcf6952", "pMT02-09_v3_r1_gcf6952", "E1961_pDNA_r1_gcf6992", "pDNA_r1_gcf7027",
                          "pMT02_r1_gcf5927", "pMT02_r1_gcf6301", "pMT02_r2_gcf5927", "pMT09_02_r1_gcf6502", "pMT09_02_r1_gcf6578", "pMT09_02_r1_gcf6881", "pMT09_02_r2_gcf6881", "pMT09_3_r1_gcf6952", "pMT09_r1_gcf6502",
                          "pMT09_r1_gcf6578", "pMT09_r1_gcf6952", "pMT09_r2_gcf6502", "pMT09_r3_gcf6578", "pMT09_v3_r1_gcf6952")) %>%
  dplyr::select(sample_id, barcode, rpm, library) %>%
  unique() %>%
  mutate(rpm = as.integer(rpm)) 

bc_df_cor[is.na(bc_df_cor)] <- 0


## Here I want to take a quick look how well individual samples correlate
# Compare pDNA data

n <- sample(1:nrow(bc_df_cor), 10000)
boundaries <- seq(from = 0.8, by = 0.05, length.out = 4)

for (i in unique(bc_df_cor$library)) {
    plt <- ggpairs(bc_df_cor %>% 
                     filter(library == i) %>%
                     spread(sample_id, rpm) %>%
                     column_to_rownames("barcode") %>%
                     dplyr::select(-library),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.1, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red") +
                   theme_bw()}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red") +
                   theme_bw()})) +
  ggtitle(paste("Correlation", i)) +
  theme(text = element_text(size = 10)) +
  xlab("rpm") +
  ylab("rpm")

print(plt)

}
```


---

## Data quality plots

```{r read count 2, out.width= "80%", fig.align= "center",  warning= FALSE, message= FALSE}
# I want to show the following:
## 1: Read distribution of matched barcodes vs. unmatched barcode

### a: total read counts per sample
for (i in unique(bc_df_filt$library)) {
  bc_df_filt_i <- bc_df_filt[bc_df_filt$library == i,]
  p <- plot_ly(bc_df_filt_i %>%
         mutate(sum_counts = ave(starcode_counts, sample, FUN = function(x) sum(x))) %>%
         dplyr::select(sample, sum_counts) %>%
         unique(), 
        x = ~sum_counts, y = ~sample, type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = paste("Number of reads per barcode per sample", i),
         xaxis = list(title = "Expected number of reads per barcode per sample"),
         yaxis = list(title = "Sample"))
  print(p)
}


### b: get a feeling for the distribution of the read counts - are there samples were there are few barcodes with high read counts?
n_highly_expressed <- data.frame("sample_id" = unique(bc_df_filt$sample_id[bc_df_filt$cell != "pDNA"]),
                                 "n_bc" = "", stringsAsFactors = F)
for (i in unique(bc_df_filt$sample_id)) {
  n_highly_expressed$n_bc[n_highly_expressed$sample_id == i] <- length(bc_df_filt$barcode[bc_df_filt$rpm > 500 & bc_df_filt$sample_id == i])
}

plot_ly(n_highly_expressed, x = ~sample_id, y = ~as.numeric(n_bc), type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Highly expressed barcodes",
         xaxis = list(title = "Number of barcodes with > 500 rpm"),
         yaxis = list(title = "Condition")) %>%
  layout(shapes = list(hline(75)))

n_highly_expressed$n_bc <- as.numeric(n_highly_expressed$n_bc)

bc_df_filt <- merge(bc_df_filt, n_highly_expressed, all = T, by = "sample_id")

## 2: What is the correlation of the cDNA bc counts with the pDNA bc counts? 
pDNA_1 <- bc_df_filt[grep(unique(bc_df_filt$pDNA[!is.na(bc_df_filt$pDNA)])[1], bc_df_filt$sample_id),] %>%
  dplyr::select(barcode, sample_id, rpm) %>%
  spread(sample_id, rpm) %>%
  column_to_rownames("barcode")
pDNA_1$pDNA_rpm <- rowMeans(pDNA_1)
pDNA_1 <- pDNA_1 %>%
  rownames_to_column(var = "barcode") %>%
  distinct(barcode, pDNA_rpm) %>%
  mutate(pDNA = unique(bc_df_filt$pDNA[!is.na(bc_df_filt$pDNA)])[1])

pDNA_2 <- bc_df_filt[grep(unique(bc_df_filt$pDNA[!is.na(bc_df_filt$pDNA)])[2], bc_df_filt$sample_id),] %>%
  dplyr::select(barcode, sample_id, rpm) %>%
  spread(sample_id, rpm) %>%
  column_to_rownames("barcode")
pDNA_2$pDNA_rpm <- rowMeans(pDNA_2)
pDNA_2 <- pDNA_2 %>%
  rownames_to_column(var = "barcode") %>%
  distinct(barcode, pDNA_rpm) %>%
  mutate(pDNA = unique(bc_df_filt$pDNA[!is.na(bc_df_filt$pDNA)])[2])

pDNA_3 <- bc_df_filt[grep(unique(bc_df_filt$pDNA[!is.na(bc_df_filt$pDNA)])[3], bc_df_filt$sample_id),] %>%
  dplyr::select(barcode, sample_id, rpm) %>%
  spread(sample_id, rpm) %>%
  column_to_rownames("barcode")
pDNA_3$pDNA_rpm <- rowMeans(pDNA_3)
pDNA_3 <- pDNA_3 %>%
  rownames_to_column(var = "barcode") %>%
  distinct(barcode, pDNA_rpm) %>%
  mutate(pDNA = unique(bc_df_filt$pDNA[!is.na(bc_df_filt$pDNA)])[3])

pDNA_4 <- bc_df_filt[grep(unique(bc_df_filt$pDNA[!is.na(bc_df_filt$pDNA)])[4], bc_df_filt$sample_id),] %>%
  dplyr::select(barcode, sample_id, rpm) %>%
  spread(sample_id, rpm) %>%
  column_to_rownames("barcode")
pDNA_4$pDNA_rpm <- rowMeans(pDNA_4)
pDNA_4 <- pDNA_4 %>%
  rownames_to_column(var = "barcode") %>%
  distinct(barcode, pDNA_rpm) %>%
  mutate(pDNA = unique(bc_df_filt$pDNA[!is.na(bc_df_filt$pDNA)])[4])

pDNA_5 <- bc_df_filt[grep(unique(bc_df_filt$pDNA[!is.na(bc_df_filt$pDNA)])[5], bc_df_filt$sample_id),] %>%
  dplyr::select(barcode, sample_id, rpm) %>%
  spread(sample_id, rpm) %>%
  column_to_rownames("barcode")
pDNA_5$pDNA_rpm <- rowMeans(pDNA_5)
pDNA_5 <- pDNA_5 %>%
  rownames_to_column(var = "barcode") %>%
  distinct(barcode, pDNA_rpm) %>%
  mutate(pDNA = unique(bc_df_filt$pDNA[!is.na(bc_df_filt$pDNA)])[5])

pDNA_all <- rbind(pDNA_1, pDNA_2, pDNA_3, pDNA_4, pDNA_5)


bc_df_filt <- merge(pDNA_all, bc_df_filt, all = T, by = c("barcode", "pDNA"))
bc_df_filt <- bc_df_filt[!is.na(bc_df_filt$sample_id),]

for (i in unique(bc_df_filt$library)) {
  bc_df_i <- bc_df_filt[bc_df_filt$library == i,] 
  if (i == "1") {
  p <- ggplot(bc_df_i, aes(x = pDNA_rpm, y = rpm)) +
  geom_bin2d(bins = 50)+
  xlim(0,500) +
  ylim(0,2000)+
  scale_color_viridis() +
  facet_wrap(~sample_id)
  
  print(p)
  }
    if (i == "2") {
  p <- ggplot(bc_df_i, aes(x = pDNA_rpm, y = rpm)) +
  geom_bin2d(bins = 50)+
  xlim(0,500) +
  ylim(0,2000)+
  scale_color_viridis() +
  facet_wrap(~sample_id)
  
  print(p)
    }
   if (i == "1+2"){
  p <- ggplot(bc_df_i, aes(x = pDNA_rpm, y = rpm)) +
  geom_bin2d(bins = 50)+
  xlim(0,500) +
  ylim(0,2000)+
  scale_color_viridis() +
  facet_wrap(~sample_id)
  
  print(p)
  }
  
}

# Generate correlation heatmaps 
for (i in unique(bc_df_filt$pDNA[!is.na(bc_df_filt$pDNA)])) {
  bc_df_i <- bc_df_filt[bc_df_filt$pDNA == i,] %>%
    dplyr::select(sample_id, rpm, barcode)  %>% 
    filter_all(any_vars(!is.na(.))) %>%
    unique() %>%
    spread(sample_id, rpm) %>% 
    filter_all(any_vars(!is.na(.))) %>%
    column_to_rownames("barcode")
  
  x <- cor(bc_df_i, method = "pearson", use = "pairwise.complete.obs")
  
  p<- pheatmap(x, border = "black", main = i)
  print(p)
}
    


cor <- data.frame("sample_id" = unique(bc_df_filt$sample_id), "cor" = "", stringsAsFactors = F)

for (i in unique(bc_df_filt$sample_id)) {
      cor$cor[cor$sample_id == i] <- stats::cor(bc_df_filt$rpm[bc_df_filt$sample_id == i], bc_df_filt$pDNA_rpm[bc_df_filt$sample_id == i], use = "pairwise.complete.obs")
}

cor <- cor %>%
  na.omit()

bc_df_filt <- merge(bc_df_filt, cor, by = c("sample_id"), all = T)
```

---

### Sample filtering

```{r out.width= "80%", fig.align= "center",  warning= FALSE, message=FALSE}
# Remove samples with low dynamic range (this most likely is due to unhappy cells, pDNA contamination or some other mistake in the sample prep)
bc_df_filt2 <- bc_df_filt %>%
  filter(n_bc > 75 | cell == "pDNA") %>% 
  dplyr::select(-n_bc)

# Remove samples with high pDNA contamination (most likely those samples were already removed in the previous step as this is a bit redundant)
bc_df_filt2 <- bc_df_filt2 %>%
  filter(cor <= 0.5 | cell == "pDNA") %>% 
  dplyr::select(-cor)

# Which samples were removed in these steps?
paste("the following samples were removed due to low quality:", setdiff(unique(bc_df_filt$sample_id), unique(bc_df_filt2$sample_id)))
```

---


### Normalization of barcode counts:  
Divide cDNA barcode counts through pDNA barcode counts to get activity

```{r normalization, out.width= "80%", fig.align= "center",  warning= FALSE, message=FALSE}
# Compute activity by dividing cDNA bc counts through pDNA bc counts
bc_df_filt2$activity <- bc_df_filt2$rpm / bc_df_filt2$pDNA_rpm

# Remove rows that could not be computed due to too little pDNA counts
bc_df_cDNA <- bc_df_filt2 %>%
  drop_na(activity)
```


---

## Calculate mean activity - filter out outlier barcodes 

```{r out.width= "80%", fig.align= "center",  warning= FALSE, message=FALSE}
# First identify and remove outlier barcodes - this removes the noise created by faulty barcode clustering etc. 
## Calculate median activity for each reporter and deviation of reporters from the median
bc_df_cDNA <- bc_df_cDNA %>%
  mutate(mean_activity = ave(activity, reporter_id, sample_id, FUN = function(x) quantile(x, 0.5))) %>%
  mutate(deviation = activity / mean_activity) %>%
  mutate(n_reporters = ave(reporter_id, reporter_id, sample_id, FUN = function(x) as.numeric(length(x))))


## Plot effect of highest outlier bc -> SNP in minP promoter
outlier <- bc_df_cDNA[bc_df_cDNA$reporter_id == "ZFP42_5bp_21bp_minP_3",] %>% 
  dplyr::select(barcode_number, activity, sample) %>%
  unique()

ggplot_custom(outlier, aes(x = barcode_number, y = activity)) +
  geom_quasirandom() +
  xlab("barcode #") + 
  labs(title = "Zfp42 reporter - TATA-box minP mutated", 
       subtitle = "bc8 AGAGGGTATATAAT -> AGAGGGGATATAAT")


## Plot effect of highest outlier bc -> Elk1 outlier
outlier <- bc_df_cDNA[bc_df_cDNA$reporter_id == "ELK1_5bp_21bp_mCMV_2",] %>% 
  dplyr::select(barcode_number, activity) %>%
  unique()

ggplot_custom(outlier, aes(x = barcode_number, y = activity)) +
  geom_quasirandom()+
  xlab("barcode #") + 
  labs(title = "Elk1 reporter - bc1 attached to Pax6 reporter")



## Plot effect of highest outlier bc -> Elk1 outlier
outlier <- bc_df_cDNA[bc_df_cDNA$reporter_id == "TCF7L2_5bp_10bp_hBGm_3",] %>% 
  dplyr::select(barcode_number, activity) %>%
  unique()

ggplot_custom(outlier, aes(x = barcode_number, y = activity)) +
  geom_quasirandom() +
  xlab("barcode #") + 
  labs(title = "Tcf3 reporter - bc1 attached to multiple (less active) reporters")



## Choose arbitrary cutoff to get rid of most extreme outliers
bc_df_cDNA_filt <- bc_df_cDNA[bc_df_cDNA$deviation > 0.25 & bc_df_cDNA$deviation < 4 & bc_df_cDNA$n_reporters > 2,]

bc_df_cDNA_filt$n_reporters <- ave(bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$reporter_id, 
                                bc_df_cDNA_filt$sample_id, FUN =
                                  function(x) as.numeric(length(x)))

bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  filter(n_reporters >= 2)

## Recalculate mean and sd
bc_df_cDNA_filt$mean_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, 
                                bc_df_cDNA_filt$sample_id, FUN =
                                  function(x) mean(x))
```


### Scaling data 

```{r out.width= "80%", fig.align= "center",  warning= FALSE}
## Compute activity relative to minimal promoter background
bc_df_cDNA_filt_neg <- bc_df_cDNA_filt %>%
  dplyr::select(sample_id, activity, reporter_id, tf, promoter) %>%
  filter(str_detect(tf, "RANDOM")) %>%
  unique() %>%
  mutate(activity = ave(activity, promoter, sample_id, FUN = function(x) mean(x))) %>%
  dplyr::select(-reporter_id) %>%
  unique() %>%
  dplyr::select("tf_activity2" = activity, promoter, sample_id) %>%
  unique()


### For library 2 I don't have random motifs, so I will take the mean of the mutated motifs instead (get rid of the 2% highest active mutated controls)
bc_df_cDNA_filt_neg2 <- bc_df_cDNA_filt %>%
  dplyr::select(sample_id, reporter_id, tf, promoter, neg_ctrls, mean_activity) %>%
  filter(neg_ctrls == "Yes") %>%
  unique() %>%
  group_by(sample_id, promoter) %>%
  dplyr::top_frac(-0.75, mean_activity) %>%
  mutate(activity = ave(mean_activity, promoter, sample_id, FUN = function(x) mean(x))) %>%
  dplyr::select(-reporter_id) %>%
  unique() %>%
  dplyr::select("tf_activity" = activity, promoter, sample_id) %>%
  unique()

bc_df_cDNA_filt$promoter[bc_df_cDNA_filt$hPGK == "Yes"] <- "minP"


x <- merge(bc_df_cDNA_filt_neg, bc_df_cDNA_filt_neg2)
plot(x$tf_activity, x$tf_activity2)
### Normalization by random has some outliers (gcf6502) -> I can better for now normalize by mutated motifs since I also have data for all samples from that


bc_df_cDNA_filt <- merge(bc_df_cDNA_filt, bc_df_cDNA_filt_neg2, by = c("sample_id", "promoter"))
bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  mutate(minP_activity = activity / tf_activity)


### Compute a p-value for each reporter with respect to the background minimal promoter activity
bc_df_cDNA_filt$library[is.na(bc_df_cDNA_filt$library)] <- "1+2"
bc_df_cDNA_filt_neg2 <- bc_df_cDNA_filt %>%
  dplyr::select(sample_id, reporter_id, tf, promoter, neg_ctrls, mean_activity, condition, library) %>%
  filter(neg_ctrls == "Yes") %>%
  unique() %>%
  group_by(sample_id, promoter) %>%
  dplyr::top_frac(-0.75, mean_activity) %>%
  mutate(id = paste(condition, promoter, library, sep = "_"))

bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  mutate(reporter_id2 = gsub("(.*)_.*(_[1-3]{1}$)", "\\1\\2", reporter_id),
         id = paste(condition, promoter, library, sep = "_"))


pval_list <- list()
for (i in unique(bc_df_cDNA_filt$reporter_id2)) {
  print(i)
  test_x <- bc_df_cDNA_filt[bc_df_cDNA_filt$reporter_id2 == i,]
  ids <- unique(bc_df_cDNA_filt$id[bc_df_cDNA_filt$reporter_id2 == i])
  pval_list[[i]] <- lapply(ids, function(x) t.test(test_x$activity[test_x$id == x], 
                                                   bc_df_cDNA_filt_neg2$mean_activity[bc_df_cDNA_filt_neg2$id == x])$p.value)
  names(pval_list[[i]]) <- ids
}

pval_df <- data.frame(bind_rows(pval_list))
rownames(pval_df) <- names(pval_list)
pval_df <- pval_df %>%
  rownames_to_column("reporter_id2") %>%
  pivot_longer(-reporter_id2, names_to = "id", values_to = "pval") %>%
  mutate(id = gsub("1[.]2", "1+2", id)) %>%
  na.omit()

bc_df_cDNA_filt <- merge(bc_df_cDNA_filt, pval_df, by = c("reporter_id2", "id"), all = T)
bc_df_cDNA_filt <- bc_df_cDNA_filt[!is.na(bc_df_cDNA_filt$sample_id),]

## Compute activity relative to NFYA reporter
bc_df_cDNA_filt_nfya <- bc_df_cDNA_filt %>%
  dplyr::select(sample_id, activity, reporter_id, tf, promoter) %>%
  filter(str_detect(tf, "NFYA"), promoter == "minP") %>%
  unique() %>%
  mutate(activity = ave(activity, sample_id, FUN = function(x) mean(x))) %>%
  dplyr::select(-reporter_id) %>%
  unique() %>%
  dplyr::select("nfya_activity" = activity, sample_id) %>%
  unique()

bc_df_cDNA_filt <- merge(bc_df_cDNA_filt, bc_df_cDNA_filt_nfya, by = "sample_id")
bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  mutate(nfya_activity = activity / nfya_activity)

# Compute mean of technical replicates
bc_df_cDNA_filt$mean_activity_sample_minP <- ave(bc_df_cDNA_filt$minP_activity, bc_df_cDNA_filt$reporter_id,
                                bc_df_cDNA_filt$sample_id, FUN =
                                  function(x) mean(x, na.rm = T))

bc_df_cDNA_filt$mean_activity_sample_nfya <- ave(bc_df_cDNA_filt$nfya_activity, bc_df_cDNA_filt$reporter_id,
                                bc_df_cDNA_filt$sample_id, FUN =
                                  function(x) mean(x, na.rm = T))
bc_df_cDNA_filt$mean_activity_sample <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id,
                                bc_df_cDNA_filt$sample_id, FUN =
                                  function(x) mean(x, na.rm = T))


# Compute activity relative to mutated motif
bc_df_cDNA_filt_mutated <- bc_df_cDNA_filt %>%
  filter(neg_ctrls == "Yes") %>%
  dplyr::select(reporter_id, 'mutated_activity' = mean_activity_sample, sample_id) %>%
  unique() %>%
  mutate(reporter_id = gsub("_neg", "", reporter_id)) %>%
  filter(str_detect(reporter_id, "MAFA_", negate = T), str_detect(reporter_id, "NR2E3_", negate = T), str_detect(reporter_id, "^T_", negate = T))

bc_df_cDNA_filt <- merge(bc_df_cDNA_filt, bc_df_cDNA_filt_mutated, by = c("reporter_id", "sample_id"), all = T)
bc_df_cDNA_filt <- bc_df_cDNA_filt[!is.na(bc_df_cDNA_filt$barcode),]
bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  mutate(mean_activity_sample_neg = mean_activity / mutated_activity)

## Statistical significance of reporters vs negative control
bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  mutate(reporter_id2 = gsub("_neg", "", reporter_id))

my.t.test.p.value <- function(x,y) {
  obj<-try(t.test(x,y), silent=TRUE)
  if (is(obj, "try-error")) return(1) else return(obj$p.value)
}
pval_list <- list()
for (i in unique(bc_df_cDNA_filt$reporter_id2)) {
  print(i)
  test_x <- bc_df_cDNA_filt[bc_df_cDNA_filt$reporter_id2 == i,]
  ids <- intersect(unique(bc_df_cDNA_filt$id[bc_df_cDNA_filt$reporter_id2 == i & bc_df_cDNA_filt$neg_ctrls == "Yes"]),
                     unique(bc_df_cDNA_filt$id[bc_df_cDNA_filt$reporter_id2 == i & bc_df_cDNA_filt$neg_ctrls == "No"]))
  p <- lapply(ids, function(x) my.t.test.p.value(test_x$activity[test_x$id == x & test_x$neg_ctrls == "No"], 
                                                   test_x$activity[test_x$id == x & test_x$neg_ctrls == "Yes"]))
  pval_list[[i]] <- p
  names(pval_list[[i]]) <- ids
}

pval_list <- pval_list[lengths(pval_list) > 0]
pval_df <- data.frame(bind_rows(pval_list))
rownames(pval_df) <- names(pval_list)
pval_df <- as.data.frame(pval_df)
pval_df <- pval_df %>%
  rownames_to_column("reporter_id2") %>%
  pivot_longer(-reporter_id2, names_to = "id", values_to = "pval_neg") %>%
  mutate(id = gsub("1[.]2", "1+2", id)) %>%
  na.omit()

bc_df_cDNA_filt <- merge(bc_df_cDNA_filt, pval_df, by = c("reporter_id2", "id"), all = T)
```


---


## Calculate correlations between technical replicates

```{r correlations_2, out.width= "80%", fig.align= "center",  warning= FALSE}
## Identify and relabel double reporters
bc_df_12 <- bc_df_cDNA_filt %>%
  filter(nchar == 12, library == "1+2")
bc_df_13 <- bc_df_cDNA_filt %>%
  filter(nchar == 13, library == "1+2")
double_reporters <- unique(bc_df_13$reporter_id[bc_df_13$reporter_id %in% bc_df_12$reporter_id])

double_reporters_df <- bc_df_cDNA_filt %>%
  filter(reporter_id %in% double_reporters, library == "1+2", nchar == 13)
double_reporters_df$reporter_id_2 <- paste(double_reporters_df$reporter_id, double_reporters_df$nchar)
bc_df_cDNA_filt$reporter_id_2 <- paste(bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$nchar)
bc_df_cDNA_filt <- bc_df_cDNA_filt[!bc_df_cDNA_filt$reporter_id_2 %in% double_reporters_df$reporter_id_2,] %>%
  dplyr::select(-reporter_id_2)
double_reporters_df <- double_reporters_df %>%
  dplyr::select(-reporter_id_2)

double_reporters_df$barcode_number[double_reporters_df$barcode_number == 1] <- 9
double_reporters_df$barcode_number[double_reporters_df$barcode_number == 2] <- 10
double_reporters_df$barcode_number[double_reporters_df$barcode_number == 3] <- 11
double_reporters_df$barcode_number[double_reporters_df$barcode_number == 4] <- 12
double_reporters_df$barcode_number[double_reporters_df$barcode_number == 5] <- 13

bc_df_cDNA_filt <- rbind(bc_df_cDNA_filt, double_reporters_df)

## Combine replicates in 8 different columns

bc_df_rep <- bc_df_cDNA_filt %>% 
  filter(commercial_reporter == "No", rand_promoter == "No", native_enhancer == "No", hPGK == "No") %>%
  dplyr::select(barcode_number, activity, sample_id, reporter_id, neg_ctrls, library) %>%
  spread(barcode_number, activity)
names(bc_df_rep) <- gsub("([1-9]{1})", "TR\\1", names(bc_df_rep))


for (i in unique(bc_df_rep$library)) {
  p <- ggscatter(bc_df_rep[bc_df_rep$library == i,], x = "TR1", y = "TR2",
                 add = "reg.line",
                 color = "neg_ctrls",
                 size = 0.5,
                 add.params = list(color = "blue", fill = "lightgray"), 
                 title = paste("rep1 vs rep2 - library:", i),
                 conf.int = TRUE, ylab = "rep2", xlab = "rep1") + 
    stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
    geom_abline(linetype = "dashed") +
    xlim(0,50) + ylim(0,50) +
    #scale_color_manual(values = colors) +
    facet_wrap(~sample_id)
  print(p)
}


# Correlation matrix plot
n <- sample(1:nrow(bc_df_rep), 5000)
boundaries <- seq(from = 0.8, by = 0.05, length.out = 4)
plt <- ggpairs(bc_df_rep %>% filter(library == "1+2") %>% dplyr::select('TR1', 'TR2', 'TR3', 'TR4', 'TR5', 'TR6', 'TR7', 'TR8'),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.1, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red") +
                   theme_bw()}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red") +
                   theme_bw()})) +
  ggtitle("Correlation Between Technial Replicates") +
  theme(text = element_text(size = 20)) +
  xlab("Reporter activity") +
  ylab("Reporter activity")

print(plt)
```


---

## Correlation between replicates

```{r correlations_3, out.width= "80%", fig.align= "center",  warning= FALSE}
# Correlation plots of the replicates
## Combine replicates of normalized data in 3 different columns
bc_df_rep <- bc_df_cDNA_filt %>% 
  dplyr::select(replicate, mean_activity_sample_minP, reporter_id, condition, neg_ctrls, library, gcf) %>% 
  unique() %>%
  spread(replicate, mean_activity_sample_minP)

colors <- c("#2D3047", "#1B998B")

for (i in unique(bc_df_rep$library)) {
  p <- ggscatter(bc_df_rep[bc_df_rep$library == i,], x = "r1", y = "r2",
                 add = "reg.line",
                 color = "neg_ctrls",
                 size = 0.5,
                 add.params = list(color = "blue", fill = "lightgray"), 
                 title = paste("rep1 vs rep2 - library:", i),
                 conf.int = TRUE, ylab = "rep2", xlab = "rep1") + 
    stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
    geom_abline(linetype = "dashed") +
    xlim(0,50) + ylim(0,50) +
    scale_color_manual(values = colors) +
    facet_wrap(~condition)
  print(p)
}
```


```{r data export, out.width= "80%", fig.align= "center",  warning= FALSE}
# Mean of the three replicates
bc_df_cDNA_filt$reporter_activity <- ave(bc_df_cDNA_filt$mean_activity_sample, bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$condition, FUN = function(x) mean(x))
bc_df_cDNA_filt$reporter_activity_minP <- ave(bc_df_cDNA_filt$mean_activity_sample_minP, bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$condition, FUN = function(x) mean(x))
bc_df_cDNA_filt$reporter_activity_nfya <- ave(bc_df_cDNA_filt$mean_activity_sample_nfya, bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$condition, FUN = function(x) mean(x))
bc_df_cDNA_filt$reporter_activity_neg <- ave(bc_df_cDNA_filt$mean_activity_sample_neg, bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$condition, FUN = function(x) mean(x))

# Polish export dataframe
bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  dplyr::select(-pDNA_rpm, -mean_activity, -deviation, -n_reporters, -path, -file, -gcf, -labguru_experiment)

# # Critical step: for reporters that have a high negative control value - change to minP values instead
# bc_df_cDNA_filt2 <- bc_df_cDNA_filt %>%
#   dplyr::filter(neg_ctrls == "No", native_enhancer == "No", hPGK == "No", commercial_reporter == "No") %>%
#   distinct(reporter_id, condition, tf, reporter_activity_minP, reporter_activity_neg) %>%
#   mutate(reporter_activity_minP_mean = ave(reporter_activity_minP, tf, condition, FUN = function(x) mean(x, na.rm = T))) %>%
#   mutate(reporter_activity_neg_mean = ave(reporter_activity_neg, tf, condition, FUN = function(x) mean(x, na.rm = T))) %>%
#   mutate(dif = reporter_activity_minP_mean / reporter_activity_neg_mean) %>%
#   group_by(tf, condition) %>%
#   mutate(cor = cor(reporter_activity_minP, reporter_activity_neg, use = "pairwise.complete.obs")) %>%
#   ungroup() %>%
#   mutate(neg_active = ifelse(dif > 3 | cor < 0.5, "yes", "no")) %>%
#   distinct(tf, condition, neg_active) 


# bc_df_cDNA_filt3 <- bc_df_cDNA_filt %>%
#   filter(neg_ctrls == "Yes", native_enhancer == "No", hPGK == "No", commercial_reporter == "No") %>%
#   distinct(reporter_id, condition, tf, reporter_activity_minP) %>%
#   mutate(tf = gsub("_neg", "", tf)) %>%
#   mutate(neg_activity = ave(reporter_activity_minP, tf, condition, FUN = mean)) %>%
#   distinct(tf, condition, neg_activity)

# bc_df_cDNA_filt4 <- merge(bc_df_cDNA_filt3, bc_df_cDNA_filt2)



# bc_df_cDNA_filt <- merge(bc_df_cDNA_filt, bc_df_cDNA_filt2, by = c("tf", "condition"), all = T) %>%
#   mutate(reporter_activity_neg = ifelse(neg_active == "yes" | is.na(reporter_activity_neg), reporter_activity_minP, reporter_activity_neg)) %>%
#   mutate(pval_neg = ifelse(neg_active == "yes" | is.na(pval_neg), pval, pval_neg)) 

## p-value adjustment
bc_df_cDNA_filt$reporter_rpm <- ave(bc_df_cDNA_filt$rpm, bc_df_cDNA_filt$reporter_id, FUN = mean)
#bc_df_cDNA_filt$pval_adj_minP <- adj_pvalues(ihw(pval ~ reporter_rpm,  data = bc_df_cDNA_filt, alpha = 0.1))
#bc_df_cDNA_filt$pval_adj_neg <- adj_pvalues(ihw(pval_neg ~ reporter_rpm,  data = bc_df_cDNA_filt, alpha = 0.1))

# Export bc_df for cDNA analysis
filename <- SetFileName("_reporter_activity_filt_combined", "mt")
setwd("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf7027_KD_mES/results/")
write.csv(bc_df_cDNA_filt, file = paste(filename,".csv", sep = ""), row.names = F)
```

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

