---
title: "Multiplexed identification of TF-specific reporters"
author:
- name: Max Trauernicht
  email: m.trauernicht@nki.nl
  affiliation: "Netherlands Cancer Institute - van Steensel lab"
date: "`r format(Sys.time(), "%d/%m/%Y")`"
output:
  pdf_document:
    toc: yes
    toc_depth: '4'
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: no
---
---

### Aim
~36,000 reporters for 86 TFs were transfected into 9 different cell types + treated with ~90 different siRNAs or various signaling molecules. In this script I will analyze TF reporter activities in detail and review how individual reporters respond to TF concentration variations.


### Description of input data
- library = with which library was this sample transfected (1, 2, or 1+2)
- nchar = barcode length (12 for library 1, 13 for library 2)
- commercial_reporter = positive control reporter sequence from commercial source
- neg_ctrls = reporter with mutated TF binding sites?
- hPGK = positive control, chunk from hPGK promoter (only in lib 1)
- native_enhancer = genomic mES enhancer sequence chunks from Miguels libraries (only lib 1)
- activity = rpm + pseudocount cDNA / rpm + pseudocount pDNA
- mean_activity_sample = mean of activity per unique reporter per sample (individually per replicate)
- reporter_activity = mean activity per unique reporter per condition (mean of replicates)
- minP_activity = activity relative to mean of minimal promoter only reporters of that sample
- mean_activity_sample_minP = mean of minP_activity per unique reporter 
- reporter_activity_minP = mean of mean_activity_sample_minP per condition
- nfya_activity = activity relative to mean of nfya reporters (constitutive positive control) of that sample
- mean_activity_sample_nfya = mean of nfya_activity per unique reporter 
- reporter_activity_nfya = mean of mean_activity_sample_nfya per condition
- mean_activity_sample_neg = mean_activity_sample relative to its paired negative control (identical reporter with mutated binding sites)
- reporter_activity_neg = mean of mean_activity_sample_neg per condition
- pval_minP = p-value of t-test comparing activity distributions of individual reporters with the distribution of minimal promoter only reporters within the same condition
- pval_neg = p-value of t-test comparing activity distributions of individual reporters with the distribution of the activities of its paired negative controls within the same condition

---

## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}
```

### Libraries 

```{r setup, out.width= "80%", fig.align= "center", warning = FALSE, message= FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(RColorBrewer)
library(ggplot2)
library(dplyr)
library(maditr)
library(tibble)
library(pheatmap)
library(ggpubr)
library(visNetwork)
library(ggbeeswarm)
library(ggforce)
library(viridis)
library(plyr)
library(igraph)
library(ggraph)
library(cowplot)
library(gridExtra)
library(pROC)
library(tidyr)
library(stringr)
library(randomForest)
library(ggrastr)
library(readr)
library(ggbiplot)
library(IHW)
library(biomaRt)
library(ggh4x)
library(ggiraph)
library(plotly)
library(AMR)
```

### Functions

```{r out.width= "80%", fig.align= "center"}
### Custom functions
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}


# Extract p-value from linear model
lmp <- function (modelobject) {
    if (class(modelobject) != "lm") stop("Not an object of class 'lm' ")
    f <- summary(modelobject)$fstatistic
    p <- pf(f[1],f[2],f[3],lower.tail=F)
    attributes(p) <- NULL
    return(p)
}

# Set custom ggplot2 theme and custom colors
theme_classic_lines <- function() {
  theme_pubr(border = F, legend = "top") +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_classic_lines_45 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 45) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_classic_lines_90 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 90) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_barplot <- function() {
  theme_pubr(border = T, legend = "none") +
            theme(panel.grid.major.x = element_line(colour = "black", size = 0.25),
                  strip.background = element_blank(),
                  strip.text = element_text(face="bold", hjust=0)
            )
}

theme_set(theme_classic_lines())

colors_diverse <- c("#264653", "#9AC1AE", "#5D987B", "#f2cc8f", "#e76f51")

ggplot_custom <- function(...) ggplot2::ggplot(...) + 
  scale_color_manual(values = colors_diverse) + 
  scale_fill_manual(values = colors_diverse)
```

### Loading data

```{r out.width= "80%", fig.align= "center"}
# Import processed bc counts from the preprocessing step
cDNA_df <- read.csv("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf7124_stimulations/results/mt20230511_reporter_activity_filt_combined.csv", header = T)

cDNA_df$stimulation[is.na(cDNA_df$stimulation)] <- "no"

# final measure of reporter activity: combination of reporter_activity_minP and reporter_activity_neg
## for most reporters I will take reporter_activity_neg, however, this cannot be done for commercial controls for example
cDNA_df <- cDNA_df %>%
  mutate(reporter_activity_neg = ifelse(tf %in% c("POU5F1::SOX2", "NR1H4::RXRA") | commercial_reporter == "Yes" | is.na(reporter_activity_neg), reporter_activity_minP, reporter_activity_neg)) 
```

---

## Investigate individual reporter activities - 1

Aim: Show that there are differences in activities between reporter variants

```{r}
## Figure 1E: Activities per cell type - density
cDNA_df2 <- cDNA_df %>%
  filter(stimulation == "no",
         commercial_reporter == "No", 
         hPGK == "No", 
         str_detect(tf, "RANDOM", negate = T), 
         native_enhancer == "No") %>%
  distinct(tf, reporter_activity_minP, cell, reporter_id, neg_ctrls) %>%
  mutate(reporter_activity_minP = ave(reporter_activity_minP, cell, reporter_id, FUN = mean)) %>%
  mutate(reporter_activity_minP = log2(reporter_activity_minP)) %>%
  distinct() 

ggplot(cDNA_df2 %>%
         mutate(cell = factor(cell, levels = c("A549", "K562", "HCT116", "MCF7", "HepG2", 
                                               "U2OS", "HEK293", "mES", "NPC"))),
       aes(x = reporter_activity_minP, fill = neg_ctrls)) +
  geom_density(alpha = .4) +
  facet_wrap(~cell, ncol = 2) +
  scale_fill_manual(values = c("Yes" = "grey70", "No" = "#DD6B48")) +
  theme_pubr(border = T)

## Explore PCA
reporter_mat <- cDNA_df %>%
  filter(stimulation == "no",
         neg_ctrls == "No",
         commercial_reporter == "No", 
         hPGK == "No", 
         str_detect(tf, "RANDOM", negate = T), 
         native_enhancer == "No") %>%
  distinct(tf, reporter_activity_neg, cell, neg_ctrls) %>%
  mutate(reporter_activity_neg = ave(reporter_activity_neg, cell, tf, FUN = mean)) %>%
  mutate(reporter_activity_neg = log2(reporter_activity_neg)) %>%
  distinct(reporter_activity_neg, tf, cell) %>%
  spread(key = "cell", value = "reporter_activity_neg") %>%
  column_to_rownames("tf") %>%
  replace(is.na(.), 0) %>%
  as.matrix()

meta_pca <- data.frame(row.names = c("A549", "HCT116", "HEK293", "HepG2", "K562", "MCF7", "mES", "NPC", "U2OS"),
                       cell = c("A549", "HCT116", "HEK293", "HepG2", "K562", "MCF7", "mES", "NPC", "U2OS"),
                       organism = c("human", "human", "human", "human", "human", "human", "mouse", "mouse", "human"))

reporter_pca2 <- prcomp(t(reporter_mat))

ggplot_pca(reporter_pca2, arrows = F, groups = meta_pca$cell) +
  theme_pubr(border = T)
```


---


## Integrate RNA-seq data

Aim: Does the reporter activity correlate with transcript abundance across conditions?
```{r}
# Load in RNA-seq data
## Set the names of the TFs for which I want to retrieve the data
tfs <- cDNA_df %>%
  filter(neg_ctrls == "No", commercial_reporter == "No", hPGK == "No", str_detect(tf, "RANDOM", negate = T)) %>%
  distinct(tf)
tfs$tf[tfs$tf == "AHR::ARNT"] <- "AHR"
tfs$tf[tfs$tf == "FOS::JUN"] <- "FOS"
tfs$tf[tfs$tf == "MAF::NFE2"] <- "NFE2"
tfs$tf[tfs$tf == "NR1H4::RXRA"] <- "NR1H4"
tfs$tf[tfs$tf == "NR4A2::RXRA"] <- "NR4A2"
tfs$tf[tfs$tf == "PPARA::RXRA"] <- "PPARA"
tfs$tf[tfs$tf == "PPARG::RXRA"] <- "PPARG"
tfs$tf[tfs$tf == "SMAD2::SMAD3::SMAD4"] <- "SMAD2"
tfs$tf[tfs$tf == "STAT1::STAT2"] <- "STAT1"
tfs$tf[tfs$tf == "VDR::RXRA"] <- "VDR"
tfs$tf[tfs$tf == "NFATc1"] <- "NFATC1"
tfs$tf[tfs$tf == "RARA:RXRA"] <- "RARA"

## Load mES data from Miguel
load('/DATA/usr/m.martinez.ara/GitLab/gurten/gurten/data/fc181121_epsure_tadec_rnaseq_expr_joshi.RData')

mES_rna <- tib_fpkm_joshi %>%
  mutate(twoi = (twoi_1 + twoi_2) / 2) %>%
  dplyr::select(symbol, twoi) %>%
  mutate(twoi = (twoi / sum(twoi))*10^6) %>% ## Transform from FPKM to TPM
  group_by(symbol) %>%
  arrange(desc(twoi)) %>%
  top_n(1) %>%
  ungroup()

mES_rna$symbol[mES_rna$symbol == "TRP53"] <- "TP53"

mES_rna <- mES_rna %>%
  setnames(c("symbol", "twoi"), c("tf", "TPM")) %>%
  #filter(mES_rna$tf %in% tfs$tf) %>%
  mutate(cell = "mES")

## Load data from protein atlas (TPM)
proteinatlas_rna <- read_tsv("/DATA/usr/m.trauernicht/data/RNA_seq/rna_celline.tsv") %>%
  dplyr::select('tf' = `Gene name`, TPM, "cell" = `Cell line`) %>%
  filter(cell %in% c("A-549", "U2OS", "Hep-G2", "HEK293", "K-562", "MCF-7", "HCT16")) %>%
  mutate(cell = gsub(" ", "", cell)) %>%
  mutate(cell = gsub("-", "", cell)) %>%
  #filter(tf %in% tfs$tf)
  mutate(tf_cell = paste(tf, cell)) %>%
  group_by(tf_cell) %>%
  slice_max(n = 1, order_by = TPM) %>%
  ungroup() %>%
  dplyr::select(-tf_cell)

## diffTF-compatible output
library("org.Hs.eg.db")
proteinatlas_rna_export <- proteinatlas_rna %>%
  mutate(ENSEMBL = mapIds(org.Hs.eg.db, keys = proteinatlas_rna$tf, keytype = "SYMBOL", column="ENSEMBL")) %>%
  filter(!tf %in% c("BORCS8-MEF2B", "GAGE12J")) %>%
  dplyr::select(-tf) %>%
  na.omit() %>%
  distinct() %>%
  spread(key = "cell", value = "TPM")
  

npc_rna <- read_delim("/shared/gcf/m.trauernicht/7086/genecounts.txt") %>%
  dplyr::select("tf" = external_gene_id, "NPC_r1" = `1_NPC_r1`, "NPC_r2" = `2_NPC_r2`, start_position, end_position) %>%
  mutate(length = end_position - start_position) %>%
  mutate(NPC_counts = (NPC_r1 + NPC_r2) / 2) %>%
  mutate(NPC_rpk = NPC_counts / (length / 1000)) %>%
  mutate(pm = sum(NPC_rpk / 1e6)) %>%
  mutate(TPM = NPC_rpk / pm) %>%
  mutate(cell = "NPC") %>%
  distinct(tf, TPM, cell) %>%
  mutate(tf = toupper(tf))  %>%
  group_by(tf) %>%
  slice_max(n = 1, order_by = TPM) %>%
  ungroup()


## Combine data from all cell types and change names back to original TF names
tf_rna <- rbind(mES_rna, proteinatlas_rna, npc_rna)
tf_rna$tf[tf_rna$tf == "AHR"] <- "AHR::ARNT"
tf_rna$tf[tf_rna$tf == "FOS"] <- "FOS::JUN"
tf_rna$tf[tf_rna$tf == "NFE2"] <- "MAF::NFE2"
tf_rna$tf[tf_rna$tf == "NR1H4"] <- "NR1H4::RXRA"
tf_rna$tf[tf_rna$tf == "NR4A2"] <- "NR4A2::RXRA"
tf_rna$tf[tf_rna$tf == "PPARA"] <- "PPARA::RXRA"
tf_rna$tf[tf_rna$tf == "PPARG"] <- "PPARG::RXRA"
tf_rna$tf[tf_rna$tf == "SMAD2"] <- "SMAD2::SMAD3::SMAD4"
tf_rna$tf[tf_rna$tf == "STAT1"] <- "STAT1::STAT2"
tf_rna$tf[tf_rna$tf == "VDR"] <- "VDR::RXRA"
tf_rna$tf[tf_rna$tf == "NFATC1"] <- "NFATc1"

tf_rna2 <- tf_rna %>% filter(tf == "POU5F1")
tf_rna2$tf[tf_rna2$tf == "POU5F1"] <- "POU5F1::SOX2"
tf_rna <- rbind(tf_rna, tf_rna2)

tf_rna2 <- tf_rna %>% filter(tf == "RARA")
tf_rna2$tf[tf_rna2$tf == "RARA"] <- "RARA:RXRA"
tf_rna <- rbind(tf_rna, tf_rna2)

tf_rna$TPM[is.na(tf_rna$TPM)] <- 0

#write_tsv(tf_rna, "/DATA/usr/m.trauernicht/data/RNA_seq/rna_tpm_all_tfs.tsv")


## Prepare tf reporter data frame
cDNA_df2 <- cDNA_df %>%
  #filter(tf != "RFX1") %>%
  dplyr::filter(neg_ctrls == "No",
                #commercial_reporter == "No",
                stimulation == "no",
                hPGK == "No",
                str_detect(tf, "RANDOM", negate = T),
                native_enhancer == "No") %>%
  distinct(tf, reporter_activity_neg, cell, reporter_id, commercial_reporter) %>%
  mutate(reporter_activity_neg = ave(reporter_activity_neg, cell, reporter_id, FUN = mean)) %>%
  mutate(reporter_activity_neg = log2(reporter_activity_neg)) %>%
  distinct() %>%
  mutate(tf = gsub("_.*", "", tf))

## Combine TF reporter data with RNA-seq data
cDNA_df3 <- merge(cDNA_df2, tf_rna, all = T) %>%
  na.omit()

## Compute spearman correlations for each individual reporter of reporter activity vs. TPM counts across the nine cell types
for (i in unique(cDNA_df3$reporter_id)) {
  if (nrow(cDNA_df3[cDNA_df3$reporter_id == i,]) > 2) {
  cDNA_df3$cor_pval[cDNA_df3$reporter_id == i] <- -log10(cor.test(cDNA_df3$reporter_activity_neg[cDNA_df3$reporter_id == i], cDNA_df3$TPM[cDNA_df3$reporter_id == i], method = "pearson", exact = F)$p.value)
  cDNA_df3$cor[cDNA_df3$reporter_id == i] <- cor(cDNA_df3$reporter_activity_neg[cDNA_df3$reporter_id == i], cDNA_df3$TPM[cDNA_df3$reporter_id == i], method = "pearson")
  }
  else {
    cDNA_df3$cor_pval[cDNA_df3$reporter_id == i] <- 1
    cDNA_df3$cor[cDNA_df3$reporter_id == i] <- 0
  }
}

## I am only interested in positive correlations, so set all negative correlations to zero
cDNA_df3 <- cDNA_df3 %>%
  mutate(cor_pval = ifelse(cor < 0, 0, cor_pval))
#
# ## Plot the correlations including the p-values
# for (i in unique(cDNA_df3$tf)) {
#
#   x <- cDNA_df3 %>%filter(tf == i)
#
#   label <- data.frame(reporter_id = x$reporter_id, label = round(x$cor_pval, 2))
#
#   p <- ggplot(x,
#               aes(x = TPM, y = reporter_activity_neg)) +
#     geom_point() +
#     geom_abline(lty = 2) +
#     geom_smooth(method = "lm", se = F) +
#     facet_wrap(~reporter_id) +
#     ggtitle(i) +
#     ylab("Reporter Activity") +
#     xlab("TPM") +
#     geom_label(data = label, aes(label = label, color = ifelse(label > 1, "red", "black")),
#                x = Inf, y = -Inf, hjust=1, vjust=0,
#                inherit.aes = FALSE) +
#     scale_color_manual(values = c("black", "red"))
#
#   print(p)
# }

## Set Inf values to max per TF
cDNA_df3 <- cDNA_df3 %>%
  group_by(tf) %>%
  mutate(cor_pval = ifelse(is.finite(cor_pval), cor_pval, max(cor_pval[is.finite(cor_pval)]))) %>%
  ungroup()

## Export correlations
rna_correlations <- cDNA_df3 %>%
  distinct(reporter_id, cor_pval, cor)

## Filter for TFs where correlations actually make sense -> minimum TPM > 0.5, maximum TPM < 0.5, dif_TPM > 5
cDNA_df4 <- cDNA_df3 %>%
  mutate(min_TPM = ave(TPM, tf, FUN = min),
         max_TPM = ave(TPM, tf, FUN = max)) %>%
  mutate(dif_TPM = max_TPM - min_TPM) %>%
  mutate(mean_tf_activity = ave(reporter_activity_neg, tf, cell, FUN = mean)) %>%
  mutate(max_mean_tf_activity = ave(mean_tf_activity, tf, FUN = max)) %>%
  filter(max_TPM > .5 & dif_TPM > 5 & min_TPM < .5 & max_mean_tf_activity > 0.9 | tf %in% c("PGR", "GATA1", "NR1I2", "NR4A1", "NFATc1",
                                                                                            "TEAD1"))


## Plot correlation p-values per TF
ggplot_custom(cDNA_df4 %>%
                filter(commercial_reporter == "No") %>%
                distinct(reporter_id, cor_pval, tf) %>%
           mutate(sign = ifelse(cor_pval < 1, "No", "Yes")) %>%
             mutate(mean_cor_pval = ave(cor_pval, tf, FUN = mean)),
       aes(x = reorder(tf, -mean_cor_pval), y = cor_pval, color = sign)) +
  geom_quasirandom() +
  geom_hline(yintercept = 1, lty = 2) +
  theme_pubr(border = T, x.text.angle = 90)

## Examplify for a couple of TFs
ggplot_custom(cDNA_df3 %>%
                filter(commercial_reporter == "No") %>%
                distinct(reporter_id, cor_pval, tf, commercial_reporter) %>%
                mutate(mean_cor_pval = ave(cor_pval, tf, FUN = mean)) %>%
                mutate(sign = ifelse(cor_pval < 1.30103, "No", "Yes")) %>%
                filter(tf %in% c("HNF4A", "GATA4", "ESR1", "THRB", "GATA1")),
       aes(x = reorder(tf, -mean_cor_pval), y = cor_pval, color = sign)) +
  geom_quasirandom() +
  geom_hline(yintercept = 1.30103, lty = 2) +
  theme_pubr(border = T)

ggplot(cDNA_df3 %>%
           filter(reporter_id %in% c("HNF1A_10bp_10bp_minP_1", "HNF1A_5bp_10bp_minP_1")) %>%
           distinct(cell, reporter_id, reporter_activity_neg, TPM),
       aes(x = TPM, y = reporter_activity_neg)) +
    geom_abline(lty = 2) +
    geom_smooth(method = "lm", color = "black") +
    geom_point(size = 2) +
    facet_wrap(~reporter_id, ncol = 1) +
    theme_pubr(border = T)

ggplot(cDNA_df3 %>%
           filter(reporter_id %in% c("POU5F1::SOX2_10bp_21bp_minP_1", "POU5F1::SOX2_5bp_21bp_minP_1")) %>%
           distinct(cell, reporter_id, reporter_activity_neg, TPM),
       aes(x = TPM, y = reporter_activity_neg)) +
    geom_abline(lty = 2) +
    geom_smooth(method = "lm", color = "black") +
    geom_point(size = 2) +
    facet_wrap(~reporter_id, ncol = 1) +
    theme_pubr(border = T)

ggplot(cDNA_df3 %>%
         filter(reporter_id %in% c("HNF4A_10bp_10bp_mCMV_2", "HNF4A_5bp_10bp_minP_1")) %>%
         distinct(cell, reporter_id, reporter_activity_neg, TPM),
       aes(x = TPM, y = reporter_activity_neg)) +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm", color = "black") +
  geom_point() +
  facet_wrap(~reporter_id, ncol = 1) +
  theme_pubr(border = T)

ggplot(cDNA_df3 %>%
         filter(reporter_id %in% c("GATA4_10bp_10bp_minP_3", "GATA4_5bp_10bp_minP_3")) %>%
         distinct(cell, reporter_id, reporter_activity_neg, TPM),
       aes(x = TPM, y = reporter_activity_neg)) +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm", color = "black") +
  geom_point(size = 2) +
  facet_wrap(~reporter_id, ncol = 1) +
  theme_pubr(border = T)

ggplot(cDNA_df3 %>%
         filter(reporter_id %in% c("ESR1_5bp_10bp_mCMV_1", "ESR1_10bp_21bp_minP_3")) %>%
         distinct(cell, reporter_id, reporter_activity_neg, TPM),
       aes(x = TPM, y = reporter_activity_neg)) +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm", color = "black") +
  geom_point() +
  facet_wrap(~reporter_id, ncol = 1) +
  theme_pubr(border = T)

ggplot(cDNA_df3 %>%
         filter(reporter_id %in% c("THRB_5bp_10bp_mCMV_3", "THRB_10bp_10bp_mCMV_3")) %>%
         distinct(cell, reporter_id, reporter_activity_neg, TPM),
       aes(x = TPM, y = reporter_activity_neg)) +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm", color = "black") +
  geom_point(size = 2) +
  facet_wrap(~reporter_id, ncol = 1) +
  theme_pubr(border = T)

ggplot(cDNA_df3 %>%
         filter(reporter_id %in% c("FOXO1_10bp_21bp_minP_2", "FOXO1_5bp_21bp_minP_2")) %>%
         distinct(cell, reporter_id, reporter_activity_neg, TPM),
       aes(x = TPM, y = reporter_activity_neg)) +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm", color = "black") +
  geom_point(size = 2) +
  facet_wrap(~reporter_id, ncol = 1) +
  theme_pubr(border = T)

ggplot(cDNA_df3 %>%
         filter(reporter_id %in% c("TFCP2L1_5bp_10bp_mCMV_1", "TFCP2L1_5bp_10bp_minP_2")) %>%
         distinct(cell, reporter_id, reporter_activity_neg, TPM),
       aes(x = TPM, y = reporter_activity_neg)) +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm", color = "black") +
  geom_point(size = 2) +
  facet_wrap(~reporter_id, ncol = 1) +
  theme_pubr(border = T)


# ## Tissue specificity of top TFs
# chosen_tfs <- cDNA_df6 %>%
#   distinct(tf, quality_score) %>%
#   mutate(quality_score = ave(quality_score, tf, FUN = mean)) %>%
#   distinct() %>%
#   arrange(desc(quality_score)) %>%
#   top_n(n = 50)
# 
# tf_expression_pattern <- read_csv("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/Fig4A.csv") %>%
#   pivot_longer(cols = -Sample, names_to = "tf", values_to = "expression") %>%
#   filter(tf %in% chosen_tfs$tf)
# 
# 
# ggplot(tf_expression_pattern %>%
#          mutate(expression_mean = ave(expression, Sample, FUN = mean)),
#        aes(x = reorder(Sample, -expression_mean),  y = expression)) +
#   geom_quasirandom() +
#   theme_classic_lines_90()
# 
# 
# cancer_models <- read_csv("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/CRISPRGeneEffect.csv") %>%
#   pivot_longer(cols = -`...1`, names_to = "tf", values_to = "expression") %>%
#   mutate(tf = gsub(" (.*)", "", tf)) %>%
#   filter(tf %in% chosen_tfs$tf)
# 
# 
# cancer_models <- cancer_models %>%
#   mutate(mean_effect = ave(expression, `...1`, FUN = mean))
# 
# 
# ggplot(cancer_models,
#        aes(x = reorder(`...1`, -mean_effect),  y = expression)) +
#   geom_quasirandom() +
#   theme_classic_lines_90()
```

---

## Add proteomics data

Aim: Similar to before: does reporter activity correlate with protein abundance across cell types?
```{r}
## Import proteomics data from depmap
proteomics_df <- read_csv("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/Proteomics_subsetted_NAsdropped.csv") %>%
  dplyr::select(-depmap_id, -lineage_1, -lineage_2, -lineage_3, -lineage_4, -lineage_5, -lineage_6) %>%
  pivot_longer(cols = -cell_line_display_name, names_to = "gene", values_to = "protein_abundance")

proteomics_df_long <- proteomics_df %>%
  mutate(gene = gsub(" (.*)", "", gene)) %>%
  filter(gene %in% tfs$tf) %>%
  mutate(min_abundance = min(protein_abundance, na.rm = T))

proteomics_df_long$cell_line_display_name[proteomics_df_long$cell_line_display_name == "HEPG2"] <- "HepG2"

proteomics_df_long$protein_abundance[is.na(proteomics_df_long$protein_abundance)] <- unique(proteomics_df_long$min_abundance)
  
proteomics_df_long <- proteomics_df_long %>%
  mutate(protein_abundance = ave(protein_abundance, gene, cell_line_display_name, FUN = function(x) mean(x, na.rm = T))) %>%
  dplyr::select(protein_abundance, "tf" = gene, "cell" = cell_line_display_name) %>%
  distinct() 

## Combine TF reporter data with RNA-seq data
cDNA_df3 <- merge(cDNA_df2, proteomics_df_long, all = T) %>%
  na.omit()

## Compute spearman correlations for each individual reporter of reporter activity vs. TPM counts across the nine cell types
for (i in unique(cDNA_df3$reporter_id)) {
  if (nrow(cDNA_df3[cDNA_df3$reporter_id == i,]) > 2) {
  cDNA_df3$cor_pval[cDNA_df3$reporter_id == i] <- -log10(cor.test(cDNA_df3$reporter_activity_neg[cDNA_df3$reporter_id == i], cDNA_df3$protein_abundance[cDNA_df3$reporter_id == i], method = "pearson", exact = F)$p.value)
  cDNA_df3$cor[cDNA_df3$reporter_id == i] <- cor(cDNA_df3$reporter_activity_neg[cDNA_df3$reporter_id == i], cDNA_df3$protein_abundance[cDNA_df3$reporter_id == i], method = "pearson")
  }
  else {
    cDNA_df3$cor_pval[cDNA_df3$reporter_id == i] <- 1
    cDNA_df3$cor[cDNA_df3$reporter_id == i] <- 0
  }
}

## I am only interested in positive correlations, so set all negative correlations to zero
cDNA_df3 <- cDNA_df3 %>%
  mutate(cor_pval = ifelse(cor < 0, 0, cor_pval))

## Set Inf values to max per TF
cDNA_df3 <- cDNA_df3 %>%
  group_by(tf) %>%
  mutate(cor_pval = ifelse(is.finite(cor_pval), cor_pval, max(cor_pval[is.finite(cor_pval)]))) %>%
  ungroup()

## Export correlations
protein_correlations <- cDNA_df3 %>%
  distinct(reporter_id, cor_pval, cor)

## Correlate RNA and protein correlations
rna_protein_cor <- merge(protein_correlations %>% dplyr::select(reporter_id, "prot_cor" = cor_pval), rna_correlations %>% dplyr::select(reporter_id, "rna_cor" = cor_pval), all = T) %>%
  replace(is.na(.), 0)

ggplot(rna_protein_cor %>%
         mutate(tf = gsub("(.*)_.*_.*_.*_.*", "\\1", reporter_id)) %>%
         mutate(source = gsub(".*_(.*)", "\\1", tf)) %>%
         mutate(source = ifelse(source == "TF-romanov", "romanov", source)) %>%
         mutate(source = ifelse(!source %in% c("romanov", "TF-seq", "promega"), "designed", source)) %>%
         mutate(tf = gsub("_.*", "", tf)),
       aes(x = prot_cor, y = rna_cor)) +
  geom_point(aes(color = source)) +
  geom_smooth(method = "lm") +
  theme_pubr() +
  facet_wrap(~tf)
  

ggplot_custom(cDNA_df3 %>%
                filter(commercial_reporter == "No") %>%
                distinct(reporter_id, cor_pval, tf, commercial_reporter) %>%
                mutate(mean_cor_pval = ave(cor_pval, tf, FUN = mean)) %>%
                mutate(sign = ifelse(cor_pval < 1.30103, "No", "Yes")) %>%
                filter(tf %in% c("HNF4A", "GATA4", "ESR1", "THRB", "GATA1")),
       aes(x = reorder(tf, -mean_cor_pval), y = cor_pval, color = sign)) +
  geom_quasirandom() +
  geom_hline(yintercept = 1.30103, lty = 2) +
  theme_pubr(border = T)

ggplot(cDNA_df3 %>%
           filter(reporter_id %in% c("GATA1_5bp_10bp_minP_3", "GATA1_10bp_10bp_minP_3")) %>%
           distinct(cell, reporter_id, reporter_activity_neg, protein_abundance),
       aes(x = protein_abundance, y = reporter_activity_neg)) +
    geom_abline(lty = 2) +
    geom_smooth(method = "lm", color = "black") +
    geom_point(size = 2) +
    facet_wrap(~reporter_id, ncol = 1) +
    theme_pubr(border = T)

ggplot(cDNA_df3 %>%
           filter(reporter_id %in% c("HOMEZ_10bp_10bp_minP_1", "HOMEZ_5bp_10bp_minP_1")) %>%
           distinct(cell, reporter_id, reporter_activity_neg, protein_abundance),
       aes(x = protein_abundance, y = reporter_activity_neg)) +
    geom_abline(lty = 2) +
    geom_smooth(method = "lm", color = "black") +
    geom_point(size = 2) +
    facet_wrap(~reporter_id, ncol = 1) +
    theme_pubr(border = T)

ggplot(cDNA_df3 %>%
         filter(reporter_id %in% c("NR3C1_10bp_21bp_hBGm_3", "NR3C1_10bp_10bp_minP_2")) %>%
         distinct(cell, reporter_id, reporter_activity_neg, protein_abundance),
       aes(x = protein_abundance, y = reporter_activity_neg)) +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm", color = "black") +
  geom_point() +
  facet_wrap(~reporter_id, ncol = 1) +
  theme_pubr(border = T)

ggplot(cDNA_df3 %>%
         filter(reporter_id %in% c("STAT3_10bp_10bp_minP_3", "STAT3_10bp_10bp_minP_1")) %>%
         distinct(cell, reporter_id, reporter_activity_neg, protein_abundance),
       aes(x = protein_abundance, y = reporter_activity_neg)) +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm", color = "black") +
  geom_point(size = 2) +
  facet_wrap(~reporter_id, ncol = 1) +
  theme_pubr(border = T)

ggplot(cDNA_df3 %>%
         filter(reporter_id %in% c("WT1_10bp_21bp_mCMV_1", "WT1_5bp_21bp_mCMV_1")) %>%
         distinct(cell, reporter_id, reporter_activity_neg, protein_abundance),
       aes(x = protein_abundance, y = reporter_activity_neg)) +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm", color = "black") +
  geom_point() +
  facet_wrap(~reporter_id, ncol = 1) +
  theme_pubr(border = T)

ggplot(cDNA_df3 %>%
         filter(reporter_id %in% c("SMAD4_10bp_21bp_minP_2", "SMAD4_5bp_21bp_minP_2")) %>%
         distinct(cell, reporter_id, reporter_activity_neg, protein_abundance),
       aes(x = protein_abundance, y = reporter_activity_neg)) +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm", color = "black") +
  geom_point(size = 2) +
  facet_wrap(~reporter_id, ncol = 1) +
  theme_pubr(border = T)

ggplot(cDNA_df3 %>%
         filter(reporter_id %in% c("SRF_10bp_21bp_mCMV_3", "SRF_5bp_21bp_mCMV_3")) %>%
         distinct(cell, reporter_id, reporter_activity_neg, protein_abundance),
       aes(x = protein_abundance, y = reporter_activity_neg)) +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm", color = "black") +
  geom_point(size = 2) +
  facet_wrap(~reporter_id, ncol = 1) +
  theme_pubr(border = T)

ggplot(cDNA_df3 %>%
         filter(reporter_id %in% c("SOX2_5bp_10bp_minP_2", "SOX2_10bp_10bp_minP_2")) %>%
         distinct(cell, reporter_id, reporter_activity_neg, protein_abundance),
       aes(x = protein_abundance, y = reporter_activity_neg)) +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm", color = "black") +
  geom_point(size = 2) +
  facet_wrap(~reporter_id, ncol = 1) +
  theme_pubr(border = T)

ggplot(cDNA_df3 %>%
         filter(reporter_id %in% c("RXRA_10bp_21bp_mCMV_3", "RXRA_5bp_21bp_mCMV_3")) %>%
         distinct(cell, reporter_id, reporter_activity_neg, protein_abundance),
       aes(x = protein_abundance, y = reporter_activity_neg)) +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm", color = "black") +
  geom_point(size = 2) +
  facet_wrap(~reporter_id, ncol = 1) +
  theme_pubr(border = T)

ggplot(cDNA_df3 %>%
         filter(reporter_id %in% c("OTX1_5bp_10bp_minP_2", "OTX1_10bp_10bp_minP_2")) %>%
         distinct(cell, reporter_id, reporter_activity_neg, protein_abundance),
       aes(x = protein_abundance, y = reporter_activity_neg)) +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm", color = "black") +
  geom_point(size = 2) +
  facet_wrap(~reporter_id, ncol = 1) +
  theme_pubr(border = T)


```
Some reporter activities correlate surprisingly well with protein abundance.

---


## Linear model - TF reporter grammar

Aim: If the measured activities can be fitted by a linear model, it is likely that the activity is true. If not, we might be looking at noise instead.

```{r log-linear-model-5, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
reporter_features <- cDNA_df %>%
  distinct(reporter_id, promoter, spacing, distance, background) %>%
  na.omit()

cDNA_df_TF <- cDNA_df %>%
  filter(neg_ctrls == "No", commercial_reporter == "No", hPGK == "No", native_enhancer == "No", stimulation %in% c("no", "pathway")) %>%
  distinct(reporter_activity_neg, promoter, spacing, distance, background, tf, condition, reporter_id, stimulation, reference_condition) %>%
  mutate(tf_condition = paste(tf, condition, sep = "_")) %>%
  # mutate(max_activity = ave(reporter_activity_neg, tf_condition, FUN = max)) %>%
  # mutate(min_activity = ave(reporter_activity_neg, tf_condition, FUN = min)) %>%
  # mutate(dif = max_activity - min_activity) %>%
  # filter(dif > 2.5) %>%
  mutate(reporter_activity_neg = log2(reporter_activity_neg)) %>%
  left_join(reference_condition %>% dplyr::select(-tf), by = c("reporter_id", "reference_condition")) %>%
  mutate(reporter_dif =  reference_activity - reporter_activity_neg) %>%
  mutate(background = as.character(background)) %>%
  mutate(promoter = as.factor(promoter)) %>%
  #mutate(reporter_activity_neg = ifelse(stimulation != "no", reporter_dif, reporter_activity_neg)) %>%
  distinct(reporter_activity_neg, promoter, spacing, distance, background, condition, tf, tf_condition, reporter_id)

## Fit linear model with design features for actual data
exp_variance <- data.frame("feature" = c("promoter", "spacing", "distance", "background", "Residuals"))
weight <- data.frame("feature" = c("(Intercept)", "promoterminP", "promoterhBGm", "promotermCMV",
                  "spacing10bps","spacing5bp", "distance10bp","distance21bp", "background1","background2", "background3"))

cDNA_df_TF$promoter <- relevel(cDNA_df_TF$promoter, ref = "minP")
cDNA_df_TF$row <- rownames(cDNA_df_TF)
cDNA_df_TF3 <- data.frame()

lm_list <- list()
for (i in unique(cDNA_df_TF$tf_condition)) {
  if (nrow(cDNA_df_TF[cDNA_df_TF$tf_condition == i,]) > 9) {
  print(i)
  lm_list[[i]] <- lm(reporter_activity_neg ~
                       distance + promoter + spacing + background,
                cDNA_df_TF[cDNA_df_TF$tf_condition == i,])
  }
}

f <- lapply(names(lm_list), function(x) summary(lm_list[[x]])$fstatistic)
names(f) <- names(lm_list)
p <- lapply(names(f), function(x) pf(f[[x]][1],f[[x]][2],f[[x]][3],lower.tail=F))
p <- data.frame(bind_rows(p)) %>% setnames("value", "pval")
p$tf_condition <- names(lm_list)

y <- lapply(names(lm_list), function(x) data.frame("activity_predicted" = lm_list[[x]]$fitted.values) %>% rownames_to_column("row") %>% mutate(tf_condition = x))
y <- data.frame(bind_rows(y)) %>% left_join(p, by = "tf_condition")

cDNA_df_TF2 <- y %>%
  left_join(cDNA_df_TF)

cDNA_df_TF3 <- rbind(cDNA_df_TF3, cDNA_df_TF2)

mean(cDNA_df_TF3$pval)

w <- lapply(names(lm_list), function(x) 
  if (length(lm_list[[x]]$coefficients) == length(summary(lm_list[[x]])$coefficients[,4])) {
  data.frame('weight' = lm_list[[x]]$coefficients, 'pval' = summary(lm_list[[x]])$coefficients[,4]) %>% mutate(tf = x) %>% rownames_to_column('feature')
  }
)

w <- data.frame(bind_rows(w))
weight <- weight %>%
  left_join(w)

x <- lapply(names(lm_list), function(x) anova(lm_list[[x]]) %>% rownames_to_column("feature") %>% mutate(tf = x))
x <- data.frame(bind_rows(x)) %>% 
  setnames(c("Sum.Sq", "Pr..F."), c("sum_sq", "pval")) %>%
  distinct(feature, sum_sq, pval, tf) %>%
  mutate(sum = ave(sum_sq, tf, FUN = sum)) %>%
  mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
  distinct(tf, rel_sum_sq, pval, feature)

exp_variance <- exp_variance %>%
  left_join(x)

## Fine-tune dfs
exp_variance$group <- "no residual"
exp_variance$group[grep("Residual", exp_variance$feature)] <- "residual"
exp_variance$rel_sum_sq[exp_variance$feature == "Residuals"] <- 100 - exp_variance$rel_sum_sq[exp_variance$feature == "Residuals"]
exp_variance$feature <- revalue(exp_variance$feature, c("Residuals"="total"))
prediction <- cDNA_df_TF3 %>%
  dplyr::select(-row) %>%
  na.omit()
weight <- weight %>%
  filter(feature != "(Intercept)")
categorie <- c("promoter", "spacing", "distance", "background")
weight$features <- gsub(paste(categorie, collapse="|"), "",weight$feature)
ml <- length(unique(weight$features))
feature <- unique(weight$features)
weight$condition <- gsub(paste(feature, collapse = "|"), "", weight$feature)
weight <- weight %>% dplyr::select(-feature)


cDNA_df_TF3 <- cDNA_df_TF3 %>%
  mutate(pval_adj = p.adjust(pval, method = "fdr"))

best_models <- cDNA_df_TF3 %>%
  distinct(tf, pval_adj, tf_condition) %>%
  group_by(tf) %>%
  arrange(desc(pval_adj)) %>%
  slice_tail(n = 1)

chosen_conditions <- cDNA_df6 %>%
  mutate(tf_condition = paste(cDNA_df6$tf, cDNA_df6$cell, sep = "_")) %>%
  mutate(tf_condition = gsub("mES$", "mES_2i_LIF", tf_condition)) %>%
  distinct(tf_condition)

not_fitted <- chosen_conditions$tf_condition[!chosen_conditions$tf_condition %in% cDNA_df_TF3$tf_condition]

cDNA_df_TF3_not_fitted <- cDNA_df_TF %>%
  filter(tf_condition %in% not_fitted) %>%
  mutate(activity_predicted = 0, pval_adj = 1)

cDNA_df_TF3 <- rbind.fill(cDNA_df_TF3_not_fitted, cDNA_df_TF3)

weight_not_fitted_1 <- weight %>% filter(tf == "THRB_K562") %>% mutate(weight = 0, pval = 0) %>% dplyr::select(-tf)
weight_not_fitted <- merge(not_fitted, weight_not_fitted_1) %>%
  dplyr::rename(tf = x)

weight <- rbind(weight, weight_not_fitted)

## Example: Predicted vs. measured for one TF
ggplot(cDNA_df_TF3 %>%
         filter(tf == "HNF4A", condition == "HepG2_FBS"),
       aes(x = activity_predicted, y = reporter_activity_neg, color = spacing)) +
  geom_abline(lty = 2) +
  geom_point(size = 2) +
  scale_color_manual(values = c("#7C9EB2", "#BFD0D9"))

weight_ex <- weight %>%
  filter(tf == "HNF4A_HepG2_FBS")

zero_weights <- weight %>%
  filter(features %in% c("minP", "1", "10bps", "10bp")) %>%
  mutate(weight = 0, pval = 1) %>%
  mutate(tf = "HNF4A_HepG2_FBS")
                   
weight_ex <- rbind(weight_ex, zero_weights)
                   
feature_names <- data.frame("features" = c("minP", "mCMV", "hBGm", "1", "2", "3", "10bps", "5bp", "10bp", "21bp"),
                            "feature_names" = c("minP", "minCMV", "minHBG", "Spacer 1", "Spacer 2", "Spacer 3", "10bp Spacer", 
                                                "5bp Spacer", "10bp Distance", "21bp Distance"))
                     
weight_ex <- weight_ex %>%
  left_join(feature_names)
                   

ggplot(weight_ex %>% 
         mutate(sign = ifelse(pval < 0.05, "Yes", "No")) %>%
         mutate(feature_names = factor(feature_names, levels = c("minP", "minCMV", "minHBG", "Spacer 1", "Spacer 2", "Spacer 3", "10bp Spacer", 
                                                                                                        "5bp Spacer", "10bp Distance", "21bp Distance"))),
       aes(x = feature_names, y = weight, alpha = sign)) +
  geom_hline(yintercept = 0, lty = 2) +
  geom_bar_interactive(stat = "identity", color = "black") +
  ylab("Weight") + 
  xlab("Reporter design feature") +
  theme_classic_lines_45() +
  scale_alpha_discrete(range = c(0.3,1))

## Explore weights for all TFs
weight2 <- weight %>%
  filter(tf %in% chosen_conditions$tf_condition) %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  mutate(weight = ifelse(pval > 0.1, 0, weight)) %>%
  distinct(tf, features, weight) %>%
  spread(key = "features", value = "weight") %>%
  column_to_rownames("tf") %>%
  filter(rowSums(., na.rm = T) != 0)

pheatmap(weight2 %>%
                  t(),
         clustering_method = "ward.D2",
         cellwidth = 10,
         cellheight = 10,
         border_color = F,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         angle_col = 90,
         cluster_cols = T,
         cutree_rows = 4)

ggplot(weight2 %>%
         pivot_longer(cols = 1:6, names_to = "feature", values_to = "weight") %>%
         mutate(mean_weight = ave(weight, feature, FUN = function(x) mean(x, na.rm = T))),
       aes(x = reorder(feature, -mean_weight), y = weight)) +
  geom_quasirandom(color = colors_diverse[2]) +
  stat_summary(fun = median, fun.min = median, fun.max = median, geom = "crossbar", color = "black", width = 0.25, lwd = 0.4) +
  ylab("Weight (log2 activity)") + xlab("") +
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12),
        axis.text.y = element_text(size = 10))

## Investigate hits with strong impact on spacer length
weight2 <- weight %>%
  filter(tf %in% chosen_conditions$tf_condition) %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  filter(condition == "spacing", pval < 0.01, abs(weight) >= 0.5)

actual_spacer_lengths <- data.frame(tf = c("CEBPB", "CREB1", "GRHL1", "HNF1A", "HNF4A", "KLF4", "MAF::NFE2", "NFE2L2", 
                                            "NFKB1", "NFYA", "NR1H4::RXRA", "NR1I2", "TCF7L2", "TFCP2L1", "TP53", "XBP1"),
                                    center_center_dist = c(10, 12, 10, 15, 16, 11, 15, 15, 13, 13, 15, 17, 10, 16, 20, 14),
                                    monomer_center_dist = c(6, 8, 6, 8, 9, 6, 10, 10, 6, 5, 8, 8, 10, 6, 5, 9),
                                    mode = c("homodimer", "homodimer", "homodimer", "homodimer", "homodimer", "homodimer", "heterodimer", 
                                             "heterodimer", "homodimer", "heterodimer", "heterodimer", "heterodimer", 
                                             "monomer", "tetramer", "tetramer", "heterodimer"))

weight_spacer <- weight2 %>%
  left_join(actual_spacer_lengths) %>%
  mutate(center_center_dist = center_center_dist + 5, monomer_center_dist = monomer_center_dist + 5)

ggplot(weight_spacer,
       aes(x = monomer_center_dist, y = weight)) +
  geom_hline(yintercept = 0, lty = 2) +
  geom_smooth(method = "lm", se = F) +
  geom_point(size = 2) +
  theme_pubr(border = T)

ggplot(weight_spacer,
       aes(x = center_center_dist, y = weight)) +
  geom_hline(yintercept = 0, lty = 2) +
  geom_smooth(method = "lm", se = F, color = "#E07A5F") +
  geom_point(size = 2, aes(shape = mode)) +
  theme_pubr(border = T)


## Overview figures: which features matter how much and for which TF?
exp_variance_plot <- exp_variance %>%
  filter(tf %in% chosen_conditions$tf_condition) %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  mutate(text = paste("TF =", tf))

total_variance <- exp_variance_plot %>%
  filter(feature == "total") %>%
  dplyr::select(tf, "total_variance" = rel_sum_sq)

exp_variance_plot <- exp_variance_plot %>%
  filter(feature != "total") %>%
  left_join(total_variance) %>%
  filter(total_variance > 50) %>%
  mutate(mean_rel_sum_sq = ave(rel_sum_sq, feature, FUN = mean))

variance_plot <- ggplot(exp_variance_plot,
         aes(x = reorder(feature, -mean_rel_sum_sq), y = rel_sum_sq, color = total_variance)) +
  geom_quasirandom(aes(text = text)) +
  stat_summary(fun = median, fun.min = median, fun.max = median, geom = "crossbar", color = "red", width = 0.25, lwd = 0.4) +
  ylab("Variance explained (%)") + xlab("") +
  labs(title = "Linear variance modelling",
       subtitle = "Expression variance predicted by features") +
  theme(text = element_text(size = 12)) +
  scale_color_gradient(low = "grey90", high = "grey10") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12),
        axis.text.y = element_text(size = 10))

ggplotly(p = variance_plot, tooltip = "text")

exp_variance2 <- exp_variance %>%
  filter(tf %in% best_models$tf_condition) %>%
  distinct(tf, feature, rel_sum_sq) %>%
  spread(key = "feature", value = "rel_sum_sq") %>%
  column_to_rownames("tf")

pheatmap(exp_variance2 %>%
                  t(),
         cellwidth = 10,
         cellheight = 10,
         border_color = F,
         color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
         angle_col = 90,
         cluster_cols = T,
         cutree_rows = 2)



## Update reporter quality score by implementing predictability
## Add reporter feature information
reporter_features <- cDNA_df %>%
  filter(native_enhancer == "No", neg_ctrls == "No", hPGK == "No") %>%
  distinct(reporter_id, promoter, spacing, distance, background)

cDNA_df6 <- cDNA_df6 %>%
  left_join(reporter_features) 

# ## Correct confidence score by linear model pvalue
# chosen_models2 <- chosen_models %>% 
#   mutate(tf_condition = paste(tf, condition, sep = "_"))
pval_prediction <- prediction %>%
  filter(tf_condition %in% chosen_conditions$tf_condition) %>%
  distinct(tf, pval) %>%
  mutate(pval = -log10(pval)) %>%
  mutate(pval = ifelse(pval > 2, 2, pval)) %>%
  dplyr::select(tf, "lm_pval" = pval)

cDNA_df6 <- cDNA_df6 %>%
  distinct() %>%
  left_join(pval_prediction) %>%
  mutate(lm_pval = ifelse(is.na(lm_pval), 1, lm_pval)) %>%
  mutate(quality_score = ifelse(lm_pval < -log10(0.1), quality_score / 2, quality_score))

predicted_activities <- prediction %>%
  filter(tf_condition %in% chosen_conditions$tf_condition) %>%
  distinct(reporter_id, activity_predicted) 

cDNA_df6 <- cDNA_df6 %>%
  left_join(predicted_activities)

# write_tsv(cDNA_df6, "/DATA/usr/m.trauernicht/projects/SuRE-TF/data/mt20230302_reporter_scores.tsv")
# write_tsv(cDNA_df_TF3, "/DATA/usr/m.trauernicht/projects/SuRE-TF/data/mt20230302_predicted_activities.tsv")
# write_tsv(weight, "/DATA/usr/m.trauernicht/projects/SuRE-TF/data/mt20230302_lm_weights.tsv")

cDNA_df7 <- cDNA_df6 %>%
  distinct() %>%
  mutate(length_tf = as.numeric(ave(tf, tf, FUN = length))) %>%
  #filter(length_tf > 15) %>%
  mutate(mean_quality_score = ave(quality_score, tf, FUN = mean)) %>%
  arrange(desc(mean_quality_score)) %>%
  mutate(quality_score2 = -quality_score) %>%
  mutate(score_rank = as.numeric(ave(quality_score2, quality, tf, FUN = rank))) %>%
  mutate(top_3 = ifelse(score_rank <= 3, "Yes", "No")) %>%
  mutate(top_3 = ifelse(quality_score < 2, "No", top_3)) %>%
  left_join(pval_prediction) #%>%
  #mutate(top_3 = ifelse(lm_pval >= 1, top_3, "Insignificant"))

ggplot(cDNA_df7 %>%
         filter(quality == "Sensitivity"),
       aes(y = quality_score, x = reorder(tf, -mean_quality_score))) +
  geom_hline(yintercept = 2, color = "red") +
  geom_quasirandom_rast(aes(shape = commercial_reporter, color = top_3), raster.dpi = 600, width = 0.25) +
  stat_summary(fun = mean, fun.min = mean, fun.max = mean, geom = "crossbar", color = "black", width = 0.25, lwd = 0.4) +
  theme_pubr(border = T, x.text.angle = 90) +
  scale_color_manual(values = c("Yes" = "#9AC1AE", "No" = "#264653", "Insignificant" = "grey"))

ggplot(cDNA_df7 %>%
         filter(quality == "Specificity"),
       aes(y = quality_score, x = reorder(tf, -mean_quality_score))) +
  geom_hline(yintercept = 2, color = "red") +
  geom_quasirandom_rast(aes(shape = commercial_reporter, color = commercial_reporter), raster.dpi = 600, width = 0.25) +
  stat_summary(fun = mean, fun.min = mean, fun.max = mean, geom = "crossbar", color = "black", width = 0.25, lwd = 0.4) +
  theme_pubr(border = T, x.text.angle = 90) +
  scale_color_manual(values = c("Yes" = "#A5A58D", "No" = "black"))
```


## Reporter quality score computation

Aim: We want to find for each TF the optimal reporters. Here I combine all data to compute reporter quality scores to find the best reporters for each TF.
```{r}
## Prepare data frame
cDNA_df2 <- cDNA_df %>%
  #filter(tf != "RFX1") %>%
  dplyr::filter(neg_ctrls == "No", 
                #commercial_reporter == "No", 
                hPGK == "No", 
                str_detect(tf, "RANDOM", negate = T), 
                native_enhancer == "No") %>%
  distinct(tf, condition, reporter_activity_neg, stimulation, reference_condition, cell, reporter_activity_neg, reporter_id, commercial_reporter) %>%
  mutate(reporter_activity_neg = log2(reporter_activity_neg)) %>%
  distinct()

# Export for shiny app
#write_tsv(cDNA_df2, "/DATA/usr/m.trauernicht/projects/SuRE-TF/data/mt20230302_all_activities.tsv")

## Make separate data frame for reference conditions - this will be used for normalizing the other conditions
reference_condition <- cDNA_df2 %>%
  filter(condition %in% reference_condition) %>%
  dplyr::select(-reference_condition, "reference_condition" = condition, -stimulation, "reference_activity" = reporter_activity_neg, reporter_id) %>%
  mutate(reference_activity = ave(reference_activity, reporter_id, reference_condition, FUN = mean)) %>%
  distinct()

## Normalize KD conditions
cDNA_df3 <- cDNA_df2 %>% 
  left_join(reference_condition) %>%
  filter(stimulation == "KD") %>%
  mutate(condition = gsub("HepG2_|mES_", "", condition)) %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  filter(tf == condition) %>%
  mutate(reporter_dif =  reference_activity - reporter_activity_neg)

## Normalize pathway stimulations
stimulations <- data.frame(condition = c("HepG2_CDCA", "HepG2_Rif", "HepG2_T09", "HepG2_Wy", "mES_2i", "mES_BX", "mES_HQ", "mES_IWP2", "mES_LIF_CH", "mES_LIF_PD", "mES_LIF_PD",
                                          "mES_vitC", "A549_Dex_100", "A549_Nutlin", "mES_Bmp4", "mES_Dex", "mES_FK", "mES_LPS", "mES_N2B27_Bmp4", "mES_N2B27_RA", "mES_N2B27_Shh", 
                                         "mES_Nutlin", "mES_RA", "mES_Shh", "mES_vitA", "U2OS_FK", "U2OS_LPS",
                                         "HCT116_cDIM12", "U2OS_Calcitriol", "U2OS_PMA", "HCT116_GSK4112", "MCF7_Hexestrol", "MCF7_DIMc", 
                                         "HCT116_Calcimycin", "HCT116_Calcimycin", "HCT116_SR8278", "HCT116_SRI011381", "HCT116_ITE", "U2OS_Aldosterone", "U2OS_SRI1078", "U2OS_Cytosporone",
                                         "U2OS_AICAR", "U2OS_Wortmannin", "U2OS_Heat", "A549_5aDHT", "A549_TCPOBOP", "K562_T3", "K562_T3", "mES_Progesterone", "mES_GSK4716"),
                           tf = c("NR1H4::RXRA", "NR1I2", "NR1H2", "PPARA::RXRA", "STAT3", "IRF3", "NFE2L2", "OTX1", "EGR1", "TCF7", "TCF7L2", "NFE2L2", "NR3C1", "TP53", "SMAD4", "NR3C1", 
                                  "CREB1", "IRF3", "SMAD4","RARA", "GLI1", "TP53", "RARA", "GLI1", "RARA", "CREB1", "IRF3", "NR4A2::RXRA", "VDR::RXRA", "NFKB1", "NR1D1", "ESR1", "NR4A1",
                                  "NFAT5", "NFATc1", "NR1D1", "SMAD2::SMAD3::SMAD4", "AHR::ARNT", "NR3C2", "RORA", "NR4A1", "TEAD", "FOXO1", "HSF1", "AR", "NR1I3", "THRA", "THRB", "PGR", "ESRRB"),
                           effect_size = c(1,1,1,1,0,0,1,1,1,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1))

cDNA_df4 <- cDNA_df2 %>%
  left_join(reference_condition) %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  filter(stimulation == "pathway") %>%
  mutate(reporter_dif =  reference_activity - reporter_activity_neg)

cDNA_df4 <- merge(cDNA_df4, stimulations) ## Only keep relevant TF reporters

cDNA_df4 <- cDNA_df4 %>%
  mutate(reporter_dif = ifelse(effect_size == 1, -reporter_dif, reporter_dif)) %>% ## Change sign of difference, based on whether the stimulation should cause an increase or decrease (effect_size)
  dplyr::select(-effect_size)

## Normalize degron conditions
cDNA_df_degron <- cDNA_df2 %>% 
  mutate(tf = gsub("_.*", "", tf)) %>%
  left_join(reference_condition) %>%
  filter(stimulation == "degron") %>%
  filter(condition != "mES_Nanog_AID") %>% ## Doesn't have a direct target (I don't have Nanog reporters)
  mutate(target_tf = toupper(gsub("mES_(.*)_AID", "\\1", condition)))

cDNA_df_degron2 <- cDNA_df_degron
cDNA_df_degron2$target_tf[cDNA_df_degron2$condition == "mES_Pou5f1_AID"] <- "SOX2"
cDNA_df_degron3 <- cDNA_df_degron
cDNA_df_degron3$target_tf[cDNA_df_degron3$condition == "mES_Pou5f1_AID"] <- "POU5F1::SOX2"
cDNA_df_degron4 <- cDNA_df_degron
cDNA_df_degron4$target_tf[cDNA_df_degron4$condition == "mES_Sox2_AID"] <- "POU5F1::SOX2"
cDNA_df_degron <- rbind(cDNA_df_degron, cDNA_df_degron2, cDNA_df_degron3, cDNA_df_degron4)
cDNA_df_degron <- cDNA_df_degron %>% 
  filter(tf == target_tf) %>%
  mutate(reporter_dif =  reference_activity - reporter_activity_neg) %>%
  dplyr::select(-target_tf)

## Combine stimulations and KDs 
perturbation_df <- rbind(cDNA_df3, cDNA_df4, cDNA_df_degron) %>% ## Combine all perturbation data
  mutate(condition_2 = paste(condition, reference_condition)) %>%
  mutate(mean_reporter_dif = ave(reporter_dif, condition_2, tf, FUN = function(x) mean(x, na.rm = T))) %>%
  group_by(tf) %>%
  arrange(desc(mean_reporter_dif)) %>% 
  top_n(1, wt = mean_reporter_dif) %>% ## Some TFs have multiple perturbations: for those I only keep the strongest perturbations (on average across all reporters)
  ungroup() %>%
  mutate(reporter_dif = ifelse(tf == "CEBPB", -reporter_dif, reporter_dif)) %>%
  distinct(reporter_id, tf, commercial_reporter, condition, reporter_dif)

## Besides the effect of the perturbation I also want to look at the raw reporter activities - for this I will consider the highest activities (over identical mutated control) in any cell type 

### Define optimal condition for each TF: either the highest expressing cell line, or the stimulated condition
reference_activities1 <- cDNA_df4 %>%
  right_join(stimulations) %>%
  filter(effect_size == 1) %>%
  distinct(tf, condition, cell, reporter_id, commercial_reporter, reporter_activity_neg) %>%
  group_by(tf, condition) %>%
  dplyr::arrange(desc(reporter_activity_neg)) %>%
  top_n(10) %>%
  ungroup() %>%
  mutate(median_reporter_activity_neg = ave(reporter_activity_neg, tf, condition, FUN = median)) %>%
  group_by(tf) %>%
  arrange(desc(median_reporter_activity_neg)) %>%
  top_n(1) %>%
  ungroup() %>%
  distinct(tf, condition)

reference_activities <- cDNA_df4 %>%
  distinct(tf, condition, cell, reporter_id, commercial_reporter, reporter_activity_neg) %>%
  right_join(reference_activities1) %>%
  dplyr::select(-cell) %>%
  setnames("condition", "cell")


reference_activities2 <- merge(cDNA_df2 %>% filter(stimulation == "no") %>% mutate(tf = gsub("_.*", "", tf)), tf_rna, by = c("cell", "tf")) %>%
  filter(TPM > 50) %>%
  mutate(mean_reporter_activity_neg = ave(reporter_activity_neg, tf, condition, FUN = mean)) %>%
  group_by(tf) %>%
  dplyr::arrange(desc(mean_reporter_activity_neg)) %>%
  top_n(1, wt = mean_reporter_activity_neg) %>%
  ungroup()


reference_activities2 <- merge(cDNA_df2 %>% filter(stimulation == "no") %>% mutate(tf = gsub("_.*", "", tf)), tf_rna, by = c("cell", "tf")) %>%
  mutate(mean_reporter_activity_neg = ave(reporter_activity_neg, tf, condition, FUN = mean)) %>%
  group_by(tf) %>%
  dplyr::arrange(desc(TPM)) %>%
  top_n(1, wt = TPM) %>%
  ungroup() %>%
  group_by(reporter_id) %>%
  dplyr::arrange(desc(mean_reporter_activity_neg)) %>%
  top_n(1, wt = mean_reporter_activity_neg) %>%
  ungroup() %>%
  mutate(reporter_activity_neg = ave(reporter_activity_neg, reporter_id, cell)) %>%
  distinct(tf, cell, reporter_id, commercial_reporter, reporter_activity_neg) %>%
  filter(!tf %in% unique(reference_activities$tf)) %>%
  mutate(reporter_activity_neg = ave(reporter_activity_neg, reporter_id, FUN = mean))

reference_activities <- rbind(reference_activities, reference_activities2)

# reference_activities <- merge(cDNA_df2 %>% mutate(tf = gsub("_.*", "", tf)), chosen_models, by = c("condition", "tf")) %>%
#   distinct(tf, cell, reporter_id, commercial_reporter, reporter_activity_neg)

## Include reporter specificities based on KD of related TFs: I want to punish reporters that reduce activity upon KD of related TFs
related_tfs <- data.frame(tf = c("CREB1", "FOS::JUN", "CEBPB", "XBP1", "CEBPB", "FOS::JUN", "CREB1", "ELK1", "ETS2", "ELK1", "ETS2", "ELK1", "FOXA1", "FOXA1", "GATA1", "GATA4", "GATA1", "HNF4A", "IRF3", "TP53", "IRX3",
                                 "KLF4", "SP1", "MAF::NFE2", "NFE2L2", "NFKB1", "NFYA", "NR1H2", "PPARA::RXRA", "NFKB1", "REL", "IRF3", "STAT3", "TCF7", "TCF7L2", "TFCP2L1", "TFCP2L1", "GRHL1", "THRB", "THRA", "TP53", "TFCP2L1",
                                 "E2F1", "E2F1", "SP1", "EGR1", "EGR1", "KLF4", "KLF4", "SP1", "SP1", "NFIA", "HSF1", "KLF4", "MEF2A", "PPARG::RXRA", "ESRRB", "POU2F1", "POU5F1", "POU5F1::SOX2", "PPARG::RXRA",
                                 "PPARA::RXRA", "NR1H2", "NR1H2", "SOX2", "SOX9", "SOX2", "SOX9", "TCF7", "TCF7L2", "TEAD1", "STAT1::STAT2"),
                          related_tf = c("ATF2", "ATF2", "ATF4", "ATF6", "CEBPG", "CREB1", "FOS::JUN", "ETS1", "ELK1", "ETS2", "ETS1", "ETV4", "FOXA2", "FOXP1", "GATA2", "GATA2", "GATA4", "HNF4G", "IRF2", "IRX3", "TP53",
                                         "SP1", "KLF4", "NFE2L2", "MAF::NFE2", "NFKB2", "NFYB", "PPARA", "NR1H2", "REL", "NFKB1", "STAT1::STAT2", "STAT1::STAT2", "TCF7L2", "TCF7", "TFCP2", "GRHL1", "TFCP2L1", "THRA", "THRB", 
                                         "TP73", "UBP1", "E2F2", "E2F3", "EGR1", "SP1", "KLF4", "SP1", "EGR1", "KLF4", "EGR1", "HIC2", "HSF2", "KLF9", "MEF2B", "NR1H2", "NR5A2", "POU5F1", "POU2F1", "POU2F1", "PPARA::RXRA",
                                         "PPARG::RXRA", "PPARA::RXRA", "PPARG::RXRA", "SOX13", "SOX13", "SOX9", "SOX2", "TCF7L1", "TCF7L1", "TEAD2", "STAT3"))

off_target_activities <- cDNA_df2 %>% 
  left_join(reference_condition) %>%
  filter(stimulation == "KD") %>%
  mutate(condition = gsub("HepG2_|mES_", "", condition)) %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  mutate(reporter_dif =  reference_activity - reporter_activity_neg) %>%
  left_join(related_tfs) %>%
  na.omit() %>%
  filter(condition == related_tf) %>%
  mutate(reporter_dif = ave(reporter_dif, reporter_id, FUN = mean)) %>% ## Some TFs have multiple related TFs for which I have KD data: I will compute the mean over the related TFs per reporter
  dplyr::select(reporter_id, tf, commercial_reporter, "off_target_dif" = reporter_dif, "off_target" = related_tf) %>%
  distinct() %>%
  group_by(reporter_id) %>%
  top_n(1) %>%
  ungroup() 

## Also: punish reporters that are active in cells where the TF is not expressed
off_target_activities2 <- cDNA_df2 %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  left_join(tf_rna) %>%
  filter(TPM == 0, stimulation == "no") %>%
  mutate(off_target_activity = ave(reporter_activity_neg, reporter_id, FUN = mean)) %>%
  distinct(reporter_id, off_target_activity, tf, commercial_reporter) %>%
  mutate(off_target_activity = ifelse(off_target_activity < 0, 0, off_target_activity)) %>%
  mutate(off_target = "Expression_0TPM")

off_target_activities <- merge(off_target_activities, off_target_activities2, all = T)
off_target_activities$off_target_dif[is.na(off_target_activities$off_target_dif)] <- 0
off_target_activities$off_target_activity[is.na(off_target_activities$off_target_activity)] <- 0

off_target_activities <- off_target_activities %>%
  pivot_longer(cols = c(off_target_dif, off_target_activity), names_to = "off_target_act") %>%
  dplyr::select(-off_target_act) %>%
  #mutate(off_target_dif = ave(value, reporter_id, FUN = function(x) sum(x))) %>%
  dplyr::select(reporter_id, "off_target_dif" = value, tf, off_target) %>%
  mutate(mean_off_target_dif = ave(off_target_dif, tf, off_target, FUN = mean)) %>%
  group_by(tf) %>%
  arrange(desc(mean_off_target_dif)) %>% 
  top_n(1, wt = mean_off_target_dif) %>% ## Some TFs have multiple perturbations: for those I only keep the strongest off-target effect (on average across all reporters)
  ungroup() %>%
  group_by(reporter_id) %>%
  arrange(desc(off_target_dif)) %>% 
  top_n(1, wt = off_target_dif) %>%
  distinct(reporter_id, off_target_dif, off_target)


cDNA_df6 <- reference_activities %>%
  left_join(perturbation_df) %>%
  left_join(rna_protein_cor) %>%
  #left_join(viper_correlations) %>%
  left_join(off_target_activities) %>%
  mutate(reporter_dif = ifelse(is.na(reporter_dif), 0, reporter_dif)) %>%
  mutate(cor_pval = base::pmax(rna_cor, prot_cor)) %>%
  #mutate(viper_cor_pval = ifelse(is.na(viper_cor_pval), 0, viper_cor_pval)) %>%
  mutate(off_target_dif = ifelse(is.na(off_target_dif), 0, off_target_dif)) %>%
  mutate(off_target_dif = -off_target_dif) %>%
  mutate(Sensitivity = reporter_dif + reporter_activity_neg + cor_pval + off_target_dif) %>%
  mutate(Specificity = reporter_dif + off_target_dif + cor_pval) %>%
  pivot_longer(cols = c("Sensitivity", "Specificity"), names_to = "quality", values_to = "quality_score") %>%
  mutate(quality_score = ifelse(quality_score < 0, 0, quality_score))
```


## Figure 2: Plot activities in chosen conditions

Aim: Get a good overview of reporter activities
```{r}
## Quasirandom plots of reporter activities
cDNA_df_plot <- cDNA_df6 %>%
  distinct(tf, reporter_id, reporter_activity_neg, condition, cell) %>%
  mutate(cell = gsub("_.*", "", cell)) %>%
  mutate(mean_tf_activity = ave(reporter_activity_neg, tf, FUN = mean)) %>%
  mutate(cell = factor(cell, levels = c("A549", "HCT116", "HEK293", "HepG2", "K562",
                                        "MCF7", "U2OS", "mES", "NPC"))) %>%
  mutate(n_tf = as.numeric(ave(tf, tf, FUN = length))) %>%
  filter(n_tf > 10)

cell_colors <- c("A549" = "#f2cc8f", "HCT116" = "#babf95", "HEK293" = "#f4f1de", 
                               "HepG2" = "#5f797b", "K562" = "#8f5d5d", "MCF7" = "#81b29a", 
                               "U2OS" = "#3d405b", "mES" = "#eab69f", "NPC" = "#e07a5f") %>%
  as.data.frame() %>%
  rownames_to_column("cell")

tfs_colors <- cDNA_df_plot %>%
  distinct(tf, cell, mean_tf_activity) %>%
  arrange(desc(mean_tf_activity)) %>%
  left_join(cell_colors)
  
  
tfs_colors_vec <- tfs_colors %>%
  column_to_rownames("tf") %>%
  dplyr::select(-cell, -mean_tf_activity) %>%
  t() %>%
  base::as.vector()

names(tfs_colors_vec) <- tfs_colors$tf


ggplot(cDNA_df_plot,
       aes(x = reorder(tf, -mean_tf_activity), y = reporter_activity_neg)) +
  geom_hline(yintercept = 1, lty = 3) + 
  geom_hline(yintercept = -1, lty = 3) + 
  geom_quasirandom(width = .2, aes(color = ifelse(abs(reporter_activity_neg) > 1, "black", "grey70")),
                   alpha = .6) +
  stat_summary(fun = median, fun.min = median, fun.max = median, geom = "crossbar", color = "red", width = 0.5, lwd = 0.4) +
  theme_pubr(x.text.angle = 90) +
  scale_color_manual(values = c("black", "grey70")) +
  theme(axis.text.x = element_text(colour=tfs_colors_vec))

## Plot corresponding cell types
ggplot(cDNA_df_plot,
       aes(x = reorder(tf, -mean_tf_activity), y = "x", fill = cell)) +
  geom_tile() +
  scale_fill_manual(values = c("A549" = "#f2cc8f", "HCT116" = "#babf95", "HEK293" = "#2a9d8f", 
                               "HepG2" = "#8ab17d", "K562" = "#e9c46a", "MCF7" = "#efb366", 
                               "U2OS" = "#f4a261", "mES" = "#ee8959", "NPC" = "#e76f51")) +
  theme_pubr(x.text.angle = 90)
```

## Figure 4: Showcase how reporter activities change in different conditions
```{r}
## Showcase example
ggplot(cDNA_df %>% 
         filter(str_detect(tf, "VDR"), condition %in% c("U2OS", "U2OS_Calcitriol"), neg_ctrls == "No", promoter != "Random") %>%
         distinct(reporter_id, commercial_reporter, condition, reporter_activity_neg, promoter) %>%
         spread(key = "condition", value = "reporter_activity_neg"),
       aes(x = log2(U2OS),
           y = log2(U2OS_Calcitriol), 
           shape = commercial_reporter, 
           color = promoter)) +
  geom_hline(yintercept = 0, lty = 2) + 
  geom_vline(xintercept = 0, lty = 2) +
  geom_abline(lty = "twodash") +
  geom_point(size = 2.5) +
  xlim(-0.5, 8) +
  ylim(-0.5, 8) +
  xlab("Activity (log2)") +
  ylab(paste("Perturbation effect (log2)")) +
  theme_pubr(border = T) +
  scale_color_manual(values = c("mCMV" = "black", "hBGm" = "grey50", "minP" = "grey80"))

ggplot(cDNA_df %>% 
         filter(str_detect(tf, "NFE2L2"), condition %in% c("HepG2", "HepG2_NFE2L2"), neg_ctrls == "No") %>%
         distinct(reporter_id, commercial_reporter, spacing, condition, reporter_activity_neg) %>%
         spread(key = "condition", value = "reporter_activity_neg"),
       aes(x = log2(HepG2),
           y = log2(HepG2_NFE2L2), 
           shape = commercial_reporter,
           color = spacing)) +
  geom_hline(yintercept = 0, lty = 2) +
  geom_vline(xintercept = 0, lty = 2) +
  geom_abline(lty = "twodash") +
  geom_point(size = 2.5) +
  #xlim(-0.5, 3.5) +
  #ylim(-0.5, 3.5) +
  xlab("Activity (log2)") +
  ylab(paste("Perturbation effect (log2)")) +
  theme_pubr(border = T) +
  scale_color_manual(values = c("#7C9EB2", "#BFD0D9"))

ggplot(cDNA_df %>% 
         filter(str_detect(tf, "TP53"), condition %in% c("HepG2", "HepG2_TP73"), neg_ctrls == "No", promoter != "Random") %>%
         distinct(reporter_id, commercial_reporter, spacing, condition, promoter, reporter_activity_neg) %>%
         spread(key = "condition", value = "reporter_activity_neg"),
       aes(x = log2(HepG2),
           y = log2(HepG2_TP73), 
           shape = commercial_reporter)) +
  geom_hline(yintercept = 0, lty = 2) +
  geom_vline(xintercept = 0, lty = 2) +
  geom_abline(lty = "twodash") +
  geom_point(size = 2.5, aes(color = promoter)) +
  xlab("Activity (log2)") +
  ylab(paste("Off-target effect (log2)")) +
  theme_pubr(border = T) +
  scale_color_manual(values = c("mCMV" = "black", "hBGm" = "grey50", "minP" = "grey80"))
```


---




## Explore activities using set of top reporters
```{r, fig.width=9, fig.height=9}
## Set of best reporters
optimal_reporters <- cDNA_df7 %>%
  filter(top_3 == "Yes")


### Activity across the 9 cell types
heatmap_cell_activities <- cDNA_df %>%
  filter(reporter_id %in% optimal_reporters$reporter_id) %>%
  dplyr::filter(neg_ctrls == "No", 
                stimulation == "no",
                hPGK == "No", 
                str_detect(tf, "RANDOM", negate = T), 
                native_enhancer == "No") %>%
  distinct(tf, reporter_activity_neg, cell) %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  mutate(reporter_activity_neg = log2(ave(reporter_activity_neg, tf, cell, FUN = mean))) %>%
  distinct() %>%
  spread(key = "tf", value = "reporter_activity_neg") %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames("cell")

myBreaks <- c(seq(-1, 0, length.out=50+1), 
              seq(0.000001, 5, length.out=50))

pheatmap(heatmap_cell_activities %>% 
           as.matrix(),
         color = colorRampPalette(c("#F2CC8F", "white", "#3D405B"))(100),
         clustering_method = "ward.D2",
         border_color = F, 
         cellwidth = 8, 
         cellheight = 8,
         angle_col = 90,
         cutree_cols = 6,
         #cutree_rows = 3,
         breaks = myBreaks)


### Activities upon KD
cDNA_df2 <- cDNA_df %>%
  filter(reporter_id %in% optimal_reporters$reporter_id) %>%
  dplyr::filter(neg_ctrls == "No", 
                #commercial_reporter == "No", 
                hPGK == "No", 
                str_detect(tf, "RANDOM", negate = T), 
                native_enhancer == "No") %>%
  distinct(tf, condition, reporter_activity_neg, stimulation, reference_condition, cell) %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  mutate(reporter_activity_neg = log2(ave(reporter_activity_neg, tf, condition, FUN = mean))) %>%
  distinct()

reference_condition <- cDNA_df2 %>%
  filter(condition %in% reference_condition) %>%
  dplyr::select(-reference_condition, "reference_condition" = condition, -stimulation, "reference_activity" = reporter_activity_neg) #%>%
  #filter(abs(reference_activity) > 1)

cDNA_df3 <- cDNA_df2 %>%
  filter(cell %in% c("HepG2", "mES")) %>%
  filter(stimulation == "KD") %>%
  left_join(reference_condition) %>%
  mutate(reporter_FC = reporter_activity_neg - reference_activity) %>%
  mutate(TF_dif = ave(reporter_FC, tf, condition, FUN = mean)) %>%
  mutate(TF_dif = ifelse(abs(TF_dif) > 0.2, TF_dif, 0))

cDNA_df3[is.na(cDNA_df3)] <- 0

cDNA_df4 <- cDNA_df3 %>%
  distinct(tf, condition, TF_dif, cell) %>%
  mutate(condition = gsub("HepG2_", "", condition)) %>%
  mutate(condition = gsub("mES_", "", condition)) %>%
  filter(condition %in% tf) %>%
  filter(tf %in% condition) %>%
  mutate(diag = ifelse(tf == condition, T, F)) 

KD_doubles <- cDNA_df4 %>%
  filter(diag == T) %>%
  mutate(n_kd = ave(tf, tf, FUN = length)) %>%
  filter(n_kd == 2) %>%
  group_by(tf) %>%
  dplyr::arrange(TF_dif, .by_group = T) %>%
  dplyr::slice(n = 1, .preserve = T) %>%
  ungroup() %>%
  mutate(condition2 = paste(condition, cell, sep = "_"))

KD_singles <- cDNA_df4 %>%
  filter(diag == T) %>%
  mutate(n_kd = ave(tf, tf, FUN = length)) %>%
  filter(n_kd == 1) %>%
  mutate(condition2 = paste(condition, cell, sep = "_"))

cDNA_df5 <- cDNA_df4 %>%
  mutate(condition2 = paste(condition, cell, sep = "_")) %>%
  filter(condition2 %in% KD_doubles$condition2 | condition2 %in% KD_singles$condition2) %>%
  filter(!condition %in% c("TP53"), !tf %in% c("TP53")) %>%
  arrange(diag)
 
ggplot(cDNA_df5,
       aes(x = tf, y = condition, fill = TF_dif)) +
  geom_tile(color = cDNA_df5$diag, size = .5) +
  scale_color_manual(values = c("black", "white")) +
  scale_fill_gradient2(low = "#99B2DD",
                        mid = "white",
                        high = "#dd6b48",
                       limits = c(-1.5,1.5),
                       oob = squish) +
  coord_fixed() +
  xlab("TF reporter") +
  ylab("KD condition") +
  theme_pubr(border = T, x.text.angle = 90) 


### TF activity changes upon stimulation 
cDNA_df2 <- cDNA_df %>%
  filter(reporter_id %in% optimal_reporters$reporter_id) %>%
  dplyr::filter(neg_ctrls == "No", 
                #commercial_reporter == "No", 
                hPGK == "No", 
                str_detect(tf, "RANDOM", negate = T), 
                native_enhancer == "No") %>%
  distinct(tf, condition, reporter_activity_neg, stimulation, reference_condition, cell) %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  mutate(reporter_activity_neg = log2(ave(reporter_activity_neg, tf, condition, FUN = mean))) %>%
  distinct()

reference_condition <- cDNA_df2 %>%
  filter(condition %in% reference_condition) %>%
  dplyr::select(-reference_condition, "reference_condition" = condition, -stimulation, "reference_activity" = reporter_activity_neg) #%>%
  #filter(abs(reference_activity) > 1)

stimulations <- stimulations %>%
  distinct(condition, tf) %>%
  mutate(target = "Yes")

cDNA_df3 <- cDNA_df2 %>%
  filter(stimulation == "pathway", condition %in% c("HCT116_Calcymicin", "HepG2_CDCA", "K562_Hemin", "K562_T3", "MCF7_Hexestrol", "mES_2i", "mES_BX", "mES_HQ",
                                                    "mES_IWP2", "mES_LIF_CH", "mES_LIF_PD", "mES_N2B27", "mES_serum_2i_LIF", "mES_vitC", "U2OS_Calcitriol", "U2OS_Heat",
                                                    "U2OS_PMA", "U2OS_Wortmannin")) %>%
  left_join(stimulations) %>%
  left_join(reference_condition) %>%
  mutate(reporter_FC = reporter_activity_neg - reference_activity) %>%
  mutate(TF_dif = ave(reporter_FC, tf, condition, FUN = mean)) %>%
  mutate(TF_dif = ifelse(abs(TF_dif) > 0.5, TF_dif, 0)) %>%
  mutate(mean_activity = ave(reporter_activity_neg, tf, condition, FUN = mean)) %>%
  mutate(mean_TF_dif = ave(TF_dif, condition, FUN = mean)) %>%
  mutate(mean_TF_dif2 = ave(TF_dif, tf, FUN = mean)) %>%
  filter(mean_TF_dif != 0, mean_TF_dif2 != 0)

target_change <- cDNA_df3 %>%
  filter(target == "Yes") %>%
  filter(TF_dif != 0)

cDNA_df3[is.na(cDNA_df3)] <- 0

cDNA_df3 <- cDNA_df3 %>%
  #filter(condition %in% target_change$condition) %>%
  distinct(tf, condition, TF_dif) %>%
  spread(tf, TF_dif) %>%
  column_to_rownames("condition")

myBreaks <- c(seq(-2, 0, length.out=50+1), 
              seq(0.000001, 2, length.out=50))

pheatmap(cDNA_df3 %>% 
                 as.matrix(),
         color = colorRampPalette(c("#99B2DD", "white", "#E07A5F"))(100),
         clustering_method = "ward.D",
         border_color = F, 
         cellwidth = 8, 
         cellheight = 8,
         #annotation_col = tf_clusters, 
         angle_col = 90,
         cutree_cols = 5,
         #cutree_rows = 3,
         breaks = myBreaks)

```













## Effect of TF perturbations of TF activity

```{r}
cDNA_df2 <- cDNA_df %>%
  filter(tf != "RFX1") %>%
  dplyr::filter(neg_ctrls == "No", 
                commercial_reporter == "No", 
                hPGK == "No", 
                str_detect(tf, "RANDOM", negate = T), 
                native_enhancer == "No") %>%
  distinct(tf, condition, reporter_activity_minP, stimulation, reference_condition) %>%
  mutate(reporter_activity_minP = log2(ave(reporter_activity_minP, tf, condition, FUN = mean))) %>%
  distinct()

reference_condition <- cDNA_df2 %>%
  filter(condition %in% reference_condition) %>%
  dplyr::select(-reference_condition, "reference_condition" = condition, -stimulation, "reference_activity" = reporter_activity_minP) #%>%
  #filter(abs(reference_activity) > 1)

cDNA_df3 <- cDNA_df2 %>%
  filter(stimulation == "degron") %>%
  left_join(reference_condition) %>%
  mutate(reporter_FC = reporter_activity_minP - reference_activity) %>%
  mutate(TF_dif = ave(reporter_FC, tf, condition, FUN = mean)) %>%
  mutate(TF_dif = ifelse(abs(TF_dif) > 0.5, TF_dif, 0)) %>%
  mutate(mean_activity = ave(reporter_activity_minP, tf, condition, FUN = mean)) #%>%
  #filter(mean_activity > 0.5)

cDNA_df3[is.na(cDNA_df3)] <- 0

cDNA_df3 <- cDNA_df3 %>%
  distinct(tf, condition, TF_dif) %>%
  spread(tf, TF_dif) %>%
  column_to_rownames("condition")%>%
  select_if(colSums(.) != 0)

myBreaks <- c(seq(-1.5, 0, length.out=50+1), 
              seq(0.000001, 1.5, length.out=50))

pheatmap(cDNA_df3 %>% 
                 as.matrix(),
         color = colorRampPalette(c("#99B2DD", "white", "#E07A5F"))(100),
         clustering_method = "ward.D",
         border_color = F, 
         cellwidth = 8, 
         cellheight = 8,
         #annotation_col = tf_clusters, 
         angle_col = 90,
         #cutree_cols = 5,
         #cutree_rows = 3,
         breaks = myBreaks)
```

## Effect of TF perturbations on TF activity

```{r}
cDNA_df2 <- cDNA_df %>%
  #filter(reporter_id %in% optimal_reporters$reporter_id) %>%
  filter(tf != "RFX1") %>%
  dplyr::filter(neg_ctrls == "No", 
                commercial_reporter == "No", 
                hPGK == "No", 
                str_detect(tf, "RANDOM", negate = T), 
                native_enhancer == "No") %>%
  distinct(tf, condition, reporter_activity_minP, stimulation, reference_condition, cell) %>%
  mutate(reporter_activity_minP = log2(ave(reporter_activity_minP, tf, condition, FUN = mean))) %>%
  distinct()

reference_condition <- cDNA_df2 %>%
  filter(condition %in% reference_condition) %>%
  dplyr::select(-reference_condition, "reference_condition" = condition, -stimulation, "reference_activity" = reporter_activity_minP) #%>%
  #filter(abs(reference_activity) > 1)

cDNA_df3 <- cDNA_df2 %>%
  filter(cell == "HepG2") %>%
  filter(stimulation == "KD") %>%
  left_join(reference_condition) %>%
  mutate(reporter_FC = reporter_activity_minP - reference_activity) %>%
  mutate(TF_dif = ave(reporter_FC, tf, condition, FUN = mean)) %>%
  mutate(TF_dif = ifelse(abs(TF_dif) > 0.2, TF_dif, 0)) %>%
  mutate(mean_activity = ave(reporter_activity_minP, tf, condition, FUN = mean)) #%>%
  #filter(mean_activity > 0.5)

cDNA_df3[is.na(cDNA_df3)] <- 0

cDNA_df4 <- cDNA_df3 %>%
  distinct(tf, condition, TF_dif) %>%
  spread(tf, TF_dif) %>%
  column_to_rownames("condition")%>%
  select_if(colSums(.) != 0)

myBreaks <- c(seq(-1, 0, length.out=50+1), 
              seq(0.000001, 1, length.out=50))

pheatmap(cDNA_df4 %>% 
                 as.matrix(),
         color = colorRampPalette(c("#99B2DD", "white", "#E07A5F"))(100),
         clustering_method = "ward.D",
         border_color = F, 
         cellwidth = 8, 
         cellheight = 8,
         #annotation_col = tf_clusters, 
         angle_col = 90,
         #cutree_cols = 5,
         #cutree_rows = 3,
         breaks = myBreaks)

cDNA_df4 <- cDNA_df3 %>%
  distinct(tf, condition, TF_dif) %>%
  mutate(condition = gsub("HepG2_", "", condition)) %>%
  filter(condition %in% tf) %>%
  filter(tf %in% condition) %>%
  mutate(diag = ifelse(tf == condition, T, F)) %>%
  arrange(diag)

ggplot(cDNA_df4,
       aes(x = tf, y = condition, fill = TF_dif)) +
  geom_tile(color = cDNA_df4$diag, size = .5) +
  scale_color_manual(values = c("black", "white")) +
  scale_fill_gradient2(low = "#99B2DD",
                        mid = "white",
                        high = "#dd6b48",
                       limits = c(-2,2),
                       oob = squish) +
  coord_fixed() +
  xlab("TF reporter") +
  ylab("KD condition") +
  ggtitle("Only TFs that are direct target of KD") +
  theme_pubr(border = T, x.text.angle = 90) +
  labs(fill="FC compared to control - mean of all reporters")
```


```{r}
cDNA_df2 <- cDNA_df %>%
  filter(tf != "RFX1") %>%
  dplyr::filter(neg_ctrls == "No", 
                commercial_reporter == "No", 
                hPGK == "No", 
                str_detect(tf, "RANDOM", negate = T), 
                native_enhancer == "No") %>%
  distinct(tf, condition, reporter_activity_minP, stimulation, reference_condition, cell) %>%
  mutate(reporter_activity_minP = log2(ave(reporter_activity_minP, tf, condition, FUN = mean))) %>%
  distinct()

reference_condition <- cDNA_df2 %>%
  filter(condition %in% reference_condition) %>%
  dplyr::select(-reference_condition, "reference_condition" = condition, -stimulation, "reference_activity" = reporter_activity_minP) #%>%
  #filter(abs(reference_activity) > 1)

cDNA_df3 <- cDNA_df2 %>%
  filter(cell == "mES") %>%
  filter(stimulation == "KD") %>%
  left_join(reference_condition) %>%
  mutate(reporter_FC = reporter_activity_minP - reference_activity) %>%
  mutate(TF_dif = ave(reporter_FC, tf, condition, FUN = mean)) %>%
  mutate(TF_dif = ifelse(abs(TF_dif) > 0.2, TF_dif, 0)) %>%
  mutate(mean_activity = ave(reporter_activity_minP, tf, condition, FUN = mean)) #%>%
  #filter(mean_activity > 0.5)

cDNA_df3[is.na(cDNA_df3)] <- 0

cDNA_df4 <- cDNA_df3 %>%
  distinct(tf, condition, TF_dif) %>%
  spread(tf, TF_dif) %>%
  column_to_rownames("condition")%>%
  select_if(colSums(.) != 0)

myBreaks <- c(seq(-1, 0, length.out=50+1), 
              seq(0.000001, 1, length.out=50))

pheatmap(cDNA_df4 %>% 
                 as.matrix(),
         color = colorRampPalette(c("#99B2DD", "white", "#E07A5F"))(100),
         clustering_method = "ward.D",
         border_color = F, 
         cellwidth = 8, 
         cellheight = 8,
         #annotation_col = tf_clusters, 
         angle_col = 90,
         #cutree_cols = 5,
         #cutree_rows = 3,
         breaks = myBreaks)

cDNA_df4 <- cDNA_df3 %>%
  distinct(tf, condition, TF_dif) %>%
  mutate(condition = gsub("mES_", "", condition)) %>%
  filter(condition %in% tf) %>%
  filter(tf %in% condition) %>%
  mutate(diag = ifelse(tf == condition, T, F)) %>%
  arrange(diag)

ggplot(cDNA_df4,
       aes(x = tf, y = condition, fill = TF_dif)) +
  geom_tile(color = cDNA_df4$diag, size = .5) +
  scale_color_manual(values = c("black", "white")) +
  scale_fill_gradient2(low = "#99B2DD",
                        mid = "white",
                        high = "#dd6b48",
                       limits = c(-1,1),
                       oob = squish) +
  coord_fixed() +
  xlab("TF reporter") +
  ylab("KD condition") +
  ggtitle("Only TFs that are direct target of KD") +
  theme_pubr(border = T, x.text.angle = 90) +
  labs(fill="FC compared to control - mean of all reporters")
```


## Compare degron data to KD data
```{r}
cDNA_df3 <- cDNA_df2 %>%
  filter(condition %in% c("mES_POU5F1", "mES_Pou5f1_AID")) %>%
  left_join(reference_condition) %>%
  mutate(reporter_FC = reporter_activity_minP - reference_activity) %>%
  mutate(TF_dif = ave(reporter_FC, tf, condition, FUN = mean)) %>%
  #mutate(TF_dif = ifelse(abs(TF_dif) > 0.2, TF_dif, 0)) %>%
  mutate(mean_activity = ave(reporter_activity_minP, tf, condition, FUN = mean)) %>%
  distinct(tf, TF_dif, condition) %>%
  spread(key = "condition", value = "TF_dif")

ggplot(cDNA_df3,
       aes(x = mES_POU5F1, y = mES_Pou5f1_AID)) +
  geom_point() +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm")

cDNA_df3 <- cDNA_df2 %>%
  filter(condition %in% c("mES_SOX2", "mES_Sox2_AID")) %>%
  left_join(reference_condition) %>%
  mutate(reporter_FC = reporter_activity_minP - reference_activity) %>%
  mutate(TF_dif = ave(reporter_FC, tf, condition, FUN = mean)) %>%
  #mutate(TF_dif = ifelse(abs(TF_dif) > 0.2, TF_dif, 0)) %>%
  mutate(mean_activity = ave(reporter_activity_minP, tf, condition, FUN = mean)) %>%
  distinct(tf, TF_dif, condition) %>%
  spread(key = "condition", value = "TF_dif")

ggplot(cDNA_df3,
       aes(x = mES_SOX2, y = mES_Sox2_AID)) +
  geom_point() +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm")
```










```{r warning = F}
## Heatmap: mean reporter activity per condition
cDNA_df4 <- cDNA_df %>%
  filter(stimulation == "no") %>%
  filter(tf != "RFX1") %>%
  dplyr::filter(neg_ctrls == "No", 
                commercial_reporter == "No", 
                hPGK == "No", 
                str_detect(tf, "RANDOM", negate = T), 
                native_enhancer == "No") %>%
  distinct(tf, condition, reporter_activity_minP) %>%
  mutate(reporter_activity_minP = ave(reporter_activity_minP, tf, condition, FUN = mean)) %>%
  mutate(log_reporter_activity_minP = log2(reporter_activity_minP)) %>%
  mutate(mean_tf_activity = ave(reporter_activity_minP, tf, FUN = mean)) %>%
  #filter(abs(mean_tf_activity) > 0.5) %>%
  distinct()

nt_df <- cDNA_df4 %>%
  filter(condition == "HepG2_NT") %>%
  mutate(log_reporter_activity_minP = ave(log_reporter_activity_minP, tf, FUN = mean)) %>%
  filter(abs(log_reporter_activity_minP) > 0.15) %>%
  dplyr::select(tf, "nt_activity" = reporter_activity_minP) %>%
  mutate(nt_activity = ave(nt_activity, tf, FUN = mean)) %>%
  distinct()


## First only plot identical matches to check direct KDs
cDNA_df5 <- merge(cDNA_df4, nt_df, all = T) %>%
  na.omit() %>%
  mutate(reporter_FC = log2(reporter_activity_minP/nt_activity)) %>%
  filter(condition %in% tf) %>%
  filter(tf %in% condition) %>%
  distinct(tf, condition, reporter_FC) %>%
  mutate(diag = ifelse(tf == condition, T, F)) %>%
  arrange(diag)

ggplot(cDNA_df5,
       aes(x = tf, y = condition, fill = reporter_FC)) +
  geom_tile(color = cDNA_df5$diag, size = .5) +
  scale_color_manual(values = c("black", "white")) +
  scale_fill_gradient2(low = "#99B2DD",
                        mid = "white",
                        high = "#dd6b48",
                       limits = c(-.5,.5),
                       oob = squish) +
  coord_fixed() +
  xlab("TF reporter") +
  ylab("KD condition") +
  ggtitle("Only TFs that are direct target of KD") +
  theme_pubr(border = T, x.text.angle = 90) +
  labs(fill="FC compared to control - mean of all reporters")


## Now also include indirect targets
cDNA_df5 <- merge(cDNA_df4, nt_df, all = T) %>%  na.omit() %>%
  mutate(reporter_FC = log2(reporter_activity_minP/nt_activity)) %>%
  #filter(condition %in% tf) %>%
  #filter(tf %in% condition) %>%
  distinct(tf, condition, reporter_FC)
 
tf_kd_targets <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf7027_KD_mES/tf_kd_targets_mES.csv") %>% 
  #filter(condition %in% cDNA_df5$condition) %>%
  setnames("target", "tf") %>%
  mutate(is_target = T) %>%
  right_join(cDNA_df5, by = c("condition", "tf")) 

tf_kd_targets$is_target[is.na(tf_kd_targets$is_target)] <- F
tf_kd_targets <- tf_kd_targets %>%
  filter(condition != "NT") %>%
  arrange(is_target)

ggplot(tf_kd_targets,
       aes(x = tf, y = condition, fill = reporter_FC)) +
  geom_tile(color = tf_kd_targets$is_target, size = .5) +
  scale_color_manual(values = c("black", "white")) +
  scale_fill_gradient2(low = "#99B2DD",
                        mid = "grey95",
                        high = "#dd6b48",
                       na.value = "white",
                       limits = c(-.5,.5),
                       oob = squish) +
  coord_fixed() +
  xlab("TF reporter") +
  ylab("KD condition") +
  ggtitle("All 'active' TFs - KD targets highlighted by box") +
  theme_pubr(border = T, x.text.angle = 90) +
  labs(fill="FC compared to control - mean of all reporters")


tf_kd_targets <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6952_KD_HepG2/tf_kd_targets.csv") %>% 
  #filter(condition %in% cDNA_df5$condition) %>%
  setnames("target", "tf") %>%
  mutate(is_target = T) %>%
  right_join(cDNA_df5, by = c("condition", "tf")) 

tf_kd_targets$is_target[is.na(tf_kd_targets$is_target)] <- F
tf_kd_targets <- tf_kd_targets %>%
  filter(condition != "NT") %>%
  arrange(is_target) %>%
  mutate(target_length = ave(is_target, tf, FUN = function(x) length(unique(x)))) %>%
  filter(target_length > 1) %>%
  dplyr::select(-target_length)

ggplot(tf_kd_targets,
       aes(x = tf, y = condition, fill = reporter_FC)) +
  geom_tile(color = tf_kd_targets$is_target, size = .5) +
  scale_color_manual(values = c("black", "white")) +
  scale_fill_gradient2(low = "#99B2DD",
                        mid = "grey95",
                        high = "#dd6b48",
                       na.value = "white",
                       limits = c(-1,1),
                       oob = squish) +
  coord_fixed() +
  xlab("TF reporter") +
  ylab("KD condition") +
  ggtitle("Only targeted TFs - KD targets highlighted by box") +
  theme_pubr(border = T, x.text.angle = 90) +
  labs(fill="FC compared to control - mean of all reporters")


tf_kd_targets <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6952_KD_HepG2/tf_kd_targets.csv") %>% 
  #filter(condition %in% cDNA_df5$condition) %>%
  setnames("target", "tf") %>%
  mutate(is_target = T) %>%
  left_join(cDNA_df5, by = c("condition", "tf")) 

tf_kd_targets$is_target[is.na(tf_kd_targets$is_target)] <- F
tf_kd_targets <- tf_kd_targets %>%
  filter(condition != "NT") %>%
  arrange(is_target)

ggplot(tf_kd_targets,
       aes(x = tf, y = condition, fill = reporter_FC)) +
  geom_tile(color = tf_kd_targets$is_target, size = .5) +
  scale_color_manual(values = c("black", "white")) +
  scale_fill_gradient2(low = "#99B2DD",
                        mid = "grey95",
                        high = "#dd6b48",
                       na.value = "white",
                       limits = c(-1,1),
                       oob = squish) +
  coord_fixed() +
  xlab("TF reporter") +
  ylab("KD condition") +
  ggtitle("Only KD targets") +
  theme_pubr(border = T, x.text.angle = 90) +
  labs(fill="FC compared to control - mean of all reporters")
```

Conclusion: KDs worked to varying degrees. Some were efficient and specific, others were unspecific, or didn't work at all.

---


## Interaction network between TFs
```{r}
# threshold <- 1
# 
# interaction_df <- cDNA_df4 %>%
#   dplyr::select("receiver" = tf, "signal" = condition, TF_dif) %>%
#   distinct() %>%
#   mutate(TF_dif_binary = "0") %>%
#   mutate(TF_dif_binary = ifelse(TF_dif > threshold, "-1", TF_dif_binary)) %>%
#   mutate(TF_dif_binary = ifelse(TF_dif < -threshold, "1", TF_dif_binary)) %>%
#   mutate(TF_dif_binary = ifelse(TF_dif > -threshold & TF_dif < threshold, "0", TF_dif_binary)) %>%
#   mutate(TF_dif_binary = as.numeric(TF_dif_binary)) %>%
#   mutate(TF_dif_binary = (ifelse(receiver == signal, 0, TF_dif_binary))) %>%
#   mutate(sum_receiver = ave(TF_dif_binary, receiver, FUN = function(x) sum(x))) %>%
#   mutate(sum_signal = ave(TF_dif_binary, signal, FUN = function(x) sum(x))) %>%
#   mutate(sum_receiver_size = ave(TF_dif_binary, receiver, FUN = function(x) sum(abs(x)))) %>%
#   mutate(sum_signal_size = ave(TF_dif_binary, signal, FUN = function(x) sum(abs(x)))) 
# 
# 
# interaction_matrix <- interaction_df %>%
#   distinct(TF_dif_binary, receiver, signal) %>%
#   spread(key = "receiver", value = "TF_dif_binary") %>%
#   column_to_rownames("signal")
# 
# pheatmap(interaction_matrix,
#          border_color = NA,
#          color = colorRampPalette(c("#d5bdaf", 'white', "#84a98c"))(100),
#          clustering_method = "ward.D")
# 
# 
# ###
# interaction_nodes <- interaction_df %>%
#   filter(signal == receiver) %>%
#   distinct(receiver, signal, sum_signal, sum_receiver, sum_signal_size, sum_receiver_size) %>%
#   mutate(sum_total = sum_receiver_size + sum_signal_size) %>%
#   mutate(class = ifelse(sum_signal <= 0, "repressor", "activator")) %>%
#   distinct(receiver, sum_total, class) 
# 
# 
# interaction_edges <- interaction_df %>%
#   filter(receiver != signal) %>%
#   distinct(receiver, signal, TF_dif_binary, TF_dif) %>%
#   mutate(weight = 0) %>%
#   mutate(weight = ifelse(TF_dif_binary == -1, 1, weight)) %>%
#   mutate(weight = ifelse(TF_dif_binary == 1, 2, weight)) %>%
#   mutate(class = ifelse(weight == 1, "repressor", "activator")) %>%
#   dplyr::select(signal, receiver, weight, class, TF_dif) %>%
#   filter(weight != 0) %>%
#   mutate(TF_dif = abs(TF_dif))
#   
# interaction_nodes <- interaction_nodes %>%
#   filter(receiver %in% c(unique(interaction_edges$receiver),unique(interaction_edges$signal)))
# 
# routes_igraph <- graph_from_data_frame(d = interaction_edges, vertices = interaction_nodes, directed = TRUE)
# 
# 
# visIgraph(routes_igraph) %>%
#     visNodes(size = 25, shape = "circle") %>%
#     visOptions(highlightNearest = list(enabled = TRUE, hover = TRUE),
#                nodesIdSelection = TRUE) %>%
#     visInteraction(keyboard = TRUE)
# 
# ggraph(routes_igraph, layout = "sugiyama") + 
#   geom_node_point(aes(size = sum_total, color = class)) +
#   geom_edge_link(aes(color = class, width = TF_dif), 
#                  alpha = 0.8,
#                  arrow = arrow(length = unit(3, 'mm')), 
#                  start_cap = circle(2, 'mm'),
#                  end_cap = circle(3, 'mm')) + 
#   scale_edge_width(range = c(0.2, 2)) +
#   scale_radius(limits = c(1, NA), range = c(1, 6)) + 
#   geom_node_text(aes(label = name), repel = TRUE) +
#   scale_edge_color_manual(values = c("activator" = colors_diverse[2], "repressor" = colors_diverse[1])) +
#   scale_color_manual(values = c("activator" = colors_diverse[2], "repressor" = colors_diverse[1])) +
#   theme_graph()

```




## Cistrome TF targets
```{r}
# bc_files = list.files('/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6952_KD_HepG2/cistrome/',
#                        full.names=T, patter='*.txt')
# bc_list <- lapply(bc_files, fread, header = T)
# names(bc_list)<- gsub('.*/cistrome//(.*?).txt', '\\1', bc_files)
# 
# cistrome_df <- bind_rows(bc_list, .id = "TF") %>%
#   distinct(score, symbol, TF)
# 
# cistrome_df <- cistrome_df %>%
#   filter(TF %in% interaction_df$receiver, symbol %in% interaction_df$receiver) %>%
#   mutate(score = ave(score, TF, symbol, FUN = mean)) %>%
#   distinct() %>%
#   filter(TF != symbol) %>%
#   filter(score > 2) %>%
#   mutate(sum_target = as.numeric(ave(symbol, symbol, FUN = function(x) length(x))))
# 
# cistrome_matrix <- cistrome_df %>%
#   dplyr::select(-sum_target) %>%
#   spread(key = "symbol", value = "score") %>%
#   column_to_rownames("TF")
# 
# cistrome_matrix[is.na(cistrome_matrix)] <- 0
# 
# pheatmap(cistrome_matrix,
#          border_color = NA,
#          color = colorRampPalette(c("#d5bdaf", 'white', "#84a98c"))(100),
#          clustering_method = "ward.D")
# 
# cistrome_nodes <- cistrome_df %>%
#   distinct(symbol, sum_target)
# 
# cistrome_edges <- cistrome_df %>%
#   filter(symbol %in% cistrome_nodes$symbol) %>%
#   distinct(TF, symbol, "weight" = score)
# 
# cistrome_igraph <- graph_from_data_frame(d = cistrome_edges, vertices = cistrome_nodes, directed = TRUE)
# 
# ggraph(cistrome_igraph, layout = "sugiyama") + 
#   geom_node_point(aes(size = sum_target)) +
#   geom_edge_link(aes(width = weight), 
#                  alpha = 0.8,
#                  arrow = arrow(length = unit(3, 'mm')), 
#                  start_cap = circle(2, 'mm'),
#                  end_cap = circle(3, 'mm')) + 
#   scale_edge_width(range = c(0.2, 2)) +
#   geom_node_text(aes(label = name), repel = TRUE) +
#   theme_graph()
```


## Cistrome TF targets
```{r}
# bc_files = list.files('/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6952_KD_HepG2/cistrome/',
#                        full.names=T, patter='*.csv')
# bc_list <- lapply(bc_files, fread, header = T)
# names(bc_list)<- gsub('.*/cistrome//.*M_.*_(.*?)hg38.*.csv', '\\1', bc_files)
# 
# cistrome_df <- bind_rows(bc_list, .id = "target") %>%
#   distinct(RP_score, "TF" = Factor, target)
# 
# cistrome_df <- cistrome_df %>%
#   filter(TF %in% interaction_df$receiver, target %in% interaction_df$receiver) %>%
#   mutate(RP_score = ave(RP_score, TF, target, FUN = mean)) %>%
#   distinct() %>%
#   filter(TF != target) %>%
#   #filter(RP_score > 0.3) %>%
#   mutate(sum_target = as.numeric(ave(target, target, FUN = function(x) length(x))))
# 
# cistrome_matrix <- cistrome_df %>%
#   dplyr::select(-sum_target) %>%
#   spread(key = "target", value = "RP_score") %>%
#   column_to_rownames("TF")
# 
# cistrome_matrix[is.na(cistrome_matrix)] <- 0
# 
# pheatmap(cistrome_matrix,
#          border_color = NA,
#          color = colorRampPalette(c("#d5bdaf", 'white', "#84a98c"))(100),
#          clustering_method = "ward.D")
# 
# cistrome_nodes <- cistrome_df %>%
#   distinct(target, sum_target)
# 
# cistrome_edges <- cistrome_df %>%
#   filter(target %in% cistrome_nodes$target) %>%
#   distinct(TF, target, "weight" = RP_score)
# 
# cistrome_igraph <- graph_from_data_frame(d = cistrome_edges, vertices = cistrome_nodes, directed = TRUE)
# 
# ggraph(cistrome_igraph, layout = "sugiyama") + 
#   geom_node_point(aes(size = sum_target)) +
#   geom_edge_link(aes(width = weight), 
#                  alpha = 0.8,
#                  arrow = arrow(length = unit(3, 'mm')), 
#                  start_cap = circle(2, 'mm'),
#                  end_cap = circle(3, 'mm')) + 
#   scale_edge_width(range = c(0.2, 2)) +
#   geom_node_text(aes(label = name), repel = TRUE) +
#   theme_graph()
```



## Reporter activity compared to background

Aim: Display reporter activities relative to background activity per tf and condition. Get an overview over all reporter activities per condition.

```{r reporter_activities, out.width= "80%", fig.align= "center"}
nt_df <- cDNA_df %>%
  dplyr::filter(condition %in% c("NT_1", "NT_2")) %>%
  dplyr::select(reporter_id, "nt_reporter_activity_minP" = reporter_activity_minP) %>%
  mutate(nt_reporter_activity_minP = ave(nt_reporter_activity_minP, reporter_id, FUN = mean)) %>%
  distinct()

cDNA_df2 <- merge(cDNA_df, nt_df, all = T) %>%
  #na.omit() %>%
  filter(neg_ctrls == "No") %>%
  mutate(reporter_FC = log2(reporter_activity_minP/nt_reporter_activity_minP)) %>%
  distinct(tf, reporter_id, condition, reporter_FC, commercial_reporter) %>%
  mutate(mean_tf_activity = ave(reporter_FC, tf, condition, FUN = mean))

## Reporter FC vs reporter activity in NT control - are the most changing reporters the ones with higher activity?
ggplot(merge(cDNA_df, nt_df, all = T) %>%
  filter(neg_ctrls == "No", tf == "SP1", condition == "SP1") %>%
  mutate(reporter_FC = log2(reporter_activity_minP/nt_reporter_activity_minP)) %>%
  distinct(tf, reporter_id, condition, reporter_FC, commercial_reporter, nt_reporter_activity_minP),
       aes(x = reporter_FC, y = nt_reporter_activity_minP)) +
  geom_point()

tf_kd_targets <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf7027_KD_mES/tf_kd_targets_mES.csv") %>%
  #filter(condition %in% cDNA_df5$condition) %>%
  setnames("target", "tf") %>%
  mutate(is_target = T) %>%
  right_join(cDNA_df2, by = c("condition", "tf"))

tf_kd_targets$is_target[is.na(tf_kd_targets$is_target)] <- F
tf_kd_targets <- tf_kd_targets %>%
  filter(condition != "NT") %>%
  arrange(is_target)

# for (i in unique(tf_kd_targets$condition)) {
#   print(i)
#   p <- ggplot(tf_kd_targets %>%
#                 filter(condition == i),
#               mapping = aes(x = reorder(tf, -mean_tf_activity), y = reporter_FC, color = is_target)) +
#     geom_boxplot(aes(alpha = 0.7)) +
#     geom_quasirandom(aes(alpha = 0.7)) +
#     scale_color_manual(values = c(colors_diverse[1], colors_diverse[3])) +
#     ggtitle(i) +
#     theme_classic_lines_90()
#
#   print(p)
# }

tf_kd_targets <- tf_kd_targets %>%
  mutate(tf = gsub("_.*", "", tf))

tf_kd_targets2 <- tf_kd_targets %>%
  filter(tf %in% cDNA_df4$tf) %>%
  mutate(is_target = ifelse(commercial_reporter == "Yes", "commercial", is_target))

for (i in unique(tf_kd_targets2$condition)) {
  print(i)
  p <- ggplot(tf_kd_targets2 %>%
                filter(condition == i),
              mapping = aes(x = reorder(tf, -mean_tf_activity), y = reporter_FC, color = is_target, group = commercial_reporter)) +
    geom_hline(yintercept = 0, lty = 2) +
    geom_quasirandom(dodge.width = .75) +
    scale_color_manual(values = c("FALSE" = colors_diverse[1], "TRUE" = colors_diverse[3], "commercial" = colors_diverse[4])) +
    ggtitle(i) +
    theme_pubr(border = T, x.text.angle = 90)

  print(p)
}

for (i in unique(tf_kd_targets2$tf)) {
  print(i)
  p <- ggplot(tf_kd_targets2 %>%
                filter(tf == i),
              mapping = aes(x = reorder(condition, -mean_tf_activity), y = reporter_FC, color = is_target, group = commercial_reporter)) +
    geom_hline(yintercept = 0, lty = 2) +
    geom_quasirandom(dodge.width = .75, raster.dpi = 600) +
    scale_color_manual(values = c("FALSE" = colors_diverse[1], "TRUE" = colors_diverse[3], "commercial" = colors_diverse[4])) +
    ggtitle(i) +
    theme_pubr(border = T, x.text.angle = 90)

  print(p)
}


```

Conclusion: Some KDs work well, some synthetic reporters are better than the positive controls.

---



## Include infered TF activity from regulon expression

Aim: Similar as before: How well does regulon activity correlate with TF reporter activity across cell types?
```{r}
# # VIPER
# library(decoupleR)
# library(dorothea)
# library(OmnipathR)
# 
# tf_rna <- read_tsv("/DATA/usr/m.trauernicht/data/RNA_seq/rna_tpm_all_tfs.tsv")
# 
# ## Import human TF regulon information
# tf_interactions <- decoupleR::get_dorothea()
# 
# ## Import mouse regulons
# tf_interactions_mouse <- decoupleR::get_dorothea(organism = "mouse") %>%
#   mutate(source = toupper(source)) %>%
#   mutate(target = toupper(target))
# 
# ## Import TF expression data
# tf_rna_condition <- tf_rna %>% 
#   distinct() %>%
#   # mutate(TPM = log2(TPM)) %>%
#   # mutate(TPM = ifelse(!is.finite(TPM), NA, TPM)) %>%
#   spread(key = "cell", value = "TPM")
# 
# ### Rename TFs so they match the official names
# tf_rna_condition$tf[tf_rna_condition$tf == "AHR::ARNT"] <- "AHR"
# tf_rna_condition$tf[tf_rna_condition$tf == "FOS::JUN"] <- "FOS"
# tf_rna_condition$tf[tf_rna_condition$tf == "MAF::NFE2"] <- "NFE2"
# tf_rna_condition$tf[tf_rna_condition$tf == "NR1H4::RXRA"] <- "NR1H4"
# tf_rna_condition$tf[tf_rna_condition$tf == "NR4A2::RXRA"] <- "NR4A2"
# tf_rna_condition$tf[tf_rna_condition$tf == "PPARA::RXRA"] <- "PPARA"
# tf_rna_condition$tf[tf_rna_condition$tf == "PPARG::RXRA"] <- "PPARG"
# tf_rna_condition$tf[tf_rna_condition$tf == "SMAD2::SMAD3::SMAD4"] <- "SMAD2"
# tf_rna_condition$tf[tf_rna_condition$tf == "STAT1::STAT2"] <- "STAT1"
# tf_rna_condition$tf[tf_rna_condition$tf == "VDR::RXRA"] <- "VDR"
# tf_rna_condition$tf[tf_rna_condition$tf == "NFATc1"] <- "NFATC1"
# tf_rna_condition$tf[tf_rna_condition$tf == "RARA:RXRA"] <- "RARA"
# 
# ## Select human expression data
# tf_rna_condition_human <- tf_rna_condition %>%
#   na.omit() %>%
#   distinct() %>%
#   column_to_rownames("tf") %>%
#   dplyr::select(-mES, -NPC) 
# 
# ## Run VIPER on human expression data and human regulons
# viper_res_human <- decoupleR::run_viper(tf_rna_condition_human, tf_interactions, minsize = 1) %>%
#   dplyr::select("tf" = source, "cell" = condition, "viper_score" = score)
# 
# ## Select mouse expression data
# tf_rna_condition_mouse <- tf_rna_condition %>%
#   na.omit() %>%
#   distinct() %>%
#   column_to_rownames("tf") %>%
#   dplyr::select(mES, NPC)
# 
# ## Run VIPER on mouse expression data and human regulons
# viper_res_mouse <- decoupleR::run_viper(tf_rna_condition_mouse, tf_interactions_mouse, minsize = 1) %>%
#   dplyr::select("tf" = source, "cell" = condition, "viper_score" = score)
# 
# ## Combine human and mouse VIPER output
# viper_res <- rbind(viper_res_mouse, viper_res_human)
# 
# ### Change names back to original TF names
# viper_res2 <- viper_res %>% filter(tf == "POU5F1")
# viper_res2$tf[viper_res2$tf == "POU5F1"] <- "POU5F1::SOX2"
# viper_res <- rbind(viper_res, viper_res2)
# 
# viper_res2 <- viper_res %>% filter(tf == "RARA")
# viper_res2$tf[viper_res2$tf == "RARA"] <- "RARA:RXRA"
# viper_res <- rbind(viper_res, viper_res2)
# 
# viper_res2 <- viper_res %>% filter(tf == "SMAD4")
# viper_res2$tf[viper_res2$tf == "SMAD4"] <- "SMAD2::SMAD3::SMAD4"
# viper_res <- rbind(viper_res, viper_res2)
# 
# viper_res$tf[viper_res$tf == "OTX2"] <- "OTX1"
# viper_res$tf[viper_res$tf == "AHR"] <- "AHR::ARNT"
# viper_res$tf[viper_res$tf == "ARNT"] <- "AHR::ARNT"
# viper_res$tf[viper_res$tf == "FOS"] <- "FOS::JUN"
# viper_res$tf[viper_res$tf == "NFE2"] <- "MAF::NFE2"
# viper_res$tf[viper_res$tf == "NR1H4"] <- "NR1H4::RXRA"
# viper_res$tf[viper_res$tf == "NR4A2"] <- "NR4A2::RXRA"
# viper_res$tf[viper_res$tf == "PPARA"] <- "PPARA::RXRA"
# viper_res$tf[viper_res$tf == "PPARG"] <- "PPARG::RXRA"
# viper_res$tf[viper_res$tf == "SMAD2"] <- "SMAD2::SMAD3::SMAD4"
# viper_res$tf[viper_res$tf == "SMAD3"] <- "SMAD2::SMAD3::SMAD4"
# viper_res$tf[viper_res$tf == "STAT1"] <- "STAT1::STAT2"
# viper_res$tf[viper_res$tf == "VDR"] <- "VDR::RXRA"
# viper_res$tf[viper_res$tf == "NFATC1"] <- "NFATc1"
# 
# #write_tsv(viper_res, "/DATA/usr/m.trauernicht/data/RNA_seq/viper_activities.tsv")
# 
# viper_res_filt <- viper_res %>%
#   filter(tf %in% unique(cDNA_df2$tf)) %>%
#   mutate(viper_score = ave(viper_score, tf, cell, FUN = mean)) %>%
#   distinct()
# 
# ## Set viper score to 0 if the TF is not expressed (somehow viper doesn't take care of this now)
# 
# 
# 
# # # Other models, don't use for now, I checked that they correlate very well, so no need to include redundant information
# # mlm_res <- decoupleR::run_mlm(mat = tf_rna_condition %>% na.omit(), network = tf_interactions)
# # ulm_res <- decoupleR::run_ulm(mat = tf_rna_condition %>% na.omit(), network = tf_interactions)
# 
# # ## Alternative workflow using the original viper algorithm
# # library(minet)
# # mim <- minet::build.mim(t(tf_rna_condition))
# # regulon <- aracne(mim)
# # regul <- viper::aracne2regulon(regulon, t(tf_rna_condition))
# # viper_res <- viper::viper(tf_rna_condition, regulon) %>%
# #   dplyr::select("tf" = source, "cell" = condition, "viper_score" = score)
# 
# 
# ## Combine TF reporter data with RNA-seq data
# cDNA_df3 <- merge(cDNA_df2, viper_res_filt, all = T) %>%
#   mutate(viper_score = ifelse(is.na(viper_score), min(viper_score, na.rm = T), viper_score))
# 
# ## Compute correlations for each individual reporter of reporter activity vs. TPM counts across the nine cell types
# for (i in unique(cDNA_df3$reporter_id)) {
#     if (nrow(cDNA_df3[cDNA_df3$reporter_id == i,]) > 2) {
#       cDNA_df3$cor_pval[cDNA_df3$reporter_id == i] <- -log10(cor.test(cDNA_df3$reporter_activity_neg[cDNA_df3$reporter_id == i], cDNA_df3$viper_score[cDNA_df3$reporter_id == i], method = "pearson", exact = F)$p.value)
#       cDNA_df3$cor[cDNA_df3$reporter_id == i] <- cor(cDNA_df3$reporter_activity_neg[cDNA_df3$reporter_id == i], cDNA_df3$viper_score[cDNA_df3$reporter_id == i], method = "pearson")
#     }
#   else {
#     cDNA_df3$cor_pval[cDNA_df3$reporter_id == i] <- 1
#     cDNA_df3$cor[cDNA_df3$reporter_id == i] <- 0
#   }
# }
# 
# ## I am only interested in positive correlations, so set all negative correlations to zero
# cDNA_df3 <- cDNA_df3 %>%
#   mutate(cor_pval = ifelse(cor < 0, 0, cor_pval))
# 
# # for (i in unique(cDNA_df3$tf)) {
# #   
# #   x <- cDNA_df3 %>%filter(tf == i)
# #   
# #   label <- data.frame(reporter_id = x$reporter_id, label = round(x$cor_pval, 2))
# #   
# #   p <- ggplot(x,
# #               aes(x = viper_score, y = reporter_activity_neg)) +
# #     geom_point() +
# #     geom_abline(lty = 2) +
# #     geom_smooth(method = "lm", se = F) +
# #     facet_wrap(~reporter_id) +
# #     ggtitle(i) +
# #     geom_label(data = label, aes(label = label, color = ifelse(label > 1, "red", "black")),
# #                x = Inf, y = -Inf, hjust=1, vjust=0,
# #                inherit.aes = FALSE) +
# #     ylab("Reporter Activity") +
# #     xlab("VIPER score") +
# #     scale_color_manual(values = c("black", "red"))
# #   
# #   print(p)
# # }
# 
# ## Export correlations
# viper_correlations <- cDNA_df3 %>%
#   dplyr::select(reporter_id, "viper_cor_pval" = cor_pval)
# 
# # ## Correlate RNA-seq vs. VIPER correlations (only significant in one of the two)
# # cor_rna_viper <- viper_correlations %>%
# #   setnames("cor", "viper_cor") %>%
# #   #mutate(viper_cor = ifelse(cor_pval < 1, 0, cor_pval)) %>%
# #   dplyr::select(-cor_pval) %>%
# #   left_join(rna_correlations) #%>%
# #   #mutate(cor = ifelse(cor_pval < 1, 0, cor_pval)) %>%
# #   #mutate(sum_cor = cor + viper_cor) %>%
# #   #filter(sum_cor != 0)
# # 
# # ggplot(cor_rna_viper,
# #        aes(x = cor, y = viper_cor)) +
# #   geom_point() +
# #   geom_abline(lty = 2) 
```





# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

