---
title: "Multiplexed identification of TF-specific reporters"
author:
- name: Max Trauernicht
  email: m.trauernicht@nki.nl
  affiliation: "Netherlands Cancer Institute - van Steensel lab"
date: "`r format(Sys.time(), "%d/%m/%Y")`"
output:
  pdf_document:
    toc: yes
    toc_depth: '4'
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: no
---
---

### Aim
35,500 reporters for 86 TFs were transfected into 9 different cell types + treated with ~90 different siRNAs or various signaling molecules. In this script I will analyze TF reporter activities in detail and review how individual reporters respond to TF concentration variations.


### Description of input data
- library = with which library was this sample transfected (1, 2, or 1+2)
- nchar = barcode length (12 for library 1, 13 for library 2)
- commercial_reporter = positive control reporter sequence from commercial source
- neg_ctrls = reporter with mutated TF binding sites?
- hPGK = positive control, chunk from hPGK promoter (only in lib 1)
- native_enhancer = genomic mES enhancer sequence chunks from Miguels libraries (only lib 1)
- activity = rpm + pseudocount cDNA / rpm + pseudocount pDNA
- mean_activity_sample = mean of activity per unique reporter per sample (individually per replicate)
- reporter_activity = mean activity per unique reporter per condition (mean of replicates)
- minP_activity = activity relative to mean of minimal promoter only reporters of that sample
- mean_activity_sample_minP = mean of minP_activity per unique reporter 
- reporter_activity_minP = mean of mean_activity_sample_minP per condition
- nfya_activity = activity relative to mean of nfya reporters (constitutive positive control) of that sample
- mean_activity_sample_nfya = mean of nfya_activity per unique reporter 
- reporter_activity_nfya = mean of mean_activity_sample_nfya per condition
- mean_activity_sample_neg = mean_activity_sample relative to its paired negative control (identical reporter with mutated binding sites)
- reporter_activity_neg = mean of mean_activity_sample_neg per condition
- pval_minP = p-value of t-test comparing activity distributions of individual reporters with the distribution of minimal promoter only reporters within the same condition
- pval_neg = p-value of t-test comparing activity distributions of individual reporters with the distribution of the activities of its paired negative controls within the same condition


### Main tested conditions:
- HepG2 + TF-KD / NT

---

## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}
```

### Libraries 

```{r setup, out.width= "80%", fig.align= "center", warning = FALSE, message= FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(RColorBrewer)
library(ggplot2)
library(dplyr)
library(maditr)
library(tibble)
library(pheatmap)
library(ggpubr)
library(visNetwork)
library(ggbeeswarm)
library(ggforce)
library(viridis)
library(plyr)
library(igraph)
library(ggraph)
library(cowplot)
library(gridExtra)
library(pROC)
library(tidyr)
library(stringr)
library(randomForest)
library(ggrastr)
library(readr)
library(ggbiplot)
library(IHW)
library(biomaRt)
library(ggh4x)
```

### Functions

```{r out.width= "80%", fig.align= "center"}
### Custom functions
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}


# Extract p-value from linear model
lmp <- function (modelobject) {
    if (class(modelobject) != "lm") stop("Not an object of class 'lm' ")
    f <- summary(modelobject)$fstatistic
    p <- pf(f[1],f[2],f[3],lower.tail=F)
    attributes(p) <- NULL
    return(p)
}

# Set custom ggplot2 theme and custom colors
theme_classic_lines <- function() {
  theme_pubr(border = F, legend = "top") +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_classic_lines_45 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 45) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_classic_lines_90 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 90) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_barplot <- function() {
  theme_pubr(border = T, legend = "none") +
            theme(panel.grid.major.x = element_line(colour = "black", size = 0.25),
                  strip.background = element_blank(),
                  strip.text = element_text(face="bold", hjust=0)
            )
}

theme_set(theme_classic_lines())

colors_diverse <- c("#264653", "#9AC1AE", "#5D987B", "#f2cc8f", "#e76f51")

ggplot_custom <- function(...) ggplot2::ggplot(...) + 
  scale_color_manual(values = colors_diverse) + 
  scale_fill_manual(values = colors_diverse)
```

### Loading data

```{r out.width= "80%", fig.align= "center"}
# Import processed bc counts from the preprocessing step
cDNA_df <- read.csv("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf7124_stimulations/results/mt20230301_reporter_activity_filt_combined.csv", header = T)

cDNA_df$stimulation[is.na(cDNA_df$stimulation)] <- "no"

# final measure of reporter activity: combination of reporter_activity_minP and reporter_activity_neg
## for most reporters I will take reporter_activity_neg, however, this cannot be done for commercial controls for example
cDNA_df <- cDNA_df %>%
  mutate(reporter_activity_neg = ifelse(tf %in% c("POU5F1::SOX2", "NR1H4::RXRA") | commercial_reporter == "Yes" | is.na(reporter_activity_neg), reporter_activity_minP, reporter_activity_neg)) 
```

---

## Investigate individual reporter activities - 1

Aim: Show that there are differences in activities between reporter variants

```{r}
cDNA_df2 <- cDNA_df %>%
  filter(neg_ctrls == "No", 
         commercial_reporter == "No", 
         hPGK == "No", 
         str_detect(tf, "RANDOM", negate = T), 
         native_enhancer == "No",
         condition %in% c("mES_2i_LIF")) %>%
  distinct(tf, condition, reporter_activity_neg, cell, reporter_activity_neg, reporter_id) %>%
  mutate(reporter_activity_neg = log2(reporter_activity_neg)) %>%
  distinct() 

ggplot(cDNA_df2 %>%
         mutate(tf_activity = ave(reporter_activity_neg, tf, FUN = mean)),
       aes(x = reorder(tf, -tf_activity), y = reporter_activity_neg)) +
  geom_hline(yintercept = 1, lty = "twodash") + 
  geom_hline(yintercept = -1, lty = "twodash") + 
  geom_quasirandom(aes(color = reporter_activity_neg), size = 2, width = 0.25) +
  stat_summary(fun = median, fun.min = median, fun.max = median, geom = "crossbar", color = "black", width = 0.5, lwd = 0.4) +
  scale_color_gradient(low = "#98C1D9", high = "#E0777D", limits = c(-2,4), oob = squish) +
  theme_pubr(x.text.angle = 90) 
```


---


## Reporter correlation across cell types

Aim: Correlate for each TF the correlation of reporter activity with RNA-seq across multiple cell types
```{r}
# Load in RNA-seq data
## Set the names of the TFs for which I want to retrieve the data
tfs <- cDNA_df %>%
  filter(neg_ctrls == "No", commercial_reporter == "No", hPGK == "No", str_detect(tf, "RANDOM", negate = T)) %>%
  distinct(tf)
tfs$tf[tfs$tf == "AHR::ARNT"] <- "AHR"
tfs$tf[tfs$tf == "FOS::JUN"] <- "FOS"
tfs$tf[tfs$tf == "MAF::NFE2"] <- "NFE2"
tfs$tf[tfs$tf == "NR1H4::RXRA"] <- "NR1H4"
tfs$tf[tfs$tf == "NR4A2::RXRA"] <- "NR4A2"
tfs$tf[tfs$tf == "PPARA::RXRA"] <- "PPARA"
tfs$tf[tfs$tf == "PPARG::RXRA"] <- "PPARG"
tfs$tf[tfs$tf == "SMAD2::SMAD3::SMAD4"] <- "SMAD2"
tfs$tf[tfs$tf == "STAT1::STAT2"] <- "STAT1"
tfs$tf[tfs$tf == "VDR::RXRA"] <- "VDR"
tfs$tf[tfs$tf == "NFATc1"] <- "NFATC1"
tfs$tf[tfs$tf == "RARA:RXRA"] <- "RARA"

## Load mES data from Miguel
load('/DATA/usr/m.martinez.ara/GitLab/gurten/gurten/data/fc181121_epsure_tadec_rnaseq_expr_joshi.RData')

mES_rna <- tib_fpkm_joshi %>%
  mutate(twoi = (twoi_1 + twoi_2) / 2) %>%
  dplyr::select(symbol, twoi) %>%
  mutate(twoi = (twoi / sum(twoi))*10^6) %>% ## Transform from FPKM to TPM
  group_by(symbol) %>%
  arrange(desc(twoi)) %>%
  top_n(1) %>%
  ungroup()

mES_rna$symbol[mES_rna$symbol == "TRP53"] <- "TP53"

mES_rna <- mES_rna %>%
  setnames(c("symbol", "twoi"), c("tf", "TPM")) %>%
  #filter(mES_rna$tf %in% tfs$tf) %>%
  mutate(cell = "mES")

## Load data from protein atlas (TPM)
proteinatlas_rna <- read_tsv("/DATA/usr/m.trauernicht/data/RNA_seq/rna_celline.tsv") %>%
  dplyr::select('tf' = `Gene name`, TPM, "cell" = `Cell line`) %>%
  filter(cell %in% c("A-549", "U2OS", "Hep-G2", "HEK293", "K-562", "MCF-7", "HCT16")) %>%
  mutate(cell = gsub(" ", "", cell)) %>%
  mutate(cell = gsub("-", "", cell)) %>%
  #filter(tf %in% tfs$tf)
  mutate(tf_cell = paste(tf, cell)) %>%
  group_by(tf_cell) %>%
  slice_max(n = 1, order_by = TPM) %>%
  ungroup() %>%
  dplyr::select(-tf_cell)

npc_rna <- read_delim("/shared/gcf/m.trauernicht/7086/genecounts.txt") %>%
  dplyr::select("tf" = external_gene_id, "NPC_r1" = `1_NPC_r1`, "NPC_r2" = `2_NPC_r2`, start_position, end_position) %>%
  mutate(length = end_position - start_position) %>%
  mutate(NPC_counts = (NPC_r1 + NPC_r2) / 2) %>%
  mutate(NPC_rpk = NPC_counts / (length / 1000)) %>%
  mutate(pm = sum(NPC_rpk / 1e6)) %>%
  mutate(TPM = NPC_rpk / pm) %>%
  mutate(cell = "NPC") %>%
  distinct(tf, TPM, cell) %>%
  mutate(tf = toupper(tf))  %>%
  group_by(tf) %>%
  slice_max(n = 1, order_by = TPM) %>%
  ungroup()


## Combine data from all cell types and change names back to original TF names
tf_rna <- rbind(mES_rna, proteinatlas_rna, npc_rna)
tf_rna$tf[tf_rna$tf == "AHR"] <- "AHR::ARNT"
tf_rna$tf[tf_rna$tf == "FOS"] <- "FOS::JUN"
tf_rna$tf[tf_rna$tf == "NFE2"] <- "MAF::NFE2"
tf_rna$tf[tf_rna$tf == "NR1H4"] <- "NR1H4::RXRA"
tf_rna$tf[tf_rna$tf == "NR4A2"] <- "NR4A2::RXRA"
tf_rna$tf[tf_rna$tf == "PPARA"] <- "PPARA::RXRA"
tf_rna$tf[tf_rna$tf == "PPARG"] <- "PPARG::RXRA"
tf_rna$tf[tf_rna$tf == "SMAD2"] <- "SMAD2::SMAD3::SMAD4"
tf_rna$tf[tf_rna$tf == "STAT1"] <- "STAT1::STAT2"
tf_rna$tf[tf_rna$tf == "VDR"] <- "VDR::RXRA"
tf_rna$tf[tf_rna$tf == "NFATC1"] <- "NFATc1"

tf_rna2 <- tf_rna %>% filter(tf == "POU5F1")
tf_rna2$tf[tf_rna2$tf == "POU5F1"] <- "POU5F1::SOX2"
tf_rna <- rbind(tf_rna, tf_rna2)

tf_rna2 <- tf_rna %>% filter(tf == "RARA")
tf_rna2$tf[tf_rna2$tf == "RARA"] <- "RARA:RXRA"
tf_rna <- rbind(tf_rna, tf_rna2)

tf_rna$TPM[is.na(tf_rna$TPM)] <- 0

#write_tsv(tf_rna, "/DATA/usr/m.trauernicht/data/RNA_seq/rna_tpm_all_tfs.tsv")


## Prepare tf reporter data frame
cDNA_df2 <- cDNA_df %>%
  #filter(tf != "RFX1") %>%
  dplyr::filter(neg_ctrls == "No",
                #commercial_reporter == "No",
                stimulation == "no",
                hPGK == "No",
                str_detect(tf, "RANDOM", negate = T),
                native_enhancer == "No") %>%
  distinct(tf, reporter_activity_neg, cell, reporter_id, commercial_reporter) %>%
  mutate(reporter_activity_neg = ave(reporter_activity_neg, cell, reporter_id, FUN = mean)) %>%
  mutate(reporter_activity_neg = log2(reporter_activity_neg)) %>%
  distinct() %>%
  mutate(tf = gsub("_.*", "", tf))

## Combine TF reporter data with RNA-seq data
cDNA_df3 <- merge(cDNA_df2, tf_rna, all = T) %>%
  na.omit()

## Compute spearman correlations for each individual reporter of reporter activity vs. TPM counts across the nine cell types
for (i in unique(cDNA_df3$reporter_id)) {
  if (nrow(cDNA_df3[cDNA_df3$reporter_id == i,]) > 2) {
  cDNA_df3$cor_pval[cDNA_df3$reporter_id == i] <- -log10(cor.test(cDNA_df3$reporter_activity_neg[cDNA_df3$reporter_id == i], cDNA_df3$TPM[cDNA_df3$reporter_id == i], method = "pearson", exact = F)$p.value)
  cDNA_df3$cor[cDNA_df3$reporter_id == i] <- cor(cDNA_df3$reporter_activity_neg[cDNA_df3$reporter_id == i], cDNA_df3$TPM[cDNA_df3$reporter_id == i], method = "pearson")
  }
  else {
    cDNA_df3$cor_pval[cDNA_df3$reporter_id == i] <- 1
    cDNA_df3$cor[cDNA_df3$reporter_id == i] <- 0
  }
}

## I am only interested in positive correlations, so set all negative correlations to zero
cDNA_df3 <- cDNA_df3 %>%
  mutate(cor_pval = ifelse(cor < 0, 0, cor_pval))
#
# ## Plot the correlations including the p-values
# for (i in unique(cDNA_df3$tf)) {
#
#   x <- cDNA_df3 %>%filter(tf == i)
#
#   label <- data.frame(reporter_id = x$reporter_id, label = round(x$cor_pval, 2))
#
#   p <- ggplot(x,
#               aes(x = TPM, y = reporter_activity_neg)) +
#     geom_point() +
#     geom_abline(lty = 2) +
#     geom_smooth(method = "lm", se = F) +
#     facet_wrap(~reporter_id) +
#     ggtitle(i) +
#     ylab("Reporter Activity") +
#     xlab("TPM") +
#     geom_label(data = label, aes(label = label, color = ifelse(label > 1, "red", "black")),
#                x = Inf, y = -Inf, hjust=1, vjust=0,
#                inherit.aes = FALSE) +
#     scale_color_manual(values = c("black", "red"))
#
#   print(p)
# }

## Set Inf values to max per TF
cDNA_df3 <- cDNA_df3 %>%
  group_by(tf) %>%
  mutate(cor_pval = ifelse(is.finite(cor_pval), cor_pval, max(cor_pval[is.finite(cor_pval)]))) %>%
  ungroup()

## Export correlations
rna_correlations <- cDNA_df3 %>%
  distinct(reporter_id, cor_pval, cor)

## Plot correlation p-values per TF
ggplot_custom(cDNA_df3 %>%
         distinct(reporter_id, cor_pval, tf, commercial_reporter) %>%
         mutate(mean_cor_pval = ave(cor_pval, tf, FUN = mean)),
       aes(x = reorder(tf, -mean_cor_pval), y = cor_pval, color = commercial_reporter)) +
  geom_point() +
  geom_hline(yintercept = 1, lty = 2) +
  theme_classic_lines_90()
```


## Include infered activity from regulon expression
```{r}
# VIPER
library(decoupleR)
library(dorothea)
library(OmnipathR)

tf_rna <- read_tsv("/DATA/usr/m.trauernicht/data/RNA_seq/rna_tpm_all_tfs.tsv")

## Import human TF regulon information
tf_interactions <- decoupleR::get_dorothea()

## Import mouse regulons
tf_interactions_mouse <- decoupleR::get_dorothea(organism = "mouse") %>%
  mutate(source = toupper(source)) %>%
  mutate(target = toupper(target))

## Import TF expression data
tf_rna_condition <- tf_rna %>% 
  distinct() %>%
  # mutate(TPM = log2(TPM)) %>%
  # mutate(TPM = ifelse(!is.finite(TPM), NA, TPM)) %>%
  spread(key = "cell", value = "TPM")

### Rename TFs so they match the official names
tf_rna_condition$tf[tf_rna_condition$tf == "AHR::ARNT"] <- "AHR"
tf_rna_condition$tf[tf_rna_condition$tf == "FOS::JUN"] <- "FOS"
tf_rna_condition$tf[tf_rna_condition$tf == "MAF::NFE2"] <- "NFE2"
tf_rna_condition$tf[tf_rna_condition$tf == "NR1H4::RXRA"] <- "NR1H4"
tf_rna_condition$tf[tf_rna_condition$tf == "NR4A2::RXRA"] <- "NR4A2"
tf_rna_condition$tf[tf_rna_condition$tf == "PPARA::RXRA"] <- "PPARA"
tf_rna_condition$tf[tf_rna_condition$tf == "PPARG::RXRA"] <- "PPARG"
tf_rna_condition$tf[tf_rna_condition$tf == "SMAD2::SMAD3::SMAD4"] <- "SMAD2"
tf_rna_condition$tf[tf_rna_condition$tf == "STAT1::STAT2"] <- "STAT1"
tf_rna_condition$tf[tf_rna_condition$tf == "VDR::RXRA"] <- "VDR"
tf_rna_condition$tf[tf_rna_condition$tf == "NFATc1"] <- "NFATC1"
tf_rna_condition$tf[tf_rna_condition$tf == "RARA:RXRA"] <- "RARA"

## Select human expression data
tf_rna_condition_human <- tf_rna_condition %>%
  na.omit() %>%
  distinct() %>%
  column_to_rownames("tf") %>%
  dplyr::select(-mES, -NPC) 

## Run VIPER on human expression data and human regulons
viper_res_human <- decoupleR::run_viper(tf_rna_condition_human, tf_interactions, minsize = 1) %>%
  dplyr::select("tf" = source, "cell" = condition, "viper_score" = score)

## Select mouse expression data
tf_rna_condition_mouse <- tf_rna_condition %>%
  na.omit() %>%
  distinct() %>%
  column_to_rownames("tf") %>%
  dplyr::select(mES, NPC)

## Run VIPER on mouse expression data and human regulons
viper_res_mouse <- decoupleR::run_viper(tf_rna_condition_mouse, tf_interactions_mouse, minsize = 1) %>%
  dplyr::select("tf" = source, "cell" = condition, "viper_score" = score)

## Combine human and mouse VIPER output
viper_res <- rbind(viper_res_mouse, viper_res_human)

### Change names back to original TF names
viper_res2 <- viper_res %>% filter(tf == "POU5F1")
viper_res2$tf[viper_res2$tf == "POU5F1"] <- "POU5F1::SOX2"
viper_res <- rbind(viper_res, viper_res2)

viper_res2 <- viper_res %>% filter(tf == "RARA")
viper_res2$tf[viper_res2$tf == "RARA"] <- "RARA:RXRA"
viper_res <- rbind(viper_res, viper_res2)

viper_res2 <- viper_res %>% filter(tf == "SMAD4")
viper_res2$tf[viper_res2$tf == "SMAD4"] <- "SMAD2::SMAD3::SMAD4"
viper_res <- rbind(viper_res, viper_res2)

viper_res$tf[viper_res$tf == "OTX2"] <- "OTX1"
viper_res$tf[viper_res$tf == "AHR"] <- "AHR::ARNT"
viper_res$tf[viper_res$tf == "ARNT"] <- "AHR::ARNT"
viper_res$tf[viper_res$tf == "FOS"] <- "FOS::JUN"
viper_res$tf[viper_res$tf == "NFE2"] <- "MAF::NFE2"
viper_res$tf[viper_res$tf == "NR1H4"] <- "NR1H4::RXRA"
viper_res$tf[viper_res$tf == "NR4A2"] <- "NR4A2::RXRA"
viper_res$tf[viper_res$tf == "PPARA"] <- "PPARA::RXRA"
viper_res$tf[viper_res$tf == "PPARG"] <- "PPARG::RXRA"
viper_res$tf[viper_res$tf == "SMAD2"] <- "SMAD2::SMAD3::SMAD4"
viper_res$tf[viper_res$tf == "SMAD3"] <- "SMAD2::SMAD3::SMAD4"
viper_res$tf[viper_res$tf == "STAT1"] <- "STAT1::STAT2"
viper_res$tf[viper_res$tf == "VDR"] <- "VDR::RXRA"
viper_res$tf[viper_res$tf == "NFATC1"] <- "NFATc1"

#write_tsv(viper_res, "/DATA/usr/m.trauernicht/data/RNA_seq/viper_activities.tsv")

viper_res_filt <- viper_res %>%
  filter(tf %in% unique(cDNA_df2$tf)) %>%
  mutate(viper_score = ave(viper_score, tf, cell, FUN = mean)) %>%
  distinct()

## Set viper score to 0 if the TF is not expressed (somehow viper doesn't take care of this now)



# # Other models, don't use for now, I checked that they correlate very well, so no need to include redundant information
# mlm_res <- decoupleR::run_mlm(mat = tf_rna_condition %>% na.omit(), network = tf_interactions)
# ulm_res <- decoupleR::run_ulm(mat = tf_rna_condition %>% na.omit(), network = tf_interactions)

# ## Alternative workflow using the original viper algorithm
# library(minet)
# mim <- minet::build.mim(t(tf_rna_condition))
# regulon <- aracne(mim)
# regul <- viper::aracne2regulon(regulon, t(tf_rna_condition))
# viper_res <- viper::viper(tf_rna_condition, regulon) %>%
#   dplyr::select("tf" = source, "cell" = condition, "viper_score" = score)


## Combine TF reporter data with RNA-seq data
cDNA_df3 <- merge(cDNA_df2, viper_res_filt, all = T) %>%
  mutate(viper_score = ifelse(is.na(viper_score), min(viper_score, na.rm = T), viper_score))

## Compute correlations for each individual reporter of reporter activity vs. TPM counts across the nine cell types
for (i in unique(cDNA_df3$reporter_id)) {
    if (nrow(cDNA_df3[cDNA_df3$reporter_id == i,]) > 2) {
      cDNA_df3$cor_pval[cDNA_df3$reporter_id == i] <- -log10(cor.test(cDNA_df3$reporter_activity_neg[cDNA_df3$reporter_id == i], cDNA_df3$viper_score[cDNA_df3$reporter_id == i], method = "pearson", exact = F)$p.value)
      cDNA_df3$cor[cDNA_df3$reporter_id == i] <- cor(cDNA_df3$reporter_activity_neg[cDNA_df3$reporter_id == i], cDNA_df3$viper_score[cDNA_df3$reporter_id == i], method = "pearson")
    }
  else {
    cDNA_df3$cor_pval[cDNA_df3$reporter_id == i] <- 1
    cDNA_df3$cor[cDNA_df3$reporter_id == i] <- 0
  }
}

## I am only interested in positive correlations, so set all negative correlations to zero
cDNA_df3 <- cDNA_df3 %>%
  mutate(cor_pval = ifelse(cor < 0, 0, cor_pval))

# for (i in unique(cDNA_df3$tf)) {
#   
#   x <- cDNA_df3 %>%filter(tf == i)
#   
#   label <- data.frame(reporter_id = x$reporter_id, label = round(x$cor_pval, 2))
#   
#   p <- ggplot(x,
#               aes(x = viper_score, y = reporter_activity_neg)) +
#     geom_point() +
#     geom_abline(lty = 2) +
#     geom_smooth(method = "lm", se = F) +
#     facet_wrap(~reporter_id) +
#     ggtitle(i) +
#     geom_label(data = label, aes(label = label, color = ifelse(label > 1, "red", "black")),
#                x = Inf, y = -Inf, hjust=1, vjust=0,
#                inherit.aes = FALSE) +
#     ylab("Reporter Activity") +
#     xlab("VIPER score") +
#     scale_color_manual(values = c("black", "red"))
#   
#   print(p)
# }

## Export correlations
viper_correlations <- cDNA_df3 %>%
  dplyr::select(reporter_id, "viper_cor_pval" = cor_pval)

# ## Correlate RNA-seq vs. VIPER correlations (only significant in one of the two)
# cor_rna_viper <- viper_correlations %>%
#   setnames("cor", "viper_cor") %>%
#   #mutate(viper_cor = ifelse(cor_pval < 1, 0, cor_pval)) %>%
#   dplyr::select(-cor_pval) %>%
#   left_join(rna_correlations) #%>%
#   #mutate(cor = ifelse(cor_pval < 1, 0, cor_pval)) %>%
#   #mutate(sum_cor = cor + viper_cor) %>%
#   #filter(sum_cor != 0)
# 
# ggplot(cor_rna_viper,
#        aes(x = cor, y = viper_cor)) +
#   geom_point() +
#   geom_abline(lty = 2) 
```


## Reporter quality score computation

Aim: We want to find for each TF the optimal reporters. Here I combine all data to compute reporter quality scores to find the best reporters for each TF.
```{r}
## Prepare data frame
cDNA_df2 <- cDNA_df %>%
  #filter(tf != "RFX1") %>%
  dplyr::filter(neg_ctrls == "No", 
                #commercial_reporter == "No", 
                hPGK == "No", 
                str_detect(tf, "RANDOM", negate = T), 
                native_enhancer == "No") %>%
  distinct(tf, condition, reporter_activity_neg, stimulation, reference_condition, cell, reporter_activity_neg, reporter_id, commercial_reporter) %>%
  mutate(reporter_activity_neg = log2(reporter_activity_neg)) %>%
  distinct()

# Export for shiny app
#write_tsv(cDNA_df2, "/DATA/usr/m.trauernicht/projects/SuRE-TF/data/mt20230302_all_activities.tsv")

## Make separate data frame for reference conditions - this will be used for normalizing the other conditions
reference_condition <- cDNA_df2 %>%
  filter(condition %in% reference_condition) %>%
  dplyr::select(-reference_condition, "reference_condition" = condition, -stimulation, "reference_activity" = reporter_activity_neg, reporter_id) %>%
  mutate(reference_activity = ave(reference_activity, reporter_id, reference_condition, FUN = mean)) %>%
  distinct()

## Normalize KD conditions
cDNA_df3 <- cDNA_df2 %>% 
  left_join(reference_condition) %>%
  filter(stimulation == "KD") %>%
  mutate(condition = gsub("HepG2_|mES_", "", condition)) %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  filter(tf == condition) %>%
  mutate(reporter_dif =  reference_activity - reporter_activity_neg)

## Normalize pathway stimulations
stimulations <- data.frame(condition = c("HepG2_CDCA", "HepG2_Rif", "HepG2_T09", "HepG2_Wy", "mES_2i", "mES_BX", "mES_HQ", "mES_IWP2", "mES_LIF_CH", "mES_LIF_PD", "mES_LIF_PD",
                                          "mES_vitC", "A549_Dex_100", "A549_Nutlin", "mES_Bmp4", "mES_Dex", "mES_FK", "mES_LPS", "mES_N2B27_Bmp4", "mES_N2B27_RA", "mES_N2B27_Shh", 
                                         "mES_Nutlin", "mES_RA", "mES_Shh", "mES_vitA", "U2OS_FK", "U2OS_LPS",
                                         "HCT116_cDIM12", "U2OS_Calcitriol", "U2OS_PMA", "HCT116_GSK4112", "MCF7_Hexestrol", "MCF7_DIMc"),
                           tf = c("NR1H4::RXRA", "NR1I2", "NR1H2", "PPARA::RXRA", "STAT3", "IRF3", "NFE2L2", "OTX1", "EGR1", "TCF7", "TCF7L2", "NFE2L2", "NR3C1", "TP53", "SMAD4", "NR3C1", 
                                  "CREB1", "IRF3", "SMAD4","RARA", "GLI1", "TP53", "RARA", "GLI1", "RARA", "CREB1", "IRF3", "NR4A2::RXRA", "VDR::RXRA", "NFKB1", "NR1D1", "ESR1", "NR4A1"),
                           effect_size = c(1,1,1,1,0,0,1,1,1,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,0,1,1))

cDNA_df4 <- cDNA_df2 %>%
  left_join(reference_condition) %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  filter(stimulation == "pathway") %>%
  mutate(reporter_dif =  reference_activity - reporter_activity_neg)

cDNA_df4 <- merge(cDNA_df4, stimulations) ## Only keep relevant TF reporters

cDNA_df4 <- cDNA_df4 %>%
  mutate(reporter_dif = ifelse(effect_size == 1, -reporter_dif, reporter_dif)) %>% ## Change sign of difference, based on whether the stimulation should cause an increase or decrease (effect_size)
  dplyr::select(-effect_size)

## Normalize degron conditions
cDNA_df_degron <- cDNA_df2 %>% 
  mutate(tf = gsub("_.*", "", tf)) %>%
  left_join(reference_condition) %>%
  filter(stimulation == "degron") %>%
  filter(condition != "mES_Nanog_AID") %>% ## Doesn't have a direct target (I don't have Nanog reporters)
  mutate(target_tf = toupper(gsub("mES_(.*)_AID", "\\1", condition)))

cDNA_df_degron2 <- cDNA_df_degron
cDNA_df_degron2$target_tf[cDNA_df_degron2$condition == "mES_Pou5f1_AID"] <- "SOX2"
cDNA_df_degron3 <- cDNA_df_degron
cDNA_df_degron3$target_tf[cDNA_df_degron3$condition == "mES_Pou5f1_AID"] <- "POU5F1::SOX2"
cDNA_df_degron4 <- cDNA_df_degron
cDNA_df_degron4$target_tf[cDNA_df_degron4$condition == "mES_Sox2_AID"] <- "POU5F1::SOX2"
cDNA_df_degron <- rbind(cDNA_df_degron, cDNA_df_degron2, cDNA_df_degron3, cDNA_df_degron4)
cDNA_df_degron <- cDNA_df_degron %>% 
  filter(tf == target_tf) %>%
  mutate(reporter_dif =  reference_activity - reporter_activity_neg) %>%
  dplyr::select(-target_tf)

## Combine stimulations and KDs 
perturbation_df <- rbind(cDNA_df3, cDNA_df4, cDNA_df_degron) %>% ## Combine all perturbation data
  mutate(condition_2 = paste(condition, reference_condition)) %>%
  mutate(mean_reporter_dif = ave(reporter_dif, condition_2, tf, FUN = function(x) mean(x, na.rm = T))) %>%
  group_by(tf) %>%
  arrange(desc(mean_reporter_dif)) %>% 
  top_n(1, wt = mean_reporter_dif) %>% ## Some TFs have multiple perturbations: for those I only keep the strongest perturbations (on average across all reporters)
  ungroup() %>%
  mutate(reporter_dif = ifelse(tf == "CEBPB", -reporter_dif, reporter_dif)) %>%
  distinct(reporter_id, tf, commercial_reporter, condition, reporter_dif)

## Besides the effect of the perturbation I also want to look at the raw reporter activities - for this I will consider the highest activities (over identical mutated control) in any cell type 

### Define optimal condition for each TF: either the highest expressing cell line, or the stimulated condition
reference_activities1 <- cDNA_df4 %>%
  right_join(stimulations) %>%
  filter(effect_size == 1) %>%
  distinct(tf, condition, cell, reporter_id, commercial_reporter, reporter_activity_neg) %>%
  group_by(tf, condition) %>%
  dplyr::arrange(desc(reporter_activity_neg)) %>%
  top_n(10) %>%
  ungroup() %>%
  mutate(median_reporter_activity_neg = ave(reporter_activity_neg, tf, condition, FUN = median)) %>%
  group_by(tf) %>%
  arrange(desc(median_reporter_activity_neg)) %>%
  top_n(1) %>%
  ungroup() %>%
  distinct(tf, condition)

reference_activities <- cDNA_df4 %>%
  distinct(tf, condition, cell, reporter_id, commercial_reporter, reporter_activity_neg) %>%
  right_join(reference_activities1) %>%
  dplyr::select(-cell) %>%
  setnames("condition", "cell")

reference_activities2 <- merge(cDNA_df2 %>% filter(stimulation == "no") %>% mutate(tf = gsub("_.*", "", tf)), tf_rna, by = c("cell", "tf")) %>%
  mutate(mean_reporter_activity_neg = ave(reporter_activity_neg, tf, condition, FUN = mean)) %>%
  group_by(tf) %>%
  dplyr::arrange(desc(TPM)) %>%
  top_n(1, wt = TPM) %>%
  ungroup() %>%
  group_by(reporter_id) %>%
  dplyr::arrange(desc(mean_reporter_activity_neg)) %>%
  top_n(1, wt = mean_reporter_activity_neg) %>%
  ungroup() %>%
  mutate(reporter_activity_neg = ave(reporter_activity_neg, reporter_id, cell)) %>%
  distinct(tf, cell, reporter_id, commercial_reporter, reporter_activity_neg) %>%
  filter(!tf %in% unique(reference_activities$tf)) %>%
  mutate(reporter_activity_neg = ave(reporter_activity_neg, reporter_id, FUN = mean))

reference_activities <- rbind(reference_activities, reference_activities2)

# reference_activities <- merge(cDNA_df2 %>% mutate(tf = gsub("_.*", "", tf)), chosen_models, by = c("condition", "tf")) %>%
#   distinct(tf, cell, reporter_id, commercial_reporter, reporter_activity_neg)

## Include reporter specificities based on KD of related TFs: I want to punish reporters that reduce activity upon KD of related TFs
related_tfs <- data.frame(tf = c("CREB1", "FOS::JUN", "CEBPB", "XBP1", "CEBPB", "FOS::JUN", "CREB1", "ELK1", "ETS2", "ELK1", "ETS2", "ELK1", "FOXA1", "FOXA1", "GATA1", "GATA4", "GATA1", "HNF4A", "IRF3", "TP53", "IRX3",
                                 "KLF4", "SP1", "MAF::NFE2", "NFE2L2", "NFKB1", "NFYA", "NR1H2", "PPARA::RXRA", "NFKB1", "REL", "IRF3", "STAT3", "TCF7", "TCF7L2", "TFCP2L1", "TFCP2L1", "GRHL1", "THRB", "THRA", "TP53", "TFCP2L1",
                                 "E2F1", "E2F1", "SP1", "EGR1", "EGR1", "KLF4", "KLF4", "SP1", "SP1", "NFIA", "HSF1", "KLF4", "MEF2A", "PPARG::RXRA", "ESRRB", "POU2F1", "POU5F1", "POU5F1::SOX2", "PPARG::RXRA",
                                 "PPARA::RXRA", "NR1H2", "NR1H2", "SOX2", "SOX9", "SOX2", "SOX9", "TCF7", "TCF7L2", "TEAD1", "STAT1::STAT2"),
                          related_tf = c("ATF2", "ATF2", "ATF4", "ATF6", "CEBPG", "CREB1", "FOS::JUN", "ETS1", "ELK1", "ETS2", "ETS1", "ETV4", "FOXA2", "FOXP1", "GATA2", "GATA2", "GATA4", "HNF4G", "IRF2", "IRX3", "TP53",
                                         "SP1", "KLF4", "NFE2L2", "MAF::NFE2", "NFKB2", "NFYB", "PPARA", "NR1H2", "REL", "NFKB1", "STAT1::STAT2", "STAT1::STAT2", "TCF7L2", "TCF7", "TFCP2", "GRHL1", "TFCP2L1", "THRA", "THRB", 
                                         "TP73", "UBP1", "E2F2", "E2F3", "EGR1", "SP1", "KLF4", "SP1", "EGR1", "KLF4", "EGR1", "HIC2", "HSF2", "KLF9", "MEF2B", "NR1H2", "NR5A2", "POU5F1", "POU2F1", "POU2F1", "PPARA::RXRA",
                                         "PPARG::RXRA", "PPARA::RXRA", "PPARG::RXRA", "SOX13", "SOX13", "SOX9", "SOX2", "TCF7L1", "TCF7L1", "TEAD2", "STAT3"))

off_target_activities <- cDNA_df2 %>% 
  left_join(reference_condition) %>%
  filter(stimulation == "KD") %>%
  mutate(condition = gsub("HepG2_|mES_", "", condition)) %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  mutate(reporter_dif =  reference_activity - reporter_activity_neg) %>%
  left_join(related_tfs) %>%
  na.omit() %>%
  filter(condition == related_tf) %>%
  mutate(reporter_dif = ave(reporter_dif, reporter_id, FUN = mean)) %>% ## Some TFs have multiple related TFs for which I have KD data: I will compute the mean over the related TFs per reporter
  dplyr::select(reporter_id, tf, commercial_reporter, "off_target_dif" = reporter_dif, "off_target" = related_tf) %>%
  distinct() %>%
  group_by(reporter_id) %>%
  top_n(1) %>%
  ungroup() 

## Also: punish reporters that are active in cells where the TF is not expressed
off_target_activities2 <- cDNA_df2 %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  left_join(tf_rna) %>%
  filter(TPM == 0, stimulation == "no") %>%
  mutate(off_target_activity = ave(reporter_activity_neg, reporter_id, FUN = mean)) %>%
  distinct(reporter_id, off_target_activity, tf, commercial_reporter) %>%
  mutate(off_target_activity = ifelse(off_target_activity < 0, 0, off_target_activity)) %>%
  mutate(off_target = "Expression_0TPM")

off_target_activities <- merge(off_target_activities, off_target_activities2, all = T)
off_target_activities$off_target_dif[is.na(off_target_activities$off_target_dif)] <- 0
off_target_activities$off_target_activity[is.na(off_target_activities$off_target_activity)] <- 0

off_target_activities <- off_target_activities %>%
  pivot_longer(cols = c(off_target_dif, off_target_activity), names_to = "off_target_act") %>%
  dplyr::select(-off_target_act) %>%
  #mutate(off_target_dif = ave(value, reporter_id, FUN = function(x) sum(x))) %>%
  dplyr::select(reporter_id, "off_target_dif" = value, tf, off_target) %>%
  mutate(mean_off_target_dif = ave(off_target_dif, tf, off_target, FUN = mean)) %>%
  group_by(tf) %>%
  arrange(desc(mean_off_target_dif)) %>% 
  top_n(1, wt = mean_off_target_dif) %>% ## Some TFs have multiple perturbations: for those I only keep the strongest off-target effect (on average across all reporters)
  ungroup() %>%
  group_by(reporter_id) %>%
  arrange(desc(off_target_dif)) %>% 
  top_n(1, wt = off_target_dif) %>%
  distinct(reporter_id, off_target_dif, off_target)


cDNA_df6 <- reference_activities %>%
  left_join(perturbation_df) %>%
  left_join(rna_correlations) %>%
  left_join(viper_correlations) %>%
  left_join(off_target_activities) %>%
  mutate(reporter_dif = ifelse(is.na(reporter_dif), 0, reporter_dif)) %>%
  mutate(viper_cor_pval = ifelse(is.na(viper_cor_pval), 0, viper_cor_pval)) %>%
  mutate(off_target_dif = ifelse(is.na(off_target_dif), 0, off_target_dif)) %>%
  mutate(off_target_dif = -off_target_dif) %>%
  mutate(Sensitivity = reporter_dif + reporter_activity_neg) %>%
  mutate(Specificity = reporter_dif + off_target_dif + cor_pval) %>%
  pivot_longer(cols = c("Sensitivity", "Specificity"), names_to = "quality", values_to = "quality_score") %>%
  mutate(quality_score = ifelse(quality_score < 0, 0, quality_score))
```



## Linear model - TF reporter grammar

Aim: If the measured activities can be fitted by a linear model, it is likely that the activity is true. If not, we might be looking at noise instead.

```{r log-linear-model-5, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
reporter_features <- cDNA_df %>%
  distinct(reporter_id, promoter, spacing, distance, background) %>%
  na.omit()

cDNA_df_TF <- cDNA_df %>%
  filter(neg_ctrls == "No", commercial_reporter == "No", hPGK == "No", native_enhancer == "No", stimulation %in% c("no", "pathway")) %>%
  distinct(reporter_activity_neg, promoter, spacing, distance, background, tf, condition, reporter_id, stimulation, reference_condition) %>%
  mutate(tf_condition = paste(tf, condition, sep = "_")) %>%
  # mutate(max_activity = ave(reporter_activity_neg, tf_condition, FUN = max)) %>%
  # mutate(min_activity = ave(reporter_activity_neg, tf_condition, FUN = min)) %>%
  # mutate(dif = max_activity - min_activity) %>%
  # filter(dif > 2.5) %>%
  mutate(reporter_activity_neg = log2(reporter_activity_neg)) %>%
  left_join(reference_condition %>% dplyr::select(-tf), by = c("reporter_id", "reference_condition")) %>%
  mutate(reporter_dif =  reference_activity - reporter_activity_neg) %>%
  mutate(background = as.character(background)) %>%
  mutate(promoter = as.factor(promoter)) %>%
  #mutate(reporter_activity_neg = ifelse(stimulation != "no", reporter_dif, reporter_activity_neg)) %>%
  distinct(reporter_activity_neg, promoter, spacing, distance, background, condition, tf, tf_condition, reporter_id)

## Fit linear model with design features for actual data
exp_variance <- data.frame("feature" = c("promoter", "spacing", "distance", "background", "spacing:background", "Residuals"))
weight <- data.frame("feature" = c("(Intercept)", "promoterminP", "promoterhBGm", "promotermCMV",
                  "spacing10bps","spacing5bp", "distance10bp","distance21bp", "background1","background2", "background3",
                  "spacing5bp:background2", "spacing5bp:background3"))

cDNA_df_TF$promoter <- relevel(cDNA_df_TF$promoter, ref = "minP")
cDNA_df_TF$row <- rownames(cDNA_df_TF)
cDNA_df_TF3 <- data.frame()

lm_list <- list()
for (i in unique(cDNA_df_TF$tf_condition)) {
  if (nrow(cDNA_df_TF[cDNA_df_TF$tf_condition == i,]) > 20) {
  print(i)
  lm_list[[i]] <- lm(reporter_activity_neg ~
                       distance + promoter + spacing*background,
                cDNA_df_TF[cDNA_df_TF$tf_condition == i,])
  }
}

f <- lapply(names(lm_list), function(x) summary(lm_list[[x]])$fstatistic)
names(f) <- names(lm_list)
p <- lapply(names(f), function(x) pf(f[[x]][1],f[[x]][2],f[[x]][3],lower.tail=F))
p <- data.frame(bind_rows(p)) %>% setnames("value", "pval")
p$tf_condition <- names(lm_list)

y <- lapply(names(lm_list), function(x) data.frame("activity_predicted" = lm_list[[x]]$fitted.values) %>% rownames_to_column("row") %>% mutate(tf_condition = x))
y <- data.frame(bind_rows(y)) %>% left_join(p, by = "tf_condition")

cDNA_df_TF2 <- y %>%
  left_join(cDNA_df_TF)

cDNA_df_TF3 <- rbind(cDNA_df_TF3, cDNA_df_TF2)

mean(cDNA_df_TF3$pval)

w <- lapply(names(lm_list), function(x) 
  if (length(lm_list[[x]]$coefficients) == length(summary(lm_list[[x]])$coefficients[,4])) {
  data.frame('weight' = lm_list[[x]]$coefficients, 'pval' = summary(lm_list[[x]])$coefficients[,4]) %>% mutate(tf = x) %>% rownames_to_column('feature')
  }
)

w <- data.frame(bind_rows(w))
weight <- weight %>%
  left_join(w)

x <- lapply(names(lm_list), function(x) anova(lm_list[[x]]) %>% rownames_to_column("feature") %>% mutate(tf = x))
x <- data.frame(bind_rows(x)) %>% 
  setnames(c("Sum.Sq", "Pr..F."), c("sum_sq", "pval")) %>%
  distinct(feature, sum_sq, pval, tf) %>%
  mutate(sum = ave(sum_sq, tf, FUN = sum)) %>%
  mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
  distinct(tf, rel_sum_sq, pval, feature)

exp_variance <- exp_variance %>%
  left_join(x)

## Fine-tune dfs
exp_variance$group <- "no residual"
exp_variance$group[grep("Residual", exp_variance$feature)] <- "residual"
exp_variance$rel_sum_sq[exp_variance$feature == "Residuals"] <- 100 - exp_variance$rel_sum_sq[exp_variance$feature == "Residuals"]
exp_variance$feature <- revalue(exp_variance$feature, c("Residuals"="total"))
prediction <- cDNA_df_TF3 %>%
  dplyr::select(-row) %>%
  na.omit()
weight <- weight %>%
  filter(feature != "(Intercept)")
categorie <- c("promoter", "spacing", "distance", "background")
weight$features <- gsub(paste(categorie, collapse="|"), "",weight$feature)
ml <- length(unique(weight$features))
feature <- unique(weight$features)
weight$condition <- gsub(paste(feature, collapse = "|"), "", weight$feature)
weight <- weight %>% dplyr::select(-feature)


cDNA_df_TF3 <- cDNA_df_TF3 %>%
  mutate(pval_adj = p.adjust(pval, method = "fdr"))

best_models <- cDNA_df_TF3 %>%
  distinct(tf, pval_adj, tf_condition) %>%
  group_by(tf) %>%
  arrange(desc(pval_adj)) %>%
  slice_tail(n = 1)


## Overview figures: which features matter how much and for which TF?
ggplot(exp_variance %>%
         filter(tf %in% best_models$tf_condition) %>%
         filter(group == "no residual") %>%
         mutate(mean_rel_sum_sq = ave(rel_sum_sq, feature, FUN = mean)),
       aes(x = reorder(feature, -mean_rel_sum_sq), y = rel_sum_sq)) +
  geom_quasirandom(color = colors_diverse[2]) +
  stat_summary(fun = median, fun.min = median, fun.max = median, geom = "crossbar", color = "black", width = 0.25, lwd = 0.4) +
  ylab("Variance explained (%)") + xlab("") +
  labs(title = "Linear variance modelling",
       subtitle = "Expression variance predicted by features") +
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12),
        axis.text.y = element_text(size = 10))

exp_variance2 <- exp_variance %>%
  filter(tf %in% best_models$tf_condition) %>%
  distinct(tf, feature, rel_sum_sq) %>%
  spread(key = "feature", value = "rel_sum_sq") %>%
  column_to_rownames("tf")

pheatmap(exp_variance2 %>%
                  t(),
         cellwidth = 10,
         cellheight = 10,
         border_color = F,
         color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
         angle_col = 90,
         cluster_cols = T,
         cutree_rows = 2)


## Update reporter quality score by implementing predictability
## Add reporter feature information
reporter_features <- cDNA_df %>%
  filter(native_enhancer == "No", neg_ctrls == "No", hPGK == "No") %>%
  distinct(reporter_id, promoter, spacing, distance, background)

cDNA_df6 <- cDNA_df6 %>%
  left_join(reporter_features) 

# ## Correct confidence score by linear model pvalue
# chosen_models2 <- chosen_models %>% 
#   mutate(tf_condition = paste(tf, condition, sep = "_"))
chosen_conditions <- cDNA_df6 %>%
  mutate(tf_condition = paste(cDNA_df6$tf, cDNA_df6$cell, sep = "_")) %>%
  mutate(tf_condition = gsub("mES$", "mES_2i_LIF", tf_condition)) %>%
  distinct(tf_condition)
  
pval_prediction <- prediction %>%
  filter(tf_condition %in% chosen_conditions$tf_condition) %>%
  distinct(tf, pval) %>%
  mutate(pval = -log10(pval)) %>%
  mutate(pval = ifelse(pval > 2, 2, pval)) %>%
  dplyr::select(tf, "lm_pval" = pval)

cDNA_df6 <- cDNA_df6 %>%
  distinct() %>%
  left_join(pval_prediction) %>%
  mutate(quality_score = ifelse(lm_pval < -log10(0.1), quality_score / 2, quality_score))

predicted_activities <- prediction %>%
  filter(tf_condition %in% chosen_conditions$tf_condition) %>%
  distinct(reporter_id, activity_predicted) 

cDNA_df6 <- cDNA_df6 %>%
  left_join(predicted_activities)

write_tsv(cDNA_df6, "/DATA/usr/m.trauernicht/projects/SuRE-TF/data/mt20230302_reporter_scores.tsv")
write_tsv(cDNA_df_TF3, "/DATA/usr/m.trauernicht/projects/SuRE-TF/data/mt20230302_predicted_activities.tsv")
write_tsv(weight, "/DATA/usr/m.trauernicht/projects/SuRE-TF/data/mt20230302_lm_weights.tsv")
```


---

## Linear model for specificity/sensitivity
```{r}
# Sensitivity
cDNA_df_TF <- cDNA_df6 %>%
  left_join(reporter_features) %>%
  mutate(background = as.character(background)) %>%
  mutate(promoter = as.factor(promoter)) %>%
  filter(commercial_reporter == "No") %>%
  mutate(tf_condition = paste(tf, cell, sep = "_")) %>%
  distinct(quality_score, quality, promoter, spacing, distance, background, tf, tf_condition, reporter_id) %>%
  filter(quality == "Sensitivity")

## Fit linear model with design features for actual data
exp_variance <- data.frame("feature" = c("promoter", "spacing", "distance", "background", "spacing:background", "Residuals"))
weight <- data.frame("feature" = c("(Intercept)", "promoterminP", "promoterhBGm", "promotermCMV",
                  "spacing10bps","spacing5bp", "distance10bp","distance21bp", "background1","background2", "background3",
                  "spacing5bp:background2", "spacing5bp:background3"))

cDNA_df_TF$promoter <- relevel(cDNA_df_TF$promoter, ref = "minP")
cDNA_df_TF$row <- rownames(cDNA_df_TF)
cDNA_df_TF3 <- data.frame()

lm_list <- list()
for (i in unique(cDNA_df_TF$tf_condition)) {
  if (nrow(cDNA_df_TF[cDNA_df_TF$tf_condition == i,]) > 20) {
  print(i)
  lm_list[[i]] <- lm(quality_score ~
                       distance + promoter + spacing*background,
                cDNA_df_TF[cDNA_df_TF$tf_condition == i,])
  }
}

f <- lapply(names(lm_list), function(x) summary(lm_list[[x]])$fstatistic)
names(f) <- names(lm_list)
p <- lapply(names(f), function(x) pf(f[[x]][1],f[[x]][2],f[[x]][3],lower.tail=F))
p <- data.frame(bind_rows(p)) %>% setnames("value", "pval")
p$tf_condition <- names(lm_list)

y <- lapply(names(lm_list), function(x) data.frame("activity_predicted" = lm_list[[x]]$fitted.values) %>% rownames_to_column("row") %>% mutate(tf_condition = x))
y <- data.frame(bind_rows(y)) %>% left_join(p, by = "tf_condition")

cDNA_df_TF2 <- y %>%
  left_join(cDNA_df_TF)

cDNA_df_TF3 <- rbind(cDNA_df_TF3, cDNA_df_TF2)

mean(cDNA_df_TF3$pval)

w <- lapply(names(lm_list), function(x) 
  if (length(lm_list[[x]]$coefficients) == length(summary(lm_list[[x]])$coefficients[,4])) {
  data.frame('weight' = lm_list[[x]]$coefficients, 'pval' = summary(lm_list[[x]])$coefficients[,4]) %>% mutate(tf = x) %>% rownames_to_column('feature')
  }
)

w <- data.frame(bind_rows(w))
weight <- weight %>%
  left_join(w)

x <- lapply(names(lm_list), function(x) anova(lm_list[[x]]) %>% rownames_to_column("feature") %>% mutate(tf = x))
x <- data.frame(bind_rows(x)) %>% 
  setnames(c("Sum.Sq", "Pr..F."), c("sum_sq", "pval")) %>%
  distinct(feature, sum_sq, pval, tf) %>%
  mutate(sum = ave(sum_sq, tf, FUN = sum)) %>%
  mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
  distinct(tf, rel_sum_sq, pval, feature)

exp_variance <- exp_variance %>%
  left_join(x)

## Fine-tune dfs
exp_variance$group <- "no residual"
exp_variance$group[grep("Residual", exp_variance$feature)] <- "residual"
exp_variance$rel_sum_sq[exp_variance$feature == "Residuals"] <- 100 - exp_variance$rel_sum_sq[exp_variance$feature == "Residuals"]
exp_variance$feature <- revalue(exp_variance$feature, c("Residuals"="total"))
prediction <- cDNA_df_TF3 %>%
  dplyr::select(-row) %>%
  na.omit()
weight <- weight %>%
  filter(feature != "(Intercept)")
categorie <- c("promoter", "spacing", "distance", "background")
weight$features <- gsub(paste(categorie, collapse="|"), "",weight$feature)
ml <- length(unique(weight$features))
feature <- unique(weight$features)
weight$condition <- gsub(paste(feature, collapse = "|"), "", weight$feature)
weight <- weight %>% dplyr::select(-feature)


cDNA_df_TF3 <- cDNA_df_TF3 %>%
  mutate(pval_adj = p.adjust(pval, method = "fdr"))

write_tsv(weight, "/DATA/usr/m.trauernicht/projects/SuRE-TF/data/mt20230302_lm_weights_sensitivity.tsv")


## Specificity
cDNA_df_TF <- cDNA_df6 %>%
  left_join(reporter_features) %>%
  mutate(background = as.character(background)) %>%
  mutate(promoter = as.factor(promoter)) %>%
  filter(commercial_reporter == "No") %>%
  mutate(tf_condition = paste(tf, cell, sep = "_")) %>%
  distinct(quality_score, quality, promoter, spacing, distance, background, tf, tf_condition, reporter_id) %>%
  filter(quality == "Specificity")

## Fit linear model with design features for actual data
exp_variance <- data.frame("feature" = c("promoter", "spacing", "distance", "background", "spacing:background", "Residuals"))
weight <- data.frame("feature" = c("(Intercept)", "promoterminP", "promoterhBGm", "promotermCMV",
                  "spacing10bps","spacing5bp", "distance10bp","distance21bp", "background1","background2", "background3",
                  "spacing5bp:background2", "spacing5bp:background3"))

cDNA_df_TF$promoter <- relevel(cDNA_df_TF$promoter, ref = "minP")
cDNA_df_TF$row <- rownames(cDNA_df_TF)
cDNA_df_TF3 <- data.frame()

lm_list <- list()
for (i in unique(cDNA_df_TF$tf_condition)) {
  if (nrow(cDNA_df_TF[cDNA_df_TF$tf_condition == i,]) > 20) {
  print(i)
  lm_list[[i]] <- lm(quality_score ~
                       distance + promoter + spacing*background,
                cDNA_df_TF[cDNA_df_TF$tf_condition == i,])
  }
}

f <- lapply(names(lm_list), function(x) summary(lm_list[[x]])$fstatistic)
names(f) <- names(lm_list)
p <- lapply(names(f), function(x) pf(f[[x]][1],f[[x]][2],f[[x]][3],lower.tail=F))
p <- data.frame(bind_rows(p)) %>% setnames("value", "pval")
p$tf_condition <- names(lm_list)

y <- lapply(names(lm_list), function(x) data.frame("activity_predicted" = lm_list[[x]]$fitted.values) %>% rownames_to_column("row") %>% mutate(tf_condition = x))
y <- data.frame(bind_rows(y)) %>% left_join(p, by = "tf_condition")

cDNA_df_TF2 <- y %>%
  left_join(cDNA_df_TF)

cDNA_df_TF3 <- rbind(cDNA_df_TF3, cDNA_df_TF2)

mean(cDNA_df_TF3$pval)

w <- lapply(names(lm_list), function(x) 
  if (length(lm_list[[x]]$coefficients) == length(summary(lm_list[[x]])$coefficients[,4])) {
  data.frame('weight' = lm_list[[x]]$coefficients, 'pval' = summary(lm_list[[x]])$coefficients[,4]) %>% mutate(tf = x) %>% rownames_to_column('feature')
  }
)

w <- data.frame(bind_rows(w))
weight <- weight %>%
  left_join(w)

x <- lapply(names(lm_list), function(x) anova(lm_list[[x]]) %>% rownames_to_column("feature") %>% mutate(tf = x))
x <- data.frame(bind_rows(x)) %>% 
  setnames(c("Sum.Sq", "Pr..F."), c("sum_sq", "pval")) %>%
  distinct(feature, sum_sq, pval, tf) %>%
  mutate(sum = ave(sum_sq, tf, FUN = sum)) %>%
  mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
  distinct(tf, rel_sum_sq, pval, feature)

exp_variance <- exp_variance %>%
  left_join(x)

## Fine-tune dfs
exp_variance$group <- "no residual"
exp_variance$group[grep("Residual", exp_variance$feature)] <- "residual"
exp_variance$rel_sum_sq[exp_variance$feature == "Residuals"] <- 100 - exp_variance$rel_sum_sq[exp_variance$feature == "Residuals"]
exp_variance$feature <- revalue(exp_variance$feature, c("Residuals"="total"))
prediction <- cDNA_df_TF3 %>%
  dplyr::select(-row) %>%
  na.omit()
weight <- weight %>%
  filter(feature != "(Intercept)")
categorie <- c("promoter", "spacing", "distance", "background")
weight$features <- gsub(paste(categorie, collapse="|"), "",weight$feature)
ml <- length(unique(weight$features))
feature <- unique(weight$features)
weight$condition <- gsub(paste(feature, collapse = "|"), "", weight$feature)
weight <- weight %>% dplyr::select(-feature)


cDNA_df_TF3 <- cDNA_df_TF3 %>%
  mutate(pval_adj = p.adjust(pval, method = "fdr"))

write_tsv(weight, "/DATA/usr/m.trauernicht/projects/SuRE-TF/data/mt20230302_lm_weights_specificity.tsv")
```













## Other plotting
```{r, fig.width=9, fig.height=9}
#####
ggplot(cDNA_df6,
       aes(x = cor, y = reporter_dif, color = commercial_reporter)) +
  geom_point() +
  geom_smooth(method = "lm") +
  theme_classic_lines_90() +
  facet_wrap(~tf, scales = "free")

ggplot(cDNA_df6,
       aes(x = cor, y = reporter_activity_neg, color = commercial_reporter)) +
  geom_point() +
  geom_smooth(method = "lm") +
  theme_classic_lines_90() +
  facet_wrap(~tf, scales = "free")

ggplot_custom(cDNA_df6,
       aes(x = reorder(tf, -mean_response_score), y = response_score, color = commercial_reporter)) +
  geom_hline(yintercept = quantile(cDNA_df6$response_score, .9, na.rm = T), lty = 2) +
  geom_hline(yintercept = quantile(cDNA_df6$response_score, .75, na.rm = T), lty = 2) +
  geom_hline(yintercept = quantile(cDNA_df6$response_score, .5, na.rm = T), lty = 2) +
  geom_point() +
  theme_pubr(x.text.angle = 90)

ggplot(cDNA_df6 %>%
         distinct(tf, rel_response_score, commercial_reporter, reporter_dif, cor_pval, off_target_dif, reporter_activity_neg) %>%
         filter(rel_response_score > .5) %>%
         pivot_longer(cols = 4:7, names_to = "feature", values_to = "importance"),
       aes(x = feature, y = importance)) +
  geom_quasirandom(aes(color = commercial_reporter)) +
  theme_classic_lines_90() +
  facet_wrap(~tf)

## Set of best reporters
optimal_reporters <- cDNA_df6 %>%
  group_by(tf) %>%
  top_n(3, rel_response_score) %>%
  distinct(reporter_id, tf)

cDNA_df2 <- cDNA_df %>%
  filter(reporter_id %in% optimal_reporters$reporter_id) %>%
  filter(tf != "RFX1") %>%
  dplyr::filter(neg_ctrls == "No", 
                commercial_reporter == "No", 
                hPGK == "No", 
                str_detect(tf, "RANDOM", negate = T), 
                native_enhancer == "No") %>%
  distinct(tf, condition, reporter_activity_neg, stimulation, reference_condition, cell) %>%
  mutate(reporter_activity_neg = log2(ave(reporter_activity_neg, tf, condition, FUN = mean))) %>%
  distinct()

reference_condition <- cDNA_df2 %>%
  filter(condition %in% reference_condition) %>%
  dplyr::select(-reference_condition, "reference_condition" = condition, -stimulation, "reference_activity" = reporter_activity_neg) #%>%
  #filter(abs(reference_activity) > 1)

cDNA_df3 <- cDNA_df2 %>%
  filter(cell == "HepG2") %>%
  filter(stimulation == "KD") %>%
  left_join(reference_condition) %>%
  mutate(reporter_FC = reporter_activity_neg - reference_activity) %>%
  mutate(TF_dif = ave(reporter_FC, tf, condition, FUN = mean)) %>%
  mutate(TF_dif = ifelse(abs(TF_dif) > 0.2, TF_dif, 0)) %>%
  mutate(mean_activity = ave(reporter_activity_neg, tf, condition, FUN = mean)) #%>%
  #filter(mean_activity > 0.5)

cDNA_df3[is.na(cDNA_df3)] <- 0

cDNA_df4 <- cDNA_df3 %>%
  distinct(tf, condition, TF_dif) %>%
  mutate(condition = gsub("HepG2_", "", condition)) %>%
  filter(condition %in% tf) %>%
  filter(tf %in% condition) %>%
  mutate(diag = ifelse(tf == condition, T, F)) %>%
  arrange(diag)

ggplot(cDNA_df4,
       aes(x = tf, y = condition, fill = TF_dif)) +
  geom_tile(color = cDNA_df4$diag, size = .5) +
  scale_color_manual(values = c("black", "white")) +
  scale_fill_gradient2(low = "#99B2DD",
                        mid = "white",
                        high = "#dd6b48",
                       limits = c(-2,2),
                       oob = squish) +
  coord_fixed() +
  xlab("TF reporter") +
  ylab("KD condition") +
  ggtitle("Only TFs that are direct target of KD") +
  theme_pubr(border = T, x.text.angle = 90) +
  labs(fill="FC compared to control - mean of all reporters")
```























---

## TF activities per cell type

Aim: Which TFs are active in which condition?

```{r}
## Heatmap: mean reporter activity per condition
cDNA_df2 <- cDNA_df %>%
  filter(stimulation == "no") %>%
  filter(tf != "RFX1") %>%
  dplyr::filter(neg_ctrls == "No", 
                commercial_reporter == "No", 
                hPGK == "No", 
                str_detect(tf, "RANDOM", negate = T), 
                native_enhancer == "No") %>%
  distinct(tf, condition, reporter_activity_neg) %>%
  mutate(reporter_activity_neg = log2(ave(reporter_activity_neg, tf, condition, FUN = mean))) %>%
  #mutate(pval_neg = -log10(ave(pval_neg, tf, condition, FUN = mean))) %>%
  distinct()

# Compare similar samples
cDNA_df4 <- cDNA_df2 %>%
  spread(condition, reporter_activity_neg)

ggplot(cDNA_df4,
       aes(x = HepG2_NT, y = HepG2)) +
  geom_point()
### Nuclear receptors are more active in HepG2_NT - this must be due to the presence of hormones in FBS

ggplot(cDNA_df4,
       aes(x = mES_2i_LIF, y = mES_Nanog_ctrl)) +
  geom_point()

#cDNA_df2$reporter_activity_neg[cDNA_df2$pval_minP < -log10(0.15) | cDNA_df2$reporter_activity_neg < 1.5 & cDNA_df2$reporter_activity_neg > -0.5] <- 0 # Select only reporters that are significantly active

cDNA_df2$reporter_activity_neg[cDNA_df2$pval_neg < -log10(0.35) | cDNA_df2$reporter_activity_neg < 1.25 & cDNA_df2$reporter_activity_neg > -0.15] <- 0 # Select only reporters that are significantly active

activity_norm <- cDNA_df2 %>%
  filter(condition %in% c("A549", "HCT116", "HEK293", "HepG2", "K562", "MCF7", "mES_2i_LIF", "NPC", "U2OS"), tf == "GBX2") %>%
  dplyr::select("activity_scale" = reporter_activity_neg, condition)
  

cDNA_df3 <- cDNA_df2 %>%
  filter(condition %in% c("A549", "HCT116", "HEK293", "HepG2", "K562", "MCF7", "mES_2i_LIF", "NPC", "U2OS")) %>%
  left_join(activity_norm) %>%
  mutate(sum = ave(reporter_activity_neg, tf, FUN = function(x) sum(x, na.rm = T))) %>%
  dplyr::filter(sum != 0) %>% # Remove TFs for which no reporter has significant activity
  dplyr::select(-sum) %>%
  mutate(reporter_activity_neg = reporter_activity_neg / activity_scale) %>%
    dplyr::select(-activity_scale) %>%
  spread(tf, reporter_activity_neg) %>%
  column_to_rownames('condition') 

myBreaks <- c(seq(-1, 0, length.out=50+1), 
              seq(0.000001, 5, length.out=50))

pheatmap(cDNA_df3 %>% 
                 as.matrix(),
         color = colorRampPalette(c("#F2CC8F", "white", "#3D405B"))(100),
         clustering_method = "ward.D",
         border_color = F, 
         cellwidth = 8, 
         cellheight = 8,
         #annotation_col = tf_clusters, 
         angle_col = 90,
         cutree_cols = 5,
         #cutree_rows = 3,
         breaks = myBreaks)


## Other ways of representing differences between cell types
cDNA_df3 <- cDNA_df2 %>% 
  filter(condition %in% c("A549", "HCT116", "HEK293", "HepG2", "K562", "MCF7", "mES_2i_LIF", "NPC", "U2OS")) %>%
  dplyr::select(-pval_neg) %>%
  mutate(mean_tf_activity = ave(reporter_activity_neg, tf, FUN = mean)) %>%
  mutate(sd_tf_activity = ave(reporter_activity_neg, tf, FUN = sd)) %>%
  mutate(max_tf_activity = ave(reporter_activity_neg, tf, FUN = max)) %>%
  mutate(min_tf_activity = ave(reporter_activity_neg, tf, FUN = min)) %>%
  mutate(dif = max_tf_activity - min_tf_activity)

ggplot(cDNA_df3,
       aes(x = reorder(tf, -mean_tf_activity), y = reporter_activity_neg)) +
  geom_point() +
  theme_classic_lines_90()

ggplot(cDNA_df3 %>%
         distinct(tf, mean_tf_activity, dif),
       aes(x = reorder(tf, -mean_tf_activity), y = dif)) +
  geom_bar(stat = "identity") +
  theme_classic_lines_90()




# pca_res <- prcomp(cDNA_df3 %>% t(), scale. = TRUE)
# loadings <- data.frame('scale' = pca_res$scale) %>%
#   dplyr::filter(scale > 1.5)
# ggbiplot(pca_res, choices = 2:3, ellipse = T)
# 
# importance_pc <- data.frame("PC1" = pca_res[["x"]][,1],
#                          "PC2" = pca_res[["x"]][,2],
#                           "PC3" = pca_res[["x"]][,3],
#                          "PC4" = pca_res[["x"]][,4]) %>%
#   rownames_to_column("tf")
# 
# ggplot(importance_pc,
#        aes(x = reorder(tf, - PC1), y = PC1)) +
#   geom_bar(stat = "identity") +
#   theme_classic_lines_90()
# 
# ggplot(importance_pc,
#        aes(x = reorder(tf, - PC2), y = PC2)) +
#   geom_bar(stat = "identity") +
#   theme_classic_lines_90()
# 
# ggplot(importance_pc,
#        aes(x = reorder(tf, - PC3), y = PC3)) +
#   geom_bar(stat = "identity") +
#   theme_classic_lines_90()
# 
# ggplot(importance_pc,
#        aes(x = reorder(tf, - PC4), y = PC4)) +
#   geom_bar(stat = "identity") +
#   theme_classic_lines_90()
```



## Effect of TF perturbations of TF activity

```{r}
cDNA_df2 <- cDNA_df %>%
  #filter(reporter_id %in% optimal_reporters$reporter_id) %>%
  filter(tf != "RFX1") %>%
  dplyr::filter(neg_ctrls == "No", 
                commercial_reporter == "No", 
                hPGK == "No", 
                str_detect(tf, "RANDOM", negate = T), 
                native_enhancer == "No") %>%
  distinct(tf, condition, reporter_activity_minP, stimulation, reference_condition) %>%
  mutate(reporter_activity_minP = log2(ave(reporter_activity_minP, tf, condition, FUN = mean))) %>%
  distinct()

reference_condition <- cDNA_df2 %>%
  filter(condition %in% reference_condition) %>%
  dplyr::select(-reference_condition, "reference_condition" = condition, -stimulation, "reference_activity" = reporter_activity_minP) #%>%
  #filter(abs(reference_activity) > 1)

cDNA_df3 <- cDNA_df2 %>%
  filter(stimulation == "pathway", condition != "HepG2_T09") %>%
  left_join(reference_condition) %>%
  mutate(reporter_FC = reporter_activity_minP - reference_activity) %>%
  mutate(TF_dif = ave(reporter_FC, tf, condition, FUN = mean)) %>%
  mutate(TF_dif = ifelse(abs(TF_dif) > 0.5, TF_dif, 0)) %>%
  mutate(mean_activity = ave(reporter_activity_minP, tf, condition, FUN = mean)) %>%
  mutate(mean_TF_dif = ave(TF_dif, condition, FUN = mean)) %>%
  mutate(mean_TF_dif2 = ave(TF_dif, tf, FUN = mean)) %>%
  filter(mean_TF_dif != 0, mean_TF_dif2 != 0)

cDNA_df3[is.na(cDNA_df3)] <- 0

cDNA_df3 <- cDNA_df3 %>%
  distinct(tf, condition, TF_dif) %>%
  spread(tf, TF_dif) %>%
  column_to_rownames("condition")

myBreaks <- c(seq(-1, 0, length.out=50+1), 
              seq(0.000001, 1, length.out=50))

pheatmap(cDNA_df3 %>% 
                 as.matrix(),
         color = colorRampPalette(c("#99B2DD", "white", "#E07A5F"))(100),
         clustering_method = "ward.D",
         border_color = F, 
         cellwidth = 8, 
         cellheight = 8,
         #annotation_col = tf_clusters, 
         angle_col = 90,
         #cutree_cols = 5,
         #cutree_rows = 3,
         breaks = myBreaks)

cDNA_df4 <- cDNA_df3 %>%
  na.omit()

pheatmap(cDNA_df4 %>% 
                 as.matrix(),
         color = colorRampPalette(c("#99B2DD", "white", "#E07A5F"))(100),
         clustering_method = "ward.D",
         border_color = F, 
         cellwidth = 8, 
         cellheight = 8,
         #annotation_col = tf_clusters, 
         angle_col = 90,
         #cutree_cols = 5,
         #cutree_rows = 3,
         breaks = myBreaks)

cDNA_df5 <- cDNA_df3 %>%
  t() %>%
  as.data.frame() %>%
  na.omit()

pheatmap(cDNA_df5 %>%
           t(),
         color = colorRampPalette(c("#99B2DD", "white", "#E07A5F"))(100),
         clustering_method = "ward.D",
         border_color = F, 
         cellwidth = 8, 
         cellheight = 8,
         #annotation_col = tf_clusters, 
         angle_col = 90,
         #cutree_cols = 5,
         #cutree_rows = 3,
         breaks = myBreaks)

cDNA_df6 <- cDNA_df3 %>%
  rownames_to_column("condition") %>%
  filter(condition %in% c("mES_2i", "mES_BX", "mES_HQ", "mES_IWP2", "mES_LIF_CH", "mES_LIF_PD", "mES_N2B27", "mES_serum_2i_LIF", "mES_vitC")) %>% 
  column_to_rownames("condition") %>%
  select_if(colSums(.) != 0)

myBreaks <- c(seq(-2, 0, length.out=50+1), 
              seq(0.000001, 2, length.out=50))

pheatmap(cDNA_df6,
         color = colorRampPalette(c("#99B2DD", "white", "#E07A5F"))(100),
         clustering_method = "ward.D",
         border_color = F, 
         cellwidth = 8, 
         cellheight = 8,
         #annotation_col = tf_clusters, 
         angle_col = 90,
         #cutree_cols = 5,
         #cutree_rows = 3,
         breaks = myBreaks)
```


## Effect of TF perturbations of TF activity

```{r}
cDNA_df2 <- cDNA_df %>%
  filter(tf != "RFX1") %>%
  dplyr::filter(neg_ctrls == "No", 
                commercial_reporter == "No", 
                hPGK == "No", 
                str_detect(tf, "RANDOM", negate = T), 
                native_enhancer == "No") %>%
  distinct(tf, condition, reporter_activity_minP, stimulation, reference_condition) %>%
  mutate(reporter_activity_minP = log2(ave(reporter_activity_minP, tf, condition, FUN = mean))) %>%
  distinct()

reference_condition <- cDNA_df2 %>%
  filter(condition %in% reference_condition) %>%
  dplyr::select(-reference_condition, "reference_condition" = condition, -stimulation, "reference_activity" = reporter_activity_minP) #%>%
  #filter(abs(reference_activity) > 1)

cDNA_df3 <- cDNA_df2 %>%
  filter(stimulation == "degron") %>%
  left_join(reference_condition) %>%
  mutate(reporter_FC = reporter_activity_minP - reference_activity) %>%
  mutate(TF_dif = ave(reporter_FC, tf, condition, FUN = mean)) %>%
  mutate(TF_dif = ifelse(abs(TF_dif) > 0.5, TF_dif, 0)) %>%
  mutate(mean_activity = ave(reporter_activity_minP, tf, condition, FUN = mean)) #%>%
  #filter(mean_activity > 0.5)

cDNA_df3[is.na(cDNA_df3)] <- 0

cDNA_df3 <- cDNA_df3 %>%
  distinct(tf, condition, TF_dif) %>%
  spread(tf, TF_dif) %>%
  column_to_rownames("condition")%>%
  select_if(colSums(.) != 0)

myBreaks <- c(seq(-1.5, 0, length.out=50+1), 
              seq(0.000001, 1.5, length.out=50))

pheatmap(cDNA_df3 %>% 
                 as.matrix(),
         color = colorRampPalette(c("#99B2DD", "white", "#E07A5F"))(100),
         clustering_method = "ward.D",
         border_color = F, 
         cellwidth = 8, 
         cellheight = 8,
         #annotation_col = tf_clusters, 
         angle_col = 90,
         #cutree_cols = 5,
         #cutree_rows = 3,
         breaks = myBreaks)
```

## Effect of TF perturbations on TF activity

```{r}
cDNA_df2 <- cDNA_df %>%
  #filter(reporter_id %in% optimal_reporters$reporter_id) %>%
  filter(tf != "RFX1") %>%
  dplyr::filter(neg_ctrls == "No", 
                commercial_reporter == "No", 
                hPGK == "No", 
                str_detect(tf, "RANDOM", negate = T), 
                native_enhancer == "No") %>%
  distinct(tf, condition, reporter_activity_minP, stimulation, reference_condition, cell) %>%
  mutate(reporter_activity_minP = log2(ave(reporter_activity_minP, tf, condition, FUN = mean))) %>%
  distinct()

reference_condition <- cDNA_df2 %>%
  filter(condition %in% reference_condition) %>%
  dplyr::select(-reference_condition, "reference_condition" = condition, -stimulation, "reference_activity" = reporter_activity_minP) #%>%
  #filter(abs(reference_activity) > 1)

cDNA_df3 <- cDNA_df2 %>%
  filter(cell == "HepG2") %>%
  filter(stimulation == "KD") %>%
  left_join(reference_condition) %>%
  mutate(reporter_FC = reporter_activity_minP - reference_activity) %>%
  mutate(TF_dif = ave(reporter_FC, tf, condition, FUN = mean)) %>%
  mutate(TF_dif = ifelse(abs(TF_dif) > 0.2, TF_dif, 0)) %>%
  mutate(mean_activity = ave(reporter_activity_minP, tf, condition, FUN = mean)) #%>%
  #filter(mean_activity > 0.5)

cDNA_df3[is.na(cDNA_df3)] <- 0

cDNA_df4 <- cDNA_df3 %>%
  distinct(tf, condition, TF_dif) %>%
  spread(tf, TF_dif) %>%
  column_to_rownames("condition")%>%
  select_if(colSums(.) != 0)

myBreaks <- c(seq(-1, 0, length.out=50+1), 
              seq(0.000001, 1, length.out=50))

pheatmap(cDNA_df4 %>% 
                 as.matrix(),
         color = colorRampPalette(c("#99B2DD", "white", "#E07A5F"))(100),
         clustering_method = "ward.D",
         border_color = F, 
         cellwidth = 8, 
         cellheight = 8,
         #annotation_col = tf_clusters, 
         angle_col = 90,
         #cutree_cols = 5,
         #cutree_rows = 3,
         breaks = myBreaks)

cDNA_df4 <- cDNA_df3 %>%
  distinct(tf, condition, TF_dif) %>%
  mutate(condition = gsub("HepG2_", "", condition)) %>%
  filter(condition %in% tf) %>%
  filter(tf %in% condition) %>%
  mutate(diag = ifelse(tf == condition, T, F)) %>%
  arrange(diag)

ggplot(cDNA_df4,
       aes(x = tf, y = condition, fill = TF_dif)) +
  geom_tile(color = cDNA_df4$diag, size = .5) +
  scale_color_manual(values = c("black", "white")) +
  scale_fill_gradient2(low = "#99B2DD",
                        mid = "white",
                        high = "#dd6b48",
                       limits = c(-2,2),
                       oob = squish) +
  coord_fixed() +
  xlab("TF reporter") +
  ylab("KD condition") +
  ggtitle("Only TFs that are direct target of KD") +
  theme_pubr(border = T, x.text.angle = 90) +
  labs(fill="FC compared to control - mean of all reporters")
```


```{r}
cDNA_df2 <- cDNA_df %>%
  filter(tf != "RFX1") %>%
  dplyr::filter(neg_ctrls == "No", 
                commercial_reporter == "No", 
                hPGK == "No", 
                str_detect(tf, "RANDOM", negate = T), 
                native_enhancer == "No") %>%
  distinct(tf, condition, reporter_activity_minP, stimulation, reference_condition, cell) %>%
  mutate(reporter_activity_minP = log2(ave(reporter_activity_minP, tf, condition, FUN = mean))) %>%
  distinct()

reference_condition <- cDNA_df2 %>%
  filter(condition %in% reference_condition) %>%
  dplyr::select(-reference_condition, "reference_condition" = condition, -stimulation, "reference_activity" = reporter_activity_minP) #%>%
  #filter(abs(reference_activity) > 1)

cDNA_df3 <- cDNA_df2 %>%
  filter(cell == "mES") %>%
  filter(stimulation == "KD") %>%
  left_join(reference_condition) %>%
  mutate(reporter_FC = reporter_activity_minP - reference_activity) %>%
  mutate(TF_dif = ave(reporter_FC, tf, condition, FUN = mean)) %>%
  mutate(TF_dif = ifelse(abs(TF_dif) > 0.2, TF_dif, 0)) %>%
  mutate(mean_activity = ave(reporter_activity_minP, tf, condition, FUN = mean)) #%>%
  #filter(mean_activity > 0.5)

cDNA_df3[is.na(cDNA_df3)] <- 0

cDNA_df4 <- cDNA_df3 %>%
  distinct(tf, condition, TF_dif) %>%
  spread(tf, TF_dif) %>%
  column_to_rownames("condition")%>%
  select_if(colSums(.) != 0)

myBreaks <- c(seq(-1, 0, length.out=50+1), 
              seq(0.000001, 1, length.out=50))

pheatmap(cDNA_df4 %>% 
                 as.matrix(),
         color = colorRampPalette(c("#99B2DD", "white", "#E07A5F"))(100),
         clustering_method = "ward.D",
         border_color = F, 
         cellwidth = 8, 
         cellheight = 8,
         #annotation_col = tf_clusters, 
         angle_col = 90,
         #cutree_cols = 5,
         #cutree_rows = 3,
         breaks = myBreaks)

cDNA_df4 <- cDNA_df3 %>%
  distinct(tf, condition, TF_dif) %>%
  mutate(condition = gsub("mES_", "", condition)) %>%
  filter(condition %in% tf) %>%
  filter(tf %in% condition) %>%
  mutate(diag = ifelse(tf == condition, T, F)) %>%
  arrange(diag)

ggplot(cDNA_df4,
       aes(x = tf, y = condition, fill = TF_dif)) +
  geom_tile(color = cDNA_df4$diag, size = .5) +
  scale_color_manual(values = c("black", "white")) +
  scale_fill_gradient2(low = "#99B2DD",
                        mid = "white",
                        high = "#dd6b48",
                       limits = c(-1,1),
                       oob = squish) +
  coord_fixed() +
  xlab("TF reporter") +
  ylab("KD condition") +
  ggtitle("Only TFs that are direct target of KD") +
  theme_pubr(border = T, x.text.angle = 90) +
  labs(fill="FC compared to control - mean of all reporters")
```


## Compare degron data to KD data
```{r}
cDNA_df3 <- cDNA_df2 %>%
  filter(condition %in% c("mES_POU5F1", "mES_Pou5f1_AID")) %>%
  left_join(reference_condition) %>%
  mutate(reporter_FC = reporter_activity_minP - reference_activity) %>%
  mutate(TF_dif = ave(reporter_FC, tf, condition, FUN = mean)) %>%
  #mutate(TF_dif = ifelse(abs(TF_dif) > 0.2, TF_dif, 0)) %>%
  mutate(mean_activity = ave(reporter_activity_minP, tf, condition, FUN = mean)) %>%
  distinct(tf, TF_dif, condition) %>%
  spread(key = "condition", value = "TF_dif")

ggplot(cDNA_df3,
       aes(x = mES_POU5F1, y = mES_Pou5f1_AID)) +
  geom_point() +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm")

cDNA_df3 <- cDNA_df2 %>%
  filter(condition %in% c("mES_SOX2", "mES_Sox2_AID")) %>%
  left_join(reference_condition) %>%
  mutate(reporter_FC = reporter_activity_minP - reference_activity) %>%
  mutate(TF_dif = ave(reporter_FC, tf, condition, FUN = mean)) %>%
  #mutate(TF_dif = ifelse(abs(TF_dif) > 0.2, TF_dif, 0)) %>%
  mutate(mean_activity = ave(reporter_activity_minP, tf, condition, FUN = mean)) %>%
  distinct(tf, TF_dif, condition) %>%
  spread(key = "condition", value = "TF_dif")

ggplot(cDNA_df3,
       aes(x = mES_SOX2, y = mES_Sox2_AID)) +
  geom_point() +
  geom_abline(lty = 2) +
  geom_smooth(method = "lm")
```










```{r warning = F}
## Heatmap: mean reporter activity per condition
cDNA_df4 <- cDNA_df %>%
  filter(stimulation == "no") %>%
  filter(tf != "RFX1") %>%
  dplyr::filter(neg_ctrls == "No", 
                commercial_reporter == "No", 
                hPGK == "No", 
                str_detect(tf, "RANDOM", negate = T), 
                native_enhancer == "No") %>%
  distinct(tf, condition, reporter_activity_minP) %>%
  mutate(reporter_activity_minP = ave(reporter_activity_minP, tf, condition, FUN = mean)) %>%
  mutate(log_reporter_activity_minP = log2(reporter_activity_minP)) %>%
  mutate(mean_tf_activity = ave(reporter_activity_minP, tf, FUN = mean)) %>%
  #filter(abs(mean_tf_activity) > 0.5) %>%
  distinct()

nt_df <- cDNA_df4 %>%
  filter(condition == "HepG2_NT") %>%
  mutate(log_reporter_activity_minP = ave(log_reporter_activity_minP, tf, FUN = mean)) %>%
  filter(abs(log_reporter_activity_minP) > 0.15) %>%
  dplyr::select(tf, "nt_activity" = reporter_activity_minP) %>%
  mutate(nt_activity = ave(nt_activity, tf, FUN = mean)) %>%
  distinct()


## First only plot identical matches to check direct KDs
cDNA_df5 <- merge(cDNA_df4, nt_df, all = T) %>%
  na.omit() %>%
  mutate(reporter_FC = log2(reporter_activity_minP/nt_activity)) %>%
  filter(condition %in% tf) %>%
  filter(tf %in% condition) %>%
  distinct(tf, condition, reporter_FC) %>%
  mutate(diag = ifelse(tf == condition, T, F)) %>%
  arrange(diag)

ggplot(cDNA_df5,
       aes(x = tf, y = condition, fill = reporter_FC)) +
  geom_tile(color = cDNA_df5$diag, size = .5) +
  scale_color_manual(values = c("black", "white")) +
  scale_fill_gradient2(low = "#99B2DD",
                        mid = "white",
                        high = "#dd6b48",
                       limits = c(-.5,.5),
                       oob = squish) +
  coord_fixed() +
  xlab("TF reporter") +
  ylab("KD condition") +
  ggtitle("Only TFs that are direct target of KD") +
  theme_pubr(border = T, x.text.angle = 90) +
  labs(fill="FC compared to control - mean of all reporters")


## Now also include indirect targets
cDNA_df5 <- merge(cDNA_df4, nt_df, all = T) %>%  na.omit() %>%
  mutate(reporter_FC = log2(reporter_activity_minP/nt_activity)) %>%
  #filter(condition %in% tf) %>%
  #filter(tf %in% condition) %>%
  distinct(tf, condition, reporter_FC)
 
tf_kd_targets <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf7027_KD_mES/tf_kd_targets_mES.csv") %>% 
  #filter(condition %in% cDNA_df5$condition) %>%
  setnames("target", "tf") %>%
  mutate(is_target = T) %>%
  right_join(cDNA_df5, by = c("condition", "tf")) 

tf_kd_targets$is_target[is.na(tf_kd_targets$is_target)] <- F
tf_kd_targets <- tf_kd_targets %>%
  filter(condition != "NT") %>%
  arrange(is_target)

ggplot(tf_kd_targets,
       aes(x = tf, y = condition, fill = reporter_FC)) +
  geom_tile(color = tf_kd_targets$is_target, size = .5) +
  scale_color_manual(values = c("black", "white")) +
  scale_fill_gradient2(low = "#99B2DD",
                        mid = "grey95",
                        high = "#dd6b48",
                       na.value = "white",
                       limits = c(-.5,.5),
                       oob = squish) +
  coord_fixed() +
  xlab("TF reporter") +
  ylab("KD condition") +
  ggtitle("All 'active' TFs - KD targets highlighted by box") +
  theme_pubr(border = T, x.text.angle = 90) +
  labs(fill="FC compared to control - mean of all reporters")


tf_kd_targets <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6952_KD_HepG2/tf_kd_targets.csv") %>% 
  #filter(condition %in% cDNA_df5$condition) %>%
  setnames("target", "tf") %>%
  mutate(is_target = T) %>%
  right_join(cDNA_df5, by = c("condition", "tf")) 

tf_kd_targets$is_target[is.na(tf_kd_targets$is_target)] <- F
tf_kd_targets <- tf_kd_targets %>%
  filter(condition != "NT") %>%
  arrange(is_target) %>%
  mutate(target_length = ave(is_target, tf, FUN = function(x) length(unique(x)))) %>%
  filter(target_length > 1) %>%
  dplyr::select(-target_length)

ggplot(tf_kd_targets,
       aes(x = tf, y = condition, fill = reporter_FC)) +
  geom_tile(color = tf_kd_targets$is_target, size = .5) +
  scale_color_manual(values = c("black", "white")) +
  scale_fill_gradient2(low = "#99B2DD",
                        mid = "grey95",
                        high = "#dd6b48",
                       na.value = "white",
                       limits = c(-1,1),
                       oob = squish) +
  coord_fixed() +
  xlab("TF reporter") +
  ylab("KD condition") +
  ggtitle("Only targeted TFs - KD targets highlighted by box") +
  theme_pubr(border = T, x.text.angle = 90) +
  labs(fill="FC compared to control - mean of all reporters")


tf_kd_targets <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6952_KD_HepG2/tf_kd_targets.csv") %>% 
  #filter(condition %in% cDNA_df5$condition) %>%
  setnames("target", "tf") %>%
  mutate(is_target = T) %>%
  left_join(cDNA_df5, by = c("condition", "tf")) 

tf_kd_targets$is_target[is.na(tf_kd_targets$is_target)] <- F
tf_kd_targets <- tf_kd_targets %>%
  filter(condition != "NT") %>%
  arrange(is_target)

ggplot(tf_kd_targets,
       aes(x = tf, y = condition, fill = reporter_FC)) +
  geom_tile(color = tf_kd_targets$is_target, size = .5) +
  scale_color_manual(values = c("black", "white")) +
  scale_fill_gradient2(low = "#99B2DD",
                        mid = "grey95",
                        high = "#dd6b48",
                       na.value = "white",
                       limits = c(-1,1),
                       oob = squish) +
  coord_fixed() +
  xlab("TF reporter") +
  ylab("KD condition") +
  ggtitle("Only KD targets") +
  theme_pubr(border = T, x.text.angle = 90) +
  labs(fill="FC compared to control - mean of all reporters")
```

Conclusion: KDs worked to varying degrees. Some were efficient and specific, others were unspecific, or didn't work at all.

---


## Interaction network between TFs
```{r}
# threshold <- 1
# 
# interaction_df <- cDNA_df4 %>%
#   dplyr::select("receiver" = tf, "signal" = condition, TF_dif) %>%
#   distinct() %>%
#   mutate(TF_dif_binary = "0") %>%
#   mutate(TF_dif_binary = ifelse(TF_dif > threshold, "-1", TF_dif_binary)) %>%
#   mutate(TF_dif_binary = ifelse(TF_dif < -threshold, "1", TF_dif_binary)) %>%
#   mutate(TF_dif_binary = ifelse(TF_dif > -threshold & TF_dif < threshold, "0", TF_dif_binary)) %>%
#   mutate(TF_dif_binary = as.numeric(TF_dif_binary)) %>%
#   mutate(TF_dif_binary = (ifelse(receiver == signal, 0, TF_dif_binary))) %>%
#   mutate(sum_receiver = ave(TF_dif_binary, receiver, FUN = function(x) sum(x))) %>%
#   mutate(sum_signal = ave(TF_dif_binary, signal, FUN = function(x) sum(x))) %>%
#   mutate(sum_receiver_size = ave(TF_dif_binary, receiver, FUN = function(x) sum(abs(x)))) %>%
#   mutate(sum_signal_size = ave(TF_dif_binary, signal, FUN = function(x) sum(abs(x)))) 
# 
# 
# interaction_matrix <- interaction_df %>%
#   distinct(TF_dif_binary, receiver, signal) %>%
#   spread(key = "receiver", value = "TF_dif_binary") %>%
#   column_to_rownames("signal")
# 
# pheatmap(interaction_matrix,
#          border_color = NA,
#          color = colorRampPalette(c("#d5bdaf", 'white', "#84a98c"))(100),
#          clustering_method = "ward.D")
# 
# 
# ###
# interaction_nodes <- interaction_df %>%
#   filter(signal == receiver) %>%
#   distinct(receiver, signal, sum_signal, sum_receiver, sum_signal_size, sum_receiver_size) %>%
#   mutate(sum_total = sum_receiver_size + sum_signal_size) %>%
#   mutate(class = ifelse(sum_signal <= 0, "repressor", "activator")) %>%
#   distinct(receiver, sum_total, class) 
# 
# 
# interaction_edges <- interaction_df %>%
#   filter(receiver != signal) %>%
#   distinct(receiver, signal, TF_dif_binary, TF_dif) %>%
#   mutate(weight = 0) %>%
#   mutate(weight = ifelse(TF_dif_binary == -1, 1, weight)) %>%
#   mutate(weight = ifelse(TF_dif_binary == 1, 2, weight)) %>%
#   mutate(class = ifelse(weight == 1, "repressor", "activator")) %>%
#   dplyr::select(signal, receiver, weight, class, TF_dif) %>%
#   filter(weight != 0) %>%
#   mutate(TF_dif = abs(TF_dif))
#   
# interaction_nodes <- interaction_nodes %>%
#   filter(receiver %in% c(unique(interaction_edges$receiver),unique(interaction_edges$signal)))
# 
# routes_igraph <- graph_from_data_frame(d = interaction_edges, vertices = interaction_nodes, directed = TRUE)
# 
# 
# visIgraph(routes_igraph) %>%
#     visNodes(size = 25, shape = "circle") %>%
#     visOptions(highlightNearest = list(enabled = TRUE, hover = TRUE),
#                nodesIdSelection = TRUE) %>%
#     visInteraction(keyboard = TRUE)
# 
# ggraph(routes_igraph, layout = "sugiyama") + 
#   geom_node_point(aes(size = sum_total, color = class)) +
#   geom_edge_link(aes(color = class, width = TF_dif), 
#                  alpha = 0.8,
#                  arrow = arrow(length = unit(3, 'mm')), 
#                  start_cap = circle(2, 'mm'),
#                  end_cap = circle(3, 'mm')) + 
#   scale_edge_width(range = c(0.2, 2)) +
#   scale_radius(limits = c(1, NA), range = c(1, 6)) + 
#   geom_node_text(aes(label = name), repel = TRUE) +
#   scale_edge_color_manual(values = c("activator" = colors_diverse[2], "repressor" = colors_diverse[1])) +
#   scale_color_manual(values = c("activator" = colors_diverse[2], "repressor" = colors_diverse[1])) +
#   theme_graph()

```




## Cistrome TF targets
```{r}
# bc_files = list.files('/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6952_KD_HepG2/cistrome/',
#                        full.names=T, patter='*.txt')
# bc_list <- lapply(bc_files, fread, header = T)
# names(bc_list)<- gsub('.*/cistrome//(.*?).txt', '\\1', bc_files)
# 
# cistrome_df <- bind_rows(bc_list, .id = "TF") %>%
#   distinct(score, symbol, TF)
# 
# cistrome_df <- cistrome_df %>%
#   filter(TF %in% interaction_df$receiver, symbol %in% interaction_df$receiver) %>%
#   mutate(score = ave(score, TF, symbol, FUN = mean)) %>%
#   distinct() %>%
#   filter(TF != symbol) %>%
#   filter(score > 2) %>%
#   mutate(sum_target = as.numeric(ave(symbol, symbol, FUN = function(x) length(x))))
# 
# cistrome_matrix <- cistrome_df %>%
#   dplyr::select(-sum_target) %>%
#   spread(key = "symbol", value = "score") %>%
#   column_to_rownames("TF")
# 
# cistrome_matrix[is.na(cistrome_matrix)] <- 0
# 
# pheatmap(cistrome_matrix,
#          border_color = NA,
#          color = colorRampPalette(c("#d5bdaf", 'white', "#84a98c"))(100),
#          clustering_method = "ward.D")
# 
# cistrome_nodes <- cistrome_df %>%
#   distinct(symbol, sum_target)
# 
# cistrome_edges <- cistrome_df %>%
#   filter(symbol %in% cistrome_nodes$symbol) %>%
#   distinct(TF, symbol, "weight" = score)
# 
# cistrome_igraph <- graph_from_data_frame(d = cistrome_edges, vertices = cistrome_nodes, directed = TRUE)
# 
# ggraph(cistrome_igraph, layout = "sugiyama") + 
#   geom_node_point(aes(size = sum_target)) +
#   geom_edge_link(aes(width = weight), 
#                  alpha = 0.8,
#                  arrow = arrow(length = unit(3, 'mm')), 
#                  start_cap = circle(2, 'mm'),
#                  end_cap = circle(3, 'mm')) + 
#   scale_edge_width(range = c(0.2, 2)) +
#   geom_node_text(aes(label = name), repel = TRUE) +
#   theme_graph()
```


## Cistrome TF targets
```{r}
# bc_files = list.files('/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6952_KD_HepG2/cistrome/',
#                        full.names=T, patter='*.csv')
# bc_list <- lapply(bc_files, fread, header = T)
# names(bc_list)<- gsub('.*/cistrome//.*M_.*_(.*?)hg38.*.csv', '\\1', bc_files)
# 
# cistrome_df <- bind_rows(bc_list, .id = "target") %>%
#   distinct(RP_score, "TF" = Factor, target)
# 
# cistrome_df <- cistrome_df %>%
#   filter(TF %in% interaction_df$receiver, target %in% interaction_df$receiver) %>%
#   mutate(RP_score = ave(RP_score, TF, target, FUN = mean)) %>%
#   distinct() %>%
#   filter(TF != target) %>%
#   #filter(RP_score > 0.3) %>%
#   mutate(sum_target = as.numeric(ave(target, target, FUN = function(x) length(x))))
# 
# cistrome_matrix <- cistrome_df %>%
#   dplyr::select(-sum_target) %>%
#   spread(key = "target", value = "RP_score") %>%
#   column_to_rownames("TF")
# 
# cistrome_matrix[is.na(cistrome_matrix)] <- 0
# 
# pheatmap(cistrome_matrix,
#          border_color = NA,
#          color = colorRampPalette(c("#d5bdaf", 'white', "#84a98c"))(100),
#          clustering_method = "ward.D")
# 
# cistrome_nodes <- cistrome_df %>%
#   distinct(target, sum_target)
# 
# cistrome_edges <- cistrome_df %>%
#   filter(target %in% cistrome_nodes$target) %>%
#   distinct(TF, target, "weight" = RP_score)
# 
# cistrome_igraph <- graph_from_data_frame(d = cistrome_edges, vertices = cistrome_nodes, directed = TRUE)
# 
# ggraph(cistrome_igraph, layout = "sugiyama") + 
#   geom_node_point(aes(size = sum_target)) +
#   geom_edge_link(aes(width = weight), 
#                  alpha = 0.8,
#                  arrow = arrow(length = unit(3, 'mm')), 
#                  start_cap = circle(2, 'mm'),
#                  end_cap = circle(3, 'mm')) + 
#   scale_edge_width(range = c(0.2, 2)) +
#   geom_node_text(aes(label = name), repel = TRUE) +
#   theme_graph()
```



## Reporter activity compared to background

Aim: Display reporter activities relative to background activity per tf and condition. Get an overview over all reporter activities per condition.

```{r reporter_activities, out.width= "80%", fig.align= "center"}
nt_df <- cDNA_df %>%
  dplyr::filter(condition %in% c("NT_1", "NT_2")) %>%
  dplyr::select(reporter_id, "nt_reporter_activity_minP" = reporter_activity_minP) %>%
  mutate(nt_reporter_activity_minP = ave(nt_reporter_activity_minP, reporter_id, FUN = mean)) %>%
  distinct()

cDNA_df2 <- merge(cDNA_df, nt_df, all = T) %>%
  #na.omit() %>%
  filter(neg_ctrls == "No") %>%
  mutate(reporter_FC = log2(reporter_activity_minP/nt_reporter_activity_minP)) %>%
  distinct(tf, reporter_id, condition, reporter_FC, commercial_reporter) %>%
  mutate(mean_tf_activity = ave(reporter_FC, tf, condition, FUN = mean))

## Reporter FC vs reporter activity in NT control - are the most changing reporters the ones with higher activity?
ggplot(merge(cDNA_df, nt_df, all = T) %>%
  filter(neg_ctrls == "No", tf == "SP1", condition == "SP1") %>%
  mutate(reporter_FC = log2(reporter_activity_minP/nt_reporter_activity_minP)) %>%
  distinct(tf, reporter_id, condition, reporter_FC, commercial_reporter, nt_reporter_activity_minP),
       aes(x = reporter_FC, y = nt_reporter_activity_minP)) +
  geom_point()

tf_kd_targets <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf7027_KD_mES/tf_kd_targets_mES.csv") %>%
  #filter(condition %in% cDNA_df5$condition) %>%
  setnames("target", "tf") %>%
  mutate(is_target = T) %>%
  right_join(cDNA_df2, by = c("condition", "tf"))

tf_kd_targets$is_target[is.na(tf_kd_targets$is_target)] <- F
tf_kd_targets <- tf_kd_targets %>%
  filter(condition != "NT") %>%
  arrange(is_target)

# for (i in unique(tf_kd_targets$condition)) {
#   print(i)
#   p <- ggplot(tf_kd_targets %>%
#                 filter(condition == i),
#               mapping = aes(x = reorder(tf, -mean_tf_activity), y = reporter_FC, color = is_target)) +
#     geom_boxplot(aes(alpha = 0.7)) +
#     geom_quasirandom(aes(alpha = 0.7)) +
#     scale_color_manual(values = c(colors_diverse[1], colors_diverse[3])) +
#     ggtitle(i) +
#     theme_classic_lines_90()
#
#   print(p)
# }

tf_kd_targets <- tf_kd_targets %>%
  mutate(tf = gsub("_.*", "", tf))

tf_kd_targets2 <- tf_kd_targets %>%
  filter(tf %in% cDNA_df4$tf) %>%
  mutate(is_target = ifelse(commercial_reporter == "Yes", "commercial", is_target))

for (i in unique(tf_kd_targets2$condition)) {
  print(i)
  p <- ggplot(tf_kd_targets2 %>%
                filter(condition == i),
              mapping = aes(x = reorder(tf, -mean_tf_activity), y = reporter_FC, color = is_target, group = commercial_reporter)) +
    geom_hline(yintercept = 0, lty = 2) +
    geom_quasirandom(dodge.width = .75) +
    scale_color_manual(values = c("FALSE" = colors_diverse[1], "TRUE" = colors_diverse[3], "commercial" = colors_diverse[4])) +
    ggtitle(i) +
    theme_pubr(border = T, x.text.angle = 90)

  print(p)
}

for (i in unique(tf_kd_targets2$tf)) {
  print(i)
  p <- ggplot(tf_kd_targets2 %>%
                filter(tf == i),
              mapping = aes(x = reorder(condition, -mean_tf_activity), y = reporter_FC, color = is_target, group = commercial_reporter)) +
    geom_hline(yintercept = 0, lty = 2) +
    geom_quasirandom(dodge.width = .75, raster.dpi = 600) +
    scale_color_manual(values = c("FALSE" = colors_diverse[1], "TRUE" = colors_diverse[3], "commercial" = colors_diverse[4])) +
    ggtitle(i) +
    theme_pubr(border = T, x.text.angle = 90)

  print(p)
}


```

Conclusion: Some KDs work well, some synthetic reporters are better than the positive controls.

---




<!-- ## Confidence level 1A - linear model -->
<!-- Aim: Predict reporter activities by reporter design using linear models. -->
<!-- I will fit linear models for each TF in each tested condition. Then I will validate these models by checking the p-value and R-squared distribution of 'active' TFs (one reporter at least 5-fold enriched over negative control) vs 'inactive' TFs and randomized data. -->
<!-- ```{r warning = F} -->
<!-- # Step 1A: Fit a linear model for each TF per condition and keep those that have a significant fit -->
<!-- cDNA_df_level1_lm <- cDNA_df %>% -->
<!--   dplyr::filter(condition %in% rownames(cDNA_df4)) %>% -->
<!--   dplyr::filter(promoter != "Random") %>% -->
<!--   dplyr::filter(hPGK == "No", commercial_reporter == "No", native_enhancer == "No", neg_ctrls == "No") %>% -->
<!--   dplyr::filter(!is.na(reporter_activity_minP)) %>% -->
<!--   mutate(pval_minP_adj = p.adjust(pval_minP, method = "fdr")) %>% -->
<!--   mutate(active = ifelse((reporter_activity_minP >= 5 & pval_minP_adj < 0.05) | (reporter_activity_minP <= 0.333 & pval_minP_adj < 0.05), "Yes", "No")) %>% -->
<!--   mutate(tf_condition = paste(tf, condition, sep = "_")) %>% -->
<!--   distinct(reporter_activity_minP, tf, condition, tf_condition, reporter_id, promoter, spacing, distance, background, active) %>% -->
<!--   mutate(reporter_activity_minP_random = ave(reporter_activity_minP, tf_condition, FUN = function(x) sample(x))) -->

<!-- ## Fit linear model with design features for actual data -->
<!-- cDNA_df_level1_lm$row <- rownames(cDNA_df_level1_lm) -->

<!-- lm_list <- list() -->
<!-- for (i in unique(cDNA_df_level1_lm$tf_condition)) { -->
<!--   if (nrow(cDNA_df_level1_lm[cDNA_df_level1_lm$tf_condition == i,]) > 10) { -->
<!--   lm_list[[i]] <- lm(log2(reporter_activity_minP) ~  -->
<!--                        promoter +  -->
<!--                        spacing +  -->
<!--                        distance +  -->
<!--                        background, -->
<!--                 cDNA_df_level1_lm[cDNA_df_level1_lm$tf_condition == i,]) -->
<!--   } -->
<!-- } -->

<!-- lm_pval_df <- data.frame("tf_condition" = names(lm_list)) -->

<!-- for (i in unique(lm_pval_df$tf_condition)) { -->
<!--   lm_pval_df$pval[lm_pval_df$tf_condition == i] <- lmp(lm_list[[i]]) -->
<!--   lm_pval_df$r2_adj[lm_pval_df$tf_condition == i] <- summary(lm_list[[i]])$adj.r.squared -->
<!--   lm_pval_df$rmse[lm_pval_df$tf_condition == i] <- sqrt(mean(lm_list[[i]]$residuals^2)) -->
<!-- } -->

<!-- ### As a check-up: proof that TFs that have reporters with high activity have a better linear model fit (because from low-activity TFs you would expect variability due to noise, and we want to make sure that the model is not fitting noise) -->
<!-- passed_tf_conditions <- cDNA_df_level1_lm %>% -->
<!--   dplyr::filter(active == "Yes") %>% -->
<!--   distinct(tf_condition, active) -->

<!-- lm_pval_df <- merge(lm_pval_df, passed_tf_conditions, all = T) -->

<!-- lm_pval_df$active[is.na(lm_pval_df$active)] <- "No" -->

<!-- lm_pval_df$pval_adj <- p.adjust(lm_pval_df$pval, method = "fdr") -->

<!-- ## Fit linear model with design features for randomly sampled data -->
<!-- lm_list_random <- list() -->
<!-- for (i in unique(cDNA_df_level1_lm$tf_condition)) { -->
<!--   if (nrow(cDNA_df_level1_lm[cDNA_df_level1_lm$tf_condition == i,]) > 10) { -->
<!--   lm_list_random[[i]] <- lm(log2(reporter_activity_minP_random) ~  -->
<!--                        promoter +  -->
<!--                        spacing +  -->
<!--                        distance +  -->
<!--                        background, -->
<!--                 cDNA_df_level1_lm[cDNA_df_level1_lm$tf_condition == i,]) -->
<!--   } -->
<!-- } -->

<!-- lm_pval_df_random <- data.frame("tf_condition" = names(lm_list_random)) -->

<!-- for (i in unique(lm_pval_df_random$tf_condition)) { -->
<!--   lm_pval_df_random$pval[lm_pval_df_random$tf_condition == i] <- lmp(lm_list_random[[i]]) -->
<!--   lm_pval_df_random$r2_adj[lm_pval_df_random$tf_condition == i] <- summary(lm_list_random[[i]])$adj.r.squared -->
<!--   lm_pval_df_random$rmse[lm_pval_df_random$tf_condition == i] <- sqrt(mean(lm_list_random[[i]]$residuals^2)) -->
<!-- } -->

<!-- ### As a check-up: proof that TFs that have reporters with high activity have a better linear model fit (because from low-activity TFs you would expect variability due to noise, and we want to make sure that the model is not fitting noise) -->
<!-- lm_pval_df_random <- merge(lm_pval_df_random, passed_tf_conditions) %>% -->
<!--   mutate(active = "random") -->

<!-- lm_pval_df_random$pval_adj <- p.adjust(lm_pval_df_random$pval, method = "fdr") -->

<!-- lm_pval_df <- rbind(lm_pval_df, lm_pval_df_random) -->

<!-- ## Sanity checks linear model -->
<!-- ggplot(lm_pval_df, -->
<!--               aes(x = pval, color = active)) + -->
<!--   geom_density(data = lm_pval_df %>% filter(active == "random"), color = colors_diverse[1], size = 1) + -->
<!--   geom_density(data = lm_pval_df %>% filter(active == "No"), color = colors_diverse[2], size = 1) + -->
<!--   geom_density(data = lm_pval_df %>% filter(active == "Yes"), color = colors_diverse[3], size = 1) + -->
<!--   theme_pubr() -->

<!-- ggplot(mapping = aes(x = r2_adj)) + -->
<!--   geom_density(data = lm_pval_df %>% filter(active == "random"), color = colors_diverse[1], size = 1) + -->
<!--   geom_density(data = lm_pval_df %>% filter(active == "No"), color = colors_diverse[2], size = 1) + -->
<!--   geom_density(data = lm_pval_df %>% filter(active == "Yes"), color = colors_diverse[3], size = 1) + -->
<!--   theme_pubr() -->

<!-- ggplot_custom(lm_pval_df %>% -->
<!--                 mutate(pval = as.numeric(pval)) %>% -->
<!--                 mutate(sign = ifelse(pval_adj < 0.05, "Yes", "No")), -->
<!--               aes(x = active, fill = sign)) + -->
<!--   geom_bar(stat = "count") + -->
<!--   ylab("Number of linear models") -->
<!-- ``` -->

<!-- Conclusion: The majority of linear models has a significant fit if at least one reporter is >5-fold enriched over its negative control. That means that these linear models work really well in predicting reporter activity. If no reporter is >5-fold activated the performance of the model drops, which is expected because we want to avoid fitting noise. Randomized data can be fitted significantly in ~5% of the cases, which is expected by random chance. -->


<!-- --- -->


<!-- ## Reporter activity distribution per TF -->

<!-- Aim: Is there a large variance in reporter activities for the same TF? How does the distribution look for all TFs? -->

<!-- ```{r warning = F} -->
<!-- reporter_heatmap <- cDNA_df %>% -->
<!--   filter(condition == "mES_2i_LIF") %>% -->
<!--   filter(native_enhancer == "No", hPGK == "No", neg_ctrls == "No", commercial_reporter == "No") %>% -->
<!--   #filter(commercial_reporter == "No") %>% -->
<!--   mutate(tf = gsub("(.*)_.*", "\\1", tf)) %>% -->
<!--   mutate(tf = gsub("(.*)_.*", "\\1", tf)) %>% -->
<!--   filter(!tf %in% c("TFEB", "SRE", "HIF1A", "ATF4", "CREL", "STAT4", "STAT5", "ATF2", "ATF6", "Myc", "STAT6", "NFE2", -->
<!--                     "IRF1", "STAT1.50", "STAT1", "CLOCK", "GATA3", "STAT1.49", "SREBP", "SREBF1", "PPARG::PPARG", "CEBPA")) %>% -->
<!--   #filter(promoter != "hBGm") %>% -->
<!--   filter(str_detect(tf, "RANDOM", negate = T)) %>% -->
<!--   mutate(tf_condition = paste(tf, condition, sep = "_")) %>% -->
<!--   distinct(reporter_id, reporter_activity_minP, tf, pval_adj_minP, condition, tf_condition) %>% -->
<!--   mutate(reporter_activity_minP = log2(reporter_activity_minP)) %>% -->
<!--   mutate(pval_adj_minP = -log10(pval_adj_minP)) %>% -->
<!--   mutate(reporter_id = gsub("^.*_([0-9]{1,2}bp_[0-9]{1,2}bp)", "\\1", reporter_id)) -->

<!-- lm_pval <- lm_pval_df %>% -->
<!--   filter(str_detect(tf_condition, "mES_2i_LIF")) %>% -->
<!--   filter(active != "random") %>% -->
<!--   filter(str_detect(tf_condition, "RANDOM", negate = T)) %>% -->
<!--   mutate(sign = ifelse(pval_adj < 0.05, "yes", "no")) %>% -->
<!--   distinct(tf_condition, sign, pval_adj) -->

<!-- reporter_heatmap <- merge(reporter_heatmap, lm_pval, all = T) -->


<!-- reporter_heatmap$act_class[reporter_heatmap$pval_adj_minP < 1.30103 | reporter_heatmap$reporter_activity_minP < 1] <- 0 -->
<!-- reporter_heatmap$act_class[reporter_heatmap$pval_adj_minP >= 1.30103 & reporter_heatmap$reporter_activity_minP >= log2(2)] <- 1 -->
<!-- reporter_heatmap$act_class[reporter_heatmap$pval_adj_minP >= 1.30103 & reporter_heatmap$reporter_activity_minP < -log2(2)] <- -1 -->
<!-- reporter_heatmap$act_class[reporter_heatmap$pval_adj_minP >= 1.30103 & reporter_heatmap$reporter_activity_minP >= log2(5)] <- 2 -->

<!-- #reporter_heatmap$reporter_activity_minP[reporter_heatmap$pval_adj_minP < 2] <- 0 -->

<!-- ggplot(data = reporter_heatmap %>% -->
<!--                 mutate(act_class = as.character(act_class)) %>% -->
<!--                 mutate(tf_activity = ave(reporter_activity_minP, tf, FUN = mean), -->
<!--                        sd_activity = ave(reporter_activity_minP, tf, FUN = function(x) sd(x)))) + -->
<!--   geom_hline(yintercept = 1, lty = "twodash") +  -->
<!--   geom_hline(yintercept = -1, lty = "twodash") +  -->
<!--   geom_quasirandom(mapping = aes(x = reorder(tf, -tf_activity), y = reporter_activity_minP, fill = act_class, alpha = sign), shape = 21, color = "black", size = 2, width = 0.25) + -->
<!--   theme_pubr(x.text.angle = 90) + -->
<!--   scale_alpha_discrete(range = c(0.2,1)) + -->
<!--   scale_fill_manual(values = c("#99B2DD", colors_diverse[1], "#EE9781", "#e76f51")) -->
<!-- ``` -->

<!-- Conclusion: There is a large spread for most TFs. For some this might be due to noise, for others the spread seems to be extraordinarily large. We need to investigate further for which TFs we can actually explain the variance and for which not.  -->

<!-- --- -->

<!-- ## Confidence level 1A - linear model -->
<!-- Aim: Many linear models produce good fits, however, within these linear models we might not be able to predict all reporter activities equally well. Some individual reporters might be far away from the fitted regression. To exclude individual reporters from which we cannot predict its activity, we will look at the RMSE of individual linear models and exclude individual reporters that deviate >2xRMSEs.  -->
<!-- ```{r warning = F} -->
<!-- # Step 1B: Use the RMSE to keep only reporters that can be predicted well by the linear model -->
<!-- rmse_df <- lm_pval_df %>% -->
<!--   filter(active != "random") %>% -->
<!--   mutate(sign = ifelse(pval_adj < 0.05, "Yes", "No")) %>% -->
<!--   distinct(tf_condition, rmse, sign) -->

<!-- cDNA_df_level1_lm_rmse <- merge(cDNA_df_level1_lm, rmse_df, all = T)  -->

<!-- predicted_values <- data.frame() -->
<!-- for (i in unique(cDNA_df_level1_lm$tf_condition)) { -->
<!--     if (nrow(cDNA_df_level1_lm[cDNA_df_level1_lm$tf_condition == i,]) > 10) { -->
<!--   predicted <- data.frame("predicted_activity" = lm_list[[i]]$fitted.values) %>%  -->
<!--     rownames_to_column("row") %>% -->
<!--     mutate(tf_condition = i) -->
<!--   predicted_values <- rbind.fill(predicted_values, predicted) -->
<!--     } -->
<!-- } -->

<!-- cDNA_df_level1_lm_rmse <- merge(cDNA_df_level1_lm_rmse, predicted_values) -->


<!-- ggplot(cDNA_df_level1_lm_rmse %>% -->
<!--                 filter(tf_condition == "STAT3_mES_2i_LIF"), -->
<!--               aes(x = log2(reporter_activity_minP), y = predicted_activity)) + -->
<!--   annotate("rect",xmin=log2(0.333),xmax=log2(5),ymin=-Inf,ymax=Inf, alpha=0.05, fill="black") + -->
<!--   geom_abline(lty = "twodash") + -->
<!--   geom_smooth(method = "lm", color = "black") + -->
<!--   geom_smooth(formula = y~x, method = "lm", position = position_nudge(y = 1), color = "red", se = F, size = 0.5) + -->
<!--   geom_smooth(formula = y~x, method = "lm", position = position_nudge(y = -1), color = "red", se = F, size = 0.5) + -->
<!--   geom_vline(xintercept = log2(5), lty = "dashed") + -->
<!--   geom_vline(xintercept = log2(0.333), lty = "dashed") + -->
<!--   geom_point(shape = 21, size = 2, aes(fill = active)) + -->
<!--   scale_fill_manual(values = c(colors_diverse[1], colors_diverse[2])) + -->
<!--   scale_color_manual(values = c("black", "red")) + -->
<!--   theme_pubr() -->


<!-- ggplot(cDNA_df_level1_lm_rmse %>% -->
<!--                 filter(tf_condition == "REL_K562_DMSO"), -->
<!--              aes(x = log2(reporter_activity_minP), y = predicted_activity)) + -->
<!--   annotate("rect",xmin=log2(0.333),xmax=log2(5),ymin=-Inf,ymax=Inf, alpha=0.05, fill="black") + -->
<!--   geom_abline(lty = "twodash") + -->
<!--   geom_smooth(method = "lm", color = "black") + -->
<!--   geom_smooth(formula = y~x, method = "lm", position = position_nudge(y = 1), color = "red", se = F, size = 0.5) + -->
<!--   geom_smooth(formula = y~x, method = "lm", position = position_nudge(y = -1), color = "red", se = F, size = 0.5) + -->
<!--   geom_vline(xintercept = log2(5), lty = "dashed") + -->
<!--   geom_vline(xintercept = log2(0.333), lty = "dashed") + -->
<!--   geom_point(shape = 21, size = 2, aes(fill = active)) + -->
<!--   scale_fill_manual(values = c(colors_diverse[1], colors_diverse[2])) + -->
<!--   scale_color_manual(values = c("black", "red")) + -->
<!--   theme_pubr() + -->
<!--   xlim(0,6) + -->
<!--   ylim(0,6) -->





<!-- ggplot(cDNA_df_level1_lm_rmse %>% -->
<!--                 filter(tf == "STAT3"), -->
<!--               aes(x = log2(reporter_activity_minP), y = predicted_activity)) + -->
<!--   annotate("rect",xmin=log2(0.333),xmax=log2(5),ymin=-Inf,ymax=Inf, alpha=0.05, fill="black") + -->
<!--   geom_abline(lty = "twodash") + -->
<!--   geom_smooth(method = "lm", color = "black") + -->
<!--   geom_smooth(formula = y~x, method = "lm", position = position_nudge(y = 1), color = "red", se = F, size = 0.5) + -->
<!--   geom_smooth(formula = y~x, method = "lm", position = position_nudge(y = -1), color = "red", se = F, size = 0.5) + -->
<!--   geom_vline(xintercept = log2(5), lty = "dashed") + -->
<!--   geom_vline(xintercept = log2(0.333), lty = "dashed") + -->
<!--   geom_point(shape = 21, size = 2, aes(fill = active)) + -->
<!--   scale_fill_manual(values = c(colors_diverse[1], colors_diverse[2])) + -->
<!--   scale_color_manual(values = c("black", "red")) + -->
<!--   facet_wrap(~tf_condition, scales = "free")+ -->
<!--   theme_pubr() + -->
<!--   xlim(-2.5,6) + -->
<!--   ylim(-2.5,6) -->


<!-- for (i in unique(cDNA_df_level1_lm_rmse$condition)) { -->
<!--   p <- ggplot(cDNA_df_level1_lm_rmse %>% -->
<!--                 filter(condition == i) %>% -->
<!--                 filter(sign == "Yes"), -->
<!--               aes(x = log2(reporter_activity_minP), y = predicted_activity)) + -->
<!--   annotate("rect",xmin=-Inf,xmax=log2(5),ymin=-Inf,ymax=Inf, alpha=0.05, fill="black") + -->
<!--   geom_abline(lty = "twodash") + -->
<!--   geom_smooth(method = "lm", color = "black") + -->
<!--   geom_smooth(formula = y~x, method = "lm", position = position_nudge(y = 1), color = "red", se = F, size = 0.5) + -->
<!--   geom_smooth(formula = y~x, method = "lm", position = position_nudge(y = -1), color = "red", se = F, size = 0.5) + -->
<!--   geom_vline(xintercept = log2(5), lty = "dashed") + -->
<!--   geom_point(shape = 21, size = 2, aes(fill = active)) + -->
<!--   scale_fill_manual(values = c(colors_diverse[1], colors_diverse[2])) + -->
<!--   scale_color_manual(values = c("black", "red")) + -->
<!--   theme_pubr() + -->
<!--     ggtitle(i) + -->
<!--     facet_wrap(~tf) -->

<!--   print(p) -->
<!-- } -->
<!-- ``` -->

<!-- Conclusion: With the RMSE method we can easily filter out individual reporters that don't match the regression. Visualizing the linear model by plotting the predicted activity against the actual activity is a good way of inspecting the quality of the individual models. For many TFs the fits look good, however there are some cases (like EGR1) where the variance in reporter activities cannot be predicted by the model.  -->

<!-- --- -->

<!-- ## Confidence level 1B: -->
<!-- Aim: We now only kept reporters from which we can predict its activity. The next step is now to remove all reporters that have a low enrichment over its negative control (less than 5-fold). We can then plot how many reporters are kept per TF in each filtering step. -->
<!-- ```{r warning = F, message = F} -->
<!-- ## Do the actual filtering -->
<!-- lm_pval_df_sign <- lm_pval_df %>% -->
<!--   filter(pval_adj < 0.05) -->

<!-- cDNA_df_level1_lm_rmse_filt <- cDNA_df_level1_lm_rmse %>% -->
<!--   filter(tf_condition %in% lm_pval_df_sign$tf_condition) -->


<!-- cDNA_df_level1_lm_rmse_filt <- cDNA_df_level1_lm_rmse_filt %>% -->
<!--   filter(lm_outlier == "No") -->
<!-- ## So, what we do is filter for significant linear models, and then remove the reporters from the significant linear models that are > 2xRMSE from the fit (= reporters from which we cannot predict the activity well, this cutoff is quite relaxed of course and could be strengthened to e.g. 1xRMSE) -->



<!-- # Step 1C: Is the reporter 5-fold enriched over its mutated control? -->
<!-- cDNA_df_level1_enriched <- cDNA_df_level1_lm_rmse_filt %>% -->
<!--   filter(active == "Yes") -->



<!-- # Plot how many reporters are passing each filtering step -->
<!-- ## Count reporters at start -->
<!-- tf_reporter_counts <- cDNA_df %>% -->
<!--   filter(condition %in% c("HepG2_CDCA", "HepG2_DMSO", "K562_DMSO", "mES_2i_LIF", "mES_2i", "mES_HQ", "mES_LIF_CH", "mES_LIF_PD", "mES_N2B27")) %>% -->
<!--   filter(hPGK == "No", commercial_reporter == "No", native_enhancer == "No", neg_ctrls == "No") %>% -->
<!--   filter(!is.na(reporter_activity_minP)) %>% -->
<!--   distinct(tf, condition, reporter_id) %>% -->
<!--   mutate(tf_condition = paste(tf, condition, sep = "_")) %>% -->
<!--   mutate(level1_A = as.numeric(ave(tf_condition, tf_condition, FUN = function(x) length(x)))) %>% -->
<!--   distinct(tf_condition, level1_A, tf, condition) -->

<!-- # ## Count reporters with good linear model -->
<!-- # tf_reporter_counts <- cDNA_df_level1_lm %>% -->
<!-- #   distinct(tf, condition, reporter_id, tf_condition) %>% -->
<!-- #   mutate(level1_lm_count = as.numeric(ave(tf_condition, tf_condition, FUN = function(x) length(x)))) %>% -->
<!-- #   distinct(tf_condition, level1_lm_count, tf, condition) %>% -->
<!-- #   right_join(tf_reporter_counts) %>% -->
<!-- #   replace(is.na(.), 0) -->

<!-- ## Count reporters that fit linear model -->
<!-- tf_reporter_counts <- cDNA_df_level1_lm_rmse_filt %>% -->
<!--   distinct(tf, condition, reporter_id, tf_condition) %>% -->
<!--   mutate(level1_B = as.numeric(ave(tf_condition, tf_condition, FUN = function(x) length(x)))) %>% -->
<!--   distinct(tf_condition, level1_B, tf, condition) %>% -->
<!--   right_join(tf_reporter_counts) %>% -->
<!--   replace(is.na(.), 0) -->

<!-- ## Count reporters with significant enrichment -->
<!-- tf_reporter_counts <- cDNA_df_level1_enriched %>% -->
<!--   distinct(tf, condition, reporter_id, tf_condition) %>% -->
<!--   mutate(level1_C = as.numeric(ave(tf_condition, tf_condition, FUN = function(x) length(x)))) %>% -->
<!--   distinct(tf_condition, level1_C, tf, condition) %>% -->
<!--   right_join(tf_reporter_counts) %>% -->
<!--   replace(is.na(.), 0) -->

<!-- ## Take only the maximum counts per condition  -->
<!-- tf_reporter_counts <- tf_reporter_counts %>% -->
<!--   distinct(tf, level1_A, level1_B, level1_C) %>% -->
<!--   group_by(tf) %>%  -->
<!--   dplyr::slice(which.max(level1_C)) %>% -->
<!--   ungroup() -->

<!-- pheatmap(tf_reporter_counts %>% -->
<!--            column_to_rownames("tf") %>% -->
<!--            mutate(level1_B = level1_B / level1_A) %>% -->
<!--            mutate(level1_C = level1_C / level1_A) %>% -->
<!--            mutate(level1_A = level1_A / level1_A) %>%  -->
<!--            t(), -->
<!--          border_color = NA, -->
<!--          color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100), -->
<!--          cellwidth = 8, -->
<!--          cellheight = 8,  -->
<!--          cluster_rows = F, -->
<!--          cluster_cols = T) -->

<!-- tf_reporter_counts_long <- tf_reporter_counts %>% -->
<!--   mutate(level1_B = level1_B / level1_A) %>% -->
<!--   mutate(level1_C = level1_C / level1_A) %>% -->
<!--   mutate(level1_A = level1_A / level1_A) %>%  -->
<!--   #filter(level1_B > 0) %>% -->
<!--   pivot_longer(contains("level"), values_to = "reporter_counts", names_to = "confidence_level") -->

<!-- ggplot_custom(tf_reporter_counts_long, -->
<!--               aes(x = confidence_level, y = reporter_counts, group = 1)) + -->
<!--     geom_point() + -->
<!--     geom_line() + -->
<!--     facet_wrap(~tf) + -->
<!--     theme_pubr(border = T, x.text.angle = 90)  -->

<!-- tf_reporter_counts_long2 <- tf_reporter_counts_long %>% -->
<!--   mutate(reporter_counts = as.character(reporter_counts)) %>% -->
<!--   mutate(rel_count = as.numeric(ave(tf, confidence_level, reporter_counts, FUN = function(x) length(x)))) %>% -->
<!--   mutate(reporter_counts = as.numeric(reporter_counts)) -->

<!-- rel_counts <- tf_reporter_counts_long2 %>% -->
<!--   distinct(tf, confidence_level, rel_count) %>% -->
<!--   filter(confidence_level != "level1_A") %>% -->
<!--   setnames("rel_count", "rel_count2") -->

<!-- rel_counts$confidence_level[rel_counts$confidence_level == "level1_B"] <- "level1_A" -->
<!-- rel_counts$confidence_level[rel_counts$confidence_level == "level1_C"] <- "level1_B" -->

<!-- tf_reporter_counts_long2 <- merge(tf_reporter_counts_long2, rel_counts, all = T) -->


<!-- ggplot_custom(tf_reporter_counts_long2, -->
<!--               aes(x = confidence_level, y = reporter_counts, group = tf)) + -->
<!--     geom_point(aes(size = rel_count)) + -->
<!--     geom_line(aes(alpha = rel_count2)) + -->
<!--     theme_pubr()  -->

<!-- ggplot_custom(tf_reporter_counts_long2 %>% -->
<!--                 filter(confidence_level == "level1_C") %>% -->
<!--                 distinct(tf, confidence_level, reporter_counts) %>% -->
<!--                 mutate(level1_pass = ifelse(reporter_counts > 0, ifelse(reporter_counts > 0.5, ">50% passed reporters", "<50% passed reporters"), "all removed")) %>% -->
<!--                 mutate(tf_count = as.numeric(ave(level1_pass, level1_pass, FUN = function(x) length(x)))) %>% -->
<!--                 distinct(level1_pass, tf_count), -->
<!--               aes(x = level1_pass, y = tf_count)) + -->
<!--   geom_bar(stat = "identity") -->


<!-- ggplot_custom(tf_reporter_counts %>% -->
<!--                 distinct(tf, level1_C) %>% -->
<!--                 mutate(reporters_per_tf = ifelse(level1_C > 0, ">= 1 reporter", "0 reporters")) %>% -->
<!--                 mutate(tf_count = as.numeric(ave(reporters_per_tf, reporters_per_tf, FUN = function(x) length(x)))) %>% -->
<!--                 distinct(reporters_per_tf, tf_count), -->
<!--               aes(x = reporters_per_tf, y = tf_count)) + -->
<!--   geom_bar(stat = "identity") -->

<!-- ``` -->

<!-- Conclusion: For 43 TFs we keep at least one reporter in confidence level 1. For 20 TFs we even keep >50% of the reporters. 20 TFs are removed because the linear models do not predict activity and from the 63 TFs from which we can build good models another 20 TFs are removed due to low activity.  -->

<!-- --- -->

<!-- ## Confidence level 2A - direct stimulation -->
<!-- Aim: We want to show that reporters are directly responding to upstream signaling pathway stimulations. For some TFs, we added specific compounds to stimulate pathways, are all reporters for those TFs responding? -->
<!-- ```{r warning = F} -->
<!-- tf_stimulations <- read_csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/tf_stimulation_overview.csv") %>% -->
<!--   filter(tf %in% cDNA_df_level1_enriched$tf) %>% -->
<!--   filter(!is.na(stimulated_condition)) %>% -->
<!--   mutate(tf_condition = paste(tf, stimulated_condition, sep = "_")) %>% -->
<!--   distinct(tf, tf_condition) %>% -->
<!--   mutate(stimulation = "stimulated") -->

<!-- tf_stimulations2 <- read_csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/tf_stimulation_overview.csv") %>% -->
<!--   filter(tf %in% cDNA_df_level1_enriched$tf) %>% -->
<!--   filter(!is.na(stimulated_condition)) %>% -->
<!--   mutate(tf_condition = paste(tf, control_condition, sep = "_")) %>% -->
<!--   distinct(tf, tf_condition) %>% -->
<!--   mutate(stimulation = "unstimulated") -->

<!-- tf_stimulations <- rbind(tf_stimulations, tf_stimulations2) -->

<!-- cDNA_df_level2 <- merge(cDNA_df_level1_lm, tf_stimulations) %>% -->
<!--   distinct(tf, reporter_id, stimulation, reporter_activity_minP) %>% -->
<!--   spread(key = "stimulation", value = "reporter_activity_minP") %>% -->
<!--   mutate(fc = log2(stimulated / unstimulated)) -->

<!-- ggplot_custom(cDNA_df_level2, -->
<!--               aes(x = tf, y = fc)) + -->
<!--   annotate("rect",xmin=-Inf,xmax=Inf,ymin=-Inf,ymax=log2(5), alpha=0.05, fill="black") + -->
<!--   geom_hline(lty = "twodash", yintercept = log2(5)) + -->
<!--   geom_quasirandom() + -->
<!--   ylab("stimulated / unstimulated (log2)") + -->
<!--   theme_pubr(x.text.angle = 45) -->
<!-- ``` -->

<!-- Conclusion:  -->

<!-- --- -->

<!-- ## Confidence level 2B - Correlation -->

<!-- Aim: We cannot directly stimulate all TFs - therefore we want to indirectly approximate TF activities and correlate those with the reporter activities. So, we regard the inferred TF activities as 'true' and then compute for each TF the correlations of each reporter with the inferred TF activities across conditions. Reporters that have a significant correlation will be regarded as 'confident' reporters.  -->

<!-- We can take different approaches to infer TF activities. Those are our options: -->
<!--   1) TF expression from RNA-seq -->
<!--   2) TF target accessibility from ATAC-seq (diffTF) -->
<!--   3) TF regulon expression using VIPER -->
<!--   4) cistrome expression (Aerts lab) -->
<!--   5) Other MPRA-based TF activities (Ernst et al, Meijsing et al) -->


<!-- ### 2B.1: RNA-seq -->
<!-- Correlate each reporter activity with the corresponding TF RNA expression across condition - label significant correlations as confident. -->

<!-- ```{r warning = F} -->

<!-- ``` -->

<!-- Conclusion: -->

<!-- --- -->

<!-- ### 2B.2: Compare reporter activities to ATAC-seq -->
<!-- Correlate each reporter activity with the corresponding TF target gene accessibility across condition - label significant correlations as confident. -->

<!-- ```{r atac_comparison, out.width= "80%", fig.align= "center"} -->
<!-- # DiffTF was run in basic and classification mode (https://doi.org/10.1016/j.celrep.2019.10.106), final output is imported: -->
<!-- ## Input peaks for diffTF were selected in separate script -->
<!-- difftf_activity <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/diffTF/output_rna_top1000_inverse/FINAL_OUTPUT/extension100/HepG2vsK562.top1000.summary.tsv") %>% -->
<!--   dplyr::select(TF, 'diffTF_activity' = weighted_meanDifference, 'classification' = classification_q0.001_final) -->
<!-- difftf_activity$classification[difftf_activity$TF == "POU5F1::SOX2"] <- difftf_activity$classification[difftf_activity$TF == "POU5F1"] -->
<!-- difftf_activity$classification[difftf_activity$TF == "RAR:RXR"] <- difftf_activity$classification[difftf_activity$TF == "RARA"] -->

<!-- # Import mean reporter activities per TF -->
<!-- reporter_activity_all <- tf_activities -->
<!-- reporter_activity_all_mean <- reporter_activity_all %>% -->
<!--   unique() %>% -->
<!--   pivot_wider(names_from = "condition", values_from = "reporter_activity_minP") %>% -->
<!--   mutate(dif_reporter = log2(K562 / HepG2)) %>% -->
<!--   dplyr::select(dif_reporter, 'TF' = tf) %>% -->
<!--   unique() -->
<!-- reporter_activity_all_mean$TF <- gsub("::.*", "_", reporter_activity_all_mean$TF) -->
<!-- reporter_activity_all_mean$TF[reporter_activity_all_mean$TF == "POU5F1_"] <- "POU5F1::SOX2" -->
<!-- reporter_activity_all_mean$TF[reporter_activity_all_mean$TF == "RARA:RXRA"] <- "RAR:RXR" -->
<!-- reporter_activity_all_mean$TF[reporter_activity_all_mean$TF == "ETS1"] <- "ETS2" -->
<!-- reporter_activity_all_mean$TF[reporter_activity_all_mean$TF == "NFATc1"] <- "NFATC1" -->
<!-- reporter_activity_all_mean$TF[reporter_activity_all_mean$TF == "TCF3"] <- "TCF7L2" -->
<!-- reporter_activity_all_mean$TF <- gsub("_", "", reporter_activity_all_mean$TF) -->


<!-- activity_comp <- merge(difftf_activity, reporter_activity_all_mean, by = "TF", all = T) %>% -->
<!--   na.omit() -->

<!-- activity_comp <- activity_comp %>% -->
<!--   mutate(diffTF_activity_norm = (diffTF_activity-mean(diffTF_activity))/sd(diffTF_activity), -->
<!--          reporter_activity_norm = (dif_reporter-mean(dif_reporter))/sd(dif_reporter)) -->

<!-- # Import class of TF -->
<!-- tf_class <- read_csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/garcia_tf-properties.csv") %>% -->
<!--   dplyr::select(TF, chromatin_regulation_mode) -->

<!-- activity_comp <- merge(activity_comp, tf_class, all = T) -->
<!-- activity_comp$chromatin_regulation_mode[activity_comp$TF == "POU5F1::SOX2"] <- activity_comp$chromatin_regulation_mode[activity_comp$TF == "POU5F1"] -->
<!-- activity_comp$chromatin_regulation_mode[activity_comp$TF == "RAR:RXR"] <- activity_comp$chromatin_regulation_mode[activity_comp$TF == "RARA"] -->
<!-- activity_comp <- activity_comp %>% na.omit() -->

<!-- ggplot(activity_comp, aes(x = reporter_activity_norm, y = diffTF_activity_norm)) + -->
<!--   geom_point(aes( color = chromatin_regulation_mode)) + -->
<!--   geom_abline(linetype = "dashed") + -->
<!--   geom_smooth(method = "lm", color = "grey", fill = "lightgrey") + -->
<!--   xlab("Differential Reporter Activity (K562/HepG2)") + -->
<!--   ylab("Differential Accessibility - diffTF (K562/HepG2)") + -->
<!--   scale_color_manual(values = c("grey",colors_diverse[c(2,5,4)])) + -->
<!--   facet_wrap(~classification) + -->
<!--   ggtitle('TF reporter activity vs. diffTF activity') -->
<!-- ``` -->

<!-- --- -->

<!-- ### 2B.3: Regulon expression -->
<!-- Correlate each reporter activity with the corresponding regulon RNA expression of the corresponding TF across condition - label significant correlations as confident. -->

<!-- ```{r warning = F} -->

<!-- ``` -->

<!-- Conclusion: -->

<!-- --- -->

<!-- ### 2B.4: Cistrome -->
<!-- Correlate TF activities  -->

<!-- ```{r warning = F} -->

<!-- ``` -->

<!-- Conclusion: -->

<!-- --- -->

<!-- ### 2B.5: Other MPRA-based TF activities (Ernst et al) -->
<!-- Correlate reporter activities to MPRA-infered TF activities - Ernst et al. uses K562 & HepG2 cells. Which reporters are approximating their imputed TF activities? Correlations might be hard because we only have 2 cell types here.  -->

<!-- ```{r mpra_comparison, out.width= "80%", fig.align= "center"} -->
<!-- # Extract TF activities for K562 & HepG2 -->
<!-- tf_activities <- cDNA_df %>% -->
<!--   filter(condition %in% c("K562_DMSO", "HepG2_DMSO"), neg_ctrls == "No") %>% -->
<!--   dplyr::select(reporter_id, reporter_activity_minP, condition, tf) %>% -->
<!--   unique() %>% -->
<!--   mutate(tf = gsub("_.*", "", tf)) %>% -->
<!--   spread(condition, reporter_activity_minP) %>% -->
<!--   mutate(dif = log2(K562_DMSO / HepG2_DMSO)) %>% -->
<!--   dplyr::select(tf, dif, reporter_id) -->

<!-- #write_csv2(tf_activities, "/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/diffTF/tf_reporter_activities_all.csv") -->

<!-- # Correlate with RNA-seq -->
<!-- load('/DATA/usr/m.martinez.ara/GitLab/gurten/gurten/data/fc181121_epsure_tadec_rnaseq_expr_joshi.RData') -->

<!-- # remove Serum data and calculate mean expression -->
<!-- tib_fpkm_joshi$twoi <- (tib_fpkm_joshi$twoi_1 + tib_fpkm_joshi$twoi_2)/2 -->
<!-- tib_fpkm_joshi <- tib_fpkm_joshi %>% -->
<!--   dplyr::select(gene_id, 'tf' = symbol, twoi) -->

<!-- tf_activities2 <- cDNA_df %>% -->
<!--   filter(condition == "mES_2i_LIF", neg_ctrls == "No") %>% -->
<!--   dplyr::select(reporter_id, reporter_activity_minP, tf) %>% -->
<!--   unique() %>% -->
<!--   mutate(tf = gsub("_.*", "", tf), -->
<!--          tf_activity = ave(reporter_activity_minP, tf, FUN = function(x) mean(x))) %>% -->
<!--   dplyr::select(tf_activity, tf) %>% -->
<!--   unique() -->

<!-- tf_activities2 <- merge(tf_activities2, tib_fpkm_joshi, by = "tf") %>% -->
<!--   mutate(tf_activity = (tf_activity-mean(tf_activity))/sd(tf_activity), -->
<!--          twoi = (twoi-mean(twoi))/sd(twoi)) -->

<!-- ggplot_custom(tf_activities2, aes(x = twoi, y = tf_activity)) + -->
<!--   geom_point() + -->
<!--   ggtitle('TF reporter activities vs. TF expression in mESCs') -->

<!-- # Correlate with regulon data (VIPER, Aerts lab papers) -->


<!-- # Correlate with TF activities from Ernst et al. (https://doi.org/10.1038/nbt.3678) -->
<!-- ernst_activities <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/ernst_tf_activities_hepg2_k562.csv", sep = ";", dec = ",", header = T, fill = T) -->

<!-- ernst_activities <- ernst_activities %>% -->
<!--   dplyr::select('tf' = TF, activity_enrichment_HepG2, activity_enrichment_K562) %>% -->
<!--   mutate(avg_score_HepG2 = ave(activity_enrichment_HepG2, tf, FUN = function(x) mean(x)), -->
<!--          avg_score_K562 = ave(activity_enrichment_K562, tf, FUN = function(x) mean(x))) %>% -->
<!--   unique() %>% -->
<!--   pivot_longer(-tf, names_to = "condition", values_to = "ernst_activity") %>% -->
<!--   mutate(condition = gsub("avg_score_", "", condition)) -->

<!-- tf_activities <- cDNA_df %>% -->
<!--   filter(condition %in% c("K562_DMSO", "HepG2_DMSO"), neg_ctrls == "No") %>% -->
<!--   dplyr::select(reporter_activity_minP, condition, tf) %>% -->
<!--   mutate(tf = gsub("_.*", "", tf), -->
<!--          condition = gsub("_DMSO", "", condition), -->
<!--          reporter_activity_minP = ave(reporter_activity_minP, condition, tf, FUN = function(x) mean(x))) %>% -->
<!--   unique() -->

<!-- ernst_activities <- merge(ernst_activities, tf_activities, by = c('tf', 'condition')) %>% -->
<!--   mutate(ernst_activity = (ernst_activity-mean(ernst_activity))/sd(ernst_activity), -->
<!--          reporter_activity_minP = (reporter_activity_minP-mean(reporter_activity_minP))/sd(reporter_activity_minP)) -->

<!-- ggplot_custom(ernst_activities,  -->
<!--               aes(x = ernst_activity, y = reporter_activity_minP)) + -->
<!--   geom_point() + -->
<!--   geom_abline(linetype = "dashed") + -->
<!--   geom_smooth(method = "lm", color = "grey", fill = "lightgrey") + -->
<!--   facet_wrap(~condition) + -->
<!--   ggtitle('TF reporter activities vs. MPRA-chunk-inferred TF activities') -->

<!-- #write_csv2(ernst_activities, "/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/diffTF/tf_reporter_activities_ernst.csv") -->
<!-- ``` -->

<!-- Conclusion: -->

<!-- --- -->








---



```{r log-linear-model-6, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# nt_df <- cDNA_df %>%
#   dplyr::filter(condition == "NT") %>%
#   dplyr::select(reporter_id, "nt_reporter_activity_minP" = reporter_activity_minP) %>%
#   mutate(nt_reporter_activity_minP = ave(nt_reporter_activity_minP, reporter_id, FUN = mean)) %>%
#   distinct()
# 
# cDNA_df_TF <- merge(cDNA_df, nt_df, all = T) %>%
#   filter(hPGK == "No", commercial_reporter == "No", native_enhancer == "No", neg_ctrls == "No") %>%
#   filter(promoter != "Random") %>%
#   filter(condition != "NT") %>%
#   filter(!is.na(reporter_activity_minP)) %>%
#   mutate(tf_condition = paste(tf, condition, sep = "_")) %>%
#   distinct(nt_reporter_activity_minP, tf, condition, tf_condition, reporter_id, promoter, spacing, distance, background) %>%
#   mutate(background = as.character(background)) %>%
#   dplyr::filter(condition == "NT")
# 
# tf_kd_targets2 <- tf_kd_targets %>%
#   filter(commercial_reporter == "No") %>%
#   distinct(condition, tf, is_target)
# 
# cDNA_df_TF <- merge(cDNA_df_TF, tf_kd_targets2) %>%
#   mutate(is_target = ifelse(tf == condition, "direct", is_target))
# 
# ## Fit linear model with design features for actual data
# exp_variance <- data.frame("feature" = c("promoter", "spacing", "distance", "background", "Residuals"))
# weight <- data.frame("feature" = c("(Intercept)", "promoterminP", "promoterhBGm", "promotermCMV",
#                   "spacing10bps","spacing5bp", "distance10bp","distance21bp", "background1","background2", "background3"))
# 
# cDNA_df_TF$promoter <- factor(cDNA_df_TF$promoter, levels = c("minP", "mCMV", "hBGm"))
# cDNA_df_TF$row <- rownames(cDNA_df_TF)
# cDNA_df_TF3 <- data.frame()
# 
# for (i in unique(cDNA_df_TF$tf_condition[cDNA_df_TF$is_target %in% c("direct", "TRUE")])) {
#   if (nrow(cDNA_df_TF[cDNA_df_TF$tf_condition == i,]) > 20) {
#   print(i)
#   x <- lm(reporter_FC ~
#                        promoter +
#                        spacing +
#                        distance +
#                        background,
#                 cDNA_df_TF[cDNA_df_TF$tf_condition == i,])
#   
#     f <- summary(x)$fstatistic
#     p <- pf(f[1],f[2],f[3],lower.tail=F)
#     names(p) <- "pval"
# 
#   y <- data.frame(x$fitted.values) %>% 
#     rownames_to_column() %>%
#     cbind(p)
#   names(y) <- c("row", "activity_predicted", "pval")
#   cDNA_df_TF2 <- y %>%
#     left_join(cDNA_df_TF)
#   cDNA_df_TF3 <- rbind(cDNA_df_TF3, cDNA_df_TF2)
# 
#   w <- data.frame('weight' = x$coefficients, 'pval' = summary(x)$coefficients[,4], tf_condition = i) %>% rownames_to_column('feature')
#   weight <- merge(weight, w, all = T)
# 
#   x <- anova(x) %>% rownames_to_column("feature") %>%
#     setnames(c("Sum Sq", "Pr(>F)"), c("sum_sq", "pval")) %>%
#     distinct(feature, sum_sq, pval) %>%
#     mutate(sum = sum(sum_sq)) %>%
#     mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
#     mutate(tf_condition = i) %>%
#     distinct(tf_condition, rel_sum_sq, pval, feature)
# 
#   exp_variance <- merge(exp_variance, x, all = T)
#   }
# }
# 
# 
# 
# ## Fine-tune dfs
# exp_variance$group <- "no residual"
# exp_variance$group[grep("Residual", exp_variance$feature)] <- "residual"
# prediction <- cDNA_df_TF3 %>%
#   dplyr::select(-row) %>%
#   na.omit()
# weight <- weight %>%
#   filter(feature != "(Intercept)")
# categorie <- c("mutated", "condition", "promoter",
#                "spacing", "distance", "background")
# weight$features <- gsub(paste(categorie, collapse="|"), "",weight$feature)
# ml <- length(unique(weight$features))
# feature <- unique(weight$features)
# weight$condition <- gsub(paste(feature, collapse = "|"), "", weight$feature)
# weight <- weight %>% dplyr::select(-feature)
# 
# 
# cDNA_df_TF3 <- cDNA_df_TF3 %>%
#   mutate(pval_adj = p.adjust(pval, method = "fdr"))
# 
# ## Pval distribution
# ggplot_custom(cDNA_df_TF3 %>% 
#          distinct(tf_condition, pval_adj, is_target),
#        aes(x = -log10(pval_adj), fill = is_target)) +
#   geom_density(alpha = .5) +
#   geom_vline(xintercept = -log10(0.01), lty = 2)
# 
# ggplot(cDNA_df_TF3 %>% 
#          distinct(tf_condition, pval_adj, is_target),
#        aes(y = -log10(pval_adj), x = is_target)) +
#   geom_quasirandom() +
#   geom_boxplot(alpha = .7) +
#   geom_hline(yintercept = -log10(0.01), lty = 2)
# 
# ggplot(cDNA_df_TF3 %>%
#          distinct(tf_condition, pval_adj, is_target, reporter_FC) %>%
#          mutate(reporter_FC = ave(reporter_FC, tf_condition, FUN = mean)) %>%
#          distinct(),
#        aes(x = reporter_FC, y = -log10(pval_adj))) +
#   geom_point() +
#   facet_wrap(~is_target)+
#   geom_hline(yintercept = -log10(0.01), lty = 2)
# 
# # Model parameters
# exp_variance$rel_sum_sq[exp_variance$feature == "Residuals"] <- 100 - exp_variance$rel_sum_sq[exp_variance$feature == "Residuals"]
# exp_variance$feature <- revalue(exp_variance$feature, c("Residuals"="total"))
# 
# tf_kd_targets3 <- cDNA_df_TF3 %>%
#   distinct(tf_condition, is_target, pval_adj)
# 
# exp_variance <- exp_variance %>%
#   left_join(tf_kd_targets3)
# 
# ggplot(exp_variance,
#        aes(x = reorder(feature, -rel_sum_sq), y = rel_sum_sq, fill = group)) +
#   scale_fill_manual(values = c("#1B998B", "#2D3047")) +
#   geom_bar(stat = "identity") +
#   ylab("Variance explained (%)") + xlab("") +
#   labs(title = "Linear variance modelling",
#        subtitle = "Expression variance predicted by features") +
#   theme(text = element_text(size = 12)) +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12),
#         axis.text.y = element_text(size = 10)) +
#   facet_wrap(~tf_condition)
# 
# exp_variance2 <- exp_variance %>%
#   filter(is_target == T | is_target == "direct") %>%
#   distinct(tf_condition, feature, rel_sum_sq) %>%
#   spread(key = "feature", value = "rel_sum_sq") %>%
#   #filter(total >= 50) %>%
#   column_to_rownames("tf_condition")
# 
# target_annotation <- exp_variance %>%
#   distinct(tf_condition, is_target) %>%
#   column_to_rownames("tf_condition")
# 
# pheatmap(exp_variance2 %>%
#                   t(),
#          cellwidth = 10,
#          cellheight = 10,
#          border_color = F,
#          color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
#          angle_col = 90,
#          cluster_cols = F,
#          annotation_col = target_annotation,
#          cutree_rows = 2)
# 
# exp_variance2 <- exp_variance %>%
#   filter(pval_adj <= 0.01) %>%
#   distinct(tf_condition, feature, rel_sum_sq) %>%
#   spread(key = "feature", value = "rel_sum_sq") %>%
#   filter(total >= 80) %>%
#   column_to_rownames("tf_condition")
# 
# pheatmap(exp_variance2 %>%
#                   t(),
#          cellwidth = 10,
#          cellheight = 10,
#          border_color = F,
#          cluster_cols = F,
#          color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
#          angle_col = 90)
# 
# ## Investigate individual KDs
# cDNA_df_TF4 <- cDNA_df_TF3 %>%
#   filter(tf_condition %in% c("PAX6_PAX6", "PAX6_THRB")) %>%
#   distinct(tf_condition, reporter_FC, reporter_id) %>%
#   spread(key = "tf_condition", value = "reporter_FC")
# 
# ggplot(cDNA_df_TF4,
#        aes(x = PAX6_PAX6, y = PAX6_THRB)) +
#   geom_point() +
#   geom_abline(lty = 2) +
#   xlim(-2.5,1) +
#   ylim(-2.5,1)
# 
# cDNA_df_TF4 <- cDNA_df_TF3 %>%
#   filter(tf_condition %in% c("FOXA1_FOXA1", "FOXA1_FOXA2")) %>%
#   distinct(tf_condition, reporter_FC, reporter_id) %>%
#   spread(key = "tf_condition", value = "reporter_FC")
# 
# ggplot(cDNA_df_TF4,
#        aes(x = FOXA1_FOXA1, y = FOXA1_FOXA2)) +
#   geom_point() +
#   geom_abline(lty = 2) +
#   xlim(-1,1) +
#   ylim(-1,1)
# 
# cDNA_df_TF4 <- cDNA_df_TF3 %>%
#   filter(tf_condition %in% c("NFE2L2_NFE2L2", "NFE2L2_MAF::NFE2")) %>%
#   distinct(tf_condition, reporter_FC, reporter_id) %>%
#   spread(key = "tf_condition", value = "reporter_FC")
# 
# ggplot(cDNA_df_TF4,
#        aes(x = NFE2L2_NFE2L2, y = `NFE2L2_MAF::NFE2`)) +
#   geom_point() +
#   geom_abline(lty = 2) +
#   xlim(-2.5,1) +
#   ylim(-2.5,1)
# 
# cDNA_df_TF4 <- cDNA_df_TF3 %>%
#   filter(tf_condition %in% c("MAF::NFE2_MAF::NFE2", "MAF::NFE2_NFE2L2")) %>%
#   distinct(tf_condition, reporter_FC, reporter_id) %>%
#   spread(key = "tf_condition", value = "reporter_FC")
# 
# ggplot(cDNA_df_TF4,
#        aes(x = `MAF::NFE2_MAF::NFE2`, y = `MAF::NFE2_NFE2L2`)) +
#   geom_point() +
#   geom_abline(lty = 2) +
#   xlim(-4,1) +
#   ylim(-4,1)
# 
# cDNA_df_TF4 <- cDNA_df_TF3 %>%
#   filter(tf_condition %in% c("CREB1_ATF2", "CREB1_FOS::JUN")) %>%
#   distinct(tf_condition, reporter_FC, reporter_id) %>%
#   spread(key = "tf_condition", value = "reporter_FC")
# 
# ggplot(cDNA_df_TF4,
#        aes(x = CREB1_ATF2, y = `CREB1_FOS::JUN`)) +
#   geom_point() +
#   geom_abline(lty = 2) +
#   xlim(-2.5,1) +
#   ylim(-2.5,1)
# 
# 
# cDNA_df_TF4 <- cDNA_df_TF3 %>%
#   filter(tf_condition %in% c("NFKB1_NFKB1", "NFKB1_REL")) %>%
#   distinct(tf_condition, reporter_FC, reporter_id) %>%
#   spread(key = "tf_condition", value = "reporter_FC")
# 
# ggplot(cDNA_df_TF4,
#        aes(x = NFKB1_NFKB1, y = NFKB1_REL)) +
#   geom_point() +
#   geom_abline(lty = 2) +
#   xlim(-1.5,1) +
#   ylim(-1.5,1)
# 
# cDNA_df_TF4 <- cDNA_df_TF3 %>%
#   filter(tf_condition %in% c("NFKB1_NFKB1", "NFKB1_NFKB2")) %>%
#   distinct(tf_condition, reporter_FC, reporter_id) %>%
#   spread(key = "tf_condition", value = "reporter_FC")
# 
# ggplot(cDNA_df_TF4,
#        aes(x = NFKB1_NFKB1, y = NFKB1_NFKB2)) +
#   geom_point() +
#   geom_abline(lty = 2) +
#   xlim(-1.5,1) +
#   ylim(-1.5,1)
# 
# 
# cDNA_df_TF4 <- cDNA_df_TF3 %>%
#   filter(tf_condition %in% c("SP1_SP1", "SP1_KLF4")) %>%
#   distinct(tf_condition, reporter_FC, reporter_id) %>%
#   spread(key = "tf_condition", value = "reporter_FC")
# 
# ggplot(cDNA_df_TF4,
#        aes(x = SP1_SP1, y = SP1_KLF4)) +
#   geom_point() +
#   geom_abline(lty = 2) +
#   xlim(-1.5,1) +
#   ylim(-1.5,1)
# 
# cDNA_df_TF4 <- cDNA_df_TF3 %>%
#   filter(tf_condition %in% c("KLF4_KLF4", "KLF4_SP1")) %>%
#   distinct(tf_condition, reporter_FC, reporter_id) %>%
#   spread(key = "tf_condition", value = "reporter_FC")
# 
# ggplot(cDNA_df_TF4,
#        aes(x = KLF4_KLF4, y = KLF4_SP1)) +
#   geom_point() +
#   geom_abline(lty = 2) +
#   xlim(-1.5,1) +
#   ylim(-1.5,1)
# 
# cDNA_df_TF4 <- cDNA_df_TF3 %>%
#   filter(tf_condition %in% c("CEBPB_CEBPB", "CEBPB_CEBPG")) %>%
#   distinct(tf_condition, reporter_FC, reporter_id) %>%
#   spread(key = "tf_condition", value = "reporter_FC")
# 
# ggplot(cDNA_df_TF4,
#        aes(x = CEBPB_CEBPB, y = CEBPB_CEBPG)) +
#   geom_point() +
#   geom_abline(lty = 2) +
#   xlim(-0.5,3) +
#   ylim(-0.5,3)
# 
# 
# ## Plot weights of parameters in model - this helps to understand the model
# level_order <- c("minP", "mCMV", "hBGm",
#                  "2", "3", "10bps", "5bp", "10bp", "21bp")
# 
# ggplot(weight %>%
#          filter(tf_condition %in% c("PAX6_PAX6", "PAX6_THRB")) %>%
#          na.omit() %>%
#                  mutate(sign = ifelse(pval < 0.05, "Yes", "No")),
#        aes(x = factor(features, level = level_order), y = weight, alpha = sign, fill = condition)) +
#   scale_fill_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B", "#C2B2B4", "#A54657")) +
#   geom_bar(stat = "identity") +
#   ylab("Weight") + xlab("") +
#   theme(text = element_text(size = 12)) +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10),
#         axis.text.y = element_text(size = 10)) +
#   facet_wrap(~tf_condition)
# 
# 
# 
# ## Final overview - model fit for each TF
# # Feature importance overview for each TF in a single plot (bar-chart including mean activities per feature)
# 
# 
# for (i in unique(cDNA_df_TF$tf_condition[cDNA_df_TF$is_target == T])) {
# 
#   # Plot to show prediction accuracy
#   sp <- ggplot(prediction %>%
#                  filter(tf_condition == i),
#                aes(x = reporter_FC, y = activity_predicted)) +
#     geom_smooth(color = "black", fill = "lightgrey", method = "lm") +
#     geom_point() +
#     ylab("Predicted KD FC (log2)") +
#     xlab("Average KD FC (log2)") +
#     scale_color_manual(values = c(colors_diverse[1],colors_diverse[2],colors_diverse[4])) +
#     ggtitle(i)
# 
# 
#   # Feature importance overview for each TF (including mean activities per feature)
#   fp <- ggplot(exp_variance[exp_variance$tf_condition == i,] %>%
#                                  na.omit() %>%
#                  mutate(sign = ifelse(pval < 0.05, "Yes", "No")),
#        aes(x = reorder(feature, -rel_sum_sq), y = rel_sum_sq, fill = group, alpha = sign)) +
#   scale_fill_manual(values = c("#1B998B", "#2D3047")) +
#   geom_bar(stat = "identity") +
#   ylab("Variance explained (%)") + xlab("") +
# 
#   theme(text = element_text(size = 12)) +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12),
#         axis.text.y = element_text(size = 10)) +
#     guides(fill = F)
# 
#   # Plot weights of parameters in model - this helps to understand the model
#   p <- ggplot(weight[weight$tf_condition == i,] %>%
#                 na.omit() %>%
#                  mutate(sign = ifelse(pval < 0.05, "Yes", "No")),
#        aes(x = factor(features, level = level_order), y = weight, alpha = sign, fill = condition)) +
#   scale_fill_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B", "#C2B2B4", "#A54657")) +
#   geom_bar(stat = "identity") +
#   ylab("Weight") + xlab("") +
# 
#   theme(text = element_text(size = 12)) +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10),
#         axis.text.y = element_text(size = 10))
# 
#   # Combine all plots
#   sp.plot <- arrangeGrob(sp,fp,p, nrow = 1)
#   grid.arrange(sp.plot)
# }
# 
# # Plot features in a heatmap
# exp_variance_heatmap <- exp_variance %>%
#   mutate(rel_sum_sq = ifelse(pval < 0.05 | is.na(pval), rel_sum_sq, 0)) %>%
#   mutate(min_pval = ave(pval, tf_condition, FUN = function(x) min(x, na.rm = T))) %>%
#   filter(min_pval < 0.05) %>%
#   dplyr::select(-group, -pval) %>%
#   spread(feature, rel_sum_sq) %>%
#   filter(total >= 75) %>%
#   mutate(tf = gsub("(^.*)_.*_.*", "\\1", tf_condition)) %>%
#   mutate(tf = gsub("(^.*)_.*", "\\1", tf)) %>%
#   mutate(condition = gsub("^.*_(.*)+", "\\1", tf_condition))
# 
# for (i in unique(exp_variance_heatmap$condition)) {
#   p <- pheatmap(exp_variance_heatmap %>%
#                   filter(condition == i) %>%
#                   column_to_rownames("tf") %>%
#                   dplyr::select(-condition, -tf_condition, -total) %>%
#                   t(),
#          cellwidth = 8,
#          cellheight = 8,
#          border_color = F,
#          color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
#          angle_col = 90,
#          main = i)
#   print(p)
# }
# 
# ## Correlation between cell types - is the variability the same between individual cell types?
# ### Implement still: select only TFs that are active in both conditions
# exp_variance_heatmap2 <- exp_variance %>%
#   filter(tf_condition %in% exp_variance_heatmap$tf_condition) %>%
#   filter(pval < 0.05) %>%
#   dplyr::select(-group, -pval) %>%
#   mutate(tf = gsub("(^.*)_.*_.*", "\\1", tf_condition)) %>%
#   mutate(tf = gsub("(^.*)_.*", "\\1", tf)) %>%
#   mutate(condition = gsub("^.*_(.*)+", "\\1", tf_condition)) %>%
#   dplyr::select(-tf_condition) %>%
#   spread(key = "condition", value = "rel_sum_sq") %>%
#   filter(feature != "total")
# 
# ggplot(exp_variance_heatmap2,
#        aes(x = HepG2_DMSO, y = K562_DMSO)) +
#   geom_point() +
#   geom_smooth(method = "lm") +
#   xlab("feature importance HepG2") +
#   ylab("feature importance K562") +
#   facet_wrap(~feature)
# 
# ggplot(exp_variance_heatmap2,
#        aes(x = mES_2i_LIF, y = K562_DMSO)) +
#   geom_point() +
#   geom_smooth(method = "lm") +
#   xlab("feature importance mES") +
#   ylab("feature importance K562")
# 
# 
# exp_variance_heatmap <- exp_variance_heatmap %>%
#   filter(str_detect(condition, "mES")) %>%
#   group_by(tf) %>%
#   top_n(1, total) %>%
#   ungroup()
# 
# conditions <- exp_variance_heatmap %>%
#   distinct(tf, condition) %>%
#   column_to_rownames("tf")
# 
# condition_colors <- list(condition = c("mES_N2B27" = "black",
#                       "mES_LIF_CH" = "grey10",
#                       "mES_LIF_PD" = "grey30",
#                       "mES_2i" = "grey50",
#                       "mES_2i_LIF" = "grey70",
#                       "mES_HQ" = "grey90",
#                       "K562_DMSO" = "#e76f51",
#                       "HepG2_CDCA" = "#9AC1AE",
#                       "HepG2_DMSO" = "#f2cc8f"))
# 
# pheatmap(exp_variance_heatmap %>%
#                   column_to_rownames("tf") %>%
#                   dplyr::select(-condition, -tf_condition, -total) %>%
#                   t(),
#          cellwidth = 8,
#          cellheight = 8,
#          border_color = F,
#          color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
#          angle_col = 90,
#          annotation_col = conditions,
#          annotation_colors = condition_colors)
# 
# # Plot weights in a heatmap
# weight_heatmap <- weight %>%
#   distinct(tf_condition, weight, features) %>%
#   mutate(total = ave(weight, tf_condition, FUN = function(x) sum(x, na.rm = T))) %>%
#   filter(!feature %in% c("hBGm", "10bp", "10bps", "1")) %>%
#   spread(features, weight) %>%
#   #filter(total >= 60) %>%
#   mutate(tf = gsub("(^.*)_.*_.*", "\\1", tf_condition)) %>%
#   mutate(tf = gsub("(^.*)_.*", "\\1", tf)) %>%
#   mutate(condition = gsub("^.*_(.*)+", "\\1", tf_condition))
# 
# weight_heatmap[is.na(weight_heatmap)] <- 0
# 
# for (i in unique(weight_heatmap$condition)) {
#   p <- pheatmap(weight_heatmap %>%
#                   filter(condition == i) %>%
#                   column_to_rownames("tf") %>%
#                   dplyr::select(-condition, -tf_condition, -total) %>%
#                   t(),
#          cellwidth = 8,
#          cellheight = 8,
#          border_color = F,
#          color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
#          angle_col = 90,
#          main = i)
#   print(p)
# }
# 
# 
# weight_heatmap2 <- weight %>%
#   filter(tf_condition %in% weight_heatmap$tf_condition) %>%
#   filter(pval < 0.05) %>%
#   dplyr::select(-pval) %>%
#   mutate(tf = gsub("(^.*)_.*_.*", "\\1", tf_condition)) %>%
#   mutate(tf = gsub("(^.*)_.*", "\\1", tf)) %>%
#   mutate(condition = gsub("^.*_(.*)+", "\\1", tf_condition)) %>%
#   dplyr::select(-tf_condition) %>%
#   spread(key = "condition", value = "weight")
# 
# ggplot(weight_heatmap2,
#        aes(x = HepG2_DMSO, y = K562_DMSO)) +
#   geom_point() +
#   geom_smooth(method = "lm") +
#   xlab("feature importance HepG2") +
#   ylab("feature importance K562") +
#   facet_wrap(~features)
# 
# for (i in unique(weight_heatmap2$features)) {
#   pc_data <- weight_heatmap2 %>% filter(features == i) %>% dplyr::select(-features) %>% replace(is.na(.), 0)
#   pca_res <- prcomp(pc_data %>% column_to_rownames('tf'), scale. = TRUE)
#   p <- ggbiplot(pca_res, ellipse = T) + ggtitle(i)
#   print(p)
# }
# 
# 
# 
# weight_heatmap <- weight_heatmap %>%
#   group_by(tf) %>%
#   top_n(1, total) %>%
#   ungroup()
# 
# conditions <- weight_heatmap %>%
#   distinct(tf, condition) %>%
#   column_to_rownames("tf")
# 
# condition_colors <- list(condition = c("mES_N2B27" = "black",
#                       "mES_LIF_CH" = "grey10",
#                       "mES_LIF_PD" = "grey30",
#                       "mES_2i" = "grey50",
#                       "mES_2i_LIF" = "grey70",
#                       "mES_HQ" = "grey90",
#                       "K562_DMSO" = "#e76f51",
#                       "HepG2_CDCA" = "#9AC1AE",
#                       "HepG2_DMSO" = "#f2cc8f"))
# 
# pheatmap(weight_heatmap %>%
#                   column_to_rownames("tf") %>%
#                   dplyr::select(-condition, -tf_condition, -total) %>%
#                   t(),
#          cellwidth = 8,
#          cellheight = 8,
#          border_color = F,
#          #color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
#          angle_col = 90,
#          annotation_col = conditions,
#          annotation_colors = condition_colors)




```



<!-- ## Native enhancer sequences + NPCs -->

<!-- ```{r} -->
<!-- ## NPC data -->
<!-- es_npc <- cDNA_df %>% -->
<!--   filter(condition %in% c("NPC", "mES_2i_LIF"), commercial_reporter == "No", native_enhancer == "No", hPGK == "No", neg_ctrls == "No") %>% -->
<!--   distinct(condition, reporter_activity_minP, tf) %>% -->
<!--   mutate(reporter_activity_minP = log2(ave(reporter_activity_minP, tf, condition, FUN = mean))) %>% -->
<!--   distinct() -->

<!-- pheatmap(es_npc %>% -->
<!--            spread(key = "condition", value = "reporter_activity_minP") %>% -->
<!--            column_to_rownames("tf") %>% -->
<!--            t(), -->
<!--          cellwidth = 8, -->
<!--          cellheight = 8, -->
<!--          border = F) -->

<!-- ## native enhancers -->
<!-- es_npc <- cDNA_df %>% -->
<!--   filter(commercial_reporter == "No", native_enhancer == "Yes", hPGK == "No", neg_ctrls == "No") %>% -->
<!--   #filter(str_detect(tf, "minP")) %>% -->
<!--   distinct(condition, reporter_activity_minP, tf, promoter, pval_adj_minP) %>% -->
<!--   mutate(reporter_activity_minP = log2(ave(reporter_activity_minP, tf, condition, FUN = mean))) %>% -->
<!--   mutate(pval_adj_minP = -log10(pval_adj_minP)) %>% -->
<!--   distinct() %>% -->
<!--   mutate(chunk = gsub("(.*)_.*_.*", "\\1", tf)) %>% -->
<!--   mutate(enhancer = gsub("(e[0-9]{1,2})_.*", "\\1", chunk)) %>% -->
<!--   distinct(chunk, enhancer, reporter_activity_minP, condition, promoter, pval_adj_minP) %>% -->
<!--   filter(condition %in% c("mES_2i_LIF", "mES_N2B27_RA", "mES_N2B27", "mES_2i", "mES_LIF_PD", "mES_LIF_CH", "NPC")) -->


<!-- ggplot(mapping = aes(x = reporter_activity_minP, y = pval_adj_minP)) + -->
<!--     geom_hline(yintercept = -log10(0.05), linetype = "twodash") + -->
<!--     geom_vline(xintercept = log2(2), linetype = "twodash") + -->
<!--     geom_vline(xintercept = -log2(2), linetype = "twodash") + -->
<!--     geom_point(data = es_npc %>%  -->
<!--                  filter(condition == "mES_2i_LIF") %>%  -->
<!--                  filter(promoter == "mCMV") %>% -->
<!--                  filter(abs(reporter_activity_minP) <= 1 | pval_adj_minP <= -log10(0.05)),  -->
<!--                size = 2, color = "grey90") + -->
<!--     geom_point(data = es_npc %>%  -->
<!--                  filter(condition == "mES_2i_LIF") %>%  -->
<!--                  filter(promoter == "mCMV") %>% -->
<!--                  filter(abs(reporter_activity_minP) > 1 & pval_adj_minP > -log10(0.05) ),  -->
<!--                size = 2, color = "grey40") + -->
<!--     theme_pubr(border = F) + -->
<!--   facet_wrap(~enhancer) -->

<!-- for (i in unique(es_npc$promoter)) { -->
<!--   p <- pheatmap(es_npc %>% -->
<!--              filter(promoter == i) %>% -->
<!--            dplyr::select(-pval_adj_minP, -promoter, -enhancer) %>% -->
<!--              distinct() %>% -->
<!--            spread(key = "condition", value = "reporter_activity_minP") %>% -->
<!--            column_to_rownames("chunk") %>% -->
<!--            t(), -->
<!--          border_color = NA, -->
<!--          cellwidth = 8, -->
<!--          cellheight = 8, -->
<!--          cluster_cols = F, -->
<!--          main = i) -->
<!--   print(p) -->
<!-- } -->
<!-- ``` -->


<!-- ```{r build tf motif matrices db2: random_promoter, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE} -->
<!-- # load motif Metadata --> PWM feature matrix -->
<!-- tib_pwm_native_enhancer  <- get_pwm_feature_matrix(motif_meta_fn = '/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/fc181127_curated_metadata_no_composite_filt.csv', -->
<!--                                         fimo_fn       = '/home/m.trauernicht/mydata/projects/SuRE-TF/data/gcf5927_stimulation-1/results/native-enhancer/fimo/fimo.tsv', -->
<!--                                         db            = 2) -->
<!-- ``` -->


<!-- ```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE} -->
<!-- ## Add groups so I can plot them individually -->
<!-- tib_pwm_native_enhancer$group <- gsub("(^e[0-9]{1,2}).*", "\\1", tib_pwm_native_enhancer$id) -->
<!-- controls <- c("minP", "mCMV", "hBGm") -->
<!-- tib_pwm_native_enhancer$group <- gsub(paste(controls, collapse = "|"), "Control", tib_pwm_native_enhancer$group) -->

<!-- load('/DATA/usr/m.martinez.ara/GitLab/gurten/gurten/data/fc181121_epsure_tadec_rnaseq_expr_joshi.RData') -->

<!-- # Take mean of 2i expression -->
<!-- tib_fpkm_joshi <- tib_fpkm_joshi %>% -->
<!--   dplyr::select(-contains('serum')) %>% -->
<!--   rowwise() %>% -->
<!--   mutate(expr_mean = mean(c(twoi_1, twoi_2))) %>% -->
<!--   ungroup() -->

<!-- # only select fpkm>0 -->
<!-- tib_fpkm_joshi_expressed <- tib_fpkm_joshi %>% filter(expr_mean>10) -->

<!-- # Change to mouse name of Tp53 -->
<!-- tib_fpkm_joshi_expressed$symbol <- gsub("TRP53", "TP53", tib_fpkm_joshi_expressed$symbol) -->

<!-- tib_pwm_native_enhancer_ES <- tib_pwm_native_enhancer[,colnames(tib_pwm_native_enhancer) %in% tib_fpkm_joshi_expressed$symbol | colnames(tib_pwm_native_enhancer) %in% c("id", "group")] -->
<!-- ``` -->


<!-- ```{r, fig.height=10, fig.width=10, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE} -->
<!-- # remove columns with only 0s and plot heatmap for each native enhancer individually -->
<!-- for (i in unique(tib_pwm_native_enhancer_ES$group)) { -->
<!--   data <- tib_pwm_native_enhancer_ES[tib_pwm_native_enhancer_ES$group == i,] %>% -->
<!--     column_to_rownames("id") %>% dplyr::select(-group) -->
<!--   p <- pheatmap(as.matrix(data[, colSums(data != 0) > 0]), -->
<!--                 color = colorRampPalette(brewer.pal(n = 7, name = "Oranges"))(100), -->
<!--                 border_color = "black", -->
<!--                 cellwidth = 8, cellheight = 10) -->
<!--   print(p) -->
<!-- } -->

<!-- ## Zic2/Hic2 bind to mCMV/hBGm -> Does this explain increased cooperativity? -->

<!-- # make list of relevant TFs -->
<!-- relevant_TFs <- c("KLF", "SP1", "SP3", "POU", "RAR", "SOX", "STAT", "^TCF", "TFCP", "TP53") -->
<!-- relevant_TFs_2 <- c("KLF4", "POU", "SOX", "STAT", "^TCF", "TFCP") -->
<!-- relevant_TFs_3 <- "TFCP" -->

<!-- tib_pwm_native_enhancer_ES <- tib_pwm_native_enhancer_ES[-grep(paste(promoters$TF, collapse = "|"), tib_pwm_native_enhancer_ES$id),] -->

<!-- data <- tib_pwm_native_enhancer_ES %>% -->
<!--   dplyr::select(KLF4, POU5F1, SOX2, TCF7, TFCP2L1, CTCF, ZIC2,id, group) %>% -->
<!--   column_to_rownames("id") %>% dplyr::select(-group) -->

<!-- pheatmap(as.matrix(t(data[, colSums(data != 0) > 0])), -->
<!--                 color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100), -->
<!--                 border_color = "black", cluster_cols = F, -->
<!--                 cellwidth = 10, cellheight = 10) -->

<!-- ``` -->


```{r fimo_model, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# # Subselect interesting TFs
# tib_pwm_native_enhancer_ES <- tib_pwm_native_enhancer_ES %>%
#   dplyr::select(KLF4, POU5F1, SOX2, TCF7, TFCP2L1, CTCF, ZIC2,id, group)
# 
# # Transform to binary
# tib_pwm_native_enhancer_ES <- tib_pwm_native_enhancer_ES %>% mutate_if(is.numeric, function(x) ifelse((x==0 | is.na(x)),"No","Yes"))
# 
# # Combine with expression data
# native_expression <- tf_activity_heatmap %>% rownames_to_column("condition") %>%
#   melt(id.vars = "condition", variable.name = "id", value.name = "expression")
# 
# tib_pwm_native_enhancer_ES_merge <- merge(tib_pwm_native_enhancer_ES, native_expression, all = T) %>%
#   dplyr::select(-group)
# 
# 
# 
# ### Calculate expression variance
# exp_variance <- data.frame(c("KLF4", "POU5F1", 
#                              "SOX2", "TCF7", "TFCP2L1", "CTCF", "ZIC2", "Residuals"))
# prediction <- data.frame(1:280)
# names(prediction) <- "id"
# names(exp_variance) <- "feature"
# weight <- data.frame(c("(Intercept)", "KLF4Yes", "POU5F1Yes", 
#                              "SOX2Yes", "TCF7Yes", "TFCP2L1Yes", "CTCFYes", "ZIC2Yes"))
# names(weight) <- "feature"
# 
# 
# tib_pwm_native_enhancer_ES_merge$row <- rownames(tib_pwm_native_enhancer_ES_merge)
# 
# 
# 
# for (i in unique(tib_pwm_native_enhancer_ES_merge$condition)) {
#   x <- lm(expression ~ KLF4 + POU5F1 + SOX2 + TCF7 + TFCP2L1 + CTCF + ZIC2, 
#         tib_pwm_native_enhancer_ES_merge[tib_pwm_native_enhancer_ES_merge$condition == i,])
#   y <- data.frame(x$fitted.values) %>% rownames_to_column()
#   par(mfrow=c(2,2))
#   plot(x, main = i)
#   names(y) <- c("row", i)
#   y$id <- 1:nrow(y)
#   prediction <- merge(prediction,y, all = T)
#   
#  
#   w <- data.frame(x$coefficients) %>% rownames_to_column()
#   names(w) <- c("feature", i)
#   weight <- merge(weight,w, all = T)
#   
#   x <- anova(x) %>% rownames_to_column("feature") %>%
#     setnames("Sum Sq", "sum_sq") %>%
#     dplyr::select(feature, sum_sq) %>%
#     mutate(sum = sum(sum_sq)) %>%
#     mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
#     setnames("rel_sum_sq", i)
#   x <- x[,c(1,4)]
# 
#   exp_variance <- merge(exp_variance, x)
# }
# 
# 
# 
# ## Prepare lm parameters for plotting
# exp_variance <- melt(exp_variance, variable.name = "id")
# exp_variance$group <- "no residual"
# exp_variance$group[grep("Residual", exp_variance$feature)] <- "residual"
# prediction <- prediction %>% dplyr::select(-id) %>% melt(variable.name = "id") %>% na.omit() %>%
#   dplyr::select(-id) %>% setnames("value", "activity_predicted")
# tib_pwm_native_enhancer_ES_merge <- merge(tib_pwm_native_enhancer_ES_merge, prediction)
# weight <- weight %>% filter(feature != "(Intercept)")
# weight <- melt(weight, variable.name = "id")
# categorie <- c("KLF4", "POU5F1", "SOX2", "TCF7", "TFCP2L1", "CTCF", "ZIC2")
# #weight$features <- gsub(paste(categorie, collapse="|"), "",weight$feature)
# #feature <- weight$features
# weight$condition <- gsub(paste(feature, collapse="|"), "", weight$feature)
# weight$value[is.na(weight$value)] <- 0
# 
# 
# # Further plotting preparations
# level_order <- c("KLF4Yes", "POU5F1Yes", "SOX2Yes", "CTCFYes", "ZIC2Yes", "TFCP2L1Yes", "TCF7Yes")
# 
# 
# # Model parameters
# exp_variance$value[exp_variance$feature == "Residuals"] <- 100 - exp_variance$value[exp_variance$feature == "Residuals"]
# exp_variance$feature <- revalue(exp_variance$feature, c("Residuals"="total"))
# 
# 
# 
# ggplot(exp_variance, 
#        aes(x = reorder(feature, -value), y = value, fill = group)) +
#   scale_fill_manual(values = c("#1B998B", "#2D3047")) +
#   geom_bar(stat = "identity") + 
#   ylab("Variance explained (%)") + xlab("") + 
#   labs(title = "Linear variance modelling", 
#        subtitle = "Expression variance predicted by features") +
#   theme_bw() +
#   theme(text = element_text(size = 12)) +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12), 
#         axis.text.y = element_text(size = 10)) +
#     guides(fill = F) + facet_wrap(~id)
# 
# ggscatter(tib_pwm_native_enhancer_ES_merge, x = "expression", y = "activity_predicted",
#           add = "reg.line", 
#           add.params = list(color = "blue", fill = "lightgray"),
#           conf.int = TRUE, ylab = "Predicted log2-activity", size = 0.5,
#           xlab = "Average log2-activity") + theme_bw() + facet_wrap(~condition) +
#   stat_cor(aes(label = ..r.label..), method = "pearson", label.x = -2, label.y = 4)
# 
# 
# 
# for (i in unique(tib_pwm_native_enhancer_ES_merge$condition)) {
#   
#   # Plot to show prediction accuracy
#   sp <- ggscatter(tib_pwm_native_enhancer_ES_merge[tib_pwm_native_enhancer_ES_merge$condition == i,], x = "expression", y = "activity_predicted",
#    add = "reg.line", 
#    add.params = list(color = "blue", fill = "lightgray"),
#    conf.int = TRUE, ylab = "Predicted log2-activity", 
#    title = i,
#    xlab = "Average log2-activity") + theme_bw()
#   
#   
#   # Feature importance overview for each TF (including mean activities per feature)
#   fp <- ggplot(exp_variance[exp_variance$id == i,], 
#        aes(x = reorder(feature, -value), y = value, fill = group)) +
#   scale_fill_manual(values = c("#1B998B", "#2D3047")) +
#   geom_bar(stat = "identity") + 
#   ylab("Variance explained (%)") + xlab("") +
#   ggtitle(i)+
#   theme_bw() +
#   theme(text = element_text(size = 12)) +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12), 
#         axis.text.y = element_text(size = 10)) +
#     guides(fill = F)
#   
#   # Plot weights of parameters in model - this helps to understand the model 
#   p <- ggplot(weight[weight$id == i,], 
#        aes(x = factor(feature, level = level_order), y = value, fill = condition)) +
#   scale_fill_manual(values = c("#DD6B48", "#1B998B", "#1B998B", "#1B998B", "#1B998B", "#1B998B", "1B998B")) +
#   ggtitle(i)+
#   geom_bar(stat = "identity") + 
#   ylab("Weight") + xlab("") +
#   theme_bw() +
#   theme(text = element_text(size = 12)) +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
#         axis.text.y = element_text(size = 10)) +
#     guides(fill = F) 
# 
#   # Combine all plots
#   sp.plot <- arrangeGrob(sp,fp,p, nrow = 2, ncol = 2)
#   grid.arrange(sp.plot)
# }
```




<!-- ## Random forest model to classify reporters (in an exploratory phase) -->

<!-- ```{r random_forest_classes, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE} -->
<!-- cDNA_df_lm <- cDNA_df %>% -->
<!--   filter(neg_ctrls == "No", commercial_reporter == "No", promoter != "Random", -->
<!--          condition %in% c("HepG2_CDCA", "HepG2_DMSO", "K562_DMSO", "mES_2i", -->
<!--                           "mES_2i_LIF", "mES_HQ", "mES_LIF_CH", "mES_LIF_PD", -->
<!--                           "mES_N2B27" -->
<!--                           ), -->
<!--          tf %in% names(cDNA_df4), -->
<!--          reporter_activity_class != -1) %>% -->
<!--   mutate(background = as.factor(background), -->
<!--          reporter_activity_class = as.factor(reporter_activity_class)) %>% -->
<!--   dplyr::select(reporter_id, tf, reporter_activity_class, condition, promoter, spacing, background, distance) %>% -->
<!--   unique() -->

<!-- mixed_classes <- cDNA_df3 %>% -->
<!--   mutate(id = paste(condition, tf, sep = "_")) %>% -->
<!--   dplyr::select(id, class) %>% -->
<!--   filter(class == 1) %>% -->
<!--   unique() -->

<!-- ## Fitting the model for each condition and tf -->
<!-- cDNA_df_lm$id <- paste(cDNA_df_lm$condition, cDNA_df_lm$tf, sep = "_") -->
<!-- cDNA_df_lm <- cDNA_df_lm %>% -->
<!--   mutate(sum = ave(id, id, FUN = function(x) length(x))) %>% -->
<!--   filter(sum >= 2, id %in% mixed_classes$id) %>% -->
<!--   dplyr::select(-sum) %>% -->
<!--   na.omit() -->

<!-- rf_features <- lapply(unique(cDNA_df_lm$id), function(x) randomForest(reporter_activity_class ~ promoter + spacing + background + distance, -->
<!--                                                                      cDNA_df_lm[cDNA_df_lm$id == x,])) -->

<!-- names(rf_features) <- unique(cDNA_df_lm$id) -->


<!-- # ROC -->
<!-- roc_rf_train <- list() -->
<!-- for (i in names(rf_features)) { -->
<!--   roc_rf_train[[i]] <- roc(response = cDNA_df_lm$reporter_activity_class[cDNA_df_lm$id == i], predictor = rf_features[[i]]$votes[,2]) -->
<!-- } -->

<!-- auc_rf <- data.frame("id" = names(roc_rf_train), -->
<!--                      "n" = 1:length(roc_rf_train), -->
<!--                      "auc" = "") -->
<!-- for (i in 1:length(roc_rf_train)) { -->
<!--   if (i == 1) { -->
<!--     plot <- plot(roc_rf_train[[i]], col="#264653", lwd=3, main="ROC curve RF") -->
<!--     auc_rf$auc[auc_rf$n == i] <- auc(plot) -->
<!--   } -->
<!--   else { -->
<!--     plot <- plot(roc_rf_train[[i]], add = TRUE, label = T) -->
<!--     auc_rf$auc[auc_rf$n == i] <- auc(plot) -->
<!--   } -->
<!-- } -->

<!-- auc_rf <- auc_rf %>% -->
<!--   mutate(condition = gsub("(.*)_.*$", "\\1", id), -->
<!--          tf = gsub(".*_.*_(.*$)", "\\1", id), -->
<!--          auc = as.numeric(auc)) -->

<!-- ggplot_custom(auc_rf, aes(x = tf, y = auc)) + -->
<!--   geom_point() + -->
<!--   geom_boxplot() + -->
<!--   geom_hline(yintercept = 0.75, linetype = "dashed") + -->
<!--   theme_classic_lines_90() -->

<!-- ggplot_custom(auc_rf, aes(x = condition, y = auc)) + -->
<!--   geom_point() + -->
<!--   geom_hline(yintercept = 0.75, linetype = "dashed", color = "red")+ -->
<!--   coord_flip() -->


<!-- auc_rf_sel <- auc_rf %>% -->
<!--   filter(auc >= 0.75) -->

<!-- rf_features <- rf_features[auc_rf_sel$id] -->


<!-- # Get the variable importances -->
<!-- explained_rf <- list() -->

<!-- for (i in names(rf_features)) { -->
<!--   explained_rf[[i]] <- explain(rf_features[[i]], data=cDNA_df_lm %>% dplyr::select(promoter, spacing, background, distance), y=cDNA_df_lm$reporter_activity_class) -->
<!-- } -->

<!-- varimps <- list() -->
<!-- for (i in names(explained_rf)) { -->
<!--   varimps[[i]] <- feature_importance(explained_rf[[i]], type = "ratio") -->
<!-- } -->
<!-- #, loss_function = '1-auc' type='ratio' -->

<!-- varimps <- bind_rows(varimps, .id = "id") -->

<!-- varimps <- varimps %>% -->
<!--   mutate(dropout_loss = ave(dropout_loss, id, variable, FUN = function(x) mean(x))) %>% -->
<!--   dplyr::select(id, 'feature' = variable, dropout_loss) %>% -->
<!--   mutate(condition = gsub("(.*)_.*$", "\\1", id), -->
<!--          tf = gsub(".*_.*_(.*$)", "\\1", id)) %>% -->
<!--   dplyr::select(-id) %>% -->
<!--   unique() -->

<!-- for (i in unique(varimps$condition)) { -->
<!--   p <- ggplot_custom(varimps %>% filter(condition == i), aes(y = dropout_loss, x = feature)) + -->
<!--     geom_point()+ -->
<!--     ylab("dropout loss (RMSE)") + -->
<!--     facet_wrap(~tf) + -->
<!--     theme_classic_lines_90() + -->
<!--     ggtitle(i) -->
<!--   print(p) -->
<!-- } -->

<!-- ggplot_custom(varimps, aes(x = feature, y = dropout_loss)) + -->
<!--   geom_boxplot() + -->
<!--   geom_quasirandom() + -->
<!--   theme_classic_lines_90() -->

<!-- ``` -->




# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

