---
title: "TF activity computation & feature analysis - TF reporter library 90 TFs - combined"
author: 
  - name: "Max Trauernicht"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute - van Steensel lab"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
---

---

### Aim
In more than 100 samples, 35,500 reporters for 86 TFs were transfected into various cells (mainly mES, K562 & HepG2), and stimulated with various compounds. To be more precise, the data consists of two different libraries, library 1, which is more focused on mES biology, and library 2, which is more focused on classical signaling pathways.All the barcode counts from the 7 different sequencing runs are combined in this analysis. In a previous scripts, sanity checks will be carried out to filter out samples with bad quality, and barcode counts will be normalized to compute reporter activitites. In this script, plots will be generated to characterize the TF reporter activities and models will be built to extract the importance of reporter features. 


---

## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}
```


### Libraries 

```{r setup, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(RColorBrewer)
library(ggplot2)
library(dplyr)
library(maditr)
library(tibble)
library(pheatmap)
library(ggpubr)
library(ggbeeswarm)
library(ggforce)
library(viridis)
library(plyr)
library(cowplot)
library(gridExtra)
library(pROC)
```

### Functions

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
### Custom functions
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}


# Custom ggplot2 themes
theme_classic_lines <- function() {
  theme_pubr(border = F, legend = "top") +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_classic_lines_45 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 45) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_classic_lines_90 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 90) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_set(theme_classic_lines())

colors_diverse <- c("#264653", "#9AC1AE", "#5D987B", "#f2cc8f", "#e76f51")

ggplot_custom <- function(...) ggplot2::ggplot(...) + 
  scale_color_manual(values = colors_diverse) + 
  scale_fill_manual(values = colors_diverse)
```

### Loading data

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Import processed bc counts from the preprocessing step
cDNA_df <- read.csv("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6578_gen2_stimulation-2/results/mt20211118_reporter_activity_filt_combined.csv", header = T)

# Filter TP53 from library 1+2 K562 data
cDNA_df2 <- cDNA_df %>%
  filter(library == "1+2" & condition == "K562_DMSO" & tf == "TP53")
cDNA_df <- anti_join(cDNA_df, cDNA_df2, by = c("reporter_id", "sample_id"))
```


## Reporter activities over background per tf and condition

```{r data_density, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# Change multiple hypothesis correction method
cDNA_df$pval_adj <- p.adjust(cDNA_df$pval, method = 'fdr')

# Plot per minimal promoter the reporter activites over the minimal promoter only and the adjusted p-value for each reporter
for (i in unique(cDNA_df$condition)) {
  
  p <- ggplot_custom(cDNA_df %>%
                filter(condition == i, neg_ctrls == "No") %>%
                dplyr::select(reporter_id, pval_adj, reporter_activity_minP, promoter) %>%
                unique(),
            aes(x = log(reporter_activity_minP), y = -log10(pval_adj),
                color = ifelse(log(reporter_activity_minP) > 1 & pval_adj < 0.05, "green", 
                               ifelse(log(reporter_activity_minP) < -1 & pval_adj < 0.05, "yellow","black")))) +
    geom_point() +
    geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
    geom_vline(xintercept = log2(2), linetype = "dashed") +
    geom_vline(xintercept = -log2(2), linetype = "dashed") +
    facet_wrap(~promoter)+
    ggtitle(i)
  
  print(p)
}


# Same plot for one condition only
ggplot_custom(cDNA_df %>%
                filter(condition == "HepG2_CDCA", neg_ctrls == "No") %>%
                dplyr::select(reporter_id, pval_adj, reporter_activity_minP, promoter) %>%
                unique() %>%
                mutate(reporter_activity_minP = log2(reporter_activity_minP),
                       pval_adj = -log10(pval_adj)),
            aes(x = reporter_activity_minP, y = pval_adj, 
                color = ifelse(reporter_activity_minP > 1 & pval_adj > -log10(0.05), ifelse(reporter_activity_minP > 3 & pval_adj > -log10(0.05), "green", "blue"), 
                               ifelse(reporter_activity_minP < -1 & pval_adj > -log10(0.05), "yellow","black")))) +
  geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  geom_vline(xintercept = log2(2), linetype = "dashed") +
  geom_vline(xintercept = -log2(2), linetype = "dashed") +
  facet_wrap(~promoter)

ggplot_custom(cDNA_df %>%
                filter(condition == "mES_2i_LIF", neg_ctrls == "No", library == "1") %>%
                dplyr::select(reporter_id, pval_neg, reporter_activity_neg, promoter) %>%
                mutate(pval_neg = p.adjust(pval_neg, method = 'fdr')) %>%
                unique(),
            aes(x = log(reporter_activity_neg), y = -log10(pval_neg), 
                color = ifelse(log(reporter_activity_neg) > 1 & pval_neg < 0.05, "green", 
                               ifelse(log(reporter_activity_neg) < -1 & pval_neg < 0.05, "yellow","black")))) +
  geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  geom_vline(xintercept = log2(2), linetype = "dashed") +
  geom_vline(xintercept = -log2(2), linetype = "dashed") +
  facet_wrap(~promoter)



# Show per condition how many reporters per TF are upregulated, downregulated, or unchanged
cDNA_df$reporter_activity_class <- 0
cDNA_df$reporter_activity_class[cDNA_df$pval_adj < 0.05 & cDNA_df$reporter_activity_minP > 2] <- 1
cDNA_df$reporter_activity_class[cDNA_df$pval_adj < 0.05 & cDNA_df$reporter_activity_minP > 8] <- 2
cDNA_df$reporter_activity_class[cDNA_df$pval_adj < 0.05 & cDNA_df$reporter_activity_minP < 0.5] <- -1


cDNA_df2 <- cDNA_df %>%
                filter(neg_ctrls == "No", commercial_reporter == "No") %>%
                dplyr::select(reporter_id, tf, reporter_activity_class, condition, promoter) %>%
                unique()

cDNA_df2 <- cDNA_df2 %>%
  mutate(class_count = ave(reporter_activity_class, tf, promoter, condition, FUN = function(x) sum(x)))


for (i in unique(cDNA_df2$condition)) {
  p <- ggplot_custom(cDNA_df2 %>%
                       filter(condition == i),
                     aes(x = reorder(tf, -class_count), fill = factor(reporter_activity_class, levels = c("0", "1", "2", "-1")))) +
    geom_bar(position = "fill") +
    geom_hline(yintercept = 0.25, linetype = "dashed") +
    geom_hline(yintercept = 0.75, linetype = "dashed") +
    theme_classic_lines_90() +
    facet_wrap(~promoter) +
    ggtitle(i)
  
  print(p)
}

ggplot_custom(cDNA_df2 %>%
                       filter(condition == "mES_2i_LIF"),
                     aes(x = reorder(tf, -class_count), fill = factor(reporter_activity_class, levels = c("0", "1", "2", "-1")))) +
    geom_bar(position = "fill") +
    geom_hline(yintercept = 0.25, linetype = "dashed") +
    geom_hline(yintercept = 0.75, linetype = "dashed") +
    theme_classic_lines_90()


## Which reporters are significantly different between two conditions?
cDNA_df3 <- cDNA_df %>%
  dplyr::select(reporter_id, condition, reporter_activity_minP, tf) %>%
  filter(condition == "mES_2i_LIF" | condition == "mES_LIF_PD") %>%
  unique() %>%
  spread(condition, reporter_activity_minP) %>%
  mutate(dif = mES_2i_LIF / mES_LIF_PD)

ggplot_custom(cDNA_df3, aes(x = tf, y = dif)) +
  geom_quasirandom() 

```


---

## For TFs that have mixed populations of reporters: what discriminates them?

```{r}
# Divide TFs per condition into different classes: activators, selective activator, inactive, selective repressor, repressor

## Determine amount of active/repressive reporters per tf
cDNA_df3 <- cDNA_df2 %>%
  mutate(class_count2 = ave(reporter_activity_class, tf, condition, FUN = function(x) length(x)),
         class_count_rel = ave(reporter_activity_class, reporter_activity_class, tf, condition, FUN = function(x) length(x))) %>%
  mutate(class_count_rel = class_count_rel / class_count2) %>%
  dplyr::select(tf, condition, class_count_rel, reporter_activity_class) %>%
  unique() %>%
  spread(reporter_activity_class, class_count_rel)

cDNA_df3$class[cDNA_df3$`0` >= 0.75] <- 0
cDNA_df3$class[cDNA_df3$`1` >= 0.75] <- 2
cDNA_df3$class[cDNA_df3$`2` >= 0.75] <- 4
cDNA_df3$class[cDNA_df3$`-1` >= 0.75] <- -2
cDNA_df3$class[cDNA_df3$`1` >= 0.25 & cDNA_df3$`1` < 0.75] <- 1
cDNA_df3$class[cDNA_df3$`1` < 0.25 & cDNA_df3$`2` < 0.25 & cDNA_df3$`1` + cDNA_df3$`2` >= 0.25] <- 1
cDNA_df3$class[cDNA_df3$`2` >= 0.25 & cDNA_df3$`2` < 0.75] <- 3
cDNA_df3$class[cDNA_df3$`-1` >= 0.25 & cDNA_df3$`-1` < 0.75] <- -2
cDNA_df3$class[cDNA_df3$`1` >= 0.085 & cDNA_df3$`-1` >= 0.085] <- -1


cDNA_df4 <- cDNA_df3 %>%
  dplyr::select(tf, condition, class) %>%
  unique() %>%
  na.omit() %>%
  filter(condition %in% c("HepG2_CDCA", "HepG2_DMSO", "K562_DMSO", "mES_2i", 
                          "mES_2i_LIF", "mES_HQ", "mES_LIF_CH", "mES_LIF_PD", 
                          "mES_N2B27")) %>%
  mutate(sum = ave(class, tf, FUN = function(x) sum(x))) %>%
  filter(sum != 0) %>%
  dplyr::select(-sum)

cDNA_df4 <- cDNA_df4 %>%
  spread(tf, class) %>%
  column_to_rownames('condition')

makeColorRampPalette <- function(colors, cutoff.fraction, num.colors.in.palette)
{
  stopifnot(length(colors) == 4)
  ramp1 <- colorRampPalette(colors[1:2])(num.colors.in.palette * cutoff.fraction)
  ramp2 <- colorRampPalette(colors[2:4])(num.colors.in.palette * (1 - cutoff.fraction))
  return(c(ramp1, ramp2))
}

cols <- makeColorRampPalette(c(colors_diverse[4], colors_diverse[1], colors_diverse[2], colors_diverse[3]), 
                             1-(4/6),
                             100)

pheatmap(as.matrix(cDNA_df4),
         color = cols,
         border_color = "white", 
         cellwidth = 8, cellheight = 8,
         angle_col = 90)
```

---

## Why are the reporters classified as repressive/activating? Is there some logic to it?

```{r}
# Which features explain the classification into activating/repressing?
cDNA_df_lm <- cDNA_df %>%
  filter(neg_ctrls == "No", commercial_reporter == "No", promoter != "Random",
         condition %in% c("HepG2_CDCA", "HepG2_DMSO", "K562_DMSO", "mES_2i", 
                          "mES_2i_LIF", "mES_HQ", "mES_LIF_CH", "mES_LIF_PD", 
                          "mES_N2B27"
                          ),
         tf %in% names(cDNA_df4)) %>%
  mutate(background = as.factor(background)) %>%
  dplyr::select(reporter_id, tf, reporter_activity_class, condition, promoter, spacing, background, distance) %>%
  unique()

cDNA_df3[is.na(cDNA_df3)] <- 0
mixed_classes <- cDNA_df3 %>%
  mutate(id = paste(condition, tf, sep = "_")) %>%
  dplyr::select(-class) %>%
  #filter(class != 0) %>%
  unique() %>%
  filter(`1` < 0.95, `0` < 0.95, `2` < 0.95)

## Fitting the model for each condition and tf
cDNA_df_lm$id <- paste(cDNA_df_lm$condition, cDNA_df_lm$tf, sep = "_")
cDNA_df_lm <- cDNA_df_lm %>%
  mutate(sum = ave(id, id, FUN = function(x) length(x))) %>%
  filter(sum >= 2, id %in% mixed_classes$id) %>%
  dplyr::select(-sum)



lm_features <- lapply(unique(cDNA_df_lm$id), function(x) lm(reporter_activity_class ~ promoter + spacing + background + distance, 
                                                                     cDNA_df_lm[cDNA_df_lm$id == x,]))

names(lm_features) <- unique(cDNA_df_lm$id)


# ROC
roc_lm <- list()
for (i in names(lm_features)) {
  roc_lm[[i]] <- roc(response = cDNA_df_lm$reporter_activity_class[cDNA_df_lm$id == i], predictor = round(lm_features[[i]]$fitted.values))
}

auc_lm <- data.frame("id" = names(roc_lm),
                     "n" = 1:length(roc_lm),
                     "auc" = "")
for (i in 1:length(roc_lm)) {
  if (i == 1) {
    plot <- plot(roc_lm[[i]], col="#264653", lwd=3, main="ROC curve RF")
    auc_lm$auc[auc_lm$n == i] <- auc(plot)
  }
  else {
    plot <- plot(roc_lm[[i]], add = TRUE, label = T)
    auc_lm$auc[auc_lm$n == i] <- auc(plot)
  }
}


auc_lm <- auc_lm %>%
  mutate(condition = gsub("(.*)_.*$", "\\1", id),
         tf = gsub(".*_.*_(.*$)", "\\1", id),
         auc = as.numeric(auc)) 

ggplot_custom(auc_lm, aes(x = tf, y = auc)) +
  geom_point() +
  geom_boxplot() +
  geom_hline(yintercept = 0.75, linetype = "dashed") +
  theme_classic_lines_90()
  
ggplot_custom(auc_lm, aes(x = condition, y = auc)) +
  geom_point() +
  geom_hline(yintercept = 0.75, linetype = "dashed", color = "red")+
  coord_flip()


# auc_lm_sel <- auc_lm %>%
#   filter(auc >= 0.75)
# 
# lm_features <- lm_features[auc_lm_sel$id]


# Extract weights
lm_weights <- list()
for (i in names(lm_features)) {
  lm_weights[[i]] <- lm_features[[i]]$coefficients
}

lm_weights <- bind_rows(lm_weights, .id = "id")

lm_weights <- lm_weights %>%
  dplyr::select(-`(Intercept)`) %>%
  pivot_longer(-id, names_to = "feature", values_to = "weight")

ggplot_custom(lm_weights, aes(x = feature, y = weight)) +
  geom_quasirandom()


# Extract features
exp_features <- list()
for (i in names(lm_features)) {
  exp_features[[i]] <- anova(lm_features[[i]]) %>%
  dplyr::select('sum_sq' = `Deviance`) %>%
    na.omit() %>%
  mutate(sum = sum(sum_sq)) %>%
  mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
  setnames("rel_sum_sq", "percent") %>%
  dplyr::select("percent")
}
exp_features <- bind_rows(exp_features, .id = "id") %>%
  rownames_to_column("feature") %>%
  mutate(feature = gsub("...[0-9]{1-3}", "", feature)) %>%
  mutate(feature = gsub("[0-9]", "", feature)) %>%
  mutate(percent = format(percent, scientific = F, digits = 0)) %>%
  mutate(percent = as.numeric(percent))

exp_features$group <- "no residual"
exp_features$group[grep("Residual", exp_features$feature)] <- "residual"
exp_features$percent[exp_features$feature == "Residuals"] <- 100 - exp_features$percent[exp_features$feature == "Residuals"]
exp_features$feature <- revalue(exp_features$feature, c("Residuals"="total"))


# Plotting the features
ggplot_custom(exp_features, 
       aes(x = percent, y = reorder(feature, -percent), fill = group)) +
  geom_bar(stat = "identity") + 
  xlab("Variance explained (%)") + 
  ylab("") 

classes <- cDNA_df3 %>%
  mutate(id = paste(condition, tf, sep = "_")) %>%
  dplyr::select(id, class) %>%
  unique()

exp_features <- merge(exp_features, classes, by = "id")

ggplot_custom(exp_features, 
       aes(y = percent, x = feature, color = group)) +
  geom_quasirandom() + 
  xlab("Variance explained (%)") + 
  ylab("") +
  facet_wrap(~class)

# Keep features from good models only
exp_features_good <- exp_features %>%
  filter(group == "residual")
exp_features_good_auc <- merge(exp_features_good, auc_lm, by = "id")

ggplot_custom(exp_features_good_auc,
              aes(x = percent/100, y = auc)) +
  geom_point()


exp_features_good <- exp_features_good %>%
  filter(percent >= 50)
exp_features_sel <- exp_features[exp_features$id %in% exp_features_good$id,]

exp_features_sel <- exp_features_sel %>%
  mutate(tf = gsub(".*_(.*)", "\\1", id),
         condition = gsub("(.*)_.*", "\\1", id))

# Plot again the features
ggplot_custom(exp_features_sel, 
       aes(y = percent, x = feature, color = group)) +
  geom_quasirandom() + 
  xlab("Variance explained (%)") + 
  ylab("")

order <- exp_features_sel %>%
                mutate(sum = ave(percent, id, FUN = function(x) sum(x))) %>%
  dplyr::select(sum, tf) %>%
  unique() %>%
  group_by(tf) %>%
  top_n(1, sum)

exp_features_sel_ord <- merge(order, exp_features_sel, by = "tf") %>%
  filter(feature != "total")
  

ggplot(exp_features_sel_ord %>%
         filter(condition %in% c("K562_DMSO", "HepG2_DMSO", "mES_2i_LIF")), 
       aes(y = percent, x = reorder(id, -sum), fill = feature, group = feature)) +
  geom_bar(stat = "identity", position = "stack") + 
  xlab("Variance explained (%)") + 
  ylab("") +
  scale_fill_manual(values = c("#e07a5f", "#3d405b", "#81b29a", "#f2cc8f"))+
  theme_classic_lines_90() 


ggplot_custom(exp_features_sel, aes(x=feature, y=percent, color=group, group = id)) + 
  geom_line() +
  geom_point()

exp_features_sel_heatmap <- exp_features_sel %>%
  filter(condition %in% c("mES_2i_LIF", "K562_DMSO", "HepG2_DMSO")) %>%
  mutate(cell_type = gsub("(.*)_.*", "\\1", condition)) %>%
  mutate(cell_type = gsub("(.*)_.*", "\\1", cell_type)) %>%
  dplyr::select(-group, -condition) %>%
  filter(feature != "total") %>%
  spread(feature, percent) %>% 
  column_to_rownames("id") %>%
  arrange(tf)

cell_type <- exp_features_sel_heatmap %>%
  dplyr::select(cell_type)

colors_pheatmap <- list('cell_type' = c("K562" = colors_diverse[1], "HepG2" = colors_diverse[5], "mES" = colors_diverse[4]))

pheatmap(as.matrix(t(exp_features_sel_heatmap %>% dplyr::select(-cell_type, -tf, -class))),
         color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
         border_color = "white", 
         cellwidth = 8, cellheight = 8,
         angle_col = 90, annotation_col = cell_type, cluster_cols = T,
         annotation_colors = colors_pheatmap)

# Explore selected ids
ids_selected <- cDNA_df_lm %>%
  filter(id %in% rownames(exp_features_sel_heatmap)) %>%
  mutate(reporter_activity_class = as.character(reporter_activity_class),
         sum_all = ave(id, id, FUN = function(x) length(x)),
         sum_class = ave(id, id, reporter_activity_class, FUN = function(x) length(x))) %>%
  mutate(sum_all = as.numeric(sum_all),
         sum_class = as.numeric(sum_class),
         frac = sum_class / sum_all)

## GRHL1, POU5F1::SOX2, MAF::NFE2
reporter_correlations <- cDNA_df %>%
  filter(tf %in% c("GRHL1", "POU5F1::SOX2", "MAF::NFE2"),
         condition %in% c("mES_2i_LIF", "K562_DMSO", "HepG2_DMSO")) %>%
  dplyr::select(reporter_id, condition, tf, pval_adj) %>%
  mutate(pval_adj = -log10(pval_adj)) %>%
  unique() %>%
  spread(condition, pval_adj)

ggplot_custom(reporter_correlations, aes(x = K562_DMSO, y = mES_2i_LIF)) +
  geom_point() +
  facet_wrap(~tf, scales = "free")
```

---

## Random forest model to classify reporters (performance mediocre) 

```{r}
cDNA_df_lm <- cDNA_df %>%
  filter(neg_ctrls == "No", commercial_reporter == "No", promoter != "Random",
         condition %in% c("HepG2_CDCA", "HepG2_DMSO", "K562_DMSO", "mES_2i", 
                          "mES_2i_LIF", "mES_HQ", "mES_LIF_CH", "mES_LIF_PD", 
                          "mES_N2B27"
                          ),
         tf %in% names(cDNA_df4),
         reporter_activity_class != -1) %>%
  mutate(background = as.factor(background),
         reporter_activity_class = as.factor(reporter_activity_class)) %>%
  dplyr::select(reporter_id, tf, reporter_activity_class, condition, promoter, spacing, background, distance) %>%
  unique()

mixed_classes <- cDNA_df3 %>%
  mutate(id = paste(condition, tf, sep = "_")) %>%
  dplyr::select(id, class) %>%
  filter(class == 1) %>%
  unique()

## Fitting the model for each condition and tf
cDNA_df_lm$id <- paste(cDNA_df_lm$condition, cDNA_df_lm$tf, sep = "_")
cDNA_df_lm <- cDNA_df_lm %>%
  mutate(sum = ave(id, id, FUN = function(x) length(x))) %>%
  filter(sum >= 2, id %in% mixed_classes$id) %>%
  dplyr::select(-sum)

rf_features <- lapply(unique(cDNA_df_lm$id), function(x) randomForest(reporter_activity_class ~ promoter + spacing + background + distance, 
                                                                     cDNA_df_lm[cDNA_df_lm$id == x,]))

names(rf_features) <- unique(cDNA_df_lm$id)


# ROC
roc_rf_train <- list()
for (i in names(rf_features)) {
  roc_rf_train[[i]] <- roc(response = cDNA_df_lm$reporter_activity_class[cDNA_df_lm$id == i], predictor = rf_features[[i]]$votes[,2])
}

auc_rf <- data.frame("id" = names(roc_rf_train),
                     "n" = 1:length(roc_rf_train),
                     "auc" = "")
for (i in 1:length(roc_rf_train)) {
  if (i == 1) {
    plot <- plot(roc_rf_train[[i]], col="#264653", lwd=3, main="ROC curve RF")
    auc_rf$auc[auc_rf$n == i] <- auc(plot)
  }
  else {
    plot <- plot(roc_rf_train[[i]], add = TRUE, label = T)
    auc_rf$auc[auc_rf$n == i] <- auc(plot)
  }
}

auc_rf <- auc_rf %>%
  mutate(condition = gsub("(.*)_.*$", "\\1", id),
         tf = gsub(".*_.*_(.*$)", "\\1", id),
         auc = as.numeric(auc)) 

ggplot_custom(auc_rf, aes(x = tf, y = auc)) +
  geom_point() +
  geom_boxplot() +
  geom_hline(yintercept = 0.75, linetype = "dashed") +
  theme_classic_lines_90()
  
ggplot_custom(auc_rf, aes(x = condition, y = auc)) +
  geom_point() +
  geom_hline(yintercept = 0.75, linetype = "dashed", color = "red")+
  coord_flip()


auc_rf_sel <- auc_rf %>%
  filter(auc >= 0.75)

rf_features <- rf_features[auc_rf_sel$id]


# Get the variable importances
explained_rf <- list()

for (i in names(rf_features)) {
  explained_rf[[i]] <- explain(rf_features[[i]], data=cDNA_df_lm %>% dplyr::select(promoter, spacing, background, distance), y=cDNA_df_lm$reporter_activity_class)
}

varimps <- list()
for (i in names(explained_rf)) {
  varimps[[i]] <- feature_importance(explained_rf[[i]], type = "ratio")
}
#, loss_function = '1-auc' type='ratio'

varimps <- bind_rows(varimps, .id = "id")

varimps <- varimps %>%
  mutate(dropout_loss = ave(dropout_loss, id, variable, FUN = function(x) mean(x))) %>%
  dplyr::select(id, 'feature' = variable, dropout_loss) %>%
  mutate(condition = gsub("(.*)_.*$", "\\1", id),
         tf = gsub(".*_.*_(.*$)", "\\1", id)) %>%
  dplyr::select(-id) %>%
  unique()

for (i in unique(varimps$condition)) {
  p <- ggplot_custom(varimps %>% filter(condition == i), aes(y = dropout_loss, x = feature)) +
    geom_point()+
    ylab("dropout loss (RMSE)") +
    facet_wrap(~tf) +
    theme_classic_lines_90() +
    ggtitle(i)
  print(p)
}

ggplot_custom(varimps, aes(x = feature, y = dropout_loss)) +
  geom_boxplot() +
  geom_quasirandom() +
  theme_classic_lines_90()

```



---

## How to select the best reporter per tf?

```{r}
# Extract TF activities for K562 & HepG2
tf_activities <- cDNA_df %>%
  filter(condition %in% c("K562_DMSO", "HepG2_DMSO"), neg_ctrls == "No") %>%
  dplyr::select(reporter_id, reporter_activity_minP, condition, tf) %>%
  unique() %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  spread(condition, reporter_activity_minP) %>%
  mutate(dif = log2(K562_DMSO / HepG2_DMSO)) %>%
  dplyr::select(tf, dif, reporter_id)

#write_csv2(tf_activities, "/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/diffTF/tf_reporter_activities_all.csv")

# Correlate with RNA-seq
load('/DATA/usr/m.martinez.ara/GitLab/gurten/gurten/data/fc181121_epsure_tadec_rnaseq_expr_joshi.RData')

# remove Serum data and calculate mean expression
tib_fpkm_joshi$twoi <- (tib_fpkm_joshi$twoi_1 + tib_fpkm_joshi$twoi_2)/2
tib_fpkm_joshi <- tib_fpkm_joshi %>%
  dplyr::select(gene_id, 'tf' = symbol, twoi)

tf_activities2 <- cDNA_df %>%
  filter(condition == "mES_2i_LIF", neg_ctrls == "No") %>%
  dplyr::select(reporter_id, reporter_activity_minP, tf) %>%
  unique() %>%
  mutate(tf = gsub("_.*", "", tf),
         tf_activity = ave(reporter_activity_minP, tf, FUN = function(x) mean(x))) %>%
  dplyr::select(tf_activity, tf) %>%
  unique()

tf_activities2 <- merge(tf_activities2, tib_fpkm_joshi, by = "tf") %>%
  mutate(tf_activity = (tf_activity-mean(tf_activity))/sd(tf_activity),
         twoi = (twoi-mean(twoi))/sd(twoi))

ggplot_custom(tf_activities2, aes(x = twoi, y = tf_activity)) +
  geom_point()

# Correlate with regulon


# Correlate with TF activities from Ernst et al. (https://doi.org/10.1038/nbt.3678)
ernst_activities <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/ernst_tf_activities_hepg2_k562.csv", sep = ";", dec = ",", header = T, fill = T)

ernst_activities <- ernst_activities %>%
  dplyr::select('tf' = TF, avg_score_HepG2, avg_score_K562) %>%
  mutate(avg_score_HepG2 = ave(avg_score_HepG2, tf, FUN = function(x) mean(x)),
         avg_score_K562 = ave(avg_score_K562, tf, FUN = function(x) mean(x))) %>%
  unique() %>%
  pivot_longer(-tf, names_to = "condition", values_to = "ernst_activity") %>%
  mutate(condition = gsub("avg_score_", "", condition))

tf_activities <- cDNA_df %>%
  filter(condition %in% c("K562_DMSO", "HepG2_DMSO"), neg_ctrls == "No") %>%
  dplyr::select(reporter_activity_minP, condition, tf) %>%
  mutate(tf = gsub("_.*", "", tf),
         condition = gsub("_DMSO", "", condition),
         reporter_activity_minP = ave(reporter_activity_minP, condition, tf, FUN = function(x) mean(x))) %>%
  unique()

ernst_activities <- merge(ernst_activities, tf_activities, by = c('tf', 'condition')) %>%
  mutate(ernst_activity = (ernst_activity-mean(ernst_activity))/sd(ernst_activity),
         reporter_activity_minP = (reporter_activity_minP-mean(reporter_activity_minP))/sd(reporter_activity_minP))

ggplot_custom(ernst_activities, 
              aes(x = ernst_activity, y = reporter_activity_minP)) +
  geom_point() +
  geom_abline(linetype = "dashed") +
  geom_smooth(method = "lm", color = "grey", fill = "lightgrey") +
  facet_wrap(~condition)

write_csv2(ernst_activities, "/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/diffTF/tf_reporter_activities_ernst.csv")
```

---

## Heat map - display activity for each TF in each condition

```{r tf_activity_heatmap, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
## Prepare dataframe
# Caculate mean TF activity per condition
tf_activity_heatmap <- cDNA_df %>%
  filter(condition %in% c("HepG2_DMSO", "K562_DMSO", "mES_2i_LIF", "mES_2i", "mES_N2B27", "mES_LIF_PD", "mES_LIF_CH"),
         neg_ctrls == "No", commercial_reporter == "No") %>%
  mutate(tf_activity = ave(reporter_activity_minP, condition, tf, FUN = function(x) mean(x))) %>%
  #mutate(tf_activity = log2(tf_activity)) %>%
        #mutate(tf_activity = ave(tf_activity, condition, FUN = function(x) (x-mean(x))/sd(x))) %>%
  dplyr::select(tf_activity, condition, tf) %>%
  unique() %>%
  spread(condition, tf_activity) %>% 
  remove_rownames %>% 
  column_to_rownames(var="tf")

# Keeping the scale in the pheatmap function
myBreaks1 <- seq(0,6,0.06)

# Keep only TFs that are in at least one condition above a certain threshold
tf_activity_heatmap <- tf_activity_heatmap[rowSums(tf_activity_heatmap > 4) >= 1,] %>%
  na.omit()

tf_activity_heatmap <- tf_activity_heatmap[,c(2,1,6,7,3,8,4,5)]

# pheatmap function
pheatmap(as.matrix(t(tf_activity_heatmap)),
         color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
         breaks = myBreaks1, border_color = "black", 
         cellwidth = 8, cellheight = 8,
         angle_col = 90)
```

---

##

```{r}

## reporter activity distribution for each experiment individually
plot1 <- cDNA_df %>% filter(cDNA_df$library =="2", cDNA_df$commercial_reporter =="No", cDNA_df$hPGK == "No", cDNA_df$native_enhancer == "No", cDNA_df$rand_promoter == "No")

ggplot(data = plot1 %>% dplyr::select(condition, tf_activity, tf, neg_ctrls) %>% unique(), aes(y = tf_activity,x = condition, alpha = 0.4)) + 
  geom_quasirandom(alpha = 0.4) + xlab("log2(cDNA/pDNA)") + 
  labs(title = "reporter activity distribution per condition") + scale_color_brewer(palette = "Set1") +  
  geom_quasirandom(aes(color = neg_ctrls, dodge.width = 0.75))


```


```{bash run fimo db2_1, eval = FALSE}
# motfn=/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/20170320_pwms_selected.meme
# odir=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/results/native-enhancer/fimo
# query=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/results/native-enhancer/cDNA_df_native.fasta

# nice -n 19 fimo --no-qvalue --thresh 1e-4 --verbosity 1 --o $odir $motfn $query 
```


```{r build tf motif matrices db2: random_promoter, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# # load motif Metadata --> PWM feature matrix
# tib_pwm_native_enhancer  <- get_pwm_feature_matrix(motif_meta_fn = '/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/fc181127_curated_metadata_no_composite_filt.csv', 
#                                         fimo_fn       = '/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/results/native-enhancer/fimo/fimo.tsv',
#                                         db            = 2)
```


```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# ## Add groups so I can plot them individually
# tib_pwm_native_enhancer$group <- gsub("(^e[0-9]{1,2}).*", "\\1", tib_pwm_native_enhancer$id)
# controls <- c("minP", "mCMV", "hBGm")
# tib_pwm_native_enhancer$group <- gsub(paste(controls, collapse = "|"), "Control", tib_pwm_native_enhancer$group)
```


```{r out.width= "80%", fig.align= "center", echo=FALSE}
# load('/DATA/usr/m.martinez.ara/GitLab/gurten/gurten/data/fc181121_epsure_tadec_rnaseq_expr_joshi.RData')
# 
# # Take mean of 2i expression
# tib_fpkm_joshi <- tib_fpkm_joshi %>%
#   dplyr::select(-contains('serum')) %>%
#   rowwise() %>%
#   mutate(expr_mean = mean(c(twoi_1, twoi_2))) %>%
#   ungroup()
```

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# # only select fpkm>0
# tib_fpkm_joshi_expressed <- tib_fpkm_joshi %>% filter(expr_mean>10)
# 
# # Change to mouse name of Tp53
# tib_fpkm_joshi_expressed$symbol <- gsub("TRP53", "TP53", tib_fpkm_joshi_expressed$symbol)
# 
# tib_pwm_native_enhancer_ES <- tib_pwm_native_enhancer[,colnames(tib_pwm_native_enhancer) %in% tib_fpkm_joshi_expressed$symbol | colnames(tib_pwm_native_enhancer) %in% c("id", "group")]
```


```{r, fig.height=10, fig.width=10, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# # remove columns with only 0s and plot heatmap for each native enhancer individually
# for (i in unique(tib_pwm_native_enhancer_ES$group)) {
#   data <- tib_pwm_native_enhancer_ES[tib_pwm_native_enhancer_ES$group == i,] %>% 
#     column_to_rownames("id") %>% dplyr::select(-group)
#   p <- pheatmap(as.matrix(data[, colSums(data != 0) > 0]),
#                 color = colorRampPalette(brewer.pal(n = 7, name = "Oranges"))(100),
#                 border_color = "black",
#                 cellwidth = 8, cellheight = 10)
#   print(p)
# }
# 
# ## Zic2/Hic2 bind to mCMV/hBGm -> Does this explain increased cooperativity?
# 
# # make list of relevant TFs
# relevant_TFs <- c("KLF", "SP1", "SP3", "POU", "RAR", "SOX", "STAT", "^TCF", "TFCP", "TP53")
# relevant_TFs_2 <- c("KLF4", "POU", "SOX", "STAT", "^TCF", "TFCP")
# relevant_TFs_3 <- "TFCP"
# 
# tib_pwm_native_enhancer_ES <- tib_pwm_native_enhancer_ES[-grep(paste(promoters$TF, collapse = "|"), tib_pwm_native_enhancer_ES$id),]
# 
# data <- tib_pwm_native_enhancer_ES %>%
#   dplyr::select(KLF4, POU5F1, SOX2, TCF7, TFCP2L1, CTCF, ZIC2,id, group) %>% 
#   column_to_rownames("id") %>% dplyr::select(-group)
# 
# pheatmap(as.matrix(t(data[, colSums(data != 0) > 0])),
#                 color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
#                 border_color = "black", cluster_cols = F, 
#                 cellwidth = 10, cellheight = 10)

```


```{r fimo_model, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# # Subselect interesting TFs
# tib_pwm_native_enhancer_ES <- tib_pwm_native_enhancer_ES %>%
#   dplyr::select(KLF4, POU5F1, SOX2, TCF7, TFCP2L1, CTCF, ZIC2,id, group)
# 
# # Transform to binary
# tib_pwm_native_enhancer_ES <- tib_pwm_native_enhancer_ES %>% mutate_if(is.numeric, function(x) ifelse((x==0 | is.na(x)),"No","Yes"))
# 
# # Combine with expression data
# native_expression <- tf_activity_heatmap %>% rownames_to_column("condition") %>%
#   melt(id.vars = "condition", variable.name = "id", value.name = "expression")
# 
# tib_pwm_native_enhancer_ES_merge <- merge(tib_pwm_native_enhancer_ES, native_expression, all = T) %>%
#   dplyr::select(-group)
# 
# 
# 
# ### Calculate expression variance
# exp_variance <- data.frame(c("KLF4", "POU5F1", 
#                              "SOX2", "TCF7", "TFCP2L1", "CTCF", "ZIC2", "Residuals"))
# prediction <- data.frame(1:280)
# names(prediction) <- "id"
# names(exp_variance) <- "feature"
# weight <- data.frame(c("(Intercept)", "KLF4Yes", "POU5F1Yes", 
#                              "SOX2Yes", "TCF7Yes", "TFCP2L1Yes", "CTCFYes", "ZIC2Yes"))
# names(weight) <- "feature"
# 
# 
# tib_pwm_native_enhancer_ES_merge$row <- rownames(tib_pwm_native_enhancer_ES_merge)
# 
# 
# 
# for (i in unique(tib_pwm_native_enhancer_ES_merge$condition)) {
#   x <- lm(expression ~ KLF4 + POU5F1 + SOX2 + TCF7 + TFCP2L1 + CTCF + ZIC2, 
#         tib_pwm_native_enhancer_ES_merge[tib_pwm_native_enhancer_ES_merge$condition == i,])
#   y <- data.frame(x$fitted.values) %>% rownames_to_column()
#   par(mfrow=c(2,2))
#   plot(x, main = i)
#   names(y) <- c("row", i)
#   y$id <- 1:nrow(y)
#   prediction <- merge(prediction,y, all = T)
#   
#  
#   w <- data.frame(x$coefficients) %>% rownames_to_column()
#   names(w) <- c("feature", i)
#   weight <- merge(weight,w, all = T)
#   
#   x <- anova(x) %>% rownames_to_column("feature") %>%
#     setnames("Sum Sq", "sum_sq") %>%
#     dplyr::select(feature, sum_sq) %>%
#     mutate(sum = sum(sum_sq)) %>%
#     mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
#     setnames("rel_sum_sq", i)
#   x <- x[,c(1,4)]
# 
#   exp_variance <- merge(exp_variance, x)
# }
# 
# 
# 
# ## Prepare lm parameters for plotting
# exp_variance <- melt(exp_variance, variable.name = "id")
# exp_variance$group <- "no residual"
# exp_variance$group[grep("Residual", exp_variance$feature)] <- "residual"
# prediction <- prediction %>% dplyr::select(-id) %>% melt(variable.name = "id") %>% na.omit() %>%
#   dplyr::select(-id) %>% setnames("value", "activity_predicted")
# tib_pwm_native_enhancer_ES_merge <- merge(tib_pwm_native_enhancer_ES_merge, prediction)
# weight <- weight %>% filter(feature != "(Intercept)")
# weight <- melt(weight, variable.name = "id")
# categorie <- c("KLF4", "POU5F1", "SOX2", "TCF7", "TFCP2L1", "CTCF", "ZIC2")
# #weight$features <- gsub(paste(categorie, collapse="|"), "",weight$feature)
# #feature <- weight$features
# weight$condition <- gsub(paste(feature, collapse="|"), "", weight$feature)
# weight$value[is.na(weight$value)] <- 0
# 
# 
# # Further plotting preparations
# level_order <- c("KLF4Yes", "POU5F1Yes", "SOX2Yes", "CTCFYes", "ZIC2Yes", "TFCP2L1Yes", "TCF7Yes")
# 
# 
# # Model parameters
# exp_variance$value[exp_variance$feature == "Residuals"] <- 100 - exp_variance$value[exp_variance$feature == "Residuals"]
# exp_variance$feature <- revalue(exp_variance$feature, c("Residuals"="total"))
# 
# 
# 
# ggplot(exp_variance, 
#        aes(x = reorder(feature, -value), y = value, fill = group)) +
#   scale_fill_manual(values = c("#1B998B", "#2D3047")) +
#   geom_bar(stat = "identity") + 
#   ylab("Variance explained (%)") + xlab("") + 
#   labs(title = "Linear variance modelling", 
#        subtitle = "Expression variance predicted by features") +
#   
#   theme(text = element_text(size = 12)) +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12), 
#         axis.text.y = element_text(size = 10)) +
#     guides(fill = F) + facet_wrap(~id)
# 
# ggscatter(tib_pwm_native_enhancer_ES_merge, x = "expression", y = "activity_predicted",
#           add = "reg.line", 
#           add.params = list(color = "blue", fill = "lightgray"),
#           conf.int = TRUE, ylab = "Predicted log2-activity", size = 0.5,
#           xlab = "Average log2-activity") +  facet_wrap(~condition) +
#   stat_cor(aes(label = ..r.label..), method = "pearson", label.x = -2, label.y = 4)
# 
# 
# 
# for (i in unique(tib_pwm_native_enhancer_ES_merge$condition)) {
#   
#   # Plot to show prediction accuracy
#   sp <- ggscatter(tib_pwm_native_enhancer_ES_merge[tib_pwm_native_enhancer_ES_merge$condition == i,], x = "expression", y = "activity_predicted",
#    add = "reg.line", 
#    add.params = list(color = "blue", fill = "lightgray"),
#    conf.int = TRUE, ylab = "Predicted log2-activity", 
#    title = i,
#    xlab = "Average log2-activity") + theme_bw()
#   
#   
#   # Feature importance overview for each TF (including mean activities per feature)
#   fp <- ggplot(exp_variance[exp_variance$id == i,], 
#        aes(x = reorder(feature, -value), y = value, fill = group)) +
#   scale_fill_manual(values = c("#1B998B", "#2D3047")) +
#   geom_bar(stat = "identity") + 
#   ylab("Variance explained (%)") + xlab("") +
#   ggtitle(i)+
#   
#   theme(text = element_text(size = 12)) +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12), 
#         axis.text.y = element_text(size = 10)) +
#     guides(fill = F)
#   
#   # Plot weights of parameters in model - this helps to understand the model 
#   p <- ggplot(weight[weight$id == i,], 
#        aes(x = factor(feature, level = level_order), y = value, fill = condition)) +
#   scale_fill_manual(values = c("#DD6B48", "#1B998B", "#1B998B", "#1B998B", "#1B998B", "#1B998B", "1B998B")) +
#   ggtitle(i)+
#   geom_bar(stat = "identity") + 
#   ylab("Weight") + xlab("") +
#   
#   theme(text = element_text(size = 12)) +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
#         axis.text.y = element_text(size = 10)) +
#     guides(fill = F) 
# 
#   # Combine all plots
#   sp.plot <- arrangeGrob(sp,fp,p, nrow = 2, ncol = 2)
#   grid.arrange(sp.plot)
# }
```



### Heatmap per TF - comparing design activities mutated vs. non-mutated
```{r heatmap_4, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
## Prepare dataframe
# Caculate mean TF activity per condition
cDNA_df$reporter <- gsub("_[0-9]{1}$", "", cDNA_df$reporter_id)
cDNA_df$tf_design_activity <- ave(cDNA_df$log_activity, 
                                  cDNA_df$condition, 
                                  cDNA_df$reporter,
                                  FUN = function(x) mean(x))


tf_activity_heatmap <- cDNA_df[cDNA_df$native_enhancer == "No" & 
                                 cDNA_df$rand_promoter == "No" &
                                 cDNA_df$hPGK == "No" & cDNA_df$library == "2",] %>% 
  dplyr::select(reporter, condition, tf_design_activity) %>% unique()

tf_activity_heatmap$tf_design_activity[!is.finite(tf_activity_heatmap$tf_design_activity)] <- 0
tf_activity_heatmap <- dcast(tf_activity_heatmap, reporter ~ condition,
                             value.var="tf_design_activity")
tf_activity_heatmap$tf <- gsub("(^.*?)_[0-9]{1,2}bp.*", "\\1", tf_activity_heatmap$reporter)
tf_activity_heatmap$tf <- gsub("_neg$", "", tf_activity_heatmap$tf)
tf_activity_heatmap <- tf_activity_heatmap %>% 
  remove_rownames %>% column_to_rownames(var="reporter")
# Keeping the scale in the pheatmap function
myBreaks1 <- seq(-3,3,0.06)

# Manually annotate heatmaps
annotation_col <- data.frame("TF" = cDNA_df$reporter[cDNA_df$native_enhancer == "No" & 
                                 cDNA_df$rand_promoter == "No" &
                                 cDNA_df$hPGK == "No"], 
                             "neg_ctrls" = cDNA_df$neg_ctrls[cDNA_df$native_enhancer == "No" & 
                                 cDNA_df$rand_promoter == "No" &
                                 cDNA_df$hPGK == "No"]) %>%
  unique() %>%
  remove_rownames %>% column_to_rownames(var="TF")

annotation_colors = list(
    neg_ctrls = c(No ="#1B998B", Yes = "#2D3047"))


# pheatmap function
for(i in unique(tf_activity_heatmap$tf)){
  pheatmap(as.matrix(t(tf_activity_heatmap[tf_activity_heatmap$tf == i, -8])),
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks1, border_color = "#565656", 
         cellwidth = 10, cellheight = 10,
         annotation_col = annotation_col, 
         annotation_colors = annotation_colors)
}

NR1H4 <- tf_activity_heatmap[grep("NR1H4::RXRA|NR1H4::RXRA_TF-seq_", tf_activity_heatmap$tf),]

pheatmap(as.matrix(t(NR1H4[NR1H4$tf == "NR1H4::RXRA"| NR1H4$tf == "NR1H4::RXRA_TF-seq_", -8])),
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks1, border_color = "#565656", 
         cellwidth = 10, cellheight = 10,
         annotation_col = annotation_col, 
         annotation_colors = annotation_colors)


pheatmap(as.matrix(t(tf_activity_heatmap[grep("NR1H2", tf_activity_heatmap$tf), -8])),
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks1, border_color = "#565656", 
         cellwidth = 10, cellheight = 10,
         annotation_col = annotation_col, 
         annotation_colors = annotation_colors)



```



### Plotting quasirandom plots instead of heatmaps
```{r}
# Plot activity per reporter
tf_activity_df <- tf_activity_heatmap %>%
  rownames_to_column(var = "reporter") 

tf_activity_df$group  <- "Other"

# HepG2_enhanced <- c("GATA4", "HNF1A", "HNF4A", "NR1D1", "NR1H4::RXRA", "NR1I2", "NR1I3", "ONECUT1", "PPARA::RXRA", "RORA", "SOX9", "TCF7", "FOXA1")
# tf_activity_df$group[grep(paste(HepG2_enhanced, collapse = "|"), tf_activity_df$tf)] <- "HepG2_enhanced"
# 
# K562_enhanced <- c("GATA1", "REL", "WT1", "EGR1", "POU2F1", "SRF", "MTF1", "CREB1", "SP1", "NFYA", "STAT3", "XBP1", "IRF3","MAF::NFE2", "THRB")
# tf_activity_df$group[grep(paste(K562_enhanced, collapse = "|"), tf_activity_df$tf)] <- "K562_enhanced"

active <-  c("NR1H4::RXRA", "NR4A2::RXRA", "STAT1::STAT2", "POU2F1", "SRF", "XBP1","MAF::NFE2","ETS2", "REL", "HNF1A", "TCF7", "FOS::JUN", "NRF1")
tf_activity_df$group[grep(paste(active, collapse = "|"), tf_activity_df$reporter)] <- "active"


tf_activity_df <- tf_activity_df[-grep("neg|promega|TF-seq|romanov|RANDOM", tf_activity_df$reporter),] %>%
  dplyr::select(tf, reporter, group, "HepG2_DMSO", "K562_DMSO")


tf_activity_df <- tf_activity_df %>%
  mutate(mean_HepG2 = ave(HepG2_DMSO, tf, FUN = function(x) mean(x)),
         sd_HepG2 = ave(HepG2_DMSO, tf , FUN = function(x) sd(x)),
         mean_K562 = ave(K562_DMSO, tf, FUN = function(x) mean(x)),
         sd_K562 = ave(K562_DMSO, tf , FUN = function(x) sd(x)))
          
tf_activity_df2 <- tf_activity_df[tf_activity_df$group != "Other",] %>% melt(id.vars = "reporter", measure.vars = c("HepG2_DMSO", "K562_DMSO")) 
tf_activity_df2$tf <- gsub("(^.*?)_[0-9]{1,2}bp.*", "\\1", tf_activity_df2$reporter)


ggplot(data = tf_activity_df2, aes(x = reorder (tf, -value), y = `value`, color = variable)) +
  geom_quasirandom() +
  
  geom_hline(yintercept = 1, linetype = "dashed") +
  ylim(-2.5,7.5) + 
  ylab("reporter activity (log2)")+
  xlab("TF") +
  geom_quasirandom(dodge.width = 0.50, aes(color = variable)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))




ggplot(data = tf_activity_df[tf_activity_df$group != "Other",] , aes(x = tf, y = `K562_DMSO`, color = `K562_DMSO`)) +
  geom_quasirandom() +
  
  geom_hline(yintercept = 1, linetype = "dashed") +
  ylab("reporter activity (log2)") +
  coord_flip() +
  scale_color_viridis() +
  ggtitle("Reporter expression of HepG2 vs K562 enhanced TFs")

ggplot(tf_activity_df, aes(x = reorder(tf, -mean), y = `K562_DMSO`)) +
  geom_bar(data = tf_activity_df %>% dplyr::select(tf, mean, sd) %>% unique(), 
           mapping = aes(x = reorder(tf, -sd), y = sd), stat = "identity", alpha = 0.5) +
  
  ylim(0,4) +
  ylab("SD reporter activity") +
  coord_flip()


ggplot(tf_activity_df %>% filter(tf == "NR1H4::RXRA"), aes(x = tf, y = `HepG2_CDCA`, color = `HepG2_CDCA`)) +
  geom_quasirandom(width = 0.1) +
  
  geom_hline(yintercept = 1, linetype = "dashed") +
  ylab("reporter activity (log2)") +
  coord_flip() +
  scale_color_viridis()



# Plot mean activity per TF vs SD within reporters
ggplot(tf_activity_df, aes(x = mean, y = sd)) +
  geom_point() +
  theme_bw()
```




### Heatmap per TF - only WT TF activities
```{r heatmap_5, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
## Prepare dataframe
# Caculate mean TF activity per condition
tf_activity_heatmap <- cDNA_df[cDNA_df$native_enhancer == "No" & 
                                 cDNA_df$hPGK == "No" &
                                 cDNA_df$neg_ctrls == "No",] %>% 
  dplyr::select(reporter, condition, tf_design_activity) %>% unique()

tf_activity_heatmap$tf_design_activity[!is.finite(tf_activity_heatmap$tf_design_activity)] <- 0
tf_activity_heatmap <- dcast(tf_activity_heatmap, reporter ~ condition,
                             value.var="tf_design_activity")
tf_activity_heatmap$tf <- gsub("(^.*?)_[0-9]{1,2}bp.*", "\\1", tf_activity_heatmap$reporter)
tf_activity_heatmap <- tf_activity_heatmap %>% 
  remove_rownames %>% column_to_rownames(var="reporter")
# Keeping the scale in the pheatmap function
myBreaks1 <- seq(-3,3,0.06)

# pheatmap function
for(i in unique(tf_activity_heatmap$tf)){
  pheatmap(as.matrix(tf_activity_heatmap[tf_activity_heatmap$tf == i,-8]),
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks1, border_color = "#565656", 
         cellwidth = 30, cellheight = 10)
}

pheatmap(as.matrix(tf_activity_heatmap[tf_activity_heatmap$tf == "HNF1A",-8]),
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         breaks = myBreaks1, border_color = "#565656", 
         cellwidth = 30, cellheight = 10)
```




### Compute activity changes relative to their negative controls
```{r, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
## TF activity relative to negative controls 
cDNA_df_ctrl <- cDNA_df[cDNA_df$native_enhancer == "No" &
                          cDNA_df$hPGK == "No" & cDNA_df$tf != "RANDOM1" &
                          cDNA_df$tf != "RANDOM2" & cDNA_df$tf != "RANDOM3" & cDNA_df$library == "2" & cDNA_df$commercial_reporter == "No",] %>% 
  mutate(reporter_2 = gsub("_neg", "", reporter_id)) %>%
  dplyr::select(reporter_2, condition, reporter_activity, neg_ctrls) %>% 
  mutate(reporter_activity = ave(reporter_activity, reporter_2, condition,
                                 neg_ctrls, FUN = function(x) mean(x))) %>%
  unique()
  

## Divide TF activity by negative control activity for each reporter
for (i in unique(cDNA_df_ctrl$reporter_2)) {
  for (j in unique(cDNA_df_ctrl$condition)) {
    if (length(cDNA_df_ctrl$reporter_activity[cDNA_df_ctrl$reporter_2 == i & 
                                     cDNA_df_ctrl$neg_ctrls == "Yes" &
                                     cDNA_df_ctrl$condition == j]) != 0 &
       length(cDNA_df_ctrl$reporter_activity[cDNA_df_ctrl$reporter_2 == i & 
                                    cDNA_df_ctrl$neg_ctrls == "No" &
                                     cDNA_df_ctrl$condition == j]) != 0) { 
        
    cDNA_df_ctrl$reporter_activity_relative[cDNA_df_ctrl$reporter_2 == i & 
                                     cDNA_df_ctrl$condition == j] <- 
      
     cDNA_df_ctrl$reporter_activity[cDNA_df_ctrl$reporter_2 == i & 
                             cDNA_df_ctrl$neg_ctrls == "No" & 
                              cDNA_df_ctrl$condition == j] /
      
      cDNA_df_ctrl$reporter_activity[cDNA_df_ctrl$reporter_2 == i & 
                              cDNA_df_ctrl$neg_ctrls == "Yes" & 
                               cDNA_df_ctrl$condition == j]
   }
    else {
     cDNA_df_ctrl$reporter_activity_relative[cDNA_df_ctrl$reporter_2 == i & 
                                       cDNA_df_ctrl$condition == j] <- 0
    }
  }
}

## Log2 transformation
cDNA_df_ctrl <- cDNA_df_ctrl %>% 
  mutate(log_reporter_activity_relative = log2(reporter_activity_relative))
  
cDNA_df_ctrl$log_reporter_activity_relative[!is.finite(cDNA_df_ctrl$log_reporter_activity_relative)] <- 0

cDNA_df_ctrl$tf <- gsub("(^.*?)_[0-9]{1,2}bp.*", "\\1", cDNA_df_ctrl$reporter_2)

## Generate heatmaps with relative log2 activities 
# Caculate mean TF activity per condition
tf_activity_heatmap <- cDNA_df_ctrl[cDNA_df_ctrl$neg_ctrls == "No",] %>% 
  dplyr::select(reporter_2, tf, condition, log_reporter_activity_relative) %>% unique()
tf_activity_heatmap2 <- tf_activity_heatmap %>% dplyr::select(-reporter_2) %>%
  mutate(log_reporter_activity_relative = 
           ave(log_reporter_activity_relative, condition, tf, FUN = function(x) mean(x))) %>%
  unique()
tf_activity_heatmap2 <- tf_activity_heatmap2 %>% 
  dcast(tf ~ condition, value.var = "log_reporter_activity_relative")
tf_activity_heatmap2 <- tf_activity_heatmap2 %>% 
  remove_rownames %>% column_to_rownames(var="tf")
tf_activity_heatmap <- tf_activity_heatmap %>% 
  mutate(log_reporter_activity_relative = 
           ave(log_reporter_activity_relative, reporter_2, condition, FUN = function(x) mean(x))) %>%
  unique()
tf_activity_heatmap <- tf_activity_heatmap %>% 
  dcast(reporter_2+tf ~ condition, value.var = "log_reporter_activity_relative")
tf_activity_heatmap <- tf_activity_heatmap %>% 
  remove_rownames %>% column_to_rownames(var="reporter_2")
# Keeping the scale in the pheatmap function
myBreaks1 <- seq(0, 3.5, 0.035)


pheatmap(as.matrix(t(tf_activity_heatmap2)),
         color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
         breaks = myBreaks1, border_color = "#565656", 
         cellwidth = 12, cellheight = 10,
         angle_col = 45)

myBreaks1 <- seq(-2,6,0.08)

# pheatmap function
tf_activity_heatmap <- tf_activity_heatmap[,c(1,7,3,2,4,5,6,8)]
for(i in unique(tf_activity_heatmap$tf[tf_activity_heatmap$tf!="ESR1"&tf_activity_heatmap$tf!="IRX3"])){
  pheatmap(as.matrix(t(tf_activity_heatmap[tf_activity_heatmap$tf == i,-1])),
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         border_color = "black",
         breaks = myBreaks1, cluster_rows = F,
         cellwidth = 10, cellheight = 10)
}
  pheatmap(as.matrix(tf_activity_heatmap[tf_activity_heatmap$tf == "HNF1A",-1]),
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
         border_color = "black",
        breaks = myBreaks1, cluster_rows = F,
         cellwidth = 10, cellheight = 10)

```


```{r}
##plot log2 relative to neg reporter activity distribution per background
cDNA_df_ctrl <- cDNA_df_ctrl %>% mutate(background = gsub(".*([0-9]{1}$)", "\\1", cDNA_df_ctrl$reporter_2), promoter = gsub(".*(minP|mCMV|hBGm|Random)_.*", "\\1", reporter_2), spacing = gsub(".*_([0-9]{1,2}bp)_[0-9]{1,2}bp_.*", "\\1", reporter_2))
cDNA_df_ctrl$celltype <- "celltype"
cDNA_df_ctrl$celltype[grep("K562", cDNA_df_ctrl$condition)] <- "K"
cDNA_df_ctrl$celltype[grep("HepG2", cDNA_df_ctrl$condition)] <- "H"

ggplot(data = cDNA_df_ctrl[cDNA_df_ctrl$background != "" &
                             cDNA_df_ctrl$tf =="HNF1A",] %>% 
         dplyr::select(log_reporter_activity_relative, reporter_2, background, condition, promoter, spacing, celltype) %>% unique(), 
       aes(y = log_reporter_activity_relative, x = background, color = promoter, shape = spacing , size= celltype, alpha = 0.4)) + 
  geom_quasirandom() + xlab("log2(cDNA/pDNA) relative to neg_ctrl") + 
  labs(title = "reporter activity distribution - backgrounds", subtitle = "HNF1A") +
  scale_color_brewer(palette = "Set2") +
  theme_bw()
```



```{r SuperPlot, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# # Plot for each tf-activity per condition per biological replicate the techical replicate distribution and means
# ## Compute relevant mean and sd values
# activity_mean <- cDNA_df %>% 
#   dplyr::select(activity, condition, rep, TF, promoter, neg_ctrls, hPGK, native_enhancer, rand_promoter) %>% 
#   group_by(condition, rep, TF, promoter) %>% 
#   mutate(activity = mean(activity))
# 
# activity_mean <- activity_mean %>% 
#   group_by(condition, TF, promoter) %>% 
#   mutate(mean = mean(activity),
#          se = sd(activity)/sqrt(length(unique((activity)))))
# 
# activity_mean <- activity_mean %>% 
#   mutate(upper = mean + se, lower = mean - se)
# 
# 
# ## All TFs
# activity_mean_1 <- activity_mean[activity_mean$promoter == "minP" & 
#                         activity_mean$neg_ctrls == "No" &
#                         activity_mean$native_enhancer == "No",] %>% unique()
# 
# ## hPGK controls
# activity_mean_2 <- activity_mean[activity_mean$promoter == "hPGK" & 
#                         activity_mean$neg_ctrls == "No" & 
#                         activity_mean$native_enhancer == "No" ,] %>% unique()
# 
# ## Native enhancers
# activity_mean_3 <- activity_mean[activity_mean$promoter == "minP" & 
#                         activity_mean$native_enhancer == "Yes",] %>% unique()
# 
# 
# ## Random promoters
# activity_mean_4 <- activity_mean[activity_mean$promoter == "Random" & 
#                         activity_mean$neg_ctrls == "No" & 
#                         activity_mean$native_enhancer == "No",] %>% unique()
# 
# ## Negative controls
# activity_mean_5 <- activity_mean[activity_mean$promoter == "minP" & 
#                         activity_mean$neg_ctrls == "Yes" & 
#                         activity_mean$native_enhancer == "No",] %>% unique()
# 
# 
# ## Generate SuperPlot
# ### Selection 1 - All TFs
# for (i in 1:4) {
#   p <- ggplot(cDNA_df[cDNA_df$neg_ctrls == "No" &
#                         cDNA_df$native_enhancer == "No",], 
#               aes(x=condition, y=activity, color=factor(rep))) + 
#     scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71")) +
#     geom_quasirandom(aes(alpha = 0.9), dodge.width = 0.4) +
#     geom_errorbar(data = activity_mean_1, 
#                   aes(ymin = lower, 
#                       ymax = upper), colour = "black") + 
#     geom_quasirandom(data = activity_mean_1, aes(alpha = 1), size=4) + 
#     stat_compare_means(data=activity_mean_1, 
#                        comparisons = list(c(1:2)), 
#                    method="t.test", paired=TRUE) + 
#     labs(title = "TF activity distribution per condition", 
#          subtitle = paste("Page", i)) +
#     facet_wrap_paginate(~TF, ncol = 3, nrow = 3, page = i) +
#     
#     ylim(0,6) +
#     theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
#           axis.text.y = element_text(size = 10)) +
#     ylab("TF reporter activity") + xlab("") + 
#     theme(text = element_text(size = 12))
#   print(p)
# }
# 
# 
# 
# ### Selection 2 - hPGK controls
# ggplot(cDNA_df[cDNA_df$promoter == "hPGK" & 
#                         cDNA_df$neg_ctrls == "No" &
#                         cDNA_df$native_enhancer == "No",], 
#               aes(x=condition, y=activity, color=factor(rep))) + 
#     scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71")) +
#     geom_quasirandom(aes(alpha = 0.9), dodge.width = 0.4) +
#     geom_errorbar(data = activity_mean_2, 
#                   aes(ymin = lower, 
#                       ymax = upper), colour = "black") + 
#     geom_quasirandom(data = activity_mean_2, aes(alpha = 1), size=4) + 
#     stat_compare_means(data=activity_mean_2, 
#                        comparisons = list(c(1:2)), 
#                    method="t.test", paired=TRUE) + 
#     labs(title = "hPGK controls: activity distribution per condition", 
#          subtitle = p) +
#     
#     theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
#           axis.text.y = element_text(size = 10)) +
#     ylab("TF reporter activity") + xlab("") + 
#     theme(text = element_text(size = 12))
# 
# 
# 
# ### Selection 3 - Native enhancers
# for (i in 1:5) {
#   p <- ggplot(cDNA_df[cDNA_df$native_enhancer == "Yes",], 
#               aes(x=condition, y=activity, color=factor(rep))) + 
#     scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71")) +
#     geom_quasirandom(aes(alpha = 0.9), dodge.width = 0.4) +
#     geom_errorbar(data = activity_mean_3, 
#                   aes(ymin = lower, 
#                       ymax = upper), colour = "black") + 
#     geom_quasirandom(data = activity_mean_3, aes(alpha = 1), size=4) + 
#     stat_compare_means(data=activity_mean_3, 
#                        comparisons = list(c(1:2)), 
#                    method="t.test", paired=TRUE) + 
#     labs(title = "Miguels native enhancer chunks per condition", 
#          subtitle = paste("Page", i)) +
#     facet_wrap_paginate(~TF, ncol = 3, nrow = 3, page = i) +
#     ylim(0,6) +
#     
#     theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
#           axis.text.y = element_text(size = 10)) +
#     ylab("TF reporter activity") + xlab("") + 
#     theme(text = element_text(size = 12))
#   print(p)
# }
# 
# 
# ### Selection 5 - All TF-negative controls
# for (i in 1:3) {
#   p <- ggplot(cDNA_df[cDNA_df$promoter == "minP" & 
#                         cDNA_df$neg_ctrls == "Yes" &
#                         cDNA_df$native_enhancer == "No",], 
#               aes(x=condition, y=activity, color=factor(rep))) + 
#     scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71")) +
#     geom_quasirandom(aes(alpha = 0.9), dodge.width = 0.4) +
#     geom_errorbar(data = activity_mean_5, 
#                   aes(ymin = lower, 
#                       ymax = upper), colour = "black") + 
#     geom_quasirandom(data = activity_mean_5, aes(alpha = 1), size=4) + 
#     stat_compare_means(data=activity_mean_5, 
#                        comparisons = list(c(1:2)), 
#                    method="t.test", paired=TRUE) + 
#     labs(title = "TF activity distribution per condition - only negative controls", 
#          subtitle = paste("Page", i)) +
#     facet_wrap_paginate(~TF, ncol = 3, nrow = 3, page = i) +
#     
#     ylim(0,6) +
#     theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
#           axis.text.y = element_text(size = 10)) +
#     ylab("TF reporter activity") + xlab("") + 
#     theme(text = element_text(size = 12))
#   print(p)
# }
```


```{r superplot_2, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# # Same as before - now compare different designs
# ## Generate facet_wrap plot for each TF - each facet is a different design 
# ## Compute values for single and combinatorial designs
# 
# 
# ## Compute relevant mean and sd values
# activity_mean <- cDNA_df %>% 
#   dplyr::select(activity, condition, rep, TF, promoter, neg_ctrls, 
#          hPGK, native_enhancer, rand_promoter, reporter_id, spacing, 
#          distance, background) %>% 
#   group_by(condition, rep, reporter_id) %>% 
#   mutate(activity = mean(activity))
# 
# activity_mean <- activity_mean %>% 
#   group_by(condition, reporter_id) %>% 
#   mutate(mean = mean(activity),
#          se = sd(activity)/sqrt(length(unique((activity)))))
# 
# activity_mean <- activity_mean %>% 
#   mutate(upper = mean + se, lower = mean - se)
# 
# activity_mean <- activity_mean[activity_mean$neg_ctrls == "No" &
#                         activity_mean$native_enhancer == "No",] %>% unique()
# 
# 
# ## Generate SuperPlot
# ### Selection 1 - All Designs separately
# for (i in 1:4) {
#   p <- ggplot(cDNA_df[cDNA_df$TF == "Trp53" & 
#                         cDNA_df$neg_ctrls == "No" &
#                         cDNA_df$native_enhancer == "No",], 
#               aes(x=condition, y=activity, color=factor(rep))) + 
#     scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71")) +
#     geom_quasirandom(aes(alpha = 0.9), dodge.width = 0.4) +
#     geom_errorbar(data = activity_mean[activity_mean$TF == "Trp53",], 
#                   aes(ymin = lower, 
#                       ymax = upper), colour = "black") + 
#     geom_quasirandom(data = activity_mean[activity_mean$TF == "Trp53",], aes(alpha = 1), size=4) + 
#     stat_compare_means(data=activity_mean[activity_mean$TF == "Trp53",], 
#                        comparisons = list(c(1:2)), 
#                    method="t.test", paired=TRUE) + 
#     labs(title = "Trp53 activity distribution per design and condition", 
#          subtitle = paste("Page", i)) +
#     facet_wrap_paginate(~reporter_id, ncol = 3, nrow = 3, page = i) +
#     
#     ylim(0,16) +
#     theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
#           axis.text.y = element_text(size = 10)) +
#     ylab("TF reporter activity") + xlab("") + 
#     theme(text = element_text(size = 12))
#   print(p)
# }
# 
# 
# ### All reporters next to each other
# ggplot(cDNA_df[cDNA_df$TF == "Trp53" & 
#                         cDNA_df$neg_ctrls == "No" &
#                         cDNA_df$native_enhancer == "No",], 
#               aes(x=reporter_id, y=log_activity)) + 
#     scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71")) +
#     geom_boxplot() +
#     labs(title = "Trp53 activity distribution per design and condition") +
#     
#     ylim(-2,6) +
#     theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
#           axis.text.y = element_text(size = 10)) +
#     ylab("TF reporter activity") + xlab("") + 
#     theme(text = element_text(size = 12))
# 
# 
# 
# 
# 
# ## Subdivide in different designs 
# ### Spacing and distance
# cDNA_df2 <- cDNA_df %>% 
#   mutate(space_dist = paste(spacing, distance)) %>% 
#   dplyr::select(activity, condition, TF, space_dist, neg_ctrls, native_enhancer, promoter) 
# 
# cDNA_df2 <- cDNA_df2[cDNA_df2$neg_ctrls == "No" &
#                        cDNA_df2$native_enhancer == "No" & 
#                        cDNA_df2$promoter != "Random" &
#                        cDNA_df2$promoter != "hPGK",] %>% unique()
# 
# 
# 
# for(i in 1:8){
#   p <- ggplot(cDNA_df2, 
#         aes(x=condition, y=activity, color=space_dist)) + 
#       scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B")) +
#       geom_quasirandom(dodge.width = 0.9) +
#       geom_boxplot(position = position_dodge(0.9), aes(alpha = 0.5)) + 
#       labs(title = "Comparing spacings: TF activity distribution per condition", 
#            subtitle = paste("Page", i)) +
#       facet_wrap_paginate(~TF, nrow = 2, ncol = 2, page = i) +
#       
#       ylim(0,6) +
#       theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
#            axis.text.y = element_text(size = 10)) +
#       ylab("TF reporter activity") + xlab("") + 
#       theme(text = element_text(size = 12))
#   print(p)
# }
# 
# 
# 
# ### Background
# cDNA_df2 <- cDNA_df %>% 
#   dplyr::select(activity, condition, TF, background, neg_ctrls, native_enhancer, promoter) 
# 
# 
# cDNA_df2 <- cDNA_df2[cDNA_df2$neg_ctrls == "No" &
#                        cDNA_df2$native_enhancer == "No" & 
#                        cDNA_df2$promoter != "Random" &
#                        cDNA_df2$promoter != "hPGK",] %>% unique()
# 
# 
# 
# for(i in 1:8){
#   p <- ggplot(cDNA_df2, 
#         aes(x=condition, y=activity, color=background)) + 
#       scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B")) +
#       geom_quasirandom(dodge.width = 0.9) +
#       geom_boxplot(position = position_dodge(0.9), aes(alpha = 0.5)) + 
#       labs(title = "Comparing backgrounds: TF activity distribution per condition", 
#            subtitle = paste("Page", i)) +
#       facet_wrap_paginate(~TF, nrow = 2, ncol = 2, page = i) +
#       
#       ylim(0,6) +
#       theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
#            axis.text.y = element_text(size = 10)) +
#       ylab("TF reporter activity") + xlab("") + 
#       theme(text = element_text(size = 12))
#   print(p)
# }
# 
# 
# ### Promoter
# cDNA_df2 <- cDNA_df %>% 
#   dplyr::select(activity, condition, TF, promoter, neg_ctrls, native_enhancer, promoter) 
# 
# 
# cDNA_df2 <- cDNA_df2[cDNA_df2$neg_ctrls == "No" &
#                        cDNA_df2$native_enhancer == "No" &
#                        cDNA_df2$promoter != "hPGK",] %>% unique()
# 
# 
# 
# for(i in 1:8){
#   p <- ggplot(cDNA_df2, 
#         aes(x=condition, y=activity, color=promoter)) + 
#       scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B")) +
#       geom_quasirandom(dodge.width = 0.9) +
#       geom_boxplot(position = position_dodge(0.9), aes(alpha = 0.5)) + 
#       labs(title = "Comparing promoters: TF activity distribution per condition", 
#            subtitle = paste("Page", i)) +
#       facet_wrap_paginate(~TF, nrow = 2, ncol = 2, page = i) +
#       
#       ylim(0,6) +
#       theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
#            axis.text.y = element_text(size = 10)) +
#       ylab("TF reporter activity") + xlab("") + 
#       theme(text = element_text(size = 12))
#   print(p)
# }
# 
# 
# ### Promoter - native enhancers
# cDNA_df2 <- cDNA_df %>% 
#   dplyr::select(activity, condition, TF, background, neg_ctrls, native_enhancer, promoter) 
# 
# 
# cDNA_df2 <- cDNA_df2[cDNA_df2$neg_ctrls == "No" &
#                        cDNA_df2$native_enhancer == "Yes" &
#                        cDNA_df2$promoter != "hPGK",] %>% unique()
# 
# 
# 
# for(i in 1:10){
#   p <- ggplot(cDNA_df2, 
#         aes(x=condition, y=activity, color=promoter)) + 
#       scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B")) +
#       geom_quasirandom(dodge.width = 0.9) +
#       geom_boxplot(position = position_dodge(0.9), aes(alpha = 0.5)) + 
#       labs(title = "Miguels native enhancers per promoter", 
#            subtitle = paste("Page", i)) +
#       facet_wrap_paginate(~TF, nrow = 2, ncol = 2, page = i) +
#       
#       ylim(0,6) +
#       theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 10), 
#            axis.text.y = element_text(size = 10)) +
#       ylab("TF reporter activity") + xlab("") + 
#       theme(text = element_text(size = 12))
#   print(p)
# }
```








### Log-linear expression modelling to explain variance - model for each TF
```{r log-linear-model-3, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
### For each TF separately - only the non-mutated 

cDNA_df_TF <- cDNA_df[cDNA_df$native_enhancer == "No" &
                        cDNA_df$hPGK == "No" & 
                        cDNA_df$neg_ctrls == "No" &
                        cDNA_df$tf != "RANDOM1"&
                        cDNA_df$tf != "RANDOM2"&
                        cDNA_df$tf != "RANDOM3" &
                        cDNA_df$commercial_reporter == "No" &
                        cDNA_df$library == "2" &
                        cDNA_df$also_lib1 == "No" &
                        cDNA_df$condition == "HepG2_CDCA",] %>% 
  mutate(log_reporter_activity = ave(log_reporter_activity, reporter_id, FUN = function(x) mean(x))) %>%
  dplyr::select(log_reporter_activity, tf, promoter, spacing, distance, background) %>%
  unique()

# cDNA_df_rel <- cDNA_df_ctrl[cDNA_df_ctrl$neg_ctrls == "No",] %>% 
#   mutate(tf = gsub("(^.*)_[0-9]{1,2}bp_[0-9]{1,2}bp_.*", "\\1", reporter_2),
#          promoter = gsub(".*(minP|mCMV|hBGm|Random)_.*", "\\1", reporter_2),
#          distance = gsub(".*_([0-9]{1,2}bp)_.*", "\\1", reporter_2),
#          spacing = gsub(".*_([0-9]{1,2}bp)_[0-9]{1,2}bp_.*", "\\1", reporter_2),
#          background = gsub(".*([0-9]{1}$)", "\\1", reporter_2)) %>%
#   dplyr::select(condition, log_reporter_activity_relative, tf, promoter, distance, spacing, background)
# 
# cDNA_df_TF <- merge(cDNA_df_TF, cDNA_df_rel, all = T)

### Calculate expression variance
exp_variance <- data.frame(c("promoter", "spacing", "distance", "background", "Residuals"))
prediction <- data.frame(1:504)
names(prediction) <- "id"
names(exp_variance) <- "feature"
weight <- data.frame(c("(Intercept)", "promotermCMV", "promoterminP", "spacing5bp", "distance21bp", "background2", "background3"))
names(weight) <- "feature"


cDNA_df_TF$row <- rownames(cDNA_df_TF)


for (i in unique(cDNA_df_TF$tf[cDNA_df_TF$tf!="RFX1" & cDNA_df_TF$tf!="THRB" & cDNA_df_TF$tf!= "TCFCP2L1"])) {
  x <- lm(log_reporter_activity ~ promoter + spacing + distance + background, 
                     cDNA_df_TF[cDNA_df_TF$tf == i,])
  y <- data.frame(x$fitted.values) %>% rownames_to_column()
  par(mfrow=c(2,2))
  plot(x, main = i)
  names(y) <- c("row", i)
  y$id <- 1:nrow(y)
  prediction <- merge(prediction,y, all = T)
  
  w <- data.frame(x$coefficients) %>% rownames_to_column()
  names(w) <- c("feature", i)
  weight <- merge(weight,w, all = T)
  
  x <- anova(x) %>% rownames_to_column("feature") %>%
    setnames("Sum Sq", "sum_sq") %>%
    dplyr::select(feature, sum_sq) %>%
    mutate(sum = sum(sum_sq)) %>%
    mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
    setnames("rel_sum_sq", i)
  x <- x[,c(1,4)]
  
  exp_variance <- merge(exp_variance, x)
}


## Prepare lm parameters for plotting
exp_variance <- melt(exp_variance, variable.name = "tf")
exp_variance$group <- "no residual"
exp_variance$group[grep("Residual", exp_variance$feature)] <- "residual"
prediction <- prediction %>% dplyr::select(-id) %>% melt(variable.name = "tf") %>% na.omit() %>%
  dplyr::select(-tf) %>% setnames("value", "activity_predicted")
cDNA_df_TF <- merge(cDNA_df_TF, prediction)
weight <- weight %>% filter(feature != "(Intercept)")
weight <- melt(weight, variable.name = "tf")
categorie <- c("mutated", "spacing", "distance", "background")
weight$features <- gsub(paste(categorie, collapse="|"), "",weight$feature)
feature <- unique(weight$features)
weight$condition <- gsub(paste(feature, collapse="|"), "", weight$feature)
weight$value[is.na(weight$value)] <- 0


# Further plotting preparations
level_order <- c("mCMV", "2", "3", "10bp", "21bp", "5bp")


# Model parameters
exp_variance$value[exp_variance$feature == "Residuals"] <- 100 - exp_variance$value[exp_variance$feature == "Residuals"]
exp_variance$feature <- revalue(exp_variance$feature, c("Residuals"="total"))



ggplot(exp_variance, 
       aes(x = reorder(feature, -value), y = value, fill = group)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  geom_bar(stat = "identity") + 
  ylab("Variance explained (%)") + xlab("") + 
  labs(title = "Linear variance modelling", 
       subtitle = "Expression variance predicted by features") +
  
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12), 
        axis.text.y = element_text(size = 10)) +
    guides(fill = F) + facet_wrap(~tf)

ggscatter(cDNA_df_TF, x = "log_reporter_activity", y = "activity_predicted",
          add = "reg.line", 
          add.params = list(color = "blue", fill = "lightgray"),
          conf.int = TRUE, ylab = "Predicted log2-activity", size = 0.5,
          xlab = "Average log2-activity") +  facet_wrap(~tf) + 
  stat_cor(aes(label = ..r.label..), method = "pearson", label.x = -2, label.y = 4)



for (i in unique(cDNA_df_TF$tf)) {
  
  # Plot to show prediction accuracy
  sp <- ggscatter(cDNA_df_TF[cDNA_df_TF$tf == i,], x = "log_reporter_activity", y = "activity_predicted",
   add = "reg.line", 
   add.params = list(color = "blue", fill = "lightgray"),
   conf.int = TRUE, ylab = "Predicted log2-activity", 
   title = i,
   xlab = "Average log2-activity") + theme_bw()
  
  
  # Feature importance overview for each TF (including mean activities per feature)
  fp <- ggplot(exp_variance[exp_variance$tf == i,], 
       aes(x = reorder(feature, -value), y = value, fill = group)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  geom_bar(stat = "identity") + 
  ylab("Variance explained (%)") + xlab("") +
  ggtitle(i)+
  
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12), 
        axis.text.y = element_text(size = 10)) +
    guides(fill = F)
  
  # Plot weights of parameters in model - this helps to understand the model 
  p <- ggplot(weight[weight$tf == i,], 
       aes(x = factor(features, level = level_order), y = value, fill = condition)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B", "#C2B2B4", "#A54657")) +
  ggtitle(i)+
  geom_bar(stat = "identity") + 
  ylab("Weight") + xlab("") +
  
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
        axis.text.y = element_text(size = 10)) +
    guides(fill = F) 

  # Combine all plots
  sp.plot <- arrangeGrob(sp,fp,p, nrow = 2, ncol = 2)
  grid.arrange(sp.plot)
}
```


















### Log-linear expression modelling to explain variance - model for each TF - only WT 
```{r log-linear-model-4, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
### For each TF separately - only the non-mutated 
cDNA_df_TF <- cDNA_df[cDNA_df$native_enhancer == "No" &
                        cDNA_df$hPGK == "No" & 
                        cDNA_df$neg_ctrls == "No" &
                        cDNA_df$tf != "RANDOM1"&
                        cDNA_df$tf != "RANDOM2"&
                        cDNA_df$tf != "RANDOM3" &
                        cDNA_df$commercial_reporter == "No" &
                        cDNA_df$library == "2" &
                        cDNA_df$also_lib1 == "No" &
                        cDNA_df$condition == "HepG2_DMSO" & 
                        cDNA_df$promoter != "Random",] %>% 
  mutate(log_reporter_activity = ave(log_reporter_activity, reporter_id, FUN = function(x) mean(x))) %>%
  dplyr::select(log_reporter_activity, tf, promoter, spacing, distance, background) %>%
  unique()

cDNA_df_TF2 <- cDNA_df[cDNA_df$native_enhancer == "No" &
                        cDNA_df$hPGK == "No" & 
                        cDNA_df$neg_ctrls == "No" &
                        cDNA_df$tf != "RANDOM1"&
                        cDNA_df$tf != "RANDOM2"&
                        cDNA_df$tf != "RANDOM3" &
                        cDNA_df$commercial_reporter == "No" &
                        cDNA_df$library == "2" &
                        cDNA_df$also_lib1 == "No" &
                        cDNA_df$condition == "HepG2_DMSO" & 
                        cDNA_df$promoter != "Random",] %>% 
  mutate(log_reporter_activity = ave(log_reporter_activity, reporter_id, FUN = function(x) mean(x))) %>%
  dplyr::select(log_reporter_activity, tf, promoter, spacing, distance, background) %>%
  unique()


### Calculate expression variance
exp_variance <- data.frame(c("promoter", "spacing", "distance", "background", "Residuals"))
prediction <- data.frame(1:504)
names(prediction) <- "id"
names(exp_variance) <- "feature"
weight <- data.frame(c("(Intercept)","promoterminP", "spacing5bp", "distance21bp", "background2", "background3"))
names(weight) <- "feature"


cDNA_df_TF$row <- rownames(cDNA_df_TF)

for (i in unique(cDNA_df_TF$tf[cDNA_df_TF$tf != "TCFCP2L1"])) {
    x <- lm(log_reporter_activity ~ promoter + spacing + distance + background, 
                     cDNA_df_TF[cDNA_df_TF$tf == i,])
  y <- data.frame(x$fitted.values) %>% rownames_to_column()
  names(y) <- c("row", i)
  y$id <- 1:nrow(y)
  prediction <- merge(prediction,y, all = T)
  
  w <- data.frame(x$coefficients) %>% rownames_to_column()
  names(w) <- c("feature", i)
  weight <- merge(weight,w, all = T)
  
  x <- anova(x) %>% rownames_to_column("feature") %>%
    setnames("Sum Sq", "sum_sq") %>%
    dplyr::select(feature, sum_sq) %>%
    mutate(sum = sum(sum_sq)) %>%
    mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
    setnames("rel_sum_sq", i)
  x <- x[,c(1,4)]
  
  exp_variance <- merge(exp_variance, x)
}


## Fine-tune dfs
exp_variance <- melt(exp_variance, variable.name = "tf")
exp_variance$group <- "no residual"
exp_variance$group[grep("Residual", exp_variance$feature)] <- "residual"
prediction <- prediction %>% dplyr::select(-id) %>% melt(variable.name = "tf") %>% na.omit() %>%
  dplyr::select(-tf) %>% setnames("value", "activity_predicted")
cDNA_df_TF <- merge(cDNA_df_TF, prediction)
weight <- weight %>% filter(feature != "(Intercept)")
weight <- melt(weight, variable.name = "tf")
categorie <- c("promoter", "spacing", "distance", "background")
weight$features <- gsub(paste(categorie, collapse="|"), "",weight$feature)
feature <- weight$features
weight$condition <- gsub(paste(feature, collapse="|"), "", weight$feature)
weight <- weight %>% dplyr::select(-feature)



# Model parameters
exp_variance$value[exp_variance$feature == "Residuals"] <- 100 - exp_variance$value[exp_variance$feature == "Residuals"]
exp_variance$feature <- revalue(exp_variance$feature, c("Residuals"="total"))

##--------------------------------------------------------------------------------------
### Calculate expression variance 2
exp_variance2 <- data.frame(c("promoter", "spacing", "distance", "background", "Residuals"))
prediction2 <- data.frame(1:504)
names(prediction2) <- "id"
names(exp_variance2) <- "feature"
weight2 <- data.frame(c("(Intercept)","promoterminP", "spacing5bp", "distance21bp", "background2", "background3"))
names(weight2) <- "feature"


cDNA_df_TF2$row <- rownames(cDNA_df_TF2)

for (i in unique(cDNA_df_TF2$tf[cDNA_df_TF2$tf != "TCFCP2L1"])) {
    x <- lm(log_reporter_activity ~ promoter + spacing + distance + background, 
                     cDNA_df_TF2[cDNA_df_TF2$tf == i,])
  y <- data.frame(x$fitted.values) %>% rownames_to_column()
  names(y) <- c("row", i)
  y$id <- 1:nrow(y)
  prediction2 <- merge(prediction2,y, all = T)
  
  w <- data.frame(x$coefficients) %>% rownames_to_column()
  names(w) <- c("feature", i)
  weight2 <- merge(weight2,w, all = T)
  
  x <- anova(x) %>% rownames_to_column("feature") %>%
    setnames("Sum Sq", "sum_sq") %>%
    dplyr::select(feature, sum_sq) %>%
    mutate(sum = sum(sum_sq)) %>%
    mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
    setnames("rel_sum_sq", i)
  x <- x[,c(1,4)]
  
  exp_variance2 <- merge(exp_variance2, x)
}


## Fine-tune dfs 2
exp_variance2 <- melt(exp_variance2, variable.name = "tf")
exp_variance2$group <- "no residual"
exp_variance2$group[grep("Residual", exp_variance2$feature)] <- "residual"
prediction2 <- prediction2 %>% dplyr::select(-id) %>% melt(variable.name = "tf") %>% na.omit() %>%
  dplyr::select(-tf) %>% setnames("value", "activity_predicted")
cDNA_df_TF2 <- merge(cDNA_df_TF2, prediction2)
weight2 <- weight2 %>% filter(feature != "(Intercept)")
weight2 <- melt(weight2, variable.name = "tf")
categorie <- c("promoter", "spacing", "distance", "background")
weight2$features <- gsub(paste(categorie, collapse="|"), "",weight2$feature)
feature2 <- weight2$features
weight2$condition <- gsub(paste(feature, collapse="|"), "", weight2$feature)
weight2 <- weight2 %>% dplyr::select(-feature)



# Model parameters
exp_variance2$value[exp_variance2$feature == "Residuals"] <- 100 - exp_variance2$value[exp_variance2$feature == "Residuals"]
exp_variance2$feature <- revalue(exp_variance2$feature, c("Residuals"="total"))

---------------------------------------------------------------------------------------------------

ggplot(exp_variance, 
       aes(x = reorder(feature, -value), y = value, fill = group)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  geom_bar(stat = "identity") + 
  ylab("Variance explained (%)") + xlab("") + 
  labs(title = "Linear variance modelling", 
       subtitle = "Expression variance predicted by features") +
  
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12), 
        axis.text.y = element_text(size = 10)) +
  facet_wrap(~tf, ncol = 6)

##beeswarm plot

weight$group <- "K562"
weight2$group <- "HepG2"

#remove TFs with total variance explained <50%
weight<- weight[weight$tf != "FOSL1" & weight$tf != "HNF1A" & weight$tf != "RFX1" & weight$tf != "FOXA1" & weight$tf != "ETS2",]
weight2<- weight2[weight2$tf != "VDR::RXRA" & weight2$tf != "MYBL2" & weight2$tf != "SOX2" & weight2$tf != "FOXA1" & weight2$tf != "SOX9" & weight2$tf != "ESR1" & 
                    weight2$tf != "RUNX2" & weight2$tf != "MEF2A" & weight2$tf != "POU5F1" & weight2$tf != "NR1D1" & weight2$tf != "RFX1" & weight2$tf != "THRB",]

weight_K562_HepG2 <- rbind(weight, weight2)

ggplot(data = weight_K562_HepG2 %>% 
         dplyr::select(tf, value, features, condition, group ) %>% unique(), 
       aes(y = value, x = condition, color = group, alpha = 0.4)) + 
  geom_quasirandom(dodge.width = 0.75)  + 
  labs(title = "Log-linear expression modeling to explain variance") +
  scale_color_brewer(palette = "Set2") +
  theme_bw()


## Plot weights of parameters in model - this helps to understand the model 
level_order <- c("minP", "2", "3", "5bp", "21bp")





## Final overview - model fit for each TF
# Feature importance overview for each TF in a single plot (bar-chart including mean activities per feature)
#for (i in unique(cDNA_df_TF$TF)) {


  
  # Plot to show prediction accuracy
  sp <- ggscatter(cDNA_df_TF[cDNA_df_TF$tf == "SRF",], x = "log_reporter_activity", y = "activity_predicted",
   add = "reg.line", 
   add.params = list(color = "blue", fill = "lightgray"),
   conf.int = TRUE, ylab = "Predicted log2-activity", 
   title = "SRF",
   xlab = "Average log2-activity") + theme_bw()
  
  
  # Feature importance overview for each TF (including mean activities per feature)
  fp <- ggplot(exp_variance[exp_variance$tf == "SRF",], 
       aes(x = reorder(feature, -value), y = value, fill = group)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  geom_bar(stat = "identity") + 
  ylab("Variance explained (%)") + xlab("") +
  
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12), 
        axis.text.y = element_text(size = 10)) +
    guides(fill = F)
  
  # Plot weights of parameters in model - this helps to understand the model 
  p <- ggplot(weight[weight$tf == "SRF",], 
       aes(x = factor(features, level = level_order), y = value, fill = condition)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B", "#C2B2B4", "#A54657")) +
  geom_bar(stat = "identity") + 
  ylab("Weight") + xlab("") +
  
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
        axis.text.y = element_text(size = 10)) +
    guides(fill = F) 

  # Combine all plots
  sp.plot <- arrangeGrob(sp,fp,p, nrow = 1, ncol = 3)
  grid.arrange(sp.plot)



```




```{r tf-properties, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
# # Import tf properties dataframe
# tf_properties <- read.csv2("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/tf_properties.csv", header = T)
# tf_properties_lamb <- read.csv2("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/complete_TF_motif_space/lambert_filt.csv", header = T) %>% setnames("Name", "TF") %>% dplyr::select(TF, DBD)
# 
# tf_properties <- merge(tf_properties, tf_properties_lamb)
# 
# convertletter <- function(x) {
#   substr(x, 2, 20) <- tolower(substr(x, 2, 20))
#   x
# }
# 
# tf_properties$TF <- convertletter(tf_properties$TF)
# 
# tf_properties <- tf_properties[tf_properties$TF %in% cDNA_df$TF,]
# 
# variance <- exp_variance[exp_variance$feature == "total",] %>% dplyr::select(TF, value) %>% unique()
# 
# expression <- cDNA_df_TF %>% dplyr::select(log_reporter_activity_relative, TF) %>% 
#   mutate(log_reporter_activity_relative = ave(log_reporter_activity_relative, TF, FUN = function(x) mean(x))) %>%
#   unique()
# 
# tf_properties <- merge(tf_properties, variance)
# tf_properties <- merge(tf_properties, expression)
# 
# x1 <- ggplot(tf_properties, aes(x = chromatin_regulation_mode, y = value)) +
#   geom_quasirandom(width = 0.2) +
#   
#   xlab("Chromatin regulation mode") +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1.0, vjust = 0))+
#   ylab("TF reporter quality")
# 
# x2 <- ggplot(tf_properties, aes(x = chromatin_regulation_mode, y = log_reporter_activity_relative)) +
#   geom_quasirandom(width = 0.2) +
#   
#   xlab("Chromatin regulation mode") +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1.0, vjust = 0))+
#   ylab("TF reporter expression")
# 
# 
# #Grid the plots to one panel
# sp.plot <- arrangeGrob(x1,x2, nrow = 1, ncol = 2)
# grid.arrange(sp.plot)
# 
# 
# y1 <- ggplot(tf_properties, aes(x = TF_superclass, y = value)) +
#   geom_quasirandom(width = 0.2) +
#   
#   theme(axis.text.x = element_text(angle = 90, hjust = 1.0, vjust = 0))+
#   xlab("TF superclass") +
#   ylab("TF reporter quality")
# 
# y2 <- ggplot(tf_properties, aes(x = TF_superclass, y = log_reporter_activity_relative)) +
#   geom_quasirandom(width = 0.2) +
#   
#   theme(axis.text.x = element_text(angle = 90, hjust = 1.0, vjust = 0))+
#   xlab("TF superclass") +
#   ylab("TF reporter expression")
# 
# sp.plot <- arrangeGrob(y1,y2, nrow = 1, ncol = 2)
# grid.arrange(sp.plot)
# 
# 
# z1 <- ggplot(tf_properties, aes(x = mode_of_regulation, y = value)) +
#   geom_quasirandom(width = 0.2) +
#   
#   xlab("Mode of regulation") +
#   ylab("TF reporter quality")
# 
# z2 <- ggplot(tf_properties, aes(x = mode_of_regulation, y = log_reporter_activity_relative)) +
#   geom_quasirandom(width = 0.2) +
#   
#   xlab("Mode of regulation") +
#   ylab("TF reporter expression")
# 
# sp.plot <- arrangeGrob(z1,z2, nrow = 1, ncol = 2)
# grid.arrange(sp.plot)
# 
# 
# x4 <- ggplot(tf_properties, aes(x = tissue_of_expression, y = value)) +
#   geom_quasirandom(width = 0.2) +
#   
#   theme(axis.text.x = element_text(angle = 90, hjust = 1.0, vjust = 0))+
#   xlab("Tissue specificity") +
#   ylab("TF reporter quality")
# 
# x5 <- ggplot(tf_properties, aes(x = tissue_of_expression, y = log_reporter_activity_relative)) +
#   geom_quasirandom(width = 0.2) +
#   
#   theme(axis.text.x = element_text(angle = 90, hjust = 1.0, vjust = 0))+
#   xlab("Tissue specificity") +
#   ylab("TF reporter expression")
# 
# sp.plot <- arrangeGrob(x4,x5, nrow = 1, ncol = 2)
# grid.arrange(sp.plot)
# 
# 
# x6 <- ggplot(tf_properties, aes(x = DBD, y = value)) +
#   geom_quasirandom(width = 0.2) +
#   
#   theme(axis.text.x = element_text(angle = 90, hjust = 1.0, vjust = 0))+
#   xlab("DNA binding domain") +
#   ylab("TF reporter quality")
# 
# x7 <- ggplot(tf_properties, aes(x = DBD, y = log_reporter_activity_relative)) +
#   geom_quasirandom(width = 0.2) +
#   
#   theme(axis.text.x = element_text(angle = 90, hjust = 1.0, vjust = 0))+
#   xlab("DNA binding domain") +
#   ylab("TF reporter expression")
# 
# sp.plot <- arrangeGrob(x6,x7, nrow = 1, ncol = 2)
# grid.arrange(sp.plot)


```



### Make the same models as before - but now per TF and per condition
```{r log-linear-model-5, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
### For each TF separately - only the non-mutated 
cDNA_df_TF <- cDNA_df[cDNA_df$native_enhancer == "No" &
                        cDNA_df$hPGK == "No" & 
                        cDNA_df$neg_ctrls == "No" &
                        cDNA_df$tf != "RANDOM1"&
                        cDNA_df$tf != "RANDOM2"&
                        cDNA_df$tf != "RANDOM3" &
                        cDNA_df$library == "2" & 
                        cDNA_df$commercial_reporter == "No",] %>% 
  mutate(log_reporter_activity = ave(log_reporter_activity, reporter_id, condition, FUN = function(x) mean(x)),
         tf = paste(tf, condition, sep = "_")) %>%
  dplyr::select(log_reporter_activity, tf, promoter, spacing, distance, background) %>%
  unique()

# cDNA_df_rel <- cDNA_df_ctrl[cDNA_df_ctrl$neg_ctrls == "No",] %>% 
#   mutate(TF = gsub("(^.*)_[0-9]{1,2}bp_[0-9]{1,2}bp_.*", "\\1", reporter_2),
#          tf = paste(TF, condition, sep = "_"),
#          promoter = gsub(".*(minP|mCMV|hBGm|Random)_.*", "\\1", reporter_2),
#          distance = gsub(".*_([0-9]{1,2}bp)_.*", "\\1", reporter_2),
#          spacing = gsub(".*_([0-9]{1,2}bp)_[0-9]{1,2}bp_.*", "\\1", reporter_2),
#          background = gsub(".*([0-9]{1}$)", "\\1", reporter_2)) %>%
#   dplyr::select(log_reporter_activity_relative, tf, promoter, distance, spacing, background) %>% unique() %>%
#   mutate(log_reporter_activity_relative = ave(log_reporter_activity_relative, 
#                                               tf, promoter, distance, spacing, background, 
#                                               FUN = function(x) mean(x)))%>%
#   setnames("tf", "TF")
# 
# cDNA_df_TF <- merge(cDNA_df_TF, cDNA_df_rel, all = T)



### Calculate expression variance
exp_variance <- data.frame(c("promoter", "spacing", "distance", "background", "Residuals"))
prediction <- data.frame(1:504)
names(prediction) <- "id"
names(exp_variance) <- "feature"
weight <- data.frame(c("(Intercept)", "promotermCMV", "promoterminP", 
                  "spacing5bp", "distance21bp", "background2", "background3"))
names(weight) <- "feature"


cDNA_df_TF$row <- rownames(cDNA_df_TF)

for (i in unique(cDNA_df_TF$tf[-grep("Irx3|Homez|Nr3c1(GR)", cDNA_df_TF$tf)])) {
  cDNA_df_TF_2 <- cDNA_df_TF[-grep("Irx3|Homez|Nr3c1(GR)", cDNA_df_TF$tf),]
  x <- lm(log_reporter_activity ~ promoter + spacing + distance + background, 
                     cDNA_df_TF[cDNA_df_TF$tf == i,])
  y <- data.frame(x$fitted.values) %>% rownames_to_column()
  names(y) <- c("row", i)
  y$id <- 1:nrow(y)
  prediction <- merge(prediction,y, all = T)
  
  w <- data.frame(x$coefficients) %>% rownames_to_column()
  names(w) <- c("feature", i)
  weight <- merge(weight,w, all = T)
  
  x <- anova(x) %>% rownames_to_column("feature") %>%
    setnames("Sum Sq", "sum_sq") %>%
    dplyr::select(feature, sum_sq) %>%
    mutate(sum = sum(sum_sq)) %>%
    mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
    setnames("rel_sum_sq", i)
  x <- x[,c(1,4)]
  
  exp_variance <- merge(exp_variance, x)
}


## Fine-tune dfs
exp_variance <- melt(exp_variance, variable.name = "tf")
exp_variance$group <- "no residual"
exp_variance$group[grep("Residual", exp_variance$feature)] <- "residual"
prediction <- prediction %>% dplyr::select(-id) %>% melt(variable.name = "tf") %>% na.omit() %>%
  dplyr::select(-tf) %>% setnames("value", "activity_predicted")
cDNA_df_TF <- merge(cDNA_df_TF, prediction)
weight <- weight %>% filter(feature != "(Intercept)")
weight <- melt(weight, variable.name = "tf")
categorie <- c("mutated", "condition", "promoter", 
               "spacing", "distance", "background")
weight$features <- gsub(paste(categorie, collapse="|"), "",weight$feature)
ml <- length(weight$features)
feature <- weight$features
weight$condition <- gsub(paste(head(feature, n = ml/2), collapse = "|"), "", weight$feature)
weight <- weight %>% dplyr::select(-feature)



# Model parameters
exp_variance$value[exp_variance$feature == "Residuals"] <- 100 - exp_variance$value[exp_variance$feature == "Residuals"]
exp_variance$feature <- revalue(exp_variance$feature, c("Residuals"="total"))

ggplot(exp_variance[grep("Trp53", exp_variance$TF),], 
       aes(x = reorder(feature, -value), y = value, fill = group)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  geom_bar(stat = "identity") + 
  ylab("Variance explained (%)") + xlab("") + 
  labs(title = "Linear variance modelling", 
       subtitle = "Expression variance predicted by features") +
  
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12), 
        axis.text.y = element_text(size = 10)) +
  facet_wrap(~TF)




## Plot weights of parameters in model - this helps to understand the model 
level_order <- c("mCMV", "minP","Random", 
                 "2", "3", "5bp", "21bp")





## Final overview - model fit for each TF
# Feature importance overview for each TF in a single plot (bar-chart including mean activities per feature)


for (i in unique(cDNA_df_TF$TF[grep("Trp53", cDNA_df_TF$TF)])) {
  
  # Plot to show prediction accuracy
  sp <- ggscatter(cDNA_df_TF[cDNA_df_TF$TF == i,], x = "log_reporter_activity", y = "activity_predicted",
   add = "reg.line", 
   add.params = list(color = "blue", fill = "lightgray"),
   conf.int = TRUE, ylab = "Predicted log2-activity", 
   title = i,
   xlab = "Average log2-activity") + theme_bw()
  
  
  # Feature importance overview for each TF (including mean activities per feature)
  fp <- ggplot(exp_variance[exp_variance$TF == i,], 
       aes(x = reorder(feature, -value), y = value, fill = group)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  geom_bar(stat = "identity") + 
  ylab("Variance explained (%)") + xlab("") +
  
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12), 
        axis.text.y = element_text(size = 10)) +
    guides(fill = F)
  
  # Plot weights of parameters in model - this helps to understand the model 
  p <- ggplot(weight[weight$TF == i,], 
       aes(x = factor(features, level = level_order), y = value, fill = condition)) +
  scale_fill_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B", "#C2B2B4", "#A54657")) +
  geom_bar(stat = "identity") + 
  ylab("Weight") + xlab("") +
  
  theme(text = element_text(size = 12)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10), 
        axis.text.y = element_text(size = 10)) 

  # Combine all plots
  sp.plot <- arrangeGrob(sp,fp,p, nrow = 2, ncol = 2)
  grid.arrange(sp.plot)
}
```

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

