---
title: "TF activity computation & feature analysis - TF reporter library 90 TFs - combined"
author: 
  - name: "Max Trauernicht"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute - van Steensel lab"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
---

---

### Aim
In more than 100 samples, 35,500 reporters for 86 TFs were transfected into various cells (mainly mES, K562 & HepG2), and stimulated with various compounds. To be more precise, the data consists of two different libraries, library 1, which is more focused on mES biology, and library 2, which is more focused on classical signaling pathways. All the barcode counts from the 7 different sequencing runs are combined in this analysis. In a previous script (mt20211021_barcode_preprocessing... .Rmd), sanity checks were be carried out to filter out samples with bad quality, and barcode counts were be normalized to compute reporter activitites. In this script, plots will be generated to characterize the TF reporter activities and models will be built to extract the importance of reporter features. 


### Description of input data
- library = with which library was this sample transfected (1, 2, or 1+2)
- nchar = barcode length (12 for library 1, 13 for library 2)
- commercial_reporter = positive control reporter sequence from commercial source
- neg_ctrls = reporter with mutated TF binding sites?
- hPGK = positive control, chunk from hPGK promoter (only in lib 1)
- native_enhancer = genomic mES enhancer sequence chunks from Miguels libraries (only lib 1)
- activity = rpm + pseudocount cDNA / rpm + pseudocount pDNA
- mean_activity_sample = mean of activity per unique reporter per sample (individually per replicate)
- reporter_activity = mean activity per unique reporter per condition (mean of replicates)
- minP_activity = activity relative to mean of minimal promoter only reporters of that sample
- mean_activity_sample_minP = mean of minP_activity per unique reporter 
- reporter_activity_minP = mean of mean_activity_sample_minP per condition
- nfya_activity = activity relative to mean of nfya reporters (constitutive positive control) of that sample
- mean_activity_sample_nfya = mean of nfya_activity per unique reporter 
- reporter_activity_nfya = mean of mean_activity_sample_nfya per condition
- mean_activity_sample_neg = mean_activity_sample relative to its paired negative control (identical reporter with mutated binding sites)
- reporter_activity_neg = mean of mean_activity_sample_neg per condition
- pval_minP = p-value of t-test comparing activity distributions of individual reporters with the distribution of minimal promoter only reporters within the same condition
- pval_neg = p-value of t-test comparing activity distributions of individual reporters with the distribution of the activities of its paired negative controls within the same condition


### Main tested conditions:
- mESC: 2i+LIF
- mESC: 2i-LIF (STAT3 inactivation)
- mESC: LIF+PD (TCF inactivation)
- mESC: LIF+CH (MEK activation)
- mESC: N2B27 (neural differentiation medium)
- mESC: HQ (NRF activation)
- NPC
- HepG2
- HepG2 + bile acids (NR activation)
- K562

*Data from TFs in library 1 in HepG2 cells was ignored for a large part of the analysis for now because the computed TF activities did not make sense, I will look into what happened there further*

---

## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}
```

### Libraries 

```{r setup, out.width= "80%", fig.align= "center", warning = FALSE, message= FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(RColorBrewer)
library(ggplot2)
library(dplyr)
library(maditr)
library(tibble)
library(pheatmap)
library(ggpubr)
library(ggbeeswarm)
library(ggforce)
library(viridis)
library(plyr)
library(cowplot)
library(gridExtra)
library(pROC)
library(tidyr)
library(stringr)
library(randomForest)
library(readr)
```

### Functions

```{r out.width= "80%", fig.align= "center"}
### Custom functions
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}


# Set custom ggplot2 theme and custom colors
theme_classic_lines <- function() {
  theme_pubr(border = F, legend = "top") +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_classic_lines_45 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 45) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_classic_lines_90 <- function() {
  theme_pubr(border = T, legend = "top", x.text.angle = 90) +
            theme(panel.grid.major = element_line(colour = "#adb5bd", size = 0.25),
                  strip.background = element_rect(fill = "#ced4da")
            )
}

theme_set(theme_classic_lines())

colors_diverse <- c("#264653", "#9AC1AE", "#5D987B", "#f2cc8f", "#e76f51")

ggplot_custom <- function(...) ggplot2::ggplot(...) + 
  scale_color_manual(values = colors_diverse) + 
  scale_fill_manual(values = colors_diverse)
```

### Loading data

```{r out.width= "80%", fig.align= "center"}
# Import processed bc counts from the preprocessing step
cDNA_df <- read.csv("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6578_gen2_stimulation-2/results/mt20211118_reporter_activity_filt_combined.csv", header = T) %>%
  dplyr::select(-tf_activity, -reporter_id2, 'pval_minP' = pval, -pval_adj, -mutated_activity)

# Filter TP53 from library 1+2 K562 data (we previously determined that we should filter these data)
cDNA_df2 <- cDNA_df %>%
  filter(library == "1+2" & condition == "K562_DMSO" & tf == "TP53")
cDNA_df <- anti_join(cDNA_df, cDNA_df2, by = c("reporter_id", "sample_id"))
```

# Plotting reporter activities

## Reporter activity compared to background

*Display reporter activities relative to background activity per tf and condition*

```{r reporter_activities, out.width= "80%", fig.align= "center"}
# Calculate fdr from the previously calculated p-values to correct for multiple hypotheses
cDNA_df$pval_adj_minP <- p.adjust(cDNA_df$pval_minP, method = 'fdr')

# Plot the reporter activities over the minimal promoter only and the adjusted p-value for each reporter
for (i in unique(cDNA_df$condition)) {
  
  p <- ggplot_custom(cDNA_df %>%
                filter(condition == i, neg_ctrls == "No") %>%
                dplyr::select(reporter_id, pval_adj_minP, reporter_activity_minP, promoter) %>%
                unique(),
            aes(x = log2(reporter_activity_minP), y = -log10(pval_adj_minP),
                color = ifelse(log2(reporter_activity_minP) > 1 & pval_adj_minP < 0.05, "green", 
                               ifelse(log2(reporter_activity_minP) < -1 & pval_adj_minP < 0.05, "yellow","black")))) +
    geom_point() +
    geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
    geom_vline(xintercept = log2(2), linetype = "dashed") +
    geom_vline(xintercept = -log2(2), linetype = "dashed") +
    facet_wrap(~promoter) +
    ggtitle(paste('reporter activity over minimal promoter only, condition:', i))
  
  print(p)
}


# Same plot for one condition only
ggplot_custom(cDNA_df %>%
                filter(condition == "HepG2_CDCA", neg_ctrls == "No") %>%
                dplyr::select(reporter_id, pval_adj_minP, reporter_activity_minP, promoter) %>%
                unique() %>%
                mutate(reporter_activity_minP = log2(reporter_activity_minP),
                       pval_adj_minP = -log10(pval_adj_minP)),
            aes(x = reporter_activity_minP, y = pval_adj_minP, 
                color = ifelse(reporter_activity_minP > 1 & pval_adj_minP > -log10(0.05), ifelse(reporter_activity_minP > 3 & pval_adj_minP > -log10(0.05), "green", "blue"), 
                               ifelse(reporter_activity_minP < -1 & pval_adj_minP > -log10(0.05), "yellow","black")))) +
  geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  geom_vline(xintercept = log2(2), linetype = "dashed") +
  geom_vline(xintercept = -log2(2), linetype = "dashed") +
  facet_wrap(~promoter) + 
  ggtitle('reporter activity over minimal promoter only, HepG2_CDCA')


# Same plot for one condition, but now plotting reporter_activity_neg and its significance
ggplot_custom(cDNA_df %>%
                filter(condition == "HepG2_CDCA", neg_ctrls == "No", library == "1+2") %>%
                dplyr::select(reporter_id, pval_neg, reporter_activity_neg, promoter) %>%
                mutate(pval_neg = p.adjust(pval_neg, method = 'fdr')) %>%
                unique(),
            aes(x = log2(reporter_activity_neg), y = -log10(pval_neg), 
                color = ifelse(log2(reporter_activity_neg) > 1 & pval_neg < 0.05, "green", 
                               ifelse(log2(reporter_activity_neg) < -1 & pval_neg < 0.05, "yellow","black")))) +
  geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  geom_vline(xintercept = log2(2), linetype = "dashed") +
  geom_vline(xintercept = -log2(2), linetype = "dashed") +
  facet_wrap(~promoter) + 
  ggtitle('reporter activity over paired negative control, HepG2_CDCA')
```

---

## Calculate fraction of active/repressive reporters

```{r reporter_fraction, out.width= "80%", fig.align= "center"}
# Stratify reporters by their activity
cDNA_df$reporter_activity_class <- 0
cDNA_df$reporter_activity_class[cDNA_df$pval_adj_minP < 0.05 & cDNA_df$reporter_activity_minP > 2] <- 1 # significantly active = significant 2-fold increase
cDNA_df$reporter_activity_class[cDNA_df$pval_adj_minP < 0.05 & cDNA_df$reporter_activity_minP > 8] <- 2 # significantly highly active = significant 8-fold increase
cDNA_df$reporter_activity_class[cDNA_df$pval_adj_minP < 0.05 & cDNA_df$reporter_activity_minP < 0.5] <- -1 # significantly repressive = significant 2-fold decrease

## Sub-select relevant data
cDNA_df2 <- cDNA_df %>%
                filter(neg_ctrls == "No", commercial_reporter == "No") %>%
                dplyr::select(reporter_id, tf, reporter_activity_class, condition, promoter) %>%
                unique()

# Calculate per condition how many reporters per TF are upregulated, downregulated, or unchanged
cDNA_df2 <- cDNA_df2 %>%
  mutate(class_count = ave(reporter_activity_class, tf, promoter, condition, FUN = function(x) sum(x)))


for (i in unique(cDNA_df2$condition)) {
  p <- ggplot_custom(cDNA_df2 %>%
                       filter(condition == i),
                     aes(x = reorder(tf, -class_count), fill = factor(reporter_activity_class, levels = c("0", "1", "2", "-1")))) +
    geom_bar(position = "fill") +
    geom_hline(yintercept = 0.25, linetype = "dashed") +
    geom_hline(yintercept = 0.75, linetype = "dashed") +
    theme_classic_lines_90() +
    facet_wrap(~promoter) +
    xlab("TF") +
    ggtitle(paste('fraction of active/repressive reporters, condition:',i))
  
  print(p)
}


## Same plot but for one condition
ggplot_custom(cDNA_df2 %>%
                       filter(condition == "mES_2i_LIF", promoter == "mCMV"),
                     aes(x = reorder(tf, -class_count), fill = factor(reporter_activity_class, levels = c("0", "1", "2", "-1")))) +
    geom_bar(position = "fill") +
    geom_hline(yintercept = 0.25, linetype = "dashed") +
    geom_hline(yintercept = 0.75, linetype = "dashed") +
    theme_classic_lines_90()+
    xlab("TF") +
    ggtitle('fraction of active/repressive reporters, mES_2i_LIF - mCMV')
```


---

## Fraction of active reporters per TF

*How many reporters are classified as active/repressive per tf and condition?*

```{r tf_activity_reporter_fraction, out.width= "80%", fig.align= "center"}
# Divide TFs per condition into different classes: activators, selective activator, inactive, selective repressor, repressor

## Determine amount of active/repressive reporters per tf
cDNA_df3 <- cDNA_df2 %>%
  mutate(class_count2 = ave(reporter_activity_class, tf, condition, FUN = function(x) length(x)),
         class_count_rel = ave(reporter_activity_class, reporter_activity_class, tf, condition, FUN = function(x) length(x))) %>%
  mutate(class_count_rel = class_count_rel / class_count2) %>%
  dplyr::select(tf, condition, class_count_rel, reporter_activity_class) %>%
  unique() %>%
  spread(reporter_activity_class, class_count_rel)

cDNA_df3$class[cDNA_df3$`0` >= 0.75] <- 0 # inactive = >75% of the reporters are inactive
cDNA_df3$class[cDNA_df3$`1` >= 0.75] <- 2 # active = >75% of the reporters are active
cDNA_df3$class[cDNA_df3$`2` >= 0.75] <- 4 # highly active = >75% of the reporters are highly active
cDNA_df3$class[cDNA_df3$`-1` >= 0.75] <- -4 # repressive = >75% of the reporters are repressive
cDNA_df3$class[cDNA_df3$`1` >= 0.25 & cDNA_df3$`1` < 0.75] <- 1 # partly active = >25% & <75% of the reporters are active
cDNA_df3$class[cDNA_df3$`1` < 0.25 & cDNA_df3$`2` < 0.25 & cDNA_df3$`1` + cDNA_df3$`2` >= 0.25] <- 1 # partly active = >25% of the reporters are active or highly active
cDNA_df3$class[cDNA_df3$`2` >= 0.25 & cDNA_df3$`2` < 0.75] <- 3 # partly highly active = >25% of the reporters are highly active
cDNA_df3$class[cDNA_df3$`-1` >= 0.25 & cDNA_df3$`-1` < 0.75] <- -2 # partly repressive = >25% & <75% of the reporters are repressive
cDNA_df3$class[cDNA_df3$`1` >= 0.085 & cDNA_df3$`-1` >= 0.085] <- -1 # mixed active/repressive = > 8.5% are active and >8.5% are repressive


## Sub-select relevant columns, for now only focus on the conditions where I have robust data from both libraries
cDNA_df4 <- cDNA_df3 %>%
  dplyr::select(tf, condition, class) %>%
  unique() %>%
  na.omit() %>%
  filter(condition %in% c("HepG2_CDCA", "HepG2_DMSO", "K562_DMSO", "mES_2i", 
                          "mES_2i_LIF", "mES_HQ", "mES_LIF_CH", "mES_LIF_PD", 
                          "mES_N2B27")) %>%
  mutate(sum = ave(class, tf, FUN = function(x) sum(x))) %>%
  filter(sum != 0) %>%
  dplyr::select(-sum)

cDNA_df4 <- cDNA_df4 %>%
  spread(tf, class) %>%
  column_to_rownames('condition')

## Make custom color palette
makeColorRampPalette <- function(colors, cutoff.fraction, num.colors.in.palette)
{
  stopifnot(length(colors) == 4)
  ramp1 <- colorRampPalette(colors[1:2])(num.colors.in.palette * cutoff.fraction)
  ramp2 <- colorRampPalette(colors[2:4])(num.colors.in.palette * (1 - cutoff.fraction))
  return(c(ramp1, ramp2))
}

cols <- makeColorRampPalette(c(colors_diverse[4], colors_diverse[1], colors_diverse[2], colors_diverse[3]), 
                             1-(4/6),
                             100)
## Plot heatmap showing fraction of active reporters
pheatmap(as.matrix(cDNA_df4),
         color = cols,
         border_color = "white", 
         cellwidth = 8, cellheight = 8,
         angle_col = 90,
         main = "fraction of active/repressive reporters per tf")
```

---

# Reporter feature modeling

## Linear model to fit classifications

*Why are the reporters classified as repressive/activating? Fit linear model to reporter classification*

```{r linear_model_classes, out.width= "80%", fig.align= "center"}
# Sub-select relevant data
cDNA_df_lm <- cDNA_df %>%
  filter(neg_ctrls == "No", commercial_reporter == "No", promoter != "Random",
         condition %in% c("HepG2_CDCA", "HepG2_DMSO", "K562_DMSO", "mES_2i", 
                          "mES_2i_LIF", "mES_HQ", "mES_LIF_CH", "mES_LIF_PD", 
                          "mES_N2B27"
                          ),
         tf %in% names(cDNA_df4)) %>%
  mutate(background = as.factor(background)) %>%
  dplyr::select(reporter_id, tf, reporter_activity_class, condition, promoter, spacing, background, distance) %>%
  unique()

cDNA_df3[is.na(cDNA_df3)] <- 0

# Only select tfs for which differences in reporter activity were observed (for the other ones it doesn't make sense to fit models)
mixed_classes <- cDNA_df3 %>%
  mutate(id = paste(condition, tf, sep = "_")) %>%
  dplyr::select(-class) %>%
  #filter(class != 0) %>%
  unique() %>%
  filter(`1` < 0.95, `0` < 0.95, `2` < 0.95)

# Select only relevant tfs from source data frame and only include ids with more than two entries
cDNA_df_lm$id <- paste(cDNA_df_lm$condition, cDNA_df_lm$tf, sep = "_")
cDNA_df_lm <- cDNA_df_lm %>%
  mutate(sum = ave(id, id, FUN = function(x) length(x))) %>%
  filter(sum >= 2, id %in% mixed_classes$id) %>%
  dplyr::select(-sum)


# Fit linear model for each id (combination of condition and tf)
lm_features <- lapply(unique(cDNA_df_lm$id), function(x) lm(reporter_activity_class ~ promoter + spacing + background + distance, 
                                                                     cDNA_df_lm[cDNA_df_lm$id == x,]))

names(lm_features) <- unique(cDNA_df_lm$id)


## Extract and plot weights from linear model
lm_weights <- list()
for (i in names(lm_features)) {
  lm_weights[[i]] <- lm_features[[i]]$coefficients
}

lm_weights <- bind_rows(lm_weights, .id = "id")

lm_weights <- lm_weights %>%
  dplyr::select(-`(Intercept)`) %>%
  pivot_longer(-id, names_to = "feature", values_to = "weight")

ggplot_custom(lm_weights, aes(x = feature, y = weight)) +
  geom_quasirandom() +
  theme_classic_lines_90() +
  ggtitle('weights of all models')


## Extract and plot features from linear model
exp_features <- list()
for (i in names(lm_features)) {
  exp_features[[i]] <- anova(lm_features[[i]]) %>%
  dplyr::select('sum_sq' = `Sum Sq`) %>%
    na.omit() %>%
  mutate(sum = sum(sum_sq)) %>%
  mutate(rel_sum_sq = (sum_sq/sum)*100) %>%
  setnames("rel_sum_sq", "percent") %>%
  dplyr::select("percent")
}
exp_features <- bind_rows(exp_features, .id = "id") %>%
  rownames_to_column("feature") %>%
  mutate(feature = gsub("...[0-9]{1-3}", "", feature)) %>%
  mutate(feature = gsub("[0-9]", "", feature)) %>%
  mutate(percent = format(percent, scientific = F, digits = 0)) %>%
  mutate(percent = as.numeric(percent))

exp_features$group <- "no residual"
exp_features$group[grep("Residual", exp_features$feature)] <- "residual"
exp_features$percent[exp_features$feature == "Residuals"] <- 100 - exp_features$percent[exp_features$feature == "Residuals"]
exp_features$feature <- revalue(exp_features$feature, c("Residuals"="total"))

ggplot_custom(exp_features, 
       aes(x = percent, y = reorder(feature, -percent), fill = group)) +
  geom_bar(stat = "identity") + 
  xlab("Variance explained (%)") + 
  ylab("") +
  ggtitle('sum of features of all models')

## Plot features separately per tf activity class
classes <- cDNA_df3 %>%
  mutate(id = paste(condition, tf, sep = "_")) %>%
  dplyr::select(id, class) %>%
  unique()

exp_features <- merge(exp_features, classes, by = "id")

ggplot_custom(exp_features, 
       aes(y = percent, x = feature, color = group)) +
  geom_quasirandom() + 
  xlab("Variance explained (%)") + 
  ylab("") +
  facet_wrap(~class) +
  theme_classic_lines_90() +
  ggtitle('features of all models per tf activity class')


## Discard poor performing linear models (<50% of variance explained)
exp_features_good <- exp_features %>%
  filter(group == "residual", percent >= 50)
exp_features_sel <- exp_features[exp_features$id %in% exp_features_good$id,]

exp_features_sel <- exp_features_sel %>%
  mutate(tf = gsub(".*_(.*)", "\\1", id),
         condition = gsub("(.*)_.*", "\\1", id))

## Plot again the features
ggplot_custom(exp_features_sel, 
       aes(y = percent, x = feature, color = group)) +
  geom_quasirandom() + 
  xlab("Variance explained (%)") + 
  ylab("") +
  ggtitle('features of all models - only good models')

## Plot features individually per id
### Order the ids
order <- exp_features_sel %>%
                mutate(sum = ave(percent, id, FUN = function(x) sum(x))) %>%
  dplyr::select(sum, tf) %>%
  unique() %>%
  group_by(tf) %>%
  top_n(1, sum)

exp_features_sel_ord <- merge(order, exp_features_sel, by = "tf") %>%
  filter(feature != "total")
  

### As bar plot
ggplot(exp_features_sel_ord %>%
         filter(condition %in% c("K562_DMSO", "HepG2_DMSO", "mES_2i_LIF")), 
       aes(y = percent, x = reorder(id, -sum), fill = feature, group = feature)) +
  geom_bar(stat = "identity", position = "stack") + 
  xlab("Variance explained (%)") + 
  ylab("") +
  scale_fill_manual(values = c("#e07a5f", "#3d405b", "#81b29a", "#f2cc8f"))+
  theme_classic_lines_90() +
  ggtitle('feature importance per tf & condition')

### As connected line-plot
ggplot_custom(exp_features_sel, aes(x=feature, y=percent, color=group, group = id)) + 
  geom_line() +
  geom_point()+
  ggtitle('feature importance per tf & condition')


### Only select the three main cell lines to simplify
exp_features_sel_heatmap <- exp_features_sel %>%
  filter(condition %in% c("mES_2i_LIF", "K562_DMSO", "HepG2_DMSO")) %>%
  mutate(cell_type = gsub("(.*)_.*", "\\1", condition)) %>%
  mutate(cell_type = gsub("(.*)_.*", "\\1", cell_type)) %>%
  dplyr::select(-group, -condition) %>%
  filter(feature != "total") %>%
  spread(feature, percent) %>% 
  column_to_rownames("id") #%>%
  #arrange(tf)

cell_type <- exp_features_sel_heatmap %>%
  dplyr::select(cell_type)

colors_pheatmap <- list('cell_type' = c("K562" = colors_diverse[1], "HepG2" = colors_diverse[5], "mES" = colors_diverse[4]))

pheatmap(as.matrix(t(exp_features_sel_heatmap %>% dplyr::select(-cell_type, -tf, -class))),
         color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
         border_color = "white", 
         cellwidth = 8, cellheight = 8,
         angle_col = 90, annotation_col = cell_type, cluster_cols = T,
         annotation_colors = colors_pheatmap,
         main = 'feature importance per tf & condition')
```

---

## Comparison of TF reporters between conditions

*Can we select the best reporter for each TF?*

```{r reporter_correlations, out.width= "80%", fig.align= "center"}
# Select conditions for which I have robust data, remove highly active reporters that I took over from library 1 and included in library 2 (because those distort the activity distributions)
cDNA_df2 <- cDNA_df %>%
  filter(condition %in% c("HepG2_DMSO", "K562_DMSO", "mES_2i_LIF", "mES_2i", "mES_N2B27", "mES_LIF_PD", "mES_LIF_CH", "mES_HQ"), library != "1+2")

tfs_lib1 <- cDNA_df %>%
  filter(library == "1") %>%
  dplyr::select(tf) %>% 
  unique()
  
cDNA_df3 <- cDNA_df %>%
  filter(condition %in% c("HepG2_DMSO", "K562_DMSO", "mES_2i_LIF", "mES_2i", "mES_N2B27", "mES_LIF_PD", "mES_LIF_CH", "mES_HQ"), library == "1+2")

cDNA_df3 <- cDNA_df3[!cDNA_df3$tf %in% tfs_lib1$tf,]

cDNA_df<- rbind(cDNA_df2, cDNA_df3)


## Compare the activities of TF reporters in different conditions
### Select TFs for which I have good models
reporter_correlations <- cDNA_df %>%
  mutate(id = paste(condition, tf, sep = "_")) %>%
  filter(id %in% cDNA_df_lm$id) %>%
  dplyr::select(reporter_id, condition, tf, reporter_activity_minP) %>%
  unique() %>%
  spread(condition, reporter_activity_minP)

ggplot_custom(reporter_correlations %>% 
                dplyr::select(-mES_2i_LIF) %>% 
                na.omit(), aes(x = log2(K562_DMSO), y = log2(HepG2_DMSO))) +
  geom_point() +
  facet_wrap(~tf) +
  geom_abline(linetype = "dashed") +
  geom_smooth(method = "lm", color = "grey", fill = "lightgrey") +
  ggtitle('comparison of reporter activities between conditions')


### Or for a single case: STAT3 upon JAK activation
reporter_correlations <- cDNA_df %>%
  dplyr::select(reporter_id, condition, tf, reporter_activity_minP) %>%
  unique() %>%
  spread(condition, reporter_activity_minP)

ggplot_custom(reporter_correlations %>% 
                filter(tf == "STAT3") %>%
                dplyr::select(mES_2i_LIF, mES_2i, tf, reporter_id) %>% 
                na.omit(), aes(x = log2(mES_2i_LIF), y = log2(mES_2i))) +
  geom_point() +
  geom_abline(linetype = "dashed") +
  geom_smooth(method = "lm", color = "grey", fill = "lightgrey") +
  ggtitle('STAT3 reporters upon JAK activation')
# Differences in sensitivity and specificity

ggplot_custom(reporter_correlations %>% 
                filter(tf == "TCF3") %>%
                dplyr::select(mES_2i_LIF, mES_LIF_PD, tf, reporter_id) %>% 
                na.omit(), aes(x = log2(mES_2i_LIF), y = log2(mES_LIF_PD))) +
  geom_point() +
  geom_abline(linetype = "dashed") +
  geom_smooth(method = "lm", color = "grey", fill = "lightgrey") +
  ggtitle('TCF3 reporters upon WNT inactivation')
# Differences in specificity

ggplot_custom(reporter_correlations %>% 
                filter(tf == "NFE2L2") %>%
                dplyr::select(mES_2i_LIF, mES_HQ, tf, reporter_id) %>% 
                na.omit(), aes(x = log2(mES_2i_LIF), y = log2(mES_HQ))) +
  geom_point() +
  geom_abline(linetype = "dashed") +
  geom_smooth(method = "lm", color = "grey", fill = "lightgrey") +
  ggtitle('NRF2 reporters upon oxidative stress induction')
# Clear case of design doesn't matter
```



---

## Compare reporter activities to other measures

*Comparison with TF expression and MPRA-inferred TF activities (Ernst et al.)*

```{r mpra_comparison, out.width= "80%", fig.align= "center"}
# Extract TF activities for K562 & HepG2
tf_activities <- cDNA_df %>%
  filter(condition %in% c("K562_DMSO", "HepG2_DMSO"), neg_ctrls == "No") %>%
  dplyr::select(reporter_id, reporter_activity_minP, condition, tf) %>%
  unique() %>%
  mutate(tf = gsub("_.*", "", tf)) %>%
  spread(condition, reporter_activity_minP) %>%
  mutate(dif = log2(K562_DMSO / HepG2_DMSO)) %>%
  dplyr::select(tf, dif, reporter_id)

#write_csv2(tf_activities, "/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/diffTF/tf_reporter_activities_all.csv")

# Correlate with RNA-seq
load('/DATA/usr/m.martinez.ara/GitLab/gurten/gurten/data/fc181121_epsure_tadec_rnaseq_expr_joshi.RData')

# remove Serum data and calculate mean expression
tib_fpkm_joshi$twoi <- (tib_fpkm_joshi$twoi_1 + tib_fpkm_joshi$twoi_2)/2
tib_fpkm_joshi <- tib_fpkm_joshi %>%
  dplyr::select(gene_id, 'tf' = symbol, twoi)

tf_activities2 <- cDNA_df %>%
  filter(condition == "mES_2i_LIF", neg_ctrls == "No") %>%
  dplyr::select(reporter_id, reporter_activity_minP, tf) %>%
  unique() %>%
  mutate(tf = gsub("_.*", "", tf),
         tf_activity = ave(reporter_activity_minP, tf, FUN = function(x) mean(x))) %>%
  dplyr::select(tf_activity, tf) %>%
  unique()

tf_activities2 <- merge(tf_activities2, tib_fpkm_joshi, by = "tf") %>%
  mutate(tf_activity = (tf_activity-mean(tf_activity))/sd(tf_activity),
         twoi = (twoi-mean(twoi))/sd(twoi))

ggplot_custom(tf_activities2, aes(x = twoi, y = tf_activity)) +
  geom_point() +
  ggtitle('TF reporter activities vs. TF expression in mESCs')

# Correlate with regulon data (VIPER, Aerts lab papers)


# Correlate with TF activities from Ernst et al. (https://doi.org/10.1038/nbt.3678)
ernst_activities <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/ernst_tf_activities_hepg2_k562.csv", sep = ";", dec = ",", header = T, fill = T)

ernst_activities <- ernst_activities %>%
  dplyr::select('tf' = TF, activity_enrichment_HepG2, activity_enrichment_K562) %>%
  mutate(avg_score_HepG2 = ave(activity_enrichment_HepG2, tf, FUN = function(x) mean(x)),
         avg_score_K562 = ave(activity_enrichment_K562, tf, FUN = function(x) mean(x))) %>%
  unique() %>%
  pivot_longer(-tf, names_to = "condition", values_to = "ernst_activity") %>%
  mutate(condition = gsub("avg_score_", "", condition))

tf_activities <- cDNA_df %>%
  filter(condition %in% c("K562_DMSO", "HepG2_DMSO"), neg_ctrls == "No") %>%
  dplyr::select(reporter_activity_minP, condition, tf) %>%
  mutate(tf = gsub("_.*", "", tf),
         condition = gsub("_DMSO", "", condition),
         reporter_activity_minP = ave(reporter_activity_minP, condition, tf, FUN = function(x) mean(x))) %>%
  unique()

ernst_activities <- merge(ernst_activities, tf_activities, by = c('tf', 'condition')) %>%
  mutate(ernst_activity = (ernst_activity-mean(ernst_activity))/sd(ernst_activity),
         reporter_activity_minP = (reporter_activity_minP-mean(reporter_activity_minP))/sd(reporter_activity_minP))

ggplot_custom(ernst_activities, 
              aes(x = ernst_activity, y = reporter_activity_minP)) +
  geom_point() +
  geom_abline(linetype = "dashed") +
  geom_smooth(method = "lm", color = "grey", fill = "lightgrey") +
  facet_wrap(~condition) +
  ggtitle('TF reporter activities vs. MPRA-chunk-inferred TF activities')

#write_csv2(ernst_activities, "/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/diffTF/tf_reporter_activities_ernst.csv")
```

---


## Compare reporter activities to ATAC-seq

*how does TF reporter activity relate to diffTF inferred TF activity?*

```{r atac_comparison, out.width= "80%", fig.align= "center"}
# DiffTF was run in basic and classification mode (https://doi.org/10.1016/j.celrep.2019.10.106), final output is imported:
## Input peaks for diffTF were selected in separate script
difftf_activity <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE-TF/ATAC_seq_hepg2/diffTF/output_rna_top1000_inverse/FINAL_OUTPUT/extension100/HepG2vsK562.top1000.summary.tsv") %>%
  dplyr::select(TF, 'diffTF_activity' = weighted_meanDifference, 'classification' = classification_q0.001_final)
difftf_activity$classification[difftf_activity$TF == "POU5F1::SOX2"] <- difftf_activity$classification[difftf_activity$TF == "POU5F1"]
difftf_activity$classification[difftf_activity$TF == "RAR:RXR"] <- difftf_activity$classification[difftf_activity$TF == "RARA"]

# Import mean reporter activities per TF
reporter_activity_all <- tf_activities
reporter_activity_all_mean <- reporter_activity_all %>%
  unique() %>%
  pivot_wider(names_from = "condition", values_from = "reporter_activity_minP") %>%
  mutate(dif_reporter = log2(K562 / HepG2)) %>%
  dplyr::select(dif_reporter, 'TF' = tf) %>%
  unique()
reporter_activity_all_mean$TF <- gsub("::.*", "_", reporter_activity_all_mean$TF)
reporter_activity_all_mean$TF[reporter_activity_all_mean$TF == "POU5F1_"] <- "POU5F1::SOX2"
reporter_activity_all_mean$TF[reporter_activity_all_mean$TF == "RARA:RXRA"] <- "RAR:RXR"
reporter_activity_all_mean$TF[reporter_activity_all_mean$TF == "ETS1"] <- "ETS2"
reporter_activity_all_mean$TF[reporter_activity_all_mean$TF == "NFATc1"] <- "NFATC1"
reporter_activity_all_mean$TF[reporter_activity_all_mean$TF == "TCF3"] <- "TCF7L2"
reporter_activity_all_mean$TF <- gsub("_", "", reporter_activity_all_mean$TF)


activity_comp <- merge(difftf_activity, reporter_activity_all_mean, by = "TF", all = T) %>%
  na.omit()

activity_comp <- activity_comp %>%
  mutate(diffTF_activity_norm = (diffTF_activity-mean(diffTF_activity))/sd(diffTF_activity),
         reporter_activity_norm = (dif_reporter-mean(dif_reporter))/sd(dif_reporter))

# Import class of TF
tf_class <- read_csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/garcia_tf-properties.csv") %>%
  dplyr::select(TF, chromatin_regulation_mode)

activity_comp <- merge(activity_comp, tf_class, all = T)
activity_comp$chromatin_regulation_mode[activity_comp$TF == "POU5F1::SOX2"] <- activity_comp$chromatin_regulation_mode[activity_comp$TF == "POU5F1"]
activity_comp$chromatin_regulation_mode[activity_comp$TF == "RAR:RXR"] <- activity_comp$chromatin_regulation_mode[activity_comp$TF == "RARA"]
activity_comp <- activity_comp %>% na.omit()

ggplot(activity_comp, aes(x = reporter_activity_norm, y = diffTF_activity_norm)) +
  geom_point(aes( color = chromatin_regulation_mode)) +
  geom_abline(linetype = "dashed") +
  geom_smooth(method = "lm", color = "grey", fill = "lightgrey") +
  xlab("Differential Reporter Activity (K562/HepG2)") +
  ylab("Differential Accessibility - diffTF (K562/HepG2)") +
  scale_color_manual(values = c("grey",colors_diverse[c(2,5,4)])) +
  facet_wrap(~classification) +
  ggtitle('TF reporter activity vs. diffTF activity')
```

---

## Plotting mean TF activities per condition

*display actual activity values per condition and TF instead of fraction of active reporters per TF*

```{r tf_activity_heatmap, out.width= "80%", fig.align= "center"}
## Prepare dataframe
# Caculate mean TF activity per condition
tf_activity_heatmap <- cDNA_df %>%
  filter(neg_ctrls == "No", commercial_reporter == "No") %>%
  mutate(tf_activity = ave(reporter_activity_minP, condition, tf, FUN = function(x) mean(x))) %>%
  #mutate(tf_activity = log2(tf_activity)) %>%
        #mutate(tf_activity = ave(tf_activity, condition, FUN = function(x) (x-mean(x))/sd(x))) %>%
  dplyr::select(tf_activity, condition, tf) %>%
  unique() %>%
  spread(condition, tf_activity) %>% 
  remove_rownames %>% 
  column_to_rownames(var="tf")

# Keeping the scale in the pheatmap function
myBreaks1 <- seq(0,6,0.06)

# Keep only TFs that are in at least one condition above a certain threshold
tf_activity_heatmap <- tf_activity_heatmap[rowSums(tf_activity_heatmap > 4) >= 1,] %>%
  na.omit()

#tf_activity_heatmap <- tf_activity_heatmap[,c(2,1,6,7,3,8,4,5)]

# pheatmap function
pheatmap(as.matrix(t(tf_activity_heatmap)),
         color = colorRampPalette(brewer.pal(n = 7, name = "Greys"))(100),
         breaks = myBreaks1, border_color = "black", 
         cellwidth = 8, cellheight = 8,
         angle_col = 90,
         main = 'reporter activities per tf and condition')
```
*This heatmap doesn't take significance into consideration and to me is less representative of tf activity than the fraction of reporter activity classes heatmap*

---


<!-- ## Linear model to explain variance of reporter activites per tf and condition (instead of activity classes now for actual activity values) -->

<!-- ```{r log-linear-model-5, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE} -->
<!-- ### For each TF separately - only the non-mutated  -->
<!-- cDNA_df_TF <- cDNA_df[cDNA_df$native_enhancer == "No" & -->
<!--                         cDNA_df$hPGK == "No" &  -->
<!--                         cDNA_df$neg_ctrls == "No" & -->
<!--                         cDNA_df$tf != "RANDOM1"& -->
<!--                         cDNA_df$tf != "RANDOM2"& -->
<!--                         cDNA_df$tf != "RANDOM3" & -->
<!--                         cDNA_df$commercial_reporter == "No",] %>%  -->
<!--   mutate(tf = paste(tf, condition, sep = "_")) %>% -->
<!--   dplyr::select(reporter_activity_minP, tf, promoter, spacing, distance, background) %>% -->
<!--   unique() -->


<!-- ### Calculate expression variance -->
<!-- exp_variance <- data.frame(c("promoter", "spacing", "distance", "background", "Residuals")) -->
<!-- prediction <- data.frame(1:504) -->
<!-- names(prediction) <- "id" -->
<!-- names(exp_variance) <- "feature" -->
<!-- weight <- data.frame(c("(Intercept)", "promotermCMV", "promoterminP",  -->
<!--                   "spacing5bp", "distance21bp", "background2", "background3")) -->
<!-- names(weight) <- "feature" -->


<!-- cDNA_df_TF$row <- rownames(cDNA_df_TF) -->

<!-- for (i in unique(cDNA_df_TF$tf[-grep("Irx3|Homez|Nr3c1(GR)", cDNA_df_TF$tf)])) { -->
<!--   cDNA_df_TF_2 <- cDNA_df_TF[-grep("Irx3|Homez|Nr3c1(GR)", cDNA_df_TF$tf),] -->
<!--   x <- lm(log_reporter_activity ~ promoter + spacing + distance + background,  -->
<!--                      cDNA_df_TF[cDNA_df_TF$tf == i,]) -->
<!--   y <- data.frame(x$fitted.values) %>% rownames_to_column() -->
<!--   names(y) <- c("row", i) -->
<!--   y$id <- 1:nrow(y) -->
<!--   prediction <- merge(prediction,y, all = T) -->

<!--   w <- data.frame(x$coefficients) %>% rownames_to_column() -->
<!--   names(w) <- c("feature", i) -->
<!--   weight <- merge(weight,w, all = T) -->

<!--   x <- anova(x) %>% rownames_to_column("feature") %>% -->
<!--     setnames("Sum Sq", "sum_sq") %>% -->
<!--     dplyr::select(feature, sum_sq) %>% -->
<!--     mutate(sum = sum(sum_sq)) %>% -->
<!--     mutate(rel_sum_sq = (sum_sq/sum)*100) %>% -->
<!--     setnames("rel_sum_sq", i) -->
<!--   x <- x[,c(1,4)] -->

<!--   exp_variance <- merge(exp_variance, x) -->
<!-- } -->


<!-- ## Fine-tune dfs -->
<!-- exp_variance <- melt(exp_variance, variable.name = "tf") -->
<!-- exp_variance$group <- "no residual" -->
<!-- exp_variance$group[grep("Residual", exp_variance$feature)] <- "residual" -->
<!-- prediction <- prediction %>% dplyr::select(-id) %>% melt(variable.name = "tf") %>% na.omit() %>% -->
<!--   dplyr::select(-tf) %>% setnames("value", "activity_predicted") -->
<!-- cDNA_df_TF <- merge(cDNA_df_TF, prediction) -->
<!-- weight <- weight %>% filter(feature != "(Intercept)") -->
<!-- weight <- melt(weight, variable.name = "tf") -->
<!-- categorie <- c("mutated", "condition", "promoter",  -->
<!--                "spacing", "distance", "background") -->
<!-- weight$features <- gsub(paste(categorie, collapse="|"), "",weight$feature) -->
<!-- ml <- length(weight$features) -->
<!-- feature <- weight$features -->
<!-- weight$condition <- gsub(paste(head(feature, n = ml/2), collapse = "|"), "", weight$feature) -->
<!-- weight <- weight %>% dplyr::select(-feature) -->



<!-- # Model parameters -->
<!-- exp_variance$value[exp_variance$feature == "Residuals"] <- 100 - exp_variance$value[exp_variance$feature == "Residuals"] -->
<!-- exp_variance$feature <- revalue(exp_variance$feature, c("Residuals"="total")) -->

<!-- ggplot(exp_variance[grep("Trp53", exp_variance$TF),],  -->
<!--        aes(x = reorder(feature, -value), y = value, fill = group)) + -->
<!--   scale_fill_manual(values = c("#1B998B", "#2D3047")) + -->
<!--   geom_bar(stat = "identity") +  -->
<!--   ylab("Variance explained (%)") + xlab("") +  -->
<!--   labs(title = "Linear variance modelling",  -->
<!--        subtitle = "Expression variance predicted by features") + -->

<!--   theme(text = element_text(size = 12)) + -->
<!--   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12),  -->
<!--         axis.text.y = element_text(size = 10)) + -->
<!--   facet_wrap(~TF) -->




<!-- ## Plot weights of parameters in model - this helps to understand the model  -->
<!-- level_order <- c("mCMV", "minP","Random",  -->
<!--                  "2", "3", "5bp", "21bp") -->





<!-- ## Final overview - model fit for each TF -->
<!-- # Feature importance overview for each TF in a single plot (bar-chart including mean activities per feature) -->


<!-- for (i in unique(cDNA_df_TF$TF[grep("Trp53", cDNA_df_TF$TF)])) { -->

<!--   # Plot to show prediction accuracy -->
<!--   sp <- ggscatter(cDNA_df_TF[cDNA_df_TF$TF == i,], x = "log_reporter_activity", y = "activity_predicted", -->
<!--    add = "reg.line",  -->
<!--    add.params = list(color = "blue", fill = "lightgray"), -->
<!--    conf.int = TRUE, ylab = "Predicted log2-activity",  -->
<!--    title = i, -->
<!--    xlab = "Average log2-activity") + theme_bw() -->


<!--   # Feature importance overview for each TF (including mean activities per feature) -->
<!--   fp <- ggplot(exp_variance[exp_variance$TF == i,],  -->
<!--        aes(x = reorder(feature, -value), y = value, fill = group)) + -->
<!--   scale_fill_manual(values = c("#1B998B", "#2D3047")) + -->
<!--   geom_bar(stat = "identity") +  -->
<!--   ylab("Variance explained (%)") + xlab("") + -->

<!--   theme(text = element_text(size = 12)) + -->
<!--   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 12),  -->
<!--         axis.text.y = element_text(size = 10)) + -->
<!--     guides(fill = F) -->

<!--   # Plot weights of parameters in model - this helps to understand the model  -->
<!--   p <- ggplot(weight[weight$TF == i,],  -->
<!--        aes(x = factor(features, level = level_order), y = value, fill = condition)) + -->
<!--   scale_fill_manual(values = c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B", "#C2B2B4", "#A54657")) + -->
<!--   geom_bar(stat = "identity") +  -->
<!--   ylab("Weight") + xlab("") + -->

<!--   theme(text = element_text(size = 12)) + -->
<!--   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1, size = 10),  -->
<!--         axis.text.y = element_text(size = 10))  -->

<!--   # Combine all plots -->
<!--   sp.plot <- arrangeGrob(sp,fp,p, nrow = 2, ncol = 2) -->
<!--   grid.arrange(sp.plot) -->
<!-- } -->
<!-- ``` -->


---


<!-- ## Random forest model to classify reporters (in an exploratory phase) -->

<!-- ```{r random_forest_classes, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE} -->
<!-- cDNA_df_lm <- cDNA_df %>% -->
<!--   filter(neg_ctrls == "No", commercial_reporter == "No", promoter != "Random", -->
<!--          condition %in% c("HepG2_CDCA", "HepG2_DMSO", "K562_DMSO", "mES_2i", -->
<!--                           "mES_2i_LIF", "mES_HQ", "mES_LIF_CH", "mES_LIF_PD", -->
<!--                           "mES_N2B27" -->
<!--                           ), -->
<!--          tf %in% names(cDNA_df4), -->
<!--          reporter_activity_class != -1) %>% -->
<!--   mutate(background = as.factor(background), -->
<!--          reporter_activity_class = as.factor(reporter_activity_class)) %>% -->
<!--   dplyr::select(reporter_id, tf, reporter_activity_class, condition, promoter, spacing, background, distance) %>% -->
<!--   unique() -->

<!-- mixed_classes <- cDNA_df3 %>% -->
<!--   mutate(id = paste(condition, tf, sep = "_")) %>% -->
<!--   dplyr::select(id, class) %>% -->
<!--   filter(class == 1) %>% -->
<!--   unique() -->

<!-- ## Fitting the model for each condition and tf -->
<!-- cDNA_df_lm$id <- paste(cDNA_df_lm$condition, cDNA_df_lm$tf, sep = "_") -->
<!-- cDNA_df_lm <- cDNA_df_lm %>% -->
<!--   mutate(sum = ave(id, id, FUN = function(x) length(x))) %>% -->
<!--   filter(sum >= 2, id %in% mixed_classes$id) %>% -->
<!--   dplyr::select(-sum) %>% -->
<!--   na.omit() -->

<!-- rf_features <- lapply(unique(cDNA_df_lm$id), function(x) randomForest(reporter_activity_class ~ promoter + spacing + background + distance, -->
<!--                                                                      cDNA_df_lm[cDNA_df_lm$id == x,])) -->

<!-- names(rf_features) <- unique(cDNA_df_lm$id) -->


<!-- # ROC -->
<!-- roc_rf_train <- list() -->
<!-- for (i in names(rf_features)) { -->
<!--   roc_rf_train[[i]] <- roc(response = cDNA_df_lm$reporter_activity_class[cDNA_df_lm$id == i], predictor = rf_features[[i]]$votes[,2]) -->
<!-- } -->

<!-- auc_rf <- data.frame("id" = names(roc_rf_train), -->
<!--                      "n" = 1:length(roc_rf_train), -->
<!--                      "auc" = "") -->
<!-- for (i in 1:length(roc_rf_train)) { -->
<!--   if (i == 1) { -->
<!--     plot <- plot(roc_rf_train[[i]], col="#264653", lwd=3, main="ROC curve RF") -->
<!--     auc_rf$auc[auc_rf$n == i] <- auc(plot) -->
<!--   } -->
<!--   else { -->
<!--     plot <- plot(roc_rf_train[[i]], add = TRUE, label = T) -->
<!--     auc_rf$auc[auc_rf$n == i] <- auc(plot) -->
<!--   } -->
<!-- } -->

<!-- auc_rf <- auc_rf %>% -->
<!--   mutate(condition = gsub("(.*)_.*$", "\\1", id), -->
<!--          tf = gsub(".*_.*_(.*$)", "\\1", id), -->
<!--          auc = as.numeric(auc)) -->

<!-- ggplot_custom(auc_rf, aes(x = tf, y = auc)) + -->
<!--   geom_point() + -->
<!--   geom_boxplot() + -->
<!--   geom_hline(yintercept = 0.75, linetype = "dashed") + -->
<!--   theme_classic_lines_90() -->

<!-- ggplot_custom(auc_rf, aes(x = condition, y = auc)) + -->
<!--   geom_point() + -->
<!--   geom_hline(yintercept = 0.75, linetype = "dashed", color = "red")+ -->
<!--   coord_flip() -->


<!-- auc_rf_sel <- auc_rf %>% -->
<!--   filter(auc >= 0.75) -->

<!-- rf_features <- rf_features[auc_rf_sel$id] -->


<!-- # Get the variable importances -->
<!-- explained_rf <- list() -->

<!-- for (i in names(rf_features)) { -->
<!--   explained_rf[[i]] <- explain(rf_features[[i]], data=cDNA_df_lm %>% dplyr::select(promoter, spacing, background, distance), y=cDNA_df_lm$reporter_activity_class) -->
<!-- } -->

<!-- varimps <- list() -->
<!-- for (i in names(explained_rf)) { -->
<!--   varimps[[i]] <- feature_importance(explained_rf[[i]], type = "ratio") -->
<!-- } -->
<!-- #, loss_function = '1-auc' type='ratio' -->

<!-- varimps <- bind_rows(varimps, .id = "id") -->

<!-- varimps <- varimps %>% -->
<!--   mutate(dropout_loss = ave(dropout_loss, id, variable, FUN = function(x) mean(x))) %>% -->
<!--   dplyr::select(id, 'feature' = variable, dropout_loss) %>% -->
<!--   mutate(condition = gsub("(.*)_.*$", "\\1", id), -->
<!--          tf = gsub(".*_.*_(.*$)", "\\1", id)) %>% -->
<!--   dplyr::select(-id) %>% -->
<!--   unique() -->

<!-- for (i in unique(varimps$condition)) { -->
<!--   p <- ggplot_custom(varimps %>% filter(condition == i), aes(y = dropout_loss, x = feature)) + -->
<!--     geom_point()+ -->
<!--     ylab("dropout loss (RMSE)") + -->
<!--     facet_wrap(~tf) + -->
<!--     theme_classic_lines_90() + -->
<!--     ggtitle(i) -->
<!--   print(p) -->
<!-- } -->

<!-- ggplot_custom(varimps, aes(x = feature, y = dropout_loss)) + -->
<!--   geom_boxplot() + -->
<!--   geom_quasirandom() + -->
<!--   theme_classic_lines_90() -->

<!-- ``` -->




# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

