---
title: "Oligo Design"
author: "Max Trauernicht"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    code_folding: show
    highlight: monochrome
    theme: journal
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
  editor_options:
    chunk_output_type: console
---

# knitr document van Steensel lab

# TF reporter oligo design

# Introduction
In this document, an oligo library of TF reporters will be generated. These oligos contain TF binding sites of ~30 selected TFs, random inactive spacing between them, followed by minimal promoters, and a barcode in the transcription unit.

# Data processing
## Path, Libraries, Parameters and Useful Functions
```{r setup, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(ggplot2)
library(seqinr)
library(seqLogo)
library(universalmotif)
library(Biostrings)
library(SimRAD)
library(gtools)
library(DNABarcodes)
library(phylotools)
library(ape)
library(magrittr)
library(dplyr)
library(readr)
library(stringr)
library(tidyr)
library(heatmaply)
library(pheatmap)
library(tibble)
library(ggseqlogo)
library(RColorBrewer)
```

### Custom functions
Functions used thoughout this script.
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}

# Function to substring the right part of the motif
substrRight <- function(x, n){
  substr(x, nchar(x)-n+1, nchar(x))
}


# Function to load PWM matrix
get_pwm_feature_matrix <- function(motif_meta_fn, fimo_fn, db = 2) {

  # validate args
  valid_dbs <- 1:2
  if(!db %in% valid_dbs)
    stop('Invalid db (database version). Please use db=1 (maintained for backward compatibility only) or db=2')

  # db=1 is maintained for backward compatibility only
  if(db == 1) {

    # read in motif metadata
    motif_meta    <- read.csv(motif_meta_fn)

    # check whether motif metadata contain essential annotations
    if(!all(c('PWM.ID', 'Cognate.TF') %in% colnames(motif_meta))) {
      message('The motif metadata file does not contain the essential columns PWM.ID and Cognate.TF')
    }

    motif_minimal <- motif_meta[, c('PWM.ID', 'Cognate.TF')]

    # load fimo output --> extract motif id, sequence id and p-value
    df <- read.table(fimo_fn)
    df <- df[, c(1, 2, 7)]

    colnames(df) <- c('PWM.ID', 'seqid', 'pval')

    # add TF id
    df <- merge(df, motif_minimal, by = 'PWM.ID')

    # group motif hits by sequence id
    l <- split(df, df[['seqid']])

    # multiple PWM and multiple hits possible. Reduce hits to one per TF, keeping best p-val only
    l <- lapply(l, function(x) {
      x_by_tf <- split(x, x[['Cognate.TF']], drop = TRUE)
      x_by_tf <- lapply(x_by_tf, function(y) y[which.min(y$pval), ])
      do.call('rbind', x_by_tf)
    })

    # initialize feature matrix
    n_tf          <- motif_minimal[['Cognate.TF']] %>%
      unique %>%
      length
    n_seq         <- length(l)
    pwm           <- matrix(1, nrow = n_seq, ncol = n_tf)
    colnames(pwm) <- (motif_minimal[['Cognate.TF']] %>% unique)

    # replace :: from names of composite motifs
    colnames(pwm) <- str_replace_all(colnames(pwm), '::', '_')

    # fill in feature matrix
    for(i in 1 : n_seq) {
      pwm[i, l[[i]][['Cognate.TF']]] <- l[[i]]$pval
    }

    # -log10 transform
    pwm           <- -1 * log10(pwm)

    # coerce to tib and return
    tib_fimo <- as_data_frame(pwm) %>%
      mutate(id = names(l))
      dplyr::select(id, everything())

  }

  # db = 2 (default)
  else {

    # load metadata
    tib_meta    <- read_csv(motif_meta_fn) %>%
      # extract tf symbol from motif id (Cognate_TF unsafe, it can be empty) and replace :: occurrences
      mutate(tf_symbol = str_remove(ID, '_[0-9]*'),
             tf_symbol = str_replace(tf_symbol, '::', '_')) %>%
      select(motif_id = `PWM ID`, tf_symbol)

    # load fimo results
    tib_fimo <- read_tsv(fimo_fn) %>%
      # extract motif id, sequence id and p-value
      select(motif_id, sequence_name, pval = `p-value`)

    # add tf symbol to fimo results
    tib_fimo <- tib_fimo %>%
      left_join(tib_meta, by = 'motif_id') %>%
      # remove hits with missing motif id (composite pwms)
      filter(!is.na(tf_symbol))

    # select best hit for each motif and sequence
    tib_fimo <- tib_fimo %>%
      group_by(sequence_name, tf_symbol) %>%
      dplyr::slice(which.min(pval)) %>%
      ungroup()

    # spread into feature matrix
    tib_fimo <- tib_fimo %>%
      mutate(pval = -1 * log10(pval)) %>%
      select(-motif_id) %>%
      spread(key = tf_symbol, value = pval, fill = 0, drop = TRUE) %>%
      # perform cosmetics on the id
      mutate(id = sequence_name) %>%
      select(-c(sequence_name)) %>%
      select(id, everything())

  }

  return(tib_fimo)

}
```


## Data import
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Import chosen TFs
tf.df.frame <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/parameter_files/TF_motifs.csv", header = T)
tf.df.annotated <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/parameter_files/TF_motifs_annotated.csv", header = T)

# Import parameters
spacings <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/parameter_files/Spacing.csv", header = T)
distances <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/parameter_files/Distance.csv", header = T)
promoters <- read.csv2("//DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/parameter_files/Promoter.csv", header = T)
barcodes <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/parameter_files/Barcode.csv", header = T)
background <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/parameter_files/Background.csv", header = T)
```

# Analysis

# Motif selection
```{r motif plots, out.width= "80%", fig.align= "center", echo = FALSE, warning = FALSE}
## Selecting motifs for the chosen TFs
# Visualize 'best' chosen motifs from Lambert et al., 2018 (Cell)
# Load and prepare the pwms
stat3 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M05480_1.94d.txt", header = T)
stat3 <- stat3[,-1]
stat3 <- t(stat3)
zfx <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M00640_1.94d.txt", header = T)
zfx <- zfx[,-1]
zfx <- t(zfx)
esrrb <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M08637_1.94d.txt", header = T)
esrrb <- esrrb[,-1]
esrrb <- t(esrrb)
smad4 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M10290_1.94d.txt", header = T)
smad4 <- smad4[,-1]
smad4 <- t(smad4)
klf4 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M08884_1.94d.txt", header = T)
klf4 <- klf4[,-1]
klf4 <- t(klf4)
tfcp2l1 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M09235_1.94d.txt", header = T)
tfcp2l1 <- tfcp2l1[,-1]
tfcp2l1 <- t(tfcp2l1)
gbx2 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M05758_1.94d.txt", header = T)
gbx2 <- gbx2[,-1]
gbx2 <- t(gbx2)
zfp42 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M09317_1.94d.txt", header = T)
zfp42 <- zfp42[,-1]
zfp42 <- t(zfp42)
trp53 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M06704_1.94d.txt", header = T)
trp53 <- trp53[,-1]
trp53 <- t(trp53)
rara <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M06033_1.94d.txt", header = T)
rara <- rara[,-1]
rara <- t(rara)
neurog2 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M08951_1.94d.txt", header = T)
neurog2 <- neurog2[,-1]
neurog2 <- t(neurog2)
pax6 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M09045_1.94d.txt", header = T)
pax6 <- pax6[,-1]
pax6 <- t(pax6)
rbpj <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M00862_1.94d.txt", header = T)
rbpj <- rbpj[,-1]
rbpj <- t(rbpj)
sp1 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M06128_1.94d.txt", header = T)
sp1 <- sp1[,-1]
sp1 <- t(sp1)
nr3c1 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M10263_1.94d.txt", header = T)
nr3c1 <- nr3c1[,-1]
nr3c1 <- t(nr3c1)
nfe2l2 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M02705_1.94d.txt", header = T)
nfe2l2 <- nfe2l2[,-1]
nfe2l2 <- t(nfe2l2)
nrf1 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M05963_1.94d.txt", header = T)
nrf1 <- nrf1[,-1]
nrf1 <- t(nrf1)
nfe2 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M08957_1.94d.txt", header = T)
nfe2 <- nfe2[,-1]
nfe2 <- t(nfe2)
nfkb1 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M10276_1.94d.txt", header = T)
nfkb1 <- nfkb1[,-1]
nfkb1 <- t(nfkb1)
creb1 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M10066_1.94d.txt", header = T)
creb1 <- creb1[,-1]
creb1 <- t(creb1)
gli1 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M00597_1.94d.txt", header = T)
gli1 <- gli1[,-1]
gli1 <- t(gli1)
irf3 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M08827_1.94d.txt", header = T)
irf3 <- irf3[,-1]
irf3 <- t(irf3)
irx3 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M03266_2.00.txt", header = T)
irx3 <- irx3[,-1]
irx3 <- t(irx3)
otx1 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M05974_1.94d.txt", header = T)
otx1 <- otx1[,-1]
otx1 <- t(otx1)
elk1 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M05658_1.94d.txt", header = T)
elk1 <- elk1[,-1]
elk1 <- t(elk1)
tcf7l2 <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M02162_1.94d.txt", header = T)
tcf7l2 <- tcf7l2[,-1]
tcf7l2 <- t(tcf7l2)
homez <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M05805_1.94d.txt", header = T)
homez <- homez[,-1]
homez <- t(homez)
nfya <- read.table("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/lambert_2018/PWMs/M06672_1.94d.txt", header = T)
nfya <- nfya[,-1]
nfya <- t(nfya)

# Combine all the pwms in a list
seqlogos <- list(stat3, zfx, esrrb, smad4, klf4, tfcp2l1, gbx2, zfp42, trp53, rara, neurog2, pax6, rbpj, sp1, nr3c1, nfe2l2, nrf1, nfe2, nfkb1, creb1, gli1, irf3, irx3, otx1, elk1, tcf7l2, homez, nfya)

## Rename the titles of the pwms
# names(seqlogos) <- c("Stat3 - M05480_1.94d", "Zfx - M00640_1.94d", "Esrrb - M08637_1.94d", "Smad4 - M10290_1.94d", "Klf4 - M08884_1.94d", "Tfcp2l1 - M09235_1.94d", "Gbx2 - M05758_1.94d", "Zfp42 - M09317_1.94d", "Trp53 - M06704_1.94d", "Rara - M06033_1.94d", "Neurog2 - M08951_1.94d", "Pax6 - M09045_1.94d", "Rbpj - M00862_1.94d", "Sp1 - M06128_1.94d", "Nr3c1 - M10263_1.94d", "Nfe2l2 - M02705_1.94d", "Nfkb1 - M10276_1.94d", "Creb1 - M10066_1.94d", "Gli1 - M00597_1.94d", "Irf3 - M08827_1.94d", "Irx3 - M03266_2.00", "Otx1 - M05974_1.94d", "Elk1 - M05658_1.94d", "Tcf7l2 - M02162_1.94d", "Homez - M05805_1.94d", "Nfya - M06672_1.94d")

names(seqlogos) <- c("Stat3", "Zfx", "Esrrb", "Smad4", "Klf4", "Tfcp2l1", "Gbx2", "Zfp42", "Trp53", "Rara", "Neurog2", "Pax6", "Rbpj", "Sp1", "Nr3c1", "Nfe2l2", "Nrf1", "NFE2", "Nfkb1", "Creb1", "Gli1", "Irf3", "Irx3", "Otx1", "Elk1", "Tcf7l2", "Homez", "Nfya")

# Create sequence plot of all pwms
ggseqlogo(seqlogos, ncol = 4) + theme_bw()+ theme(strip.background =element_rect(fill="#E6A08A"))

```

# Create DNA barcodes
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# # Create barcodes with length = 12
# barc <- create.dnabarcodes(n = 12, dist = 3, filter.triplets = T,
#                            filter.gc = T, filter.self_complementary = T, cores = 24)

# write.csv2(barc, file = "/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/parameter_files/barcodes.csv")
barc <- read.csv2("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/output/barcodes.csv", row.names = 1)
colnames(barc) <- "barcode"

# Filter out ATGs
barc <- barc[-grep("ATG",barc$barcode),]

# Filter out EcoRI & NheI sites
ecori_nhei <- c("GAATTC","GCTAGC")
barc <- barc[-grep(paste(ecori_nhei, collapse = "|"),barc)]
```





## Adding dimensions to the DF - sequences will be added in later stages
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
tf.df <- tf.df.frame
tf.df$Motif <- as.character(tf.df$Motif)

# Create whole df by merging all conditions
tf.df <- Reduce(function(x, y) merge(x, y, all=TRUE), list(tf.df, spacings, distances,
                                                            barcodes, promoters, background))

# Adding the DNA sequence from 5' to 3'

## Constant 5' primer sequence
tf.df$Primer1_seq <- ""

## Add motif to position 1
tf.df$motif1 <- tf.df$Motif

## Spacer sequence between TF motifs
tf.df$Space1 <- ""

## Do the same for the other 3 repeats
tf.df$motif2 <- tf.df$Motif
tf.df$Space2 <- ""
tf.df$motif3 <- tf.df$Motif

## The third spacing and the fourth motif is only included for the motifs shorter than 14 bp
# tf.df$Space3 <- 
#   ifelse (nchar(tf.df$motif1) < 16, NA,"")
# tf.df$motif4 <- 
#   ifelse (nchar(tf.df$motif1) < 16, tf.df$Motif,"")
tf.df$Space3 <- ""
tf.df$motif4 <- tf.df$Motif
  
# Sequence from last TF-motif to start of minimal promoter (10, 21, 21 incl. polyA)
tf.df$Distance_seq <- ""

## Minimal promoter
tf.df$Promoter_sequence <- "TAGAGGGTATATAATGGAAGCTCGACTTCCAG"
tf.df$Promoter_sequence[tf.df$Promoter == "mCMV"] <- "GGCGTTTACTATGGGAGGTCTATATAAGCAGAGCTCGTTTAGTGAACCGTCAGATC"
tf.df$Promoter_sequence[tf.df$Promoter == "hBGm"] <- "GGGCTGGGCATAAAAGTCAGGGCAGAGCCATCTATTGCTTACATTTGCTTCT"

## S1 Illumina adapter
tf.df$S1_primer <- "CACGACGCTCTTCCGATCT"

## Barcode
tf.df$barcode <- ""

## 3' Primer sequence
tf.df$Primer2_seq <- ""
```










# Add random sequence promoter for some conditions & check TF binding
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Add random 32 bp promoter
random_promoter <- 1:2000
random_promoter <- as.data.frame(random_promoter)
random_promoter$promoter <- ""

# Generate 2000 random sequences 
# GC content ~50%
set.seed(483972)
for (i in 1:2000) {
  random_promoter$promoter[random_promoter$random_promoter ==i] <- sim.DNAseq(32, GCfreq = 0.5)
}

random_promoter_space <- random_promoter
random_promoter_space$seq.name <- paste(random_promoter_space$random_promoter)


# Assemble sequence to test
random_promoter_space$seq.text <- paste(random_promoter_space$promoter)

# Write fasta file to run on FIMO script
random_promoter_space <- random_promoter_space[,c(-1:-2)]
# dat2fasta(random_promoter_space, outfile = "/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/parameter_files/output/fimo/random_promoter.fasta")     
```


 
## Run FIMO script again 
```{bash run fimo db2_1, eval = FALSE}
# motfn=/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/20170320_pwms_selected.meme
# odir=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/parameter_files/output/fimo/random_promoter
# query=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/parameter_files/output/fimo/random_promoter.fasta

# nice -n 19 fimo --no-qvalue --thresh 1e-4 --verbosity 1 --o $odir $motfn $query 
```



## load fimo results
We built a TF motif matrix using -log10 transformed FIMO scores. We used this feature encoding throughout the rest of this analysis, unless otherwise stated.

```{r build tf motif matrices db2: random_promoter, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# load motif Metadata --> PWM feature matrix
tib_pwm_random_promoter  <- get_pwm_feature_matrix(motif_meta_fn = '/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/fc181127_curated_metadata_no_composite_filt.csv', 
                                        fimo_fn       = '/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/random_promoter/fimo.tsv',
                                        db            = 2)
```


## visualize fimo results
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
## Select a random sequence which has no TF binding
# convert to binary
id <- tib_pwm_random_promoter$id 
tib_pwm_random_promoter_binary <- tib_pwm_random_promoter %>% mutate_if(is.numeric, function(x) ifelse((x==0 | is.na(x)),0,1)) 
tib_pwm_random_promoter_binary$id <- id
tib_pwm_random_promoter_binary_top <- tib_pwm_random_promoter_binary

# compute rowsums to get cumulative binding
tib_pwm_random_promoter_binary_top$binding <- rowSums(tib_pwm_random_promoter_binary_top[,2:ncol(tib_pwm_random_promoter_binary_top)])

# select only cumulative binding and id
tib_pwm_random_promoter_binary_top <- tib_pwm_random_promoter_binary_top %>%
  select(id,binding)

# Identify spacers that were exlcuded from fimo script (due to 0 hits)
space_id <- unique(tib_pwm_random_promoter_binary_top$id)
iteration <- 1:2000
space_nohit <- iteration[! iteration %in% space_id]
# 203 of the 2000 random sequences don't have any TF binding -> select one of those

# Select inactive 32 bp sequences
random_promoter_inactive <- random_promoter$promoter[random_promoter$random_promoter %in% space_nohit]

# Remove any NheI or EcoRI elements
random_promoter_inactive <- random_promoter_inactive[-grep(paste(ecori_nhei, collapse = "|"), random_promoter_inactive)]
# 199 sequences are left

# Also exclude these sites from random promoter - S1 adapter overlap region
random_promoter_inactive <- as.data.frame(random_promoter_inactive)
random_promoter_inactive$S1 <- "CACGACGCTCTTCCGATCT"
random_promoter_inactive$seq <- paste(random_promoter_inactive$random_promoter_inactive,
                                      random_promoter_inactive$S1, sep = "")
random_promoter_inactive <- random_promoter_inactive[-grep(paste(ecori_nhei, collapse = "|"), random_promoter_inactive$seq),]
random_promoter_inactive <- random_promoter_inactive$random_promoter_inactive
# 197 sequences are left

# Remove ATGs from random promoter
random_promoter_inactive <- random_promoter_inactive[-grep("ATG", random_promoter_inactive)]
# this leaves us with 122 possible sequences

# Just select the first random sequence promoter for a single condition 
# Select 10bp spacer, 21bp distance, minP, background 1 - keep all replicates & all TFs
# Bind to tf.df
rd.promoter <- tf.df
rd.promoter <- rd.promoter[rd.promoter$Spacing == "10bp" & rd.promoter$Distance == "21bp" & 
                             rd.promoter$Background == 1 & rd.promoter$Promoter == "minP",]
rd.promoter$Promoter <- "Random"
rd.promoter$Promoter_sequence <- random_promoter_inactive[1]
tf.df <- rbind(tf.df, rd.promoter)
```










## Now we want to find an optimal 5' primer sequence
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Generate 2000 random 18bp primer adapters and test in combination with first motif
motif_primer_1 <- tf.df %>% select(TF, Motif) %>% unique()
motif_primer_1$primer <- ""
iteration <- 1:2000
motif_primer_1 <- merge(motif_primer_1, iteration, all=T)

# Generate 1000 random sequences in front of first TF motif
# GC content ~50%
set.seed(23478)
for (i in 1:2000) {
  motif_primer_1$primer[motif_primer_1$y ==i] <- sim.DNAseq(18, GCfreq = 0.5)
}

motif_primer_1_space <- motif_primer_1
motif_primer_1_space$seq.name <- paste(motif_primer_1_space$TF, motif_primer_1_space$y, sep = "_")


# Assemble sequence to test: 4 rim bases of the TF motifs and the spacer sequence
motif_primer_1_space$seq.text <- paste(motif_primer_1_space$primer, 
                                       substr(motif_primer_1_space$Motif, 1, 4), sep = "")

# Write fasta file to run on FIMO script
motif_primer_1_space_export <- motif_primer_1_space
motif_primer_1_space_export <- motif_primer_1_space %>% select (seq.name, seq.text)
# dat2fasta(motif_primer_1_space_export, outfile = "/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/primer_1_2.fasta")     
```


 
## Run FIMO script again 
```{bash run fimo db2_2, eval = FALSE}
# motfn=/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/20170320_pwms_selected.meme
# odir=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/primer_1_2
# query=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/primer_1_2.fasta

# nice -n 19 fimo --no-qvalue --thresh 1e-4 --verbosity 1 --o $odir $motfn $query 
```



## load fimo results
We built a TF motif matrix using -log10 transformed FIMO scores. We used this feature encoding throughout the rest of this analysis, unless otherwise stated.

```{r build tf motif matrices db2: primer_1, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# load motif Metadata --> PWM feature matrix
tib_pwm_primer_1  <- get_pwm_feature_matrix(motif_meta_fn = '/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/fc181127_curated_metadata_no_composite_filt.csv', 
                                        fimo_fn       = '/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/primer_1_2/fimo.tsv',
                                        db            = 2)
```


## visualize fimo results
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# convert to binary
tib_pwm_primer_1_binary <- tib_pwm_primer_1 %>% mutate_if(is.numeric, function(x) ifelse((x==0 | is.na(x)),0,1)) 
tib_pwm_primer_1_binary_top <- tib_pwm_primer_1_binary

# compute rowsums to get cumulative binding
tib_pwm_primer_1_binary_top$binding <- rowSums(tib_pwm_primer_1_binary_top[,2:ncol(tib_pwm_primer_1_binary_top)])

# select only cumulative binding and id
tib_pwm_primer_1_binary_top <- tib_pwm_primer_1_binary_top %>%
  select(id,binding)

# make third column wiht only the space_id - then sum up the binding scores per space_id
tib_pwm_primer_1_binary_top$space <- gsub(".*_(.*)", "\\1", tib_pwm_primer_1_binary_top$id)
tib_pwm_primer_1_binary_top <- tib_pwm_primer_1_binary_top %>%
  select(-id)
tib_pwm_primer_1_binary_top$cum_binding <- ave(tib_pwm_primer_1_binary_top$binding, tib_pwm_primer_1_binary_top$space,
                                             FUN = sum)
tib_pwm_primer_1_binary_top <- tib_pwm_primer_1_binary_top %>%
  select(-binding)
tib_pwm_primer_1_binary_top <- unique(tib_pwm_primer_1_binary_top)

# Identify spacers that were exlcuded from fimo script (due to 0 hits)
space_id <- unique(tib_pwm_primer_1_binary_top$space)
iteration <- 1:2000
space_nohit <- iteration[! iteration %in% space_id]

# Remove EcoRI/NheI sites
remove_primer_1 <- motif_primer_1_space[grep(paste(ecori_nhei, collapse = "|"), motif_primer_1_space$seq.text),]
remove_primer_1 <- unique(remove_primer_1$y)
space_nohit <- space_nohit[!space_nohit %in% remove_primer_1]
# 20 tested sequences have no TF binding -> Choose one with good primer conditions
```

## Make final selection for 18bp primer sequence
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Select sequences from original file
spacers_selected <- space_nohit
motif_primer_1 <- unique(motif_primer_1[motif_primer_1$y %in% spacers_selected,3])

# Primer sequence should end with C/G
motif_primer_1_G <- motif_primer_1[grep(".*[C;G]$", motif_primer_1)]
# 10 possible primer sequences are left
# Chosen primer sequence:
motif_primer_1_G[5]
# Which is: CGGAGCGAACCGAGTTAG
```









## Now we want to find an optimal 3' primer sequence
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Subselect one spacing and neighboring TF motifs, make rows for random spacings that will be created
motif_primer_2 <- 1:2000
motif_primer_2 <- as.data.frame(motif_primer_2)
motif_primer_2$primer <- ""

# Generate 2000 random sequences 
# GC content ~50%
set.seed(483972)
for (i in 1:2000) {
  motif_primer_2$primer[motif_primer_2$motif_primer_2 ==i] <- sim.DNAseq(18, GCfreq = 0.5)
}

motif_primer_2_space <- motif_primer_2
motif_primer_2_space$seq.name <- paste(motif_primer_2_space$motif_primer_2)


# Assemble sequence to test: 4 rim bases of the TF motifs and the spacer sequence
motif_primer_2_space$seq.text <- paste(motif_primer_2_space$primer)

# Write fasta file to run on FIMO script
motif_primer_2_space_export <- motif_primer_2_space
motif_primer_2_space_export <- motif_primer_2_space[,c(-1:-2)]
# dat2fasta(motif_primer_2_space_export, outfile = "/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/primer_2.fasta")     
```


 
## Run FIMO script again 
```{bash run fimo db2_3, eval = FALSE}
# motfn=/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/20170320_pwms_selected.meme
# odir=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/primer_2
# query=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/primer_2.fasta

# nice -n 19 fimo --no-qvalue --thresh 1e-4 --verbosity 1 --o $odir $motfn $query 
```



## load fimo results
We built a TF motif matrix using -log10 transformed FIMO scores. We used this feature encoding throughout the rest of this analysis, unless otherwise stated.

```{r build tf motif matrices db2: primer_2, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# load motif Metadata --> PWM feature matrix
tib_pwm_primer_2  <- get_pwm_feature_matrix(motif_meta_fn = '/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/fc181127_curated_metadata_no_composite_filt.csv', 
                                        fimo_fn       = '/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/primer_2/fimo.tsv',
                                        db            = 2)
```


## visualize fimo results
```{r out.width= "100%", fig.align= "center", echo=FALSE}
## We make a pre-selection of the top 15 spacers first to narrow down the selection
# convert to binary
id <- tib_pwm_primer_2$id 
tib_pwm_primer_2_binary <- tib_pwm_primer_2 %>% mutate_if(is.numeric, function(x) ifelse((x==0 | is.na(x)),0,1)) 
tib_pwm_primer_2_binary$id <- id
tib_pwm_primer_2_binary_top <- tib_pwm_primer_2_binary

# compute rowsums to get cumulative binding
tib_pwm_primer_2_binary_top$binding <- rowSums(tib_pwm_primer_2_binary_top[,2:ncol(tib_pwm_primer_2_binary_top)])

# select only cumulative binding and id
tib_pwm_primer_2_binary_top <- tib_pwm_primer_2_binary_top %>%
  select(id,binding)

# make third column wiht only the space_id - then sum up the binding scores per space_id
tib_pwm_primer_2_binary_top$space <- gsub(".*_(.*)", "\\1", tib_pwm_primer_2_binary_top$id)
tib_pwm_primer_2_binary_top <- tib_pwm_primer_2_binary_top %>%
  select(-id)
tib_pwm_primer_2_binary_top$cum_binding <- ave(tib_pwm_primer_2_binary_top$binding, tib_pwm_primer_2_binary_top$space,
                                             FUN = sum)
tib_pwm_primer_2_binary_top <- tib_pwm_primer_2_binary_top %>%
  select(-binding)
tib_pwm_primer_2_binary_top <- unique(tib_pwm_primer_2_binary_top)

# Identify spacers that were exlcuded from fimo script (due to 0 hits)
space_id <- unique(tib_pwm_primer_2_binary_top$space)
iteration <- 1:2000
space_nohit <- iteration[! iteration %in% space_id]

# Remove EcoRI/NheI sites
remove_primer_2 <- motif_primer_2_space[grep(paste(ecori_nhei, collapse = "|"), motif_primer_2_space$seq.text),]
remove_primer_2 <- unique(remove_primer_2$seq.name)
space_nohit <- space_nohit[!space_nohit %in% remove_primer_2]
# 971 out of 2000 primer adapters don't bind any TFs
```

## Make final selection for 18bp primer sequence
```{r out.width= "100%", fig.align= "center", echo=FALSE}
# Select sequences from original file
spacers_selected <- space_nohit
motif_primer_2 <- unique(motif_primer_2[motif_primer_2$motif_primer_2 %in% spacers_selected,2])

# Primer sequence should end with C/G
motif_primer_2_G <- motif_primer_2[grep(".*[C;G]$", motif_primer_2)]

# Remove ATGs
motif_primer_2_G <- motif_primer_2_G[-grep("ATG", motif_primer_2_G)]
# All of those seem to be fine as a 5' primer sequence
# Chosen sequence:
motif_primer_2_G[11]
# Which is: CATCGTCGCATCCAAGAG
```















# Generate random spacings between TF motifs - screen those for TF binding
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
## Generate random spacings
## First select top 9 backgrounds for 5 bp spacings (3 backgrounds * 3 spacings per background)

# Subselect one spacing and neighboring TF motifs, make rows for random spacings that will be created
motif <- unique(tf.df[,c(1,9:11)])
iteration <- 1:623
motif <- merge(motif, iteration, all=T)

# Generate 623 random spacings between TF motifs
# GC content ~50%
set.seed(9456)
for (i in 1:623) {
  motif$Space1[motif$y ==i] <- sim.DNAseq(5, GCfreq = 0.5)
}

motif_5bp <- motif

# Assemble sequence to test: 4 rim bases of the TF motifs and the spacer sequence
motif_5bp$seq.name <- paste(motif_5bp$TF, motif_5bp$y, sep = "_")
motif_5bp$seq.text <- paste(substrRight(motif_5bp$motif1, 4), 
                       motif_5bp$Space1, substr(motif_5bp$motif2, 1, 4), sep = "")


# Write fasta file to run on FIMO script
motif_5bp_export <- motif_5bp[,c(-1:-5)]
# dat2fasta(motif_5bp_export, outfile = "/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/5bp_spacer_2.fasta")       
```


## Run FIMO script
```{bash run fimo db2_4, eval = FALSE}
# motfn=/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/20170320_pwms_selected.meme
# odir=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/space_1_3
# query=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/5bp_spacer_2.fasta

# nice -n 19 fimo --no-qvalue --thresh 1e-4 --verbosity 1 --o $odir $motfn $query 
```



## load fimo results
We built a TF motif matrix using -log10 transformed FIMO scores. We used this feature encoding throughout the rest of this analysis, unless otherwise stated. For some analysis, both Nanog and Klf2 libraries were combined.

```{r build tf motif matrices db2: 5bp_spacer, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# load motif Metadata --> PWM feature matrix
tib_pwm_5bp_spacer  <- get_pwm_feature_matrix(motif_meta_fn = '/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/fc181127_curated_metadata_no_composite_filt.csv', 
                                        fimo_fn       = '/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/space_1_3/fimo.tsv',
                                        db            = 2)
```


## visualize fimo results
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
## We make a pre-selection of the top 9 spacers first to narrow down the selection
# convert to binary
tib_pwm_5bp_spacer_binary <- tib_pwm_5bp_spacer %>% mutate_if(is.numeric, function(x) ifelse((x==0 | is.na(x)),0,1)) 
tib_pwm_5bp_spacer_binary_top <- tib_pwm_5bp_spacer_binary

# compute rowsums to get cumulative binding
tib_pwm_5bp_spacer_binary_top$binding <- rowSums(tib_pwm_5bp_spacer_binary_top[,2:ncol(tib_pwm_5bp_spacer_binary_top)])

# select only cumulative binding and id
tib_pwm_5bp_spacer_binary_top <- tib_pwm_5bp_spacer_binary_top %>%
  select(id,binding)

# make third column wiht only the space_id - then sum up the binding scores per space_id
tib_pwm_5bp_spacer_binary_top$space <- gsub(".*_(.*)", "\\1", tib_pwm_5bp_spacer_binary_top$id)
tib_pwm_5bp_spacer_binary_top <- tib_pwm_5bp_spacer_binary_top %>%
  select(-id)
tib_pwm_5bp_spacer_binary_top$cum_binding <- ave(tib_pwm_5bp_spacer_binary_top$binding, tib_pwm_5bp_spacer_binary_top$space,
                                             FUN = sum)
tib_pwm_5bp_spacer_binary_top <- tib_pwm_5bp_spacer_binary_top %>%
  select(-binding)
tib_pwm_5bp_spacer_binary_top <- unique(tib_pwm_5bp_spacer_binary_top)

# Identify spacers that were exlcuded from fimo script (due to 0 hits)
space_id <- unique(tib_pwm_5bp_spacer_binary_top$space)
space_nohit <- iteration[!iteration %in% space_id]

# Remove EcoRI/NheI sites
remove_5bp <- motif_5bp[grep(paste(ecori_nhei, collapse = "|"), motif_5bp$seq.text),]
remove_5bp <- unique(remove_5bp$y)
space_nohit <- space_nohit[!space_nohit %in% remove_5bp]

# This revealed the best 2 hits - now add 7 more to complete 9 
tib_pwm_5bp_spacer_binary_top <- tib_pwm_5bp_spacer_binary_top[order(tib_pwm_5bp_spacer_binary_top$cum_binding),]

# Look at the top 12 to see which TF binds to them (as they all have 1 hit for a TF)
low_binding_space <- tib_pwm_5bp_spacer_binary_top$space[1:12]
low_binding_space <- low_binding_space[!low_binding_space %in% remove_5bp]

# Subselect top 12 from original df
low_binding_space2 <- merge(low_binding_space, unique(tf.df$TF))


low_binding_space2$space <- paste(low_binding_space2$y, low_binding_space2$x, sep = "_")

tib_pwm_5bp_spacer_binary_select <- tib_pwm_5bp_spacer_binary[tib_pwm_5bp_spacer_binary$id %in%
                                                                low_binding_space2$space,]



# convert to matrix - leave out id
tib_pwm_5bp_spacer_binaryMatrix <- as.matrix(select(tib_pwm_5bp_spacer_binary_select,-id))

# assign ids as rownames of matrix
rownames(tib_pwm_5bp_spacer_binaryMatrix) <- tib_pwm_5bp_spacer_binary_select$id
```

```{r, fig.height=10, fig.width=10, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
#remove columns with only 0s and plot heatmap
heatmaply(tib_pwm_5bp_spacer_binaryMatrix[, colSums(tib_pwm_5bp_spacer_binaryMatrix != 0) > 0])
```

## Make final selection for 5bp spacers
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Top 4 was already selected based on 0 TF hits
spacers_selected <- space_nohit

# Based on the heatmap above the following spacer IDs were chosen: 
spacers_5bp_selected <- union(spacers_selected, c("320", "283", "28", "362", "564", "286", "228"))

# Select sequences from original file
motif <- motif[motif$y %in% spacers_5bp_selected,]
spacers_5bp_selected  <- unique(motif[,3])
```










## Choose 10 bp sequence between TF motifs instead of 5 bp
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Subselect one spacing and neighboring TF motifs, make rows for random spacings that will be created
motif_10bp <- unique(motif[,c(-3,-5)])
motif_10bp$Space2 <- ""
iteration <- 1:2000
motif_10bp <- merge(motif_10bp, iteration, all=T)

# Generate 2000 random spacings between TF motifs
# GC content ~50%
set.seed(12823)
for (i in 1:2000) {
  motif_10bp$Space2[motif_10bp$y ==i] <- sim.DNAseq(10, GCfreq = 0.5)
}

motif_10bp_space <- motif_10bp
motif_10bp_space$seq.name <- paste(motif_10bp_space$TF, motif_10bp_space$y, sep = "_")


# Assemble sequence to test: 4 rim bases of the TF motifs and the spacer sequence
motif_10bp_space$seq.text <- paste(substrRight(motif_10bp_space$motif1, 4), 
                       motif_10bp_space$Space2, 
                       substr(motif_10bp_space$motif2, 1, 4), sep = "")

# Write fasta file to run on FIMO script
motif_10bp_space_export <- motif_10bp_space
motif_10bp_space_export <- motif_10bp_space_export[,c(-1:-5)]
# dat2fasta(motif_10bp_space_export, outfile = "/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/10bp_spacer_2_2.fasta")     
```


 
## Run FIMO script again 
```{bash run fimo db2_5, eval = FALSE}
# motfn=/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/20170320_pwms_selected.meme
# odir=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/space_2_3
# query=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/10bp_spacer_2_2.fasta

# nice -n 19 fimo --no-qvalue --thresh 1e-4 --verbosity 1 --o $odir $motfn $query 
```



## load fimo results
We built a TF motif matrix using -log10 transformed FIMO scores. We used this feature encoding throughout the rest of this analysis, unless otherwise stated. For some analysis, both Nanog and Klf2 libraries were combined.

```{r build tf motif matrices db2: 10bp_spacer, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# load motif Metadata --> PWM feature matrix
tib_pwm_10bp_spacer  <- get_pwm_feature_matrix(motif_meta_fn = '/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/fc181127_curated_metadata_no_composite_filt.csv', 
                                        fimo_fn       = '/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/space_2_3/fimo.tsv',
                                        db            = 2)
```


## visualize fimo results
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
## We make a selection of 9 10bp spacer
# convert to binary
tib_pwm_10bp_spacer_binary <- tib_pwm_10bp_spacer %>% mutate_if(is.numeric, function(x) ifelse((x==0 | is.na(x)),0,1)) 
tib_pwm_10bp_spacer_binary_top <- tib_pwm_10bp_spacer_binary

# compute rowsums to get cumulative binding
tib_pwm_10bp_spacer_binary_top$binding <- rowSums(tib_pwm_10bp_spacer_binary_top[,2:ncol(tib_pwm_10bp_spacer_binary_top)])

# select only cumulative binding and id
tib_pwm_10bp_spacer_binary_top <- tib_pwm_10bp_spacer_binary_top %>%
  select(id,binding)

# make third column wiht only the space_id - then sum up the binding scores per space_id
tib_pwm_10bp_spacer_binary_top$space <- gsub(".*_(.*)", "\\1", tib_pwm_10bp_spacer_binary_top$id)
tib_pwm_10bp_spacer_binary_top <- tib_pwm_10bp_spacer_binary_top %>%
  select(-id)
tib_pwm_10bp_spacer_binary_top$cum_binding <- ave(tib_pwm_10bp_spacer_binary_top$binding, tib_pwm_10bp_spacer_binary_top$space,
                                             FUN = sum)
tib_pwm_10bp_spacer_binary_top <- tib_pwm_10bp_spacer_binary_top %>%
  select(-binding)
tib_pwm_10bp_spacer_binary_top <- unique(tib_pwm_10bp_spacer_binary_top)

# Identify spacers that were exlcuded from fimo script (due to 0 hits)
space_id <- unique(tib_pwm_10bp_spacer_binary_top$space)
iteration <- 1:2000
space_nohit <- iteration[! iteration %in% space_id]
# We have 6 hits here already

# Remove EcoRI/NheI sites
remove_10bp <- motif_10bp_space[grep(paste(ecori_nhei, collapse = "|"), motif_10bp_space$seq.text),]
remove_10bp <- unique(remove_10bp$y)
space_nohit <- space_nohit[!space_nohit %in% remove_10bp]
# Still 6 hits



# Select top 3 additional sequences - investigate top 11 (which each have 1 or 2 TF binding hits)
tib_pwm_10bp_spacer_binary_top <- tib_pwm_10bp_spacer_binary_top[order(tib_pwm_10bp_spacer_binary_top$cum_binding),]
low_binding_space <- tib_pwm_10bp_spacer_binary_top$space[1:11]
low_binding_space <- low_binding_space[!low_binding_space %in% remove_10bp]

# Subselect top 13 from original df
low_binding_space2 <- merge(low_binding_space, unique(tf.df$TF))
low_binding_space2$space <- paste(low_binding_space2$y, low_binding_space2$x, sep = "_")

tib_pwm_10bp_spacer_binary_select <- tib_pwm_10bp_spacer_binary[tib_pwm_10bp_spacer_binary$id %in%
                                                                low_binding_space2$space,]




# convert to matrix - leave out id
tib_pwm_10bp_spacer_binaryMatrix <- as.matrix(select(tib_pwm_10bp_spacer_binary_select,-id))

# assign ids as rownames of matrix
rownames(tib_pwm_10bp_spacer_binaryMatrix) <- tib_pwm_10bp_spacer_binary_select$id
```

```{r, fig.height=10, fig.width=10, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
#remove columns with only 0s and plot heatmap
heatmaply(tib_pwm_10bp_spacer_binaryMatrix[, colSums(tib_pwm_10bp_spacer_binaryMatrix != 0) > 0])
```


## Make final selection for 10bp spacers
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Based on the heatmap above we choose the top 3 hits: 
spacers_10bp_selected <- union(space_nohit, c("362", "1761", "292"))

# Select sequences from original file
spacers_10bp_selected <- unique(motif_10bp[motif_10bp$y %in% spacers_10bp_selected,4])
```








## Now we want to find 3 optimal spacings for in between the last TF motif and the core promoter - 10bp distance
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Subselect one spacing and neighboring TF motifs, make rows for random spacings that will be created
motif_promoter <- tf.df %>% 
  select(TF, Motif, Promoter_sequence)
motif_promoter <- unique(motif_promoter)
motif_promoter$dist1 <- ""
iteration <- 1:1000
motif_promoter <- merge(motif_promoter, iteration, all=T)

# Generate 1000 random spacings between last TF motif and core promoter
# GC content ~50%
set.seed(4125)
for (i in 1:1000) {
  motif_promoter$dist1[motif_promoter$y ==i] <- sim.DNAseq(10, GCfreq = 0.5)
}

motif_promoter_space <- motif_promoter
motif_promoter_space$seq.name <- paste(motif_promoter_space$TF, motif_promoter_space$y, sep = "_")


# Assemble sequence to test: 4 rim bases of the TF motifs and the spacer sequence
motif_promoter_space$seq.text <- paste(substrRight(motif_promoter_space$Motif, 4), 
                       motif_promoter_space$dist1, 
                       substr(motif_promoter_space$Promoter_sequence, 1, 4), sep = "")

# Write fasta file to run on FIMO script
motif_promoter_space_export <- motif_promoter_space
motif_promoter_space_export <- motif_promoter_space_export[,c(-1:-5)]
# dat2fasta(motif_promoter_space_export, outfile = "/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/promoter_spacer_1_3.fasta")     
```


 
## Run FIMO script again 
```{bash run fimo db2_6, eval = FALSE}
# motfn=/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/20170320_pwms_selected.meme
# odir=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/promoter_space_1_5
# query=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/promoter_spacer_1_3.fasta

# nice -n 19 fimo --no-qvalue --thresh 1e-4 --verbosity 1 --o $odir $motfn $query 
```



## load fimo results
We built a TF motif matrix using -log10 transformed FIMO scores. We used this feature encoding throughout the rest of this analysis, unless otherwise stated. For some analysis, both Nanog and Klf2 libraries were combined.

```{r build tf motif matrices db2: promoter_spacer, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# load motif Metadata --> PWM feature matrix
tib_pwm_promoter_spacer  <- get_pwm_feature_matrix(motif_meta_fn = '/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/fc181127_curated_metadata_no_composite_filt.csv', 
                                        fimo_fn       = '/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/promoter_space_1_5/fimo.tsv',
                                        db            = 2)
```


## visualize fimo results
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
## We make a pre-selection of the top 9 spacers first to narrow down the selection
# convert to binary
tib_pwm_promoter_spacer_binary <- tib_pwm_promoter_spacer %>% mutate_if(is.numeric, function(x) ifelse((x==0 | is.na(x)),0,1)) 
tib_pwm_promoter_spacer_binary_top <- tib_pwm_promoter_spacer_binary

# compute rowsums to get cumulative binding
tib_pwm_promoter_spacer_binary_top$binding <- rowSums(tib_pwm_promoter_spacer_binary_top[,2:ncol(tib_pwm_promoter_spacer_binary_top)])

# select only cumulative binding and id
tib_pwm_promoter_spacer_binary_top <- tib_pwm_promoter_spacer_binary_top %>%
  select(id,binding)

# make third column wiht only the space_id - then sum up the binding scores per space_id
tib_pwm_promoter_spacer_binary_top$space <- gsub(".*_(.*)", "\\1", tib_pwm_promoter_spacer_binary_top$id)
tib_pwm_promoter_spacer_binary_top <- tib_pwm_promoter_spacer_binary_top %>%
  select(-id)
tib_pwm_promoter_spacer_binary_top$cum_binding <- ave(tib_pwm_promoter_spacer_binary_top$binding, tib_pwm_promoter_spacer_binary_top$space,
                                             FUN = sum)
tib_pwm_promoter_spacer_binary_top <- tib_pwm_promoter_spacer_binary_top %>%
  select(-binding)
tib_pwm_promoter_spacer_binary_top <- unique(tib_pwm_promoter_spacer_binary_top)

# Identify spacers that were exlcuded from fimo script (due to 0 hits)
space_id <- unique(tib_pwm_promoter_spacer_binary_top$space)
iteration <- 1:1000
space_nohit <- iteration[! iteration %in% space_id]
# This gives 11 hits

# Remove EcoRI/NheI sites
remove_promoter <- motif_promoter_space[grep(paste(ecori_nhei, collapse = "|"), motif_promoter_space$seq.text),]
remove_promoter <- unique(remove_promoter$y)
space_nohit <- space_nohit[!space_nohit %in% remove_promoter]
# still 11 hits
```


## Make final selection for promoter spacers
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Based on the heatmap above the following spacer ID was chosen
spacers_selected_promoter_1 <- space_nohit[1:3]

# Select sequences from original file
spacers_selected_promoter_1 <- unique(motif_promoter[motif_promoter$y %in% spacers_selected_promoter_1,4])
```









## Do the same, but now use 21bp spacer between TF motif and core promoter
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Subselect one spacing and neighboring TF motifs, make rows for random spacings that will be created
motif_promoter <- tf.df %>% 
  select(TF, Motif, Promoter_sequence)
motif_promoter <- unique(motif_promoter)
motif_promoter$dist1 <- ""
iteration <- 1:500
motif_promoter <- merge(motif_promoter, iteration, all=T)

# Generate 500 random spacings between last TF motif and core promoter
# GC content ~50%
set.seed(434853)
for (i in 1:500) {
  motif_promoter$dist1[motif_promoter$y ==i] <- sim.DNAseq(21, GCfreq = 0.5)
}

motif_promoter_space <- motif_promoter
motif_promoter_space$seq.name <- paste(motif_promoter_space$TF, motif_promoter_space$y, sep = "_")


# Assemble sequence to test: 4 rim bases of the TF motifs and the spacer sequence
motif_promoter_space$seq.text <- paste(substrRight(motif_promoter_space$Motif, 4), 
                       motif_promoter_space$dist1, 
                       substr(motif_promoter_space$Promoter_sequence, 1, 4), sep = "")

# Write fasta file to run on FIMO script
motif_promoter_space_export <- motif_promoter_space
motif_promoter_space_export <- motif_promoter_space_export[,c(-1:-5)]
# dat2fasta(motif_promoter_space_export, outfile = "/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/promoter_spacer_2_2.fasta")     
```


 
## Run FIMO script again 
```{bash run fimo db2_7, eval = FALSE}
# motfn=/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/20170320_pwms_selected.meme
# odir=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/promoter_space_2_5
# query=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/promoter_spacer_2_2.fasta

# nice -n 19 fimo --no-qvalue --thresh 1e-4 --verbosity 1 --o $odir $motfn $query 
```



## load fimo results
We built a TF motif matrix using -log10 transformed FIMO scores. We used this feature encoding throughout the rest of this analysis, unless otherwise stated. 

```{r build tf motif matrices db2: promoter_spacer_2, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# load motif Metadata --> PWM feature matrix
tib_pwm_promoter_spacer_2  <- get_pwm_feature_matrix(motif_meta_fn = '/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/fc181127_curated_metadata_no_composite_filt.csv', 
                                        fimo_fn       = '/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/promoter_space_2_5/fimo.tsv',
                                        db            = 2)
```


## visualize fimo results
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# convert to binary
tib_pwm_promoter_spacer_2_binary <- tib_pwm_promoter_spacer_2 %>% mutate_if(is.numeric, function(x) ifelse((x==0 | is.na(x)),0,1)) 
tib_pwm_promoter_spacer_2_binary_top <- tib_pwm_promoter_spacer_2_binary

# compute rowsums to get cumulative binding
tib_pwm_promoter_spacer_2_binary_top$binding <- rowSums(tib_pwm_promoter_spacer_2_binary_top[,2:ncol(tib_pwm_promoter_spacer_2_binary_top)])

# select only cumulative binding and id
tib_pwm_promoter_spacer_2_binary_top <- tib_pwm_promoter_spacer_2_binary_top %>%
  select(id,binding)

# make third column wiht only the space_id - then sum up the binding scores per space_id
tib_pwm_promoter_spacer_2_binary_top$space <- gsub(".*_(.*)", "\\1", tib_pwm_promoter_spacer_2_binary_top$id)
tib_pwm_promoter_spacer_2_binary_top <- tib_pwm_promoter_spacer_2_binary_top %>%
  select(-id)
tib_pwm_promoter_spacer_2_binary_top$cum_binding <- ave(tib_pwm_promoter_spacer_2_binary_top$binding, tib_pwm_promoter_spacer_2_binary_top$space,
                                             FUN = sum)
tib_pwm_promoter_spacer_2_binary_top <- tib_pwm_promoter_spacer_2_binary_top %>%
  select(-binding)
tib_pwm_promoter_spacer_2_binary_top <- unique(tib_pwm_promoter_spacer_2_binary_top)

# Identify spacers that were exlcuded from fimo script (due to 0 hits)
space_id <- unique(tib_pwm_promoter_spacer_2_binary_top$space)
iteration <- 1:500
space_nohit <- iteration[! iteration %in% space_id]
# This gives 1 hit

# Remove EcoRI/NheI sites
remove_promoter_2 <- motif_promoter_space[grep(paste(ecori_nhei, collapse = "|"), motif_promoter_space$seq.text),]
remove_promoter_2 <- unique(remove_promoter_2$y)
space_nohit <- space_nohit[!space_nohit %in% remove_promoter_2]
# Still 1 hit


# Select top 2 additional sequences - investigate top 3 (which each have 2 TF binding hits)
tib_pwm_promoter_spacer_2_binary_top <- tib_pwm_promoter_spacer_2_binary_top[order(tib_pwm_promoter_spacer_2_binary_top$cum_binding),]
low_binding_space <- tib_pwm_promoter_spacer_2_binary_top$space[1:3]
low_binding_space <- low_binding_space[!low_binding_space %in% remove_10bp]

# Subselect top 13 from original df
low_binding_space2 <- merge(low_binding_space, unique(tf.df$TF))
low_binding_space2$space <- paste(low_binding_space2$y, low_binding_space2$x, sep = "_")

tib_pwm_promoter_spacer_2_binary_select <-
  tib_pwm_promoter_spacer_2_binary[tib_pwm_promoter_spacer_2_binary$id 
                                   %in%low_binding_space2$space,]




# convert to matrix - leave out id
tib_pwm_promoter_spacer_2_binaryMatrix <- as.matrix(select(tib_pwm_promoter_spacer_2_binary_select,-id))

# assign ids as rownames of matrix
rownames(tib_pwm_promoter_spacer_2_binaryMatrix) <- tib_pwm_promoter_spacer_2_binary_select$id
```

```{r, fig.height=10, fig.width=10, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
#remove columns with only 0s and plot heatmap
heatmaply(tib_pwm_promoter_spacer_2_binaryMatrix[, colSums(tib_pwm_promoter_spacer_2_binaryMatrix != 0) > 0])
```


## Make final selection for 21bp promoter spacers
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Based on the heatmap above we choose the top 3 hits: avoid high-risk TFs like NfKB1
spacers_selected_promoter_2 <- union(space_nohit, c("330", "300"))

# Select sequences from original file
spacers_selected_promoter_2 <- unique(motif_promoter[motif_promoter$y %in% spacers_selected_promoter_2,4])
```





















## Adding 3x random 11bp 'motifs' that serve as negative control
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# 5' primer sequence left side - random motif - 3x 5bp spacing / 3x 10 bp spacing
random_motif <- unique(motif_primer_1_space %>% select(Motif, seq.name)) 
random_motif$Motif <- ""
random_motif$seq.name <- 1:2000
random_motif <- unique(random_motif)
names(random_motif) <- c("motif", "iteration")

# Generate 2000 random 11bp motifs
# GC content ~50%
set.seed(023873)
for (i in 1:2000) {
  random_motif$motif[random_motif$iteration ==i] <- sim.DNAseq(11, GCfreq = 0.5)
}

# Generate all possible combinations of these motifs with primer sequences, spacers, etc.
random_motif2 <- random_motif[1:25,]
names(random_motif2) <- c("spacing", "repeat")
spacers <- list(motif_primer_1_G[5], spacers_5bp_selected[1], spacers_5bp_selected[2],
                 spacers_5bp_selected[3], spacers_5bp_selected[4], spacers_5bp_selected[5],
                 spacers_5bp_selected[6], spacers_5bp_selected[7], spacers_5bp_selected[8],
                 spacers_5bp_selected[9], spacers_10bp_selected[1], spacers_10bp_selected[2],
                 spacers_10bp_selected[3], spacers_10bp_selected[4], spacers_10bp_selected[5],
                 spacers_10bp_selected[6], spacers_10bp_selected[7], spacers_10bp_selected[8],
                 spacers_10bp_selected[9], spacers_selected_promoter_1[1], spacers_selected_promoter_1[2],
                 spacers_selected_promoter_1[3], 
                 spacers_selected_promoter_2[1], spacers_selected_promoter_2[2], 
                 spacers_selected_promoter_2[3])
random_motif2$spacing <- spacers
  


random_motif <- merge(random_motif, random_motif2)

random_motif_2 <- random_motif
random_motif_2$seq.name <- paste(random_motif_2$`repeat`, random_motif_2$iteration, sep = "_")


# Assemble sequence to test: 4 rim bases of the TF motifs and the spacer sequence
random_motif_2$seq.text <- paste(substrRight(random_motif_2$spacing, 4), 
                       random_motif_2$motif, 
                       substr(random_motif_2$spacing, 1, 4), sep = "")

# Write fasta file to run on FIMO script
random_motif_2_exp <- random_motif_2 %>% select(seq.name, seq.text)
# dat2fasta(random_motif_2_exp, outfile = "/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/random_motif.fasta")     
```


 
## Run FIMO script again 
```{bash run fimo db2_8, eval = FALSE}
# motfn=/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/20170320_pwms_selected.meme
# odir=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/random_motif
# query=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/random_motif.fasta

# nice -n 19 fimo --no-qvalue --thresh 1e-4 --verbosity 1 --o $odir $motfn $query 
```



## load fimo results
We built a TF motif matrix using -log10 transformed FIMO scores. We used this feature encoding throughout the rest of this analysis, unless otherwise stated.

```{r build tf motif matrices db2: rd_motif, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# load motif Metadata --> PWM feature matrix
tib_pwm_random_motif  <- get_pwm_feature_matrix(motif_meta_fn = '/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/fc181127_curated_metadata_no_composite_filt.csv', 
                                        fimo_fn       = '/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/random_motif/fimo.tsv',
                                        db            = 2)
```


## visualize fimo results
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# convert to binary
tib_pwm_random_motif_binary <- tib_pwm_random_motif %>% mutate_if(is.numeric, function(x) ifelse((x==0 | is.na(x)),0,1)) 
tib_pwm_random_motif_binary_top <- tib_pwm_random_motif_binary

# compute rowsums to get cumulative binding
tib_pwm_random_motif_binary_top$binding <- rowSums(tib_pwm_random_motif_binary_top[,2:ncol(tib_pwm_random_motif_binary_top)])

# select only cumulative binding and id
tib_pwm_random_motif_binary_top <- tib_pwm_random_motif_binary_top %>%
  select(id,binding)

# make third column wiht only the space_id - then sum up the binding scores per space_id
tib_pwm_random_motif_binary_top$space <- gsub(".*_(.*)", "\\1", tib_pwm_random_motif_binary_top$id)
tib_pwm_random_motif_binary_top <- tib_pwm_random_motif_binary_top %>%
  select(-id)
tib_pwm_random_motif_binary_top$cum_binding <- ave(tib_pwm_random_motif_binary_top$binding, tib_pwm_random_motif_binary_top$space,
                                             FUN = sum)
tib_pwm_random_motif_binary_top <- tib_pwm_random_motif_binary_top %>%
  select(-binding)
tib_pwm_random_motif_binary_top <- unique(tib_pwm_random_motif_binary_top)

# Identify spacers that were exlcuded from fimo script (due to 0 hits)
space_id <- unique(tib_pwm_random_motif_binary_top$space)
iteration <- 1:2000
space_nohit <- iteration[! iteration %in% space_id]

# Remove EcoRI/NheI sites
remove_primer_1 <- random_motif_2[grep(paste(ecori_nhei, collapse = "|"), random_motif_2$seq.text),]
remove_primer_1 <- unique(remove_primer_1$iteration)
space_nohit <- space_nohit[!space_nohit %in% remove_primer_1]
# 22 tested sequences have no TF binding
```

## Make final selection random motifs
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Select sequences from original file
motifs_selected <- space_nohit
random_motif_2 <- unique(random_motif_2[random_motif_2$iteration %in% motifs_selected,1])
## These random motifs could be used
tf.df$Motif[tf.df$TF == "Random1"] <- random_motif_2[1]
tf.df$Motif[tf.df$TF == "Random2"] <- random_motif_2[2]
tf.df$Motif[tf.df$TF == "Random3"] <- random_motif_2[3]
tf.df$motif1[tf.df$TF == "Random1"] <- random_motif_2[1]
tf.df$motif1[tf.df$TF == "Random2"] <- random_motif_2[2]
tf.df$motif1[tf.df$TF == "Random3"] <- random_motif_2[3]
tf.df$motif2[tf.df$TF == "Random1"] <- random_motif_2[1]
tf.df$motif2[tf.df$TF == "Random2"] <- random_motif_2[2]
tf.df$motif2[tf.df$TF == "Random3"] <- random_motif_2[3]
tf.df$motif3[tf.df$TF == "Random1"] <- random_motif_2[1]
tf.df$motif3[tf.df$TF == "Random2"] <- random_motif_2[2]
tf.df$motif3[tf.df$TF == "Random3"] <- random_motif_2[3]
tf.df$motif4[tf.df$TF == "Random1"] <- random_motif_2[1]
tf.df$motif4[tf.df$TF == "Random2"] <- random_motif_2[2]
tf.df$motif4[tf.df$TF == "Random3"] <- random_motif_2[3]
```
# This approach  seems to fish out negative sequences quite well (as can be seen in the later results)















# Combining all sequences to get final oligo design
```{r out.width= "100%", fig.align= "center", echo=FALSE}
# Add sequences 5' to 3'
## Add 5 bp spacings
tf.df$Space1[tf.df$Spacing == "5bp" & tf.df$Background == 1] <- spacers_5bp_selected[1]
tf.df$Space2[tf.df$Spacing == "5bp" & tf.df$Background == 1] <- spacers_5bp_selected[2]
tf.df$Space3[tf.df$Spacing == "5bp" & tf.df$Background == 1] <- spacers_5bp_selected[3]
# tf.df$Space3[tf.df$Spacing == "5bp" & tf.df$Background == 1] <- 
#   ifelse (nchar(tf.df$motif1) < 16, spacers_5bp_selected[3],"")
tf.df$Space1[tf.df$Spacing == "5bp" & tf.df$Background == 2] <- spacers_5bp_selected[4]
tf.df$Space2[tf.df$Spacing == "5bp" & tf.df$Background == 2] <- spacers_5bp_selected[5]
tf.df$Space3[tf.df$Spacing == "5bp" & tf.df$Background == 2] <- spacers_5bp_selected[6]
# tf.df$Space3[tf.df$Spacing == "5bp" & tf.df$Background == 2] <- 
#   ifelse (nchar(tf.df$motif1) < 16, spacers_5bp_selected[6],"")
tf.df$Space1[tf.df$Spacing == "5bp" & tf.df$Background == 3] <- spacers_5bp_selected[7]
tf.df$Space2[tf.df$Spacing == "5bp" & tf.df$Background == 3] <- spacers_5bp_selected[8]
tf.df$Space3[tf.df$Spacing == "5bp" & tf.df$Background == 3] <- spacers_5bp_selected[9]
# tf.df$Space3[tf.df$Spacing == "5bp" & tf.df$Background == 3] <- 
#   ifelse (nchar(tf.df$motif1) < 16, spacers_5bp_selected[9],"")

## Add 10 bp spacings
tf.df$Space1[tf.df$Spacing == "10bp" & tf.df$Background == 1] <- spacers_10bp_selected[1]
tf.df$Space2[tf.df$Spacing == "10bp" & tf.df$Background == 1] <- spacers_10bp_selected[2]
tf.df$Space3[tf.df$Spacing == "10bp" & tf.df$Background == 1] <- spacers_10bp_selected[3]
# tf.df$Space3[tf.df$Spacing == "10bp" & tf.df$Background == 1] <- 
#   ifelse (nchar(tf.df$motif1) < 16, spacers_10bp_selected[3],"")
tf.df$Space1[tf.df$Spacing == "10bp" & tf.df$Background == 2] <- spacers_10bp_selected[4]
tf.df$Space2[tf.df$Spacing == "10bp" & tf.df$Background == 2] <- spacers_10bp_selected[5]
tf.df$Space3[tf.df$Spacing == "10bp" & tf.df$Background == 2] <- spacers_10bp_selected[6]
# tf.df$Space3[tf.df$Spacing == "10bp" & tf.df$Background == 2] <- 
#   ifelse (nchar(tf.df$motif1) < 16, spacers_10bp_selected[6],"")
tf.df$Space1[tf.df$Spacing == "10bp" & tf.df$Background == 3] <- spacers_10bp_selected[7]
tf.df$Space2[tf.df$Spacing == "10bp" & tf.df$Background == 3] <- spacers_10bp_selected[8]
tf.df$Space3[tf.df$Spacing == "10bp" & tf.df$Background == 3] <- spacers_10bp_selected[9]
# tf.df$Space3[tf.df$Spacing == "10bp" & tf.df$Background == 3] <- 
#   ifelse (nchar(tf.df$motif1) < 16, spacers_10bp_selected[9],"")

## Add spacing in front of promoter - 10 bp
tf.df$Distance_seq[tf.df$Distance == "10bp" & tf.df$Background == 1] <- spacers_selected_promoter_1[1]
tf.df$Distance_seq[tf.df$Distance == "10bp" & tf.df$Background == 2] <- spacers_selected_promoter_1[2]
tf.df$Distance_seq[tf.df$Distance == "10bp" & tf.df$Background == 3] <- spacers_selected_promoter_1[3]

## Add spacing in front of promoter - 21 bp
tf.df$Distance_seq[tf.df$Distance == "21bp" & tf.df$Background == 1] <- spacers_selected_promoter_2[1]
tf.df$Distance_seq[tf.df$Distance == "21bp" & tf.df$Background == 2] <- spacers_selected_promoter_2[2]
tf.df$Distance_seq[tf.df$Distance == "21bp" & tf.df$Background == 3] <- spacers_selected_promoter_2[3]

## Add 5' primer sequence: This one had the best primer qualities
tf.df$Primer1_seq <- motif_primer_1_G[5]

## Add 3' primer sequence: This one had the best primer qualities
tf.df$Primer2_seq <- motif_primer_2_G[11]
```



## Include positive controls - hPGK promoter
```{r}
# hPGK promoter as a positive control: select 3' 183 bp region of the hPGK
hPGK <- "cagctttgctccttcgctttctgggctcagaggctgggaaggggtgggtccgggggcgggctcaggggcgggctcaggggcggggcgggcgcccgaaggtcctccggaggcccggcattctgcacgcttcaaaagcgcacgtctgccgcgctgttctcctcttcctcatctccgggcctttcg"

# Add this sequence in place of promoter
pgk <- tf.df[1:20,]
pgk$TF <- "ctrl"
pgk$Motif <- ""
pgk$Spacing <- ""
pgk$Distance <- ""
pgk$Promoter <- "hPGK"
pgk$Barcode <- 1:20
pgk$Background <- ""
pgk$motif1 <- ""
pgk$Space1 <- ""
pgk$motif2 <- ""
pgk$Space2 <- ""
pgk$motif3 <- ""
pgk$Space3 <- ""
pgk$motif4 <- ""
pgk$Distance_seq <- ""
pgk$Promoter_sequence <- hPGK
tf.df <- rbind(tf.df, pgk)

# Export pgk for adding the control to different libraries
filename <- SetFileName("_pgk", "mt")
setwd("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/")
write.csv(pgk, file = paste(filename,".csv", sep = ""), row.names = F)
```



## Include positive controls - Klf2 native enhancer
```{r}
# Klf2 native enhancer controls
## Compute region size
l <- max(nchar(paste(tf.df$motif1, tf.df$Space1, tf.df$motif2, tf.df$Space2, 
           tf.df$motif3, tf.df$Space3, tf.df$motif4, tf.df$Distance_seq, sep = "")))
## 123bp is the longest possible region -> we can select 100bp to make it more easy

## Import the selected enhancers
### E6_Klf2 (mm10 - chr8:71900631-71901081) -> medium enhancer
e6_klf2 <- "CAGACGTTACTGGCCTTTCTTCCTGACACCCTGTCCTGTCAGCAGCGGGAGTCCGAGGTGGCTGAGATTCACTGGCCCTGTGACCCCTGAACCAAGGAATTTCTTAAGAGATAGTAGGAACCCTGGACTCAGACAAGTGAGAGGCCAGTGTGCGGGCGTGGATGGGGCGGGGTTGTGGGAGGGGCTCTCGGTCTTTCAGCTCCCTGGAGTCCAATCCCACTGGAGCTTCTGCGCAGCTGCACCCGGGGTGCATTGTGCCGGGTTGGGGGTGGGGTGATGAGCTGAAGGTAGGAGTAAGGGTGGGGTGAGCTGGGGGCGGAGCTGAGGGAGGGCTGGGTGTGTTGGGGTCGGGGTTGGAGCCCAGTTACAAGTTCTCTGAGCAGCTAGGATACCTGAGAGACTTTGCATTCAAGAAAAATATTCTTTAAAACACTAAGTTGCCCTCTTGTG"

### E11_Klf2 (mm10 - chr8:71952937-71953387) -> strong enhancer
e11_klf2 <- "ACATGGGAACTGAACCGAGGACCTCTAGAAAAGTAACGTGCGCTCTTAACCACTGAGTCATCTCTCCAGCCCCTTAAGATGGTCTCACTATGTGGCCCTAGAACTGAAAAAGTATAACTCAAAAGCAAAATTCTAGACCTTTAGGCCCAGGCTTGAGAATGCGACCTCTGTGACTCAATGCTCCAAGGTATGCTAATCATAAAACAGATAACGACATTCCTCCATCCACCCGCTTCCTGGGTTCAAGGAGTGTTCATTCAAAGAAAGTCATCCTGACCCACCCTGACCTGCTGGAACCGGCTATGCAAATGTGCTATTGTTAGGGGCTCTTAGAAACTGTCTTGAGAAATAACTCGCACAATTACCAGGTATTCCTCGTGACTTGTAACTTTGTACTCCCTTGTGACTCTTAACTAGTATTTTTGGTATTTTCCAACAACGCCCTCCGCA"

### E19_Klf2 (mm10 - chr8:71984525-71984975) -> no enhancer
e19_klf2 <- "CTGTTCCAGTTCTTTGGGAACCATGTGAAGACCAAGCTGCACACCTGTTACACATGTGTAGGTATACTGAGGGCTACCCCGTGTATGTTCTTTGGTTGGTGGTTCAGTCTCTGAGAGCTCCAAGGGTCCAGGTTATTTGAATCTGCTAGTCTTCCTGTGGAAGTCCTAACCCCTTAAGAGCCTGAAATCCTTCCTCCTAATAAGAGTCCCCAAGCTCCATCTACTGTTTGGCTATGGTATTTTTATTAACAAGCTGGAGACCTAAGCTGGGCAGGTTCTGAACTACTGTAATCCTCTAAGCTATTTATGTCCCAGCCGCAGGCCCCATGGTACTTGCAGTTTGAGTCTCCCCCTGGCTTCCTATGCTCCATGTGTATCCTCATGGTGAATGCCTTGGCGACCTGTTCATCACAACTCCTCCTGGATTCTCTGCCTCATGGCCACCTCGCC"

### E93_Klf2 (mm10 - chr8:72659953-72660403) -> no enhancer
e93_klf2 <- "AAGGCAAAGACAGATGGATCTCTGGAGCTCACTGACAGCCAAATTAGGCAATTGATGAGCTCTGGGTTCTTGGAGTGTTGCTGTCTCAAGAGATAAACTGTGGGCTGGAGAGATGGTTCAGAAGTTAGGACCTCCTGCTGTAATCACAGAAGATCCAGCACCCACATCAGGTGGCTCATATATGCTAGTGACTCCAGCTCCTGGGGAATCTGACTCCCTCTCCTGGCCTTTGTGGGTACTGCACACACAATAGTCCATAATCACCATTTAAGTACTTAGAGGGAGGACTAAAAGTCACACACACACACACACACACACACACACACACACACACACACACACACACACGGCCCTCGTGTTTGTAACCCCACTTCTCAGTTCTTTTTCTATGCTCCTAAGTTTATTTTCCCAAAGGCCTAACAATCATGTTGAGTTTCAAAGGAGAACTGG"

### E97_Klf2 (mm10 - chr8:72676845-72677295) -> strong enhancer
e97_klf2 <- "ACACAAAACAAATAAGAATGCATACATAAAAAGTTGGCGGGGACCAGGCCACATTTACCACTTAAAATGGAGCCAGGTGAATGCTCTGATGCAATGCTAGTCTATTGTTAAACTCACCACCAGGGGTTCTTTAGTAAATACCTGATTACGCTATTCCTCTGGACCTAGTGAAGAAACCTGTACCAGGGGAAGTCCCTTCCACTAACTCTTTCATGTGACAATCCACCTATTTGCTAGGCCATTGTGTATTCCTGTGTTTGGGTGTGACTCAGCTATTGTCCTAAGTAATTACTATGCAGACTACCCCTGAGCTTTCCTAGTTCTGTACTTTGTAATGCCTAATTAGTATCACAGTCTCTACTAGAAGTGAATTTAAGTGTTACTGAATAGGTAACATTCTTAGTGAAGTCCAAGCTCAGGGTCGGCTTCAAGGATTTTCTAGGAACCATT"

## Take 50bp sliding window from 450bp enhancer region -> 9 fragments with 100bp fragment length each
tfs <- c("e6_klf2", "e11_klf2", "e19_klf2", "e93_klf2", "e97_klf2")
fragment <- 1:8

enhancers <- merge(tfs, fragment, all = T)
enhancers <- paste(enhancers$x, enhancers$y, sep = "_") 

klf2_minP <- tf.df[tf.df$Promoter == "minP",]
klf2_minP <- klf2_minP[1:320,]
klf2_minP$Barcode[1:40] <- 1
klf2_minP$Barcode[41:80] <- 2
klf2_minP$Barcode[81:120] <- 3
klf2_minP$Barcode[121:160] <- 4
klf2_minP$Barcode[161:200] <- 5
klf2_minP$Barcode[201:240] <- 6
klf2_minP$Barcode[241:280] <- 7
klf2_minP$Barcode[281:320] <- 8
klf2_minP$TF <- enhancers
klf2_minP$Motif <- ""
klf2_minP$Spacing <- ""
klf2_minP$Distance <- ""
klf2_minP$Background <- ""
klf2_minP$Space1 <- ""
klf2_minP$motif2 <- ""
klf2_minP$Space2 <- ""
klf2_minP$motif3 <- ""
klf2_minP$Space3 <- ""
klf2_minP$motif4 <- ""
klf2_minP$Distance_seq <- ""


klf2_minP$motif1[klf2_minP$TF == "e6_klf2_1"] <- substr(e6_klf2, 1, 100)
klf2_minP$motif1[klf2_minP$TF == "e11_klf2_1"] <- substr(e11_klf2, 1, 100)
klf2_minP$motif1[klf2_minP$TF == "e19_klf2_1"] <- substr(e19_klf2, 1, 100)
klf2_minP$motif1[klf2_minP$TF == "e93_klf2_1"] <- substr(e93_klf2, 1, 100)
klf2_minP$motif1[klf2_minP$TF == "e97_klf2_1"] <- substr(e97_klf2, 1, 100)

klf2_minP$motif1[klf2_minP$TF == "e6_klf2_2"] <- substr(e6_klf2, 51, 150)
klf2_minP$motif1[klf2_minP$TF == "e11_klf2_2"] <- substr(e11_klf2, 51, 150)
klf2_minP$motif1[klf2_minP$TF == "e19_klf2_2"] <- substr(e19_klf2, 51, 150)
klf2_minP$motif1[klf2_minP$TF == "e93_klf2_2"] <- substr(e93_klf2, 51, 150)
klf2_minP$motif1[klf2_minP$TF == "e97_klf2_2"] <- substr(e97_klf2, 51, 150)

klf2_minP$motif1[klf2_minP$TF == "e6_klf2_3"] <- substr(e6_klf2, 101, 200)
klf2_minP$motif1[klf2_minP$TF == "e11_klf2_3"] <- substr(e11_klf2, 101, 200)
klf2_minP$motif1[klf2_minP$TF == "e19_klf2_3"] <- substr(e19_klf2, 101, 200)
klf2_minP$motif1[klf2_minP$TF == "e93_klf2_3"] <- substr(e93_klf2, 101, 200)
klf2_minP$motif1[klf2_minP$TF == "e97_klf2_3"] <- substr(e97_klf2, 101, 200)

klf2_minP$motif1[klf2_minP$TF == "e6_klf2_4"] <- substr(e6_klf2, 151, 250)
klf2_minP$motif1[klf2_minP$TF == "e11_klf2_4"] <- substr(e11_klf2, 151, 250)
klf2_minP$motif1[klf2_minP$TF == "e19_klf2_4"] <- substr(e19_klf2, 151, 250)
klf2_minP$motif1[klf2_minP$TF == "e93_klf2_4"] <- substr(e93_klf2, 151, 250)
klf2_minP$motif1[klf2_minP$TF == "e97_klf2_4"] <- substr(e97_klf2, 151, 250)

klf2_minP$motif1[klf2_minP$TF == "e6_klf2_5"] <- substr(e6_klf2, 201, 300)
klf2_minP$motif1[klf2_minP$TF == "e11_klf2_5"] <- substr(e11_klf2, 201, 300)
klf2_minP$motif1[klf2_minP$TF == "e19_klf2_5"] <- substr(e19_klf2, 201, 300)
klf2_minP$motif1[klf2_minP$TF == "e93_klf2_5"] <- substr(e93_klf2, 201, 300)
klf2_minP$motif1[klf2_minP$TF == "e97_klf2_5"] <- substr(e97_klf2, 201, 300)

klf2_minP$motif1[klf2_minP$TF == "e6_klf2_6"] <- substr(e6_klf2, 251, 350)
klf2_minP$motif1[klf2_minP$TF == "e11_klf2_6"] <- substr(e11_klf2, 251, 350)
klf2_minP$motif1[klf2_minP$TF == "e19_klf2_6"] <- substr(e19_klf2, 251, 350)
klf2_minP$motif1[klf2_minP$TF == "e93_klf2_6"] <- substr(e93_klf2, 251, 350)
klf2_minP$motif1[klf2_minP$TF == "e97_klf2_6"] <- substr(e97_klf2, 251, 350)

klf2_minP$motif1[klf2_minP$TF == "e6_klf2_7"] <- substr(e6_klf2, 301, 400)
klf2_minP$motif1[klf2_minP$TF == "e11_klf2_7"] <- substr(e11_klf2, 301, 400)
klf2_minP$motif1[klf2_minP$TF == "e19_klf2_7"] <- substr(e19_klf2, 301, 400)
klf2_minP$motif1[klf2_minP$TF == "e93_klf2_7"] <- substr(e93_klf2, 301, 400)
klf2_minP$motif1[klf2_minP$TF == "e97_klf2_7"] <- substr(e97_klf2, 301, 400)

klf2_minP$motif1[klf2_minP$TF == "e6_klf2_8"] <- substr(e6_klf2, 351, 450)
klf2_minP$motif1[klf2_minP$TF == "e11_klf2_8"] <- substr(e11_klf2, 351, 450)
klf2_minP$motif1[klf2_minP$TF == "e19_klf2_8"] <- substr(e19_klf2, 351, 450)
klf2_minP$motif1[klf2_minP$TF == "e93_klf2_8"] <- substr(e93_klf2, 351, 450)
klf2_minP$motif1[klf2_minP$TF == "e97_klf2_8"] <- substr(e97_klf2, 351, 450)


klf2_mCMV <- tf.df[tf.df$Promoter == "mCMV",]
klf2_mCMV <- klf2_mCMV[1:320,]
klf2_mCMV$Barcode[1:40] <- 1
klf2_mCMV$Barcode[41:80] <- 2
klf2_mCMV$Barcode[81:120] <- 3
klf2_mCMV$Barcode[121:160] <- 4
klf2_mCMV$Barcode[161:200] <- 5
klf2_mCMV$Barcode[201:240] <- 6
klf2_mCMV$Barcode[241:280] <- 7
klf2_mCMV$Barcode[281:320] <- 8
klf2_mCMV$TF <- enhancers
klf2_mCMV$Motif <- ""
klf2_mCMV$Spacing <- ""
klf2_mCMV$Distance <- ""
klf2_mCMV$Background <- ""
klf2_mCMV$Space1 <- ""
klf2_mCMV$motif2 <- ""
klf2_mCMV$Space2 <- ""
klf2_mCMV$motif3 <- ""
klf2_mCMV$Space3 <- ""
klf2_mCMV$motif4 <- ""
klf2_mCMV$Distance_seq <- ""


klf2_mCMV$motif1[klf2_mCMV$TF == "e6_klf2_1"] <- substr(e6_klf2, 1, 100)
klf2_mCMV$motif1[klf2_mCMV$TF == "e11_klf2_1"] <- substr(e11_klf2, 1, 100)
klf2_mCMV$motif1[klf2_mCMV$TF == "e19_klf2_1"] <- substr(e19_klf2, 1, 100)
klf2_mCMV$motif1[klf2_mCMV$TF == "e93_klf2_1"] <- substr(e93_klf2, 1, 100)
klf2_mCMV$motif1[klf2_mCMV$TF == "e97_klf2_1"] <- substr(e97_klf2, 1, 100)

klf2_mCMV$motif1[klf2_mCMV$TF == "e6_klf2_2"] <- substr(e6_klf2, 51, 150)
klf2_mCMV$motif1[klf2_mCMV$TF == "e11_klf2_2"] <- substr(e11_klf2, 51, 150)
klf2_mCMV$motif1[klf2_mCMV$TF == "e19_klf2_2"] <- substr(e19_klf2, 51, 150)
klf2_mCMV$motif1[klf2_mCMV$TF == "e93_klf2_2"] <- substr(e93_klf2, 51, 150)
klf2_mCMV$motif1[klf2_mCMV$TF == "e97_klf2_2"] <- substr(e97_klf2, 51, 150)

klf2_mCMV$motif1[klf2_mCMV$TF == "e6_klf2_3"] <- substr(e6_klf2, 101, 200)
klf2_mCMV$motif1[klf2_mCMV$TF == "e11_klf2_3"] <- substr(e11_klf2, 101, 200)
klf2_mCMV$motif1[klf2_mCMV$TF == "e19_klf2_3"] <- substr(e19_klf2, 101, 200)
klf2_mCMV$motif1[klf2_mCMV$TF == "e93_klf2_3"] <- substr(e93_klf2, 101, 200)
klf2_mCMV$motif1[klf2_mCMV$TF == "e97_klf2_3"] <- substr(e97_klf2, 101, 200)

klf2_mCMV$motif1[klf2_mCMV$TF == "e6_klf2_4"] <- substr(e6_klf2, 151, 250)
klf2_mCMV$motif1[klf2_mCMV$TF == "e11_klf2_4"] <- substr(e11_klf2, 151, 250)
klf2_mCMV$motif1[klf2_mCMV$TF == "e19_klf2_4"] <- substr(e19_klf2, 151, 250)
klf2_mCMV$motif1[klf2_mCMV$TF == "e93_klf2_4"] <- substr(e93_klf2, 151, 250)
klf2_mCMV$motif1[klf2_mCMV$TF == "e97_klf2_4"] <- substr(e97_klf2, 151, 250)

klf2_mCMV$motif1[klf2_mCMV$TF == "e6_klf2_5"] <- substr(e6_klf2, 201, 300)
klf2_mCMV$motif1[klf2_mCMV$TF == "e11_klf2_5"] <- substr(e11_klf2, 201, 300)
klf2_mCMV$motif1[klf2_mCMV$TF == "e19_klf2_5"] <- substr(e19_klf2, 201, 300)
klf2_mCMV$motif1[klf2_mCMV$TF == "e93_klf2_5"] <- substr(e93_klf2, 201, 300)
klf2_mCMV$motif1[klf2_mCMV$TF == "e97_klf2_5"] <- substr(e97_klf2, 201, 300)

klf2_mCMV$motif1[klf2_mCMV$TF == "e6_klf2_6"] <- substr(e6_klf2, 251, 350)
klf2_mCMV$motif1[klf2_mCMV$TF == "e11_klf2_6"] <- substr(e11_klf2, 251, 350)
klf2_mCMV$motif1[klf2_mCMV$TF == "e19_klf2_6"] <- substr(e19_klf2, 251, 350)
klf2_mCMV$motif1[klf2_mCMV$TF == "e93_klf2_6"] <- substr(e93_klf2, 251, 350)
klf2_mCMV$motif1[klf2_mCMV$TF == "e97_klf2_6"] <- substr(e97_klf2, 251, 350)

klf2_mCMV$motif1[klf2_mCMV$TF == "e6_klf2_7"] <- substr(e6_klf2, 301, 400)
klf2_mCMV$motif1[klf2_mCMV$TF == "e11_klf2_7"] <- substr(e11_klf2, 301, 400)
klf2_mCMV$motif1[klf2_mCMV$TF == "e19_klf2_7"] <- substr(e19_klf2, 301, 400)
klf2_mCMV$motif1[klf2_mCMV$TF == "e93_klf2_7"] <- substr(e93_klf2, 301, 400)
klf2_mCMV$motif1[klf2_mCMV$TF == "e97_klf2_7"] <- substr(e97_klf2, 301, 400)

klf2_mCMV$motif1[klf2_mCMV$TF == "e6_klf2_8"] <- substr(e6_klf2, 351, 450)
klf2_mCMV$motif1[klf2_mCMV$TF == "e11_klf2_8"] <- substr(e11_klf2, 351, 450)
klf2_mCMV$motif1[klf2_mCMV$TF == "e19_klf2_8"] <- substr(e19_klf2, 351, 450)
klf2_mCMV$motif1[klf2_mCMV$TF == "e93_klf2_8"] <- substr(e93_klf2, 351, 450)
klf2_mCMV$motif1[klf2_mCMV$TF == "e97_klf2_8"] <- substr(e97_klf2, 351, 450)


klf2_hBGm <- tf.df[tf.df$Promoter == "hBGm",]
klf2_hBGm <- klf2_hBGm[1:320,]
klf2_hBGm$Barcode[1:40] <- 1
klf2_hBGm$Barcode[41:80] <- 2
klf2_hBGm$Barcode[81:120] <- 3
klf2_hBGm$Barcode[121:160] <- 4
klf2_hBGm$Barcode[161:200] <- 5
klf2_hBGm$Barcode[201:240] <- 6
klf2_hBGm$Barcode[241:280] <- 7
klf2_hBGm$Barcode[281:320] <- 8
klf2_hBGm$TF <- enhancers
klf2_hBGm$Motif <- ""
klf2_hBGm$Spacing <- ""
klf2_hBGm$Distance <- ""
klf2_hBGm$Background <- ""
klf2_hBGm$Space1 <- ""
klf2_hBGm$motif2 <- ""
klf2_hBGm$Space2 <- ""
klf2_hBGm$motif3 <- ""
klf2_hBGm$Space3 <- ""
klf2_hBGm$motif4 <- ""
klf2_hBGm$Distance_seq <- ""


klf2_hBGm$motif1[klf2_hBGm$TF == "e6_klf2_1"] <- substr(e6_klf2, 1, 100)
klf2_hBGm$motif1[klf2_hBGm$TF == "e11_klf2_1"] <- substr(e11_klf2, 1, 100)
klf2_hBGm$motif1[klf2_hBGm$TF == "e19_klf2_1"] <- substr(e19_klf2, 1, 100)
klf2_hBGm$motif1[klf2_hBGm$TF == "e93_klf2_1"] <- substr(e93_klf2, 1, 100)
klf2_hBGm$motif1[klf2_hBGm$TF == "e97_klf2_1"] <- substr(e97_klf2, 1, 100)

klf2_hBGm$motif1[klf2_hBGm$TF == "e6_klf2_2"] <- substr(e6_klf2, 51, 150)
klf2_hBGm$motif1[klf2_hBGm$TF == "e11_klf2_2"] <- substr(e11_klf2, 51, 150)
klf2_hBGm$motif1[klf2_hBGm$TF == "e19_klf2_2"] <- substr(e19_klf2, 51, 150)
klf2_hBGm$motif1[klf2_hBGm$TF == "e93_klf2_2"] <- substr(e93_klf2, 51, 150)
klf2_hBGm$motif1[klf2_hBGm$TF == "e97_klf2_2"] <- substr(e97_klf2, 51, 150)

klf2_hBGm$motif1[klf2_hBGm$TF == "e6_klf2_3"] <- substr(e6_klf2, 101, 200)
klf2_hBGm$motif1[klf2_hBGm$TF == "e11_klf2_3"] <- substr(e11_klf2, 101, 200)
klf2_hBGm$motif1[klf2_hBGm$TF == "e19_klf2_3"] <- substr(e19_klf2, 101, 200)
klf2_hBGm$motif1[klf2_hBGm$TF == "e93_klf2_3"] <- substr(e93_klf2, 101, 200)
klf2_hBGm$motif1[klf2_hBGm$TF == "e97_klf2_3"] <- substr(e97_klf2, 101, 200)

klf2_hBGm$motif1[klf2_hBGm$TF == "e6_klf2_4"] <- substr(e6_klf2, 151, 250)
klf2_hBGm$motif1[klf2_hBGm$TF == "e11_klf2_4"] <- substr(e11_klf2, 151, 250)
klf2_hBGm$motif1[klf2_hBGm$TF == "e19_klf2_4"] <- substr(e19_klf2, 151, 250)
klf2_hBGm$motif1[klf2_hBGm$TF == "e93_klf2_4"] <- substr(e93_klf2, 151, 250)
klf2_hBGm$motif1[klf2_hBGm$TF == "e97_klf2_4"] <- substr(e97_klf2, 151, 250)

klf2_hBGm$motif1[klf2_hBGm$TF == "e6_klf2_5"] <- substr(e6_klf2, 201, 300)
klf2_hBGm$motif1[klf2_hBGm$TF == "e11_klf2_5"] <- substr(e11_klf2, 201, 300)
klf2_hBGm$motif1[klf2_hBGm$TF == "e19_klf2_5"] <- substr(e19_klf2, 201, 300)
klf2_hBGm$motif1[klf2_hBGm$TF == "e93_klf2_5"] <- substr(e93_klf2, 201, 300)
klf2_hBGm$motif1[klf2_hBGm$TF == "e97_klf2_5"] <- substr(e97_klf2, 201, 300)

klf2_hBGm$motif1[klf2_hBGm$TF == "e6_klf2_6"] <- substr(e6_klf2, 251, 350)
klf2_hBGm$motif1[klf2_hBGm$TF == "e11_klf2_6"] <- substr(e11_klf2, 251, 350)
klf2_hBGm$motif1[klf2_hBGm$TF == "e19_klf2_6"] <- substr(e19_klf2, 251, 350)
klf2_hBGm$motif1[klf2_hBGm$TF == "e93_klf2_6"] <- substr(e93_klf2, 251, 350)
klf2_hBGm$motif1[klf2_hBGm$TF == "e97_klf2_6"] <- substr(e97_klf2, 251, 350)

klf2_hBGm$motif1[klf2_hBGm$TF == "e6_klf2_7"] <- substr(e6_klf2, 301, 400)
klf2_hBGm$motif1[klf2_hBGm$TF == "e11_klf2_7"] <- substr(e11_klf2, 301, 400)
klf2_hBGm$motif1[klf2_hBGm$TF == "e19_klf2_7"] <- substr(e19_klf2, 301, 400)
klf2_hBGm$motif1[klf2_hBGm$TF == "e93_klf2_7"] <- substr(e93_klf2, 301, 400)
klf2_hBGm$motif1[klf2_hBGm$TF == "e97_klf2_7"] <- substr(e97_klf2, 301, 400)

klf2_hBGm$motif1[klf2_hBGm$TF == "e6_klf2_8"] <- substr(e6_klf2, 351, 450)
klf2_hBGm$motif1[klf2_hBGm$TF == "e11_klf2_8"] <- substr(e11_klf2, 351, 450)
klf2_hBGm$motif1[klf2_hBGm$TF == "e19_klf2_8"] <- substr(e19_klf2, 351, 450)
klf2_hBGm$motif1[klf2_hBGm$TF == "e93_klf2_8"] <- substr(e93_klf2, 351, 450)
klf2_hBGm$motif1[klf2_hBGm$TF == "e97_klf2_8"] <- substr(e97_klf2, 351, 450)


klf2_Random <- tf.df[tf.df$Promoter == "Random",]
klf2_Random <- klf2_Random[1:320,]
klf2_Random$Barcode[1:40] <- 1
klf2_Random$Barcode[41:80] <- 2
klf2_Random$Barcode[81:120] <- 3
klf2_Random$Barcode[121:160] <- 4
klf2_Random$Barcode[161:200] <- 5
klf2_Random$Barcode[201:240] <- 6
klf2_Random$Barcode[241:280] <- 7
klf2_Random$Barcode[281:320] <- 8
klf2_Random$TF <- enhancers
klf2_Random$Motif <- ""
klf2_Random$Spacing <- ""
klf2_Random$Distance <- ""
klf2_Random$Background <- ""
klf2_Random$Space1 <- ""
klf2_Random$motif2 <- ""
klf2_Random$Space2 <- ""
klf2_Random$motif3 <- ""
klf2_Random$Space3 <- ""
klf2_Random$motif4 <- ""
klf2_Random$Distance_seq <- ""


klf2_Random$motif1[klf2_Random$TF == "e6_klf2_1"] <- substr(e6_klf2, 1, 100)
klf2_Random$motif1[klf2_Random$TF == "e11_klf2_1"] <- substr(e11_klf2, 1, 100)
klf2_Random$motif1[klf2_Random$TF == "e19_klf2_1"] <- substr(e19_klf2, 1, 100)
klf2_Random$motif1[klf2_Random$TF == "e93_klf2_1"] <- substr(e93_klf2, 1, 100)
klf2_Random$motif1[klf2_Random$TF == "e97_klf2_1"] <- substr(e97_klf2, 1, 100)

klf2_Random$motif1[klf2_Random$TF == "e6_klf2_2"] <- substr(e6_klf2, 51, 150)
klf2_Random$motif1[klf2_Random$TF == "e11_klf2_2"] <- substr(e11_klf2, 51, 150)
klf2_Random$motif1[klf2_Random$TF == "e19_klf2_2"] <- substr(e19_klf2, 51, 150)
klf2_Random$motif1[klf2_Random$TF == "e93_klf2_2"] <- substr(e93_klf2, 51, 150)
klf2_Random$motif1[klf2_Random$TF == "e97_klf2_2"] <- substr(e97_klf2, 51, 150)

klf2_Random$motif1[klf2_Random$TF == "e6_klf2_3"] <- substr(e6_klf2, 101, 200)
klf2_Random$motif1[klf2_Random$TF == "e11_klf2_3"] <- substr(e11_klf2, 101, 200)
klf2_Random$motif1[klf2_Random$TF == "e19_klf2_3"] <- substr(e19_klf2, 101, 200)
klf2_Random$motif1[klf2_Random$TF == "e93_klf2_3"] <- substr(e93_klf2, 101, 200)
klf2_Random$motif1[klf2_Random$TF == "e97_klf2_3"] <- substr(e97_klf2, 101, 200)

klf2_Random$motif1[klf2_Random$TF == "e6_klf2_4"] <- substr(e6_klf2, 151, 250)
klf2_Random$motif1[klf2_Random$TF == "e11_klf2_4"] <- substr(e11_klf2, 151, 250)
klf2_Random$motif1[klf2_Random$TF == "e19_klf2_4"] <- substr(e19_klf2, 151, 250)
klf2_Random$motif1[klf2_Random$TF == "e93_klf2_4"] <- substr(e93_klf2, 151, 250)
klf2_Random$motif1[klf2_Random$TF == "e97_klf2_4"] <- substr(e97_klf2, 151, 250)

klf2_Random$motif1[klf2_Random$TF == "e6_klf2_5"] <- substr(e6_klf2, 201, 300)
klf2_Random$motif1[klf2_Random$TF == "e11_klf2_5"] <- substr(e11_klf2, 201, 300)
klf2_Random$motif1[klf2_Random$TF == "e19_klf2_5"] <- substr(e19_klf2, 201, 300)
klf2_Random$motif1[klf2_Random$TF == "e93_klf2_5"] <- substr(e93_klf2, 201, 300)
klf2_Random$motif1[klf2_Random$TF == "e97_klf2_5"] <- substr(e97_klf2, 201, 300)

klf2_Random$motif1[klf2_Random$TF == "e6_klf2_6"] <- substr(e6_klf2, 251, 350)
klf2_Random$motif1[klf2_Random$TF == "e11_klf2_6"] <- substr(e11_klf2, 251, 350)
klf2_Random$motif1[klf2_Random$TF == "e19_klf2_6"] <- substr(e19_klf2, 251, 350)
klf2_Random$motif1[klf2_Random$TF == "e93_klf2_6"] <- substr(e93_klf2, 251, 350)
klf2_Random$motif1[klf2_Random$TF == "e97_klf2_6"] <- substr(e97_klf2, 251, 350)

klf2_Random$motif1[klf2_Random$TF == "e6_klf2_7"] <- substr(e6_klf2, 301, 400)
klf2_Random$motif1[klf2_Random$TF == "e11_klf2_7"] <- substr(e11_klf2, 301, 400)
klf2_Random$motif1[klf2_Random$TF == "e19_klf2_7"] <- substr(e19_klf2, 301, 400)
klf2_Random$motif1[klf2_Random$TF == "e93_klf2_7"] <- substr(e93_klf2, 301, 400)
klf2_Random$motif1[klf2_Random$TF == "e97_klf2_7"] <- substr(e97_klf2, 301, 400)

klf2_Random$motif1[klf2_Random$TF == "e6_klf2_8"] <- substr(e6_klf2, 351, 450)
klf2_Random$motif1[klf2_Random$TF == "e11_klf2_8"] <- substr(e11_klf2, 351, 450)
klf2_Random$motif1[klf2_Random$TF == "e19_klf2_8"] <- substr(e19_klf2, 351, 450)
klf2_Random$motif1[klf2_Random$TF == "e93_klf2_8"] <- substr(e93_klf2, 351, 450)
klf2_Random$motif1[klf2_Random$TF == "e97_klf2_8"] <- substr(e97_klf2, 351, 450)



# Merge the dfs of the klf enhancers
klf2 <- rbind(klf2_minP, klf2_mCMV, klf2_hBGm, klf2_Random)

# Bind the klf enhancer df to the tf.df
tf.df <- rbind(tf.df, klf2)


# Export klf enhancers for adding the control to different libraries
filename <- SetFileName("_klf2", "mt")
setwd("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/")
write.csv(klf2, file = paste(filename,".csv", sep = ""), row.names = F)
```














## Checking TF binding scores of the library - only in front of the core promoter
```{r out.width= "100%", fig.align= "center", echo=FALSE}
# Select sequence in front of core promoter
tf.df.check <- tf.df
tf.df.check$seq.name <- paste(tf.df.check$TF, 
                    tf.df.check$Spacing, tf.df.check$Distance, 
                    tf.df.check$Background, sep = ".")
tf.df.check$seq.name[tf.df.check$TF == paste(enhancers, collapse = "|")] <- 
  paste(tf.df.check$TF[tf.df.check$TF == paste(enhancers, collapse = "|")])
tf.df.check$seq.name[tf.df.check$TF == "ctrl"] <- "ctrl"
tf.df.check$seq.text <- paste(tf.df.check$Primer1_seq, 
                    tf.df.check$motif1, tf.df.check$Space1, tf.df.check$motif2, 
                    tf.df.check$Space2,tf.df.check$motif3, 
                    tf.df.check$Space3, tf.df.check$motif4,
                    tf.df.check$Distance_seq, sep = "")
tf.df.check$seq.text[tf.df.check$TF == "ctrl"] <- tf.df.check$Promoter_sequence[tf.df.check$TF == "ctrl"]
tf.df.check <- tf.df.check %>% select(seq.name, seq.text)
tf.df.check <- unique(tf.df.check)
# dat2fasta(tf.df.check, outfile = "/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/tf_check_2.fasta")     
```


 
## Run FIMO script again 
```{bash run fimo db2_9, eval = FALSE}
# motfn=/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/20170320_pwms_selected.meme
# odir=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/tf_check_6
# query=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/tf_check_2.fasta

# nice -n 19 fimo --no-qvalue --thresh 1e-4 --verbosity 1 --o $odir $motfn $query 
```



## load fimo results
We built a TF motif matrix using -log10 transformed FIMO scores. We used this feature encoding throughout the rest of this analysis, unless otherwise stated. 

```{r build tf motif matrices db2: tf_check, out.width= "100%", fig.align= "center", echo=FALSE}
# load motif Metadata --> PWM feature matrix
tib_pwm_tf_check  <- get_pwm_feature_matrix(motif_meta_fn = '/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/fc181127_curated_metadata_no_composite_filt.csv', 
                                        fimo_fn       = '/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/tf_check_6/fimo.tsv',
                                        db            = 2)
```


```{r, fig.height=4, fig.width=4, out.width= "100%", fig.align= "center", echo=FALSE, eval=FALSE}
# Generate heatmaps in a loop for each TF
# Select only hits
tib_pwm_tf_check$TF <- gsub("[.*.|_].*","\\1",tib_pwm_tf_check$id)
tib_pwm_tf_check_2 <- tib_pwm_tf_check %>% 
  remove_rownames %>% column_to_rownames(var="id")

# De-select controls for now
tib_pwm_tf_check_2 <- tib_pwm_tf_check_2[tib_pwm_tf_check_2$TF != "ctrl",]
tib_pwm_tf_check_2 <- tib_pwm_tf_check_2[!grepl('klf2',tib_pwm_tf_check_2$TF),]

# Create breaks for heatmap (to make legend between heatmaps comparable)
myBreaks1 <- seq(0,8,0.08)

 # Prepare annotation for the heatmap rows
tib_pwm_tf_check_3 <- tib_pwm_tf_check_2
tib_pwm_tf_check_3$neg[grep("neg", rownames(tib_pwm_tf_check_3))] <- "disrupted"
tib_pwm_tf_check_3$neg[-grep("neg", rownames(tib_pwm_tf_check_3))] <- "high-affinity"
neg <- tib_pwm_tf_check_3[,which(colnames(tib_pwm_tf_check_3) == "neg"),drop=FALSE]
colors <- list(
   neg = c("high-affinity"= "#B3E2CD", "disrupted"="#FDCDAC"))


# Create heatmaps for each TF
for (i in unique(tib_pwm_tf_check_2$TF)) {
  data <- tib_pwm_tf_check_2[tib_pwm_tf_check_2$TF == i,]
  data <- data[,colSums(data != 0) > 0]
p <- pheatmap(as.matrix(data %>% select (-TF)), 
              main = paste(i),
              border_color = "#000000",
              annotation_row = neg,
              annotation_colors = colors,
              breaks = myBreaks1)
print(p)
}
```



## Look at only expressed TFs in mESCs
```{r out.width= "100%", fig.align= "center", echo=FALSE}
load('/DATA/usr/m.martinez.ara/GitLab/gurten/gurten/data/fc181121_epsure_tadec_rnaseq_expr_joshi.RData')

# Take mean of 2i expression
tib_fpkm_joshi <- tib_fpkm_joshi %>%
  select(-contains('serum')) %>%
  rowwise() %>%
  mutate(expr_mean = mean(c(twoi_1, twoi_2))) %>%
  ungroup()
```

## Filter expressed TFs
```{r out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# only select fpkm>0
tib_fpkm_joshi_expressed <- tib_fpkm_joshi %>% filter(expr_mean>2)

# # identify which TFs can't be found in the RNA-seq dataset
tib_pwm_tf_check$TF <- gsub("[.*.|_].*","\\1",tib_pwm_tf_check$id)
not_found_TFs <- colnames(tib_pwm_tf_check %>% select(-TF))
not_found_TFs_2 <- not_found_TFs[!not_found_TFs %in% tib_fpkm_joshi$symbol]
not_found_TFs_2[-1]
## Most of those seem to not be important/or redundant heterodimers - just change tp53/63/73
names(tib_pwm_tf_check)[names(tib_pwm_tf_check) == 'TP53'] <- 'TRP53'

# Select expressed TFs from pwm dataset
tib_pwm_tf_check_expressed <- tib_pwm_tf_check %>%  select(id, TF, one_of(tib_fpkm_joshi_expressed$symbol))
tib_pwm_tf_check_expressed_2 <- tib_pwm_tf_check_expressed %>% 
  remove_rownames %>% column_to_rownames(var="id")

# De-select controls for now
tib_pwm_tf_check_expressed_2 <- tib_pwm_tf_check_expressed_2[tib_pwm_tf_check_expressed$TF != "ctrl",]
tib_pwm_tf_check_expressed_2 <- tib_pwm_tf_check_expressed_2[!grepl('klf2',tib_pwm_tf_check_expressed$TF),]
```

# Again, generate heatmap, this time of only expressed TFs
```{r, fig.height=4, fig.width=4, out.width= "100%", fig.align= "center", echo=FALSE, eval=FALSE}
# Generate heatmaps in a loop for each TF
# Select only hits
tib_pwm_tf_check_expressed_2 <- tib_pwm_tf_check_expressed_2[tib_pwm_tf_check_expressed_2$TF != "Random2",]


# Prepare annotation for the heatmap rows
tib_pwm_tf_check_expressed_3 <- tib_pwm_tf_check_expressed_2
tib_pwm_tf_check_expressed_3$neg[grep("neg", rownames(tib_pwm_tf_check_expressed_3))] <- "disrupted"
tib_pwm_tf_check_expressed_3$neg[-grep("neg", rownames(tib_pwm_tf_check_expressed_3))] <- "high-affinity"
neg <- tib_pwm_tf_check_expressed_3[,which(colnames(tib_pwm_tf_check_expressed_3) == "neg"),drop=FALSE]
colors <- list(
   neg = c("high-affinity"= "#B3E2CD", "disrupted"="#FDCDAC"))



for (i in unique(tib_pwm_tf_check_expressed_2$TF)) {
  data <- tib_pwm_tf_check_expressed_2[tib_pwm_tf_check_expressed_2$TF == i,]
  data <- data[,colSums(data != 0) > 0]
p <- pheatmap(as.matrix(data %>% select (-TF)), 
              main = paste(i),
              border_color = "#000000",
              annotation_row = neg,
              annotation_colors = colors,
              breaks = myBreaks1)
print(p)
}
```









# Compare our binding specificity to the ones used in O'Connell et al. 2016 (TF-seq)
```{r out.width= "100%", fig.align= "center", echo=FALSE}
# Import synthesized sequences from TF-seq paper
tf_seq_oligos <- read.csv2("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/RegulatoryElements_OConnell2016.csv", header = T)
tf_seq_oligos <- tf_seq_oligos %>% select(Transcription.Factor, Synthesis.sequence)
tf_seq_oligos <- tf_seq_oligos[c(-58,-59),]
names(tf_seq_oligos) <- c("seq.name","seq.text")
# dat2fasta(tf_seq_oligos, outfile = "/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/tf_seq.fasta")
```

## Run FIMO script again 
```{bash run fimo db2_10, eval = FALSE}
# motfn=/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/20170320_pwms_selected.meme
# odir=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/tf_seq
# query=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/tf_seq.fasta

# nice -n 19 fimo --no-qvalue --thresh 1e-4 --verbosity 1 --o $odir $motfn $query 
```



## load fimo results
We built a TF motif matrix using -log10 transformed FIMO scores. We used this feature encoding throughout the rest of this analysis, unless otherwise stated. 

```{r build tf motif matrices db2: tf_seq, out.width= "100%", fig.align= "center", echo=FALSE}
# load motif Metadata --> PWM feature matrix
tib_pwm_tf_seq  <- get_pwm_feature_matrix(motif_meta_fn = '/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/fc181127_curated_metadata_no_composite_filt.csv', 
                                        fimo_fn       = '/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/tf_seq/fimo.tsv',
                                        db            = 2)
```


```{r, fig.height=4, fig.width=4, out.width= "100%", fig.align= "center", echo=FALSE, eval=FALSE}
# Generate heatmap
# Select only hits
tib_pwm_tf_seq$order <- 1
tib_pwm_tf_seq$order[11:20] <- 2
tib_pwm_tf_seq$order[21:30] <- 3
tib_pwm_tf_seq$order[31:40] <- 4
tib_pwm_tf_seq$order[41:50] <- 5
tib_pwm_tf_seq$order[51:56] <- 6
tib_pwm_tf_seq_2 <- tib_pwm_tf_seq %>% 
  remove_rownames %>% column_to_rownames(var="id")

# Split heatmap in pieces - only observe 10 reporters at a time
for (i in unique(tib_pwm_tf_seq_2$order)) {
  data <- tib_pwm_tf_seq_2[tib_pwm_tf_seq_2$order == i,]
  data <- data[,colSums(data != 0) > 0]
p <- pheatmap(as.matrix(data %>% select (-order)), 
              main = paste(i),
              border_color = "#000000",
              breaks = myBreaks1)
print(p)
}
```






## Compare specificity between my oligos and the TF-seq oligos
```{r, fig.height=10, fig.width=10, out.width= "100%", fig.align= "center", echo=FALSE}
# Motif occurencies (binary): how many TFs are binding, TF-seq vs. my approach
# Add column type
tib_pwm_tf_seq <- tib_pwm_tf_seq %>% mutate(Type= "TF-seq")
tib_pwm_tf_check_2 <- tib_pwm_tf_check%>% mutate(Type= "My-TFs")
tib_pwm_tf_check_2 <- tib_pwm_tf_check_2[-grep("neg", tib_pwm_tf_check_2$id),]
tib_pwm_tf_check_2 <- tib_pwm_tf_check_2%>% select(-TF)
tib_pwm_all <- bind_rows(tib_pwm_tf_seq, tib_pwm_tf_check_2)

# binary
tib_pwm_all_binary <- tib_pwm_all %>% mutate_if(is.numeric, function(x) ifelse((x==0 | is.na(x)),0,1)) 
tib_pwm_all_binary$MotifOccurencies <- rowSums(tib_pwm_all_binary %>% select(-id, -Type))


# Generate plot
tib_pwm_all_binary%>%
  select(id,Type,MotifOccurencies)%>%
  ggplot(aes(Type, MotifOccurencies, fill=Type))+
  geom_dotplot(binaxis = "y", stackdir = "center",binwidth = 1, alpha=0.7)+
  stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
               geom = "crossbar", width = 0.5)+
  ggtitle("Motif Occurencies per oligo:
my oligos vs. TF-seq oligos")+
  theme_bw(base_size = 15) + scale_y_continuous(breaks=seq(0,100, 10))
```


# Compare only overlapping TFs in the two datasets
```{r}
# Filter for only overlapping TFs in both datasets
tf_seq_overlapping <- c("CREB1", "ELK1", "GLI1", "IRF3::IRF3", "NFE2L2" ,"NFYA",
                        "NFKB1/RELA", "NICD::CBF1", "NR3C1", "RARA", "SMAD2::SMAD4",
                        "SP1", "STAT3::STAT3", "TRP53")
tib_pwm_tf_seq <- tib_pwm_tf_seq[grep(paste(tf_seq_overlapping, collapse = "|"), tib_pwm_tf_seq$id),]
my_tf_overlapping <- c("Creb1", "Elk1", "Gli1", "Irf3", "Nfe2l2" ,"Nfya",
                        "Nfkb1", "Rbpj", "Nr3c1", "Rara", "Smad4",
                        "Sp1", "Stat3", "Trp53")
tib_pwm_tf_check <- tib_pwm_tf_check[grep(paste(my_tf_overlapping, collapse = "|"), tib_pwm_tf_check$id),]
tib_pwm_tf_check <- tib_pwm_tf_check%>% mutate(Type= "My-TFs")
tib_pwm_tf_check <- tib_pwm_tf_check[-grep("neg", tib_pwm_tf_check$id),]
tib_pwm_tf_check <- tib_pwm_tf_check%>% select(-TF)

# Combine dfs and compare
tib_pwm_all <- bind_rows(tib_pwm_tf_seq, tib_pwm_tf_check)

# binary
tib_pwm_all_binary <- tib_pwm_all %>% mutate_if(is.numeric, function(x) ifelse((x==0 | is.na(x)),0,1)) 
tib_pwm_all_binary$MotifOccurencies <- rowSums(tib_pwm_all_binary %>% select(-id, -Type))


# Generate plot
tib_pwm_all_binary%>%
  select(id,Type,MotifOccurencies)%>%
  ggplot(aes(Type, MotifOccurencies, fill=Type))+
  geom_dotplot(binaxis = "y", stackdir = "center",binwidth = 1, alpha=0.7)+
  stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
               geom = "crossbar", width = 0.5)+
  ggtitle("Motif Occurencies per oligo:
my oligos vs. TF-seq oligos")+
  theme_bw(base_size = 15) + scale_y_continuous(breaks=seq(0,100, 10))



# Take a closer look at heatmaps to confirm
tib_pwm_all$TF <- gsub("[.*.].*", "\\1", tib_pwm_all$id)
new_names <- c("Creb1", "Elk1", "Gli1", "Irf3", "Nfe2l2", "Nfkb1", 
               "Nfya", "Rbpj", "Nr3c1(GR)", "Rara", "Smad4", "Sp1", 
               "Stat3", "Trp53")
tib_pwm_all$TF[1:14] <-new_names
tib_pwm_all[is.na(tib_pwm_all)] <- 0
tib_pwm_all <- tib_pwm_all %>% 
  remove_rownames %>% column_to_rownames(var="id")

# Prepare annotation for the heatmap rows
type <- tib_pwm_all[,which(colnames(tib_pwm_all) == "Type"),drop=FALSE]
colors <- list(
  Type = c(`TF-seq`= "#B3E2CD", `My-TFs`="#FDCDAC"))




# Generate heatmap with both groups
for (i in unique(tib_pwm_all$TF)) {
  data <- tib_pwm_all[tib_pwm_all$TF == i,]
  data <- data[,colSums(data != 0) > 0]
p <- pheatmap(as.matrix(data %>% select (-TF, -Type)), 
              main = paste(i),
              annotation_row = type,
              annotation_colors = colors,
              border_color = "#000000",
              # breaks = myBreaks1)
)
print(p)
}
```
## So: My oligos seem to perform well compared to the TF-seq oligos, even though they achieve somewhat higher specificity for some reporters -> how?






# Compare binding strength for correct TF 








# Compare published backgrounds (Davis et al. 2019, biorxiv) with my new backgrounds
```{r out.width= "100%", fig.align= "center", echo=FALSE}
#bg41:
bg41 <- "TGTTCAGAAGGGCCAGAAATGCCAAGGACTCAGGGGAGGAGAATTAAGTCAGAGAGTTTCATTACTGAGTGTTGTTTGACTTTGTTGTCACGGATTCATTTAACCATCTCTCTACCATGGTAAAAATGTGTATCCTATGTCCAGTATGAA"
bg52 <- "CCAGGGAAAGCAGTCGGTGAGACCAGGCACAGTAGGATAGTTAGTTAGCTACCACATGTAAAGCTGAGAGCAGATGGTGCCGTTGATATAGAGCGCAGGAATGTGCGTGTTTATGCGGGTGCGTTTCTGTGCGTGTGCGTGTGGAAACCA"

# Create whole df by merging all conditions
tf.df_davis <- Reduce(function(x, y) merge(x, y, all=TRUE), list(tf.df.frame, spacings, distances,
                                                            barcodes, promoters, background))

# Add random TF motifs
tf.df_davis$Motif <- as.character(tf.df_davis$Motif)
set.seed(423)
tf.df_davis$Motif[tf.df_davis$TF == "Random1"] <- random_motif_2[1]
tf.df_davis$Motif[tf.df_davis$TF == "Random2"] <- random_motif_2[2]
tf.df_davis$Motif[tf.df_davis$TF == "Random3"] <- random_motif_2[3]

tf.df_davis <- tf.df_davis[tf.df_davis$Background != 3,]

# Adding the DNA sequence from 5' to 3'

## Constant 5' primer sequence
## Add 5' primer sequence: This one had the best primer qualities
tf.df_davis$Primer1_seq <- motif_primer_1_G[5]

# Add motifs to position 1
tf.df_davis$motif1 <- tf.df_davis$Motif

## Spacer sequence between TF motifs (5, 10, 21 bp)
tf.df_davis$Space1[tf.df_davis$Spacing == "5bp"] <- substr(bg41, 0, 5)
tf.df_davis$Space1[tf.df_davis$Spacing == "10bp"] <- substr(bg41, 0, 10)
tf.df_davis$Space1[tf.df_davis$Spacing == "5bp" & tf.df_davis$Background == 2] <- substr(bg52, 0, 5)
tf.df_davis$Space1[tf.df_davis$Spacing == "10bp" & tf.df_davis$Background == 2] <- substr(bg52, 0, 10)

## Do the same for the other 3 repeats
tf.df_davis$motif2 <- tf.df_davis$Motif
tf.df_davis$Space2[tf.df_davis$Spacing == "5bp"] <- substr(bg41, 6, 10)
tf.df_davis$Space2[tf.df_davis$Spacing == "10bp"] <- substr(bg41, 11, 20)
tf.df_davis$Space2[tf.df_davis$Spacing == "5bp" & tf.df_davis$Background == 2] <- substr(bg52, 6, 10)
tf.df_davis$Space2[tf.df_davis$Spacing == "10bp" & tf.df_davis$Background == 2] <- substr(bg52, 11, 20)
tf.df_davis$motif3 <- tf.df_davis$Motif

## The third spacing and the fourth motif is added
# tf.df_davis$Space3[tf.df_davis$Spacing == "5bp"] <- 
#   ifelse (nchar(tf.df_davis$motif1[tf.df_davis$Spacing == "5bp"]) < 14, substr(bg41, 11, 15),"")
tf.df_davis$Space3[tf.df_davis$Spacing == "5bp"] <- substr(bg41, 11, 15)
# tf.df_davis$Space3[tf.df_davis$Spacing == "5bp" & tf.df_davis$Background == 2] <- 
#   ifelse (nchar(tf.df_davis$motif1[tf.df_davis$Spacing == "5bp" & tf.df_davis$Background == 2]) < 14, 
#           substr(bg52, 11, 15),"")
tf.df_davis$Space3[tf.df_davis$Spacing == "5bp" & tf.df_davis$Background == 2] <- substr(bg52, 11, 15)


# tf.df_davis$Space3[tf.df_davis$Spacing == "10bp"] <- 
#   ifelse (nchar(tf.df_davis$motif1[tf.df_davis$Spacing == "10bp"]) < 14, substr(bg41, 21, 30),"")  
tf.df_davis$Space3[tf.df_davis$Spacing == "10bp"] <- substr(bg41, 21, 30)
# tf.df_davis$Space3[tf.df_davis$Spacing == "10bp" & tf.df_davis$Background == 2] <- 
#   ifelse (nchar(tf.df_davis$motif1[tf.df_davis$Spacing == "10bp" & tf.df_davis$Background == 2]) < 14, 
#           substr(bg52, 21, 30),"")  
tf.df_davis$Space3[tf.df_davis$Spacing == "10bp" & tf.df_davis$Background == 2] <- substr(bg52, 21, 30)

# tf.df_davis$motif4 <- 
#   ifelse (nchar(tf.df_davis$motif1) < 14, as.character(tf.df_davis$Motif),"")
tf.df_davis$motif4 <- as.character(tf.df_davis$Motif)
  
  
# Sequence from last TF-motif to start of minimal promoter (10, 21)
tf.df_davis$Distance_seq[tf.df_davis$Distance == "10bp" & tf.df_davis$Spacing == "5bp"] <- substr(bg41, 16, 25)
tf.df_davis$Distance_seq[tf.df_davis$Distance == "10bp" & tf.df_davis$Spacing == "10bp"] <- substr(bg41, 31, 40)
tf.df_davis$Distance_seq[tf.df_davis$Distance == "21bp" & tf.df_davis$Spacing == "5bp"] <- substr(bg41, 16, 36)
tf.df_davis$Distance_seq[tf.df_davis$Distance == "21bp" & tf.df_davis$Spacing == "10bp"] <- substr(bg41, 31, 51)
tf.df_davis$Distance_seq[tf.df_davis$Distance == "10bp" & tf.df_davis$Spacing == "5bp" & 
                           tf.df_davis$Background == 2] <- substr(bg52, 16, 25)
tf.df_davis$Distance_seq[tf.df_davis$Distance == "10bp" & tf.df_davis$Spacing == "10bp" & 
                           tf.df_davis$Background == 2] <- substr(bg52, 31, 40)
tf.df_davis$Distance_seq[tf.df_davis$Distance == "21bp" & tf.df_davis$Spacing == "5bp" & 
                           tf.df_davis$Background == 2] <- substr(bg52, 16, 36)
tf.df_davis$Distance_seq[tf.df_davis$Distance == "21bp" & tf.df_davis$Spacing == "10bp" & 
                           tf.df_davis$Background == 2] <- substr(bg52, 31, 51)
```



## Checking TF binding scores of the library - only in front of transcription unit (as promoters might have messy TF binding)
```{r out.width= "100%", fig.align= "center", echo=FALSE}
tf.df_davis$seq.name <- paste(tf.df_davis$TF, 
                    tf.df_davis$Spacing, tf.df_davis$Distance, 
                    tf.df_davis$Background, sep = "_")
tf.df_davis$seq.text <- paste(tf.df_davis$Primer1_seq, 
                    tf.df_davis$motif1, tf.df_davis$Space1, tf.df_davis$motif2, 
                    tf.df_davis$Space2,tf.df_davis$motif3, 
                    tf.df_davis$Space3, tf.df_davis$motif4,
                    tf.df_davis$Distance_seq, sep = "")
tf.df_davis <- tf.df_davis %>% select(seq.name, seq.text)
tf.df_davis <- unique(tf.df_davis)
# dat2fasta(tf.df_davis, outfile = "/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/tf_bg_davis.fasta")     
```


 
## Run FIMO script again 
```{bash run fimo db2_11, eval = FALSE}
# motfn=/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/20170320_pwms_selected.meme
# odir=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/tf_bg_davis_3
# query=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/tf_bg_davis.fasta

# nice -n 19 fimo --no-qvalue --thresh 1e-4 --verbosity 1 --o $odir $motfn $query 
```



## load fimo results
We built a TF motif matrix using -log10 transformed FIMO scores. We used this feature encoding throughout the rest of this analysis, unless otherwise stated. 

```{r build tf motif matrices db2: tf_bg_davis, out.width= "100%", fig.align= "center", echo=FALSE}
# load motif Metadata --> PWM feature matrix
tib_pwm_tf_bg_davis  <- get_pwm_feature_matrix(motif_meta_fn = '/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/fc181127_curated_metadata_no_composite_filt.csv', 
                                        fimo_fn       = '/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/tf_bg_davis_3/fimo.tsv',
                                        db            = 2)
```


```{r, fig.height=4, fig.width=4, out.width= "100%", fig.align= "center", echo=FALSE, eval=FALSE}
# Generate heatmaps in a loop for each TF
# Select only hits
tib_pwm_tf_bg_davis$TF <- gsub("[.*_].*","\\1",tib_pwm_tf_bg_davis$id)
tib_pwm_tf_bg_davis_2 <- tib_pwm_tf_bg_davis %>% 
  remove_rownames %>% column_to_rownames(var="id")

for (i in unique(tib_pwm_tf_bg_davis_2$TF)) {
  data <- tib_pwm_tf_bg_davis_2[tib_pwm_tf_bg_davis_2$TF == i,]
  data <- data[,colSums(data != 0) > 0]
p <- pheatmap(as.matrix(data %>% select (-TF)), 
              main = paste(i),
              border_color = "#000000",
              breaks = myBreaks1)
print(p)
}
```



## Compare specificity between my random backgrounds and 2 of the backgrounds used in Davis et al
```{r, fig.height=10, fig.width=10, out.width= "100%", fig.align= "center", echo=FALSE}
# Motif occurencies (binary): how many TFs are binding, TF-seq vs. my approach
# Add column type
tib_pwm_tf_bg_davis <- tib_pwm_tf_bg_davis %>% mutate(Type = "Davis-bg")
tib_pwm_tf_check_3 <- tib_pwm_tf_check %>% mutate(Type = "My-bg")
tib_pwm_tf_bg_davis <- tib_pwm_tf_bg_davis[-grep("neg", tib_pwm_tf_bg_davis$id),]
# tib_pwm_tf_bg_davis <- tib_pwm_tf_bg_davis %>% select(-TF)
tib_pwm_bg_all <- bind_rows(tib_pwm_tf_bg_davis, tib_pwm_tf_check_3)

# binary
tib_pwm_all_bg_binary <- tib_pwm_bg_all %>% mutate_if(is.numeric, function(x) ifelse((x==0 | is.na(x)),0,1)) 
tib_pwm_all_bg_binary$MotifOccurencies <- rowSums(tib_pwm_all_bg_binary %>% select(-Type, -id))

# Generate plot
tib_pwm_all_bg_binary%>%
  select(id,Type,MotifOccurencies)%>%
  ggplot(aes(Type, MotifOccurencies, fill=Type))+
  geom_dotplot(binaxis = "y", stackdir = "center",binwidth = 1, alpha=0.7)+
  stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
               geom = "crossbar", width = 0.5)+
  ggtitle("Motif Occurencies per oligo:
Davis background vs. my background")+
  theme_bw(base_size = 15)
```













## Filter barcode list - remove EcoRI/NheI sites
```{r out.width= "100%", fig.align= "center", echo=FALSE}
## Remove barcodes that create EcoRI/NheI sites in overap with S1 primer & 3' primer sequences 
# Import all created 12bp barcodes: 80.901 unique barcodes
barcode_selection <- as.data.frame(barc)
barcode_selection$S1 <- "CACGACGCTCTTCCGATCT"
barcode_selection$Primer2_seq <- motif_primer_2_G[11]
barcode_selection$seq <- paste(barcode_selection$S1, barcode_selection$barc, 
                            barcode_selection$Primer2_seq, sep = "")
barcode_selection <- barcode_selection[-grep(paste(ecori_nhei, collapse = "|"), barcode_selection$seq),]
barcode_list <- barcode_selection$barc
# 80.737 barcodes are left after filtering
```




## Filter barcode list - only keep inactive barcodes
```{r out.width= "100%", fig.align= "center", echo=FALSE}
# Assemble sequence to test: S1 primer sequence - barcode - primer 2 adapter 
barcode_selection$seq.name <- 1:length(barcode_selection$seq)
barcode_selection$seq.text <- paste(substrRight(barcode_selection$S1, 4), 
                       barcode_selection$barc, 
                       substr(barcode_selection$Primer2_seq, 1, 4), sep = "")
barcode_selection <- barcode_selection[barcode_selection$barc %in% barcode_list,]

# Write fasta file to run on FIMO script
barcode_selection_export <- barcode_selection
barcode_selection_export <- barcode_selection_export[,c(-1:-4)]
# dat2fasta(barcode_selection_export, outfile = "/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/barcodes.fasta") 
```


## Run FIMO script again 
```{bash run fimo db2_12, eval = FALSE}
# motfn=/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/20170320_pwms_selected.meme
# odir=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/barcodes
# query=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/barcodes.fasta

# nice -n 19 fimo --no-qvalue --thresh 1e-4 --verbosity 1 --o $odir $motfn $query 
```



## load fimo results
We built a TF motif matrix using -log10 transformed FIMO scores. We used this feature encoding throughout the rest of this analysis, unless otherwise stated. 

```{r build tf motif matrices db2: barcodes, out.width= "100%", fig.align= "center", echo=FALSE}
# load motif Metadata --> PWM feature matrix
tib_pwm_barcodes  <- get_pwm_feature_matrix(motif_meta_fn = '/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/fc181127_curated_metadata_no_composite_filt.csv', 
                                        fimo_fn       = '/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/fimo/barcodes/fimo.tsv',
                                        db            = 2)
```


## visualize fimo results
```{r out.width= "100%", fig.align= "center", echo=FALSE}
## Remove barcodes that have TF binding
# convert to binary
id <- tib_pwm_barcodes$id 
tib_pwm_barcodes_binary <- tib_pwm_barcodes %>% mutate_if(is.numeric, function(x) ifelse((x==0 | is.na(x)),0,1)) 
tib_pwm_barcodes_binary$id <- id
tib_pwm_barcodes_binary_top <- tib_pwm_barcodes_binary

# compute rowsums to get cumulative binding
tib_pwm_barcodes_binary_top$binding <- rowSums(tib_pwm_barcodes_binary_top[,2:ncol(tib_pwm_barcodes_binary_top)])

# select only cumulative binding and id
tib_pwm_barcodes_binary_top <- tib_pwm_barcodes_binary_top %>%
  select(id,binding)

# make third column wiht only the space_id - then sum up the binding scores per space_id
tib_pwm_barcodes_binary_top$space <- gsub(".*_(.*)", "\\1", tib_pwm_barcodes_binary_top$id)
tib_pwm_barcodes_binary_top <- tib_pwm_barcodes_binary_top %>%
  select(-id)
tib_pwm_barcodes_binary_top$cum_binding <- ave(tib_pwm_barcodes_binary_top$binding, tib_pwm_barcodes_binary_top$space,
                                             FUN = sum)
tib_pwm_barcodes_binary_top <- tib_pwm_barcodes_binary_top %>%
  select(-binding)
tib_pwm_barcodes_binary_top <- unique(tib_pwm_barcodes_binary_top)

# ~50% of the barcodes trigger TF binding - some of them even bind a lot of TFs - remove all barcodes that bind TFs
barc_list <- barcode_selection %>% select(barc, seq.name)
barc_list <- barc_list[!barc_list$seq.name %in% tib_pwm_barcodes_binary_top$space,]
barc_list <- barc_list$barc
# 39.669 barcodes are remaining

# Export these barcodes for further use
filename <- SetFileName("_barcodes_validated", "mt")

# Write csv file
setwd("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/")
write.csv(barc_list, file = paste(filename,".csv", sep = ""), row.names = F)
```


## Add barcodes
```{r out.width= "100%", fig.align= "center", echo=FALSE}
## Randomizing barcodes
set.seed(123)
barc_list <- sample(barc_list)
tf.df$barcode <- barc_list[1:nrow(tf.df)]

# Check if there are any duplicate barcodes
paste("duplicate barcodes: ", nrow(tf.df[duplicated(tf.df$barcode),]), sep ="")

# Save for later export
tf.df.long.export <- tf.df
```
















# Check-up & Visualizations
```{r out.width= "100%", fig.align= "center", echo=FALSE}
## Get main characteristics of designed oligo pool
# TF motif length distribution
tf.df2 <- unique(tf.df[,c(1,2)])
tf.df2 <- tf.df2[1:29,]


ggplot(tf.df2, aes(x = nchar(Motif))) +
   geom_histogram(aes(y=..count..), colour="black", fill="#E69F00", binwidth = 1)+ 
  xlab("Length TF motif") + ylab("Frequency") +
    labs(title = "Overview of TF motif lengths", 
         subtitle = paste("Total TFs =", length(unique(tf.df2$TF)), "|",  
         "max motif length =", max(nchar(tf.df2$Motif)), "|",
         "min motif length = ", min(nchar(tf.df2$Motif)), "|",
         "mean motif length = ", round(mean(nchar(tf.df2$Motif))))) +
  theme_classic() + scale_x_continuous(breaks=seq(0,18,1)) + 
  scale_y_continuous(breaks=seq(0,6,1)) 






# Length distribution
# Save intermediate df for later purpose
tf.array <- tf.df
tf.df$seq.text <- paste(tf.df$Primer1_seq, tf.df$motif1, tf.df$Space1, tf.df$motif2,
                    tf.df$Space2, tf.df$motif3, tf.df$Space3, tf.df$motif4, 
                    tf.df$Distance_seq, tf.df$Promoter_sequence, tf.df$S1_primer,
                    tf.df$barcode, tf.df$Primer2_seq, sep = "")
x <- tf.df[grep(paste(ecori_nhei, collapse = "|"),tf.df$seq.text),]
tf.df <- tf.df[,c(-8:-20)]
tf.df$nchar <- nchar(tf.df$seq.text)

ggplot(tf.df, aes(x = nchar)) +
   geom_histogram(aes(y=..count..), colour="black", fill="#E69F00", binwidth = 5)+ 
  xlab("Length Oligo (binwidth = 5)") + ylab("Frequency") +
    labs(title = "Overview of oligo lengths in pool", 
         subtitle = paste("Total oligos =", nrow(tf.df), "|",
                          "max oligo length =", max(nchar(tf.df$seq.text)), "|",
                          "min oligo length = ", min(nchar(tf.df$seq.text)), "|",
                          "mean oligo length = ", round(mean(nchar(tf.df$seq.text))))) +
  theme_classic()
```


















# Exporting data
```{r out.width= "100%", fig.align= "center", echo=FALSE}
# Export sequences for ordering DNA oligos
## Rename the parameters to make them more clear
tf.df.export <- tf.df
tf.df.export$Spacing[-grep("klf2|ctrl", tf.df.export$TF)] <- 
  paste("s", tf.df.export$Spacing, sep = "-")
tf.df.export$Distance[-grep("klf2|ctrl", tf.df.export$TF)] <- 
  paste("d", tf.df.export$Distance, sep = "-")
tf.df.export$Barcode <- paste("bc", tf.df.export$Barcode, sep = "-")
tf.df.export$Background[-grep("klf2|ctrl", tf.df.export$TF)] <- 
  paste("bg", tf.df.export$Background, sep = "-")

## Generate a single name for each oligo
tf.df.export$seq.name <- 
  gsub("(_)+", "_", with(tf.df.export,paste(tf.df.export$TF, tf.df.export$Promoter, 
                                            tf.df.export$Spacing, tf.df.export$Distance,
                                            tf.df.export$Background, tf.df.export$Barcode, sep = "_")))

## Exporting
tf.df.export <- tf.df.export %>% select(seq.name, seq.text)
filename <- SetFileName("_oligo_pool", "mt")

# Write csv file
setwd("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/")
write.csv(tf.df.export, file = paste(filename,".csv", sep = ""), row.names = F)

# Write fasta file
setwd("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/")
dat2fasta(tf.df.export, outfile = paste(filename,".fasta", sep = ""))  

# Write .csv file for tf-array
setwd("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/TF-reporter_library_design/output/")
filename <- SetFileName("_tf-array", "mt")
write.csv(tf.df.long.export, file = paste(filename,".csv", sep = ""), row.names = F)
```

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

