---
title: "pDNA insert matching"
author: "Max Trauernicht"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: 
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---

# knitr document van Steensel lab


# Introduction
I sequenced the complete insert of the pDNA library of pMT02. I already extracted all sequences in front of the 3' adapter from the sequences data and added counts to identical sequences by starcode. I now want to make an overview about how many pDNA insert sequences in the pDNA still match the designed inserts. 


## Description of Data
How to make a good rendering table: 
```{r table1, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
tabl <- "
| column1	|	column2	|	column3	|	
|----|----|----|
|1	|	2	|	3	|	
|a	|	b	|	c	|	
"
cat(tabl) # output the table in a format good for HTML/PDF/docx conversion
```

# Data processing
## Path, Libraries, Parameters and Useful Functions
```{r setup, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(seqinr)
library(ShortRead)
library(plyr)
library(maditr)
library(phylotools)
library(tidyr)
library(readr)
library(dplyr)
library(ggplot2)
library(ggbeeswarm)
library(vwr)
library(d3r)
library(sunburstR)
library(LncFinder)
library(plotly)
```

### Custom functions
Functions used thoughout this script.
```{r, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}

ReadFasta<-function(file) {
   # Read the file line by line
   fasta<-readLines(file)
   # Identify header lines
   ind<-grep(">", fasta)
   # Identify the sequence lines
   s<-data.frame(ind=ind, from=ind+1, to=c((ind-1)[-1], length(fasta)))
   # Process sequence lines
   seqs<-rep(NA, length(ind))
   for(i in 1:length(ind)) {
      seqs[i]<-paste(fasta[s$from[i]:s$to[i]], collapse="")
   }
   # Create a data frame 
   DF<-data.frame(name=gsub(">", "", fasta[ind]), sequence=seqs)
   # Return the data frame as a result object from the function
   return(DF)
}

# Function to load PWM matrix
get_pwm_feature_matrix <- function(motif_meta_fn, fimo_fn, db = 2) {

  # validate args
  valid_dbs <- 1:2
  if(!db %in% valid_dbs)
    stop('Invalid db (database version). Please use db=1 (maintained for backward compatibility only) or db=2')

  # db=1 is maintained for backward compatibility only
  if(db == 1) {

    # read in motif metadata
    motif_meta    <- read.csv(motif_meta_fn)

    # check whether motif metadata contain essential annotations
    if(!all(c('PWM.ID', 'Cognate.TF') %in% colnames(motif_meta))) {
      message('The motif metadata file does not contain the essential columns PWM.ID and Cognate.TF')
    }

    motif_minimal <- motif_meta[, c('PWM.ID', 'Cognate.TF')]

    # load fimo output --> extract motif id, sequence id and p-value
    df <- read.table(fimo_fn)
    df <- df[, c(1, 2, 7)]

    colnames(df) <- c('PWM.ID', 'seqid', 'pval')

    # add TF id
    df <- merge(df, motif_minimal, by = 'PWM.ID')

    # group motif hits by sequence id
    l <- split(df, df[['seqid']])

    # multiple PWM and multiple hits possible. Reduce hits to one per TF, keeping best p-val only
    l <- lapply(l, function(x) {
      x_by_tf <- split(x, x[['Cognate.TF']], drop = TRUE)
      x_by_tf <- lapply(x_by_tf, function(y) y[which.min(y$pval), ])
      do.call('rbind', x_by_tf)
    })

    # initialize feature matrix
    n_tf          <- motif_minimal[['Cognate.TF']] %>%
      unique %>%
      length
    n_seq         <- length(l)
    pwm           <- matrix(1, nrow = n_seq, ncol = n_tf)
    colnames(pwm) <- (motif_minimal[['Cognate.TF']] %>% unique)

    # replace :: from names of composite motifs
    colnames(pwm) <- str_replace_all(colnames(pwm), '::', '_')

    # fill in feature matrix
    for(i in 1 : n_seq) {
      pwm[i, l[[i]][['Cognate.TF']]] <- l[[i]]$pval
    }

    # -log10 transform
    pwm           <- -1 * log10(pwm)

    # coerce to tib and return
    tib_fimo <- as_data_frame(pwm) %>%
      mutate(id = names(l))
      dplyr::select(id, everything())

  }

  # db = 2 (default)
  else {

    # load metadata
    tib_meta    <- read_csv(motif_meta_fn) %>%
      # extract tf symbol from motif id (Cognate_TF unsafe, it can be empty) and replace :: occurrences
      mutate(tf_symbol = str_remove(ID, '_[0-9]*'),
             tf_symbol = str_replace(tf_symbol, '::', '_')) %>%
      select(motif_id = `PWM ID`, tf_symbol)

    # load fimo results
    tib_fimo <- read_tsv(fimo_fn) %>%
      # extract motif id, sequence id and p-value
      select(motif_id, sequence_name, pval = `p-value`)

    # add tf symbol to fimo results
    tib_fimo <- tib_fimo %>%
      left_join(tib_meta, by = 'motif_id') %>%
      # remove hits with missing motif id (composite pwms)
      filter(!is.na(tf_symbol))

    # select best hit for each motif and sequence
    tib_fimo <- tib_fimo %>%
      group_by(sequence_name, tf_symbol) %>%
      dplyr::slice(which.min(pval)) %>%
      ungroup()

    # spread into feature matrix
    tib_fimo <- tib_fimo %>%
      mutate(pval = -1 * log10(pval)) %>%
      select(-motif_id) %>%
      spread(key = tf_symbol, value = pval, fill = 0, drop = TRUE) %>%
      # perform cosmetics on the id
      mutate(id = sequence_name) %>%
      select(-c(sequence_name)) %>%
      select(id, everything())

  }

  return(tib_fimo)

}
```


## Data import
```{r, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Import reference sequences
ref_seq <- ReadFasta("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/output/mt20191205_oligo_pool.fasta")


# Remove adapters from reference sequence (cause these are not in the sequencing data)
ref_seq$sequence <- gsub("CGGAGCGAACCGAGTTAG", "", ref_seq$sequence)
ref_seq$sequence <- gsub("CATCGTCGCATCCAAGAG", "", ref_seq$sequence)

# Split up in insert and barcode part
ref_seq$barcode <- gsub(".*([A-Z]{12})$", "\\1", ref_seq$sequence)
ref_seq$insert <- gsub("(.*)[A-Z]{12}$", "\\1", ref_seq$sequence)

# Import sequencing files
pDNA_seq <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/pDNA_insert_seq/pDNA-seq-starcode.tsv", col_names = c("sequence", "number"))
pDNA_seq$barcode <- gsub(".*([A-Z]{12})$", "\\1", pDNA_seq$sequence)
pDNA_seq$insert <- gsub("(.*)[A-Z]{12}$", "\\1", pDNA_seq$sequence)

# Filter pDNA data - we should have at least 15 pDNA counts - we want to be sure that we're looking at relevant things
pDNA_seq <- pDNA_seq[pDNA_seq$number >= 15,]



pDNA_seq_clust <- read_tsv("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/pDNA_insert_seq/pDNA-seq-starcode-d1.tsv", col_names = c("sequence", "number"))
pDNA_seq_clust$barcode <- gsub(".*([A-Z]{12})$", "\\1", pDNA_seq_clust$sequence)
pDNA_seq_clust$insert <- gsub("(.*)[A-Z]{12}$", "\\1", pDNA_seq_clust$sequence)
```

# Analysis


## Match barcodes
```{r, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Match barcodes with original data
ref_seq_bc <- ref_seq %>% select(name, barcode)
pDNA_seq_bc <- pDNA_seq %>% select(barcode) %>% unique() %>% mutate(id = "pDNA")
match_bc <- merge(ref_seq_bc, pDNA_seq_bc, by = "barcode", all = T)
```





## Plot how many barcodes are found in pDNA data
```{r, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
## Identify the unmapped fraction
match_df_ref <- match_bc[!is.na(match_bc$name),]
n_match <- nrow(match_df_ref[!is.na(match_df_ref$id),])
n_nomatch <- nrow(match_df_ref[is.na(match_df_ref$id),])

# Create donut chart
data <- data.frame(
  lbls=c("matched:", "unmatched:"),
  count=c(n_match, n_nomatch)
)
 
# Compute percentages
data$fraction <- data$count / sum(data$count)
data$percentage <- data$fraction * 100


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)

# Compute the bottom of each rectangle
data$ymin <- c(0, head(data$ymax, n=-1))

# Compute label position
data$labelPosition <- (data$ymax + data$ymin) / 2

# Compute a good label
data$label <- paste0(data$lbls, "\n", round(data$percentage), "%")

# Make the plot
ggplot(data, aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=lbls)) +
  geom_rect() +
  geom_text(x=2, aes(y=labelPosition, label=label, color=lbls), size=5) + # x here controls label position (inner / outer)
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  scale_color_manual(values = c("#1B998B", "#2D3047")) +
  labs(title = "percentage of barcodes found back in pDNA-seq") +
  coord_polar(theta="y") +
  xlim(c(0, 4)) +
  theme_void() +
  theme(legend.position = "none")
```



## How many raw complete sequences match with the design?
```{r, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Match barcodes with original data
ref_seq_seq <- ref_seq %>% select(name, sequence)
pDNA_seq_seq <- pDNA_seq %>% select(sequence) %>% unique() %>% mutate(id = "pDNA")
match_seq <- merge(ref_seq_seq, pDNA_seq_seq, by = "sequence", all = T)

## Identify the unmapped fraction
match_df_ref_seq <- match_seq[!is.na(match_seq$name),]
n_match <- nrow(match_df_ref_seq[!is.na(match_df_ref_seq$id),])
n_nomatch <- nrow(match_df_ref_seq[is.na(match_df_ref_seq$id),])

# Create donut chart
data <- data.frame(
  lbls=c("matched:", "unmatched:"),
  count=c(n_match, n_nomatch)
)
 
# Compute percentages
data$fraction <- data$count / sum(data$count)
data$percentage <- data$fraction * 100


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)

# Compute the bottom of each rectangle
data$ymin <- c(0, head(data$ymax, n=-1))

# Compute label position
data$labelPosition <- (data$ymax + data$ymin) / 2

# Compute a good label
data$label <- paste0(data$lbls, "\n", round(data$percentage), "%")

# Make the plot
ggplot(data, aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=lbls)) +
  geom_rect() +
  geom_text(x=2, aes(y=labelPosition, label=label, color=lbls), size=5) + # x here controls label position (inner / outer)
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  scale_color_manual(values = c("#1B998B", "#2D3047")) +
  labs(title = "percentage of inserts found back in pDNA-seq") +
  coord_polar(theta="y") +
  xlim(c(0, 4)) +
  theme_void() +
  theme(legend.position = "none")
```




## Now we want to know the read distribution of matched/unmatched sequences
```{r, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
## Only select barcodes from design in pDNA data
pDNA_seq_insert <- pDNA_seq[pDNA_seq$barcode %in% ref_seq$barcode,] %>% select(sequence, number)
ref_seq_insert <- ref_seq %>% select(name, sequence)
match_insert <- merge(pDNA_seq_insert, ref_seq_insert, by = "sequence", all = T)

## Identify the unmapped fraction
match_df_reads <- match_insert[!is.na(match_insert$number),]
n_match <- sum(match_df_reads$number[!is.na(match_df_reads$name)])
n_nomatch <- sum(match_df_reads$number[is.na(match_df_reads$name)])

# Create donut chart
data <- data.frame(
  lbls=c("matched:", "unmatched:"),
  count=c(n_match, n_nomatch)
)
 
# Compute percentages
data$fraction <- data$count / sum(data$count)
data$percentage <- data$fraction * 100


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)

# Compute the bottom of each rectangle
data$ymin <- c(0, head(data$ymax, n=-1))

# Compute label position
data$labelPosition <- (data$ymax + data$ymin) / 2

# Compute a good label
data$label <- paste0(data$lbls, "\n", round(data$percentage), "%")

# Make the plot
ggplot(data, aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=lbls)) +
  geom_rect() +
  geom_text(x=2, aes(y=labelPosition, label=label, color=lbls), size=5) + # x here controls label position (inner / outer)
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  scale_color_manual(values = c("#1B998B", "#2D3047")) +
  labs(title = "percentage of reads from unmatched inserts (only matched barcodes)") +
  coord_polar(theta="y") +
  xlim(c(0, 4)) +
  theme_void() +
  theme(legend.position = "none")

## Histogram of read distributions matched vs. unmatched
match_df_reads$match <- "Yes"
match_df_reads$match[is.na(match_df_reads$name)] <- "No"

ggplot(match_df_reads, aes(x=number, color=match)) +
  geom_histogram(fill="white", position="dodge", bins = 500)+
  theme(legend.position="top")+
  scale_color_grey()+scale_fill_grey() +
  theme_classic()
```


## Correlate to GC contenct
```{r}
# Load reference file
ref_seq <- seqinr::read.fasta("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/output/mt20191205_oligo_pool.fasta")

# Compute gc contents
gc <- compute_GC(ref_seq)

# Plot gc distribution
density <- density(gc$GC.content)

plot_ly(x = ~density$x, y = ~density$y, type = 'scatter', mode = 'lines', fill = 'tozeroy') %>% 
  layout(xaxis = list(title = 'GC content'),
         yaxis = list(title = 'Density'))


match_seq <- match_insert[!is.na(match_insert$name),]
gc <- gc %>% rownames_to_column(var = "name")
gc <- merge(gc, match_seq)
gc[is.na(gc$number),] <- 0
gc <- gc[apply(gc[,c(-1,-3,-5)], 1, function(x) !all(x==0)),]
gc$number <- ave(gc$number, FUN = function(x) x/sum(x) *1000000)




plot_ly(data = gc, x = ~GC.content, y = ~number) %>% 
  layout(xaxis = list(title = 'GC content'),
         yaxis = list(title = 'reads per million'))
```



# Barcodes attached to wrong insert?
```{r, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Select only matched barcodes
pDNA_seq_insert <- pDNA_seq[pDNA_seq$barcode %in% ref_seq$barcode,]

# Add bc-id to the barcodes found in pDNA
ref_seq_insert <- ref_seq %>% select(barcode, name) %>% setnames("name", "bc-match")
pDNA_seq_insert <- merge(pDNA_seq_insert, ref_seq_insert)

# Add insert-id to the inserts found in pDNA
ref_seq_insert <- ref_seq %>% select(insert, name) %>% setnames("name", "insert-match")
ref_seq_insert$`insert-match` <- gsub("(.*)_bc-[0-9]$", "\\1", ref_seq_insert$`insert-match`)
pDNA_seq_insert <- merge(pDNA_seq_insert, ref_seq_insert)

# Count occurences where bc matches insert
pDNA_seq_insert$`bc-match` <- gsub("(.*)_bc-[0-9]$", "\\1", pDNA_seq_insert$`bc-match`)
for (i in 1:nrow(pDNA_seq_insert)) {
  pDNA_seq_insert$match[i] <- identical(pDNA_seq_insert$`bc-match`[i], pDNA_seq_insert$`insert-match`[i])
  
}





## Identify the unmapped fraction
n_match <- sum(pDNA_seq_insert$number[pDNA_seq_insert$match == T])
n_nomatch <- sum(pDNA_seq_insert$number[pDNA_seq_insert$match == F])

# Create donut chart
data <- data.frame(
  lbls=c("matched:", "unmatched:"),
  count=c(n_match, n_nomatch)
)
 
# Compute percentages
data$fraction <- data$count / sum(data$count)
data$percentage <- data$fraction * 100


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)

# Compute the bottom of each rectangle
data$ymin <- c(0, head(data$ymax, n=-1))

# Compute label position
data$labelPosition <- (data$ymax + data$ymin) / 2

# Compute a good label
data$label <- paste0(data$lbls, "\n", round(data$percentage), "%")

# Make the plot
ggplot(data, aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=lbls)) +
  geom_rect() +
  geom_text(x=2, aes(y=labelPosition, label=label, color=lbls), size=5) + # x here controls label position (inner / outer)
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  scale_color_manual(values = c("#1B998B", "#2D3047")) +
  labs(title = "percentage of reads where bc attached to wrong insert ") +
  coord_polar(theta="y") +
  xlim(c(0, 4)) +
  theme_void() +
  theme(legend.position = "none")


ggplot(pDNA_seq_insert, aes(x=number, fill=match)) +
  geom_histogram(position="dodge", bins = 50)+
  theme(legend.position="top")+
  scale_fill_grey()+
  labs(title = "Histogram: Read count distribution", 
       subtitle = "bc attached to correct insert vs. attached to wrong insert")+
  ylab("counts") + xlab("read counts in pDNA-seq data")+
  theme_bw()
```

# Identify those barcodes that are attached to a wrong insert
Clearly wrongly assigned barcodes can be assigned to the correct insert
Barcodes that are attached to a mixed population of inserts should to be excluded from any analysis where this plasmid library was used

```{r, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
mismatch_df <- pDNA_seq_insert[pDNA_seq_insert$match == F,] %>% select(barcode, number) %>% unique()
mismatch_df <- ddply(mismatch_df,~barcode, summarise, number = sum(number))

match_df <- pDNA_seq_insert[pDNA_seq_insert$match == T,]  %>% select(barcode, number) %>% unique()
match_df <- ddply(match_df,~barcode, summarise, number = sum(number))
setnames(match_df, "number", "match")
setnames(mismatch_df, "number", "mismatch")
matching_df <- merge(match_df, mismatch_df, all = T)
matching_df$match[is.na(matching_df$match)] <- 0.1
matching_df$mismatch[is.na(matching_df$mismatch)] <- 0.1

# Calculate match/mismatch ratio
matching_df$ratio <- matching_df$match / matching_df$mismatch
matching_df$type <- "good matching"
matching_df$type[log2(matching_df$ratio) < 5 & log2(matching_df$ratio) >= -5] <- "unclear matching"
matching_df$type[log2(matching_df$ratio) < -5] <- "incorrect matching"

# Plot match/mismatch distribution
ggplot(matching_df, 
       aes(x = match, y = mismatch, colour = type), alpha = 0.5) +
  geom_point(size = 1) + 
  xlim(0,270) + ylim(0,270) +
  xlab("number of reads: barcodes match with insert") + 
  ylab("number of reads: barcodes do not match insert") +
  scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71")) +
  theme_bw() + theme(legend.position = c(.8, .8))


ggplot(matching_df, aes(x = "match/mismatch ratio", y = log2(ratio), colour = type), alpha = 0.5) +
  geom_quasirandom() + theme_bw() + ylim(-10,15) + 
  scale_color_manual(values = c("#1B998B", "#2D3047", "#FF9B71")) +
  theme(legend.position = c(.8, .8))

# Fraction of barcodes with match/mismatch
n_match <- nrow(matching_df[matching_df$mismatch == 0.1,])
n_nomatch <- nrow(matching_df[matching_df$mismatch > 0.1,])

# Create donut chart
data <- data.frame(
  lbls=c("matched:", "unmatched:"),
  count=c(n_match, n_nomatch)
)
 
# Compute percentages
data$fraction <- data$count / sum(data$count)
data$percentage <- data$fraction * 100


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)


# Compute the cumulative percentages (top of each rectangle)
data$ymax <- cumsum(data$fraction)

# Compute the bottom of each rectangle
data$ymin <- c(0, head(data$ymax, n=-1))

# Compute label position
data$labelPosition <- (data$ymax + data$ymin) / 2

# Compute a good label
data$label <- paste0(data$lbls, "\n", round(data$percentage), "%")

# Make the plot
ggplot(data, aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=lbls)) +
  geom_rect() +
  geom_text(x=2, aes(y=labelPosition, label=label, color=lbls), size=5) + # x here controls label position (inner / outer)
  scale_fill_manual(values = c("#1B998B", "#2D3047")) +
  scale_color_manual(values = c("#1B998B", "#2D3047")) +
  labs(title = "percentage at least 1 bc wrongly attached to its insert") +
  coord_polar(theta="y") +
  xlim(c(0, 4)) +
  theme_void() +
  theme(legend.position = "none")
```


# Barcode re-evaluation 
```{r, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
# Barcodes with a mixed match of correct and incorrect insert need to be excluded from the analysis
matching_df_exclude <- matching_df[log2(matching_df$ratio) < 5 & log2(matching_df$ratio) >= -5,]

# Barcodes with incorrect matching can be assigned to new insert if only 1 insert has all barcodes
matching_df_incorrect <- matching_df[log2(matching_df$ratio) < -5,]



# Overview of type of barcode attachment distributio
# Fraction of barcodes with match/mismatch
n_total <- nrow(ref_seq)
n_match <- nrow(matching_df)
n_nomatch <- n_total - n_match
n_correct <- nrow(matching_df[log2(matching_df$ratio) >= 5,])
n_exclude <- nrow(matching_df[log2(matching_df$ratio) < 5 & log2(matching_df$ratio) >= -5,])
n_incorrect <- nrow(matching_df[log2(matching_df$ratio) < -5,])


dat <- data.frame(
  x = rep("x", each = 4),
  condition = c("4. not enough data","3. correct match", "2. ambiguous match", "1. incorrect match"),
  size = c(n_nomatch, n_correct, n_exclude, n_incorrect),
  stringsAsFactors = T
)

dat$percentage <- 100*(round(dat$size/n_total,2))

dat$pos <- c(16.5, 65, 98, 98.5)

ggplot(dat, aes(fill=condition, y=percentage, x = x)) + 
    geom_bar(position="stack", stat = "identity") +
  theme_bw()+scale_fill_grey()+ 
  geom_text(data=dat, 
            aes(x = x, y = pos,
                label = paste0(percentage,"%")), size=4) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())

```






# Investigate the mutational load of the barcodes with a good match
```{r, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
## Only select barcodes from design in pDNA data
matching_df_correct <- matching_df[matching_df$ratio > 2,]
pDNA_seq_insert <- pDNA_seq[pDNA_seq$barcode %in% matching_df$barcode,] %>% 
  select(barcode, insert, number) %>%
  setnames("insert", "pDNA")
ref_seq_insert <- ref_seq %>% select(barcode, insert) %>% setnames("insert", "ref")
match_insert <- merge(pDNA_seq_insert, ref_seq_insert, by = "barcode")
match_insert$lv <- levenshtein.distance(match_insert$pDNA, match_insert$ref)

## Read distribution vs. TF reporter length
ggplot(data = match_insert, aes(x = lv)) + 
  geom_histogram(color = "#2D3047") + xlab("Levenshtein distance") + 
  labs(title = "Sum pDNA counts vs. at levenshtein distances") + 
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 12), 
          axis.text.y = element_text(size = 12)) +
  ylab("read count")
```


# Investigate mutational load of only Trp53 constructs (as they are especially complex to PCR up)
```{r, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
## Only select barcodes from design in pDNA data
matching_df_correct <- matching_df[matching_df$ratio > 2,]
pDNA_seq_insert <- pDNA_seq[pDNA_seq$barcode %in% matching_df$barcode,] %>% 
  select(barcode, insert, number) %>%
  setnames("insert", "pDNA")
p53_bc <- ref_seq$barcode[grep("Trp53", ref_seq$name)]
pDNA_seq_insert <- pDNA_seq_insert[pDNA_seq_insert$barcode %in% p53_bc,]
ref_seq_insert <- ref_seq %>% select(barcode, insert) %>% setnames("insert", "ref")
match_insert <- merge(pDNA_seq_insert, ref_seq_insert, by = "barcode")
match_insert$lv <- levenshtein.distance(match_insert$pDNA, match_insert$ref)

## Read distribution vs. TF reporter length
ggplot(data = match_insert, aes(x = lv)) + 
  geom_histogram(color = "#2D3047") + xlab("Levenshtein distance of designed insert vs sequenced insert") + 
  labs(title = "Trp53 reporters: is the sequence correct in the pDNA?") + 
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 12), 
          axis.text.y = element_text(size = 12)) +
  ylab("reporter counts")
```



## Exporting data
```{r}
# Export barcodes that are attached to multiple inserts
bc_exclude <- matching_df_exclude$barcode %>% unique()
write.csv(bc_exclude, "/DATA/usr/m.trauernicht/projects/SuRE-TF/data/pDNA_insert_seq/bc_exclude.csv")

# Export barcodes that are attached to the wrong insert
bc_replace <- pDNA_seq_incorrect %>% select(barcode, `bc-match`, `insert-match`) %>% unique()
write.csv(bc_replace, "/DATA/usr/m.trauernicht/projects/SuRE-TF/data/pDNA_insert_seq/bc_replace.csv")
```

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

