---
title: "Barcode processing - TF reporter library - generation 2 - mES"
author: 
  - name: "Max Trauernicht"
    email: "m.trauernicht@nki.nl"
    affiliation: "Netherlands Cancer Institute - van Steensel lab"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: 
  html_document:
    theme: united
    highlight: pygments
    fig_caption: yes
    code_folding: hide
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
---

---

### Aim
18,000 TF reporters were transfected into mESCs and stimulated in different conditions, sequencing data yielded barcode counts of these experiments. These counts will be processed in this script. 

---

## Setup {.tabset}

<!-- little HTML script to do indentation of the table of contents -->
<script>
    $(document).ready(function() {
      $items = $('div#TOC li');
      $items.each(function(idx) {
        num_ul = $(this).parentsUntil('#TOC').length;
        $(this).css({'text-indent': num_ul * 10, 'padding-left': 0});
      });

    });
</script>

```{css, echo = FALSE}
div.sourceCode {
  overflow-x: hidden;
}
```


### Libraries 

```{r setup, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(data.table)
library(plyr)
library(stringr)
library(ggpubr)
library(GGally)
library(vwr)
library(dplyr)
library(tibble)
library(plotly)
library(ggbeeswarm)
library(haven)
library(readr)
library(parallel)
library(RColorBrewer)
library(gridExtra)
library(LncFinder)
library(tidyr)
library(grr)
library(viridis)
library(DESeq2)
library(PCAtools)
library(pheatmap)
```


### Functions

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Custom functions
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}

ReadFasta<-function(file) {
   # Read the file line by line
   fasta<-readLines(file)
   # Identify header lines
   ind<-grep(">", fasta)
   # Identify the sequence lines
   s<-data.frame(ind=ind, from=ind+1, to=c((ind-1)[-1], length(fasta)))
   # Process sequence lines
   seqs<-rep(NA, length(ind))
   for(i in 1:length(ind)) {
      seqs[i]<-paste(fasta[s$from[i]:s$to[i]], collapse="")
   }
   # Create a data frame 
   DF<-data.frame(name=gsub(">", "", fasta[ind]), sequence=seqs)
   # Return the data frame as a result object from the function
   return(DF)
}

# From Fede:
# ggpairs custom functions
corColor <- function(data, mapping, color = I("black"), sizeRange = c(1, 3), ...) {

  x   <- eval_data_col(data, mapping$x)
  y   <- eval_data_col(data, mapping$y)
  r   <- cor(x, y, "pairwise.complete.obs")
  rt  <- format(r, digits = 3)
  tt  <- as.character(rt)
  cex <- max(sizeRange)

  # helper function to calculate a useable size
  percent_of_range <- function(percent, range) {
    percent * diff(range) + min(range, na.rm = TRUE)
  }

  # plot correlation coefficient
  p <- ggally_text(label = tt, mapping = aes(), xP = 0.5, yP = 0.5,
                   size = I(percent_of_range(cex * abs(r), sizeRange)), color = color, ...) +
    theme(panel.grid.minor=element_blank(),
          panel.grid.major=element_blank())

  corColors <- RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")[2:6]

  if (r <= boundaries[1]) {
    corCol <- corColors[1]
  } else if (r <= boundaries[2]) {
    corCol <- corColors[2]
  } else if (r < boundaries[3]) {
    corCol <- corColors[3]
  } else if (r < boundaries[4]) {
    corCol <- corColors[4]
  } else {
    corCol <- corColors[5]
  }

  p <- p +
    theme(panel.background = element_rect(fill = corCol))

  return(p)
}
```


### Loading data

```{r data import, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Import barcode counts per experiment
## Load all extracted 13 bp read counts
bc_files = list.files('/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6578_gen2_stimulation-2/results/',
                       full.names=T, patter='*_barcode_counts.tsv')
bc_files <- bc_files[-grep("pMT02", bc_files)]
bc_list_lib_1 <- lapply(bc_files, fread, header = FALSE)
names(bc_list_lib_1)<- gsub('.*//6578_[0-9]{1,2}_(.*?)_[CGAT]{6}.*_barcode_counts.tsv', 
                                    '\\1', 
                                    bc_files)

bc_annotation <- read.csv("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/library_design/gen-2/mt20210326_oligo_pool_gen2.csv", header = T) %>%
  mutate(barcode = gsub(".*([A-Z]{13})CATCGTCGCATCCAAGAG$", "\\1", seq.text),
         tf = gsub("(^.*)_s-.*", "\\1", seq.name),
         spacing = gsub("^.*_s-(.*bp)_d.*", "\\1", seq.name),
         distance = gsub("^.*_d-(.*bp).*", "\\1", seq.name),
         barcode_number = gsub("^.*_bc-(.*)", "\\1", seq.name),
         promoter = gsub("^.*_bg-[1-3]{1}_(.*)_bc.*", "\\1", seq.name),
         background = gsub("^.*_bg-(.*)_.*_.*", "\\1", seq.name)) %>%
  dplyr::select(-seq.name, -seq.text)

# Annotate positive controls
bc_annotation$commercial_reporter <- "No"
bc_annotation$commercial_reporter[grep("TF-seq|romanov|promega", bc_annotation$promoter)] <- "Yes"

bc_annotation$tf[bc_annotation$commercial_reporter == "Yes"] <- gsub("(^.*_TF-seq|romanov|promega|promega_long)_.*", "\\1",
                                                                                 bc_annotation$tf[bc_annotation$commercial_reporter == "Yes"])
bc_annotation$spacing[bc_annotation$commercial_reporter == "Yes"] <- ""
bc_annotation$background[bc_annotation$commercial_reporter == "Yes"] <- ""
bc_annotation$promoter[bc_annotation$commercial_reporter == "Yes"] <- gsub(".*bp_(.*)_bc.*", "\\1", bc_annotation$promoter[bc_annotation$commercial_reporter == "Yes"])

bc_annotation$tf<- gsub("ETS2", "ETS1", bc_annotation$tf)
```


### Creating count data frames

```{r cluster_compare, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Generate one long data frame from the list of data frames and combine with barcode information
for (i in 1:length(bc_list_lib_1)) {
  if (i == 1) {
  bc_df <- data.frame(bc_list_lib_1[i])
  bc_df[3] <- names(bc_list_lib_1[i])
  names(bc_df) <- c("barcode", "starcode_counts", "sample")
  }
  else {
  bc_df_i <- data.frame(bc_list_lib_1[i])
  bc_df_i[3] <- names(bc_list_lib_1[i])
  names(bc_df_i) <- c("barcode", "starcode_counts", "sample")
  bc_df <- rbind(bc_df, bc_df_i)
  }
}

bc_df <- merge(bc_df, bc_annotation, all = T)


bc_df$starcode_counts[is.na(bc_df$starcode_counts)] <- 0

# First compute reads per million to estimate the relative counts in their respective sample
bc_df <- bc_df %>%
  mutate(rpm = ave(starcode_counts, sample, FUN = function(x) (x+1) / sum(x) *1e6 ))

bc_df_filt <- bc_df[!is.na(bc_df$tf),]
bc_df_unmatched <- bc_df[is.na(bc_df$tf),]
```


### Compare differently clustered pDNA data

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Make dataframe to compare raw counts between conditions
bc_df_filt <- bc_df_filt[!is.na(bc_df_filt$sample),]
bc_df_cor <- bc_df_filt %>%
  dplyr::select(-starcode_counts)

bc_df_cor <- bc_df_cor %>%
  mutate(rpm = as.integer(rpm)) %>%
  spread(sample, rpm) 

bc_df_cor[is.na(bc_df_cor)] <- 0


## First compare differently processed pDNA sequencing data
# Compare library 1
plot(bc_df_cor$pMT09_r1_r2, bc_df_cor$pMT09_r3)

# Compare different samples

## K562 rep 2 vs rep 3
plot(bc_df_cor$`LIF_PD_r3`, bc_df_cor$`LIF_PD_r2`)

# Correlation matrix plot
n <- sample(1:nrow(bc_df_cor), 5000)
boundaries <- seq(from = 0.8, by = 0.05, length.out = 4)
plt <- ggpairs(bc_df_cor[,9:28],
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.1, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red") +
                   theme_bw()}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red") +
                   theme_bw()})) +
  ggtitle("Correlation Between All Samples") +
  # theme(text = element_text(size = 20)) +
  xlab("Counts") +
  ylab("Counts")

print(plt)





# Convert into data.frame with row.names for deseq2
bc_df_cor2 <- bc_df_cor[,c(1,9:26)] %>%
  column_to_rownames("barcode")


#######################################
## Initialize deseq2

# Prepare metadata ready for deseq
metadata <- data.frame("sample_name" = names(bc_df_cor2)) %>%
  separate(sample_name, c("condition", "condition_2", "replicate"), sep = "_", remove = F) %>%
  mutate(condition = paste(condition, condition_2, sep = "_")) %>%
  dplyr::select(-condition_2) %>%
  mutate(replicate = gsub(".*_(.*$)", "\\1", sample_name)) %>%
  mutate(condition = gsub("_r[1-3]$", "", condition))

# Initialize
rnaseq_dds <- DESeqDataSetFromMatrix(countData = bc_df_cor2,
                                     colData = metadata,
                                     design= ~condition)

# Execute deseq2
rnaseq_dds <- DESeq(rnaseq_dds)


#######################################
## PCA analysis - quality control

# Get the "normalized" values and create PCA plot
# Note that "normalized" is simply log2 + 0.01 transformed normalized counts!
rnaseq_dds_norm <- normTransform(rnaseq_dds, pc = 0.01)
plt <- plotPCA(rnaseq_dds_norm, intgroup = c("condition", "replicate"), ntop = 50)

plt + 
  theme_bw() +
  theme(aspect.ratio = 1) +
  scale_shape_manual()

## PCA plot alternative
rnaseq_dds_norm <- assay(vst(rnaseq_dds))
p <- pca(bc_df_cor2, metadata = metadata %>% 
           column_to_rownames("sample_name") %>%
           mutate(cell = paste(replicate, condition, sep = "_")), removeVar = 0.1)
biplot(p, 
       colby = "condition", 
       shape = "replicate", 
       legendPosition = "right",
       legendLabSize = 8,
       legendTitleSize = 8,
       legendIconSize = 3, 
       lab = "")


## Correlation heatmap
bc_df_cor2$mean <- rowMeans(bc_df_cor2)
bc_df_cor2 <- bc_df_cor2 %>%
  filter(mean > 200) %>%
  dplyr::select(-mean)

cor_df <- cor(bc_df_cor2, method = "pearson")

pheatmap(cor_df, border_color = "black")
```



### Annotate data frame

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
## Annotate controls
# Annotate the mutated motif of each TF
bc_df$neg_ctrls <- "No"
bc_df$neg_ctrls[grep("neg", bc_df$tf)] <- "Yes"

# Annotate hPGK postive control
bc_df$hPGK <- "No"
bc_df$hPGK[grep("hPGK", bc_df$tf)] <- "Yes"

# Annotate random promoter control
bc_df$rand_promoter <- "No"
bc_df$rand_promoter[grep("Random", bc_df$promoter)] <- "Yes"

# Add replicate and condition columns
bc_df$replicate <- gsub("(_r[1-3]$)", "\\1", bc_df$sample)
bc_df$condition <- gsub("^r[1-3]_(.*)", "\\1", bc_df$sample)



bc_df <- bc_df[!is.na(bc_df$sample),]
```




## Data quality plots

```{r read count 2, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message= FALSE}
# I want to show the following:
## 1: Read distribution of matched barcodes vs. unmatched barcode

### a: total read counts per sample
plot_ly(bc_df[!is.na(bc_df$condition),] %>%
         mutate(sum_counts = ave(starcode_counts, sample, FUN = function(x) sum(x) / 36000)) %>%
         dplyr::select(sample, sum_counts) %>%
         unique(), 
        x = ~sum_counts, y = ~sample, type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Number of reads per barcode per sample",
         xaxis = list(title = "Expected number of reads per barcode per sample"),
         yaxis = list(title = "Sample"))


### b: read counts per TF
bc_df_filt <- bc_df[!is.na(bc_df$tf),]
ggplot(bc_df_filt %>%
         filter(neg_ctrls == "No", commercial_reporter == "No", hPGK == "No", rand_promoter == "No"), aes(x = rpm, y = tf)) +
  geom_violin() +
  theme_pubr(border = T) +   
  theme(axis.text.y = element_text(size = 6)) +
  xlim(0,5000) +
  facet_wrap(~condition)

ggplot(bc_df_filt %>%
         filter(neg_ctrls == "No", commercial_reporter == "No", hPGK == "No", rand_promoter == "No"), aes(x = rpm, y = tf)) +
  geom_violin() +
  theme_pubr(border = T) +   
  theme(axis.text.y = element_text(size = 6)) +
  xlim(0,1000) +
  facet_wrap(~condition)

bc_df_2 <- bc_df[bc_df$rpm <= 250,]
bc_df_2 <- bc_df_2[bc_df_2$rpm >= 0.5,]
bc_df_2 <- bc_df_2[!is.na(bc_df_2$tf),]

ggplot(bc_df_2, aes(x = rpm)) +
  geom_histogram(binwidth = 20) +
  theme_bw() +
  xlim(0,250)+
  ylim(0,6000)+
  facet_wrap(~sample)+
  theme(strip.background =element_rect(fill="#D6D5C9"))

ggplot(bc_df[bc_df$rpm >= 500 & !is.na(bc_df$tf),], aes(x = rpm)) +
  geom_histogram(binwidth = 20) +
  theme_bw() +
  xlim(250,1000)+
  ylim(0,50)+
  facet_wrap(~sample)+
  theme(strip.background =element_rect(fill="#D6D5C9"))

n_highly_expressed <- data.frame("sample" = unique(bc_df$sample),
                                 "n_bc" = "", stringsAsFactors = F)
for (i in unique(bc_df$sample)) {
  n_highly_expressed$n_bc[n_highly_expressed$sample == i] <- length(bc_df$barcode[bc_df$rpm > 500 & bc_df$sample == i])
}

plot_ly(n_highly_expressed, x = ~sample, y = ~as.numeric(n_bc), type = 'bar',
             marker = list(color = '#D6D5C9',
                           line = list(color = 'rgb(8,48,107)', width = 1.5))) %>% 
  layout(title = "Highly expressed barcodes",
         xaxis = list(title = "Number of barcodes with > 500 rpm"),
         yaxis = list(title = "Condition"))


## 2: How many barcodes can I find back at which cutoff? + What is the percentage of barcode reads that match the design at which cutoff?
## Identify the unmapped fraction
bc_fraction <- data.frame("sample" = unique(bc_df$sample))
bc_fraction$bcs_found <- ""
rpm_cutoff <- data.frame("cutoff" = seq(0.0001,15,0.5))
bc_fraction <- merge(bc_fraction, rpm_cutoff)

for (i in unique(bc_fraction$cutoff)) {
  for (j in unique(bc_df$sample)) {
    bc_fraction$bcs_found[bc_fraction$cutoff == i & bc_fraction$sample == j] <- 
      nrow(bc_df_filt[bc_df_filt$rpm >= i & bc_df_filt$sample == j ,])/
      nrow(bc_annotation) *100
  }
}



## How many reads match to designed barcodes?
bc_reads <- data.frame("sample" = unique(bc_df$sample))
bc_reads$bc_reads <- ""
bc_reads <- merge(bc_reads, rpm_cutoff)

bc_df$match <- "Yes"
bc_df$match[is.na(bc_df$library)] <- "No"

for (i in unique(bc_reads$cutoff)) {
  for (j in unique(bc_df$sample)) {
      bc_reads$bc_reads[bc_reads$cutoff == i & bc_reads$sample == j] <- 
        sum(na.omit(bc_df_filt$rpm[bc_df_filt$rpm >= i & bc_df_filt$sample == j]))/
      sum(bc_df$rpm[bc_df$rpm >= i & bc_df$sample == j & bc_df$match == "No"]) *100
  }
}

bc_fraction <- merge(bc_fraction, bc_reads)
bc_fraction$bcs_found <- as.numeric(bc_fraction$bcs_found)
bc_fraction$bc_reads <- as.numeric(bc_fraction$bc_reads)

#c("#1B998B", "#2D3047", "#FF9B71", "#ECDD7B")
# Plot to evaluate data quality per cutoff
ggplot(bc_fraction) +
  geom_point(aes(x = cutoff, y = bcs_found), color = '#1B998B', size = 1) +
  geom_line(aes(x = cutoff, y = bcs_found), color = '#1B998B', size = 1) +
  geom_point(aes(x = cutoff, y = bc_reads), color = 'black', size = 1) +
  geom_line(aes(x = cutoff, y = bc_reads), color = 'black', size = 1) +
  theme_bw()+
  xlab("rpm cutoff")+
  ylab("total barcodes (green) and matched barcode reads (black) detected (%)")+
  facet_wrap(~sample)+
  theme(strip.background =element_rect(fill="#D6D5C9"))

## 3: What is the correlation of the cDNA bc counts with the pDNA bc counts? 
pDNA <- data.frame("pDNA" = bc_df$rpm[bc_df$sample == "pMT09_r1_r2"],
                   "barcode"= bc_df$barcode[bc_df$sample == "pMT09_r1_r2"], stringsAsFactors = F)
bc_df_2 <- merge(pDNA, bc_df, all = T)

ggplot(bc_df_2, aes(x = pDNA, y = rpm)) +
  geom_bin2d(bins = 50)+
  xlim(0,500) +
  ylim(0,2000)+
  theme_bw()+
  scale_color_viridis() +
  facet_wrap(~sample)

cor <- data.frame("sample" = unique(bc_df_2$sample), "cor" = "", stringsAsFactors = F)

for (i in unique(bc_df_2$sample)) {
  x <- bc_df_2$rpm[bc_df_2$sample == i]
  y <- bc_df_2[bc_df_2$sample == i,] %>% dplyr::select(barcode, pDNA) %>% unique()
  cor$cor[cor$sample == i] <- stats::cor(x, y$pDNA, use = "pairwise.complete.obs")
}

bc_df <- merge(bc_df, cor, by = "sample", all = T)

## 4: Correlation plots of the replicates
## Combine replicates of normalized data in 3 different columns
bc_df$condition <- gsub("_r[1-3]$", "", bc_df$condition)
bc_df$replicate <- gsub(".*_(r[1-3]$)", "\\1", bc_df$sample)
bc_df_rep <- bc_df[!is.na(bc_df$tf),] %>% 
  dplyr::select(replicate, rpm, condition, barcode, neg_ctrls) %>% 
  unique() %>%
  spread(replicate, rpm)

colors <- c("#2D3047", "#1B998B")

ggscatter(bc_df_rep, x = "r1", y = "r2",
   add = "reg.line",
   color = "neg_ctrls",
   size = 1,
   add.params = list(color = "blue", fill = "lightgray"), title = "rep1 vs rep2 ",
   conf.int = TRUE, ylab = "rep2", xlab = "rep1") + 
  stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
  geom_abline(linetype = "dashed") +
  xlim(0,4000) + ylim(0,4000) +
  scale_color_manual(values = colors)+
  facet_wrap(~condition)

ggscatter(bc_df_rep, x = "r1", y = "r3",
   add = "reg.line",
   color = "neg_ctrls",
   size = 1,
   add.params = list(color = "blue", fill = "lightgray"), title = "rep1 vs rep2 ",
   conf.int = TRUE, ylab = "rep2", xlab = "rep1") + 
  stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
  geom_abline(linetype = "dashed") +
  xlim(0,4000) + ylim(0,4000) +
  scale_color_manual(values = colors)+
  facet_wrap(~condition)

ggscatter(bc_df_rep, x = "r2", y = "r3",
   add = "reg.line",
   color = "neg_ctrls",
   size = 1,
   add.params = list(color = "blue", fill = "lightgray"), title = "rep1 vs rep2 ",
   conf.int = TRUE, ylab = "rep2", xlab = "rep1") + 
  stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
  geom_abline(linetype = "dashed") +
  xlim(0,4000) + ylim(0,4000) +
  scale_color_manual(values = colors)+
  facet_wrap(~condition)
```

---

### Sample filtering

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Remove samples with low dynamic range
for (i in unique(bc_df$sample)) {
  bc_df$n_bc[bc_df$sample == i] <- length(bc_df$barcode[bc_df$rpm > 500 & bc_df$sample == i])
}
bc_df <- bc_df[bc_df$n_bc >= 100 | bc_df$sample == "pMT09_r1_r2" | bc_df$sample == "pMT09_r3",] %>% dplyr::select(-n_bc)

# Remove samples with high pDNA contamination
bc_df <- bc_df[bc_df$cor <= 0.85 | bc_df$sample == "pMT09_r1_r2" | bc_df$sample == "pMT09_r3",] %>% dplyr::select(-cor)

# Remove all non-matching reads
bc_df <- bc_df[!is.na(bc_df$tf),]
```

---


### Normalization of barcode counts:  
Divide cDNA barcode counts through pDNA barcode counts to get activity

```{r normalization, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# Add pDNA data (only use barcodes above a certain pDNA representation to exclude noisy data)
pDNA <- bc_df[bc_df$sample == "pMT09_r1_r2" | bc_df$sample == "pMT09_r3",] %>%
  dplyr::select(barcode, sample, rpm) %>% 
  unique() %>%
  dcast(barcode ~ sample) %>%
  mutate(pDNA_counts_rpm = (pMT09_r1_r2 + pMT09_r3) / 2) %>%
  dplyr::select(barcode, pDNA_counts_rpm)

pDNA <- pDNA[pDNA$pDNA_counts_rpm > 7,]

bc_df <- merge(bc_df, pDNA)

# Compute activity by dividing cDNA bc counts through pDNA bc counts
bc_df$activity <- bc_df$rpm / bc_df$pDNA_counts_rpm
```


---

## Calculate mean activity - filter out outlier barcodes 

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE, message=FALSE}
# First identify and remove outlier barcodes - this removes the noise created by faulty barcode clustering etc. 
## Calculate mean and SD for each reporter
bc_df_cDNA <- bc_df[-grep("pMT09", bc_df$sample),]
bc_df_cDNA$reporter_id <- paste(bc_df_cDNA$tf, bc_df_cDNA$spacing, 
                                bc_df_cDNA$distance, bc_df_cDNA$promoter,
                                bc_df_cDNA$background, sep = "_")
bc_df_cDNA$mean_activity <- ave(bc_df_cDNA$activity, bc_df_cDNA$reporter_id, 
                                bc_df_cDNA$sample, FUN =
                                  function(x) mean(x))
bc_df_cDNA$sd_activity <- ave(bc_df_cDNA$activity, bc_df_cDNA$reporter_id, 
                              bc_df_cDNA$sample,  FUN =
                                  function(x) sd(x))

## Remove data points that are 2xSD away from the mean
bc_df_cDNA$upper_activity <- bc_df_cDNA$mean_activity + (2 * bc_df_cDNA$sd_activity)
bc_df_cDNA$lower_activity <- bc_df_cDNA$mean_activity - (2 * bc_df_cDNA$sd_activity)

bc_df_cDNA$low_outlier <- bc_df_cDNA$activity - bc_df_cDNA$lower_activity
bc_df_cDNA$high_outlier <- bc_df_cDNA$upper_activity - bc_df_cDNA$activity

## Choose arbitrary cutoff to get rid of most extreme outliers
bc_df_cDNA_filt <- bc_df_cDNA[bc_df_cDNA$low_outlier > -0.3 & bc_df_cDNA$high_outlier > -2,]

## Recalculate mean and sd
bc_df_cDNA_filt$mean_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, 
                                bc_df_cDNA_filt$sample, FUN =
                                  function(x) mean(x))
bc_df_cDNA_filt$sd_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, 
                              bc_df_cDNA_filt$sample,  FUN =
                                  function(x) sd(x))


## Remove reporters with high SD
bc_df_cDNA_filt$outlier <- "No"
bc_df_cDNA_filt$outlier[bc_df_cDNA_filt$mean_activity > 5 & bc_df_cDNA_filt$sd_activity > 1.5* bc_df_cDNA_filt$mean_activity] <- "Yes"
bc_df_cDNA_filt <- bc_df_cDNA_filt[bc_df_cDNA_filt$outlier == "No",]
```


```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE}
# # Remove barcodes with multiple inserts attached
# bc_exclude <- read.csv("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/pDNA_seq/bc_exclude.csv") %>% dplyr::select(-X) %>% setnames("x", "barcode")
# exclude <- bc_df_cDNA_filt[!bc_df_cDNA_filt$barcode %in% bc_exclude$barcode,]
# 
# # Reassign barcodes with wrong inserts attached
# bc_replace <- read.csv("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/SuRE_TF_1/pDNA_seq/bc_replace.csv") %>% dplyr::select(-X, -bc.match)
# change.df <- exclude[exclude$barcode %in% bc_replace$barcode,]
# exclude <- exclude[!exclude$barcode %in% bc_replace$barcode,]
# 
# change.df$bc.number <- 9
# change.df <- merge(change.df, bc_replace)
# e <- c("e11", "e93", "e97")
# change.df.e <- change.df[grep(paste(e, collapse = "|"), change.df$insert.match),]
# change.df <- change.df[-grep(paste(e, collapse = "|"), change.df$insert.match),]
# change.df <- change.df %>%
#   mutate(TF = gsub("(^.*?)_.*", "\\1", insert.match),
#          promoter = gsub(".*(minP|mCMV|hBGm|Random)_.*", "\\1", insert.match),
#          distance = gsub(".*_d-([0-9]{1,2}bp)_.*", "\\1", insert.match),
#          spacing = gsub(".*_s-([0-9]{1,2}bp)_.*", "\\1", insert.match),
#          background = gsub(".*([0-9]{1}$)", "\\1", insert.match),
#          reporter_id = paste(TF, spacing, distance, promoter, background, sep = "_"))
# change.df <- change.df %>% dplyr::select(-insert.match)
# 
# change.df.e <- change.df.e %>%
#   mutate(TF = gsub("(.*?)_(minP|mCMV|hBGm|Random)$", "\\1", insert.match),
#          promoter = gsub(".*(minP|mCMV|hBGm|Random)", "\\1", insert.match),
#          spacing = "",
#          distance = "",
#          background = "",
#          reporter_id = paste(TF, promoter, sep ="_"))
# change.df.e <- change.df.e %>% dplyr::select(-insert.match)
# 
# 
# exclude$reporter_id <- gsub("___", "_", exclude$reporter_id)
# exclude$reporter_id <- gsub("_0", "", exclude$reporter_id)
# 
# 
# 
# exclude <- rbind(exclude, change.df, change.df.e)
# 
# bc_df_cDNA_filt <- exclude
```


### Scaling data 

```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE}
# # Scale data to 1 for negative controls
bc_df_cDNA_filt_neg <- bc_df_cDNA_filt %>%
  dplyr::select(sample, activity, reporter_id, neg_ctrls, rand_promoter) %>%
  filter(neg_ctrls == "Yes", rand_promoter == "Yes") %>%
  unique() %>%
  mutate(activity = ave(activity, FUN = function(x) mean(x))) %>%
  dplyr::select(-reporter_id) %>%
  unique() %>%
  dplyr::select("background_activity" = activity) %>%
  unique()

bc_df_cDNA_filt <- merge(bc_df_cDNA_filt, bc_df_cDNA_filt_neg)
bc_df_cDNA_filt <- bc_df_cDNA_filt %>%
  mutate(activity = activity / background_activity)

bc_df_cDNA_filt$mean_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id,
                                bc_df_cDNA_filt$sample, FUN =
                                  function(x) mean(x))
#bc_df_cDNA_filt <- bc_df_cDNA_filt[!is.na(bc_df_cDNA_filt$condition),]
```


---


## Calculate correlations between technical replicates

```{r correlations_2, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE}
## Combine replicates in 8 different columns
bc_df_rep <- bc_df_cDNA_filt[bc_df_cDNA_filt$hPGK == "No" &
                          bc_df_cDNA_filt$rand_promoter == "No",] %>% 
  filter(commercial_reporter == "No") %>%
  dplyr::select(barcode_number, activity, sample, reporter_id, neg_ctrls) %>%
  spread(barcode_number, activity)
names(bc_df_rep) <- gsub("([1-5]{1})", "TR\\1", names(bc_df_rep))

ggscatter(bc_df_rep, x = "TR1", y = "TR2",
                 add = "reg.line",
                 color = "neg_ctrls",
                 size = 0.5,
                 add.params = list(color = "blue", fill = "lightgray"), 
                 title = "rep1 vs rep2",
                 conf.int = TRUE, ylab = "rep2", xlab = "rep1") + 
    stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
    geom_abline(linetype = "dashed") +
    xlim(0,50) + ylim(0,50) +
    scale_color_manual(values = colors) +
    facet_wrap(~sample)

ggscatter(bc_df_rep, x = "TR2", y = "TR4",
                 add = "reg.line",
                 color = "neg_ctrls",
                 size = 0.5,
                 add.params = list(color = "blue", fill = "lightgray"), 
                 title = "rep1 vs rep2",
                 conf.int = TRUE, ylab = "rep2", xlab = "rep4") + 
    stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
    geom_abline(linetype = "dashed") +
    xlim(0,50) + ylim(0,50) +
    scale_color_manual(values = colors) +
    facet_wrap(~sample)




# Correlation matrix plot
n <- sample(1:nrow(bc_df_rep), 5000)
boundaries <- seq(from = 0.8, by = 0.05, length.out = 4)
plt <- ggpairs(bc_df_rep %>% dplyr::select('TR1', 'TR2', 'TR3', 'TR4', 'TR5'),
               upper = list(continuous = corColor),
               lower = list(continuous = function(data, mapping, ...) {
                   ggally_points(data = data[n, ], mapping = mapping, alpha = 0.1, size = 0.5) +
                   geom_abline(slope = 1, lty = "dashed", col = "red") +
                   theme_bw()}),
               diag = list(continuous = function(data, mapping, ...) {
                   ggally_densityDiag(data = data, mapping = mapping, alpha = 0.3, fill = "red") +
                   theme_bw()})) +
  ggtitle("Correlation Between Technial Replicates") +
  theme(text = element_text(size = 20)) +
  xlab("Reporter activity") +
  ylab("Reporter activity")

print(plt)
```


---

## Correlation between replicates

```{r correlations_3, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE}
# Correlation plots of the replicates
## Combine replicates of normalized data in 3 different columns
bc_df_rep <- bc_df_cDNA_filt %>% 
  dplyr::select(replicate, mean_activity, reporter_id, condition, neg_ctrls) %>% 
  unique() %>%
  spread(replicate, mean_activity)

colors <- c("#2D3047", "#1B998B")

ggscatter(bc_df_rep, x = "r1", y = "r2",
                 add = "reg.line",
                 color = "neg_ctrls",
                 size = 0.5,
                 add.params = list(color = "blue", fill = "lightgray"), 
                 title = "rep1 vs rep2",
                 conf.int = TRUE, ylab = "r2", xlab = "r1") + 
    stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
    geom_abline(linetype = "dashed") +
    xlim(0,50) + ylim(0,50) +
    scale_color_manual(values = colors) +
    facet_wrap(~condition)


ggscatter(bc_df_rep, x = "r1", y = "r3",
                 add = "reg.line",
                 color = "neg_ctrls",
                 size = 0.5,
                 add.params = list(color = "blue", fill = "lightgray"), 
                 title = "rep1 vs rep2",
                 conf.int = TRUE, ylab = "r3", xlab = "r1") + 
    stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
    geom_abline(linetype = "dashed") +
    xlim(0,50) + ylim(0,50) +
    scale_color_manual(values = colors) +
    facet_wrap(~condition)


ggscatter(bc_df_rep, x = "r3", y = "r2",
                 add = "reg.line",
                 color = "neg_ctrls",
                 size = 0.5,
                 add.params = list(color = "blue", fill = "lightgray"), 
                 title = "rep1 vs rep2",
                 conf.int = TRUE, ylab = "r2", xlab = "r3") + 
    stat_cor(method = "pearson", label.x = 4, label.y = 0) + 
    geom_abline(linetype = "dashed") +
    xlim(0,50) + ylim(0,50) +
    scale_color_manual(values = colors) +
    facet_wrap(~condition)

```


```{r out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE}
# Mean of the three replicates
bc_df_cDNA_filt$reporter_activity <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$condition, FUN = function(x) mean(x))
bc_df_cDNA_filt$reporter_activity_sd <- ave(bc_df_cDNA_filt$activity, bc_df_cDNA_filt$reporter_id, bc_df_cDNA_filt$condition, FUN = function(x) sd(x))
```


```{r data export, out.width= "80%", fig.align= "center", echo=FALSE, warning= FALSE}
# Polish export dataframe
bc_df_cDNA_filt <- bc_df_cDNA_filt %>% 
  dplyr::select(-upper_activity, -lower_activity, 
         -high_outlier, -low_outlier,) %>% 
  setnames(old = c("mean_activity", "sd_activity"), 
           new = c("replicate_activity", "replicate_activity_sd")) %>% 
  mutate(log_activity = log2(activity),
         log_reporter_activity = log2(reporter_activity))


# Export bc_df for cDNA analysis
filename <- SetFileName("_reporter_activity_filt", "mt")
setwd("/DATA/usr/m.trauernicht/projects/SuRE-TF/data/gcf6578_gen2_stimulation-2/results/")
write.csv(bc_df_cDNA_filt, file = paste(filename,".csv", sep = ""), row.names = F)
```

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

