<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2024-05-15" />

<title>Optimized reporters for multiplexed detection of transcription factor activity</title>

<script src="site_libs/header-attrs-2.13/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Optimized reporters for multiplexed detection of transcription factor activity</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Introduction</a>
</li>
<li>
  <a href="pwm_comparison.html">TFBS analysis</a>
</li>
<li>
  <a href="barcode_preprocessing_all.html">Data pre-processing</a>
</li>
<li>
  <a href="TF_reporter_variant_analysis.html">Detailed TF Reporter Activity Analysis</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Optimized reporters for multiplexed
detection of transcription factor activity</h1>
<h4 class="date">2024-05-15</h4>

</div>


<div id="identifying-sensitive-and-specific-reporters-for-86-tfs"
class="section level1">
<h1>Identifying sensitive and specific reporters for 86 TFs</h1>
<div class="float">
<img
src="https://github.com/mtrauernicht/SuRE_TF/assets/57003758/da5f67fe-06a4-4adb-99e1-071462f4fedd"
alt="1_reporter_library_design" />
<div class="figcaption">1_reporter_library_design</div>
</div>
<p><strong>Introduction:</strong><br />
Transcriptional reporters for 86 TFs were systematically designed and
probed in 9 cell types and dozens of TF perturbation conditions with the
aim to identify the most sensitive reporters per TF.</p>
<p><strong>TF reporter library design:</strong><br />
The designed first library contains:</p>
<ul>
<li><p>General reporter design: with 4 identical TF binding sites
followed by a minimal promoter and a barcode in the transcription
unit</p></li>
<li><p>In total reporters for 86 different TFs</p></li>
<li><p>10 or 5 bp spacing between the TF binding sites</p></li>
<li><p>10 or 21 bp distance from the TF binding sites to the minimal
promoter</p></li>
<li><p>3 different spacer sequences</p></li>
<li><p>3 different minimal promoters</p></li>
<li><p>5-8 barcodes per TF reporter</p></li>
<li><p>In total ~36,000 uniquely barcoded reporters per TF</p></li>
</ul>
<p>All TF reporters were designed using FIMO. This way, the spacings
were designed to be inactive, while the TF binding sites were ensured to
be intact.</p>
<p><strong>Experimental setup:</strong><br />
The library was transfected into:</p>
<ul>
<li><p>9 different cell types</p></li>
<li><p>almost 100 TF perturbation conditions:</p></li>
<li><p>TF knockdown</p></li>
<li><p>TF overexpression</p></li>
<li><p>Signaling pathway perturbation</p></li>
<li><p>TF degradation</p></li>
</ul>
<p>The cells were grown in these condition for 24h after transfection
before RNA isolation. Barcoded transcripts were then reverse transcribed
and amplified before sequencing.</p>
<p><strong>Sequencing data analysis:</strong></p>
<ul>
<li><p>Raw sequencing data were processed by counting the barcodes and
clustering the barcodes using starcode.</p></li>
<li><p>Barcode counts in the cDNA were normalized by the counts in the
plasmid library.</p></li>
<li><p>Activities were normalized to promoter-only reporter
activities</p></li>
</ul>
<hr />
<p><strong>Repository guide:</strong></p>
<ul>
<li><p><code>barcode_preprocessing.Rmd</code> contains all quality
checks and reporter activity computation steps</p></li>
<li><p><code>TF_reporter_variant_analysis.Rmd</code> contains all
further analyses and generation of figures</p></li>
<li><p><code>py_analysis</code> folder contains jupyter notebook
versions of TF_reporter_variant_analysis.Rmd that were used to generate
some figures</p></li>
<li><p><code>library_design</code> folder contains all code used to
generate the TF reporter sequences and an analysis of the selected
motifs</p></li>
<li><p><code>tf_transcript_abundance_computation.Rmd</code> contains
code to compute the transcript abundances (from RNA-seq) of all TFs in
all cell types</p></li>
<li><p><code>raw_data_analysis</code> folder contains the scripts that
were used to generate barcode counts from the raw sequencing
files</p></li>
<li><p>Mutated TFBSs were generated by our collaborators from the
Bussemaker lab: <a
href="https://github.com/BussemakerLab/TranscriptionFactorActivityReporters"
class="uri">https://github.com/BussemakerLab/TranscriptionFactorActivityReporters</a></p></li>
</ul>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
